# Comparing `tmp/draupnir-0.0.28.tar.gz` & `tmp/draupnir-0.0.29.tar.gz`

## filetype from file(1)

```diff
@@ -1 +1 @@
-gzip compressed data, was "draupnir-0.0.28.tar", last modified: Fri Jun  2 13:43:22 2023, max compression
+gzip compressed data, was "draupnir-0.0.29.tar", last modified: Tue Apr  2 22:02:55 2024, max compression
```

## Comparing `draupnir-0.0.28.tar` & `draupnir-0.0.29.tar`

### file list

```diff
@@ -1,44 +1,44 @@
-drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2023-06-02 13:43:22.081260 draupnir-0.0.28/
--rw-r--r--   0 lys       (1000) lys       (1000)     1071 2022-02-03 12:22:28.000000 draupnir-0.0.28/LICENSE
--rw-r--r--   0 lys       (1000) lys       (1000)       15 2022-02-04 14:46:10.000000 draupnir-0.0.28/MANIFEST.in
--rw-rw-r--   0 lys       (1000) lys       (1000)    10675 2023-06-02 13:43:22.081260 draupnir-0.0.28/PKG-INFO
--rw-r--r--   0 lys       (1000) lys       (1000)     9922 2023-06-02 13:40:23.000000 draupnir-0.0.28/README.md
--rw-r--r--   0 lys       (1000) lys       (1000)      103 2022-02-03 12:19:06.000000 draupnir-0.0.28/pyproject.toml
--rw-r--r--   0 lys       (1000) lys       (1000)      748 2023-06-02 13:43:22.081260 draupnir-0.0.28/setup.cfg
--rw-r--r--   0 lys       (1000) lys       (1000)     3138 2023-06-02 13:40:22.000000 draupnir-0.0.28/setup.py
-drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2023-06-02 13:43:22.021259 draupnir-0.0.28/src/
-drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2023-06-02 13:43:22.029259 draupnir-0.0.28/src/draupnir/
--rw-r--r--   0 lys       (1000) lys       (1000)    15019 2022-03-01 17:48:39.000000 draupnir-0.0.28/src/draupnir/3D.py
--rw-r--r--   0 lys       (1000) lys       (1000)      496 2022-03-30 13:01:42.000000 draupnir-0.0.28/src/draupnir/__init__.py
-drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2023-06-02 13:43:22.029259 draupnir-0.0.28/src/draupnir/data/
--rw-r--r--   0 lys       (1000) lys       (1000)     1578 2021-07-28 14:11:37.000000 draupnir-0.0.28/src/draupnir/data/AA_properties.txt
--rw-r--r--   0 lys       (1000) lys       (1000) 32494553 2021-07-28 14:11:52.000000 draupnir-0.0.28/src/draupnir/data/PfamPdbMap.txt
--rw-r--r--   0 lys       (1000) lys       (1000)    37246 2022-10-08 16:43:04.000000 draupnir-0.0.28/src/draupnir/datasets.py
--rw-r--r--   0 lys       (1000) lys       (1000)     9362 2022-03-22 15:01:49.000000 draupnir-0.0.28/src/draupnir/draw_tree.py
--rw-r--r--   0 lys       (1000) lys       (1000)     8722 2022-06-27 13:16:05.000000 draupnir-0.0.28/src/draupnir/guides.py
-drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2023-06-02 13:43:22.081260 draupnir-0.0.28/src/draupnir/infer_angles/
--rw-r--r--   0 lys       (1000) lys       (1000)      330 2022-02-15 14:19:01.000000 draupnir-0.0.28/src/draupnir/infer_angles/__init__.py
--rw-r--r--   0 lys       (1000) lys       (1000)     5319 2021-08-19 13:41:22.000000 draupnir-0.0.28/src/draupnir/infer_angles/calculate_coords.py
--rw-r--r--   0 lys       (1000) lys       (1000)     1362 2021-08-13 17:18:45.000000 draupnir-0.0.28/src/draupnir/infer_angles/nerf.py
--rw-r--r--   0 lys       (1000) lys       (1000)     9437 2021-08-13 17:41:01.000000 draupnir-0.0.28/src/draupnir/infer_angles/pnerf.py
--rw-r--r--   0 lys       (1000) lys       (1000)    19190 2022-02-13 16:56:12.000000 draupnir-0.0.28/src/draupnir/infer_angles/superposition.py
--rw-r--r--   0 lys       (1000) lys       (1000)    59299 2022-10-07 18:02:03.000000 draupnir-0.0.28/src/draupnir/load_utils.py
--rw-r--r--   0 lys       (1000) lys       (1000)   183274 2022-10-18 18:37:56.000000 draupnir-0.0.28/src/draupnir/main.py
--rw-r--r--   0 lys       (1000) lys       (1000)    67482 2022-10-07 16:10:38.000000 draupnir-0.0.28/src/draupnir/models.py
--rw-r--r--   0 lys       (1000) lys       (1000)    22323 2022-10-14 12:25:08.000000 draupnir-0.0.28/src/draupnir/models_utils.py
--rw-r--r--   0 lys       (1000) lys       (1000)    17819 2023-05-10 11:36:14.000000 draupnir-0.0.28/src/draupnir/mutual_information.py
--rw-r--r--   0 lys       (1000) lys       (1000)    81700 2022-11-08 11:12:02.000000 draupnir-0.0.28/src/draupnir/plots.py
--rw-r--r--   0 lys       (1000) lys       (1000)     4493 2022-03-22 17:45:42.000000 draupnir-0.0.28/src/draupnir/train.py
--rw-r--r--   0 lys       (1000) lys       (1000)   108500 2022-12-19 20:46:58.000000 draupnir-0.0.28/src/draupnir/utils.py
-drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2023-06-02 13:43:22.029259 draupnir-0.0.28/src/draupnir.egg-info/
--rw-r--r--   0 lys       (1000) lys       (1000)    10675 2023-06-02 13:43:21.000000 draupnir-0.0.28/src/draupnir.egg-info/PKG-INFO
--rw-r--r--   0 lys       (1000) lys       (1000)      951 2023-06-02 13:43:22.000000 draupnir-0.0.28/src/draupnir.egg-info/SOURCES.txt
--rw-r--r--   0 lys       (1000) lys       (1000)        1 2023-06-02 13:43:21.000000 draupnir-0.0.28/src/draupnir.egg-info/dependency_links.txt
--rw-r--r--   0 lys       (1000) lys       (1000)        1 2022-03-02 13:52:18.000000 draupnir-0.0.28/src/draupnir.egg-info/not-zip-safe
--rw-r--r--   0 lys       (1000) lys       (1000)      229 2023-06-02 13:43:21.000000 draupnir-0.0.28/src/draupnir.egg-info/requires.txt
--rw-r--r--   0 lys       (1000) lys       (1000)      120 2023-06-02 13:43:21.000000 draupnir-0.0.28/src/draupnir.egg-info/top_level.txt
-drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2023-06-02 13:43:22.081260 draupnir-0.0.28/tests/
--rw-r--r--   0 lys       (1000) lys       (1000)     6414 2022-03-30 15:40:03.000000 draupnir-0.0.28/tests/test_accuracy.py
--rw-r--r--   0 lys       (1000) lys       (1000)     2173 2022-03-30 16:33:30.000000 draupnir-0.0.28/tests/test_examples.py
--rw-r--r--   0 lys       (1000) lys       (1000)     3793 2022-03-30 13:55:34.000000 draupnir-0.0.28/tests/test_load.py
--rw-r--r--   0 lys       (1000) lys       (1000)     4483 2022-03-29 19:20:16.000000 draupnir-0.0.28/tests/test_norworking.py
+drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2024-04-02 22:02:55.505727 draupnir-0.0.29/
+-rw-r--r--   0 lys       (1000) lys       (1000)     1071 2022-02-03 12:22:28.000000 draupnir-0.0.29/LICENSE
+-rw-r--r--   0 lys       (1000) lys       (1000)       15 2022-02-04 14:46:10.000000 draupnir-0.0.29/MANIFEST.in
+-rw-r--r--   0 lys       (1000) lys       (1000)    11167 2024-04-02 22:02:55.505727 draupnir-0.0.29/PKG-INFO
+-rw-r--r--   0 lys       (1000) lys       (1000)     9945 2024-02-23 18:22:33.000000 draupnir-0.0.29/README.md
+-rw-r--r--   0 lys       (1000) lys       (1000)      103 2022-02-03 12:19:06.000000 draupnir-0.0.29/pyproject.toml
+-rw-r--r--   0 lys       (1000) lys       (1000)      748 2024-04-02 22:02:55.505727 draupnir-0.0.29/setup.cfg
+-rw-r--r--   0 lys       (1000) lys       (1000)     3138 2024-04-02 21:58:37.000000 draupnir-0.0.29/setup.py
+drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2024-04-02 22:02:55.481727 draupnir-0.0.29/src/
+drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2024-04-02 22:02:55.485727 draupnir-0.0.29/src/draupnir/
+-rw-r--r--   0 lys       (1000) lys       (1000)    15019 2022-03-01 17:48:39.000000 draupnir-0.0.29/src/draupnir/3D.py
+-rw-r--r--   0 lys       (1000) lys       (1000)      496 2022-03-30 13:01:42.000000 draupnir-0.0.29/src/draupnir/__init__.py
+drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2024-04-02 22:02:55.485727 draupnir-0.0.29/src/draupnir/data/
+-rw-r--r--   0 lys       (1000) lys       (1000)     1578 2021-07-28 14:11:37.000000 draupnir-0.0.29/src/draupnir/data/AA_properties.txt
+-rw-r--r--   0 lys       (1000) lys       (1000) 32494553 2021-07-28 14:11:52.000000 draupnir-0.0.29/src/draupnir/data/PfamPdbMap.txt
+-rw-r--r--   0 lys       (1000) lys       (1000)    37249 2024-04-01 21:31:59.000000 draupnir-0.0.29/src/draupnir/datasets.py
+-rw-r--r--   0 lys       (1000) lys       (1000)     9362 2022-03-22 15:01:49.000000 draupnir-0.0.29/src/draupnir/draw_tree.py
+-rw-r--r--   0 lys       (1000) lys       (1000)    11193 2024-04-02 21:27:05.000000 draupnir-0.0.29/src/draupnir/guides.py
+drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2024-04-02 22:02:55.505727 draupnir-0.0.29/src/draupnir/infer_angles/
+-rw-r--r--   0 lys       (1000) lys       (1000)      330 2022-02-15 14:19:01.000000 draupnir-0.0.29/src/draupnir/infer_angles/__init__.py
+-rw-r--r--   0 lys       (1000) lys       (1000)     5319 2021-08-19 13:41:22.000000 draupnir-0.0.29/src/draupnir/infer_angles/calculate_coords.py
+-rw-r--r--   0 lys       (1000) lys       (1000)     1362 2021-08-13 17:18:45.000000 draupnir-0.0.29/src/draupnir/infer_angles/nerf.py
+-rw-r--r--   0 lys       (1000) lys       (1000)     9437 2021-08-13 17:41:01.000000 draupnir-0.0.29/src/draupnir/infer_angles/pnerf.py
+-rw-r--r--   0 lys       (1000) lys       (1000)    19190 2022-02-13 16:56:12.000000 draupnir-0.0.29/src/draupnir/infer_angles/superposition.py
+-rw-r--r--   0 lys       (1000) lys       (1000)    60195 2024-04-02 17:24:44.000000 draupnir-0.0.29/src/draupnir/load_utils.py
+-rw-r--r--   0 lys       (1000) lys       (1000)   182199 2024-04-02 21:09:37.000000 draupnir-0.0.29/src/draupnir/main.py
+-rw-r--r--   0 lys       (1000) lys       (1000)    68853 2024-04-02 21:55:50.000000 draupnir-0.0.29/src/draupnir/models.py
+-rw-r--r--   0 lys       (1000) lys       (1000)    23103 2024-04-02 21:16:58.000000 draupnir-0.0.29/src/draupnir/models_utils.py
+-rw-r--r--   0 lys       (1000) lys       (1000)    17819 2023-05-10 11:36:14.000000 draupnir-0.0.29/src/draupnir/mutual_information.py
+-rw-r--r--   0 lys       (1000) lys       (1000)    81724 2024-02-23 21:46:00.000000 draupnir-0.0.29/src/draupnir/plots.py
+-rw-r--r--   0 lys       (1000) lys       (1000)     6840 2024-04-02 21:24:45.000000 draupnir-0.0.29/src/draupnir/train.py
+-rw-r--r--   0 lys       (1000) lys       (1000)   109487 2024-04-02 21:36:45.000000 draupnir-0.0.29/src/draupnir/utils.py
+drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2024-04-02 22:02:55.505727 draupnir-0.0.29/src/draupnir.egg-info/
+-rw-r--r--   0 lys       (1000) lys       (1000)    11167 2024-04-02 22:02:55.000000 draupnir-0.0.29/src/draupnir.egg-info/PKG-INFO
+-rw-r--r--   0 lys       (1000) lys       (1000)      951 2024-04-02 22:02:55.000000 draupnir-0.0.29/src/draupnir.egg-info/SOURCES.txt
+-rw-r--r--   0 lys       (1000) lys       (1000)        1 2024-04-02 22:02:55.000000 draupnir-0.0.29/src/draupnir.egg-info/dependency_links.txt
+-rw-r--r--   0 lys       (1000) lys       (1000)        1 2022-03-02 13:52:18.000000 draupnir-0.0.29/src/draupnir.egg-info/not-zip-safe
+-rw-r--r--   0 lys       (1000) lys       (1000)      229 2024-04-02 22:02:55.000000 draupnir-0.0.29/src/draupnir.egg-info/requires.txt
+-rw-r--r--   0 lys       (1000) lys       (1000)      120 2024-04-02 22:02:55.000000 draupnir-0.0.29/src/draupnir.egg-info/top_level.txt
+drwxrwxr-x   0 lys       (1000) lys       (1000)        0 2024-04-02 22:02:55.505727 draupnir-0.0.29/tests/
+-rw-r--r--   0 lys       (1000) lys       (1000)     6414 2022-03-30 15:40:03.000000 draupnir-0.0.29/tests/test_accuracy.py
+-rw-r--r--   0 lys       (1000) lys       (1000)     2173 2022-03-30 16:33:30.000000 draupnir-0.0.29/tests/test_examples.py
+-rw-r--r--   0 lys       (1000) lys       (1000)     3793 2022-03-30 13:55:34.000000 draupnir-0.0.29/tests/test_load.py
+-rw-r--r--   0 lys       (1000) lys       (1000)     4483 2022-03-29 19:20:16.000000 draupnir-0.0.29/tests/test_norworking.py
```

### Comparing `draupnir-0.0.28/LICENSE` & `draupnir-0.0.29/LICENSE`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/PKG-INFO` & `draupnir-0.0.29/README.md`

 * *Files 4% similar despite different names*

```diff
@@ -1,24 +1,7 @@
-Metadata-Version: 2.1
-Name: draupnir
-Version: 0.0.28
-Summary: Ancestral sequence reconstruction using a tree structured Ornstein Uhlenbeck variational autoencoder
-Home-page: https://github.com/LysSanzMoreta/DRAUPNIR_ASR
-Author: Lys Sanz Moreta
-Author-email: lys.sanz.moreta@outlook.com
-Project-URL: Changelog, https://github.com/LysSanzMoreta/DRAUPNIR_ASR/blob/master/CHANGELOG.rst
-Project-URL: Issue Tracker, https://github.com/LysSanzMoreta/DRAUPNIR_ASR/issues
-Classifier: Development Status :: 4 - Beta
-Classifier: License :: OSI Approved :: MIT License
-Classifier: Programming Language :: Python :: 3.8
-Classifier: Programming Language :: Python :: 3 :: Only
-Requires-Python: >=3.7, <4
-Description-Content-Type: text/markdown
-License-File: LICENSE
-
 
 DRAUPNIR: "Beta library version for performing ASR using a tree-structured Variational Autoencoder"
 
 <p align="center">
 <img src="https://github.com/LysSanzMoreta/DRAUPNIR_ASR/blob/main/draupnir/src/draupnir/images/draupnir_logo.png" height="auto" width="790" style="border-radius:50%">
 </p>
 
@@ -150,17 +133,17 @@
 
 **Where are my ancestral sequences?**
 
 - In each of the folders there should be a fasta file <dataset-name>_sampled_nodes_seq.fasta
 
 - Each of the sequences in the file should be identified as <node-name-input-tree>//<tree-level-order>\_sample\_<sample-number>
 
-    -Node-name-input-tree: Original name of the node in the given input tree
+    -<node-name-input-tree>: Original name of the node in the given/user's input tree.
 
-    -Tree-level-order: Position of the node in tree-level order in the tree
+    -<tree-level-order>: Position of the node in the tree expressed in tree-level order.
     
     >Node_A1//1.0_sample_0
 
     
 **If this library is useful for your research please cite:**
 
 ```
```

### Comparing `draupnir-0.0.28/README.md` & `draupnir-0.0.29/PKG-INFO`

 * *Files 13% similar despite different names*

```diff
@@ -1,7 +1,40 @@
+Metadata-Version: 2.1
+Name: draupnir
+Version: 0.0.29
+Summary: Ancestral sequence reconstruction using a tree structured Ornstein Uhlenbeck variational autoencoder
+Home-page: https://github.com/LysSanzMoreta/DRAUPNIR_ASR
+Author: Lys Sanz Moreta
+Author-email: lys.sanz.moreta@outlook.com
+Project-URL: Changelog, https://github.com/LysSanzMoreta/DRAUPNIR_ASR/blob/master/CHANGELOG.rst
+Project-URL: Issue Tracker, https://github.com/LysSanzMoreta/DRAUPNIR_ASR/issues
+Classifier: Development Status :: 4 - Beta
+Classifier: License :: OSI Approved :: MIT License
+Classifier: Programming Language :: Python :: 3.8
+Classifier: Programming Language :: Python :: 3 :: Only
+Requires-Python: >=3.7, <4
+Description-Content-Type: text/markdown
+License-File: LICENSE
+Requires-Dist: pyro-ppl>1.6.0
+Requires-Dist: biopython>1.78
+Requires-Dist: pandas>1.0.1
+Requires-Dist: matplotlib>3.3.4
+Requires-Dist: ete3>3.1.1
+Requires-Dist: dgl>0.6.1
+Requires-Dist: dill>0.3.3
+Requires-Dist: seaborn>=0.11.2
+Requires-Dist: pytorch-ignite>0.4.4
+Requires-Dist: scipy>1.5.4
+Requires-Dist: scikit-learn>0.24.1
+Requires-Dist: umap-learn>0.5.2
+Requires-Dist: gdown>=4.3.1
+Requires-Dist: ProDy>2.0.0
+Requires-Dist: PyQt5>=5.15.7
+Requires-Dist: lxml>=4.9.1
+
 
 DRAUPNIR: "Beta library version for performing ASR using a tree-structured Variational Autoencoder"
 
 <p align="center">
 <img src="https://github.com/LysSanzMoreta/DRAUPNIR_ASR/blob/main/draupnir/src/draupnir/images/draupnir_logo.png" height="auto" width="790" style="border-radius:50%">
 </p>
 
@@ -133,17 +166,17 @@
 
 **Where are my ancestral sequences?**
 
 - In each of the folders there should be a fasta file <dataset-name>_sampled_nodes_seq.fasta
 
 - Each of the sequences in the file should be identified as <node-name-input-tree>//<tree-level-order>\_sample\_<sample-number>
 
-    -Node-name-input-tree: Original name of the node in the given input tree
+    -<node-name-input-tree>: Original name of the node in the given/user's input tree.
 
-    -Tree-level-order: Position of the node in tree-level order in the tree
+    -<tree-level-order>: Position of the node in the tree expressed in tree-level order.
     
     >Node_A1//1.0_sample_0
 
     
 **If this library is useful for your research please cite:**
 
 ```
```

### Comparing `draupnir-0.0.28/setup.cfg` & `draupnir-0.0.29/setup.cfg`

 * *Files 1% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 [metadata]
 name = draupnir-artistworking43
-version = 0.0.28
+version = 0.0.29
 author = Lys Sanz Moreta
 author_email = lys.sanz.moreta@outlook.com
 description = Draupnir: Ancestral protein sequence reconstruction using a tree-structured Ornstein-Uhlenbeck variational autoencoder
 long_description = file: README.md
 long_description_content_type = text/markdown
 url = https://github.com/LysSanzMoreta/DRAUPNIR_ASR
 project_urls =
```

### Comparing `draupnir-0.0.28/setup.py` & `draupnir-0.0.29/setup.py`

 * *Files 1% similar despite different names*

```diff
@@ -16,15 +16,15 @@
 
 # Get the long description from the README file
 long_description = (here / 'README.md').read_text(encoding='utf-8')
 
 # Arguments marked as "Required" below must be included for upload to PyPI.
 # Fields marked as "Optional" may be commented out.
 setup(name='draupnir',
-      version='0.0.28',
+      version='0.0.29',
       # list folders, not files
       packages=find_packages('src'),
       package_dir={'': 'src'},
       py_modules=[splitext(basename(path))[0] for path in glob('src/draupnir/*.py')],
       #scripts=['bin/script1.py'],
       package_data={'draupnir': ['data/*']},
       description= 'Ancestral sequence reconstruction using a tree structured Ornstein Uhlenbeck variational autoencoder',
```

### Comparing `draupnir-0.0.28/src/draupnir/3D.py` & `draupnir-0.0.29/src/draupnir/3D.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/src/draupnir/data/AA_properties.txt` & `draupnir-0.0.29/src/draupnir/data/AA_properties.txt`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/src/draupnir/data/PfamPdbMap.txt` & `draupnir-0.0.29/src/draupnir/data/PfamPdbMap.txt`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/src/draupnir/datasets.py` & `draupnir-0.0.29/src/draupnir/datasets.py`

 * *Files 1% similar despite different names*

```diff
@@ -78,15 +78,15 @@
     SettingsConfig = namedtuple("SettingsConfig", ["one_hot_encoding", "model_design", "aligned_seq","data_folder","full_name","tree_file"],module="settings_config")
     if args.one_hot_encoded:
         warnings.warn("Draupnir was constructed to be used with integers for Categorical likelihood,not OneHotCategorical. And blosum-encoding for the guide. You can build"
                       "the one-hot-encoded dataset for other purposes")
 
     #script_dir = os.path.dirname(os.path.abspath(__file__))
     if not use_custom:
-        warnings.warn("You have selected a pre-defined dataset, if not present, it will be downloaded. Otherwise set use_custom to True")
+        warnings.warn("You have selected a pre-defined dataset, if not present, it will be downloaded. Otherwise set args.use_custom to True")
         root_sequence_name = available_datasets()[0][name]
         full_name = available_datasets()[1][name]
         storage_folder = os.path.abspath(os.path.join(os.path.dirname(__file__), "data")) #changed from "datasets/default"
         dir_name = '{}/{}'.format(storage_folder,name)
         dict_urls = {
                 "aminopeptidase":"https://drive.google.com/drive/folders/1fLsOJbD1hczX15NW0clCgL6Yf4mnx_yl?usp=sharing",
                 "benchmark_randall_original_naming":"https://drive.google.com/drive/folders/1oE5-22lqcobZMIguatOU_Ki3N2Fl9b4e?usp=sharing",
@@ -473,16 +473,14 @@
         dataset_test[i, 0, 0] =  len(str(internal_fasta_dict[key][0]).replace("-","")) # Fill in the sequence lenght
         dataset_test[i, 2:,0] = internal_fasta_dict[key][1]
 
     leaves_names_test = internal_fasta_dict.keys()
     dataset_test = np.array(dataset_test, dtype="float64")
     dataset_test = torch.from_numpy(dataset_test)
     return dataset_test,leaves_names_test,max_lenght_internal_aligned
-
-
 def load_coral_fluorescent_proteins_ancestral_sequences(name,ancestral_file,tree_levelorder_names,aa_probs):
     """Loads the 5 ancestral root nodes from the coral fluorescent proteins as the data set test.
     :param str name
     :param str ancestral_file
     :param list tree_levelorder_names
     :param int aa_probs
     """
```

### Comparing `draupnir-0.0.28/src/draupnir/draw_tree.py` & `draupnir-0.0.29/src/draupnir/draw_tree.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/src/draupnir/guides.py` & `draupnir-0.0.29/src/draupnir/guides.py`

 * *Files 25% similar despite different names*

```diff
@@ -1,16 +1,24 @@
 """
 =======================
 2022: Lys Sanz Moreta
 Draupnir : Ancestral protein sequence reconstruction using a tree-structured Ornstein-Uhlenbeck variational autoencoder
 =======================
 """
+from types import MethodType
+
+import pyro
 from pyro.contrib.easyguide import EasyGuide
 from pyro.nn import PyroParam
+import torch.nn as nn
+import torch
+import torch.distributions.constraints as constraints
+import pyro.distributions as dist
 from draupnir.models import *
+import draupnir.models_utils as DraupnirModelsUtils
 class DRAUPNIRGUIDES(EasyGuide):
     def __init__(self,draupnir_model,ModelLoad, Draupnir):
         super(DRAUPNIRGUIDES, self).__init__(draupnir_model)
         self.guide_type = ModelLoad.args.select_guide
         self.draupnir = Draupnir
         self.encoder_rnn_input_size = self.draupnir.aa_probs
         self.dataset_train_blosum = self.draupnir.dataset_train_blosum
@@ -18,114 +26,146 @@
         self.batch_by_clade = self.draupnir.batch_by_clade
         if self.draupnir.pretrained_params is not None:
             self.h_0_GUIDE = nn.Parameter(self.draupnir.pretrained_params["h_0_GUIDE"], requires_grad=True).to(self.draupnir.device)
         else:
             self.h_0_GUIDE = nn.Parameter(torch.randn(self.draupnir.gru_hidden_dim), requires_grad=True).to(self.draupnir.device)
         self.encoder = RNNEncoder(self.draupnir.align_seq_len, self.draupnir.aa_probs,self.draupnir.n_leaves, self.draupnir.gru_hidden_dim, self.draupnir.z_dim, self.encoder_rnn_input_size,self.draupnir.kappa_addition,self.draupnir.num_layers,self.draupnir.pretrained_params)
         self.embeddingencoder = EmbedComplexEncoder(self.draupnir.aa_probs,self.draupnir.embedding_dim,self.draupnir.pretrained_params)
-        self.alpha = PyroParam(torch.randn(3),constraint=constraints.positive,event_dim=1) #constraint=constraints.interval(0., 10.)--->TODO:Event dimension??
-        self.sigma_n = PyroParam(torch.randn(self.draupnir.z_dim),constraint=constraints.positive, event_dim=1) #TODO: Change to randn
-        self.sigma_f = PyroParam(torch.randn(self.draupnir.z_dim),constraint=constraints.positive,event_dim=1)
-        self.lambd = PyroParam(torch.randn(self.draupnir.z_dim),constraint=constraints.positive, event_dim=1)
+        self.alpha = PyroParam(dist.HalfNormal(torch.tensor([1.0])).sample([3]),constraint=constraints.positive,event_dim=0) #constraint=constraints.interval(0., 10.)--->TODO:Event dimension??
+        self.sigma_n = PyroParam(dist.HalfNormal(torch.tensor([1.0])).sample([self.draupnir.z_dim]),constraint=constraints.positive,event_dim=0)
+        self.sigma_f = PyroParam(dist.HalfNormal(torch.tensor([1.0])).sample([self.draupnir.z_dim]),constraint=constraints.positive,event_dim=0)
+        self.lambd = PyroParam(dist.HalfNormal(torch.tensor([1.0])).sample([self.draupnir.z_dim]),constraint=constraints.positive,event_dim=0)
+
         if self.draupnir.plating:
             self.encoder_splitted_leaves_indexes = list(torch.tensor_split(torch.arange(self.draupnir.n_leaves), int(self.draupnir.n_leaves / self.draupnir.plate_size)) * self.draupnir.num_epochs)
 
-    def guide(self, family_data, patristic_matrix, cladistic_matrix, data_blosum, batch_blosum):
-        """data_blosum is the data encoded with blosum vectors
-        batch_blosum is the weighted average of blosum scores per column alignment for a batch of sequences"""
+    def guide(self, family_data, patristic_matrix, cladistic_matrix, data_blosum, batch_blosum=None,map_estimates=None):
+        """
+        :param patristic_matrix: matrix of patristic distances (branch lengths) between the nodes in the tree
+        :param cladistic_matrix: matrix of cladistic distances between the nodes in the tree
+        :param data_blosum : data encoded with blosum vectors
+        :param batch_blosum : weighted average of blosum scores per column alignment for a batch of sequences"""
         if self.batch_size == None or self.batch_size > 1:
             if self.batch_by_clade:
                 return self.guide_batch_by_clade(family_data, patristic_matrix, cladistic_matrix, data_blosum,
                                                  batch_blosum)
             else:
                 return self.guide_batch(family_data, patristic_matrix, cladistic_matrix, data_blosum,
                                         batch_blosum=None)
         else:
             return self.guide_noplating(family_data, patristic_matrix, cladistic_matrix, data_blosum,
                                         batch_blosum=None)
 
-    def guide_noplating(self,family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum):
+    def guide_noplating(self,family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum=None,map_estimates=None):
         """
         :param tensor data_blosum here is the ENTIRE data encoded in blosum vector form instead of integers ---> EQUAL to self.dataset_train_blosum
         """
         #aminoacid_sequences = family_data[:, 2:, 0]
-        #Highlight: Initialize variational parameters of the OU process
-        alpha = pyro.sample("alpha", dist.Delta(self.alpha))
-        sigma_n = pyro.sample("sigma_n", dist.Delta(self.sigma_n))
-        sigma_f = pyro.sample("sigma_f", dist.Delta(self.sigma_f))
-        lambd = pyro.sample("lambd", dist.Delta(self.lambd))
+
+        alpha = self.alpha
+        sigma_n = self.sigma_n
+        sigma_f = self.sigma_f
+        lambd = self.lambd
+
+
         pyro.module("encoder", self.encoder)
         pyro.module("embeddingsencoder", self.embeddingencoder)
-        #Highlight: embed the amino acids represented by their respective blosum scores (data_blosim=self.dataset_train_blosum)
-        aminoacid_sequences = self.embeddingencoder(self.dataset_train_blosum) #remember for the corals the aa_prob is 24
-        #aminoacid_sequences = self.dataset_train_blosum
-        encoder_h_0 = self.h_0_GUIDE.expand(self.encoder.num_layers * 2, aminoacid_sequences.shape[0],self.draupnir.gru_hidden_dim).contiguous()
-        #Highlight: Everything, n_leaves and n_z, is independent (we can plate over any of them , is fine)
-        with pyro.plate("plate_guide", aminoacid_sequences.shape[0], dim=-1):
-            z_mean,z_variance = self.encoder(aminoacid_sequences,encoder_h_0) #[n,z_dim]
-            latent_z = pyro.sample("latent_z",dist.Normal(z_mean.T,z_variance.T)) #[z_dim,n]
+        with pyro.plate("plate_batch", dim=-1, device=self.draupnir.device):
+            #Highlight: embed the amino acids represented by their respective blosum scores (data_blosim=self.dataset_train_blosum)
+            aminoacid_sequences = self.embeddingencoder(self.dataset_train_blosum) #remember for the corals the aa_prob is 24
+            #aminoacid_sequences = self.dataset_train_blosum
+            encoder_h_0 = self.h_0_GUIDE.expand(self.encoder.num_layers * 2, aminoacid_sequences.shape[0],self.draupnir.gru_hidden_dim).contiguous()
+            encoder_output = self.encoder(aminoacid_sequences,encoder_h_0) #[n,z_dim]
+            z_loc, z_scale = encoder_output["z_loc"], encoder_output["z_scale"]
+            latent_z = pyro.sample("latent_z",dist.Normal(z_loc.T,z_scale.T)).to_event(1) #[z_dim,n]
+            print("Guide Latent space: {}".format(latent_z.shape))
             assert latent_z.shape == (self.draupnir.z_dim,aminoacid_sequences.shape[0])
+
         return {"alpha":alpha,
                 "sigma_n":sigma_n,
                 "sigma_f":sigma_f,
                 "lambd":lambd,
-                "latent_z":latent_z}
+                "z_loc": z_loc,
+                "z_scale": z_scale,
+                "latent_z": latent_z,
+                "rnn_final_bidirectional": encoder_output["rnn_final_bidirectional"],
+                "rnn_final_hidden_state": encoder_output["rnn_final_hidden_state"],
+                "rnn_hidden_states": encoder_output["rnn_hidden_states"],
+                }
 
-    def guide_batch(self, family_data, patristic_matrix_sorted, cladistic_matrix, data_blosum, batch_blosum):
+    def guide_batch(self, family_data, patristic_matrix_sorted, cladistic_matrix, data_blosum, batch_blosum=None,map_estimates=None):
         """
         :param tensor data_blosum here is the BATCH data encoded in blosum vector form instead of integers
         """
-        # aminoacid_sequences = family_data[:, 2:, 0]
-        # Highlight: Initialize variational parameters of the OU process
-        alpha = pyro.sample("alpha", dist.Delta(self.alpha))
-        sigma_n = pyro.sample("sigma_n", dist.Delta(self.sigma_n))
-        sigma_f = pyro.sample("sigma_f", dist.Delta(self.sigma_f))
-        lambd = pyro.sample("lambd", dist.Delta(self.lambd))
         pyro.module("encoder", self.encoder)
         pyro.module("embeddingsencoder", self.embeddingencoder)
-        # Highlight: embed the amino acids represented by their respective blosum scores
-        aminoacid_sequences = self.embeddingencoder(data_blosum)  # remember for the corals the aa_prob is 24
-        # aminoacid_sequences = self.dataset_train_blosum
-        encoder_h_0 = self.h_0_GUIDE.expand(self.encoder.num_layers * 2, aminoacid_sequences.shape[0],
-                                            self.draupnir.gru_hidden_dim).contiguous()
-        # Highlight: Everything, n_leaves and n_z, is independent (we can plate over any of them , is fine)
-        with pyro.plate("plate_guide", aminoacid_sequences.shape[0], dim=-1):
-            z_mean, z_variance = self.encoder(aminoacid_sequences, encoder_h_0)  # [n,z_dim]
-            latent_z = pyro.sample("latent_z", dist.Normal(z_mean.T, z_variance.T))  # [z_dim,n]
+        # aminoacid_sequences = family_data[:, 2:, 0]
+        with pyro.plate("plate_batch", dim=-1, device=self.draupnir.device):
+            # alpha = pyro.sample("alpha", dist.HalfNormal(1).expand_by([3, ]).to_event(1))
+            # sigma_f = pyro.sample("sigma_f", dist.HalfNormal(alpha[0]).expand_by([self.draupnir.z_dim, ]).to_event(1))  # rate of mean reversion/selection strength---> signal variance #removed .to_event(1)...
+            # sigma_n = pyro.sample("sigma_n",dist.HalfNormal(alpha[1]).expand_by([self.draupnir.z_dim, ]).to_event(1))  # Gaussian noise
+            # lambd = pyro.sample("lambd", dist.HalfNormal(alpha[2]).expand_by([self.draupnir.z_dim, ]).to_event(1))  # characteristic length-scale
+
+            alpha = pyro.sample("alpha", dist.Delta(self.alpha).to_event(1))
+            sigma_n = pyro.sample("sigma_n", dist.Delta(self.sigma_n).to_event(1))
+            sigma_f = pyro.sample("sigma_f", dist.Delta(self.sigma_f).to_event(1))
+            lambd = pyro.sample("lambd", dist.Delta(self.lambd).to_event(1))
+            # Highlight: embed the amino acids represented by their respective blosum scores
+            aminoacid_sequences = self.embeddingencoder(data_blosum)  # remember for the corals the aa_prob is 24
+            # aminoacid_sequences = self.dataset_train_blosum
+            encoder_h_0 = self.h_0_GUIDE.expand(self.encoder.num_layers * 2, aminoacid_sequences.shape[0],
+                                                self.draupnir.gru_hidden_dim).contiguous()
+            # Highlight: Everything, n_leaves and n_z, is independent (we can plate over any of them , is fine)
+            #with pyro.plate("plate_guide", aminoacid_sequences.shape[0], dim=-1):
+            encoder_output = self.encoder(aminoacid_sequences, encoder_h_0)  # [n,z_dim]
+            z_loc,z_scale = encoder_output["z_loc"],encoder_output["z_scale"]
+            latent_z = pyro.sample("latent_z", dist.Normal(z_loc.T, z_scale.T))  # [z_dim,n]
             assert latent_z.shape == (self.draupnir.z_dim, aminoacid_sequences.shape[0])
         return {"alpha": alpha,
                 "sigma_n": sigma_n,
                 "sigma_f": sigma_f,
                 "lambd": lambd,
-                "latent_z": latent_z}
+                "z_loc": z_loc,
+                "z_scale": z_scale,
+                "latent_z": latent_z,
+                "rnn_final_bidirectional":encoder_output["rnn_final_bidirectional"],
+                "rnn_final_hidden_state": encoder_output["rnn_final_hidden_state"],
+                "rnn_hidden_states": encoder_output["rnn_hidden_states"],
+                }
 
-    def guide_batch_by_clade(self, family_data, patristic_matrix_sorted, cladistic_matrix, data_blosum, batch_blosum):
+    def guide_batch_by_clade(self, family_data, patristic_matrix_sorted, cladistic_matrix, data_blosum, batch_blosum=None,map_estimates=None):
 
         """
         :param tensor data_blosum here is the CLADE-BATCHED data in blosum vector form instead of integers
         :param batch_blosum is the weighted average of the blosum vectors for the clade per column site in the MSA"""
         # aminoacid_sequences = family_data[:, 2:, 0]
-        # Highlight: Initialize variational parameters of the OU process
-        alpha = pyro.sample("alpha", dist.Delta(self.alpha))
-        sigma_n = pyro.sample("sigma_n", dist.Delta(self.sigma_n))
-        sigma_f = pyro.sample("sigma_f", dist.Delta(self.sigma_f))
-        lambd = pyro.sample("lambd", dist.Delta(self.lambd))
         pyro.module("encoder", self.encoder)
         pyro.module("embeddingsencoder", self.embeddingencoder)
-        # Highlight: embed the amino acids represented by their respective blosum scores
-        aminoacid_sequences = self.embeddingencoder(data_blosum)  # remember for the corals the aa_prob is 24
-        encoder_h_0 = self.h_0_GUIDE.expand(self.encoder.num_layers * 2, aminoacid_sequences.shape[0],
-                                            self.draupnir.gru_hidden_dim).contiguous()
-        # Highlight: Everything, n_leaves and n_z, is independent (we can plate over any of them , is fine)
-        with pyro.plate("plate_guide", aminoacid_sequences.shape[0], dim=-1):
-            z_mean, z_variance = self.encoder(aminoacid_sequences, encoder_h_0)  # [n,z_dim]
-            latent_z = pyro.sample("latent_z", dist.Normal(z_mean.T, z_variance.T))  # [z_dim,n]
+        with pyro.plate("plate_batch", dim=-1, device=self.draupnir.device):
+            alpha = pyro.sample("alpha", dist.HalfNormal(1).expand_by([3, ]).to_event(1))
+            sigma_f = pyro.sample("sigma_f", dist.HalfNormal(alpha[0]).expand_by([self.draupnir.z_dim, ]).to_event(1))  # rate of mean reversion/selection strength---> signal variance #removed .to_event(1)...
+            sigma_n = pyro.sample("sigma_n",dist.HalfNormal(alpha[1]).expand_by([self.draupnir.z_dim, ]).to_event(1))  # Gaussian noise
+            lambd = pyro.sample("lambd", dist.HalfNormal(alpha[2]).expand_by([self.draupnir.z_dim, ]).to_event(1))  # characteristic length-scale
+            # Highlight: embed the amino acids represented by their respective blosum scores
+            aminoacid_sequences = self.embeddingencoder(data_blosum)  # remember for the corals the aa_prob is 24
+            encoder_h_0 = self.h_0_GUIDE.expand(self.encoder.num_layers * 2, aminoacid_sequences.shape[0],
+                                                self.draupnir.gru_hidden_dim).contiguous()
+            encoder_output = self.encoder(aminoacid_sequences, encoder_h_0)  # [n,z_dim]
+            z_loc, z_scale = encoder_output["z_loc"], encoder_output["z_scale"]
+            latent_z = pyro.sample("latent_z", dist.Normal(z_loc.T, z_scale.T))  # [z_dim,n]
             assert latent_z.shape == (self.draupnir.z_dim, aminoacid_sequences.shape[0])
+
         return {"alpha": alpha,
                 "sigma_n": sigma_n,
                 "sigma_f": sigma_f,
                 "lambd": lambd,
-                "latent_z": latent_z}
+                "z_loc": z_loc,
+                "z_scale": z_scale,
+                "latent_z": latent_z,
+                "rnn_final_bidirectional": encoder_output["rnn_final_bidirectional"],
+                "rnn_final_hidden_state": encoder_output["rnn_final_hidden_state"],
+                "rnn_hidden_states": encoder_output["rnn_hidden_states"]
+                }
```

### Comparing `draupnir-0.0.28/src/draupnir/infer_angles/calculate_coords.py` & `draupnir-0.0.29/src/draupnir/infer_angles/calculate_coords.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/src/draupnir/infer_angles/nerf.py` & `draupnir-0.0.29/src/draupnir/infer_angles/nerf.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/src/draupnir/infer_angles/pnerf.py` & `draupnir-0.0.29/src/draupnir/infer_angles/pnerf.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/src/draupnir/infer_angles/superposition.py` & `draupnir-0.0.29/src/draupnir/infer_angles/superposition.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/src/draupnir/load_utils.py` & `draupnir-0.0.29/src/draupnir/load_utils.py`

 * *Files 2% similar despite different names*

```diff
@@ -542,26 +542,27 @@
         cladistic_matrix_train = cladistic_matrix[train_indx_patristic]
         cladistic_matrix_train = cladistic_matrix_train[:, train_indx_patristic]
         cladistic_matrix_train = cladistic_matrix_train[idx]
     assert Dataset_train[:,0,1].tolist() == patristic_matrix_train[1:,0].tolist()
     #Highlight: Need to invert the dict mapping for later
     correspondence_dict = {v: k for k, v in correspondence_dict.items()}
     return patristic_matrix_train,patristic_matrix_test,cladistic_matrix_train,cladistic_matrix_test,Dataset_test,Dataset_train,correspondence_dict
-def datasets_pretreatment(name,root_sequence_name,train_load,test_load,additional_load,build_config,device,settings_config,script_dir):
+def datasets_pretreatment(name,root_sequence_name,train_load,test_load,additional_load,build_config,args,settings_config,script_dir):
     """ Loads the ancestral sequences depending on the data set, when available. Corrects and sorts all matrices so that they are ordered equally
     :param str name: dataset_name
     :param str root_sequence_name: for the default simulated datasets, we need an additional name string to retrieve the ancestral sequences
     :param namedtuple train_load: namedtuple with all tensors concerning the leaves
     :param namedtuple test_load
     :param namedtuple additional_load
     :param namedtuple build_config
     :param device: torch cpu or torch gpu
     :param namedtuple settings_config
     :param str script_dir
     """
+    device = args.device
     #Highlight: Loading for special test datasets
     if name.startswith("simulation"):
         dataset_test,internal_names_test,max_len_test = DraupnirDatasets.load_simulations_ancestral_sequences(name,
                                                                                         settings_config,
                                                                                         build_config.align_seq_len,#TODO: Hopefully this is always correct
                                                                                         additional_load.tree_levelorder_names,
                                                                                         root_sequence_name,
@@ -762,48 +763,76 @@
 
     def __getitem__(self, index): #sets a[i]
         clade_name = self.clades_names[index]
         clade_data = self.clades_data[index]
         clade_patristic = self.clades_patristic[index]
         clade_blosum_weighted = self.clades_blosum_weighted[index]
         clade_data_blosum = self.clades_data_blosum[index]
-        return {'clade_name': clade_name, 'clade_data': clade_data,'clade_patristic': clade_patristic ,'clade_blosum_weighted':clade_blosum_weighted,'clade_data_blosum':clade_data_blosum}
+        return {'clade_name': clade_name,
+                'clade_data': clade_data,
+                'clade_patristic': clade_patristic ,
+                'clade_blosum_weighted':clade_blosum_weighted,
+                'clade_data_blosum':clade_data_blosum}
     def __len__(self):
         return len(self.clades_names)
+
 class SplittedDataset(Dataset):
     def __init__(self, batches_names, batches_data, batches_patristic, batches_blosum_weighted,batches_data_blosums):
         self.batches_names = batches_names
         self.batches_data = batches_data
         self.batches_patristic = batches_patristic
         self.batches_blosum_weighted = batches_blosum_weighted
         self.batches_data_blosum = batches_data_blosums
 
     def __getitem__(self, index):  # sets a[i]
         batch_name = self.batches_names[index]
         batch_data = self.batches_data[index]
         batch_patristic = self.batches_patristic[index]
         batch_blosum_weighted = self.batches_blosum_weighted[index]
         batch_data_blosum = self.batches_data_blosum[index]
-        return {'batch_name': batch_name, 'batch_data': batch_data, 'batch_patristic': batch_patristic,'batch_blosum_weighted': batch_blosum_weighted,'batch_data_blosum':batch_data_blosum}
+        return {'batch_name': batch_name,
+                'batch_data': batch_data,
+                'batch_patristic': batch_patristic,
+                'batch_blosum_weighted': batch_blosum_weighted,
+                'batch_data_blosum':batch_data_blosum}
 
     def __len__(self):
         return len(self.batches_names)
+
+class CustomDataset(Dataset):
+    def __init__(self, data_array_blosum,data_array_int,data_array_onehot,patristic_matrix):
+        self.batch_data_blosum = data_array_blosum
+        self.batch_data_int = data_array_int
+        self.batch_data_onehot = data_array_onehot
+    def __getitem__(self, index):  # sets a[i]
+        batch_data_blosum = self.batch_data_blosum[index]
+        batch_data_int = self.batch_data_int[index]
+        batch_data_onehot = self.batch_data_onehot[index]
+
+        return {'batch_data_blosum': batch_data_blosum,
+                'batch_data_int':batch_data_int,
+                'batch_data_onehot':batch_data_onehot
+                }
+    def __len__(self):
+        return len(self.batch_data_blosum)
+
+
 def setup_data_loaders(dataset,patristic_matrix_train,clades_dict,blosum,build_config,args,method="batch_dim_0", use_cuda=True):
     """Loads the data set into the model. There are 3 modalities of data loading:
     a) No batching, load the entire data set (batch_dim_0, batch_size=1)
     b) batching, split evenly the data set, the batch size is automatically calculated if None is given (batch_dim_0, batch_size>1 or None)
-    c) Clade batching, loads the data divided by tree clades, the latent space is not splitted, full inference (args.batch_by_clade=True: 1 batch= 1 clade )
+    c) Clade batching, loads the data divided by the tree's clades: the latent space is not splitted, full inference (args.batch_by_clade=True: 1 batch= 1 clade )
     :param dataset
     :param patristic_matrix_train
     :param clades_dict: dictionary containing the tree nodes divided by clades
     :param blosum: BLOSUM matrix
     :param namedtuple build_config
     :param namedtuple args
-    :param method: batching method"""
-    '''If a clade_dict is present it will Load each clade one at the time. Otherwise a predefined batch size is used'''
+    :param method: batching method
+    NOTES: If a clade_dict is present it will Load each clade one at the time. Otherwise a predefined batch size is used"""
     # torch.manual_seed(0)    # For same random split of train/test set every time the code runs!
     kwargs = {'num_workers': 0, 'pin_memory': use_cuda}  # pin-memory has to do with transferring CPU tensors to GPU
     patristic_matrix_train = patristic_matrix_train.detach().cpu()  # otherwise it cannot be used with the train loader
     n_seqs = dataset.shape[0]
     if method == "batch_dim_0":
         if args.batch_size == 1 : #only 1 batch // plating
             train_loader = DataLoader(dataset.cpu(),batch_size=build_config.batch_size,shuffle=False,**kwargs)
```

### Comparing `draupnir-0.0.28/src/draupnir/main.py` & `draupnir-0.0.29/src/draupnir/main.py`

 * *Files 4% similar despite different names*

```diff
@@ -423,15 +423,15 @@
                                          results_dir,
                                          build_config.triTSNE)
     DraupnirPlots.plot_latent_space_pca_by_clade(latent_space_full,
                                     additional_load,
                                     args.num_epochs,
                                     results_dir)
 def save_and_select_model(args,build_config, model_load, patristic_matrix_train,patristic_matrix_full,script_dir,results_dir):
-    """Selects the Draupnir derivation model to use. It saves the model and guides code to a file for easy checking
+    """Selects the class of Draupnir derivation model to use. It saves the entire model's and guide's code to a file for debugging/reproducibility
     :param namedtuple args
     :param namedtuple build_config
     :param namedtuple model_load
     :param tensor patristic_matrix_train [n_leaves+1, n_leaves +1]
     :param tensor patristic_matrix_full [n_leaves + n_internal + 1, n_leaves + n_internal + 1]
     :param str script_dir
     :param str results_dir
@@ -462,15 +462,15 @@
         if args.use_blosum:
             Draupnir = DraupnirModels.DRAUPNIRModel_classic(model_load)
         else:
             Draupnir = DraupnirModels.DRAUPNIRModel_classic_no_blosum(model_load)
         patristic_matrix_model = patristic_matrix_train
     else:  # batching
         print("Batching! (batch_size == None or batch_size != 1)")
-        assert args.select_guide == "variational", "Batching does not support args.select_guide == delta_map"
+        assert args.select_guide == "variational", "Batching does not support args.select_guide == delta_map, please select args.select_guide== variational"
         Draupnir = DraupnirModels.DRAUPNIRModel_batching(model_load)
         patristic_matrix_model = patristic_matrix_train
 
     plating_info = ["WITH PLATING" if args.plating else "WITHOUT plating"][0]
     print("Using model {} {}".format(Draupnir.get_class(),plating_info))
     text_file = open("{}/Hyperparameters_{}_{}epochs.txt".format(results_dir, now.strftime("%Y_%m_%d_%Hh%Mmin%Ss%fms"),args.num_epochs), "a")
     text_file.write("Model Class:  {} \n".format(Draupnir.get_class()))
@@ -628,15 +628,15 @@
     text_file.close()
     pretrained_params_dict_model = torch.load(model_dict_dir)
     pretrained_params_dict_guide = torch.load(guide_dict_dir)
 
 
     model_load = ModelLoad(z_dim=int(params_config["z_dim"]),
                            align_seq_len=align_seq_len,
-                           device=device,
+                           device=args.device,
                            args=args,
                            build_config=build_config,
                            leaves_nodes=dataset_train[:, 0, 1],
                            n_tree_levels=len(additional_info.tree_by_levels_dict),
                            gru_hidden_dim=int(params_config["gru_hidden_dim"]),
                            pretrained_params=None,
                            aa_frequencies=aa_frequencies,
@@ -934,41 +934,38 @@
                    additional_load,
                    additional_info,
                    build_config,
                    settings_config,
                    params_config,
                    n_samples,
                    args,
-                   device,
                    script_dir,
                    results_dir,
                    graph_coo=None,
                    clades_dict=None):
     """Trains Draupnir-OU by performing SVI inference with non-batched training
     :param namedtuple train_load : contains several tensors and other data structures related to the training leaves. See see utils.create_dataset for information on the input data format
     :param namedtuple test_load : contains several tensors and other data structures related to the test internal nodes
     :param namedtuple additional_load
     :param namedtuple additional_info
     :param namedtuple build_config
     :param namedtuple settings_config
     :param dict params_config
     :param int n_samples
     :param namedtuple args
-    :param str device
     :param str script_dir
     :param str results_dir
     :param graph graph_coo: graph that embedds the tree into a COO graph that works with pytorch geometric
     :param dict clades_dict"""
     align_seq_len = build_config.align_seq_len
     if not additional_load.correspondence_dict:
         correspondence_dict = dict(zip(list(range(len(additional_load.tree_levelorder_names))), additional_load.tree_levelorder_names))
     else:
         correspondence_dict = additional_load.correspondence_dict
-    #device = [torch.device("cuda") if args.use_cuda else torch.device("cpu")][0]
-    device = torch.device("cuda" if args.use_cuda else "cpu") #TODO: Check that this works
+    device = args.device
     blosum = additional_info.blosum.to(device)
     aa_frequencies = additional_load.aa_frequencies.to(device)
     dataset_train = train_load.dataset_train.to(device)
     patristic_matrix_train = train_load.patristic_matrix_train.to(device)
     patristic_matrix_full = additional_load.patristic_matrix_full.to(device)
     patristic_matrix_test = test_load.patristic_matrix_test.to(device)
     dataset_test = test_load.dataset_test.to(device)
@@ -1026,25 +1023,27 @@
     text_file = open("{}/Hyperparameters_{}_{}epochs.txt".format(results_dir, now.strftime("%Y_%m_%d_%Hh%Mmin%Ss%fms"),args.num_epochs), "a")
     text_file.write("ELBO :  {} \n".format(str(elbo)))
     text_file.write("Guide :  {} \n".format(str(guide.__class__)))
 
     #Highlight: Select optimizer/scheduler
     if args.use_scheduler:
         print("Using a learning rate scheduler on top of the optimizer!")
-        adam_args = {"lr": params_config["lr"], "betas": (params_config["beta1"], params_config["beta2"]), "eps": params_config["eps"],
+        adam_args = {"lr": params_config["lr"],
+                     "betas": (params_config["beta1"],params_config["beta2"]),
+                     "eps": params_config["eps"],
                      "weight_decay": params_config["weight_decay"]}
         optim = torch.optim.Adam #Highlight: For the scheduler we need to use TORCH.optim not PYRO.optim, and there is no clipped adam in torch
         #Highlight: "Reduce LR on plateau: Scheduler: Reduce learning rate when a metric has stopped improving."
         optim = pyro.optim.ReduceLROnPlateau({'optimizer': optim, 'optim_args': adam_args})
 
     else:
         clippedadam_args = {"lr": params_config["lr"], "betas": (params_config["beta1"], params_config["beta2"]), "eps": params_config["eps"],
                      "weight_decay": params_config["weight_decay"], "clip_norm": params_config["clip_norm"], "lrd": params_config["lrd"]}
         optim = pyro.optim.ClippedAdam(clippedadam_args)
-    def load_tune_params(load_params):
+    def load_tune_params(load_params):#TODO: Remove?
         """Loading pretrained parameters and allowing to tune them"""
         if load_params:
             pyro.clear_param_store()
             tune_folder = "/home/lys/Dropbox/PhD/DRAUPNIR/PLOTS_GP_VAE_PF00400_2021_11_17_22h59min34s252388ms_30000epochs"#delta map
             #tune_folder = ""
             print("WARNING: Loading pretrained model dict from {} !!!!".format(tune_folder))
             optim_dir = None
@@ -1060,58 +1059,60 @@
     svi = SVI(Draupnir.model, guide,optim,elbo)
     text_file.write("Optimizer :  {} \n".format(optim))
 
     check_point_epoch = [50 if args.num_epochs < 100 else (args.num_epochs / 100)][0]
 
     batching_method = ["batch_dim_0" if not args.batch_by_clade else "batch_by_clade"][0]
     train_loader = DraupnirLoadUtils.setup_data_loaders(dataset_train, patristic_matrix_train,clades_dict,blosum,build_config,args,method=batching_method, use_cuda=args.use_cuda)
-
-    training_method= lambda f, svi, patristic_matrix_model, cladistic_matrix_full,dataset_train_blosum,train_loader,args: lambda svi, patristic_matrix_model, cladistic_matrix_full,train_loader,args: f(svi, patristic_matrix_model, cladistic_matrix_full,dataset_train_blosum,train_loader,args)
-
-    if args.batch_by_clade and clades_dict:
-        training_function = training_method(DraupnirTrain.train_batch_clade,svi, patristic_matrix_model, cladistic_matrix_full,dataset_train_blosum,train_loader,args)
-    elif args.batch_size == 1:#no batching or plating
-        training_function = training_method(DraupnirTrain.train, svi, patristic_matrix_model,cladistic_matrix_full,dataset_train_blosum, train_loader, args)
-
-    else:#batching
-        training_function = training_method(DraupnirTrain.train_batch, svi, patristic_matrix_model,
-                                            cladistic_matrix_full,dataset_train_blosum, train_loader, args)
+    map_estimates = None
+    training_function_input = {"patristic_matrix_model":patristic_matrix_model,
+                   "cladistic_matrix_full":cladistic_matrix_full,
+                   "dataset_train_blosum":dataset_train_blosum,
+                   "train_loader":train_loader,
+                   "args":args,
+                   "map_estimates":map_estimates,
+                   "args":args}
+    training_function = DraupnirTrain.select_training_function(clades_dict,svi, training_function_input)
 
     ######################
     ####Training Loop#####
     ######################
     train_loss = []
     entropy = []
     average_pid_list = []
     std_pid_list = []
     #gradient_norms = defaultdict(list)
     start_total = time.time()
     epoch = 0
     epoch_count=0
     added_epochs = 0
     output_file = open("{}/output.log".format(results_dir),"w")
+
     while epoch < args.num_epochs:
         if check_point_epoch > 0 and epoch > 0 and epoch % check_point_epoch == 0:
             DraupnirPlots.plot_ELBO(train_loss, results_dir)
             DraupnirPlots.plot_entropy(entropy, results_dir)
             plot_percent_id(average_pid_list, std_pid_list, results_dir)
         start = time.time()
-        total_epoch_loss_train = training_function(svi, patristic_matrix_model, cladistic_matrix_full,train_loader,args)
+
+        map_estimates = guide(dataset_train, patristic_matrix_train, cladistic_matrix_train, dataset_train_blosum,
+                              None,
+                              None)
+        training_function_input["map_estimates"] = map_estimates
+        total_epoch_loss_train = training_function(svi, training_function_input)
         memory_usage_mib = torch.cuda.max_memory_allocated()*9.5367*1e-7 #convert byte to MiB
         stop = time.time()
         train_loss.append(float(total_epoch_loss_train)) #convert to float because otherwise it's kept in torch's history
         print("[epoch %03d]  average training loss: %.4f %.5g time/epoch %.2f MiB/epoch" % (epoch_count, total_epoch_loss_train, stop - start,memory_usage_mib))
         print("[epoch %03d]  average training loss: %.4f %.5g time/epoch %.2f MiB/epoch" % (epoch_count, total_epoch_loss_train, stop - start,memory_usage_mib),file=output_file)
         print("Current total time : {}".format(str(datetime.timedelta(seconds=stop-start_total))),file=output_file)
         # # Register hooks to monitor gradient norms.
         # for name_i, value in pyro.get_param_store().named_parameters():
         #     value.register_hook(lambda g, name_i=name_i: gradient_norms[name_i].append(g.norm().item()))
         #map_estimates = guide(dataset_train,patristic_matrix_train,cladistic_matrix_train,batch_blosum=None) #only saving 1 sample
-        map_estimates = guide(dataset_train,patristic_matrix_train,cladistic_matrix_train,dataset_train_blosum,batch_blosum=None) #only saving 1 sample
-
         map_estimates = {val: key.detach() for val, key in map_estimates.items()}
         sample_out_train = Draupnir.sample(map_estimates,
                                            1,
                                            dataset_train,
                                            patristic_matrix_full,
                                            cladistic_matrix_train,
                                            use_argmax=True,
@@ -1460,15 +1461,14 @@
                    additional_load,
                    additional_info,
                    build_config,
                    settings_config,
                    params_config,
                    n_samples,
                    args,
-                   device,
                    script_dir,
                    results_dir,
                    graph_coo=None,
                    clades_dict=None):
     """Trains Draupnir-OU by performing SVI inference with conditionally independent batched training
     :param namedtuple train_load : contains several tensors and other data structures related to the training leaves. See see utils.create_dataset for information on the input data format
     :param namedtuple test_load : contains several tensors and other data structures related to the test internal nodes
@@ -1486,15 +1486,15 @@
     :param dict clades_dict"""
     align_seq_len = build_config.align_seq_len
     if not additional_load.correspondence_dict:
         correspondence_dict = dict(
             zip(list(range(len(additional_load.tree_levelorder_names))), additional_load.tree_levelorder_names))
     else:
         correspondence_dict = additional_load.correspondence_dict
-    device = torch.device("cuda" if args.use_cuda else "cpu") #TODO: Check that this works
+    device = args.device
     blosum = additional_info.blosum.to(device)
     aa_frequencies = additional_load.aa_frequencies.to(device)
     dataset_train = train_load.dataset_train.to(device)
     patristic_matrix_train = train_load.patristic_matrix_train.to(device)
     patristic_matrix_full = additional_load.patristic_matrix_full.to(device)
     patristic_matrix_test = test_load.patristic_matrix_test.to(device)
     dataset_test = test_load.dataset_test.to(device)
@@ -1601,26 +1601,27 @@
     svi = SVI(Draupnir.model, guide, optim,elbo)  # TODO: TraceMeanField_ELBO() http://docs.pyro.ai/en/0.3.0-release/inference_algos.html#pyro.infer.trace_mean_field_elbo.TraceMeanField_ELBO
     text_file.write("Optimizer :  {} \n".format(optim))
 
     check_point_epoch = [50 if args.num_epochs < 100 else (args.num_epochs / 100)][0]
 
     batching_method = ["batch_dim_0" if not args.batch_by_clade else "batch_by_clade"][0]
     train_loader = DraupnirLoadUtils.setup_data_loaders(dataset_train, patristic_matrix_train,clades_dict,blosum,build_config,args,method=batching_method, use_cuda=args.use_cuda)
-
-
-    training_method= lambda f, svi, patristic_matrix_model, cladistic_matrix_full,dataset_train_blosum,train_loader,args: lambda svi, patristic_matrix_model, cladistic_matrix_full,train_loader,args: f(svi, patristic_matrix_model, cladistic_matrix_full,dataset_train_blosum,train_loader,args)
-
-    if args.batch_by_clade and clades_dict:
-        training_function = training_method(DraupnirTrain.train_batch_clade,svi, patristic_matrix_model, cladistic_matrix_full,dataset_train_blosum,train_loader,args)
-    elif args.batch_size == 1:#no batching or plating
-        training_function = training_method(DraupnirTrain.train, svi, patristic_matrix_model,cladistic_matrix_full,dataset_train_blosum, train_loader, args)
-
-    else:#batching
-        training_function = training_method(DraupnirTrain.train_batch, svi, patristic_matrix_model,
-                                            cladistic_matrix_full,dataset_train_blosum, train_loader, args)
+    # map_estimates = None
+    # training_function = DraupnirTrain.select_training_function(clades_dict, svi, patristic_matrix_model,
+    #                                                            cladistic_matrix_full, dataset_train_blosum,
+    #                                                            train_loader, args, map_estimates)
+    map_estimates = None
+    training_function_input = {"patristic_matrix_model":patristic_matrix_model,
+                   "cladistic_matrix_full":cladistic_matrix_full,
+                   "dataset_train_blosum":dataset_train_blosum,
+                   "train_loader":train_loader,
+                   "args":args,
+                   "map_estimates":map_estimates,
+                   "args":args}
+    training_function = DraupnirTrain.select_training_function(clades_dict,svi, training_function_input)
 
     ######################
     ####Training Loop#####
     ######################
     n_train_seqs = dataset_train.shape[0]
     blocks_train = DraupnirModelsUtils.intervals(n_train_seqs // build_config.batch_size, n_train_seqs)
     train_loss = []
@@ -1631,26 +1632,29 @@
     added_epochs = 0
     output_file = open("{}/output.log".format(results_dir), "w")
     while epoch < args.num_epochs:
         if check_point_epoch > 0 and epoch > 0 and epoch % check_point_epoch == 0:
             DraupnirPlots.plot_ELBO(train_loss, results_dir)
             DraupnirPlots.plot_entropy(entropy, results_dir)
         start = time.time()
-        total_epoch_loss_train = training_function(svi, patristic_matrix_model, cladistic_matrix_full, train_loader,args)
+        map_estimates = guide(dataset_train, patristic_matrix_train, cladistic_matrix_train, dataset_train_blosum,
+                              batch_blosum=None,
+                              map_estimates=None)  # only saving 1 sample
+        training_function_input["map_estimates"] = map_estimates
+        total_epoch_loss_train = training_function(svi, training_function_input)
 
         memory_usage_mib = torch.cuda.max_memory_allocated() * 9.5367 * 1e-7  # convert byte to MiB
         stop = time.time()
         train_loss.append(float(total_epoch_loss_train))  # convert to float because otherwise it's kept in torch's history
         print("[epoch %03d]  average training loss: %.4f %.5g time/epoch %.2f MiB/epoch" % (
         epoch_count, total_epoch_loss_train, stop - start, memory_usage_mib))
         print("[epoch %03d]  average training loss: %.4f %.5g time/epoch %.2f MiB/epoch" % (
         epoch_count, total_epoch_loss_train, stop - start, memory_usage_mib), file=output_file)
         print("Current total time : {}".format(str(datetime.timedelta(seconds=stop - start_total))), file=output_file)
 
-        map_estimates = guide(dataset_train, patristic_matrix_train, cladistic_matrix_train, dataset_train_blosum,batch_blosum=None)  # only saving 1 sample
         map_estimates = {val: key.detach() for val, key in map_estimates.items()}
         sample_out_train = Draupnir.sample_batched(map_estimates,
                                            n_samples,
                                            dataset_train,
                                            patristic_matrix_full,
                                            patristic_matrix_test,
                                            batch_idx=blocks_train[0],#only perform testing on one of the batches
@@ -1950,15 +1954,14 @@
                    additional_load,
                    additional_info,
                    build_config,
                    settings_config,
                    params_config,
                    n_samples,
                    args,
-                   device,
                    script_dir,
                    results_dir,
                    graph_coo=None,
                    clades_dict=None):
     """Trains Draupnir-OU by performing SVI inference with batched training, where each batch is a clade in the tree #TODO: scale ELBO by batch size
     :param namedtuple train_load : contains several tensors and other data structures related to the training leaves. See see utils.create_dataset for information on the input data format
     :param namedtuple test_load : contains several tensors and other data structures related to the test internal nodes
@@ -1977,15 +1980,15 @@
     print("Batching by clade")
     align_seq_len = build_config.align_seq_len
     if not additional_load.correspondence_dict:
         correspondence_dict = dict(
             zip(list(range(len(additional_load.tree_levelorder_names))), additional_load.tree_levelorder_names))
     else:
         correspondence_dict = additional_load.correspondence_dict
-    device = torch.device("cuda" if args.use_cuda else "cpu") #TODO: Check that this works
+    device = args.device
     blosum = additional_info.blosum.to(device)
     aa_frequencies = additional_load.aa_frequencies.to(device)
     dataset_train = train_load.dataset_train.to(device)
     patristic_matrix_train = train_load.patristic_matrix_train.to(device)
     patristic_matrix_full = additional_load.patristic_matrix_full.to(device)
     patristic_matrix_test = test_load.patristic_matrix_test.to(device)
     dataset_test = test_load.dataset_test.to(device)
@@ -2102,23 +2105,23 @@
     training_method = lambda f, svi, patristic_matrix_model, cladistic_matrix_full, dataset_train_blosum, train_loader,args: lambda svi, patristic_matrix_model, cladistic_matrix_full, train_loader, args: f(svi,
                                                                                                                     patristic_matrix_model,
                                                                                                                     cladistic_matrix_full,
                                                                                                                     dataset_train_blosum,
                                                                                                                     train_loader,
                                                                                                                     args)
 
-    if args.batch_by_clade and clades_dict:
-        training_function = training_method(DraupnirTrain.train_batch_clade, svi, patristic_matrix_model,
-                                            cladistic_matrix_full, dataset_train_blosum, train_loader, args)
-    elif args.batch_size == 1:  # no batching or plating
-        training_function = training_method(DraupnirTrain.train, svi, patristic_matrix_model, cladistic_matrix_full,
-                                            dataset_train_blosum, train_loader, args)
-    else:  # batching
-        training_function = training_method(DraupnirTrain.train_batch, svi, patristic_matrix_model,
-                                            cladistic_matrix_full, dataset_train_blosum, train_loader, args)
+    map_estimates = None
+    training_function_input = {"patristic_matrix_model":patristic_matrix_model,
+                   "cladistic_matrix_full":cladistic_matrix_full,
+                   "dataset_train_blosum":dataset_train_blosum,
+                   "train_loader":train_loader,
+                   "args":args,
+                   "map_estimates":map_estimates,
+                   "args":args}
+    training_function = DraupnirTrain.select_training_function(clades_dict,svi, training_function_input)
 
     ######################
     ####Training Loop#####
     ######################
     #n_train_seqs = dataset_train.shape[0]
     #blocks_train = DraupnirModelsUtils.intervals(n_train_seqs // build_config.batch_size, n_train_seqs) #TODO: Fix to have the clades or just take a slice?
     blocks_train = [[value] if isinstance(value,int) else value for value in clades_dict.values()]
@@ -2131,27 +2134,27 @@
     added_epochs = 0
     output_file = open("{}/output.log".format(results_dir), "w")
     while epoch < args.num_epochs:
         if check_point_epoch > 0 and epoch > 0 and epoch % check_point_epoch == 0:
             DraupnirPlots.plot_ELBO(train_loss, results_dir)
             DraupnirPlots.plot_entropy(entropy, results_dir)
         start = time.time()
-        total_epoch_loss_train = training_function(svi, patristic_matrix_model, cladistic_matrix_full, train_loader,
-                                                   args)
+        map_estimates = guide(dataset_train, patristic_matrix_train, cladistic_matrix_train, dataset_train_blosum,batch_blosum=None,map_estimates=None)  # only saving 1 sample
+        training_function_input["map_estimates"] = map_estimates
+        total_epoch_loss_train = training_function(svi, training_function_input)
         memory_usage_mib = torch.cuda.max_memory_allocated() * 9.5367 * 1e-7  # convert byte to MiB
         stop = time.time()
         train_loss.append(float(total_epoch_loss_train))  # convert to float because otherwise it's kept in torch's history
         print("[epoch %03d]  average training loss: %.4f %.5g time/epoch %.2f MiB/epoch" % (
         epoch_count, total_epoch_loss_train, stop - start, memory_usage_mib))
         print("[epoch %03d]  average training loss: %.4f %.5g time/epoch %.2f MiB/epoch" % (
         epoch_count, total_epoch_loss_train, stop - start, memory_usage_mib), file=output_file)
         print("Current total time : {}".format(str(datetime.timedelta(seconds=stop - start_total))), file=output_file)
 
-        map_estimates = guide(dataset_train, patristic_matrix_train, cladistic_matrix_train, dataset_train_blosum,
-                              batch_blosum=None)  # only saving 1 sample
+
         map_estimates = {val: key.detach() for val, key in map_estimates.items()}
         sample_out_train = Draupnir.sample_batched(map_estimates,
                                            n_samples,
                                            dataset_train,
                                            patristic_matrix_full,
                                            patristic_matrix_test,
                                            batch_idx=blocks_train[0],#only perform testing on one of the batches
@@ -2784,31 +2787,30 @@
     global config
     n_runs = 20
     for i in range(n_runs):
         config = generate_config()
         print(config)
         proc= subprocess.Popen(args=[sys.executable,"Draupnir_example.py","--parameter-search","True","--config-dict",str(config).replace("'", '"')],stdout=open('Random_Search_results.txt', 'a')) #stdout=open(os.devnull, 'wb'),stderr=open(os.devnull, 'wb')
         proc.communicate()
-def run(name,root_sequence_name,args,device,settings_config,build_config,script_dir):
+def run(name,root_sequence_name,args,settings_config,build_config,script_dir):
     """Loads and pre-treats the data for inference, executes Draupnir model for training or for sampling.
     :param str name
     :param str root_sequence_name
     :param namedtuple args
-    :param device: torch device
     :param namedtuple settings_config
     :param namedtuple build_config
     :param str script_dir
     """
     results_dir = "{}/PLOTS_Draupnir_{}_{}_{}epochs_{}".format(script_dir, name, now.strftime("%Y_%m_%d_%Hh%Mmin%Ss%fms"),args.num_epochs, args.select_guide)
 
     print("Loading datasets....")
     param_config = config_build(args)
     train_load,test_load,additional_load,build_config = load_data(name,settings_config,build_config,param_config,results_dir,script_dir,args)
     additional_info=DraupnirUtils.extra_processing(additional_load.ancestor_info_numbers, additional_load.patristic_matrix_full,results_dir,args,build_config)
-    train_load,test_load,additional_load= DraupnirLoadUtils.datasets_pretreatment(name,root_sequence_name,train_load,test_load,additional_load,build_config,device,settings_config,script_dir)
+    train_load,test_load,additional_load= DraupnirLoadUtils.datasets_pretreatment(name,root_sequence_name,train_load,test_load,additional_load,build_config,args,settings_config,script_dir)
     torch.save(torch.get_rng_state(),"{}/rng_key.torch".format(results_dir))
     if args.one_hot_encoded:
         raise ValueError("Please set one_hot_encoding to False")
     print("Starting Draupnir ...")
     print("Dataset: {}".format(name))
     print("Number epochs: {}".format(args.num_epochs))
     print("Z/latent Size: {}".format(param_config["z_dim"]))
@@ -2831,15 +2833,14 @@
                         additional_load,
                         additional_info,
                         build_config,
                         settings_config,
                         param_config,
                         args.n_samples,
                         args,
-                        device,
                         script_dir,
                         results_dir,
                         graph_coo,
                         clades_dict)
     elif args.batch_size == None or args.batch_size > 1:
         print("Batching, splits the OU stochastic process, no guarantee on latent space with tree structure")
         if args.batch_by_clade:
@@ -2848,30 +2849,28 @@
                         additional_load,
                         additional_info,
                         build_config,
                         settings_config,
                         param_config,
                         args.n_samples,
                         args,
-                        device,
                         script_dir,
                         results_dir,
                         graph_coo,
                         clades_dict)
         else:
             draupnir_train_batching(train_load,
                         test_load,
                         additional_load,
                         additional_info,
                         build_config,
                         settings_config,
                         param_config,
                         args.n_samples,
                         args,
-                        device,
                         script_dir,
                         results_dir,
                         graph_coo,
                         clades_dict)
     #TODO: draupnir_tuning
     else:
         print("Training Draupnir with the entire tree at once, not batching")
@@ -2880,15 +2879,14 @@
                        additional_load,
                        additional_info,
                        build_config,
                        settings_config,
                        param_config,
                        args.n_samples,
                        args,
-                       device,
                        script_dir,
                        results_dir,
                        graph_coo,
                        clades_dict)
```

### Comparing `draupnir-0.0.28/src/draupnir/models.py` & `draupnir-0.0.29/src/draupnir/models.py`

 * *Files 1% similar despite different names*

```diff
@@ -61,19 +61,19 @@
         self.n_internal = len(self.internal_nodes)
         self.n_all = self.n_leaves + self.n_internal
         self.num_layers = 1 #TODO: Remove
         self.h_0_MODEL = nn.Parameter(torch.randn(self.gru_hidden_dim), requires_grad=True).to(self.device)
         if ModelLoad.args.use_cuda:
             self.cuda()
     @abstractmethod
-    def guide(self, family_data, patristic_matrix, cladistic_matrix, data_blosum, batch_blosum):
+    def guide(self, family_data, patristic_matrix, cladistic_matrix, data_blosum, batch_blosum,guide_map_estimates):
         raise NotImplementedError
 
     @abstractmethod
-    def model(self, family_data, patristic_matrix, cladistic_matrix, data_blosum, batch_blosum):
+    def model(self, family_data, patristic_matrix, cladistic_matrix, data_blosum, batch_blosum,guide_map_estimates):
         raise NotImplementedError
 
     @abstractmethod
     def sample(self, map_estimates, n_samples, family_data_test, patristic_matrix, cladistic_matrix, use_test=True):
         raise NotImplementedError
     @abstractmethod
     def get_class(self):
@@ -81,40 +81,50 @@
         name = str(full_name).split(".")[-1].replace("'>","")
         return name
     def gp_prior(self,patristic_matrix_sorted):
         """Computes an Ornstein Ulenbeck process prior over the latent space, representing the evolutionary process.
         The Gaussian prior consists of a Ornstein - Ulenbeck kernel that uses the patristic distances tu build a covariance matrix"""
         # Highlight; OU kernel parameters
         alpha = pyro.sample("alpha", dist.HalfNormal(1).expand_by([3, ]).to_event(1)) + 1e-6 #TODO: Change to another distribution less centered around 0
-        sigma_f = pyro.sample("sigma_f", dist.HalfNormal(alpha[0]).expand_by([self.z_dim, ]).to_event(1))  +1e-6 # rate of mean reversion/selection strength---> signal variance #removed .to_event(1)...
+        sigma_f = pyro.sample("sigma_f", dist.HalfNormal(alpha[0]).expand_by([self.z_dim, ]).to_event(1)) +1e-6 # rate of mean reversion/selection strength---> signal variance #removed .to_event(1)...
         sigma_n = pyro.sample("sigma_n", dist.HalfNormal(alpha[1]).expand_by([self.z_dim, ]).to_event(1))  +1e-6 # Gaussian noise
         lambd = pyro.sample("lambd", dist.HalfNormal(alpha[2]).expand_by([self.z_dim, ]).to_event(1))  +1e-6 # characteristic length-scale
+
+        alpha = DraupnirUtils.squeeze_tensor(1,alpha)
+        sigma_f = DraupnirUtils.squeeze_tensor(1,sigma_f)
+        sigma_n = DraupnirUtils.squeeze_tensor(1,sigma_n)
+        lambd = DraupnirUtils.squeeze_tensor(1,lambd)
+
         # Highlight: Sample the latent space from MultivariateNormal with GP prior on covariance
         patristic_matrix = patristic_matrix_sorted[1:, 1:]
         OU_covariance = OUKernel_Fast(sigma_f, sigma_n, lambd).forward(patristic_matrix)
         OU_mean = torch.zeros((patristic_matrix.shape[0],)).unsqueeze(0)
+        # print("Model Covariance: {}".format(OU_covariance.shape))
+        # print("Model Mean: {}".format(OU_mean.shape))
+
         if self.leaves_testing:
-            assert OU_covariance.shape == (self.z_dim, self.n_all, self.n_all)
+            assert OU_covariance.shape == (self.z_dim, self.n_all, self.n_all),f"Expected shape {(self.z_dim, self.n_all, self.n_all)}, got {OU_covariance.shape}"
             assert OU_mean.shape == (1, self.n_all)
         else:
-            assert OU_covariance.shape == (self.z_dim, self.n_leaves, self.n_leaves)
+
+            assert OU_covariance.shape == (self.z_dim, self.n_leaves, self.n_leaves), f"Expected shape {(self.z_dim, self.n_leaves, self.n_leaves)}, got {OU_covariance.shape}"
             assert OU_mean.shape == (1,self.n_leaves)
         #noise = 1e-15 + torch.eye(OU_covariance.shape[1])
         #https://github.com/pyro-ppl/pyro/issues/702
         #https://forum.pyro.ai/t/runtimeerror-during-cholesky-decomposition/1216/2---> fix runtime error with choleky decomposition
         #https://forum.pyro.ai/t/using-constraints-within-an-nn-module/486
         #OU_covariance = transform_to(constraints.lower_cholesky)(OU_covariance) #check that this does not affect performance
         latent_space = pyro.sample('latent_z', dist.MultivariateNormal(OU_mean, OU_covariance ).to_event(1)) #[z_dim=30,n_nodes] #+ noise[None,:,:]
+        #print("Model Latent space: {}".format(latent_space.shape))
         latent_space = latent_space.T
         return latent_space
     def gp_prior_batched(self,patristic_matrix_sorted):
         "Computes a Gaussian prior over the latent space. The Gaussian prior consists of a Ornstein - Ulenbeck kernel that uses the patristic distances to build a covariance matrix"
         # Highlight; OU kernel parameters #TODO: Add noise to OU parameters to avoid error in cholesky decomposition
         alpha = pyro.sample("alpha", dist.HalfNormal(1).expand_by([3, ]).to_event(1))
-        #alpha = pyro.sample("alpha", dist.Weibull(1,5).expand_by([3, ]).to_event(1))
         sigma_f = pyro.sample("sigma_f", dist.HalfNormal(alpha[0]).expand_by([self.z_dim, ]).to_event(1))  # rate of mean reversion/selection strength---> signal variance #removed .to_event(1)...
         sigma_n = pyro.sample("sigma_n", dist.HalfNormal(alpha[1]).expand_by([self.z_dim, ]).to_event(1))  # Gaussian noise
         lambd = pyro.sample("lambd", dist.HalfNormal(alpha[2]).expand_by([self.z_dim, ]).to_event(1))  # characteristic length-scale
         # Highlight: Sample the latent space from MultivariateNormal with GP prior on covariance
         patristic_matrix = patristic_matrix_sorted[1:, 1:]
         OU_covariance = OUKernel_Fast(sigma_f, sigma_n, lambd).forward(patristic_matrix)
         OU_mean = torch.zeros((patristic_matrix.shape[0],)).unsqueeze(0)
@@ -124,40 +134,40 @@
         #https://github.com/pyro-ppl/pyro/issues/702
         #https://forum.pyro.ai/t/runtimeerror-during-cholesky-decomposition/1216/2---> fix runtime error with choleky decomposition
         #https://forum.pyro.ai/t/using-constraints-within-an-nn-module/486
         #OU_covariance = transform_to(constraints.lower_cholesky)(OU_covariance) #check that this does not affect performance
         latent_space = pyro.sample('latent_z', dist.MultivariateNormal(OU_mean, OU_covariance ).to_event(1)) #[z_dim=30,n_nodes] #+ noise[None,:,:]
         latent_space = latent_space.T
         return latent_space
-
     def map_sampling(self,map_estimates,patristic_matrix_full):
         "Use map sampling for leaves prediction/testing, when internal nodes are not available"
         test_indexes = (patristic_matrix_full[1:, 0][..., None] == self.internal_nodes).any(-1) #indexes of the leaves selected for testing
         latent_space_internal = map_estimates["latent_z"][:, test_indexes].T
         assert latent_space_internal.shape == (self.n_internal, self.z_dim)
         return latent_space_internal
     def conditional_sampling(self,map_estimates, patristic_matrix):
             """Conditional sampling the internal nodes given the leaves from a Multivariate Normal according to page 698 at Pattern Recognition and ML (Bishop)
             :param map_estimates: dictionary conatining the MAP estimates for the OU process parameters
             :param patristic_matrix: full patristic matrix"""
-            sigma_f = map_estimates["sigma_f"]
-            sigma_n = map_estimates["sigma_n"]
-            lambd = map_estimates["lambd"]
+            sigma_f = DraupnirUtils.squeeze_tensor(1,map_estimates["sigma_f"])
+            sigma_n = DraupnirUtils.squeeze_tensor(1,map_estimates["sigma_n"])
+            lambd = DraupnirUtils.squeeze_tensor(1,map_estimates["lambd"])
+
             internal_indexes = (patristic_matrix[1:, 0][..., None] == self.internal_nodes).any(-1)
 
             #n_internal = family_data_test.shape[0]
             # Highlight: Sample the ancestors conditiones on the leaves (by using the full patristic matrix). See Page 689 at Patter Recongnition and Ml (Bishop)
             # Highlight: Formula is: p(xa|xb) = N (x|µa|b, Λ−1aa ) , a = test/internal; b= train/leaves
             patristic_matrix_full = patristic_matrix[1:, 1:]
             assert patristic_matrix_full.shape == (self.n_all, self.n_all), "Remember to use the entire/full patristic matrix for conditional sampling!"
             OU = OUKernel_Fast(sigma_f, sigma_n, lambd)
             OU_covariance_full = OU.forward(patristic_matrix_full)
             # Highlight: Calculate the inverse of the covariance matrix Λ ≡ Σ−1
             inverse_full = torch.linalg.inv(OU_covariance_full)  # [z_dim,n_test+n_train,n_test+n_train]
-            assert inverse_full.shape == (self.z_dim, self.n_all, self.n_all)
+            assert inverse_full.shape == (self.z_dim, self.n_all, self.n_all), f"Expected dimensions : {(self.z_dim, self.n_all, self.n_all)}, got {inverse_full.shape}"
             # Highlight: B.49 Λ−1aa
             inverse_internal = inverse_full[:, internal_indexes, :]
             inverse_internal = inverse_internal[:, :, internal_indexes]  # [z_dim,n_test,n_test]
             assert inverse_internal.shape == (self.z_dim, self.n_internal, self.n_internal)
             # Highlight: Conditional mean Mean ---->B-50:  µa|b = µa − Λ−1aa Λab(xb − µb)
             # Highlight: µa
             OU_mean_internal = torch.zeros((self.n_internal,))  # [n_internal,]
@@ -173,37 +183,38 @@
             # Highlight:µb
             OU_mean_leaves = torch.zeros((self.n_leaves,))
             # Highlight:µa|b---> Splitted Equation  B-50
             part1 = torch.matmul(torch.linalg.inv(inverse_internal), inverse_internal_leaves)  # [z_dim,n_test,n_train]
             part2 = xb - OU_mean_leaves[None, :]  # [z_dim,n_train]
             OU_mean = OU_mean_internal[None, :, None] - torch.matmul(part1, part2[:, :,None])  # [:,n_test,:] - [z_dim,n_test,None]
             assert OU_mean.squeeze(-1).shape == (self.z_dim, self.n_internal)
+
             latent_space = dist.MultivariateNormal(OU_mean.squeeze(-1), torch.linalg.inv(inverse_internal) + 1e-6).to_event(1).sample()
             latent_space = latent_space.T
             assert latent_space.shape == (self.n_internal, self.z_dim)
             return latent_space
     def conditional_samplingMAP(self,map_estimates, patristic_matrix):
             """Conditional sampling the internal nodes given the leaves from a Multivariate Normal according to page 698 at Pattern Recognition and ML (Bishop)
             :param map_estimates: dictionary conatining the MAP estimates for the OU process parameters
             :param patristic_matrix: full patristic matrix"""
-            sigma_f = map_estimates["sigma_f"]
-            sigma_n = map_estimates["sigma_n"]
-            lambd = map_estimates["lambd"]
+            sigma_f = DraupnirUtils.squeeze_tensor(1, map_estimates["sigma_f"])
+            sigma_n = DraupnirUtils.squeeze_tensor(1, map_estimates["sigma_n"])
+            lambd = DraupnirUtils.squeeze_tensor(1, map_estimates["lambd"])
             internal_indexes = (patristic_matrix[1:, 0][..., None] == self.internal_nodes).any(-1)
 
             #n_internal = family_data_test.shape[0]
             # Highlight: Sample the ancestors conditiones on the leaves (by using the full patristic matrix). See Page 689 at Patter Recongnition and Ml (Bishop)
             # Highlight: Formula is: p(xa|xb) = N (x|µa|b, Λ−1aa ) , a = test/internal; b= train/leaves
             patristic_matrix_full = patristic_matrix[1:, 1:]
             assert patristic_matrix_full.shape == (self.n_all, self.n_all), "Remember to use the entire/full patristic matrix for conditional sampling!"
             OU = OUKernel_Fast(sigma_f, sigma_n, lambd)
             OU_covariance_full = OU.forward(patristic_matrix_full)
             # Highlight: Calculate the inverse of the covariance matrix Λ ≡ Σ−1
             inverse_full = torch.linalg.inv(OU_covariance_full)  # [z_dim,n_test+n_train,n_test+n_train]
-            assert inverse_full.shape == (self.z_dim, self.n_all, self.n_all)
+            assert inverse_full.shape == (self.z_dim, self.n_all, self.n_all), f"Expected dimensions : {(self.z_dim, self.n_all, self.n_all)}, got {inverse_full.shape}"
             # Highlight: B.49 Λ−1aa
             inverse_internal = inverse_full[:, internal_indexes, :]
             inverse_internal = inverse_internal[:, :, internal_indexes]  # [z_dim,n_test,n_test]
             assert inverse_internal.shape == (self.z_dim, self.n_internal, self.n_internal)
             # Highlight: Conditional mean Mean ---->B-50:  µa|b = µa − Λ−1aa Λab(xb − µb)
             # Highlight: µa
             OU_mean_internal = torch.zeros((self.n_internal,))  # [n_internal,]
@@ -225,17 +236,17 @@
             assert OU_mean.squeeze(-1).shape == (self.z_dim, self.n_internal)
             latent_space = dist.MultivariateNormal(OU_mean.squeeze(-1), torch.linalg.inv(inverse_internal)).to_event(1).sample()
             latent_space = latent_space.T
             assert latent_space.shape == (self.n_internal, self.z_dim)
             return OU_mean.squeeze(-1).T
     def conditional_sampling_batch(self,map_estimates, patristic_matrix):
             """Conditional sampling from Multivariate Normal according to page 698 at Pattern Recognition and ML (Bishop)"""
-            sigma_f = map_estimates["sigma_f"] + 1e-6
-            sigma_n = map_estimates["sigma_n"] + 1e-6
-            lambd = map_estimates["lambd"] + 1e-6
+            sigma_f = DraupnirUtils.squeeze_tensor(1, map_estimates["sigma_f"]) + 1e-6
+            sigma_n = DraupnirUtils.squeeze_tensor(1, map_estimates["sigma_n"])  + 1e-6
+            lambd = DraupnirUtils.squeeze_tensor(1, map_estimates["lambd"]) + 1e-6
             internal_indexes = (patristic_matrix[1:, 0][..., None] == self.internal_nodes_batch).any(-1)
             #n_internal = family_data_test.shape[0]
             # Highlight: Sample the ancestors conditiones on the leaves (by using the full patristic matrix). See Page 689 at Patter Recongnition and Ml (Bishop)
             # Highlight: Formula is: p(xa|xb) = N (x|µa|b, Λ−1aa ) , a = test/internal; b= train/leaves
             patristic_matrix_batch = patristic_matrix[1:, 1:]
             assert patristic_matrix_batch.shape == (self.n_leaves_internal_batch, self.n_leaves_internal_batch), "Here we are using a slice of the patristic matrix with size n_leaves_batch = batch_size!"
             OU = OUKernel_Fast(sigma_f, sigma_n, lambd)
@@ -269,46 +280,50 @@
             latent_space = dist.MultivariateNormal(OU_mean.squeeze(-1), torch.linalg.inv(Inverse_internal)).to_event(1).sample()
             #latent_space = dist.MultivariateNormal(OU_mean.squeeze(-1), torch.cholesky_inverse(Inverse_internal) + 1e-6).to_event(1).sample()
 
             latent_space = latent_space.T
             assert latent_space.shape == (self.n_internal_batch, self.z_dim)
             return latent_space
 
-class DRAUPNIRModel_classic(DRAUPNIRModelClass):
-    """Implements the ordinary version of Draupnir. It receives as an input the entire leaves dataset, uses a GRU as the mapping function and blosum embeddings"""
+class DRAUPNIRModel_classic_test(DRAUPNIRModelClass):
+    """Implements the ordinary version of Draupnir as described in the paper. It receives as an input the entire leaves dataset,
+     uses a GRU as the mapping function and blosum embeddings"""
     def __init__(self,ModelLoad):
         DRAUPNIRModelClass.__init__(self,ModelLoad)
         self.rnn_input_size = self.z_dim + self.aa_probs
         self.num_layers = 1
-        self.decoder = RNNDecoder_Tiling(self.align_seq_len, self.aa_probs, self.gru_hidden_dim, self.z_dim, self.rnn_input_size,self.kappa_addition,self.num_layers,self.pretrained_params)
+        self.decoder = RNNDecoder_Tiling_new(self.align_seq_len, self.aa_probs, self.gru_hidden_dim, self.z_dim,
+                                         self.rnn_input_size,self.kappa_addition,self.num_layers,self.pretrained_params)
         self.embed = EmbedComplex(self.aa_probs,self.embedding_dim, self.pretrained_params)
-    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum=None):
+    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum=None,guide_map_estimates=None):
         aminoacid_sequences = family_data[:, 2:, 0]
         #batch_nodes = family_data[:, 0, 1]
         #batch_indexes = (patristic_matrix_sorted[1:, 0][..., None] == batch_nodes).any(-1)
         # Highlight: Register GRU module
         pyro.module("embeddings",self.embed)
         pyro.module("decoder", self.decoder)
-        # Highlight: GP prior over the latent space
-        latent_space = self.gp_prior(patristic_matrix_sorted)
-
-        # Highlight: MAP the latent space to logits using the Decoder from a Seq2seq model with/without attention
-        latent_space = latent_space.repeat(1,self.align_seq_len).reshape(latent_space.shape[0],self.align_seq_len,self.z_dim) #[n_nodes,max_seq,z_dim] #This can maybe be done with new axis solely
-        blosum = self.blosum_weighted.repeat(latent_space.shape[0],1).reshape(latent_space.shape[0],self.align_seq_len,self.aa_probs) #[n_nodes,max_seq,21]
-        blosum = self.embed(blosum)
-        latent_space = torch.cat((latent_space,blosum),dim=2) #[n_nodes,align_seq_len,z_dim + 21]
-        decoder_hidden = self.h_0_MODEL.expand(self.decoder.num_layers * 2, latent_space.shape[0],
-                                               self.gru_hidden_dim).contiguous()  # bidirectional
 
-        with pyro.plate("plate_len", aminoacid_sequences.shape[1], dim=-1):
-            with pyro.plate("plate_seq", aminoacid_sequences.shape[0], dim=-2):
-                logits = self.decoder.forward(
-                    input=latent_space,
-                    hidden=decoder_hidden)
-                pyro.sample("aa_sequences", dist.Categorical(logits=logits),obs=aminoacid_sequences)  # aa_seq = [n_nodes,align_seq_len]
+        with pyro.plate("plate_batch", dim=-1, device=self.device):
+            # Highlight: GP prior over the latent space
+            latent_space = self.gp_prior(patristic_matrix_sorted)
+            # Highlight: MAP the latent space to logits using the Decoder from a Seq2seq model with/without attention
+            latent_space = latent_space.repeat(1,self.align_seq_len).reshape(latent_space.shape[0],self.align_seq_len,self.z_dim) #[n_nodes,max_seq,z_dim] #This can maybe be done with new axis solely
+            blosum = self.blosum_weighted.repeat(latent_space.shape[0],1).reshape(latent_space.shape[0],self.align_seq_len,self.aa_probs) #[n_nodes,max_seq,21]
+            blosum = self.embed(blosum)
+            latent_space = torch.cat((latent_space,blosum),dim=2) #[n_nodes,align_seq_len,z_dim + 21]
+            decoder_hidden = self.h_0_MODEL.expand(self.decoder.num_layers * 2, latent_space.shape[0],
+                                                       self.gru_hidden_dim).contiguous()  # bidirectional
+
+            # if guide_map_estimates is not None:
+            #     decoder_hidden = guide_map_estimates["rnn_final_bidirectional"]
+            with pyro.plate("plate_len", dim=-2):#with pyro.plate("plate_len", dim=-2, device=self.device):
+                    logits = self.decoder.forward(
+                        input=latent_space,
+                        hidden=decoder_hidden)
+                    pyro.sample("aa_sequences", dist.Categorical(logits=logits),obs=aminoacid_sequences)  # aa_seq = [n_nodes,align_seq_len]
 
     def sample(self, map_estimates, n_samples, family_data_test, patristic_matrix,cladistic_matrix,use_argmax=False,use_test=True,use_test2=False):
         if use_test2: #MAP estimate
             assert patristic_matrix[1:,1:].shape == (self.n_all,self.n_all)
             latent_space = self.conditional_samplingMAP(map_estimates,patristic_matrix)
             n_nodes = self.n_internal #I had to split it up because of some weird data cases (coral), otherwise family_data_test.shape[0] would have sufficed
         elif use_test:# Marginal posterior
@@ -351,33 +366,32 @@
 class DRAUPNIRModel_classic_no_blosum(DRAUPNIRModelClass):
     """Implements the ordinary version of Draupnir without blosum embeddings.
     It receives as an input the entire leaf dataset, uses a GRU as the mapping function WITHOUT blosum embeddings"""
     def __init__(self,ModelLoad):
         DRAUPNIRModelClass.__init__(self,ModelLoad)
         self.rnn_input_size = self.z_dim
         self.decoder = RNNDecoder_Tiling(self.align_seq_len, self.aa_probs, self.gru_hidden_dim, self.z_dim, self.rnn_input_size,self.kappa_addition,self.num_layers,self.pretrained_params)
-    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum = None):
+    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum = None,guide_map_estimates=None):
         aminoacid_sequences = family_data[:, 2:, 0]
         batch_nodes = family_data[:, 0, 1]
         # Highlight: Register GRU module
         pyro.module("decoder", self.decoder)
-        # Highlight: GP prior over the latent space
-        latent_space = self.gp_prior(patristic_matrix_sorted)
-        # Highlight: MAP the latent space to logits using the Decoder from a Seq2seq model with/without attention
-        latent_space = latent_space.repeat(1,self.align_seq_len).reshape(latent_space.shape[0],self.align_seq_len,self.z_dim) #[n_nodes,max_seq,z_dim]
-        decoder_hidden = self.h_0_MODEL.expand(self.decoder.num_layers * 2, latent_space.shape[0],
-                                               self.gru_hidden_dim).contiguous()  # bidirectional
-
-        with pyro.plate("plate_len", aminoacid_sequences.shape[1], dim=-1):
-            with pyro.plate("plate_seq", aminoacid_sequences.shape[0], dim=-2) :
-                logits = self.decoder.forward(
-                    input=latent_space,
-                    hidden=decoder_hidden)
-                pyro.sample("aa_sequences", dist.Categorical(logits=logits),
-                            obs=aminoacid_sequences)  # aa_seq = [n_nodes,align_seq_len]
+        with pyro.plate("plate_batch", dim=-1, device=self.device):
+            # Highlight: GP prior over the latent space
+            latent_space = self.gp_prior(patristic_matrix_sorted)
+            # Highlight: MAP the latent space to logits using the Decoder from a Seq2seq model with/without attention
+            latent_space = latent_space.repeat(1,self.align_seq_len).reshape(latent_space.shape[0],self.align_seq_len,self.z_dim) #[n_nodes,max_seq,z_dim]
+            decoder_hidden = self.h_0_MODEL.expand(self.decoder.num_layers * 2, latent_space.shape[0],
+                                                   self.gru_hidden_dim).contiguous()  # bidirectional
+
+            with pyro.plate("plate_len", aminoacid_sequences.shape[1], dim=-2):
+                    logits = self.decoder.forward(
+                        input=latent_space,
+                        hidden=decoder_hidden)
+                    pyro.sample("aa_sequences", dist.Categorical(logits=logits),obs=aminoacid_sequences)  # aa_seq = [n_nodes,align_seq_len]
 
     def sample(self, map_estimates, n_samples, family_data_test, patristic_matrix,cladistic_matrix,use_argmax=False,use_test=True,use_test2=False):
         if use_test or use_test2:
             assert patristic_matrix[1:,1:].shape == (self.n_all,self.n_all)
             latent_space = self.conditional_sampling(map_estimates,patristic_matrix)
             n_nodes = self.n_internal #I had to split it up because of some weird data cases (coral), otherwise family_data_test.shape[0] would have sufficed
         else:
@@ -423,15 +437,15 @@
         self.decoder = RNNDecoder_Tiling(self.align_seq_len, self.aa_probs, self.gru_hidden_dim, self.z_dim, self.rnn_input_size,self.kappa_addition,self.num_layers,self.pretrained_params)
         self.embed = EmbedComplex(self.aa_probs,self.embedding_dim, self.pretrained_params)
         self.splitted_leaves_indexes = list(torch.tensor_split(torch.arange(self.n_leaves), int(self.n_leaves / self.plate_size)) * self.num_epochs)
         if self.plate_unordered:
             self.model = self.model_unordered
         else:
             self.model = self.model_ordered
-    def model_ordered(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum = None):
+    def model_ordered(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum = None,guide_map_estimates=None):
         aminoacid_sequences = family_data[:, 2:, 0]
         batch_nodes = family_data[:, 0, 1]
         #batch_indexes = (patristic_matrix_sorted[1:, 0][..., None] == batch_nodes).any(-1)
         # Highlight: Register GRU module
         pyro.module("embeddings",self.embed)
         pyro.module("decoder", self.decoder)
         # Highlight: GP prior over the latent space
@@ -439,24 +453,24 @@
         # Highlight: MAP the latent space to logits using the Decoder from a Seq2seq model with/without attention
         latent_space = latent_space.repeat(1,self.align_seq_len).reshape(latent_space.shape[0],self.align_seq_len,self.z_dim) #[n_nodes,max_seq,z_dim]
         blosum = self.blosum_weighted.repeat(latent_space.shape[0],1).reshape(latent_space.shape[0],self.align_seq_len,self.aa_probs) #[n_nodes,max_seq,21]
         blosum = self.embed(blosum) #TODO: Introduce a noise variable to be able to deal with more random mutations?
         latent_space = torch.cat((latent_space,blosum),dim=2) #[n_nodes,align_seq_len,z_dim + 21]
 
         decoder_hidden = self.h_0_MODEL.expand(self.decoder.num_layers * 2, latent_space.shape[0],
-                                               self.gru_hidden_dim).contiguous()  #TODO: Should the subsamples are sharing the same initial hidden state?
+                                               self.gru_hidden_dim).contiguous()
 
         with pyro.plate("plate_len", aminoacid_sequences.shape[1], dim=-1):
             with pyro.plate("plate_seq", aminoacid_sequences.shape[0], dim=-2, subsample=self.splitted_leaves_indexes.pop(0)) as indx:  # Highlight: Ordered subsampling
                 logits = self.decoder.forward(
                     input=latent_space[indx],
                     hidden=decoder_hidden[:,indx])
                 pyro.sample("aa_sequences", dist.Categorical(logits=logits),obs=aminoacid_sequences[indx])  # aa_seq = [n_nodes,align_seq_len]
 
-    def model_unordered(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum = None):
+    def model_unordered(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum = None,guide_map_estimates=None):
         aminoacid_sequences = family_data[:, 2:, 0]
         batch_nodes = family_data[:, 0, 1]
         #batch_indexes = (patristic_matrix_sorted[1:, 0][..., None] == batch_nodes).any(-1)
         # Highlight: Register GRU module
         pyro.module("embeddings",self.embed)
         pyro.module("decoder", self.decoder)
         # Highlight: GP prior over the latent space
@@ -464,15 +478,15 @@
         # Highlight: MAP the latent space to logits using the Decoder from a Seq2seq model with/without attention
         latent_space = latent_space.repeat(1,self.align_seq_len).reshape(latent_space.shape[0],self.align_seq_len,self.z_dim) #[n_nodes,max_seq,z_dim]
         blosum = self.blosum_weighted.repeat(latent_space.shape[0],1).reshape(latent_space.shape[0],self.align_seq_len,self.aa_probs) #[n_nodes,max_seq,21]
         blosum = self.embed(blosum) #TODO: Introduce a noise variable to be able to deal with more random mutations?
         latent_space = torch.cat((latent_space,blosum),dim=2) #[n_nodes,align_seq_len,z_dim + 21]
 
         decoder_hidden = self.h_0_MODEL.expand(self.decoder.num_layers * 2, latent_space.shape[0],
-                                               self.gru_hidden_dim).contiguous()  #TODO: Should the subsamples are sharing the same initial hidden state?
+                                               self.gru_hidden_dim).contiguous()
 
         with pyro.plate("plate_len", aminoacid_sequences.shape[1], dim=-1):
             with pyro.plate("plate_seq",aminoacid_sequences.shape[0],dim=-2,subsample_size=self.plate_size) as indx:#Highlight: Random subsampling
             #with pyro.plate("plate_seq", aminoacid_sequences.shape[0], dim=-2,subsample=self.splitted_leaves_indexes.pop(0)) as indx:  # Highlight: Ordered subsampling
                 logits = self.decoder.forward(
                     input=latent_space[indx],
                     hidden=decoder_hidden[:,indx])
@@ -523,15 +537,15 @@
     def __init__(self,ModelLoad):
         DRAUPNIRModelClass.__init__(self,ModelLoad)
         self.rnn_input_size = self.z_dim + self.aa_probs
         self.decoder = RNNDecoder_Tiling(self.align_seq_len, self.aa_probs, self.gru_hidden_dim, self.z_dim, self.rnn_input_size,self.kappa_addition,self.num_layers,self.pretrained_params)
         self.embed = EmbedComplex(self.aa_probs,self.embedding_dim, self.pretrained_params)
         self.internal_nodes_batch = None
         self.n_leaves_internal_batch = None
-    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum):
+    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum,guide_map_estimates=None):
         aminoacid_sequences = family_data[:, 2:, 0]
         batch_nodes = family_data[:, 0, 1]
         batch_indexes = (patristic_matrix_sorted[1:, 0][..., None] == batch_nodes).any(-1)
         # Highlight: Register GRU module
         pyro.module("embeddings",self.embed)
         pyro.module("decoder", self.decoder)
         # Highlight: GP prior over the latent space
@@ -650,15 +664,15 @@
     Clade batch training and sampling."""
     def __init__(self,ModelLoad):
         DRAUPNIRModelClass.__init__(self,ModelLoad)
         self.rnn_input_size = self.z_dim + self.aa_probs
         #self.decoder_attention = RNNAttentionDecoder(self.n_leaves, self.max_seq_len, self.aa_prob, self.gru_hidden_dim,self.rnn_input_size,self.embedding_dim, self.z_dim, self.kappa_addition)
         self.decoder = RNNDecoder_Tiling(self.align_seq_len, self.aa_probs, self.gru_hidden_dim, self.z_dim, self.rnn_input_size,self.kappa_addition,self.num_layers,self.pretrained_params)
         self.embed = EmbedComplex(self.aa_probs,self.embedding_dim, self.pretrained_params)
-    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum):
+    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum,guide_map_estimates=None):
         aminoacid_sequences = family_data[:, 2:, 0]
         #batch_nodes = family_data[:, 0, 1]
         #batch_indexes = (patristic_matrix_sorted[1:, 0][..., None] == batch_nodes).any(-1)
         # Highlight: Register GRU module
         pyro.module("embeddings",self.embed)
         pyro.module("decoder", self.decoder)
         self.n_leaves_batch = family_data.shape[0]
@@ -777,15 +791,15 @@
     """Leaves training and testing. Train on full leave latent space (train + test), only observe the pre-selected train leaves"""
     def __init__(self,ModelLoad):
         DRAUPNIRModelClass.__init__(self,ModelLoad)
         self.rnn_input_size = self.z_dim + self.aa_probs
         #self.decoder_attention = RNNAttentionDecoder(self.n_leaves, self.align_seq_len, self.aa_probs, self.gru_hidden_dim,self.rnn_input_size,self.embedding_dim, self.z_dim, self.kappa_addition)
         self.decoder = RNNDecoder_Tiling(self.align_seq_len, self.aa_probs, self.gru_hidden_dim, self.z_dim, self.rnn_input_size,self.kappa_addition,self.num_layers,self.pretrained_params)
         self.embed = EmbedComplex(self.aa_probs,self.embedding_dim, self.pretrained_params)
-    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum=None):
+    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum=None,guide_map_estimates=None):
         aminoacid_sequences = family_data[:, 2:, 0]
         #angles = family_data[:, 2:, 1:3]
         train_nodes = family_data[:, 0, 1]
         train_indexes = (patristic_matrix_sorted[1:, 0][..., None] == train_nodes).any(-1)
         # Highlight: Register GRU module
         pyro.module("embeddings",self.embed)
         pyro.module("decoder", self.decoder)
@@ -860,15 +874,15 @@
     """Leaves training and testing.Predicting both ANGLES and AA sequences. Working on full or partial leaves space depending on the
     leaves_testing argument stated in datasets.py"""
     def __init__(self,ModelLoad):
         DRAUPNIRModelClass.__init__(self,ModelLoad)
         self.rnn_input_size = self.z_dim + self.aa_probs
         self.decoder = RNNDecoder_Tiling_Angles(self.align_seq_len, self.aa_probs, self.gru_hidden_dim, self.z_dim, self.rnn_input_size,self.kappa_addition,self.num_layers,self.pretrained_params)
         self.embed = EmbedComplex(self.aa_probs,self.embedding_dim, self.pretrained_params)
-    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum=None):
+    def model(self, family_data, patristic_matrix_sorted,cladistic_matrix,data_blosum,batch_blosum=None,guide_map_estimates=None):
         aminoacid_sequences = family_data[:, 2:, 0]
         angles = family_data[:, 2:, 1:3]
         angles_mask = torch.where(angles == 0., angles, 1.).type(angles.dtype) #keep as 0 the gaps and set to 1 where there is an observation
         train_nodes = family_data[:, 0, 1]
         train_indexes = (patristic_matrix_sorted[1:, 0][..., None] == train_nodes).any(-1)
         # Highlight: Register GRU module
         pyro.module("embeddings",self.embed)
```

### Comparing `draupnir-0.0.28/src/draupnir/models_utils.py` & `draupnir-0.0.29/src/draupnir/models_utils.py`

 * *Files 9% similar despite different names*

```diff
@@ -30,103 +30,113 @@
         self.aa_prob = aa_prob
         self.num_layers = num_layers
         self.kappa_addition = kappa_addition
         self.softmax = nn.Softmax()
         self.logsoftmax = nn.LogSoftmax(dim=-1)
         self.relu = nn.ReLU()
         self.tanh = nn.Tanh()
-        self.fc1 = nn.Linear(2 * self.gru_hidden_dim, self.gru_hidden_dim)
-        self.linear_means = nn.Linear(self.gru_hidden_dim, self.z_dim)
-        self.linear_std = nn.Linear(self.gru_hidden_dim, self.z_dim)
+        self.nndim = int(self.gru_hidden_dim/2)
+        self.fc1 = nn.Linear(self.gru_hidden_dim, self.nndim)
+        self.linear_means = nn.Linear(self.nndim, self.z_dim)
+        self.linear_std = nn.Linear(self.nndim, self.z_dim)
 
         self.softplus = nn.Softplus()
 
         self.rnn = nn.GRU(input_size=self.rnn_input_size,
                           hidden_size=self.gru_hidden_dim,
                           batch_first=True,
                           bidirectional=True,
                           num_layers=self.num_layers,
                           dropout=0.0)
 
 
     def forward(self, input, hidden):
 
-        rnn_output, rnn_hidden = self.rnn(input, hidden)  # [n_nodes,align_seq_len,gru_dim] | [1,n_nodes,gru_dim]
-        rnn_output = self.fc1(rnn_output[:,-1]) #pick the last state of the sequence given by the GRU
-        output_means = self.linear_means(rnn_output)
-        output_std = self.softplus(self.linear_std(rnn_output))
-        return output_means,output_std
-class RNNEncoder_no_mean(nn.Module):
-    def __init__(self, align_seq_len,aa_prob,n_leaves,gru_hidden_dim, z_dim,rnn_input_size, kappa_addition,num_layers,pretrained_params):
-        super(RNNEncoder_no_mean, self).__init__()
+        rnn_hidden_states, rnn_final_bidirectional = self.rnn(input, hidden)  # [n_nodes,align_seq_len,gru_dim] | [1,n_nodes,gru_dim]
+        forward_out_r,backward_out_r = rnn_hidden_states[:,:,:self.gru_hidden_dim],rnn_hidden_states[:,:,self.gru_hidden_dim:]
+        rnn_hidden_states = forward_out_r + backward_out_r
+        rnn_final_hidden_state = self.fc1(rnn_hidden_states[:,-1]) #pick the last state of the sequence given by the GRU
+        z_loc = self.linear_means(rnn_final_hidden_state)
+        z_scale = self.softplus(self.linear_std(rnn_final_hidden_state))
+
+        return  {"z_loc":z_loc,
+                 "z_scale":z_scale,
+                 "rnn_final_bidirectional":rnn_final_bidirectional,
+                 "rnn_hidden_states":rnn_hidden_states,
+                 "rnn_final_hidden_state":rnn_final_hidden_state}
+        #return output_means,output_std
+class RNNDecoder_Tiling(nn.Module):
+    def __init__(self, align_seq_len,aa_prob,gru_hidden_dim, z_dim,rnn_input_size, kappa_addition,num_layers,pretrained_params):
+        super(RNNDecoder_Tiling, self).__init__()
         self.gru_hidden_dim = gru_hidden_dim
         self.z_dim = z_dim
-        self.n_leaves = n_leaves
         self.rnn_input_size = rnn_input_size
         self.align_seq_len = align_seq_len
         self.aa_prob = aa_prob
         self.num_layers = num_layers
         self.kappa_addition = kappa_addition
         self.softmax = nn.Softmax()
         self.logsoftmax = nn.LogSoftmax(dim=-1)
         self.relu = nn.ReLU()
         self.tanh = nn.Tanh()
-        self.fc1 = nn.Linear(2 * self.gru_hidden_dim, self.gru_hidden_dim)
-        #self.linear_means = nn.Linear(self.gru_hidden_dim, self.z_dim)
-        self.linear_std = nn.Linear(self.gru_hidden_dim, self.z_dim)
-
-        self.softplus = nn.Softplus()
 
+        self.fc1 = nn.Linear(2 * self.gru_hidden_dim, self.gru_hidden_dim)
+        self.linear_probs = nn.Linear(self.gru_hidden_dim, self.aa_prob)
         self.rnn = nn.GRU(input_size=self.rnn_input_size,
                           hidden_size=self.gru_hidden_dim,
                           batch_first=True,
                           bidirectional=True,
                           num_layers=self.num_layers,
                           dropout=0.0)
+            #rnn.state_dict()
 
 
     def forward(self, input, hidden):
-
         rnn_output, rnn_hidden = self.rnn(input, hidden)  # [n_nodes,align_seq_len,gru_dim] | [1,n_nodes,gru_dim]
-        rnn_output = self.fc1(rnn_output[:,-1]) #pick the last state of the sequence given by the GRU
-        #output_means = self.linear_means(rnn_output)
-        output_std = self.softplus(self.linear_std(rnn_output))
-        return output_std
-class RNNDecoder_Tiling(nn.Module):
+        #forward_out = rnn_output[:, :, :self.gru_hidden_dim]
+        #backward_out = rnn_output[:, :, self.gru_hidden_dim:]
+        #rnn_output_out = torch.cat((forward_out, backward_out), dim=2)
+        output_logits = self.logsoftmax(self.linear_probs(self.fc1(rnn_output)))  # [n_nodes,align_seq_len,aa_probs]
+        return output_logits
+class RNNDecoder_Tiling_new(nn.Module):
     def __init__(self, align_seq_len,aa_prob,gru_hidden_dim, z_dim,rnn_input_size, kappa_addition,num_layers,pretrained_params):
-        super(RNNDecoder_Tiling, self).__init__()
+        super(RNNDecoder_Tiling_new, self).__init__()
         self.gru_hidden_dim = gru_hidden_dim
         self.z_dim = z_dim
         self.rnn_input_size = rnn_input_size
         self.align_seq_len = align_seq_len
         self.aa_prob = aa_prob
         self.num_layers = num_layers
         self.kappa_addition = kappa_addition
         self.softmax = nn.Softmax()
         self.logsoftmax = nn.LogSoftmax(dim=-1)
         self.relu = nn.ReLU()
         self.tanh = nn.Tanh()
 
-        self.fc1 = nn.Linear(2 * self.gru_hidden_dim, self.gru_hidden_dim)
-        self.linear_probs = nn.Linear(self.gru_hidden_dim, self.aa_prob)
+        self.fc1 = nn.Linear(self.gru_hidden_dim, int(self.gru_hidden_dim/2))
+        self.linear_probs = nn.Linear(int(self.gru_hidden_dim/2), self.aa_prob)
         self.rnn = nn.GRU(input_size=self.rnn_input_size,
                           hidden_size=self.gru_hidden_dim,
                           batch_first=True,
                           bidirectional=True,
                           num_layers=self.num_layers,
                           dropout=0.0)
             #rnn.state_dict()
 
 
     def forward(self, input, hidden):
-        rnn_output, rnn_hidden = self.rnn(input, hidden)  # [n_nodes,align_seq_len,gru_dim] | [1,n_nodes,gru_dim]
-        #forward_out = rnn_output[:, :, :self.gru_hidden_dim]
-        #backward_out = rnn_output[:, :, self.gru_hidden_dim:]
-        #rnn_output_out = torch.cat((forward_out, backward_out), dim=2)
-        output_logits = self.logsoftmax(self.linear_probs(self.fc1(rnn_output)))  # [n_nodes,align_seq_len,aa_probs]
+        rnn_hidden_states, rnn_final_hidden = self.rnn(input, hidden)  # [n_nodes,align_seq_len,gru_dim] | [1,n_nodes,gru_dim]
+        forward_hidden_states = rnn_hidden_states[:, :, :self.gru_hidden_dim]
+        backward_hidden_states = rnn_hidden_states[:, :, self.gru_hidden_dim:]
+        rnn_hidden_states = forward_hidden_states + backward_hidden_states
+        #rnn_final_hidden = rnn_hidden_states[:,-1]
+
+        rnn_final_hidden = self.fc1(rnn_hidden_states)
+        output_logits = self.logsoftmax(self.linear_probs(rnn_final_hidden))  # [n_nodes,align_seq_len,aa_probs]
+        #output_logits = self.logsoftmax(self.linear_probs(self.fc1(rnn_final_hidden)))  # [n_nodes,align_seq_len,aa_probs]
         return output_logits
 class RNNDecoder_Tiling_Angles(nn.Module):
     def __init__(self, align_seq_len,aa_prob,gru_hidden_dim, z_dim,rnn_input_size, kappa_addition,num_layers,pretrained_params):
         super(RNNDecoder_Tiling_Angles, self).__init__()
         self.gru_hidden_dim = gru_hidden_dim
         self.z_dim = z_dim
         self.rnn_input_size = rnn_input_size
@@ -199,15 +209,14 @@
         #backward_out = rnn_output[:, :, self.gru_hidden_dim:]
         #rnn_output_out = torch.cat((forward_out, backward_out), dim=2)
         output = self.fc1(rnn_output)
         output_logits = self.logsoftmax(self.fc2_probs(self.fc1_probs(output)))  # [n_nodes,align_seq_len,aa_probs]
         output_means = self.tanh(self.fc2_means(self.fc1_means(output)))*math.pi
         output_kappas = self.kappa_addition + self.softplus(self.fc2_kappas(self.fc1_kappas(output)))
         return output_logits,output_means,output_kappas
-
 class Embed(nn.Module):
     def __init__(self,aa_probs,embedding_dim,pretrained_params):
         super(Embed, self).__init__()
         self.aa_probs = aa_probs
         self.embedding_dim = embedding_dim
         self.fc1 = nn.Linear(self.aa_probs,self.aa_probs)
         self.softmax = nn.Softmax(dim=-1)
@@ -222,30 +231,27 @@
         self.softmax = nn.Softmax(dim=-1)
         self.fc1 = nn.Linear(self.aa_probs,self.embedding_dim)
         self.fc2 = nn.Linear(self.embedding_dim,self.aa_probs)
     def forward(self,input):
         output = self.fc1(input) #.type(torch.cuda.IntTensor)
         output = self.softmax(self.fc2(output))
         return output
-
 class EmbedComplexEncoder(nn.Module):
     def __init__(self,aa_probs,embedding_dim,pretrained_params):
         super(EmbedComplexEncoder, self).__init__()
         self.aa_probs = aa_probs
         self.embedding_dim = embedding_dim
         self.softmax = nn.Softmax(dim=-1)
         self.fc1 = nn.Linear(self.aa_probs,self.embedding_dim)
         self.fc2 = nn.Linear(self.embedding_dim,self.aa_probs)
 
     def forward(self,input):
         output = self.fc1(input) #.type(torch.cuda.IntTensor)
         output = self.softmax(self.fc2(output))
         return output
-
-
 class GPKernel(ABC):
     @abstractmethod
     def preforward(self, t1: torch.Tensor,t2: torch.Tensor) -> torch.Tensor:
         raise NotImplementedError
     @abstractmethod
     def forward(self, t: torch.Tensor) -> torch.Tensor:
         raise NotImplementedError
@@ -273,15 +279,14 @@
     def preforward(self, t1: torch.Tensor, t2: torch.Tensor) -> torch.Tensor:
         raise NotImplementedError
 
     def forward(self, t: torch.Tensor) -> torch.Tensor:
         first_term = self.sigma_f ** 2
         second_term = torch.exp(-t / self.lamb)
         return first_term * second_term + self.sigma_n ** 2 * torch.eye(t.shape[0])
-
 class OUKernel_Fast(GPKernel):
     """ Kernel that computes the covariance matrix for a z Ornstein Ulenbeck processes. As stated in Equation 2.1 https://arxiv.org/pdf/1208.0628.pdf
     :param tensor sigma_f: Quantifies the intensity of inherited variation ---> Signal variance
     :param tensor lamb: Characteristic length-scale of the evolutionary dynamics (equivalent to the inverse of the strength of selection)---> Distance between data points (nodes),larger l implies that the noise should be bigger to capture big point fluctuations
     :param tensor sigma_n:quantifies the intensity of specific variation(i.e. variation unattributable to the phylogeny)--->Gaussian Noise,intensity of specific variation--> how much to let the sequence vary ---> so max branch lengh?
     **References:**
     "Ancestral Inference from Functional Data: Statistical Methods and Numerical Examples"
```

### Comparing `draupnir-0.0.28/src/draupnir/mutual_information.py` & `draupnir-0.0.29/src/draupnir/mutual_information.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/src/draupnir/plots.py` & `draupnir-0.0.29/src/draupnir/plots.py`

 * *Files 0% similar despite different names*

```diff
@@ -289,15 +289,15 @@
 
     aa_sequences_predictions = aa_sequences_predictions.cpu().numpy()
     dataset_test = dataset_test.cpu()
     n_test = dataset_test.shape[0]
     if n_test > 100:
         torch.manual_seed(0)
         percentage_test = 50/n_test #get 50 seq only
-        test_indx = (torch.rand(size=(n_test,)) < percentage_test).int().bool() #create n  (n= n internal nodes) random True and False, Supposedly 45% of the values will be 1(True)
+        test_indx = (torch.rand(size=(n_test,)) < percentage_test).int().bool().to(dataset_test.device) #create n  (n= n internal nodes) random True and False, Supposedly 45% of the values will be 1(True)
         dataset_test = dataset_test[test_indx]
         nodes_indexes = dataset_test[:, 0, 1]
         dataset_children_predicted = aa_sequences_predictions[:,np.isin(aa_sequences_predictions[0, :, 1], nodes_indexes)]
         children_indexes_str = nodes_indexes.numpy().astype(str).tolist()
         incorrectly_predicted_sites_df = incorrectly_predicted_sites_df[incorrectly_predicted_sites_df.index.str.split("//").str[1].isin(children_indexes_str)]
```

### Comparing `draupnir-0.0.28/src/draupnir/utils.py` & `draupnir-0.0.29/src/draupnir/utils.py`

 * *Files 1% similar despite different names*

```diff
@@ -59,6275 +59,6275 @@
 000003a0: 7570 6e69 724d 6f64 656c 5574 696c 730a  upnirModelUtils.
 000003b0: 2342 696f 7079 7468 6f6e 0a69 6d70 6f72  #Biopython.impor
 000003c0: 7420 4269 6f2e 5044 4220 6173 2050 4442  t Bio.PDB as PDB
 000003d0: 0a66 726f 6d20 4269 6f2e 5044 422e 506f  .from Bio.PDB.Po
 000003e0: 6c79 7065 7074 6964 6520 696d 706f 7274  lypeptide import
 000003f0: 2050 5042 7569 6c64 6572 2c20 4361 5050   PPBuilder, CaPP
 00000400: 4275 696c 6465 720a 6672 6f6d 2042 696f  Builder.from Bio
-00000410: 2e44 6174 612e 5343 4f50 4461 7461 2069  .Data.SCOPData i
-00000420: 6d70 6f72 7420 7072 6f74 6569 6e5f 6c65  mport protein_le
-00000430: 7474 6572 735f 3374 6f31 0a66 726f 6d20  tters_3to1.from 
-00000440: 4269 6f2e 5365 7152 6563 6f72 6420 696d  Bio.SeqRecord im
-00000450: 706f 7274 2053 6571 5265 636f 7264 0a66  port SeqRecord.f
-00000460: 726f 6d20 4269 6f2e 5365 7120 696d 706f  rom Bio.Seq impo
-00000470: 7274 2053 6571 0a66 726f 6d20 4269 6f20  rt Seq.from Bio 
-00000480: 696d 706f 7274 2042 696f 7079 7468 6f6e  import Biopython
-00000490: 5761 726e 696e 670a 6672 6f6d 2042 696f  Warning.from Bio
-000004a0: 2069 6d70 6f72 7420 416c 6967 6e49 4f2c   import AlignIO,
-000004b0: 2053 6571 494f 0a66 726f 6d20 4269 6f2e   SeqIO.from Bio.
-000004c0: 5044 422e 5044 424c 6973 7420 696d 706f  PDB.PDBList impo
-000004d0: 7274 2050 4442 4c69 7374 0a69 6d70 6f72  rt PDBList.impor
-000004e0: 7420 4269 6f2e 416c 6967 6e0a 6672 6f6d  t Bio.Align.from
-000004f0: 2042 696f 2e41 6c69 676e 2e41 7070 6c69   Bio.Align.Appli
-00000500: 6361 7469 6f6e 7320 696d 706f 7274 204d  cations import M
-00000510: 6166 6674 436f 6d6d 616e 646c 696e 650a  afftCommandline.
-00000520: 6672 6f6d 2042 696f 2e50 6879 6c6f 2e54  from Bio.Phylo.T
-00000530: 7265 6543 6f6e 7374 7275 6374 696f 6e20  reeConstruction 
-00000540: 696d 706f 7274 202a 0a66 726f 6d20 4269  import *.from Bi
-00000550: 6f20 696d 706f 7274 2050 6879 6c6f 0a23  o import Phylo.#
-00000560: 4e75 6d70 790a 696d 706f 7274 206e 756d  Numpy.import num
-00000570: 7079 2061 7320 6e70 0a69 6d70 6f72 7420  py as np.import 
-00000580: 6e75 6d70 792e 7261 6e64 6f6d 2061 7320  numpy.random as 
-00000590: 6e70 720a 696d 706f 7274 2070 616e 6461  npr.import panda
-000005a0: 7320 6173 2070 640a 696d 706f 7274 206d  s as pd.import m
-000005b0: 6174 706c 6f74 6c69 620a 6d61 7470 6c6f  atplotlib.matplo
-000005c0: 746c 6962 2e75 7365 2827 4167 6727 290a  tlib.use('Agg').
-000005d0: 6465 6620 6161 5f70 726f 7065 7274 6965  def aa_propertie
-000005e0: 7328 6161 5f70 726f 6273 2c64 6174 615f  s(aa_probs,data_
-000005f0: 666f 6c64 6572 293a 0a20 2020 2022 2222  folder):.    """
-00000600: 2043 7265 6174 6573 2061 2064 6963 7469   Creates a dicti
-00000610: 6f6e 6172 7920 7769 7468 2061 6d69 6e6f  onary with amino
-00000620: 2061 6369 6420 7072 6f70 6572 7469 6573   acid properties
-00000630: 2c20 6578 7472 6163 7465 6420 6672 6f6d  , extracted from
-00000640: 0a20 2020 2068 7474 7073 3a2f 2f77 7777  .    https://www
-00000650: 2e73 6967 6d61 616c 6472 6963 682e 636f  .sigmaaldrich.co
-00000660: 6d2f 6c69 6665 2d73 6369 656e 6365 2f6d  m/life-science/m
-00000670: 6574 6162 6f6c 6f6d 6963 732f 6c65 6172  etabolomics/lear
-00000680: 6e69 6e67 2d63 656e 7465 722f 616d 696e  ning-center/amin
-00000690: 6f2d 6163 6964 2d72 6566 6572 656e 6365  o-acid-reference
-000006a0: 2d63 6861 7274 2e68 746d 6c0a 2020 2020  -chart.html.    
-000006b0: 3a70 6172 616d 2069 6e74 2061 615f 7072  :param int aa_pr
-000006c0: 6f62 733a 2061 6d69 6e6f 2061 6369 6420  obs: amino acid 
-000006d0: 7072 6f62 6162 696c 6974 6965 7320 7573  probabilities us
-000006e0: 6564 2074 6f20 6578 7472 6163 7420 7468  ed to extract th
-000006f0: 6520 7479 7065 7320 6f66 2061 6d69 6e6f  e types of amino
-00000700: 2061 6369 6473 2070 7265 7365 6e74 2069   acids present i
-00000710: 6e20 7468 6520 6461 7461 7365 740a 2020  n the dataset.  
-00000720: 2020 3a70 6172 616d 2073 7472 2064 6174    :param str dat
-00000730: 615f 666f 6c64 6572 3a20 7061 7468 2074  a_folder: path t
-00000740: 6f20 7768 6572 6520 6472 6175 706e 6972  o where draupnir
-00000750: 2069 7320 6265 656e 2065 7865 6375 7465   is been execute
-00000760: 6420 6672 6f6d 2222 220a 0a20 2020 2061  d from"""..    a
-00000770: 615f 7479 7065 7320 3d20 6c69 7374 2861  a_types = list(a
-00000780: 6d69 6e6f 6163 6964 5f6e 616d 6573 5f64  minoacid_names_d
-00000790: 6963 7428 6161 5f70 726f 6273 292e 6b65  ict(aa_probs).ke
-000007a0: 7973 2829 290a 2020 2020 7374 6f72 6167  ys()).    storag
-000007b0: 655f 666f 6c64 6572 203d 206f 732e 7061  e_folder = os.pa
-000007c0: 7468 2e61 6273 7061 7468 286f 732e 7061  th.abspath(os.pa
-000007d0: 7468 2e6a 6f69 6e28 6f73 2e70 6174 682e  th.join(os.path.
-000007e0: 6469 726e 616d 6528 5f5f 6669 6c65 5f5f  dirname(__file__
-000007f0: 292c 2022 6461 7461 2229 290a 2020 2020  ), "data")).    
-00000800: 6161 5f70 726f 7065 7274 6965 7320 3d20  aa_properties = 
-00000810: 7064 2e72 6561 645f 6373 7628 227b 7d2f  pd.read_csv("{}/
-00000820: 4141 5f70 726f 7065 7274 6965 732e 7478  AA_properties.tx
-00000830: 7422 2e66 6f72 6d61 7428 7374 6f72 6167  t".format(storag
-00000840: 655f 666f 6c64 6572 292c 7365 703d 225c  e_folder),sep="\
-00000850: 732b 2229 0a20 2020 2061 615f 696e 666f  s+").    aa_info
-00000860: 203d 2064 6566 6175 6c74 6469 6374 2829   = defaultdict()
-00000870: 0a20 2020 2066 6f72 2069 6e64 782c 6161  .    for indx,aa
-00000880: 2069 6e20 6161 5f70 726f 7065 7274 6965   in aa_propertie
-00000890: 732e 6974 6572 726f 7773 2829 3a0a 2020  s.iterrows():.  
-000008a0: 2020 2020 2020 6966 2061 615b 2241 6262        if aa["Abb
-000008b0: 722e 225d 2069 6e20 6161 5f74 7970 6573  r."] in aa_types
-000008c0: 3a0a 2020 2020 2020 2020 2020 2020 6161  :.            aa
-000008d0: 5f6e 756d 6265 7220 3d20 6161 5f74 7970  _number = aa_typ
-000008e0: 6573 2e69 6e64 6578 2861 615b 2241 6262  es.index(aa["Abb
-000008f0: 722e 225d 290a 2020 2020 2020 2020 2020  r."]).          
-00000900: 2020 6161 5f69 6e66 6f5b 6161 5f6e 756d    aa_info[aa_num
-00000910: 6265 725d 203d 205b 666c 6f61 7428 6161  ber] = [float(aa
-00000920: 5b22 4d6f 6c65 6375 6c61 725f 5765 6967  ["Molecular_Weig
-00000930: 6874 225d 292c 666c 6f61 7428 6161 5b22  ht"]),float(aa["
-00000940: 704b 6131 225d 292c 666c 6f61 7428 6161  pKa1"]),float(aa
-00000950: 5b22 706c 3422 5d29 5d0a 2020 2020 7265  ["pl4"])].    re
-00000960: 7475 726e 2061 615f 696e 666f 0a0a 6465  turn aa_info..de
-00000970: 6620 7661 6c69 6461 7465 5f73 6571 7565  f validate_seque
-00000980: 6e63 655f 616c 7068 6162 6574 2873 6571  nce_alphabet(seq
-00000990: 293a 0a20 2020 2022 2222 0a20 2020 2043  ):.    """.    C
-000009a0: 6865 636b 7320 7468 6174 2074 6865 2073  hecks that the s
-000009b0: 6571 7565 6e63 6573 2066 726f 6d20 616e  equences from an
-000009c0: 2061 6c69 676e 6d65 6e74 206f 6e6c 7920   alignment only 
-000009d0: 636f 6e74 6169 6e73 2076 616c 7565 7320  contains values 
-000009e0: 6672 6f6d 206f 6e65 206f 6620 7468 6520  from one of the 
-000009f0: 7072 6f74 6569 6e20 616c 7068 6162 6574  protein alphabet
-00000a00: 7320 2870 726f 7465 696e 3231 206f 7220  s (protein21 or 
-00000a10: 7072 6f74 6569 6e32 3170 6c75 7329 2e20  protein21plus). 
-00000a20: 5265 6a65 6374 2044 4e41 206f 7220 524e  Reject DNA or RN
-00000a30: 4120 7365 7175 656e 6365 730a 2020 2020  A sequences.    
-00000a40: 3a70 6172 616d 2073 7472 2073 6571 3a20  :param str seq: 
-00000a50: 7365 7175 656e 6365 206f 6620 6368 6172  sequence of char
-00000a60: 6163 7465 7273 0a20 2020 2022 2222 0a20  acters.    """. 
-00000a70: 2020 2061 6c70 6861 6265 7473 203d 207b     alphabets = {
-00000a80: 2764 6e61 273a 2072 652e 636f 6d70 696c  'dna': re.compil
-00000a90: 6528 275e 5b61 6367 746e 5d2a 2427 2c20  e('^[acgtn]*$', 
-00000aa0: 7265 2e49 292c 0a20 2020 2020 2020 2020  re.I),.         
-00000ab0: 2020 2020 2770 726f 7465 696e 3231 273a      'protein21':
-00000ac0: 2072 652e 636f 6d70 696c 6528 275e 5b2d   re.compile('^[-
-00000ad0: 6163 6465 6667 6869 6b6c 6d6e 7071 7273  acdefghiklmnpqrs
-00000ae0: 7476 7779 5d2a 2427 2c20 666c 6167 733d  tvwy]*$', flags=
-00000af0: 7265 2e49 474e 4f52 4543 4153 4529 2c0a  re.IGNORECASE),.
-00000b00: 2020 2020 2020 2020 2020 2020 2770 726f              'pro
-00000b10: 7465 696e 3231 706c 7573 273a 2072 652e  tein21plus': re.
-00000b20: 636f 6d70 696c 6528 275e 5b2d 6163 6465  compile('^[-acde
-00000b30: 6667 6869 6b6c 6d6e 7071 7273 7476 7779  fghiklmnpqrstvwy
-00000b40: 627a 785d 2a24 272c 2066 6c61 6773 3d20  bzx]*$', flags= 
-00000b50: 7265 2e49 474e 4f52 4543 4153 4529 7d0a  re.IGNORECASE)}.
-00000b60: 0a20 2020 2069 6620 616c 7068 6162 6574  .    if alphabet
-00000b70: 735b 2264 6e61 225d 2e73 6561 7263 6828  s["dna"].search(
-00000b80: 7374 7228 7365 7129 2920 6973 206e 6f74  str(seq)) is not
-00000b90: 204e 6f6e 653a 2072 6169 7365 2056 616c   None: raise Val
-00000ba0: 7565 4572 726f 7228 2250 6c65 6173 6520  ueError("Please 
-00000bb0: 7573 6520 616d 696e 6f20 6163 6964 7320  use amino acids 
-00000bc0: 696e 2079 6f75 7220 7365 7175 656e 6365  in your sequence
-00000bd0: 732c 2061 6363 6570 7465 6420 616c 7068  s, accepted alph
-00000be0: 6162 6574 7320 6172 6520 7072 6f74 6569  abets are protei
-00000bf0: 6e32 313a 202d 6163 6465 6667 6869 6b6c  n21: -acdefghikl
-00000c00: 6d6e 7071 7273 7476 7779 206f 7220 7072  mnpqrstvwy or pr
-00000c10: 6f74 6569 6e32 3170 6c75 733a 202d 2a61  otein21plus: -*a
-00000c20: 6364 6566 6768 696b 6c6d 6e70 7172 7374  cdefghiklmnpqrst
-00000c30: 7677 7962 7a78 2229 0a0a 2020 2020 6966  vwybzx")..    if
-00000c40: 2061 6c70 6861 6265 7473 5b22 7072 6f74   alphabets["prot
-00000c50: 6569 6e32 3122 5d2e 7365 6172 6368 2873  ein21"].search(s
-00000c60: 7472 2873 6571 2929 2069 7320 6e6f 7420  tr(seq)) is not 
-00000c70: 4e6f 6e65 3a0a 2020 2020 2020 2020 6161  None:.        aa
-00000c80: 5f70 726f 6273 203d 2032 310a 2020 2020  _probs = 21.    
-00000c90: 2020 2020 7265 7475 726e 2061 615f 7072      return aa_pr
-00000ca0: 6f62 730a 2020 2020 6966 2061 6c70 6861  obs.    if alpha
-00000cb0: 6265 7473 5b22 7072 6f74 6569 6e32 3170  bets["protein21p
-00000cc0: 6c75 7322 5d2e 7365 6172 6368 2873 7472  lus"].search(str
-00000cd0: 2873 6571 2929 2069 7320 6e6f 7420 4e6f  (seq)) is not No
-00000ce0: 6e65 3a0a 2020 2020 2020 2020 6161 5f70  ne:.        aa_p
-00000cf0: 726f 6273 203d 2032 340a 2020 2020 2020  robs = 24.      
-00000d00: 2020 7265 7475 726e 2061 615f 7072 6f62    return aa_prob
-00000d10: 730a 2020 2020 656c 7365 3a0a 2020 2020  s.    else:.    
-00000d20: 2020 2020 7261 6973 6520 5661 6c75 6545      raise ValueE
-00000d30: 7272 6f72 2822 596f 7572 2073 6571 7565  rror("Your seque
-00000d40: 6e63 6573 2063 6f6e 7461 696e 206e 6f74  nces contain not
-00000d50: 2061 6c6c 6f77 6564 2063 6861 7261 6374   allowed charact
-00000d60: 6572 732e 2041 7661 696c 6162 6c65 2061  ers. Available a
-00000d70: 6c70 6861 6265 7473 2061 7265 3a20 7b70  lphabets are: {p
-00000d80: 726f 7465 696e 3231 7d3a 202d 6163 6465  rotein21}: -acde
-00000d90: 6667 6869 6b6c 6d6e 7071 7273 7476 7779  fghiklmnpqrstvwy
-00000da0: 206f 7220 7b70 726f 7465 696e 3231 706c   or {protein21pl
-00000db0: 7573 7d20 2d2a 6163 6465 6667 6869 6b6c  us} -*acdefghikl
-00000dc0: 6d6e 7071 7273 7476 7779 627a 782e 2049  mnpqrstvwybzx. I
-00000dd0: 6620 796f 7572 2073 6571 7565 6e63 6520  f your sequence 
-00000de0: 636f 6e74 6169 6e73 2073 746f 7020 636f  contains stop co
-00000df0: 646f 6e73 2070 6572 6861 7073 2079 6f75  dons perhaps you
-00000e00: 2063 616e 2074 7269 6d20 7468 656d 2e22   can trim them."
-00000e10: 290a 0a64 6566 2061 6d69 6e6f 6163 6964  )..def aminoacid
-00000e20: 5f6e 616d 6573 5f64 6963 7428 6161 5f70  _names_dict(aa_p
-00000e30: 726f 6273 293a 0a20 2020 2022 2222 2052  robs):.    """ R
-00000e40: 6574 7572 6e73 2061 6e20 616d 696e 6f61  eturns an aminoa
-00000e50: 6369 6420 6173 736f 6369 6174 6564 2074  cid associated t
-00000e60: 6f20 6120 696e 7465 6765 7220 7661 6c75  o a integer valu
-00000e70: 650a 2020 2020 3a70 6172 616d 2069 6e74  e.    :param int
-00000e80: 2061 615f 7072 6f62 733a 2061 6d69 6e6f   aa_probs: amino
-00000e90: 2061 6369 6420 7072 6f62 6162 696c 6974   acid probabilit
-00000ea0: 6965 732c 2074 6869 7320 6e75 6d62 6572  ies, this number
-00000eb0: 2063 6f72 7265 6c61 7465 7320 746f 2074   correlates to t
-00000ec0: 6865 206e 756d 6265 7220 6f66 2064 6966  he number of dif
-00000ed0: 6665 7265 6e74 2061 6120 7479 7065 7320  ferent aa types 
-00000ee0: 696e 2074 6865 2069 6e70 7574 2061 6c69  in the input ali
-00000ef0: 676e 6d65 6e74 2222 220a 2020 2020 6966  gnment""".    if
-00000f00: 2061 615f 7072 6f62 7320 3d3d 2032 313a   aa_probs == 21:
-00000f10: 0a20 2020 2020 2020 2061 6d69 6e6f 6163  .        aminoac
-00000f20: 6964 5f6e 616d 6573 203d 207b 222d 223a  id_names = {"-":
-00000f30: 302c 2252 223a 312c 2248 223a 322c 224b  0,"R":1,"H":2,"K
-00000f40: 223a 332c 2244 223a 342c 2245 223a 352c  ":3,"D":4,"E":5,
-00000f50: 2253 223a 362c 2254 223a 372c 224e 223a  "S":6,"T":7,"N":
-00000f60: 382c 2251 223a 392c 2243 223a 3130 2c22  8,"Q":9,"C":10,"
-00000f70: 4722 3a31 312c 2250 223a 3132 2c22 4122  G":11,"P":12,"A"
-00000f80: 3a31 332c 2256 223a 3134 2c22 4922 3a31  :13,"V":14,"I":1
-00000f90: 352c 224c 223a 3136 2c22 4d22 3a31 372c  5,"L":16,"M":17,
-00000fa0: 2246 223a 3138 2c22 5922 3a31 392c 2257  "F":18,"Y":19,"W
-00000fb0: 223a 3230 7d0a 2020 2020 2020 2020 7265  ":20}.        re
-00000fc0: 7475 726e 2061 6d69 6e6f 6163 6964 5f6e  turn aminoacid_n
-00000fd0: 616d 6573 0a20 2020 2069 6620 6161 5f70  ames.    if aa_p
-00000fe0: 726f 6273 203d 3d20 3232 3a20 2369 6e63  robs == 22: #inc
-00000ff0: 6c75 6465 7320 7374 6f70 2063 6f64 6f6e  ludes stop codon
-00001000: 732d 2d2d 3e20 6669 7820 696e 2043 7265  s---> fix in Cre
-00001010: 6174 6520 626c 6f73 756d 0a20 2020 2020  ate blosum.     
-00001020: 2020 2061 6d69 6e6f 6163 6964 5f6e 616d     aminoacid_nam
-00001030: 6573 203d 207b 222d 223a 302c 222a 223a  es = {"-":0,"*":
-00001040: 302c 2252 223a 312c 2248 223a 322c 224b  0,"R":1,"H":2,"K
-00001050: 223a 332c 2244 223a 342c 2245 223a 352c  ":3,"D":4,"E":5,
-00001060: 2253 223a 362c 2254 223a 372c 224e 223a  "S":6,"T":7,"N":
-00001070: 382c 2251 223a 392c 2243 223a 3130 2c22  8,"Q":9,"C":10,"
-00001080: 4722 3a31 312c 2250 223a 3132 2c22 4122  G":11,"P":12,"A"
-00001090: 3a31 332c 2256 223a 3134 2c22 4922 3a31  :13,"V":14,"I":1
-000010a0: 352c 224c 223a 3136 2c22 4d22 3a31 372c  5,"L":16,"M":17,
-000010b0: 2246 223a 3138 2c22 5922 3a31 392c 2257  "F":18,"Y":19,"W
-000010c0: 223a 3230 7d0a 2020 2020 2020 2020 7265  ":20}.        re
-000010d0: 7475 726e 2061 6d69 6e6f 6163 6964 5f6e  turn aminoacid_n
-000010e0: 616d 6573 0a20 2020 2065 6c69 6620 6161  ames.    elif aa
-000010f0: 5f70 726f 6273 203e 2032 323a 0a20 2020  _probs > 22:.   
-00001100: 2020 2020 2061 6d69 6e6f 6163 6964 5f6e       aminoacid_n
-00001110: 616d 6573 203d 207b 222d 223a 302c 2252  ames = {"-":0,"R
-00001120: 223a 312c 2248 223a 322c 224b 223a 332c  ":1,"H":2,"K":3,
-00001130: 2244 223a 342c 2245 223a 352c 2253 223a  "D":4,"E":5,"S":
-00001140: 362c 2254 223a 372c 224e 223a 382c 2251  6,"T":7,"N":8,"Q
-00001150: 223a 392c 2243 223a 3130 2c22 4722 3a31  ":9,"C":10,"G":1
-00001160: 312c 2250 223a 3132 2c22 4122 3a31 332c  1,"P":12,"A":13,
-00001170: 2256 223a 3134 2c22 4922 3a31 352c 224c  "V":14,"I":15,"L
-00001180: 223a 3136 2c22 4d22 3a31 372c 2246 223a  ":16,"M":17,"F":
-00001190: 3138 2c22 5922 3a31 392c 2257 223a 3230  18,"Y":19,"W":20
-000011a0: 2c22 4222 3a32 312c 225a 223a 3232 2c22  ,"B":21,"Z":22,"
-000011b0: 5822 3a32 337d 0a20 2020 2020 2020 2072  X":23}.        r
-000011c0: 6574 7572 6e20 616d 696e 6f61 6369 645f  eturn aminoacid_
-000011d0: 6e61 6d65 730a 6465 6620 6372 6561 7465  names.def create
-000011e0: 5f62 6c6f 7375 6d28 6161 5f70 726f 6273  _blosum(aa_probs
-000011f0: 2c73 7562 735f 6d61 7472 6978 5f6e 616d  ,subs_matrix_nam
-00001200: 6529 3a0a 2020 2020 2222 220a 2020 2020  e):.    """.    
-00001210: 4275 696c 6473 2061 6e20 6172 7261 7920  Builds an array 
-00001220: 636f 6e74 6169 6e69 6e67 2074 6865 2062  containing the b
-00001230: 6c6f 7375 6d20 7363 6f72 6573 2070 6572  losum scores per
-00001240: 2063 6861 7261 6374 6572 0a20 2020 203a   character.    :
-00001250: 7061 7261 6d20 6161 5f70 726f 6273 3a20  param aa_probs: 
-00001260: 616d 696e 6f20 6163 6964 2070 726f 6261  amino acid proba
-00001270: 6269 6c69 7469 6573 2c20 6465 7465 726d  bilities, determ
-00001280: 696e 6573 2074 6865 2063 686f 6963 6520  ines the choice 
-00001290: 6f66 2042 4c4f 5355 4d20 6d61 7472 6978  of BLOSUM matrix
-000012a0: 0a20 2020 203a 7061 7261 6d20 7374 7220  .    :param str 
-000012b0: 7375 6273 5f6d 6174 7269 785f 6e61 6d65  subs_matrix_name
-000012c0: 3a20 6e61 6d65 206f 6620 7468 6520 7375  : name of the su
-000012d0: 6273 7469 7475 7469 6f6e 206d 6174 7269  bstitution matri
-000012e0: 782c 2063 6865 636b 2061 7661 696c 6162  x, check availab
-000012f0: 696c 6974 7920 6174 202f 686f 6d65 2f6c  ility at /home/l
-00001300: 7973 2f61 6e61 636f 6e64 6133 2f70 6b67  ys/anaconda3/pkg
-00001310: 732f 6269 6f70 7974 686f 6e2d 312e 3736  s/biopython-1.76
-00001320: 2d70 7933 3768 3531 3639 3039 615f 302f  -py37h516909a_0/
-00001330: 6c69 622f 7079 7468 6f6e 332e 372f 7369  lib/python3.7/si
-00001340: 7465 2d70 6163 6b61 6765 732f 4269 6f2f  te-packages/Bio/
-00001350: 416c 6967 6e2f 7375 6273 7469 7475 7469  Align/substituti
-00001360: 6f6e 5f6d 6174 7269 6365 732f 6461 7461  on_matrices/data
-00001370: 2222 220a 0a20 2020 2069 6620 6161 5f70  """..    if aa_p
-00001380: 726f 6273 203e 2032 3120 616e 6420 6e6f  robs > 21 and no
-00001390: 7420 7375 6273 5f6d 6174 7269 785f 6e61  t subs_matrix_na
-000013a0: 6d65 2e73 7461 7274 7377 6974 6828 2250  me.startswith("P
-000013b0: 414d 2229 3a0a 2020 2020 2020 2020 7761  AM"):.        wa
-000013c0: 726e 696e 6773 2e77 6172 6e28 2259 6f75  rnings.warn("You
-000013d0: 7220 6461 7461 7365 7420 636f 6e74 6169  r dataset contai
-000013e0: 6e73 2073 7065 6369 616c 2061 6d69 6e6f  ns special amino
-000013f0: 2061 6369 6473 2e20 5377 6974 6368 696e   acids. Switchin
-00001400: 6720 796f 7572 2073 7562 7374 6974 7574  g your substitut
-00001410: 696f 6e20 6d61 7472 6978 2074 6f20 5041  ion matrix to PA
-00001420: 4d37 3022 290a 2020 2020 2020 2020 7375  M70").        su
-00001430: 6273 5f6d 6174 7269 785f 6e61 6d65 203d  bs_matrix_name =
-00001440: 2022 5041 4d37 3022 0a20 2020 2073 7562   "PAM70".    sub
-00001450: 735f 6d61 7472 6978 203d 2042 696f 2e41  s_matrix = Bio.A
-00001460: 6c69 676e 2e73 7562 7374 6974 7574 696f  lign.substitutio
-00001470: 6e5f 6d61 7472 6963 6573 2e6c 6f61 6428  n_matrices.load(
-00001480: 7375 6273 5f6d 6174 7269 785f 6e61 6d65  subs_matrix_name
-00001490: 290a 2020 2020 6161 5f6c 6973 7420 3d20  ).    aa_list = 
-000014a0: 6c69 7374 2861 6d69 6e6f 6163 6964 5f6e  list(aminoacid_n
-000014b0: 616d 6573 5f64 6963 7428 6161 5f70 726f  ames_dict(aa_pro
-000014c0: 6273 292e 6b65 7973 2829 290a 2020 2020  bs).keys()).    
-000014d0: 696e 6465 785f 6761 7020 3d20 6161 5f6c  index_gap = aa_l
-000014e0: 6973 742e 696e 6465 7828 222d 2229 0a20  ist.index("-"). 
-000014f0: 2020 2061 615f 6c69 7374 5b69 6e64 6578     aa_list[index
-00001500: 5f67 6170 5d20 3d20 222a 2220 2369 6e20  _gap] = "*" #in 
-00001510: 7468 6520 626c 6f73 756d 206d 6174 7269  the blosum matri
-00001520: 7820 6761 7073 2061 7265 2072 6570 7265  x gaps are repre
-00001530: 7361 6e74 6564 2061 7320 2a0a 0a20 2020  santed as *..   
-00001540: 2073 7562 735f 6469 6374 203d 2064 6566   subs_dict = def
-00001550: 6175 6c74 6469 6374 2829 0a20 2020 2073  aultdict().    s
-00001560: 7562 735f 6172 7261 7920 3d20 6e70 2e7a  ubs_array = np.z
-00001570: 6572 6f73 2828 6c65 6e28 6161 5f6c 6973  eros((len(aa_lis
-00001580: 7429 202c 206c 656e 2861 615f 6c69 7374  t) , len(aa_list
-00001590: 2920 2929 0a20 2020 2066 6f72 2069 2c20  ) )).    for i, 
-000015a0: 6161 5f31 2069 6e20 656e 756d 6572 6174  aa_1 in enumerat
-000015b0: 6528 6161 5f6c 6973 7429 3a0a 2020 2020  e(aa_list):.    
-000015c0: 2020 2020 666f 7220 6a2c 2061 615f 3220      for j, aa_2 
-000015d0: 696e 2065 6e75 6d65 7261 7465 2861 615f  in enumerate(aa_
-000015e0: 6c69 7374 293a 0a20 2020 2020 2020 2020  list):.         
-000015f0: 2020 2069 6620 6161 5f31 2021 3d20 222a     if aa_1 != "*
-00001600: 2220 616e 6420 6161 5f32 2021 3d20 222a  " and aa_2 != "*
-00001610: 223a 0a20 2020 2020 2020 2020 2020 2020  ":.             
-00001620: 2020 2073 7562 735f 6469 6374 5b28 6161     subs_dict[(aa
-00001630: 5f31 2c61 615f 3229 5d20 3d20 7375 6273  _1,aa_2)] = subs
-00001640: 5f6d 6174 7269 785b 2861 615f 312c 2061  _matrix[(aa_1, a
-00001650: 615f 3229 5d0a 2020 2020 2020 2020 2020  a_2)].          
-00001660: 2020 2020 2020 7375 6273 5f64 6963 745b        subs_dict[
-00001670: 2861 615f 322c 2061 615f 3129 5d20 3d20  (aa_2, aa_1)] = 
-00001680: 7375 6273 5f6d 6174 7269 785b 2861 615f  subs_matrix[(aa_
-00001690: 312c 2061 615f 3229 5d0a 2020 2020 2020  1, aa_2)].      
-000016a0: 2020 2020 2020 7375 6273 5f61 7272 6179        subs_array
-000016b0: 5b69 2c20 6a5d 203d 2073 7562 735f 6d61  [i, j] = subs_ma
-000016c0: 7472 6978 5b28 6161 5f31 2c20 6161 5f32  trix[(aa_1, aa_2
-000016d0: 295d 0a20 2020 2020 2020 2020 2020 2073  )].            s
-000016e0: 7562 735f 6172 7261 795b 6a2c 2069 5d20  ubs_array[j, i] 
-000016f0: 3d20 7375 6273 5f6d 6174 7269 785b 2861  = subs_matrix[(a
-00001700: 615f 322c 2061 615f 3129 5d0a 0a20 2020  a_2, aa_1)]..   
-00001710: 206e 616d 6573 203d 206e 702e 636f 6e63   names = np.conc
-00001720: 6174 656e 6174 6528 286e 702e 6172 7261  atenate((np.arra
-00001730: 7928 5b66 6c6f 6174 2822 2d69 6e66 2229  y([float("-inf")
-00001740: 5d29 2c20 6e70 2e61 7261 6e67 6528 302c  ]), np.arange(0,
-00001750: 6161 5f70 726f 6273 2929 290a 2020 2020  aa_probs))).    
-00001760: 7375 6273 5f61 7272 6179 203d 206e 702e  subs_array = np.
-00001770: 635f 5b20 6e70 2e61 7261 6e67 6528 302c  c_[ np.arange(0,
-00001780: 6161 5f70 726f 6273 292c 2073 7562 735f  aa_probs), subs_
-00001790: 6172 7261 7920 5d0a 2020 2020 7375 6273  array ].    subs
-000017a0: 5f61 7272 6179 203d 206e 702e 636f 6e63  _array = np.conc
-000017b0: 6174 656e 6174 6528 286e 616d 6573 5b4e  atenate((names[N
-000017c0: 6f6e 652c 3a5d 2c73 7562 735f 6172 7261  one,:],subs_arra
-000017d0: 7929 2c61 7869 733d 3029 0a0a 2020 2020  y),axis=0)..    
-000017e0: 7265 7475 726e 2073 7562 735f 6172 7261  return subs_arra
-000017f0: 792c 2073 7562 735f 6469 6374 0a64 6566  y, subs_dict.def
-00001800: 2064 6976 6964 655f 696e 746f 5f6d 6f6e   divide_into_mon
-00001810: 6f70 6879 6c65 7469 635f 636c 6164 6573  ophyletic_clades
-00001820: 2874 7265 652c 7374 6f72 6167 655f 666f  (tree,storage_fo
-00001830: 6c64 6572 2c6e 616d 6529 3a0a 2020 2020  lder,name):.    
-00001840: 2222 220a 2020 2020 4469 7669 6465 7320  """.    Divides 
-00001850: 7468 6520 7472 6565 2069 6e74 6f20 6d6f  the tree into mo
-00001860: 6e6f 7068 796c 6574 6963 2063 6c61 6465  nophyletic clade
-00001870: 733a 0a20 2020 2053 6565 2068 7474 7073  s:.    See https
-00001880: 3a2f 2f77 7777 2e6d 756e 2e63 612f 6269  ://www.mun.ca/bi
-00001890: 6f6c 6f67 792f 7363 6172 722f 5461 786f  ology/scarr/Taxo
-000018a0: 6e5f 7479 7065 732e 6874 6d6c 0a20 2020  n_types.html.   
-000018b0: 2049 6d70 6c65 6d65 6e74 6174 696f 6e20   Implementation 
-000018c0: 6261 7365 6420 6f6e 3a20 6874 7470 733a  based on: https:
-000018d0: 2f2f 7777 772e 6269 6f73 7461 7273 2e6f  //www.biostars.o
-000018e0: 7267 2f70 2f39 3734 3039 2f0a 0a20 2020  rg/p/97409/..   
-000018f0: 2054 6865 2072 6561 736f 6e61 626c 6520   The reasonable 
-00001900: 636c 6164 6520 6469 7669 7369 6f6e 2063  clade division c
-00001910: 7269 7465 7269 6120 7365 656d 7320 746f  riteria seems to
-00001920: 2067 726f 7570 2061 6c6c 2074 686f 7365   group all those
-00001930: 206e 6f64 6573 2077 686f 7365 2064 6973   nodes whose dis
-00001940: 7461 6e63 6520 746f 2074 6865 2069 6e74  tance to the int
-00001950: 6572 6e61 6c20 6973 206c 6f77 6572 2074  ernal is lower t
-00001960: 6f20 7468 6520 6f76 6572 616c 6c20 6176  o the overall av
-00001970: 6572 6167 6520 6469 7374 616e 6365 2066  erage distance f
-00001980: 726f 6d20 6561 6368 206c 6561 6620 746f  rom each leaf to
-00001990: 2074 6865 2072 6f6f 740a 0a20 2020 203a   the root..    :
-000019a0: 7061 7261 6d20 6574 6533 2d74 7265 6520  param ete3-tree 
-000019b0: 7472 6565 3a20 6574 6533 2074 7265 6520  tree: ete3 tree 
-000019c0: 636c 6173 730a 2020 2020 3a70 6172 616d  class.    :param
-000019d0: 2073 7472 2073 746f 7261 6765 5f66 6f6c   str storage_fol
-000019e0: 6465 723a 2066 6f6c 6465 7220 7768 6572  der: folder wher
-000019f0: 6520 746f 2073 746f 7265 2074 6865 2072  e to store the r
-00001a00: 6573 756c 7473 2c20 696e 2074 6869 7320  esults, in this 
-00001a10: 6361 7365 2061 2064 6963 7469 6f6e 6172  case a dictionar
-00001a20: 7920 636f 6e74 6169 6e69 6e67 207b 2263  y containing {"c
-00001a30: 6c61 6465 5f6e 756d 6265 7222 3a20 5b6e  lade_number": [n
-00001a40: 6f64 6573 206c 6973 745d 7d0a 2020 2020  odes list]}.    
-00001a50: 3a6e 616d 6520 7374 7220 6e61 6d65 3a20  :name str name: 
-00001a60: 6e61 6d65 206f 6620 7468 6520 6461 7461  name of the data
-00001a70: 2073 6574 2070 726f 6a65 6374 0a0a 2020   set project..  
-00001a80: 2020 2222 220a 0a20 2020 2064 6566 206d    """..    def m
-00001a90: 6561 6e28 6172 7261 7929 3a0a 2020 2020  ean(array):.    
-00001aa0: 2020 2020 2222 2243 616c 6375 6c61 7465      """Calculate
-00001ab0: 7320 6272 616e 6368 206c 656e 6774 6820  s branch length 
-00001ac0: 6176 6572 6167 6522 2222 0a20 2020 2020  average""".     
-00001ad0: 2020 2072 6574 7572 6e20 7375 6d28 6172     return sum(ar
-00001ae0: 7261 7929 202f 2066 6c6f 6174 286c 656e  ray) / float(len
-00001af0: 2861 7272 6179 2929 0a0a 2020 2020 6465  (array))..    de
-00001b00: 6620 6361 6368 655f 6469 7374 616e 6365  f cache_distance
-00001b10: 7328 7472 6565 293a 0a20 2020 2020 2020  s(tree):.       
-00001b20: 2022 2222 5072 6563 616c 6375 6c61 7465   """Precalculate
-00001b30: 2064 6973 7461 6e63 6573 206f 6620 616c   distances of al
-00001b40: 6c20 6e6f 6465 7320 746f 2074 6865 2072  l nodes to the r
-00001b50: 6f6f 7422 2222 0a20 2020 2020 2020 206e  oot""".        n
-00001b60: 6f64 6532 726f 6f74 6469 7374 203d 207b  ode2rootdist = {
-00001b70: 7472 6565 3a20 307d 0a20 2020 2020 2020  tree: 0}.       
-00001b80: 2066 6f72 206e 6f64 6520 696e 2074 7265   for node in tre
-00001b90: 652e 6974 6572 5f64 6573 6365 6e64 616e  e.iter_descendan
-00001ba0: 7473 2827 7072 656f 7264 6572 2729 3a0a  ts('preorder'):.
-00001bb0: 2020 2020 2020 2020 2020 2020 6e6f 6465              node
-00001bc0: 3272 6f6f 7464 6973 745b 6e6f 6465 5d20  2rootdist[node] 
-00001bd0: 3d20 6e6f 6465 2e64 6973 7420 2b20 6e6f  = node.dist + no
-00001be0: 6465 3272 6f6f 7464 6973 745b 6e6f 6465  de2rootdist[node
-00001bf0: 2e75 705d 0a20 2020 2020 2020 2072 6574  .up].        ret
-00001c00: 7572 6e20 6e6f 6465 3272 6f6f 7464 6973  urn node2rootdis
-00001c10: 740a 0a20 2020 2064 6566 2062 7569 6c64  t..    def build
-00001c20: 5f63 6c61 6465 7328 7472 6565 2c6e 616d  _clades(tree,nam
-00001c30: 6529 3a0a 2020 2020 2020 2020 2222 2257  e):.        """W
-00001c40: 6865 6e20 6120 636c 7573 7465 7269 6e67  hen a clustering
-00001c50: 2063 6f6e 6469 7469 6f6e 2069 7320 6d65   condition is me
-00001c60: 742c 2069 7420 636f 6c6c 6170 7365 7320  t, it collapses 
-00001c70: 7468 6520 7472 6565 2061 7420 7468 6174  the tree at that
-00001c80: 206e 6f64 6520 746f 2075 6e69 6679 2061   node to unify a
-00001c90: 6c6c 2074 6865 206c 6561 7665 7320 696e  ll the leaves in
-00001ca0: 2074 6861 7420 636c 7573 7465 7220 696e   that cluster in
-00001cb0: 746f 2027 6f6e 6527 206c 6561 7665 732e  to 'one' leaves.
-00001cc0: 2041 6674 6572 2c20 7765 2072 6561 640a   After, we read.
-00001cd0: 2020 2020 2020 2020 7468 6520 636f 6c6c          the coll
-00001ce0: 6170 7365 6420 7472 6565 2069 6e74 6f20  apsed tree into 
-00001cf0: 6120 6469 6374 696f 6e61 7279 2074 6861  a dictionary tha
-00001d00: 7420 636f 6e74 6169 6e73 207b 636c 6164  t contains {clad
-00001d10: 6520 6e75 6d62 6572 3a7b 2269 6e74 6572  e number:{"inter
-00001d20: 6e61 6c22 3a5b 6e6f 6465 7320 6e75 6d62  nal":[nodes numb
-00001d30: 6572 735d 2c22 6c65 6176 6573 223a 5b6e  ers],"leaves":[n
-00001d40: 6f64 6520 6e75 6d62 6572 735d 7d7d 2222  ode numbers]}}""
-00001d50: 220a 2020 2020 2020 2020 2320 6361 6368  ".        # cach
-00001d60: 6520 7468 6520 7469 7020 636f 6e74 656e  e the tip conten
-00001d70: 7420 6f66 2065 6163 6820 6e6f 6465 2074  t of each node t
-00001d80: 6f20 7265 6475 6365 2074 6865 206e 756d  o reduce the num
-00001d90: 6265 7220 6f66 2074 696d 6573 2074 6865  ber of times the
-00001da0: 2074 7265 6520 6973 2074 7261 7665 7273   tree is travers
-00001db0: 6564 0a20 2020 2020 2020 206e 6f64 6532  ed.        node2
-00001dc0: 7469 7073 203d 2074 7265 652e 6765 745f  tips = tree.get_
-00001dd0: 6361 6368 6564 5f63 6f6e 7465 6e74 2829  cached_content()
-00001de0: 0a20 2020 2020 2020 2072 6f6f 745f 6469  .        root_di
-00001df0: 7374 616e 6365 203d 2063 6163 6865 5f64  stance = cache_d
-00001e00: 6973 7461 6e63 6573 2874 7265 6529 2020  istances(tree)  
-00001e10: 2320 6469 7374 616e 6365 7320 6f66 2065  # distances of e
-00001e20: 6163 6820 6f66 2074 6865 206e 6f64 6573  ach of the nodes
-00001e30: 2074 6f20 7468 6520 726f 6f74 0a20 2020   to the root.   
-00001e40: 2020 2020 2061 7665 7261 6765 5f72 6f6f       average_roo
-00001e50: 745f 6469 7374 616e 6365 203d 206d 6561  t_distance = mea
-00001e60: 6e28 726f 6f74 5f64 6973 7461 6e63 652e  n(root_distance.
-00001e70: 7661 6c75 6573 2829 290a 2020 2020 2020  values()).      
-00001e80: 2020 7374 645f 726f 6f74 5f64 6973 7461    std_root_dista
-00001e90: 6e63 6520 3d20 7374 6174 6973 7469 6373  nce = statistics
-00001ea0: 2e73 7464 6576 2872 6f6f 745f 6469 7374  .stdev(root_dist
-00001eb0: 616e 6365 2e76 616c 7565 7328 2929 0a20  ance.values()). 
-00001ec0: 2020 2020 2020 206e 5f6c 6561 7665 7320         n_leaves 
-00001ed0: 3d20 6c65 6e28 7472 6565 2e67 6574 5f6c  = len(tree.get_l
-00001ee0: 6561 7665 7328 2929 0a20 2020 2020 2020  eaves()).       
-00001ef0: 2023 544f 444f 3a20 6175 746f 6d61 7469   #TODO: automati
-00001f00: 7a65 2063 6c75 7374 6572 696e 6720 636f  ze clustering co
-00001f10: 6e64 6974 696f 6e0a 2020 2020 2020 2020  ndition.        
-00001f20: 6966 206e 5f6c 6561 7665 7320 3e3d 2031  if n_leaves >= 1
-00001f30: 3030 206f 7220 6e61 6d65 2e65 6e64 7377  00 or name.endsw
-00001f40: 6974 6828 225f 7375 6274 7265 6522 293a  ith("_subtree"):
-00001f50: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
-00001f60: 6e61 6d65 2069 6e20 5b22 5046 3030 3039  name in ["PF0009
-00001f70: 3622 2c22 5046 3030 3430 3022 5d3a 0a20  6","PF00400"]:. 
-00001f80: 2020 2020 2020 2020 2020 2020 2020 2063                 c
-00001f90: 6c75 7374 6572 696e 675f 636f 6e64 6974  lustering_condit
-00001fa0: 696f 6e20 3d20 6176 6572 6167 655f 726f  ion = average_ro
-00001fb0: 6f74 5f64 6973 7461 6e63 6520 232b 2030  ot_distance #+ 0
-00001fc0: 2e33 2a73 7464 5f72 6f6f 745f 6469 7374  .3*std_root_dist
-00001fd0: 616e 6365 0a20 2020 2020 2020 2020 2020  ance.           
-00001fe0: 2065 6c73 653a 0a20 2020 2020 2020 2020   else:.         
-00001ff0: 2020 2020 2020 2063 6c75 7374 6572 696e         clusterin
-00002000: 675f 636f 6e64 6974 696f 6e20 3d20 6176  g_condition = av
-00002010: 6572 6167 655f 726f 6f74 5f64 6973 7461  erage_root_dista
-00002020: 6e63 6520 2d32 2a73 7464 5f72 6f6f 745f  nce -2*std_root_
-00002030: 6469 7374 616e 6365 0a20 2020 2020 2020  distance.       
-00002040: 2065 6c69 6620 6e61 6d65 2069 6e20 5b22   elif name in ["
-00002050: 436f 7261 6c5f 616c 6c22 2c22 436f 7261  Coral_all","Cora
-00002060: 6c5f 4661 7669 696e 6122 2c22 5348 335f  l_Faviina","SH3_
-00002070: 7066 3030 3031 385f 6c61 7267 6572 5f74  pf00018_larger_t
-00002080: 6861 6e5f 3330 6161 225d 206f 7220 2263  han_30aa"] or "c
-00002090: 616c 6369 746f 6e69 6e22 2069 6e20 6e61  alcitonin" in na
-000020a0: 6d65 3a0a 2020 2020 2020 2020 2020 2020  me:.            
-000020b0: 636c 7573 7465 7269 6e67 5f63 6f6e 6469  clustering_condi
-000020c0: 7469 6f6e 203d 2061 7665 7261 6765 5f72  tion = average_r
-000020d0: 6f6f 745f 6469 7374 616e 6365 202d 2073  oot_distance - s
-000020e0: 7464 5f72 6f6f 745f 6469 7374 616e 6365  td_root_distance
-000020f0: 0a20 2020 2020 2020 2065 6c73 653a 0a20  .        else:. 
-00002100: 2020 2020 2020 2020 2020 2063 6c75 7374             clust
-00002110: 6572 696e 675f 636f 6e64 6974 696f 6e20  ering_condition 
-00002120: 3d20 6176 6572 6167 655f 726f 6f74 5f64  = average_root_d
-00002130: 6973 7461 6e63 650a 0a20 2020 2020 2020  istance..       
-00002140: 2066 6f72 206e 6f64 6520 696e 2074 7265   for node in tre
-00002150: 652e 6765 745f 6465 7363 656e 6461 6e74  e.get_descendant
-00002160: 7328 2770 7265 6f72 6465 7227 293a 0a20  s('preorder'):. 
-00002170: 2020 2020 2020 2020 2020 2069 6620 6e6f             if no
-00002180: 7420 6e6f 6465 2e69 735f 6c65 6166 2829  t node.is_leaf()
-00002190: 3a20 2023 2066 6f72 2069 6e74 6572 6e61  :  # for interna
-000021a0: 6c20 6e6f 6465 730a 2020 2020 2020 2020  l nodes.        
-000021b0: 2020 2020 2020 2020 6176 675f 6469 7374          avg_dist
-000021c0: 616e 6365 5f74 6f5f 7469 7073 203d 206d  ance_to_tips = m
-000021d0: 6561 6e28 5b72 6f6f 745f 6469 7374 616e  ean([root_distan
-000021e0: 6365 5b74 6970 5d20 2d20 726f 6f74 5f64  ce[tip] - root_d
-000021f0: 6973 7461 6e63 655b 6e6f 6465 5d20 666f  istance[node] fo
-00002200: 7220 7469 7020 696e 206e 6f64 6532 7469  r tip in node2ti
-00002210: 7073 5b6e 6f64 655d 5d29 2020 2320 6176  ps[node]])  # av
-00002220: 6572 6167 6520 6469 7374 616e 6365 2066  erage distance f
-00002230: 726f 6d20 7468 6520 696e 7465 726e 616c  rom the internal
-00002240: 206e 6f64 6520 746f 2061 6c6c 2069 7427   node to all it'
-00002250: 7320 706f 7373 6962 6c65 2064 6572 6976  s possible deriv
-00002260: 6564 206c 6561 7665 730a 2020 2020 2020  ed leaves.      
-00002270: 2020 2020 2020 2020 2020 6966 2061 7667            if avg
-00002280: 5f64 6973 7461 6e63 655f 746f 5f74 6970  _distance_to_tip
-00002290: 7320 3c20 636c 7573 7465 7269 6e67 5f63  s < clustering_c
-000022a0: 6f6e 6469 7469 6f6e 3a0a 2020 2020 2020  ondition:.      
-000022b0: 2020 2020 2020 2020 2020 2020 2020 236e                #n
-000022c0: 6f64 652e 6e61 6d65 202b 3d20 2720 434f  ode.name += ' CO
-000022d0: 4c4c 4150 5345 4420 6176 675f 643a 2567  LLAPSED avg_d:%g
-000022e0: 207b 2573 7d27 2025 2028 6176 675f 6469   {%s}' % (avg_di
-000022f0: 7374 616e 6365 5f74 6f5f 7469 7073 2c20  stance_to_tips, 
-00002300: 272c 272e 6a6f 696e 285b 7469 702e 6e61  ','.join([tip.na
-00002310: 6d65 2066 6f72 2074 6970 2069 6e20 6e6f  me for tip in no
-00002320: 6465 3274 6970 735b 6e6f 6465 5d5d 2929  de2tips[node]]))
-00002330: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00002340: 2020 2020 206e 6f64 652e 6e61 6d65 202b       node.name +
-00002350: 3d20 2720 434f 4c4c 4150 5345 4420 6176  = ' COLLAPSED av
-00002360: 675f 643a 2567 206c 6561 7665 733a 7b25  g_d:%g leaves:{%
-00002370: 737d 2069 6e74 6572 6e61 6c3a 7b25 737d  s} internal:{%s}
-00002380: 2720 2520 2861 7667 5f64 6973 7461 6e63  ' % (avg_distanc
-00002390: 655f 746f 5f74 6970 732c 2027 2c27 2e6a  e_to_tips, ','.j
-000023a0: 6f69 6e28 5b74 6970 2e6e 616d 6520 666f  oin([tip.name fo
-000023b0: 7220 7469 7020 696e 206e 6f64 6532 7469  r tip in node2ti
-000023c0: 7073 5b6e 6f64 655d 5d29 2c27 2c27 2e6a  ps[node]]),','.j
-000023d0: 6f69 6e28 5b69 6e74 6572 6e61 6c2e 6e61  oin([internal.na
-000023e0: 6d65 2066 6f72 2069 6e74 6572 6e61 6c20  me for internal 
-000023f0: 696e 206e 6f64 652e 6974 6572 5f64 6573  in node.iter_des
-00002400: 6365 6e64 616e 7473 2829 2069 6620 6e6f  cendants() if no
-00002410: 7420 696e 7465 726e 616c 2e69 735f 6c65  t internal.is_le
-00002420: 6166 2829 5d29 290a 2020 2020 2020 2020  af()])).        
-00002430: 2020 2020 2020 2020 2020 2020 6e6f 6465              node
-00002440: 2e61 6464 5f66 6561 7475 7265 7328 636f  .add_features(co
-00002450: 6c6c 6170 7365 643d 5472 7565 290a 2020  llapsed=True).  
+00000410: 2e44 6174 612e 4955 5041 4344 6174 6120  .Data.IUPACData 
+00000420: 696d 706f 7274 2070 726f 7465 696e 5f6c  import protein_l
+00000430: 6574 7465 7273 5f33 746f 310a 6672 6f6d  etters_3to1.from
+00000440: 2042 696f 2e53 6571 5265 636f 7264 2069   Bio.SeqRecord i
+00000450: 6d70 6f72 7420 5365 7152 6563 6f72 640a  mport SeqRecord.
+00000460: 6672 6f6d 2042 696f 2e53 6571 2069 6d70  from Bio.Seq imp
+00000470: 6f72 7420 5365 710a 6672 6f6d 2042 696f  ort Seq.from Bio
+00000480: 2069 6d70 6f72 7420 4269 6f70 7974 686f   import Biopytho
+00000490: 6e57 6172 6e69 6e67 0a66 726f 6d20 4269  nWarning.from Bi
+000004a0: 6f20 696d 706f 7274 2041 6c69 676e 494f  o import AlignIO
+000004b0: 2c20 5365 7149 4f0a 6672 6f6d 2042 696f  , SeqIO.from Bio
+000004c0: 2e50 4442 2e50 4442 4c69 7374 2069 6d70  .PDB.PDBList imp
+000004d0: 6f72 7420 5044 424c 6973 740a 696d 706f  ort PDBList.impo
+000004e0: 7274 2042 696f 2e41 6c69 676e 0a66 726f  rt Bio.Align.fro
+000004f0: 6d20 4269 6f2e 416c 6967 6e2e 4170 706c  m Bio.Align.Appl
+00000500: 6963 6174 696f 6e73 2069 6d70 6f72 7420  ications import 
+00000510: 4d61 6666 7443 6f6d 6d61 6e64 6c69 6e65  MafftCommandline
+00000520: 0a66 726f 6d20 4269 6f2e 5068 796c 6f2e  .from Bio.Phylo.
+00000530: 5472 6565 436f 6e73 7472 7563 7469 6f6e  TreeConstruction
+00000540: 2069 6d70 6f72 7420 2a0a 6672 6f6d 2042   import *.from B
+00000550: 696f 2069 6d70 6f72 7420 5068 796c 6f0a  io import Phylo.
+00000560: 234e 756d 7079 0a69 6d70 6f72 7420 6e75  #Numpy.import nu
+00000570: 6d70 7920 6173 206e 700a 696d 706f 7274  mpy as np.import
+00000580: 206e 756d 7079 2e72 616e 646f 6d20 6173   numpy.random as
+00000590: 206e 7072 0a69 6d70 6f72 7420 7061 6e64   npr.import pand
+000005a0: 6173 2061 7320 7064 0a69 6d70 6f72 7420  as as pd.import 
+000005b0: 6d61 7470 6c6f 746c 6962 0a6d 6174 706c  matplotlib.matpl
+000005c0: 6f74 6c69 622e 7573 6528 2741 6767 2729  otlib.use('Agg')
+000005d0: 0a64 6566 2061 615f 7072 6f70 6572 7469  .def aa_properti
+000005e0: 6573 2861 615f 7072 6f62 732c 6461 7461  es(aa_probs,data
+000005f0: 5f66 6f6c 6465 7229 3a0a 2020 2020 2222  _folder):.    ""
+00000600: 2220 4372 6561 7465 7320 6120 6469 6374  " Creates a dict
+00000610: 696f 6e61 7279 2077 6974 6820 616d 696e  ionary with amin
+00000620: 6f20 6163 6964 2070 726f 7065 7274 6965  o acid propertie
+00000630: 732c 2065 7874 7261 6374 6564 2066 726f  s, extracted fro
+00000640: 6d0a 2020 2020 6874 7470 733a 2f2f 7777  m.    https://ww
+00000650: 772e 7369 676d 6161 6c64 7269 6368 2e63  w.sigmaaldrich.c
+00000660: 6f6d 2f6c 6966 652d 7363 6965 6e63 652f  om/life-science/
+00000670: 6d65 7461 626f 6c6f 6d69 6373 2f6c 6561  metabolomics/lea
+00000680: 726e 696e 672d 6365 6e74 6572 2f61 6d69  rning-center/ami
+00000690: 6e6f 2d61 6369 642d 7265 6665 7265 6e63  no-acid-referenc
+000006a0: 652d 6368 6172 742e 6874 6d6c 0a20 2020  e-chart.html.   
+000006b0: 203a 7061 7261 6d20 696e 7420 6161 5f70   :param int aa_p
+000006c0: 726f 6273 3a20 616d 696e 6f20 6163 6964  robs: amino acid
+000006d0: 2070 726f 6261 6269 6c69 7469 6573 2075   probabilities u
+000006e0: 7365 6420 746f 2065 7874 7261 6374 2074  sed to extract t
+000006f0: 6865 2074 7970 6573 206f 6620 616d 696e  he types of amin
+00000700: 6f20 6163 6964 7320 7072 6573 656e 7420  o acids present 
+00000710: 696e 2074 6865 2064 6174 6173 6574 0a20  in the dataset. 
+00000720: 2020 203a 7061 7261 6d20 7374 7220 6461     :param str da
+00000730: 7461 5f66 6f6c 6465 723a 2070 6174 6820  ta_folder: path 
+00000740: 746f 2077 6865 7265 2064 7261 7570 6e69  to where draupni
+00000750: 7220 6973 2062 6565 6e20 6578 6563 7574  r is been execut
+00000760: 6564 2066 726f 6d22 2222 0a0a 2020 2020  ed from"""..    
+00000770: 6161 5f74 7970 6573 203d 206c 6973 7428  aa_types = list(
+00000780: 616d 696e 6f61 6369 645f 6e61 6d65 735f  aminoacid_names_
+00000790: 6469 6374 2861 615f 7072 6f62 7329 2e6b  dict(aa_probs).k
+000007a0: 6579 7328 2929 0a20 2020 2073 746f 7261  eys()).    stora
+000007b0: 6765 5f66 6f6c 6465 7220 3d20 6f73 2e70  ge_folder = os.p
+000007c0: 6174 682e 6162 7370 6174 6828 6f73 2e70  ath.abspath(os.p
+000007d0: 6174 682e 6a6f 696e 286f 732e 7061 7468  ath.join(os.path
+000007e0: 2e64 6972 6e61 6d65 285f 5f66 696c 655f  .dirname(__file_
+000007f0: 5f29 2c20 2264 6174 6122 2929 0a20 2020  _), "data")).   
+00000800: 2061 615f 7072 6f70 6572 7469 6573 203d   aa_properties =
+00000810: 2070 642e 7265 6164 5f63 7376 2822 7b7d   pd.read_csv("{}
+00000820: 2f41 415f 7072 6f70 6572 7469 6573 2e74  /AA_properties.t
+00000830: 7874 222e 666f 726d 6174 2873 746f 7261  xt".format(stora
+00000840: 6765 5f66 6f6c 6465 7229 2c73 6570 3d22  ge_folder),sep="
+00000850: 5c73 2b22 290a 2020 2020 6161 5f69 6e66  \s+").    aa_inf
+00000860: 6f20 3d20 6465 6661 756c 7464 6963 7428  o = defaultdict(
+00000870: 290a 2020 2020 666f 7220 696e 6478 2c61  ).    for indx,a
+00000880: 6120 696e 2061 615f 7072 6f70 6572 7469  a in aa_properti
+00000890: 6573 2e69 7465 7272 6f77 7328 293a 0a20  es.iterrows():. 
+000008a0: 2020 2020 2020 2069 6620 6161 5b22 4162         if aa["Ab
+000008b0: 6272 2e22 5d20 696e 2061 615f 7479 7065  br."] in aa_type
+000008c0: 733a 0a20 2020 2020 2020 2020 2020 2061  s:.            a
+000008d0: 615f 6e75 6d62 6572 203d 2061 615f 7479  a_number = aa_ty
+000008e0: 7065 732e 696e 6465 7828 6161 5b22 4162  pes.index(aa["Ab
+000008f0: 6272 2e22 5d29 0a20 2020 2020 2020 2020  br."]).         
+00000900: 2020 2061 615f 696e 666f 5b61 615f 6e75     aa_info[aa_nu
+00000910: 6d62 6572 5d20 3d20 5b66 6c6f 6174 2861  mber] = [float(a
+00000920: 615b 224d 6f6c 6563 756c 6172 5f57 6569  a["Molecular_Wei
+00000930: 6768 7422 5d29 2c66 6c6f 6174 2861 615b  ght"]),float(aa[
+00000940: 2270 4b61 3122 5d29 2c66 6c6f 6174 2861  "pKa1"]),float(a
+00000950: 615b 2270 6c34 225d 295d 0a20 2020 2072  a["pl4"])].    r
+00000960: 6574 7572 6e20 6161 5f69 6e66 6f0a 0a64  eturn aa_info..d
+00000970: 6566 2076 616c 6964 6174 655f 7365 7175  ef validate_sequ
+00000980: 656e 6365 5f61 6c70 6861 6265 7428 7365  ence_alphabet(se
+00000990: 7129 3a0a 2020 2020 2222 220a 2020 2020  q):.    """.    
+000009a0: 4368 6563 6b73 2074 6861 7420 7468 6520  Checks that the 
+000009b0: 7365 7175 656e 6365 7320 6672 6f6d 2061  sequences from a
+000009c0: 6e20 616c 6967 6e6d 656e 7420 6f6e 6c79  n alignment only
+000009d0: 2063 6f6e 7461 696e 7320 7661 6c75 6573   contains values
+000009e0: 2066 726f 6d20 6f6e 6520 6f66 2074 6865   from one of the
+000009f0: 2070 726f 7465 696e 2061 6c70 6861 6265   protein alphabe
+00000a00: 7473 2028 7072 6f74 6569 6e32 3120 6f72  ts (protein21 or
+00000a10: 2070 726f 7465 696e 3231 706c 7573 292e   protein21plus).
+00000a20: 2052 656a 6563 7420 444e 4120 6f72 2052   Reject DNA or R
+00000a30: 4e41 2073 6571 7565 6e63 6573 0a20 2020  NA sequences.   
+00000a40: 203a 7061 7261 6d20 7374 7220 7365 713a   :param str seq:
+00000a50: 2073 6571 7565 6e63 6520 6f66 2063 6861   sequence of cha
+00000a60: 7261 6374 6572 730a 2020 2020 2222 220a  racters.    """.
+00000a70: 2020 2020 616c 7068 6162 6574 7320 3d20      alphabets = 
+00000a80: 7b27 646e 6127 3a20 7265 2e63 6f6d 7069  {'dna': re.compi
+00000a90: 6c65 2827 5e5b 6163 6774 6e5d 2a24 272c  le('^[acgtn]*$',
+00000aa0: 2072 652e 4929 2c0a 2020 2020 2020 2020   re.I),.        
+00000ab0: 2020 2020 2027 7072 6f74 6569 6e32 3127       'protein21'
+00000ac0: 3a20 7265 2e63 6f6d 7069 6c65 2827 5e5b  : re.compile('^[
+00000ad0: 2d61 6364 6566 6768 696b 6c6d 6e70 7172  -acdefghiklmnpqr
+00000ae0: 7374 7677 795d 2a24 272c 2066 6c61 6773  stvwy]*$', flags
+00000af0: 3d72 652e 4947 4e4f 5245 4341 5345 292c  =re.IGNORECASE),
+00000b00: 0a20 2020 2020 2020 2020 2020 2027 7072  .            'pr
+00000b10: 6f74 6569 6e32 3170 6c75 7327 3a20 7265  otein21plus': re
+00000b20: 2e63 6f6d 7069 6c65 2827 5e5b 2d61 6364  .compile('^[-acd
+00000b30: 6566 6768 696b 6c6d 6e70 7172 7374 7677  efghiklmnpqrstvw
+00000b40: 7962 7a78 5d2a 2427 2c20 666c 6167 733d  ybzx]*$', flags=
+00000b50: 2072 652e 4947 4e4f 5245 4341 5345 297d   re.IGNORECASE)}
+00000b60: 0a0a 2020 2020 6966 2061 6c70 6861 6265  ..    if alphabe
+00000b70: 7473 5b22 646e 6122 5d2e 7365 6172 6368  ts["dna"].search
+00000b80: 2873 7472 2873 6571 2929 2069 7320 6e6f  (str(seq)) is no
+00000b90: 7420 4e6f 6e65 3a20 7261 6973 6520 5661  t None: raise Va
+00000ba0: 6c75 6545 7272 6f72 2822 506c 6561 7365  lueError("Please
+00000bb0: 2075 7365 2061 6d69 6e6f 2061 6369 6473   use amino acids
+00000bc0: 2069 6e20 796f 7572 2073 6571 7565 6e63   in your sequenc
+00000bd0: 6573 2c20 6163 6365 7074 6564 2061 6c70  es, accepted alp
+00000be0: 6861 6265 7473 2061 7265 2070 726f 7465  habets are prote
+00000bf0: 696e 3231 3a20 2d61 6364 6566 6768 696b  in21: -acdefghik
+00000c00: 6c6d 6e70 7172 7374 7677 7920 6f72 2070  lmnpqrstvwy or p
+00000c10: 726f 7465 696e 3231 706c 7573 3a20 2d2a  rotein21plus: -*
+00000c20: 6163 6465 6667 6869 6b6c 6d6e 7071 7273  acdefghiklmnpqrs
+00000c30: 7476 7779 627a 7822 290a 0a20 2020 2069  tvwybzx")..    i
+00000c40: 6620 616c 7068 6162 6574 735b 2270 726f  f alphabets["pro
+00000c50: 7465 696e 3231 225d 2e73 6561 7263 6828  tein21"].search(
+00000c60: 7374 7228 7365 7129 2920 6973 206e 6f74  str(seq)) is not
+00000c70: 204e 6f6e 653a 0a20 2020 2020 2020 2061   None:.        a
+00000c80: 615f 7072 6f62 7320 3d20 3231 0a20 2020  a_probs = 21.   
+00000c90: 2020 2020 2072 6574 7572 6e20 6161 5f70       return aa_p
+00000ca0: 726f 6273 0a20 2020 2069 6620 616c 7068  robs.    if alph
+00000cb0: 6162 6574 735b 2270 726f 7465 696e 3231  abets["protein21
+00000cc0: 706c 7573 225d 2e73 6561 7263 6828 7374  plus"].search(st
+00000cd0: 7228 7365 7129 2920 6973 206e 6f74 204e  r(seq)) is not N
+00000ce0: 6f6e 653a 0a20 2020 2020 2020 2061 615f  one:.        aa_
+00000cf0: 7072 6f62 7320 3d20 3234 0a20 2020 2020  probs = 24.     
+00000d00: 2020 2072 6574 7572 6e20 6161 5f70 726f     return aa_pro
+00000d10: 6273 0a20 2020 2065 6c73 653a 0a20 2020  bs.    else:.   
+00000d20: 2020 2020 2072 6169 7365 2056 616c 7565       raise Value
+00000d30: 4572 726f 7228 2259 6f75 7220 7365 7175  Error("Your sequ
+00000d40: 656e 6365 7320 636f 6e74 6169 6e20 6e6f  ences contain no
+00000d50: 7420 616c 6c6f 7765 6420 6368 6172 6163  t allowed charac
+00000d60: 7465 7273 2e20 4176 6169 6c61 626c 6520  ters. Available 
+00000d70: 616c 7068 6162 6574 7320 6172 653a 207b  alphabets are: {
+00000d80: 7072 6f74 6569 6e32 317d 3a20 2d61 6364  protein21}: -acd
+00000d90: 6566 6768 696b 6c6d 6e70 7172 7374 7677  efghiklmnpqrstvw
+00000da0: 7920 6f72 207b 7072 6f74 6569 6e32 3170  y or {protein21p
+00000db0: 6c75 737d 202d 2a61 6364 6566 6768 696b  lus} -*acdefghik
+00000dc0: 6c6d 6e70 7172 7374 7677 7962 7a78 2e20  lmnpqrstvwybzx. 
+00000dd0: 4966 2079 6f75 7220 7365 7175 656e 6365  If your sequence
+00000de0: 2063 6f6e 7461 696e 7320 7374 6f70 2063   contains stop c
+00000df0: 6f64 6f6e 7320 7065 7268 6170 7320 796f  odons perhaps yo
+00000e00: 7520 6361 6e20 7472 696d 2074 6865 6d2e  u can trim them.
+00000e10: 2229 0a0a 6465 6620 616d 696e 6f61 6369  ")..def aminoaci
+00000e20: 645f 6e61 6d65 735f 6469 6374 2861 615f  d_names_dict(aa_
+00000e30: 7072 6f62 7329 3a0a 2020 2020 2222 2220  probs):.    """ 
+00000e40: 5265 7475 726e 7320 616e 2061 6d69 6e6f  Returns an amino
+00000e50: 6163 6964 2061 7373 6f63 6961 7465 6420  acid associated 
+00000e60: 746f 2061 2069 6e74 6567 6572 2076 616c  to a integer val
+00000e70: 7565 0a20 2020 203a 7061 7261 6d20 696e  ue.    :param in
+00000e80: 7420 6161 5f70 726f 6273 3a20 616d 696e  t aa_probs: amin
+00000e90: 6f20 6163 6964 2070 726f 6261 6269 6c69  o acid probabili
+00000ea0: 7469 6573 2c20 7468 6973 206e 756d 6265  ties, this numbe
+00000eb0: 7220 636f 7272 656c 6174 6573 2074 6f20  r correlates to 
+00000ec0: 7468 6520 6e75 6d62 6572 206f 6620 6469  the number of di
+00000ed0: 6666 6572 656e 7420 6161 2074 7970 6573  fferent aa types
+00000ee0: 2069 6e20 7468 6520 696e 7075 7420 616c   in the input al
+00000ef0: 6967 6e6d 656e 7422 2222 0a20 2020 2069  ignment""".    i
+00000f00: 6620 6161 5f70 726f 6273 203d 3d20 3231  f aa_probs == 21
+00000f10: 3a0a 2020 2020 2020 2020 616d 696e 6f61  :.        aminoa
+00000f20: 6369 645f 6e61 6d65 7320 3d20 7b22 2d22  cid_names = {"-"
+00000f30: 3a30 2c22 5222 3a31 2c22 4822 3a32 2c22  :0,"R":1,"H":2,"
+00000f40: 4b22 3a33 2c22 4422 3a34 2c22 4522 3a35  K":3,"D":4,"E":5
+00000f50: 2c22 5322 3a36 2c22 5422 3a37 2c22 4e22  ,"S":6,"T":7,"N"
+00000f60: 3a38 2c22 5122 3a39 2c22 4322 3a31 302c  :8,"Q":9,"C":10,
+00000f70: 2247 223a 3131 2c22 5022 3a31 322c 2241  "G":11,"P":12,"A
+00000f80: 223a 3133 2c22 5622 3a31 342c 2249 223a  ":13,"V":14,"I":
+00000f90: 3135 2c22 4c22 3a31 362c 224d 223a 3137  15,"L":16,"M":17
+00000fa0: 2c22 4622 3a31 382c 2259 223a 3139 2c22  ,"F":18,"Y":19,"
+00000fb0: 5722 3a32 307d 0a20 2020 2020 2020 2072  W":20}.        r
+00000fc0: 6574 7572 6e20 616d 696e 6f61 6369 645f  eturn aminoacid_
+00000fd0: 6e61 6d65 730a 2020 2020 6966 2061 615f  names.    if aa_
+00000fe0: 7072 6f62 7320 3d3d 2032 323a 2023 696e  probs == 22: #in
+00000ff0: 636c 7564 6573 2073 746f 7020 636f 646f  cludes stop codo
+00001000: 6e73 2d2d 2d3e 2066 6978 2069 6e20 4372  ns---> fix in Cr
+00001010: 6561 7465 2062 6c6f 7375 6d0a 2020 2020  eate blosum.    
+00001020: 2020 2020 616d 696e 6f61 6369 645f 6e61      aminoacid_na
+00001030: 6d65 7320 3d20 7b22 2d22 3a30 2c22 2a22  mes = {"-":0,"*"
+00001040: 3a30 2c22 5222 3a31 2c22 4822 3a32 2c22  :0,"R":1,"H":2,"
+00001050: 4b22 3a33 2c22 4422 3a34 2c22 4522 3a35  K":3,"D":4,"E":5
+00001060: 2c22 5322 3a36 2c22 5422 3a37 2c22 4e22  ,"S":6,"T":7,"N"
+00001070: 3a38 2c22 5122 3a39 2c22 4322 3a31 302c  :8,"Q":9,"C":10,
+00001080: 2247 223a 3131 2c22 5022 3a31 322c 2241  "G":11,"P":12,"A
+00001090: 223a 3133 2c22 5622 3a31 342c 2249 223a  ":13,"V":14,"I":
+000010a0: 3135 2c22 4c22 3a31 362c 224d 223a 3137  15,"L":16,"M":17
+000010b0: 2c22 4622 3a31 382c 2259 223a 3139 2c22  ,"F":18,"Y":19,"
+000010c0: 5722 3a32 307d 0a20 2020 2020 2020 2072  W":20}.        r
+000010d0: 6574 7572 6e20 616d 696e 6f61 6369 645f  eturn aminoacid_
+000010e0: 6e61 6d65 730a 2020 2020 656c 6966 2061  names.    elif a
+000010f0: 615f 7072 6f62 7320 3e20 3232 3a0a 2020  a_probs > 22:.  
+00001100: 2020 2020 2020 616d 696e 6f61 6369 645f        aminoacid_
+00001110: 6e61 6d65 7320 3d20 7b22 2d22 3a30 2c22  names = {"-":0,"
+00001120: 5222 3a31 2c22 4822 3a32 2c22 4b22 3a33  R":1,"H":2,"K":3
+00001130: 2c22 4422 3a34 2c22 4522 3a35 2c22 5322  ,"D":4,"E":5,"S"
+00001140: 3a36 2c22 5422 3a37 2c22 4e22 3a38 2c22  :6,"T":7,"N":8,"
+00001150: 5122 3a39 2c22 4322 3a31 302c 2247 223a  Q":9,"C":10,"G":
+00001160: 3131 2c22 5022 3a31 322c 2241 223a 3133  11,"P":12,"A":13
+00001170: 2c22 5622 3a31 342c 2249 223a 3135 2c22  ,"V":14,"I":15,"
+00001180: 4c22 3a31 362c 224d 223a 3137 2c22 4622  L":16,"M":17,"F"
+00001190: 3a31 382c 2259 223a 3139 2c22 5722 3a32  :18,"Y":19,"W":2
+000011a0: 302c 2242 223a 3231 2c22 5a22 3a32 322c  0,"B":21,"Z":22,
+000011b0: 2258 223a 3233 7d0a 2020 2020 2020 2020  "X":23}.        
+000011c0: 7265 7475 726e 2061 6d69 6e6f 6163 6964  return aminoacid
+000011d0: 5f6e 616d 6573 0a64 6566 2063 7265 6174  _names.def creat
+000011e0: 655f 626c 6f73 756d 2861 615f 7072 6f62  e_blosum(aa_prob
+000011f0: 732c 7375 6273 5f6d 6174 7269 785f 6e61  s,subs_matrix_na
+00001200: 6d65 293a 0a20 2020 2022 2222 0a20 2020  me):.    """.   
+00001210: 2042 7569 6c64 7320 616e 2061 7272 6179   Builds an array
+00001220: 2063 6f6e 7461 696e 696e 6720 7468 6520   containing the 
+00001230: 626c 6f73 756d 2073 636f 7265 7320 7065  blosum scores pe
+00001240: 7220 6368 6172 6163 7465 720a 2020 2020  r character.    
+00001250: 3a70 6172 616d 2061 615f 7072 6f62 733a  :param aa_probs:
+00001260: 2061 6d69 6e6f 2061 6369 6420 7072 6f62   amino acid prob
+00001270: 6162 696c 6974 6965 732c 2064 6574 6572  abilities, deter
+00001280: 6d69 6e65 7320 7468 6520 6368 6f69 6365  mines the choice
+00001290: 206f 6620 424c 4f53 554d 206d 6174 7269   of BLOSUM matri
+000012a0: 780a 2020 2020 3a70 6172 616d 2073 7472  x.    :param str
+000012b0: 2073 7562 735f 6d61 7472 6978 5f6e 616d   subs_matrix_nam
+000012c0: 653a 206e 616d 6520 6f66 2074 6865 2073  e: name of the s
+000012d0: 7562 7374 6974 7574 696f 6e20 6d61 7472  ubstitution matr
+000012e0: 6978 2c20 6368 6563 6b20 6176 6169 6c61  ix, check availa
+000012f0: 6269 6c69 7479 2061 7420 2f68 6f6d 652f  bility at /home/
+00001300: 6c79 732f 616e 6163 6f6e 6461 332f 706b  lys/anaconda3/pk
+00001310: 6773 2f62 696f 7079 7468 6f6e 2d31 2e37  gs/biopython-1.7
+00001320: 362d 7079 3337 6835 3136 3930 3961 5f30  6-py37h516909a_0
+00001330: 2f6c 6962 2f70 7974 686f 6e33 2e37 2f73  /lib/python3.7/s
+00001340: 6974 652d 7061 636b 6167 6573 2f42 696f  ite-packages/Bio
+00001350: 2f41 6c69 676e 2f73 7562 7374 6974 7574  /Align/substitut
+00001360: 696f 6e5f 6d61 7472 6963 6573 2f64 6174  ion_matrices/dat
+00001370: 6122 2222 0a0a 2020 2020 6966 2061 615f  a"""..    if aa_
+00001380: 7072 6f62 7320 3e20 3231 2061 6e64 206e  probs > 21 and n
+00001390: 6f74 2073 7562 735f 6d61 7472 6978 5f6e  ot subs_matrix_n
+000013a0: 616d 652e 7374 6172 7473 7769 7468 2822  ame.startswith("
+000013b0: 5041 4d22 293a 0a20 2020 2020 2020 2077  PAM"):.        w
+000013c0: 6172 6e69 6e67 732e 7761 726e 2822 596f  arnings.warn("Yo
+000013d0: 7572 2064 6174 6173 6574 2063 6f6e 7461  ur dataset conta
+000013e0: 696e 7320 7370 6563 6961 6c20 616d 696e  ins special amin
+000013f0: 6f20 6163 6964 732e 2053 7769 7463 6869  o acids. Switchi
+00001400: 6e67 2079 6f75 7220 7375 6273 7469 7475  ng your substitu
+00001410: 7469 6f6e 206d 6174 7269 7820 746f 2050  tion matrix to P
+00001420: 414d 3730 2229 0a20 2020 2020 2020 2073  AM70").        s
+00001430: 7562 735f 6d61 7472 6978 5f6e 616d 6520  ubs_matrix_name 
+00001440: 3d20 2250 414d 3730 220a 2020 2020 7375  = "PAM70".    su
+00001450: 6273 5f6d 6174 7269 7820 3d20 4269 6f2e  bs_matrix = Bio.
+00001460: 416c 6967 6e2e 7375 6273 7469 7475 7469  Align.substituti
+00001470: 6f6e 5f6d 6174 7269 6365 732e 6c6f 6164  on_matrices.load
+00001480: 2873 7562 735f 6d61 7472 6978 5f6e 616d  (subs_matrix_nam
+00001490: 6529 0a20 2020 2061 615f 6c69 7374 203d  e).    aa_list =
+000014a0: 206c 6973 7428 616d 696e 6f61 6369 645f   list(aminoacid_
+000014b0: 6e61 6d65 735f 6469 6374 2861 615f 7072  names_dict(aa_pr
+000014c0: 6f62 7329 2e6b 6579 7328 2929 0a20 2020  obs).keys()).   
+000014d0: 2069 6e64 6578 5f67 6170 203d 2061 615f   index_gap = aa_
+000014e0: 6c69 7374 2e69 6e64 6578 2822 2d22 290a  list.index("-").
+000014f0: 2020 2020 6161 5f6c 6973 745b 696e 6465      aa_list[inde
+00001500: 785f 6761 705d 203d 2022 2a22 2023 696e  x_gap] = "*" #in
+00001510: 2074 6865 2062 6c6f 7375 6d20 6d61 7472   the blosum matr
+00001520: 6978 2067 6170 7320 6172 6520 7265 7072  ix gaps are repr
+00001530: 6573 616e 7465 6420 6173 202a 0a0a 2020  esanted as *..  
+00001540: 2020 7375 6273 5f64 6963 7420 3d20 6465    subs_dict = de
+00001550: 6661 756c 7464 6963 7428 290a 2020 2020  faultdict().    
+00001560: 7375 6273 5f61 7272 6179 203d 206e 702e  subs_array = np.
+00001570: 7a65 726f 7328 286c 656e 2861 615f 6c69  zeros((len(aa_li
+00001580: 7374 2920 2c20 6c65 6e28 6161 5f6c 6973  st) , len(aa_lis
+00001590: 7429 2029 290a 2020 2020 666f 7220 692c  t) )).    for i,
+000015a0: 2061 615f 3120 696e 2065 6e75 6d65 7261   aa_1 in enumera
+000015b0: 7465 2861 615f 6c69 7374 293a 0a20 2020  te(aa_list):.   
+000015c0: 2020 2020 2066 6f72 206a 2c20 6161 5f32       for j, aa_2
+000015d0: 2069 6e20 656e 756d 6572 6174 6528 6161   in enumerate(aa
+000015e0: 5f6c 6973 7429 3a0a 2020 2020 2020 2020  _list):.        
+000015f0: 2020 2020 6966 2061 615f 3120 213d 2022      if aa_1 != "
+00001600: 2a22 2061 6e64 2061 615f 3220 213d 2022  *" and aa_2 != "
+00001610: 2a22 3a0a 2020 2020 2020 2020 2020 2020  *":.            
+00001620: 2020 2020 7375 6273 5f64 6963 745b 2861      subs_dict[(a
+00001630: 615f 312c 6161 5f32 295d 203d 2073 7562  a_1,aa_2)] = sub
+00001640: 735f 6d61 7472 6978 5b28 6161 5f31 2c20  s_matrix[(aa_1, 
+00001650: 6161 5f32 295d 0a20 2020 2020 2020 2020  aa_2)].         
+00001660: 2020 2020 2020 2073 7562 735f 6469 6374         subs_dict
+00001670: 5b28 6161 5f32 2c20 6161 5f31 295d 203d  [(aa_2, aa_1)] =
+00001680: 2073 7562 735f 6d61 7472 6978 5b28 6161   subs_matrix[(aa
+00001690: 5f31 2c20 6161 5f32 295d 0a20 2020 2020  _1, aa_2)].     
+000016a0: 2020 2020 2020 2073 7562 735f 6172 7261         subs_arra
+000016b0: 795b 692c 206a 5d20 3d20 7375 6273 5f6d  y[i, j] = subs_m
+000016c0: 6174 7269 785b 2861 615f 312c 2061 615f  atrix[(aa_1, aa_
+000016d0: 3229 5d0a 2020 2020 2020 2020 2020 2020  2)].            
+000016e0: 7375 6273 5f61 7272 6179 5b6a 2c20 695d  subs_array[j, i]
+000016f0: 203d 2073 7562 735f 6d61 7472 6978 5b28   = subs_matrix[(
+00001700: 6161 5f32 2c20 6161 5f31 295d 0a0a 2020  aa_2, aa_1)]..  
+00001710: 2020 6e61 6d65 7320 3d20 6e70 2e63 6f6e    names = np.con
+00001720: 6361 7465 6e61 7465 2828 6e70 2e61 7272  catenate((np.arr
+00001730: 6179 285b 666c 6f61 7428 222d 696e 6622  ay([float("-inf"
+00001740: 295d 292c 206e 702e 6172 616e 6765 2830  )]), np.arange(0
+00001750: 2c61 615f 7072 6f62 7329 2929 0a20 2020  ,aa_probs))).   
+00001760: 2073 7562 735f 6172 7261 7920 3d20 6e70   subs_array = np
+00001770: 2e63 5f5b 206e 702e 6172 616e 6765 2830  .c_[ np.arange(0
+00001780: 2c61 615f 7072 6f62 7329 2c20 7375 6273  ,aa_probs), subs
+00001790: 5f61 7272 6179 205d 0a20 2020 2073 7562  _array ].    sub
+000017a0: 735f 6172 7261 7920 3d20 6e70 2e63 6f6e  s_array = np.con
+000017b0: 6361 7465 6e61 7465 2828 6e61 6d65 735b  catenate((names[
+000017c0: 4e6f 6e65 2c3a 5d2c 7375 6273 5f61 7272  None,:],subs_arr
+000017d0: 6179 292c 6178 6973 3d30 290a 0a20 2020  ay),axis=0)..   
+000017e0: 2072 6574 7572 6e20 7375 6273 5f61 7272   return subs_arr
+000017f0: 6179 2c20 7375 6273 5f64 6963 740a 6465  ay, subs_dict.de
+00001800: 6620 6469 7669 6465 5f69 6e74 6f5f 6d6f  f divide_into_mo
+00001810: 6e6f 7068 796c 6574 6963 5f63 6c61 6465  nophyletic_clade
+00001820: 7328 7472 6565 2c73 746f 7261 6765 5f66  s(tree,storage_f
+00001830: 6f6c 6465 722c 6e61 6d65 293a 0a20 2020  older,name):.   
+00001840: 2022 2222 0a20 2020 2044 6976 6964 6573   """.    Divides
+00001850: 2074 6865 2074 7265 6520 696e 746f 206d   the tree into m
+00001860: 6f6e 6f70 6879 6c65 7469 6320 636c 6164  onophyletic clad
+00001870: 6573 3a0a 2020 2020 5365 6520 6874 7470  es:.    See http
+00001880: 733a 2f2f 7777 772e 6d75 6e2e 6361 2f62  s://www.mun.ca/b
+00001890: 696f 6c6f 6779 2f73 6361 7272 2f54 6178  iology/scarr/Tax
+000018a0: 6f6e 5f74 7970 6573 2e68 746d 6c0a 2020  on_types.html.  
+000018b0: 2020 496d 706c 656d 656e 7461 7469 6f6e    Implementation
+000018c0: 2062 6173 6564 206f 6e3a 2068 7474 7073   based on: https
+000018d0: 3a2f 2f77 7777 2e62 696f 7374 6172 732e  ://www.biostars.
+000018e0: 6f72 672f 702f 3937 3430 392f 0a0a 2020  org/p/97409/..  
+000018f0: 2020 5468 6520 7265 6173 6f6e 6162 6c65    The reasonable
+00001900: 2063 6c61 6465 2064 6976 6973 696f 6e20   clade division 
+00001910: 6372 6974 6572 6961 2073 6565 6d73 2074  criteria seems t
+00001920: 6f20 6772 6f75 7020 616c 6c20 7468 6f73  o group all thos
+00001930: 6520 6e6f 6465 7320 7768 6f73 6520 6469  e nodes whose di
+00001940: 7374 616e 6365 2074 6f20 7468 6520 696e  stance to the in
+00001950: 7465 726e 616c 2069 7320 6c6f 7765 7220  ternal is lower 
+00001960: 746f 2074 6865 206f 7665 7261 6c6c 2061  to the overall a
+00001970: 7665 7261 6765 2064 6973 7461 6e63 6520  verage distance 
+00001980: 6672 6f6d 2065 6163 6820 6c65 6166 2074  from each leaf t
+00001990: 6f20 7468 6520 726f 6f74 0a0a 2020 2020  o the root..    
+000019a0: 3a70 6172 616d 2065 7465 332d 7472 6565  :param ete3-tree
+000019b0: 2074 7265 653a 2065 7465 3320 7472 6565   tree: ete3 tree
+000019c0: 2063 6c61 7373 0a20 2020 203a 7061 7261   class.    :para
+000019d0: 6d20 7374 7220 7374 6f72 6167 655f 666f  m str storage_fo
+000019e0: 6c64 6572 3a20 666f 6c64 6572 2077 6865  lder: folder whe
+000019f0: 7265 2074 6f20 7374 6f72 6520 7468 6520  re to store the 
+00001a00: 7265 7375 6c74 732c 2069 6e20 7468 6973  results, in this
+00001a10: 2063 6173 6520 6120 6469 6374 696f 6e61   case a dictiona
+00001a20: 7279 2063 6f6e 7461 696e 696e 6720 7b22  ry containing {"
+00001a30: 636c 6164 655f 6e75 6d62 6572 223a 205b  clade_number": [
+00001a40: 6e6f 6465 7320 6c69 7374 5d7d 0a20 2020  nodes list]}.   
+00001a50: 203a 6e61 6d65 2073 7472 206e 616d 653a   :name str name:
+00001a60: 206e 616d 6520 6f66 2074 6865 2064 6174   name of the dat
+00001a70: 6120 7365 7420 7072 6f6a 6563 740a 0a20  a set project.. 
+00001a80: 2020 2022 2222 0a0a 2020 2020 6465 6620     """..    def 
+00001a90: 6d65 616e 2861 7272 6179 293a 0a20 2020  mean(array):.   
+00001aa0: 2020 2020 2022 2222 4361 6c63 756c 6174       """Calculat
+00001ab0: 6573 2062 7261 6e63 6820 6c65 6e67 7468  es branch length
+00001ac0: 2061 7665 7261 6765 2222 220a 2020 2020   average""".    
+00001ad0: 2020 2020 7265 7475 726e 2073 756d 2861      return sum(a
+00001ae0: 7272 6179 2920 2f20 666c 6f61 7428 6c65  rray) / float(le
+00001af0: 6e28 6172 7261 7929 290a 0a20 2020 2064  n(array))..    d
+00001b00: 6566 2063 6163 6865 5f64 6973 7461 6e63  ef cache_distanc
+00001b10: 6573 2874 7265 6529 3a0a 2020 2020 2020  es(tree):.      
+00001b20: 2020 2222 2250 7265 6361 6c63 756c 6174    """Precalculat
+00001b30: 6520 6469 7374 616e 6365 7320 6f66 2061  e distances of a
+00001b40: 6c6c 206e 6f64 6573 2074 6f20 7468 6520  ll nodes to the 
+00001b50: 726f 6f74 2222 220a 2020 2020 2020 2020  root""".        
+00001b60: 6e6f 6465 3272 6f6f 7464 6973 7420 3d20  node2rootdist = 
+00001b70: 7b74 7265 653a 2030 7d0a 2020 2020 2020  {tree: 0}.      
+00001b80: 2020 666f 7220 6e6f 6465 2069 6e20 7472    for node in tr
+00001b90: 6565 2e69 7465 725f 6465 7363 656e 6461  ee.iter_descenda
+00001ba0: 6e74 7328 2770 7265 6f72 6465 7227 293a  nts('preorder'):
+00001bb0: 0a20 2020 2020 2020 2020 2020 206e 6f64  .            nod
+00001bc0: 6532 726f 6f74 6469 7374 5b6e 6f64 655d  e2rootdist[node]
+00001bd0: 203d 206e 6f64 652e 6469 7374 202b 206e   = node.dist + n
+00001be0: 6f64 6532 726f 6f74 6469 7374 5b6e 6f64  ode2rootdist[nod
+00001bf0: 652e 7570 5d0a 2020 2020 2020 2020 7265  e.up].        re
+00001c00: 7475 726e 206e 6f64 6532 726f 6f74 6469  turn node2rootdi
+00001c10: 7374 0a0a 2020 2020 6465 6620 6275 696c  st..    def buil
+00001c20: 645f 636c 6164 6573 2874 7265 652c 6e61  d_clades(tree,na
+00001c30: 6d65 293a 0a20 2020 2020 2020 2022 2222  me):.        """
+00001c40: 5768 656e 2061 2063 6c75 7374 6572 696e  When a clusterin
+00001c50: 6720 636f 6e64 6974 696f 6e20 6973 206d  g condition is m
+00001c60: 6574 2c20 6974 2063 6f6c 6c61 7073 6573  et, it collapses
+00001c70: 2074 6865 2074 7265 6520 6174 2074 6861   the tree at tha
+00001c80: 7420 6e6f 6465 2074 6f20 756e 6966 7920  t node to unify 
+00001c90: 616c 6c20 7468 6520 6c65 6176 6573 2069  all the leaves i
+00001ca0: 6e20 7468 6174 2063 6c75 7374 6572 2069  n that cluster i
+00001cb0: 6e74 6f20 276f 6e65 2720 6c65 6176 6573  nto 'one' leaves
+00001cc0: 2e20 4166 7465 722c 2077 6520 7265 6164  . After, we read
+00001cd0: 0a20 2020 2020 2020 2074 6865 2063 6f6c  .        the col
+00001ce0: 6c61 7073 6564 2074 7265 6520 696e 746f  lapsed tree into
+00001cf0: 2061 2064 6963 7469 6f6e 6172 7920 7468   a dictionary th
+00001d00: 6174 2063 6f6e 7461 696e 7320 7b63 6c61  at contains {cla
+00001d10: 6465 206e 756d 6265 723a 7b22 696e 7465  de number:{"inte
+00001d20: 726e 616c 223a 5b6e 6f64 6573 206e 756d  rnal":[nodes num
+00001d30: 6265 7273 5d2c 226c 6561 7665 7322 3a5b  bers],"leaves":[
+00001d40: 6e6f 6465 206e 756d 6265 7273 5d7d 7d22  node numbers]}}"
+00001d50: 2222 0a20 2020 2020 2020 2023 2063 6163  "".        # cac
+00001d60: 6865 2074 6865 2074 6970 2063 6f6e 7465  he the tip conte
+00001d70: 6e74 206f 6620 6561 6368 206e 6f64 6520  nt of each node 
+00001d80: 746f 2072 6564 7563 6520 7468 6520 6e75  to reduce the nu
+00001d90: 6d62 6572 206f 6620 7469 6d65 7320 7468  mber of times th
+00001da0: 6520 7472 6565 2069 7320 7472 6176 6572  e tree is traver
+00001db0: 7365 640a 2020 2020 2020 2020 6e6f 6465  sed.        node
+00001dc0: 3274 6970 7320 3d20 7472 6565 2e67 6574  2tips = tree.get
+00001dd0: 5f63 6163 6865 645f 636f 6e74 656e 7428  _cached_content(
+00001de0: 290a 2020 2020 2020 2020 726f 6f74 5f64  ).        root_d
+00001df0: 6973 7461 6e63 6520 3d20 6361 6368 655f  istance = cache_
+00001e00: 6469 7374 616e 6365 7328 7472 6565 2920  distances(tree) 
+00001e10: 2023 2064 6973 7461 6e63 6573 206f 6620   # distances of 
+00001e20: 6561 6368 206f 6620 7468 6520 6e6f 6465  each of the node
+00001e30: 7320 746f 2074 6865 2072 6f6f 740a 2020  s to the root.  
+00001e40: 2020 2020 2020 6176 6572 6167 655f 726f        average_ro
+00001e50: 6f74 5f64 6973 7461 6e63 6520 3d20 6d65  ot_distance = me
+00001e60: 616e 2872 6f6f 745f 6469 7374 616e 6365  an(root_distance
+00001e70: 2e76 616c 7565 7328 2929 0a20 2020 2020  .values()).     
+00001e80: 2020 2073 7464 5f72 6f6f 745f 6469 7374     std_root_dist
+00001e90: 616e 6365 203d 2073 7461 7469 7374 6963  ance = statistic
+00001ea0: 732e 7374 6465 7628 726f 6f74 5f64 6973  s.stdev(root_dis
+00001eb0: 7461 6e63 652e 7661 6c75 6573 2829 290a  tance.values()).
+00001ec0: 2020 2020 2020 2020 6e5f 6c65 6176 6573          n_leaves
+00001ed0: 203d 206c 656e 2874 7265 652e 6765 745f   = len(tree.get_
+00001ee0: 6c65 6176 6573 2829 290a 2020 2020 2020  leaves()).      
+00001ef0: 2020 2354 4f44 4f3a 2061 7574 6f6d 6174    #TODO: automat
+00001f00: 697a 6520 636c 7573 7465 7269 6e67 2063  ize clustering c
+00001f10: 6f6e 6469 7469 6f6e 0a20 2020 2020 2020  ondition.       
+00001f20: 2069 6620 6e5f 6c65 6176 6573 203e 3d20   if n_leaves >= 
+00001f30: 3130 3020 6f72 206e 616d 652e 656e 6473  100 or name.ends
+00001f40: 7769 7468 2822 5f73 7562 7472 6565 2229  with("_subtree")
+00001f50: 3a0a 2020 2020 2020 2020 2020 2020 6966  :.            if
+00001f60: 206e 616d 6520 696e 205b 2250 4630 3030   name in ["PF000
+00001f70: 3936 222c 2250 4630 3034 3030 225d 3a0a  96","PF00400"]:.
+00001f80: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00001f90: 636c 7573 7465 7269 6e67 5f63 6f6e 6469  clustering_condi
+00001fa0: 7469 6f6e 203d 2061 7665 7261 6765 5f72  tion = average_r
+00001fb0: 6f6f 745f 6469 7374 616e 6365 2023 2b20  oot_distance #+ 
+00001fc0: 302e 332a 7374 645f 726f 6f74 5f64 6973  0.3*std_root_dis
+00001fd0: 7461 6e63 650a 2020 2020 2020 2020 2020  tance.          
+00001fe0: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
+00001ff0: 2020 2020 2020 2020 636c 7573 7465 7269          clusteri
+00002000: 6e67 5f63 6f6e 6469 7469 6f6e 203d 2061  ng_condition = a
+00002010: 7665 7261 6765 5f72 6f6f 745f 6469 7374  verage_root_dist
+00002020: 616e 6365 202d 322a 7374 645f 726f 6f74  ance -2*std_root
+00002030: 5f64 6973 7461 6e63 650a 2020 2020 2020  _distance.      
+00002040: 2020 656c 6966 206e 616d 6520 696e 205b    elif name in [
+00002050: 2243 6f72 616c 5f61 6c6c 222c 2243 6f72  "Coral_all","Cor
+00002060: 616c 5f46 6176 6969 6e61 222c 2253 4833  al_Faviina","SH3
+00002070: 5f70 6630 3030 3138 5f6c 6172 6765 725f  _pf00018_larger_
+00002080: 7468 616e 5f33 3061 6122 5d20 6f72 2022  than_30aa"] or "
+00002090: 6361 6c63 6974 6f6e 696e 2220 696e 206e  calcitonin" in n
+000020a0: 616d 653a 0a20 2020 2020 2020 2020 2020  ame:.           
+000020b0: 2063 6c75 7374 6572 696e 675f 636f 6e64   clustering_cond
+000020c0: 6974 696f 6e20 3d20 6176 6572 6167 655f  ition = average_
+000020d0: 726f 6f74 5f64 6973 7461 6e63 6520 2d20  root_distance - 
+000020e0: 7374 645f 726f 6f74 5f64 6973 7461 6e63  std_root_distanc
+000020f0: 650a 2020 2020 2020 2020 656c 7365 3a0a  e.        else:.
+00002100: 2020 2020 2020 2020 2020 2020 636c 7573              clus
+00002110: 7465 7269 6e67 5f63 6f6e 6469 7469 6f6e  tering_condition
+00002120: 203d 2061 7665 7261 6765 5f72 6f6f 745f   = average_root_
+00002130: 6469 7374 616e 6365 0a0a 2020 2020 2020  distance..      
+00002140: 2020 666f 7220 6e6f 6465 2069 6e20 7472    for node in tr
+00002150: 6565 2e67 6574 5f64 6573 6365 6e64 616e  ee.get_descendan
+00002160: 7473 2827 7072 656f 7264 6572 2729 3a0a  ts('preorder'):.
+00002170: 2020 2020 2020 2020 2020 2020 6966 206e              if n
+00002180: 6f74 206e 6f64 652e 6973 5f6c 6561 6628  ot node.is_leaf(
+00002190: 293a 2020 2320 666f 7220 696e 7465 726e  ):  # for intern
+000021a0: 616c 206e 6f64 6573 0a20 2020 2020 2020  al nodes.       
+000021b0: 2020 2020 2020 2020 2061 7667 5f64 6973           avg_dis
+000021c0: 7461 6e63 655f 746f 5f74 6970 7320 3d20  tance_to_tips = 
+000021d0: 6d65 616e 285b 726f 6f74 5f64 6973 7461  mean([root_dista
+000021e0: 6e63 655b 7469 705d 202d 2072 6f6f 745f  nce[tip] - root_
+000021f0: 6469 7374 616e 6365 5b6e 6f64 655d 2066  distance[node] f
+00002200: 6f72 2074 6970 2069 6e20 6e6f 6465 3274  or tip in node2t
+00002210: 6970 735b 6e6f 6465 5d5d 2920 2023 2061  ips[node]])  # a
+00002220: 7665 7261 6765 2064 6973 7461 6e63 6520  verage distance 
+00002230: 6672 6f6d 2074 6865 2069 6e74 6572 6e61  from the interna
+00002240: 6c20 6e6f 6465 2074 6f20 616c 6c20 6974  l node to all it
+00002250: 2773 2070 6f73 7369 626c 6520 6465 7269  's possible deri
+00002260: 7665 6420 6c65 6176 6573 0a20 2020 2020  ved leaves.     
+00002270: 2020 2020 2020 2020 2020 2069 6620 6176             if av
+00002280: 675f 6469 7374 616e 6365 5f74 6f5f 7469  g_distance_to_ti
+00002290: 7073 203c 2063 6c75 7374 6572 696e 675f  ps < clustering_
+000022a0: 636f 6e64 6974 696f 6e3a 0a20 2020 2020  condition:.     
+000022b0: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+000022c0: 6e6f 6465 2e6e 616d 6520 2b3d 2027 2043  node.name += ' C
+000022d0: 4f4c 4c41 5053 4544 2061 7667 5f64 3a25  OLLAPSED avg_d:%
+000022e0: 6720 7b25 737d 2720 2520 2861 7667 5f64  g {%s}' % (avg_d
+000022f0: 6973 7461 6e63 655f 746f 5f74 6970 732c  istance_to_tips,
+00002300: 2027 2c27 2e6a 6f69 6e28 5b74 6970 2e6e   ','.join([tip.n
+00002310: 616d 6520 666f 7220 7469 7020 696e 206e  ame for tip in n
+00002320: 6f64 6532 7469 7073 5b6e 6f64 655d 5d29  ode2tips[node]])
+00002330: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+00002340: 2020 2020 2020 6e6f 6465 2e6e 616d 6520        node.name 
+00002350: 2b3d 2027 2043 4f4c 4c41 5053 4544 2061  += ' COLLAPSED a
+00002360: 7667 5f64 3a25 6720 6c65 6176 6573 3a7b  vg_d:%g leaves:{
+00002370: 2573 7d20 696e 7465 726e 616c 3a7b 2573  %s} internal:{%s
+00002380: 7d27 2025 2028 6176 675f 6469 7374 616e  }' % (avg_distan
+00002390: 6365 5f74 6f5f 7469 7073 2c20 272c 272e  ce_to_tips, ','.
+000023a0: 6a6f 696e 285b 7469 702e 6e61 6d65 2066  join([tip.name f
+000023b0: 6f72 2074 6970 2069 6e20 6e6f 6465 3274  or tip in node2t
+000023c0: 6970 735b 6e6f 6465 5d5d 292c 272c 272e  ips[node]]),','.
+000023d0: 6a6f 696e 285b 696e 7465 726e 616c 2e6e  join([internal.n
+000023e0: 616d 6520 666f 7220 696e 7465 726e 616c  ame for internal
+000023f0: 2069 6e20 6e6f 6465 2e69 7465 725f 6465   in node.iter_de
+00002400: 7363 656e 6461 6e74 7328 2920 6966 206e  scendants() if n
+00002410: 6f74 2069 6e74 6572 6e61 6c2e 6973 5f6c  ot internal.is_l
+00002420: 6561 6628 295d 2929 0a20 2020 2020 2020  eaf()])).       
+00002430: 2020 2020 2020 2020 2020 2020 206e 6f64               nod
+00002440: 652e 6164 645f 6665 6174 7572 6573 2863  e.add_features(c
+00002450: 6f6c 6c61 7073 6564 3d54 7275 6529 0a20  ollapsed=True). 
 00002460: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002470: 2020 6e6f 6465 2e69 6d67 5f73 7479 6c65    node.img_style
-00002480: 5b27 6472 6177 5f64 6573 6365 6e64 616e  ['draw_descendan
-00002490: 7473 275d 203d 2046 616c 7365 0a0a 2020  ts'] = False..  
-000024a0: 2020 2020 2020 666f 7220 6e20 696e 2074        for n in t
-000024b0: 7265 652e 7365 6172 6368 5f6e 6f64 6573  ree.search_nodes
-000024c0: 2863 6f6c 6c61 7073 6564 3d54 7275 6529  (collapsed=True)
-000024d0: 3a0a 2020 2020 2020 2020 2020 2020 666f  :.            fo
-000024e0: 7220 6368 696c 6420 696e 206e 2e67 6574  r child in n.get
-000024f0: 5f63 6869 6c64 7265 6e28 293a 0a20 2020  _children():.   
-00002500: 2020 2020 2020 2020 2020 2020 2063 6869               chi
-00002510: 6c64 2e64 6574 6163 6828 290a 2020 2020  ld.detach().    
-00002520: 2020 2020 2370 7269 6e74 2874 7265 652e      #print(tree.
-00002530: 6765 745f 6173 6369 6928 7368 6f77 5f69  get_ascii(show_i
-00002540: 6e74 6572 6e61 6c3d 5472 7565 2929 0a20  nternal=True)). 
-00002550: 2020 2020 2020 2069 203d 2030 0a20 2020         i = 0.   
-00002560: 2020 2020 2063 6c61 6465 5f64 6963 745f       clade_dict_
-00002570: 616c 6c20 3d20 6465 6661 756c 7464 6963  all = defaultdic
-00002580: 7428 6c61 6d62 6461 3a20 6465 6661 756c  t(lambda: defaul
-00002590: 7464 6963 7428 2929 0a20 2020 2020 2020  tdict()).       
-000025a0: 2066 6f72 206e 2069 6e20 7472 6565 2e74   for n in tree.t
-000025b0: 7261 7665 7273 6528 293a 0a20 2020 2020  raverse():.     
-000025c0: 2020 2020 2020 2020 2020 2069 6620 6e2e             if n.
-000025d0: 6973 5f6c 6561 6628 2920 616e 6420 2243  is_leaf() and "C
-000025e0: 4f4c 4c41 5053 4544 2220 696e 206e 2e6e  OLLAPSED" in n.n
-000025f0: 616d 653a 2023 636f 6c6c 6170 7365 6420  ame: #collapsed 
-00002600: 6c65 6166 2028 6974 2069 7320 6120 636c  leaf (it is a cl
-00002610: 6164 6520 6f6e 2069 7427 7320 6f77 6e29  ade on it's own)
-00002620: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00002630: 2020 2020 2063 6c61 6465 5f6e 616d 6573       clade_names
-00002640: 5f6c 6561 7665 7320 3d20 6e2e 6e61 6d65  _leaves = n.name
-00002650: 5b6e 2e6e 616d 652e 6669 6e64 2822 6c65  [n.name.find("le
-00002660: 6176 6573 3a7b 2229 202b 2038 3a6e 2e6e  aves:{") + 8:n.n
-00002670: 616d 652e 6669 6e64 2822 7d22 295d 2e73  ame.find("}")].s
-00002680: 706c 6974 2822 2c22 290a 2020 2020 2020  plit(",").      
-00002690: 2020 2020 2020 2020 2020 2020 2020 636c                cl
-000026a0: 6164 655f 6e61 6d65 735f 696e 7465 726e  ade_names_intern
-000026b0: 616c 203d 205b 6e2e 6e61 6d65 2e73 706c  al = [n.name.spl
-000026c0: 6974 2822 2022 295b 305d 5d0a 2020 2020  it(" ")[0]].    
+00002470: 2020 206e 6f64 652e 696d 675f 7374 796c     node.img_styl
+00002480: 655b 2764 7261 775f 6465 7363 656e 6461  e['draw_descenda
+00002490: 6e74 7327 5d20 3d20 4661 6c73 650a 0a20  nts'] = False.. 
+000024a0: 2020 2020 2020 2066 6f72 206e 2069 6e20         for n in 
+000024b0: 7472 6565 2e73 6561 7263 685f 6e6f 6465  tree.search_node
+000024c0: 7328 636f 6c6c 6170 7365 643d 5472 7565  s(collapsed=True
+000024d0: 293a 0a20 2020 2020 2020 2020 2020 2066  ):.            f
+000024e0: 6f72 2063 6869 6c64 2069 6e20 6e2e 6765  or child in n.ge
+000024f0: 745f 6368 696c 6472 656e 2829 3a0a 2020  t_children():.  
+00002500: 2020 2020 2020 2020 2020 2020 2020 6368                ch
+00002510: 696c 642e 6465 7461 6368 2829 0a20 2020  ild.detach().   
+00002520: 2020 2020 2023 7072 696e 7428 7472 6565       #print(tree
+00002530: 2e67 6574 5f61 7363 6969 2873 686f 775f  .get_ascii(show_
+00002540: 696e 7465 726e 616c 3d54 7275 6529 290a  internal=True)).
+00002550: 2020 2020 2020 2020 6920 3d20 300a 2020          i = 0.  
+00002560: 2020 2020 2020 636c 6164 655f 6469 6374        clade_dict
+00002570: 5f61 6c6c 203d 2064 6566 6175 6c74 6469  _all = defaultdi
+00002580: 6374 286c 616d 6264 613a 2064 6566 6175  ct(lambda: defau
+00002590: 6c74 6469 6374 2829 290a 2020 2020 2020  ltdict()).      
+000025a0: 2020 666f 7220 6e20 696e 2074 7265 652e    for n in tree.
+000025b0: 7472 6176 6572 7365 2829 3a0a 2020 2020  traverse():.    
+000025c0: 2020 2020 2020 2020 2020 2020 6966 206e              if n
+000025d0: 2e69 735f 6c65 6166 2829 2061 6e64 2022  .is_leaf() and "
+000025e0: 434f 4c4c 4150 5345 4422 2069 6e20 6e2e  COLLAPSED" in n.
+000025f0: 6e61 6d65 3a20 2363 6f6c 6c61 7073 6564  name: #collapsed
+00002600: 206c 6561 6620 2869 7420 6973 2061 2063   leaf (it is a c
+00002610: 6c61 6465 206f 6e20 6974 2773 206f 776e  lade on it's own
+00002620: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+00002630: 2020 2020 2020 636c 6164 655f 6e61 6d65        clade_name
+00002640: 735f 6c65 6176 6573 203d 206e 2e6e 616d  s_leaves = n.nam
+00002650: 655b 6e2e 6e61 6d65 2e66 696e 6428 226c  e[n.name.find("l
+00002660: 6561 7665 733a 7b22 2920 2b20 383a 6e2e  eaves:{") + 8:n.
+00002670: 6e61 6d65 2e66 696e 6428 227d 2229 5d2e  name.find("}")].
+00002680: 7370 6c69 7428 222c 2229 0a20 2020 2020  split(",").     
+00002690: 2020 2020 2020 2020 2020 2020 2020 2063                 c
+000026a0: 6c61 6465 5f6e 616d 6573 5f69 6e74 6572  lade_names_inter
+000026b0: 6e61 6c20 3d20 5b6e 2e6e 616d 652e 7370  nal = [n.name.sp
+000026c0: 6c69 7428 2220 2229 5b30 5d5d 0a20 2020  lit(" ")[0]].   
 000026d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000026e0: 636c 6164 655f 6e61 6d65 735f 696e 7465  clade_names_inte
-000026f0: 726e 616c 202b 3d20 6e2e 6e61 6d65 5b6e  rnal += n.name[n
-00002700: 2e6e 616d 652e 6669 6e64 2822 696e 7465  .name.find("inte
-00002710: 726e 616c 3a7b 2229 202b 2031 303a 5d2e  rnal:{") + 10:].
-00002720: 7374 7269 7028 227d 2229 2e73 706c 6974  strip("}").split
-00002730: 2822 2c22 290a 2020 2020 2020 2020 2020  (",").          
-00002740: 2020 2020 2020 2020 2020 636c 6164 655f            clade_
-00002750: 6469 6374 5f61 6c6c 5b22 436c 6164 655f  dict_all["Clade_
-00002760: 7b7d 222e 666f 726d 6174 2869 295d 5b22  {}".format(i)]["
-00002770: 6c65 6176 6573 225d 203d 2073 6574 2863  leaves"] = set(c
-00002780: 6c61 6465 5f6e 616d 6573 5f6c 6561 7665  lade_names_leave
-00002790: 7329 2020 2320 7265 6d6f 7665 2064 7570  s)  # remove dup
-000027a0: 6c69 6361 7465 730a 2020 2020 2020 2020  licates.        
-000027b0: 2020 2020 2020 2020 2020 2020 636c 6164              clad
-000027c0: 655f 6469 6374 5f61 6c6c 5b22 436c 6164  e_dict_all["Clad
-000027d0: 655f 7b7d 222e 666f 726d 6174 2869 295d  e_{}".format(i)]
-000027e0: 5b22 696e 7465 726e 616c 225d 203d 206c  ["internal"] = l
-000027f0: 6973 7428 6669 6c74 6572 284e 6f6e 652c  ist(filter(None,
-00002800: 7365 7428 636c 6164 655f 6e61 6d65 735f  set(clade_names_
-00002810: 696e 7465 726e 616c 2929 2920 2373 6f6d  internal))) #som
-00002820: 6574 696d 6573 2074 6865 2020 6e6f 6465  etimes the  node
-00002830: 2073 7472 696e 6773 2061 7265 2065 6d70   strings are emp
-00002840: 7479 0a20 2020 2020 2020 2020 2020 2020  ty.             
-00002850: 2020 2020 2020 2069 202b 3d20 310a 2020         i += 1.  
-00002860: 2020 2020 2020 2020 2020 2020 2020 656c                el
-00002870: 6966 206e 6f74 206e 2e69 735f 6c65 6166  if not n.is_leaf
-00002880: 2829 3a20 2369 6620 7468 6520 6e6f 6465  (): #if the node
-00002890: 2069 7320 696e 7465 726e 616c 0a20 2020   is internal.   
+000026e0: 2063 6c61 6465 5f6e 616d 6573 5f69 6e74   clade_names_int
+000026f0: 6572 6e61 6c20 2b3d 206e 2e6e 616d 655b  ernal += n.name[
+00002700: 6e2e 6e61 6d65 2e66 696e 6428 2269 6e74  n.name.find("int
+00002710: 6572 6e61 6c3a 7b22 2920 2b20 3130 3a5d  ernal:{") + 10:]
+00002720: 2e73 7472 6970 2822 7d22 292e 7370 6c69  .strip("}").spli
+00002730: 7428 222c 2229 0a20 2020 2020 2020 2020  t(",").         
+00002740: 2020 2020 2020 2020 2020 2063 6c61 6465             clade
+00002750: 5f64 6963 745f 616c 6c5b 2243 6c61 6465  _dict_all["Clade
+00002760: 5f7b 7d22 2e66 6f72 6d61 7428 6929 5d5b  _{}".format(i)][
+00002770: 226c 6561 7665 7322 5d20 3d20 7365 7428  "leaves"] = set(
+00002780: 636c 6164 655f 6e61 6d65 735f 6c65 6176  clade_names_leav
+00002790: 6573 2920 2023 2072 656d 6f76 6520 6475  es)  # remove du
+000027a0: 706c 6963 6174 6573 0a20 2020 2020 2020  plicates.       
+000027b0: 2020 2020 2020 2020 2020 2020 2063 6c61               cla
+000027c0: 6465 5f64 6963 745f 616c 6c5b 2243 6c61  de_dict_all["Cla
+000027d0: 6465 5f7b 7d22 2e66 6f72 6d61 7428 6929  de_{}".format(i)
+000027e0: 5d5b 2269 6e74 6572 6e61 6c22 5d20 3d20  ]["internal"] = 
+000027f0: 6c69 7374 2866 696c 7465 7228 4e6f 6e65  list(filter(None
+00002800: 2c73 6574 2863 6c61 6465 5f6e 616d 6573  ,set(clade_names
+00002810: 5f69 6e74 6572 6e61 6c29 2929 2023 736f  _internal))) #so
+00002820: 6d65 7469 6d65 7320 7468 6520 206e 6f64  metimes the  nod
+00002830: 6520 7374 7269 6e67 7320 6172 6520 656d  e strings are em
+00002840: 7074 790a 2020 2020 2020 2020 2020 2020  pty.            
+00002850: 2020 2020 2020 2020 6920 2b3d 2031 0a20          i += 1. 
+00002860: 2020 2020 2020 2020 2020 2020 2020 2065                 e
+00002870: 6c69 6620 6e6f 7420 6e2e 6973 5f6c 6561  lif not n.is_lea
+00002880: 6628 293a 2023 6966 2074 6865 206e 6f64  f(): #if the nod
+00002890: 6520 6973 2069 6e74 6572 6e61 6c0a 2020  e is internal.  
 000028a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000028b0: 2063 6c61 6465 5f6e 616d 6573 5f6c 6561   clade_names_lea
-000028c0: 7665 7320 3d20 5b5d 0a20 2020 2020 2020  ves = [].       
-000028d0: 2020 2020 2020 2020 2020 2020 2063 6c61               cla
-000028e0: 6465 5f6e 616d 6573 5f69 6e74 6572 6e61  de_names_interna
-000028f0: 6c20 3d20 5b5d 0a20 2020 2020 2020 2020  l = [].         
-00002900: 2020 2020 2020 2020 2020 2063 6c61 6465             clade
-00002910: 5f6e 616d 6573 5f69 6e74 6572 6e61 6c20  _names_internal 
-00002920: 2b3d 205b 6e2e 6e61 6d65 5d0a 2020 2020  += [n.name].    
+000028b0: 2020 636c 6164 655f 6e61 6d65 735f 6c65    clade_names_le
+000028c0: 6176 6573 203d 205b 5d0a 2020 2020 2020  aves = [].      
+000028d0: 2020 2020 2020 2020 2020 2020 2020 636c                cl
+000028e0: 6164 655f 6e61 6d65 735f 696e 7465 726e  ade_names_intern
+000028f0: 616c 203d 205b 5d0a 2020 2020 2020 2020  al = [].        
+00002900: 2020 2020 2020 2020 2020 2020 636c 6164              clad
+00002910: 655f 6e61 6d65 735f 696e 7465 726e 616c  e_names_internal
+00002920: 202b 3d20 5b6e 2e6e 616d 655d 0a20 2020   += [n.name].   
 00002930: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002940: 666f 7220 6465 7363 656e 6461 6e74 2069  for descendant i
-00002950: 6e20 6e2e 6974 6572 5f64 6573 6365 6e64  n n.iter_descend
-00002960: 616e 7473 2829 3a0a 2020 2020 2020 2020  ants():.        
+00002940: 2066 6f72 2064 6573 6365 6e64 616e 7420   for descendant 
+00002950: 696e 206e 2e69 7465 725f 6465 7363 656e  in n.iter_descen
+00002960: 6461 6e74 7328 293a 0a20 2020 2020 2020  dants():.       
 00002970: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002980: 6966 2064 6573 6365 6e64 616e 742e 6973  if descendant.is
-00002990: 5f6c 6561 6628 293a 0a20 2020 2020 2020  _leaf():.       
+00002980: 2069 6620 6465 7363 656e 6461 6e74 2e69   if descendant.i
+00002990: 735f 6c65 6166 2829 3a0a 2020 2020 2020  s_leaf():.      
 000029a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000029b0: 2020 2020 2069 6620 227b 2220 6e6f 7420       if "{" not 
-000029c0: 696e 2064 6573 6365 6e64 616e 742e 6e61  in descendant.na
-000029d0: 6d65 3a20 2369 7420 7761 7320 6120 7075  me: #it was a pu
-000029e0: 7265 206c 6561 660a 2020 2020 2020 2020  re leaf.        
+000029b0: 2020 2020 2020 6966 2022 7b22 206e 6f74        if "{" not
+000029c0: 2069 6e20 6465 7363 656e 6461 6e74 2e6e   in descendant.n
+000029d0: 616d 653a 2023 6974 2077 6173 2061 2070  ame: #it was a p
+000029e0: 7572 6520 6c65 6166 0a20 2020 2020 2020  ure leaf.       
 000029f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002a00: 2020 2020 2020 2020 636c 6164 655f 6e61          clade_na
-00002a10: 6d65 735f 6c65 6176 6573 202b 3d20 5b64  mes_leaves += [d
-00002a20: 6573 6365 6e64 616e 742e 6e61 6d65 5d0a  escendant.name].
-00002a30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002a40: 2020 2020 2020 2020 2020 2020 656c 7365              else
-00002a50: 3a20 2369 7320 6120 636f 6c6c 6170 7365  : #is a collapse
-00002a60: 6420 6c65 6176 650a 2020 2020 2020 2020  d leave.        
+00002a00: 2020 2020 2020 2020 2063 6c61 6465 5f6e           clade_n
+00002a10: 616d 6573 5f6c 6561 7665 7320 2b3d 205b  ames_leaves += [
+00002a20: 6465 7363 656e 6461 6e74 2e6e 616d 655d  descendant.name]
+00002a30: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00002a40: 2020 2020 2020 2020 2020 2020 2065 6c73               els
+00002a50: 653a 2023 6973 2061 2063 6f6c 6c61 7073  e: #is a collaps
+00002a60: 6564 206c 6561 7665 0a20 2020 2020 2020  ed leave.       
 00002a70: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002a80: 2020 2020 2020 2020 636c 6164 655f 6e61          clade_na
-00002a90: 6d65 735f 6c65 6176 6573 202b 3d20 6465  mes_leaves += de
-00002aa0: 7363 656e 6461 6e74 2e6e 616d 655b 6465  scendant.name[de
-00002ab0: 7363 656e 6461 6e74 2e6e 616d 652e 6669  scendant.name.fi
-00002ac0: 6e64 2822 6c65 6176 6573 3a7b 2229 2b38  nd("leaves:{")+8
-00002ad0: 3a64 6573 6365 6e64 616e 742e 6e61 6d65  :descendant.name
-00002ae0: 2e66 696e 6428 227d 2229 5d2e 7370 6c69  .find("}")].spli
-00002af0: 7428 222c 2229 0a20 2020 2020 2020 2020  t(",").         
+00002a80: 2020 2020 2020 2020 2063 6c61 6465 5f6e           clade_n
+00002a90: 616d 6573 5f6c 6561 7665 7320 2b3d 2064  ames_leaves += d
+00002aa0: 6573 6365 6e64 616e 742e 6e61 6d65 5b64  escendant.name[d
+00002ab0: 6573 6365 6e64 616e 742e 6e61 6d65 2e66  escendant.name.f
+00002ac0: 696e 6428 226c 6561 7665 733a 7b22 292b  ind("leaves:{")+
+00002ad0: 383a 6465 7363 656e 6461 6e74 2e6e 616d  8:descendant.nam
+00002ae0: 652e 6669 6e64 2822 7d22 295d 2e73 706c  e.find("}")].spl
+00002af0: 6974 2822 2c22 290a 2020 2020 2020 2020  it(",").        
 00002b00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002b10: 2020 2020 2020 2063 6c61 6465 5f6e 616d         clade_nam
-00002b20: 6573 5f69 6e74 6572 6e61 6c20 2b3d 205b  es_internal += [
-00002b30: 6465 7363 656e 6461 6e74 2e6e 616d 652e  descendant.name.
-00002b40: 7370 6c69 7428 2220 2229 5b30 5d5d 0a20  split(" ")[0]]. 
+00002b10: 2020 2020 2020 2020 636c 6164 655f 6e61          clade_na
+00002b20: 6d65 735f 696e 7465 726e 616c 202b 3d20  mes_internal += 
+00002b30: 5b64 6573 6365 6e64 616e 742e 6e61 6d65  [descendant.name
+00002b40: 2e73 706c 6974 2822 2022 295b 305d 5d0a  .split(" ")[0]].
 00002b50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002b60: 2020 2020 2020 2020 2020 2020 2020 2063                 c
-00002b70: 6c61 6465 5f6e 616d 6573 5f69 6e74 6572  lade_names_inter
-00002b80: 6e61 6c20 2b3d 2064 6573 6365 6e64 616e  nal += descendan
-00002b90: 742e 6e61 6d65 5b64 6573 6365 6e64 616e  t.name[descendan
-00002ba0: 742e 6e61 6d65 2e66 696e 6428 2269 6e74  t.name.find("int
-00002bb0: 6572 6e61 6c3a 7b22 292b 3130 3a5d 2e73  ernal:{")+10:].s
-00002bc0: 7472 6970 2822 7d22 292e 7370 6c69 7428  trip("}").split(
-00002bd0: 222c 2229 0a20 2020 2020 2020 2020 2020  ",").           
-00002be0: 2020 2020 2020 2020 2020 2020 2065 6c73               els
-00002bf0: 653a 2023 6164 6420 7468 6520 696e 7465  e: #add the inte
-00002c00: 726e 616c 206e 6f64 6520 616c 736f 2074  rnal node also t
-00002c10: 6f20 6974 2773 2063 6c61 6465 0a20 2020  o it's clade.   
+00002b60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00002b70: 636c 6164 655f 6e61 6d65 735f 696e 7465  clade_names_inte
+00002b80: 726e 616c 202b 3d20 6465 7363 656e 6461  rnal += descenda
+00002b90: 6e74 2e6e 616d 655b 6465 7363 656e 6461  nt.name[descenda
+00002ba0: 6e74 2e6e 616d 652e 6669 6e64 2822 696e  nt.name.find("in
+00002bb0: 7465 726e 616c 3a7b 2229 2b31 303a 5d2e  ternal:{")+10:].
+00002bc0: 7374 7269 7028 227d 2229 2e73 706c 6974  strip("}").split
+00002bd0: 2822 2c22 290a 2020 2020 2020 2020 2020  (",").          
+00002be0: 2020 2020 2020 2020 2020 2020 2020 656c                el
+00002bf0: 7365 3a20 2361 6464 2074 6865 2069 6e74  se: #add the int
+00002c00: 6572 6e61 6c20 6e6f 6465 2061 6c73 6f20  ernal node also 
+00002c10: 746f 2069 7427 7320 636c 6164 650a 2020  to it's clade.  
 00002c20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002c30: 2020 2020 2020 2020 2063 6c61 6465 5f6e           clade_n
-00002c40: 616d 6573 5f69 6e74 6572 6e61 6c20 2b3d  ames_internal +=
-00002c50: 205b 6465 7363 656e 6461 6e74 2e6e 616d   [descendant.nam
-00002c60: 655d 0a0a 2020 2020 2020 2020 2020 2020  e]..            
-00002c70: 2020 2020 2020 2020 636c 6164 655f 6469          clade_di
-00002c80: 6374 5f61 6c6c 5b22 436c 6164 655f 7b7d  ct_all["Clade_{}
-00002c90: 222e 666f 726d 6174 2869 295d 5b22 6c65  ".format(i)]["le
-00002ca0: 6176 6573 225d 203d 2073 6574 2863 6c61  aves"] = set(cla
-00002cb0: 6465 5f6e 616d 6573 5f6c 6561 7665 7329  de_names_leaves)
-00002cc0: 2023 7265 6d6f 7665 2064 7570 6c69 6361   #remove duplica
-00002cd0: 7465 730a 2020 2020 2020 2020 2020 2020  tes.            
-00002ce0: 2020 2020 2020 2020 636c 6164 655f 6469          clade_di
-00002cf0: 6374 5f61 6c6c 5b22 436c 6164 655f 7b7d  ct_all["Clade_{}
-00002d00: 222e 666f 726d 6174 2869 295d 5b22 696e  ".format(i)]["in
-00002d10: 7465 726e 616c 225d 203d 206c 6973 7428  ternal"] = list(
-00002d20: 6669 6c74 6572 284e 6f6e 652c 7365 7428  filter(None,set(
-00002d30: 636c 6164 655f 6e61 6d65 735f 696e 7465  clade_names_inte
-00002d40: 726e 616c 2929 2920 2373 6f6d 6574 696d  rnal))) #sometim
-00002d50: 6573 2074 6865 2020 6e6f 6465 2073 7472  es the  node str
-00002d60: 696e 6773 2061 7265 2065 6d70 7479 0a20  ings are empty. 
+00002c30: 2020 2020 2020 2020 2020 636c 6164 655f            clade_
+00002c40: 6e61 6d65 735f 696e 7465 726e 616c 202b  names_internal +
+00002c50: 3d20 5b64 6573 6365 6e64 616e 742e 6e61  = [descendant.na
+00002c60: 6d65 5d0a 0a20 2020 2020 2020 2020 2020  me]..           
+00002c70: 2020 2020 2020 2020 2063 6c61 6465 5f64           clade_d
+00002c80: 6963 745f 616c 6c5b 2243 6c61 6465 5f7b  ict_all["Clade_{
+00002c90: 7d22 2e66 6f72 6d61 7428 6929 5d5b 226c  }".format(i)]["l
+00002ca0: 6561 7665 7322 5d20 3d20 7365 7428 636c  eaves"] = set(cl
+00002cb0: 6164 655f 6e61 6d65 735f 6c65 6176 6573  ade_names_leaves
+00002cc0: 2920 2372 656d 6f76 6520 6475 706c 6963  ) #remove duplic
+00002cd0: 6174 6573 0a20 2020 2020 2020 2020 2020  ates.           
+00002ce0: 2020 2020 2020 2020 2063 6c61 6465 5f64           clade_d
+00002cf0: 6963 745f 616c 6c5b 2243 6c61 6465 5f7b  ict_all["Clade_{
+00002d00: 7d22 2e66 6f72 6d61 7428 6929 5d5b 2269  }".format(i)]["i
+00002d10: 6e74 6572 6e61 6c22 5d20 3d20 6c69 7374  nternal"] = list
+00002d20: 2866 696c 7465 7228 4e6f 6e65 2c73 6574  (filter(None,set
+00002d30: 2863 6c61 6465 5f6e 616d 6573 5f69 6e74  (clade_names_int
+00002d40: 6572 6e61 6c29 2929 2023 736f 6d65 7469  ernal))) #someti
+00002d50: 6d65 7320 7468 6520 206e 6f64 6520 7374  mes the  node st
+00002d60: 7269 6e67 7320 6172 6520 656d 7074 790a  rings are empty.
 00002d70: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002d80: 2020 2069 202b 3d20 310a 2020 2020 2020     i += 1.      
-00002d90: 2020 2020 2020 2020 2020 656c 7365 3a23            else:#
-00002da0: 4e6f 6e20 636f 6c6c 6170 7365 6420 6c65  Non collapsed le
-00002db0: 6176 6573 0a20 2020 2020 2020 2020 2020  aves.           
-00002dc0: 2020 2020 2020 2020 2070 6173 730a 2020           pass.  
-00002dd0: 2020 2020 2020 636c 6164 655f 6469 6374        clade_dict
-00002de0: 5f6c 6561 7665 7320 3d20 6465 6661 756c  _leaves = defaul
-00002df0: 7464 6963 7428 290a 2020 2020 2020 2020  tdict().        
-00002e00: 6920 3d20 300a 2020 2020 2020 2020 666f  i = 0.        fo
-00002e10: 7220 6e20 696e 2074 7265 652e 7472 6176  r n in tree.trav
-00002e20: 6572 7365 2822 7072 656f 7264 6572 2229  erse("preorder")
-00002e30: 3a0a 2020 2020 2020 2020 2020 2020 6966  :.            if
-00002e40: 206e 2e69 735f 6c65 6166 2829 3a0a 2020   n.is_leaf():.  
-00002e50: 2020 2020 2020 2020 2020 2020 2020 6966                if
-00002e60: 2022 7b22 206e 6f74 2069 6e20 6e2e 6e61   "{" not in n.na
-00002e70: 6d65 3a0a 2020 2020 2020 2020 2020 2020  me:.            
-00002e80: 2020 2020 2020 2020 636c 6164 655f 6e61          clade_na
-00002e90: 6d65 735f 6c65 6176 6573 203d 205b 6e2e  mes_leaves = [n.
-00002ea0: 6e61 6d65 5d0a 2020 2020 2020 2020 2020  name].          
-00002eb0: 2020 2020 2020 656c 7365 3a0a 2020 2020        else:.    
+00002d80: 2020 2020 6920 2b3d 2031 0a20 2020 2020      i += 1.     
+00002d90: 2020 2020 2020 2020 2020 2065 6c73 653a             else:
+00002da0: 234e 6f6e 2063 6f6c 6c61 7073 6564 206c  #Non collapsed l
+00002db0: 6561 7665 730a 2020 2020 2020 2020 2020  eaves.          
+00002dc0: 2020 2020 2020 2020 2020 7061 7373 0a20            pass. 
+00002dd0: 2020 2020 2020 2063 6c61 6465 5f64 6963         clade_dic
+00002de0: 745f 6c65 6176 6573 203d 2064 6566 6175  t_leaves = defau
+00002df0: 6c74 6469 6374 2829 0a20 2020 2020 2020  ltdict().       
+00002e00: 2069 203d 2030 0a20 2020 2020 2020 2066   i = 0.        f
+00002e10: 6f72 206e 2069 6e20 7472 6565 2e74 7261  or n in tree.tra
+00002e20: 7665 7273 6528 2270 7265 6f72 6465 7222  verse("preorder"
+00002e30: 293a 0a20 2020 2020 2020 2020 2020 2069  ):.            i
+00002e40: 6620 6e2e 6973 5f6c 6561 6628 293a 0a20  f n.is_leaf():. 
+00002e50: 2020 2020 2020 2020 2020 2020 2020 2069                 i
+00002e60: 6620 227b 2220 6e6f 7420 696e 206e 2e6e  f "{" not in n.n
+00002e70: 616d 653a 0a20 2020 2020 2020 2020 2020  ame:.           
+00002e80: 2020 2020 2020 2020 2063 6c61 6465 5f6e           clade_n
+00002e90: 616d 6573 5f6c 6561 7665 7320 3d20 5b6e  ames_leaves = [n
+00002ea0: 2e6e 616d 655d 0a20 2020 2020 2020 2020  .name].         
+00002eb0: 2020 2020 2020 2065 6c73 653a 0a20 2020         else:.   
 00002ec0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00002ed0: 636c 6164 655f 6e61 6d65 735f 6c65 6176  clade_names_leav
-00002ee0: 6573 203d 206e 2e6e 616d 655b 6e2e 6e61  es = n.name[n.na
-00002ef0: 6d65 2e66 696e 6428 226c 6561 7665 733a  me.find("leaves:
-00002f00: 7b22 2920 2b20 383a 6e2e 6e61 6d65 2e66  {") + 8:n.name.f
-00002f10: 696e 6428 227d 2229 5d2e 7370 6c69 7428  ind("}")].split(
-00002f20: 222c 2229 0a20 2020 2020 2020 2020 2020  ",").           
-00002f30: 2020 2020 2063 6c61 6465 5f64 6963 745f       clade_dict_
-00002f40: 6c65 6176 6573 5b22 436c 6164 655f 7b7d  leaves["Clade_{}
-00002f50: 222e 666f 726d 6174 2869 295d 203d 2063  ".format(i)] = c
-00002f60: 6c61 6465 5f6e 616d 6573 5f6c 6561 7665  lade_names_leave
-00002f70: 730a 2020 2020 2020 2020 2020 2020 2020  s.              
-00002f80: 2020 6920 2b3d 2031 0a0a 2020 2020 2020    i += 1..      
-00002f90: 2020 7265 7475 726e 2063 6c61 6465 5f64    return clade_d
-00002fa0: 6963 745f 6c65 6176 6573 2c63 6c61 6465  ict_leaves,clade
-00002fb0: 5f64 6963 745f 616c 6c0a 0a20 2020 2063  _dict_all..    c
-00002fc0: 6c61 6465 5f64 6963 745f 6c65 6176 6573  lade_dict_leaves
-00002fd0: 2c63 6c61 6465 5f64 6963 745f 616c 6c20  ,clade_dict_all 
-00002fe0: 3d20 6275 696c 645f 636c 6164 6573 2874  = build_clades(t
-00002ff0: 7265 652c 6e61 6d65 290a 2020 2020 2348  ree,name).    #H
-00003000: 6967 686c 6967 6874 3a20 636c 6164 6573  ighlight: clades
-00003010: 5f64 6963 745f 616c 6c20 636f 6e74 6169  _dict_all contai
-00003020: 6e73 2065 6163 6820 636c 6164 6527 7320  ns each clade's 
-00003030: 696e 7465 726e 616c 2061 6e64 206c 6561  internal and lea
-00003040: 7665 7320 6e6f 6465 732c 2063 6c61 6465  ves nodes, clade
-00003050: 735f 6469 6374 5f6c 6561 7665 7320 6f6e  s_dict_leaves on
-00003060: 6c79 2063 6f6e 7461 696e 7320 7468 6520  ly contains the 
-00003070: 6c65 6176 6573 206f 6620 6561 6368 2063  leaves of each c
-00003080: 6c61 6465 0a0a 2020 2020 6469 6c6c 2e64  lade..    dill.d
-00003090: 756d 7028 636c 6164 655f 6469 6374 5f61  ump(clade_dict_a
-000030a0: 6c6c 2c20 6f70 656e 2827 7b7d 2f7b 7d5f  ll, open('{}/{}_
-000030b0: 436c 6164 6573 5f64 6963 745f 616c 6c2e  Clades_dict_all.
-000030c0: 7027 2e66 6f72 6d61 7428 7374 6f72 6167  p'.format(storag
-000030d0: 655f 666f 6c64 6572 2c6e 616d 6529 2c20  e_folder,name), 
-000030e0: 2777 6227 2929 232c 7072 6f74 6f63 6f6c  'wb'))#,protocol
-000030f0: 3d70 6963 6b6c 652e 4849 4748 4553 545f  =pickle.HIGHEST_
-00003100: 5052 4f54 4f43 4f4c 290a 2020 2020 7069  PROTOCOL).    pi
-00003110: 636b 6c65 2e64 756d 7028 636c 6164 655f  ckle.dump(clade_
-00003120: 6469 6374 5f6c 6561 7665 732c 206f 7065  dict_leaves, ope
-00003130: 6e28 277b 7d2f 7b7d 5f43 6c61 6465 735f  n('{}/{}_Clades_
-00003140: 6469 6374 5f6c 6561 7665 732e 7027 2e66  dict_leaves.p'.f
-00003150: 6f72 6d61 7428 7374 6f72 6167 655f 666f  ormat(storage_fo
-00003160: 6c64 6572 2c6e 616d 6529 2c20 2777 6227  lder,name), 'wb'
-00003170: 292c 7072 6f74 6f63 6f6c 3d70 6963 6b6c  ),protocol=pickl
-00003180: 652e 4849 4748 4553 545f 5052 4f54 4f43  e.HIGHEST_PROTOC
-00003190: 4f4c 290a 6465 6620 6361 6c63 756c 6174  OL).def calculat
-000031a0: 655f 636c 6f73 6573 745f 6c65 6176 6573  e_closest_leaves
-000031b0: 286e 616d 652c 7472 6565 2c73 746f 7261  (name,tree,stora
-000031c0: 6765 5f66 6f6c 6465 7229 3a0a 2020 2020  ge_folder):.    
-000031d0: 2222 2220 4372 6561 7465 7320 6120 6469  """ Creates a di
-000031e0: 6374 696f 6e61 7279 2074 6861 7420 636f  ctionary that co
-000031f0: 6e74 6169 6e73 2074 6865 2063 6c6f 7365  ntains the close
-00003200: 7374 206c 6561 7665 2074 6f20 616e 2069  st leave to an i
-00003210: 6e74 6572 6e61 6c20 6e6f 6465 207b 696e  nternal node {in
-00003220: 7465 726e 616c 5f6e 6f64 653a 6c65 6176  ternal_node:leav
-00003230: 657d 0a20 2020 203a 7061 7261 6d20 7374  e}.    :param st
-00003240: 7220 6e61 6d65 3a20 6461 7461 2073 6574  r name: data set
-00003250: 2070 726f 6a65 6374 206e 616d 650a 2020   project name.  
-00003260: 2020 3a70 6172 616d 2065 7465 332d 7472    :param ete3-tr
-00003270: 6565 2074 7265 653a 2045 7465 3320 7472  ee tree: Ete3 tr
-00003280: 6565 2063 6c61 7373 206f 626a 6563 740a  ee class object.
-00003290: 2020 2020 3a70 6172 616d 2073 7472 2073      :param str s
-000032a0: 746f 7261 6765 5f66 6f6c 6465 723a 2066  torage_folder: f
-000032b0: 6f6c 6465 7220 7768 6572 6520 746f 2064  older where to d
-000032c0: 756d 7020 7468 6520 6f75 7470 7574 0a20  ump the output. 
-000032d0: 2020 2022 2222 0a20 2020 2063 6c6f 7365     """.    close
-000032e0: 7374 5f6c 6561 7665 735f 6469 6374 3d64  st_leaves_dict=d
-000032f0: 6566 6175 6c74 6469 6374 2829 2023 636c  efaultdict() #cl
-00003300: 6f73 6573 7420 6c65 6176 6520 746f 2061  osest leave to a
-00003310: 6e20 696e 7465 726e 616c 206e 6f64 650a  n internal node.
-00003320: 2020 2020 666f 7220 6e6f 6465 2069 6e20      for node in 
-00003330: 7472 6565 2e74 7261 7665 7273 6528 293a  tree.traverse():
-00003340: 0a20 2020 2020 2020 2069 6620 6e6f 7420  .        if not 
-00003350: 6e6f 6465 2e69 735f 6c65 6166 2829 3a20  node.is_leaf(): 
-00003360: 2369 6620 6974 2773 2061 6e20 696e 7465  #if it's an inte
-00003370: 726e 616c 206e 6f64 650a 2020 2020 2020  rnal node.      
-00003380: 2020 2020 2020 7465 726d 696e 616c 5f6e        terminal_n
-00003390: 6f64 6520 3d20 616c 6c28 6e6f 6465 2e69  ode = all(node.i
-000033a0: 735f 6c65 6166 2829 2066 6f72 206e 6f64  s_leaf() for nod
-000033b0: 6520 696e 206e 6f64 652e 6765 745f 6368  e in node.get_ch
-000033c0: 696c 6472 656e 2829 290a 2020 2020 2020  ildren()).      
-000033d0: 2020 2020 2020 6966 2074 6572 6d69 6e61        if termina
-000033e0: 6c5f 6e6f 6465 3a0a 2020 2020 2020 2020  l_node:.        
-000033f0: 2020 2020 2020 2020 636c 6f73 6573 745f          closest_
-00003400: 6c65 6176 6573 5f64 6963 745b 6e6f 6465  leaves_dict[node
-00003410: 2e6e 616d 655d 203d 205b 6e6f 6465 2e6e  .name] = [node.n
-00003420: 616d 6520 666f 7220 6e6f 6465 2069 6e20  ame for node in 
-00003430: 6e6f 6465 2e67 6574 5f63 6869 6c64 7265  node.get_childre
-00003440: 6e28 295d 0a20 2020 2020 2020 2020 2020  n()].           
-00003450: 2065 6c73 653a 0a20 2020 2020 2020 2020   else:.         
-00003460: 2020 2020 2020 2063 6c6f 7365 7374 5f6c         closest_l
-00003470: 6561 7665 735f 6469 6374 5b6e 6f64 652e  eaves_dict[node.
-00003480: 6e61 6d65 5d20 3d20 5b6e 6f64 652e 6765  name] = [node.ge
-00003490: 745f 636c 6f73 6573 745f 6c65 6166 2829  t_closest_leaf()
-000034a0: 5b30 5d2e 6e61 6d65 5d0a 2020 2020 7069  [0].name].    pi
-000034b0: 636b 6c65 2e64 756d 7028 636c 6f73 6573  ckle.dump(closes
-000034c0: 745f 6c65 6176 6573 5f64 6963 742c 206f  t_leaves_dict, o
-000034d0: 7065 6e28 277b 7d2f 7b7d 5f43 6c6f 7365  pen('{}/{}_Close
-000034e0: 7374 5f6c 6561 7665 735f 6469 6374 2e70  st_leaves_dict.p
-000034f0: 272e 666f 726d 6174 2873 746f 7261 6765  '.format(storage
-00003500: 5f66 6f6c 6465 722c 6e61 6d65 292c 2027  _folder,name), '
-00003510: 7762 2729 2c70 726f 746f 636f 6c3d 7069  wb'),protocol=pi
-00003520: 636b 6c65 2e48 4947 4845 5354 5f50 524f  ckle.HIGHEST_PRO
-00003530: 544f 434f 4c29 0a64 6566 2063 616c 6375  TOCOL).def calcu
-00003540: 6c61 7465 5f64 6972 6563 746c 795f 6c69  late_directly_li
-00003550: 6e6b 6564 5f6e 6f64 6573 286e 616d 652c  nked_nodes(name,
-00003560: 7472 6565 2c73 746f 7261 6765 5f66 6f6c  tree,storage_fol
-00003570: 6465 7229 3a0a 2020 2020 2222 2243 7265  der):.    """Cre
-00003580: 6174 6573 2061 2064 6963 7469 6f6e 6172  ates a dictionar
-00003590: 7920 7468 6174 2063 6f6e 7461 696e 7320  y that contains 
-000035a0: 7468 6520 3220 6368 696c 6472 656e 206e  the 2 children n
-000035b0: 6f64 6573 2064 6972 6563 746c 7920 6c69  odes directly li
-000035c0: 6e6b 6564 2074 6f20 6120 6e6f 6465 2028  nked to a node (
-000035d0: 6e6f 7420 616c 6c20 7468 6520 6368 696c  not all the chil
-000035e0: 6472 656e 2066 726f 6d20 7468 6174 206e  dren from that n
-000035f0: 6f64 6529 207b 6e6f 6465 3a63 6869 6c64  ode) {node:child
-00003600: 7265 6e7d 0a20 2020 203a 7061 7261 6d20  ren}.    :param 
-00003610: 7374 7220 6e61 6d65 3a20 6461 7461 2073  str name: data s
-00003620: 6574 2070 726f 6a65 6374 206e 616d 650a  et project name.
-00003630: 2020 2020 3a70 6172 616d 2065 7465 332d      :param ete3-
-00003640: 7472 6565 2074 7265 653a 2045 7465 3320  tree tree: Ete3 
-00003650: 7472 6565 2063 6c61 7373 206f 626a 6563  tree class objec
-00003660: 740a 2020 2020 3a70 6172 616d 2073 7472  t.    :param str
-00003670: 2073 746f 7261 6765 5f66 6f6c 6465 723a   storage_folder:
-00003680: 2066 6f6c 6465 7220 7768 6572 6520 746f   folder where to
-00003690: 2064 756d 7020 7468 6520 6f75 7470 7574   dump the output
-000036a0: 2222 220a 2020 2020 636c 6f73 6573 745f  """.    closest_
-000036b0: 6368 696c 6472 656e 5f64 6963 743d 6465  children_dict=de
-000036c0: 6661 756c 7464 6963 7428 290a 2020 2020  faultdict().    
-000036d0: 666f 7220 6e6f 6465 2069 6e20 7472 6565  for node in tree
-000036e0: 2e74 7261 7665 7273 6528 293a 0a20 2020  .traverse():.   
-000036f0: 2020 2020 2063 6c6f 7365 7374 5f63 6869       closest_chi
-00003700: 6c64 7265 6e5f 6469 6374 5b6e 6f64 652e  ldren_dict[node.
-00003710: 6e61 6d65 5d20 3d20 5b6e 6f64 652e 6e61  name] = [node.na
-00003720: 6d65 2066 6f72 206e 6f64 6520 696e 206e  me for node in n
-00003730: 6f64 652e 6765 745f 6368 696c 6472 656e  ode.get_children
-00003740: 2829 5d0a 2020 2020 7069 636b 6c65 2e64  ()].    pickle.d
-00003750: 756d 7028 636c 6f73 6573 745f 6368 696c  ump(closest_chil
-00003760: 6472 656e 5f64 6963 742c 206f 7065 6e28  dren_dict, open(
-00003770: 277b 7d2f 7b7d 5f43 6c6f 7365 7374 5f63  '{}/{}_Closest_c
-00003780: 6869 6c64 7265 6e5f 6469 6374 2e70 272e  hildren_dict.p'.
-00003790: 666f 726d 6174 2873 746f 7261 6765 5f66  format(storage_f
-000037a0: 6f6c 6465 722c 6e61 6d65 292c 2027 7762  older,name), 'wb
-000037b0: 2729 2c70 726f 746f 636f 6c3d 7069 636b  '),protocol=pick
-000037c0: 6c65 2e48 4947 4845 5354 5f50 524f 544f  le.HIGHEST_PROTO
-000037d0: 434f 4c29 0a64 6566 2063 616c 6375 6c61  COL).def calcula
-000037e0: 7465 5f64 6573 6365 6e64 616e 7473 286e  te_descendants(n
-000037f0: 616d 652c 7472 6565 2c73 746f 7261 6765  ame,tree,storage
-00003800: 5f66 6f6c 6465 7229 3a0a 2020 2020 2222  _folder):.    ""
-00003810: 2243 7265 6174 6573 2061 2064 6963 7469  "Creates a dicti
-00003820: 6f6e 6172 7920 7468 6174 2063 6f6e 7461  onary that conta
-00003830: 696e 7320 616c 6c20 7468 6520 696e 7465  ins all the inte
-00003840: 726e 616c 206e 6f64 6573 2061 6e64 206c  rnal nodes and l
-00003850: 6561 7665 7320 7468 6174 2064 6573 6365  eaves that desce
-00003860: 6e64 2066 726f 6d20 7468 6174 2069 6e74  nd from that int
-00003870: 6572 6e61 6c20 6e6f 6465 207b 696e 7465  ernal node {inte
-00003880: 726e 616c 5f6e 6f64 653a 6465 7363 656e  rnal_node:descen
-00003890: 6461 6e74 737d 0a20 2020 203a 7061 7261  dants}.    :para
-000038a0: 6d20 7374 7220 6e61 6d65 3a20 6461 7461  m str name: data
-000038b0: 2073 6574 2070 726f 6a65 6374 206e 616d   set project nam
-000038c0: 650a 2020 2020 3a70 6172 616d 2065 7465  e.    :param ete
-000038d0: 332d 7472 6565 2074 7265 653a 2045 7465  3-tree tree: Ete
-000038e0: 3320 7472 6565 2063 6c61 7373 206f 626a  3 tree class obj
-000038f0: 6563 740a 2020 2020 3a70 6172 616d 2073  ect.    :param s
-00003900: 7472 2073 746f 7261 6765 5f66 6f6c 6465  tr storage_folde
-00003910: 723a 2066 6f6c 6465 7220 7768 6572 6520  r: folder where 
-00003920: 746f 2064 756d 7020 7468 6520 6f75 7470  to dump the outp
-00003930: 7574 2222 220a 2020 2020 636c 6f73 6573  ut""".    closes
-00003940: 745f 6465 7363 656e 6461 6e74 735f 6469  t_descendants_di
-00003950: 6374 203d 2064 6566 6175 6c74 6469 6374  ct = defaultdict
-00003960: 286c 616d 6264 613a 2064 6566 6175 6c74  (lambda: default
-00003970: 6469 6374 2829 290a 2020 2020 666f 7220  dict()).    for 
-00003980: 6e6f 6465 2069 6e20 7472 6565 2e74 7261  node in tree.tra
-00003990: 7665 7273 6528 293a 0a20 2020 2020 2020  verse():.       
-000039a0: 2069 6620 6e6f 7420 6e6f 6465 2e69 735f   if not node.is_
-000039b0: 6c65 6166 2829 3a0a 2020 2020 2020 2020  leaf():.        
-000039c0: 2020 2020 6465 7363 656e 6461 6e74 5f6c      descendant_l
-000039d0: 6561 7665 7320 3d20 5b5d 0a20 2020 2020  eaves = [].     
-000039e0: 2020 2020 2020 2064 6573 6365 6e64 616e         descendan
-000039f0: 745f 696e 7465 726e 616c 203d 205b 6e6f  t_internal = [no
-00003a00: 6465 2e6e 616d 655d 0a20 2020 2020 2020  de.name].       
-00003a10: 2020 2020 2066 6f72 2064 6573 6365 6e64       for descend
-00003a20: 616e 7420 696e 206e 6f64 652e 6974 6572  ant in node.iter
-00003a30: 5f64 6573 6365 6e64 616e 7473 2829 3a0a  _descendants():.
-00003a40: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003a50: 6966 2064 6573 6365 6e64 616e 742e 6973  if descendant.is
-00003a60: 5f6c 6561 6628 293a 0a20 2020 2020 2020  _leaf():.       
-00003a70: 2020 2020 2020 2020 2020 2020 2064 6573               des
-00003a80: 6365 6e64 616e 745f 6c65 6176 6573 2e61  cendant_leaves.a
-00003a90: 7070 656e 6428 6465 7363 656e 6461 6e74  ppend(descendant
-00003aa0: 2e6e 616d 6529 0a20 2020 2020 2020 2020  .name).         
-00003ab0: 2020 2020 2020 2065 6c73 653a 0a20 2020         else:.   
+00002ed0: 2063 6c61 6465 5f6e 616d 6573 5f6c 6561   clade_names_lea
+00002ee0: 7665 7320 3d20 6e2e 6e61 6d65 5b6e 2e6e  ves = n.name[n.n
+00002ef0: 616d 652e 6669 6e64 2822 6c65 6176 6573  ame.find("leaves
+00002f00: 3a7b 2229 202b 2038 3a6e 2e6e 616d 652e  :{") + 8:n.name.
+00002f10: 6669 6e64 2822 7d22 295d 2e73 706c 6974  find("}")].split
+00002f20: 2822 2c22 290a 2020 2020 2020 2020 2020  (",").          
+00002f30: 2020 2020 2020 636c 6164 655f 6469 6374        clade_dict
+00002f40: 5f6c 6561 7665 735b 2243 6c61 6465 5f7b  _leaves["Clade_{
+00002f50: 7d22 2e66 6f72 6d61 7428 6929 5d20 3d20  }".format(i)] = 
+00002f60: 636c 6164 655f 6e61 6d65 735f 6c65 6176  clade_names_leav
+00002f70: 6573 0a20 2020 2020 2020 2020 2020 2020  es.             
+00002f80: 2020 2069 202b 3d20 310a 0a20 2020 2020     i += 1..     
+00002f90: 2020 2072 6574 7572 6e20 636c 6164 655f     return clade_
+00002fa0: 6469 6374 5f6c 6561 7665 732c 636c 6164  dict_leaves,clad
+00002fb0: 655f 6469 6374 5f61 6c6c 0a0a 2020 2020  e_dict_all..    
+00002fc0: 636c 6164 655f 6469 6374 5f6c 6561 7665  clade_dict_leave
+00002fd0: 732c 636c 6164 655f 6469 6374 5f61 6c6c  s,clade_dict_all
+00002fe0: 203d 2062 7569 6c64 5f63 6c61 6465 7328   = build_clades(
+00002ff0: 7472 6565 2c6e 616d 6529 0a20 2020 2023  tree,name).    #
+00003000: 4869 6768 6c69 6768 743a 2063 6c61 6465  Highlight: clade
+00003010: 735f 6469 6374 5f61 6c6c 2063 6f6e 7461  s_dict_all conta
+00003020: 696e 7320 6561 6368 2063 6c61 6465 2773  ins each clade's
+00003030: 2069 6e74 6572 6e61 6c20 616e 6420 6c65   internal and le
+00003040: 6176 6573 206e 6f64 6573 2c20 636c 6164  aves nodes, clad
+00003050: 6573 5f64 6963 745f 6c65 6176 6573 206f  es_dict_leaves o
+00003060: 6e6c 7920 636f 6e74 6169 6e73 2074 6865  nly contains the
+00003070: 206c 6561 7665 7320 6f66 2065 6163 6820   leaves of each 
+00003080: 636c 6164 650a 0a20 2020 2064 696c 6c2e  clade..    dill.
+00003090: 6475 6d70 2863 6c61 6465 5f64 6963 745f  dump(clade_dict_
+000030a0: 616c 6c2c 206f 7065 6e28 277b 7d2f 7b7d  all, open('{}/{}
+000030b0: 5f43 6c61 6465 735f 6469 6374 5f61 6c6c  _Clades_dict_all
+000030c0: 2e70 272e 666f 726d 6174 2873 746f 7261  .p'.format(stora
+000030d0: 6765 5f66 6f6c 6465 722c 6e61 6d65 292c  ge_folder,name),
+000030e0: 2027 7762 2729 2923 2c70 726f 746f 636f   'wb'))#,protoco
+000030f0: 6c3d 7069 636b 6c65 2e48 4947 4845 5354  l=pickle.HIGHEST
+00003100: 5f50 524f 544f 434f 4c29 0a20 2020 2070  _PROTOCOL).    p
+00003110: 6963 6b6c 652e 6475 6d70 2863 6c61 6465  ickle.dump(clade
+00003120: 5f64 6963 745f 6c65 6176 6573 2c20 6f70  _dict_leaves, op
+00003130: 656e 2827 7b7d 2f7b 7d5f 436c 6164 6573  en('{}/{}_Clades
+00003140: 5f64 6963 745f 6c65 6176 6573 2e70 272e  _dict_leaves.p'.
+00003150: 666f 726d 6174 2873 746f 7261 6765 5f66  format(storage_f
+00003160: 6f6c 6465 722c 6e61 6d65 292c 2027 7762  older,name), 'wb
+00003170: 2729 2c70 726f 746f 636f 6c3d 7069 636b  '),protocol=pick
+00003180: 6c65 2e48 4947 4845 5354 5f50 524f 544f  le.HIGHEST_PROTO
+00003190: 434f 4c29 0a64 6566 2063 616c 6375 6c61  COL).def calcula
+000031a0: 7465 5f63 6c6f 7365 7374 5f6c 6561 7665  te_closest_leave
+000031b0: 7328 6e61 6d65 2c74 7265 652c 7374 6f72  s(name,tree,stor
+000031c0: 6167 655f 666f 6c64 6572 293a 0a20 2020  age_folder):.   
+000031d0: 2022 2222 2043 7265 6174 6573 2061 2064   """ Creates a d
+000031e0: 6963 7469 6f6e 6172 7920 7468 6174 2063  ictionary that c
+000031f0: 6f6e 7461 696e 7320 7468 6520 636c 6f73  ontains the clos
+00003200: 6573 7420 6c65 6176 6520 746f 2061 6e20  est leave to an 
+00003210: 696e 7465 726e 616c 206e 6f64 6520 7b69  internal node {i
+00003220: 6e74 6572 6e61 6c5f 6e6f 6465 3a6c 6561  nternal_node:lea
+00003230: 7665 7d0a 2020 2020 3a70 6172 616d 2073  ve}.    :param s
+00003240: 7472 206e 616d 653a 2064 6174 6120 7365  tr name: data se
+00003250: 7420 7072 6f6a 6563 7420 6e61 6d65 0a20  t project name. 
+00003260: 2020 203a 7061 7261 6d20 6574 6533 2d74     :param ete3-t
+00003270: 7265 6520 7472 6565 3a20 4574 6533 2074  ree tree: Ete3 t
+00003280: 7265 6520 636c 6173 7320 6f62 6a65 6374  ree class object
+00003290: 0a20 2020 203a 7061 7261 6d20 7374 7220  .    :param str 
+000032a0: 7374 6f72 6167 655f 666f 6c64 6572 3a20  storage_folder: 
+000032b0: 666f 6c64 6572 2077 6865 7265 2074 6f20  folder where to 
+000032c0: 6475 6d70 2074 6865 206f 7574 7075 740a  dump the output.
+000032d0: 2020 2020 2222 220a 2020 2020 636c 6f73      """.    clos
+000032e0: 6573 745f 6c65 6176 6573 5f64 6963 743d  est_leaves_dict=
+000032f0: 6465 6661 756c 7464 6963 7428 2920 2363  defaultdict() #c
+00003300: 6c6f 7365 7374 206c 6561 7665 2074 6f20  losest leave to 
+00003310: 616e 2069 6e74 6572 6e61 6c20 6e6f 6465  an internal node
+00003320: 0a20 2020 2066 6f72 206e 6f64 6520 696e  .    for node in
+00003330: 2074 7265 652e 7472 6176 6572 7365 2829   tree.traverse()
+00003340: 3a0a 2020 2020 2020 2020 6966 206e 6f74  :.        if not
+00003350: 206e 6f64 652e 6973 5f6c 6561 6628 293a   node.is_leaf():
+00003360: 2023 6966 2069 7427 7320 616e 2069 6e74   #if it's an int
+00003370: 6572 6e61 6c20 6e6f 6465 0a20 2020 2020  ernal node.     
+00003380: 2020 2020 2020 2074 6572 6d69 6e61 6c5f         terminal_
+00003390: 6e6f 6465 203d 2061 6c6c 286e 6f64 652e  node = all(node.
+000033a0: 6973 5f6c 6561 6628 2920 666f 7220 6e6f  is_leaf() for no
+000033b0: 6465 2069 6e20 6e6f 6465 2e67 6574 5f63  de in node.get_c
+000033c0: 6869 6c64 7265 6e28 2929 0a20 2020 2020  hildren()).     
+000033d0: 2020 2020 2020 2069 6620 7465 726d 696e         if termin
+000033e0: 616c 5f6e 6f64 653a 0a20 2020 2020 2020  al_node:.       
+000033f0: 2020 2020 2020 2020 2063 6c6f 7365 7374           closest
+00003400: 5f6c 6561 7665 735f 6469 6374 5b6e 6f64  _leaves_dict[nod
+00003410: 652e 6e61 6d65 5d20 3d20 5b6e 6f64 652e  e.name] = [node.
+00003420: 6e61 6d65 2066 6f72 206e 6f64 6520 696e  name for node in
+00003430: 206e 6f64 652e 6765 745f 6368 696c 6472   node.get_childr
+00003440: 656e 2829 5d0a 2020 2020 2020 2020 2020  en()].          
+00003450: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
+00003460: 2020 2020 2020 2020 636c 6f73 6573 745f          closest_
+00003470: 6c65 6176 6573 5f64 6963 745b 6e6f 6465  leaves_dict[node
+00003480: 2e6e 616d 655d 203d 205b 6e6f 6465 2e67  .name] = [node.g
+00003490: 6574 5f63 6c6f 7365 7374 5f6c 6561 6628  et_closest_leaf(
+000034a0: 295b 305d 2e6e 616d 655d 0a20 2020 2070  )[0].name].    p
+000034b0: 6963 6b6c 652e 6475 6d70 2863 6c6f 7365  ickle.dump(close
+000034c0: 7374 5f6c 6561 7665 735f 6469 6374 2c20  st_leaves_dict, 
+000034d0: 6f70 656e 2827 7b7d 2f7b 7d5f 436c 6f73  open('{}/{}_Clos
+000034e0: 6573 745f 6c65 6176 6573 5f64 6963 742e  est_leaves_dict.
+000034f0: 7027 2e66 6f72 6d61 7428 7374 6f72 6167  p'.format(storag
+00003500: 655f 666f 6c64 6572 2c6e 616d 6529 2c20  e_folder,name), 
+00003510: 2777 6227 292c 7072 6f74 6f63 6f6c 3d70  'wb'),protocol=p
+00003520: 6963 6b6c 652e 4849 4748 4553 545f 5052  ickle.HIGHEST_PR
+00003530: 4f54 4f43 4f4c 290a 6465 6620 6361 6c63  OTOCOL).def calc
+00003540: 756c 6174 655f 6469 7265 6374 6c79 5f6c  ulate_directly_l
+00003550: 696e 6b65 645f 6e6f 6465 7328 6e61 6d65  inked_nodes(name
+00003560: 2c74 7265 652c 7374 6f72 6167 655f 666f  ,tree,storage_fo
+00003570: 6c64 6572 293a 0a20 2020 2022 2222 4372  lder):.    """Cr
+00003580: 6561 7465 7320 6120 6469 6374 696f 6e61  eates a dictiona
+00003590: 7279 2074 6861 7420 636f 6e74 6169 6e73  ry that contains
+000035a0: 2074 6865 2032 2063 6869 6c64 7265 6e20   the 2 children 
+000035b0: 6e6f 6465 7320 6469 7265 6374 6c79 206c  nodes directly l
+000035c0: 696e 6b65 6420 746f 2061 206e 6f64 6520  inked to a node 
+000035d0: 286e 6f74 2061 6c6c 2074 6865 2063 6869  (not all the chi
+000035e0: 6c64 7265 6e20 6672 6f6d 2074 6861 7420  ldren from that 
+000035f0: 6e6f 6465 2920 7b6e 6f64 653a 6368 696c  node) {node:chil
+00003600: 6472 656e 7d0a 2020 2020 3a70 6172 616d  dren}.    :param
+00003610: 2073 7472 206e 616d 653a 2064 6174 6120   str name: data 
+00003620: 7365 7420 7072 6f6a 6563 7420 6e61 6d65  set project name
+00003630: 0a20 2020 203a 7061 7261 6d20 6574 6533  .    :param ete3
+00003640: 2d74 7265 6520 7472 6565 3a20 4574 6533  -tree tree: Ete3
+00003650: 2074 7265 6520 636c 6173 7320 6f62 6a65   tree class obje
+00003660: 6374 0a20 2020 203a 7061 7261 6d20 7374  ct.    :param st
+00003670: 7220 7374 6f72 6167 655f 666f 6c64 6572  r storage_folder
+00003680: 3a20 666f 6c64 6572 2077 6865 7265 2074  : folder where t
+00003690: 6f20 6475 6d70 2074 6865 206f 7574 7075  o dump the outpu
+000036a0: 7422 2222 0a20 2020 2063 6c6f 7365 7374  t""".    closest
+000036b0: 5f63 6869 6c64 7265 6e5f 6469 6374 3d64  _children_dict=d
+000036c0: 6566 6175 6c74 6469 6374 2829 0a20 2020  efaultdict().   
+000036d0: 2066 6f72 206e 6f64 6520 696e 2074 7265   for node in tre
+000036e0: 652e 7472 6176 6572 7365 2829 3a0a 2020  e.traverse():.  
+000036f0: 2020 2020 2020 636c 6f73 6573 745f 6368        closest_ch
+00003700: 696c 6472 656e 5f64 6963 745b 6e6f 6465  ildren_dict[node
+00003710: 2e6e 616d 655d 203d 205b 6e6f 6465 2e6e  .name] = [node.n
+00003720: 616d 6520 666f 7220 6e6f 6465 2069 6e20  ame for node in 
+00003730: 6e6f 6465 2e67 6574 5f63 6869 6c64 7265  node.get_childre
+00003740: 6e28 295d 0a20 2020 2070 6963 6b6c 652e  n()].    pickle.
+00003750: 6475 6d70 2863 6c6f 7365 7374 5f63 6869  dump(closest_chi
+00003760: 6c64 7265 6e5f 6469 6374 2c20 6f70 656e  ldren_dict, open
+00003770: 2827 7b7d 2f7b 7d5f 436c 6f73 6573 745f  ('{}/{}_Closest_
+00003780: 6368 696c 6472 656e 5f64 6963 742e 7027  children_dict.p'
+00003790: 2e66 6f72 6d61 7428 7374 6f72 6167 655f  .format(storage_
+000037a0: 666f 6c64 6572 2c6e 616d 6529 2c20 2777  folder,name), 'w
+000037b0: 6227 292c 7072 6f74 6f63 6f6c 3d70 6963  b'),protocol=pic
+000037c0: 6b6c 652e 4849 4748 4553 545f 5052 4f54  kle.HIGHEST_PROT
+000037d0: 4f43 4f4c 290a 6465 6620 6361 6c63 756c  OCOL).def calcul
+000037e0: 6174 655f 6465 7363 656e 6461 6e74 7328  ate_descendants(
+000037f0: 6e61 6d65 2c74 7265 652c 7374 6f72 6167  name,tree,storag
+00003800: 655f 666f 6c64 6572 293a 0a20 2020 2022  e_folder):.    "
+00003810: 2222 4372 6561 7465 7320 6120 6469 6374  ""Creates a dict
+00003820: 696f 6e61 7279 2074 6861 7420 636f 6e74  ionary that cont
+00003830: 6169 6e73 2061 6c6c 2074 6865 2069 6e74  ains all the int
+00003840: 6572 6e61 6c20 6e6f 6465 7320 616e 6420  ernal nodes and 
+00003850: 6c65 6176 6573 2074 6861 7420 6465 7363  leaves that desc
+00003860: 656e 6420 6672 6f6d 2074 6861 7420 696e  end from that in
+00003870: 7465 726e 616c 206e 6f64 6520 7b69 6e74  ternal node {int
+00003880: 6572 6e61 6c5f 6e6f 6465 3a64 6573 6365  ernal_node:desce
+00003890: 6e64 616e 7473 7d0a 2020 2020 3a70 6172  ndants}.    :par
+000038a0: 616d 2073 7472 206e 616d 653a 2064 6174  am str name: dat
+000038b0: 6120 7365 7420 7072 6f6a 6563 7420 6e61  a set project na
+000038c0: 6d65 0a20 2020 203a 7061 7261 6d20 6574  me.    :param et
+000038d0: 6533 2d74 7265 6520 7472 6565 3a20 4574  e3-tree tree: Et
+000038e0: 6533 2074 7265 6520 636c 6173 7320 6f62  e3 tree class ob
+000038f0: 6a65 6374 0a20 2020 203a 7061 7261 6d20  ject.    :param 
+00003900: 7374 7220 7374 6f72 6167 655f 666f 6c64  str storage_fold
+00003910: 6572 3a20 666f 6c64 6572 2077 6865 7265  er: folder where
+00003920: 2074 6f20 6475 6d70 2074 6865 206f 7574   to dump the out
+00003930: 7075 7422 2222 0a20 2020 2063 6c6f 7365  put""".    close
+00003940: 7374 5f64 6573 6365 6e64 616e 7473 5f64  st_descendants_d
+00003950: 6963 7420 3d20 6465 6661 756c 7464 6963  ict = defaultdic
+00003960: 7428 6c61 6d62 6461 3a20 6465 6661 756c  t(lambda: defaul
+00003970: 7464 6963 7428 2929 0a20 2020 2066 6f72  tdict()).    for
+00003980: 206e 6f64 6520 696e 2074 7265 652e 7472   node in tree.tr
+00003990: 6176 6572 7365 2829 3a0a 2020 2020 2020  averse():.      
+000039a0: 2020 6966 206e 6f74 206e 6f64 652e 6973    if not node.is
+000039b0: 5f6c 6561 6628 293a 0a20 2020 2020 2020  _leaf():.       
+000039c0: 2020 2020 2064 6573 6365 6e64 616e 745f       descendant_
+000039d0: 6c65 6176 6573 203d 205b 5d0a 2020 2020  leaves = [].    
+000039e0: 2020 2020 2020 2020 6465 7363 656e 6461          descenda
+000039f0: 6e74 5f69 6e74 6572 6e61 6c20 3d20 5b6e  nt_internal = [n
+00003a00: 6f64 652e 6e61 6d65 5d0a 2020 2020 2020  ode.name].      
+00003a10: 2020 2020 2020 666f 7220 6465 7363 656e        for descen
+00003a20: 6461 6e74 2069 6e20 6e6f 6465 2e69 7465  dant in node.ite
+00003a30: 725f 6465 7363 656e 6461 6e74 7328 293a  r_descendants():
+00003a40: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00003a50: 2069 6620 6465 7363 656e 6461 6e74 2e69   if descendant.i
+00003a60: 735f 6c65 6166 2829 3a0a 2020 2020 2020  s_leaf():.      
+00003a70: 2020 2020 2020 2020 2020 2020 2020 6465                de
+00003a80: 7363 656e 6461 6e74 5f6c 6561 7665 732e  scendant_leaves.
+00003a90: 6170 7065 6e64 2864 6573 6365 6e64 616e  append(descendan
+00003aa0: 742e 6e61 6d65 290a 2020 2020 2020 2020  t.name).        
+00003ab0: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
 00003ac0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003ad0: 2064 6573 6365 6e64 616e 745f 696e 7465   descendant_inte
-00003ae0: 726e 616c 2e61 7070 656e 6428 6465 7363  rnal.append(desc
-00003af0: 656e 6461 6e74 2e6e 616d 6529 0a20 2020  endant.name).   
-00003b00: 2020 2020 2020 2020 2063 6c6f 7365 7374           closest
-00003b10: 5f64 6573 6365 6e64 616e 7473 5f64 6963  _descendants_dic
-00003b20: 745b 6e6f 6465 2e6e 616d 655d 5b22 696e  t[node.name]["in
-00003b30: 7465 726e 616c 225d 203d 2064 6573 6365  ternal"] = desce
-00003b40: 6e64 616e 745f 696e 7465 726e 616c 0a20  ndant_internal. 
-00003b50: 2020 2020 2020 2020 2020 2063 6c6f 7365             close
-00003b60: 7374 5f64 6573 6365 6e64 616e 7473 5f64  st_descendants_d
-00003b70: 6963 745b 6e6f 6465 2e6e 616d 655d 5b22  ict[node.name]["
-00003b80: 6c65 6176 6573 225d 203d 2064 6573 6365  leaves"] = desce
-00003b90: 6e64 616e 745f 6c65 6176 6573 0a20 2020  ndant_leaves.   
-00003ba0: 2064 696c 6c2e 6475 6d70 2863 6c6f 7365   dill.dump(close
-00003bb0: 7374 5f64 6573 6365 6e64 616e 7473 5f64  st_descendants_d
-00003bc0: 6963 742c 206f 7065 6e28 277b 7d2f 7b7d  ict, open('{}/{}
-00003bd0: 5f44 6573 6365 6e64 616e 7473 5f64 6963  _Descendants_dic
-00003be0: 742e 7027 2e66 6f72 6d61 7428 7374 6f72  t.p'.format(stor
-00003bf0: 6167 655f 666f 6c64 6572 2c6e 616d 6529  age_folder,name)
-00003c00: 2c20 2777 6227 2929 232c 7072 6f74 6f63  , 'wb'))#,protoc
-00003c10: 6f6c 3d70 6963 6b6c 652e 4849 4748 4553  ol=pickle.HIGHES
-00003c20: 545f 5052 4f54 4f43 4f4c 290a 6465 6620  T_PROTOCOL).def 
-00003c30: 5066 616d 5f70 6172 7365 7228 6661 6d69  Pfam_parser(fami
-00003c40: 6c79 5f6e 616d 652c 6461 7461 5f66 6f6c  ly_name,data_fol
-00003c50: 6465 722c 6669 7273 745f 6d61 7463 683d  der,first_match=
-00003c60: 4661 6c73 652c 7570 6461 7465 5f70 6661  False,update_pfa
-00003c70: 6d3d 4661 6c73 6529 3a0a 2020 2020 2222  m=False):.    ""
-00003c80: 2243 7265 6174 6573 2061 2064 6963 7469  "Creates a dicti
-00003c90: 6f6e 6172 7920 636f 6e74 6169 6e69 6e67  onary containing
-00003ca0: 2074 6865 2050 4442 2066 696c 6573 2061   the PDB files a
-00003cb0: 6e64 2074 6865 2073 6571 7565 6e63 6520  nd the sequence 
-00003cc0: 696e 666f 726d 6174 696f 6e20 2863 6861  information (cha
-00003cd0: 696e 2c20 7265 7369 6475 6573 2e2e 2e29  in, residues...)
-00003ce0: 0a20 2020 203a 7061 7261 6d20 7374 7220  .    :param str 
-00003cf0: 6661 6d69 6c79 5f6e 616d 653a 2070 6661  family_name: pfa
-00003d00: 6d20 6661 6d69 6c79 206e 616d 650a 2020  m family name.  
-00003d10: 2020 3a70 6172 616d 2062 6f6f 6c20 6669    :param bool fi
-00003d20: 7273 745f 6d61 7463 683a 2054 7275 6520  rst_match: True 
-00003d30: 2d3e 2050 6963 6b73 206f 6e6c 7920 7468  -> Picks only th
-00003d40: 6520 6669 7273 7420 666f 756e 6420 7072  e first found pr
-00003d50: 6f74 6569 6e20 616e 6420 6e6f 7420 6974  otein and not it
-00003d60: 2773 2064 7570 6c69 6361 7465 732c 2046  's duplicates, F
-00003d70: 616c 7365 202d 3e20 5461 6b65 7320 6576  alse -> Takes ev
-00003d80: 6572 7920 6176 6169 6c61 626c 6520 7374  ery available st
-00003d90: 7275 6374 7572 652c 2069 6e63 6c75 6469  ructure, includi
-00003da0: 6e67 2064 7570 6c69 6361 7465 730a 2020  ng duplicates.  
-00003db0: 2020 3a70 6172 616d 2062 6f6f 6c20 7570    :param bool up
-00003dc0: 6461 7465 5f70 6661 6d3a 2054 7275 6520  date_pfam: True 
-00003dd0: 2d3e 2044 6f77 6e6c 6f61 6473 2061 6e64  -> Downloads and
-00003de0: 2073 6176 6520 7468 6520 6c61 7465 7374   save the latest
-00003df0: 2070 6661 6d20 7665 7273 696f 6e20 6672   pfam version fr
-00003e00: 6f6d 2068 7474 703a 2f2f 6674 702e 6562  om http://ftp.eb
-00003e10: 692e 6163 2e75 6b2f 7075 622f 6461 7461  i.ac.uk/pub/data
-00003e20: 6261 7365 732f 5066 616d 2f63 7572 7265  bases/Pfam/curre
-00003e30: 6e74 5f72 656c 6561 7365 2f70 6462 6d61  nt_release/pdbma
-00003e40: 702e 677a 223b 2046 616c 7365 202d 3e20  p.gz"; False -> 
-00003e50: 5573 6520 7468 6520 7374 6f72 6564 2076  Use the stored v
-00003e60: 6572 7369 6f6e 0a20 2020 2022 2222 0a20  ersion.    """. 
-00003e70: 2020 2069 6620 7570 6461 7465 5f70 6661     if update_pfa
-00003e80: 6d3a 0a20 2020 2020 2020 2074 7279 3a0a  m:.        try:.
-00003e90: 2020 2020 2020 2020 2020 2020 7375 6270              subp
-00003ea0: 726f 6365 7373 2e63 616c 6c28 2272 6d20  rocess.call("rm 
-00003eb0: 2d72 6620 7064 626d 6170 2e67 7a20 7064  -rf pdbmap.gz pd
-00003ec0: 626d 6170 2229 0a20 2020 2020 2020 2065  bmap").        e
-00003ed0: 7863 6570 743a 0a20 2020 2020 2020 2020  xcept:.         
-00003ee0: 2020 2070 6173 730a 2020 2020 2020 2020     pass.        
-00003ef0: 7375 6270 726f 6365 7373 2e63 616c 6c28  subprocess.call(
-00003f00: 2277 6765 7420 6874 7470 3a2f 2f66 7470  "wget http://ftp
-00003f10: 2e65 6269 2e61 632e 756b 2f70 7562 2f64  .ebi.ac.uk/pub/d
-00003f20: 6174 6162 6173 6573 2f50 6661 6d2f 6375  atabases/Pfam/cu
-00003f30: 7272 656e 745f 7265 6c65 6173 652f 7064  rrent_release/pd
-00003f40: 626d 6170 2e67 7a22 2c73 6865 6c6c 3d54  bmap.gz",shell=T
-00003f50: 7275 6529 0a20 2020 2020 2020 2073 7562  rue).        sub
-00003f60: 7072 6f63 6573 732e 6361 6c6c 2822 6775  process.call("gu
-00003f70: 6e7a 6970 2070 6462 6d61 702e 677a 222c  nzip pdbmap.gz",
-00003f80: 7368 656c 6c3d 5472 7565 290a 2020 2020  shell=True).    
-00003f90: 2020 2020 7375 6270 726f 6365 7373 2e63      subprocess.c
-00003fa0: 616c 6c28 226d 7620 7064 626d 6170 2050  all("mv pdbmap P
-00003fb0: 6661 6d50 6462 4d61 702e 7478 7422 2c20  famPdbMap.txt", 
-00003fc0: 7368 656c 6c3d 5472 7565 290a 0a20 2020  shell=True)..   
-00003fd0: 2070 7269 6e74 2822 5265 6164 696e 6720   print("Reading 
-00003fe0: 616e 6420 6372 6561 7469 6e67 2070 6661  and creating pfa
-00003ff0: 6d20 6469 6374 696f 6e61 7279 2e2e 2e2e  m dictionary....
-00004000: 2229 0a20 2020 2073 746f 7261 6765 5f66  ").    storage_f
-00004010: 6f6c 6465 7220 3d20 6f73 2e70 6174 682e  older = os.path.
-00004020: 6162 7370 6174 6828 6f73 2e70 6174 682e  abspath(os.path.
-00004030: 6a6f 696e 286f 732e 7061 7468 2e64 6972  join(os.path.dir
-00004040: 6e61 6d65 285f 5f66 696c 655f 5f29 2c20  name(__file__), 
-00004050: 2264 6174 6122 2929 0a20 2020 2064 6620  "data")).    df 
-00004060: 3d20 7064 2e72 6561 645f 6373 7628 227b  = pd.read_csv("{
-00004070: 7d2f 5066 616d 5064 624d 6170 2e74 7874  }/PfamPdbMap.txt
-00004080: 222e 666f 726d 6174 2873 746f 7261 6765  ".format(storage
-00004090: 5f66 6f6c 6465 7229 2c20 7365 703d 225c  _folder), sep="\
-000040a0: 7422 2c20 6572 726f 725f 6261 645f 6c69  t", error_bad_li
-000040b0: 6e65 733d 4661 6c73 652c 2065 6e67 696e  nes=False, engin
-000040c0: 653d 2770 7974 686f 6e27 2c20 696e 6465  e='python', inde
-000040d0: 785f 636f 6c3d 4661 6c73 652c 0a20 2020  x_col=False,.   
+00003ad0: 2020 6465 7363 656e 6461 6e74 5f69 6e74    descendant_int
+00003ae0: 6572 6e61 6c2e 6170 7065 6e64 2864 6573  ernal.append(des
+00003af0: 6365 6e64 616e 742e 6e61 6d65 290a 2020  cendant.name).  
+00003b00: 2020 2020 2020 2020 2020 636c 6f73 6573            closes
+00003b10: 745f 6465 7363 656e 6461 6e74 735f 6469  t_descendants_di
+00003b20: 6374 5b6e 6f64 652e 6e61 6d65 5d5b 2269  ct[node.name]["i
+00003b30: 6e74 6572 6e61 6c22 5d20 3d20 6465 7363  nternal"] = desc
+00003b40: 656e 6461 6e74 5f69 6e74 6572 6e61 6c0a  endant_internal.
+00003b50: 2020 2020 2020 2020 2020 2020 636c 6f73              clos
+00003b60: 6573 745f 6465 7363 656e 6461 6e74 735f  est_descendants_
+00003b70: 6469 6374 5b6e 6f64 652e 6e61 6d65 5d5b  dict[node.name][
+00003b80: 226c 6561 7665 7322 5d20 3d20 6465 7363  "leaves"] = desc
+00003b90: 656e 6461 6e74 5f6c 6561 7665 730a 2020  endant_leaves.  
+00003ba0: 2020 6469 6c6c 2e64 756d 7028 636c 6f73    dill.dump(clos
+00003bb0: 6573 745f 6465 7363 656e 6461 6e74 735f  est_descendants_
+00003bc0: 6469 6374 2c20 6f70 656e 2827 7b7d 2f7b  dict, open('{}/{
+00003bd0: 7d5f 4465 7363 656e 6461 6e74 735f 6469  }_Descendants_di
+00003be0: 6374 2e70 272e 666f 726d 6174 2873 746f  ct.p'.format(sto
+00003bf0: 7261 6765 5f66 6f6c 6465 722c 6e61 6d65  rage_folder,name
+00003c00: 292c 2027 7762 2729 2923 2c70 726f 746f  ), 'wb'))#,proto
+00003c10: 636f 6c3d 7069 636b 6c65 2e48 4947 4845  col=pickle.HIGHE
+00003c20: 5354 5f50 524f 544f 434f 4c29 0a64 6566  ST_PROTOCOL).def
+00003c30: 2050 6661 6d5f 7061 7273 6572 2866 616d   Pfam_parser(fam
+00003c40: 696c 795f 6e61 6d65 2c64 6174 615f 666f  ily_name,data_fo
+00003c50: 6c64 6572 2c66 6972 7374 5f6d 6174 6368  lder,first_match
+00003c60: 3d46 616c 7365 2c75 7064 6174 655f 7066  =False,update_pf
+00003c70: 616d 3d46 616c 7365 293a 0a20 2020 2022  am=False):.    "
+00003c80: 2222 4372 6561 7465 7320 6120 6469 6374  ""Creates a dict
+00003c90: 696f 6e61 7279 2063 6f6e 7461 696e 696e  ionary containin
+00003ca0: 6720 7468 6520 5044 4220 6669 6c65 7320  g the PDB files 
+00003cb0: 616e 6420 7468 6520 7365 7175 656e 6365  and the sequence
+00003cc0: 2069 6e66 6f72 6d61 7469 6f6e 2028 6368   information (ch
+00003cd0: 6169 6e2c 2072 6573 6964 7565 732e 2e2e  ain, residues...
+00003ce0: 290a 2020 2020 3a70 6172 616d 2073 7472  ).    :param str
+00003cf0: 2066 616d 696c 795f 6e61 6d65 3a20 7066   family_name: pf
+00003d00: 616d 2066 616d 696c 7920 6e61 6d65 0a20  am family name. 
+00003d10: 2020 203a 7061 7261 6d20 626f 6f6c 2066     :param bool f
+00003d20: 6972 7374 5f6d 6174 6368 3a20 5472 7565  irst_match: True
+00003d30: 202d 3e20 5069 636b 7320 6f6e 6c79 2074   -> Picks only t
+00003d40: 6865 2066 6972 7374 2066 6f75 6e64 2070  he first found p
+00003d50: 726f 7465 696e 2061 6e64 206e 6f74 2069  rotein and not i
+00003d60: 7427 7320 6475 706c 6963 6174 6573 2c20  t's duplicates, 
+00003d70: 4661 6c73 6520 2d3e 2054 616b 6573 2065  False -> Takes e
+00003d80: 7665 7279 2061 7661 696c 6162 6c65 2073  very available s
+00003d90: 7472 7563 7475 7265 2c20 696e 636c 7564  tructure, includ
+00003da0: 696e 6720 6475 706c 6963 6174 6573 0a20  ing duplicates. 
+00003db0: 2020 203a 7061 7261 6d20 626f 6f6c 2075     :param bool u
+00003dc0: 7064 6174 655f 7066 616d 3a20 5472 7565  pdate_pfam: True
+00003dd0: 202d 3e20 446f 776e 6c6f 6164 7320 616e   -> Downloads an
+00003de0: 6420 7361 7665 2074 6865 206c 6174 6573  d save the lates
+00003df0: 7420 7066 616d 2076 6572 7369 6f6e 2066  t pfam version f
+00003e00: 726f 6d20 6874 7470 3a2f 2f66 7470 2e65  rom http://ftp.e
+00003e10: 6269 2e61 632e 756b 2f70 7562 2f64 6174  bi.ac.uk/pub/dat
+00003e20: 6162 6173 6573 2f50 6661 6d2f 6375 7272  abases/Pfam/curr
+00003e30: 656e 745f 7265 6c65 6173 652f 7064 626d  ent_release/pdbm
+00003e40: 6170 2e67 7a22 3b20 4661 6c73 6520 2d3e  ap.gz"; False ->
+00003e50: 2055 7365 2074 6865 2073 746f 7265 6420   Use the stored 
+00003e60: 7665 7273 696f 6e0a 2020 2020 2222 220a  version.    """.
+00003e70: 2020 2020 6966 2075 7064 6174 655f 7066      if update_pf
+00003e80: 616d 3a0a 2020 2020 2020 2020 7472 793a  am:.        try:
+00003e90: 0a20 2020 2020 2020 2020 2020 2073 7562  .            sub
+00003ea0: 7072 6f63 6573 732e 6361 6c6c 2822 726d  process.call("rm
+00003eb0: 202d 7266 2070 6462 6d61 702e 677a 2070   -rf pdbmap.gz p
+00003ec0: 6462 6d61 7022 290a 2020 2020 2020 2020  dbmap").        
+00003ed0: 6578 6365 7074 3a0a 2020 2020 2020 2020  except:.        
+00003ee0: 2020 2020 7061 7373 0a20 2020 2020 2020      pass.       
+00003ef0: 2073 7562 7072 6f63 6573 732e 6361 6c6c   subprocess.call
+00003f00: 2822 7767 6574 2068 7474 703a 2f2f 6674  ("wget http://ft
+00003f10: 702e 6562 692e 6163 2e75 6b2f 7075 622f  p.ebi.ac.uk/pub/
+00003f20: 6461 7461 6261 7365 732f 5066 616d 2f63  databases/Pfam/c
+00003f30: 7572 7265 6e74 5f72 656c 6561 7365 2f70  urrent_release/p
+00003f40: 6462 6d61 702e 677a 222c 7368 656c 6c3d  dbmap.gz",shell=
+00003f50: 5472 7565 290a 2020 2020 2020 2020 7375  True).        su
+00003f60: 6270 726f 6365 7373 2e63 616c 6c28 2267  bprocess.call("g
+00003f70: 756e 7a69 7020 7064 626d 6170 2e67 7a22  unzip pdbmap.gz"
+00003f80: 2c73 6865 6c6c 3d54 7275 6529 0a20 2020  ,shell=True).   
+00003f90: 2020 2020 2073 7562 7072 6f63 6573 732e       subprocess.
+00003fa0: 6361 6c6c 2822 6d76 2070 6462 6d61 7020  call("mv pdbmap 
+00003fb0: 5066 616d 5064 624d 6170 2e74 7874 222c  PfamPdbMap.txt",
+00003fc0: 2073 6865 6c6c 3d54 7275 6529 0a0a 2020   shell=True)..  
+00003fd0: 2020 7072 696e 7428 2252 6561 6469 6e67    print("Reading
+00003fe0: 2061 6e64 2063 7265 6174 696e 6720 7066   and creating pf
+00003ff0: 616d 2064 6963 7469 6f6e 6172 792e 2e2e  am dictionary...
+00004000: 2e22 290a 2020 2020 7374 6f72 6167 655f  .").    storage_
+00004010: 666f 6c64 6572 203d 206f 732e 7061 7468  folder = os.path
+00004020: 2e61 6273 7061 7468 286f 732e 7061 7468  .abspath(os.path
+00004030: 2e6a 6f69 6e28 6f73 2e70 6174 682e 6469  .join(os.path.di
+00004040: 726e 616d 6528 5f5f 6669 6c65 5f5f 292c  rname(__file__),
+00004050: 2022 6461 7461 2229 290a 2020 2020 6466   "data")).    df
+00004060: 203d 2070 642e 7265 6164 5f63 7376 2822   = pd.read_csv("
+00004070: 7b7d 2f50 6661 6d50 6462 4d61 702e 7478  {}/PfamPdbMap.tx
+00004080: 7422 2e66 6f72 6d61 7428 7374 6f72 6167  t".format(storag
+00004090: 655f 666f 6c64 6572 292c 2073 6570 3d22  e_folder), sep="
+000040a0: 5c74 222c 2065 7272 6f72 5f62 6164 5f6c  \t", error_bad_l
+000040b0: 696e 6573 3d46 616c 7365 2c20 656e 6769  ines=False, engi
+000040c0: 6e65 3d27 7079 7468 6f6e 272c 2069 6e64  ne='python', ind
+000040d0: 6578 5f63 6f6c 3d46 616c 7365 2c0a 2020  ex_col=False,.  
 000040e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000040f0: 2020 6e61 6d65 733d 5b22 5044 425f 3122    names=["PDB_1"
-00004100: 2c20 2243 6861 696e 222c 2022 456d 7074  , "Chain", "Empt
-00004110: 7922 2c20 2246 756e 6374 696f 6e22 2c20  y", "Function", 
-00004120: 2246 616d 696c 7922 2c20 2255 6e69 7072  "Family", "Unipr
-00004130: 6f74 222c 2022 5265 7369 6475 6573 225d  ot", "Residues"]
-00004140: 2920 2354 4f44 4f3a 206d 6f76 6520 746f  ) #TODO: move to
-00004150: 2064 6174 610a 2020 2020 6661 6d69 6c79   data.    family
-00004160: 203d 2064 665b 6466 5b27 4661 6d69 6c79   = df[df['Family
-00004170: 275d 2e73 7472 2e63 6f6e 7461 696e 7328  '].str.contains(
-00004180: 6661 6d69 6c79 5f6e 616d 6529 5d0a 2020  family_name)].  
-00004190: 2020 6966 2066 6972 7374 5f6d 6174 6368    if first_match
-000041a0: 3a20 2331 3135 2023 6f6e 6c79 2074 616b  : #115 #only tak
-000041b0: 6573 2074 6865 2066 6972 7374 2070 6462  es the first pdb
-000041c0: 2066 696c 6520 666f 756e 6420 666f 7220   file found for 
-000041d0: 6561 6368 2073 6571 7565 6e63 650a 2020  each sequence.  
-000041e0: 2020 2020 2020 6669 7273 745f 756e 6971        first_uniq
-000041f0: 7565 5f6d 6174 6368 203d 2066 616d 696c  ue_match = famil
-00004200: 792e 6772 6f75 7062 7928 2755 6e69 7072  y.groupby('Unipr
-00004210: 6f74 2729 2e68 6561 6428 3129 2e72 6573  ot').head(1).res
-00004220: 6574 5f69 6e64 6578 2864 726f 703d 5472  et_index(drop=Tr
-00004230: 7565 290a 2020 2020 2020 2020 6669 7273  ue).        firs
-00004240: 745f 756e 6971 7565 5f6d 6174 6368 5b22  t_unique_match["
-00004250: 5044 425f 3122 5d20 3d20 6669 7273 745f  PDB_1"] = first_
-00004260: 756e 6971 7565 5f6d 6174 6368 5b27 5044  unique_match['PD
-00004270: 425f 3127 5d2e 7374 722e 7265 706c 6163  B_1'].str.replac
-00004280: 6528 7227 3b27 2c20 2727 290a 2020 2020  e(r';', '').    
-00004290: 2020 2020 7064 625f 6c69 7374 203d 2066      pdb_list = f
-000042a0: 6972 7374 5f75 6e69 7175 655f 6d61 7463  irst_unique_matc
-000042b0: 685b 2250 4442 5f31 225d 2e74 6f6c 6973  h["PDB_1"].tolis
-000042c0: 7428 290a 2020 2020 2020 2020 7265 7320  t().        res 
-000042d0: 3d20 5b22 5265 7369 6475 6573 225d 202a  = ["Residues"] *
-000042e0: 206c 656e 2870 6462 5f6c 6973 7429 0a20   len(pdb_list). 
-000042f0: 2020 2020 2020 2072 6573 6964 7565 735f         residues_
-00004300: 6c69 7374 203d 2066 6972 7374 5f75 6e69  list = first_uni
-00004310: 7175 655f 6d61 7463 685b 2252 6573 6964  que_match["Resid
-00004320: 7565 7322 5d2e 7374 722e 7265 706c 6163  ues"].str.replac
-00004330: 6528 7227 3b27 2c20 2727 292e 746f 6c69  e(r';', '').toli
-00004340: 7374 2829 0a20 2020 2020 2020 2063 6861  st().        cha
-00004350: 696e 203d 205b 2243 6861 696e 225d 202a  in = ["Chain"] *
-00004360: 206c 656e 2870 6462 5f6c 6973 7429 0a20   len(pdb_list). 
-00004370: 2020 2020 2020 2063 6861 696e 735f 6c69         chains_li
-00004380: 7374 203d 2066 6972 7374 5f75 6e69 7175  st = first_uniqu
-00004390: 655f 6d61 7463 685b 2243 6861 696e 225d  e_match["Chain"]
-000043a0: 2e73 7472 2e72 6570 6c61 6365 2872 273b  .str.replace(r';
-000043b0: 272c 2027 2729 2e74 6f6c 6973 7428 290a  ', '').tolist().
-000043c0: 2020 2020 2020 2020 7066 616d 5f64 6963          pfam_dic
-000043d0: 7420 3d20 7b61 3a20 7b62 3a20 632c 2064  t = {a: {b: c, d
-000043e0: 3a20 657d 2066 6f72 2061 2c20 622c 2063  : e} for a, b, c
-000043f0: 2c20 642c 2065 2069 6e20 7a69 7028 7064  , d, e in zip(pd
-00004400: 625f 6c69 7374 2c20 7265 732c 2072 6573  b_list, res, res
-00004410: 6964 7565 735f 6c69 7374 2c20 6368 6169  idues_list, chai
-00004420: 6e2c 2063 6861 696e 735f 6c69 7374 297d  n, chains_list)}
-00004430: 0a20 2020 2065 6c73 653a 2023 3634 3120  .    else: #641 
-00004440: 2363 6f6e 7461 696e 7320 616c 6c20 7064  #contains all pd
-00004450: 6220 6669 6c65 7320 2864 7570 6c69 6361  b files (duplica
-00004460: 7465 7329 2066 6f72 2065 6163 6820 7365  tes) for each se
-00004470: 7175 656e 6365 0a20 2020 2020 2020 2066  quence.        f
-00004480: 616d 696c 795f 6772 6f75 7073 203d 2066  amily_groups = f
-00004490: 616d 696c 792e 6772 6f75 7062 7928 2755  amily.groupby('U
-000044a0: 6e69 7072 6f74 2729 5b5b 2250 4442 5f31  niprot')[["PDB_1
-000044b0: 222c 2022 4368 6169 6e22 2c20 2246 756e  ", "Chain", "Fun
-000044c0: 6374 696f 6e22 2c20 2246 616d 696c 7922  ction", "Family"
-000044d0: 2c20 2252 6573 6964 7565 7322 5d5d 2e61  , "Residues"]].a
-000044e0: 6767 286c 616d 6264 6120 783a 206c 6973  gg(lambda x: lis
-000044f0: 7428 7829 290a 2020 2020 2020 2020 7064  t(x)).        pd
-00004500: 625f 6c69 7374 203d 7375 6d28 6661 6d69  b_list =sum(fami
-00004510: 6c79 5f67 726f 7570 735b 2250 4442 5f31  ly_groups["PDB_1
-00004520: 225d 2e74 6f6c 6973 7428 292c 5b5d 2920  "].tolist(),[]) 
-00004530: 236d 6170 2873 7472 2c20 7661 6c75 655f  #map(str, value_
-00004540: 6c69 7374 2920 232e 7374 722e 7265 706c  list) #.str.repl
-00004550: 6163 6528 7227 3b27 2c20 2727 290a 2020  ace(r';', '').  
-00004560: 2020 2020 2020 7064 625f 6c69 7374 203d        pdb_list =
-00004570: 205b 7064 622e 7265 706c 6163 6528 7227   [pdb.replace(r'
-00004580: 3b27 2c20 2727 2920 666f 7220 7064 6220  ;', '') for pdb 
-00004590: 696e 2070 6462 5f6c 6973 745d 0a20 2020  in pdb_list].   
-000045a0: 2020 2020 2072 6573 203d 205b 2252 6573       res = ["Res
-000045b0: 6964 7565 7322 5d2a 6c65 6e28 7064 625f  idues"]*len(pdb_
-000045c0: 6c69 7374 290a 2020 2020 2020 2020 7265  list).        re
-000045d0: 7369 6475 6573 5f6c 6973 743d 7375 6d28  sidues_list=sum(
-000045e0: 6661 6d69 6c79 5f67 726f 7570 735b 2252  family_groups["R
-000045f0: 6573 6964 7565 7322 5d2e 746f 6c69 7374  esidues"].tolist
-00004600: 2829 2c5b 5d29 0a20 2020 2020 2020 2072  (),[]).        r
-00004610: 6573 6964 7565 735f 6c69 7374 203d 205b  esidues_list = [
-00004620: 7265 732e 7265 706c 6163 6528 7227 3b27  res.replace(r';'
-00004630: 2c20 2727 2920 666f 7220 7265 7320 696e  , '') for res in
-00004640: 2072 6573 6964 7565 735f 6c69 7374 5d0a   residues_list].
-00004650: 2020 2020 2020 2020 6368 6169 6e20 3d20          chain = 
-00004660: 5b22 4368 6169 6e22 5d20 2a20 6c65 6e28  ["Chain"] * len(
-00004670: 7064 625f 6c69 7374 290a 2020 2020 2020  pdb_list).      
-00004680: 2020 6368 6169 6e73 5f6c 6973 7420 3d20    chains_list = 
-00004690: 7375 6d28 6661 6d69 6c79 5f67 726f 7570  sum(family_group
-000046a0: 735b 2243 6861 696e 225d 2e74 6f6c 6973  s["Chain"].tolis
-000046b0: 7428 292c 5b5d 290a 2020 2020 2020 2020  t(),[]).        
-000046c0: 6368 6169 6e73 5f6c 6973 7420 3d20 5b63  chains_list = [c
-000046d0: 6861 696e 5b2d 325d 2066 6f72 2063 6861  hain[-2] for cha
-000046e0: 696e 2069 6e20 6368 6169 6e73 5f6c 6973  in in chains_lis
-000046f0: 745d 0a20 2020 2020 2020 2070 6661 6d5f  t].        pfam_
-00004700: 6469 6374 203d 207b 613a 7b62 3a63 2c64  dict = {a:{b:c,d
-00004710: 3a65 7d20 666f 7220 612c 2062 2c20 632c  :e} for a, b, c,
-00004720: 642c 6520 696e 207a 6970 2870 6462 5f6c  d,e in zip(pdb_l
-00004730: 6973 742c 2072 6573 2c72 6573 6964 7565  ist, res,residue
-00004740: 735f 6c69 7374 2c63 6861 696e 2c20 6368  s_list,chain, ch
-00004750: 6169 6e73 5f6c 6973 7429 7d0a 2020 2020  ains_list)}.    
-00004760: 7265 7475 726e 2070 6661 6d5f 6469 6374  return pfam_dict
-00004770: 2c20 7064 625f 6c69 7374 0a64 6566 2074  , pdb_list.def t
-00004780: 7265 655f 7061 6972 5f66 6f72 5f72 6f6f  ree_pair_for_roo
-00004790: 7469 6e67 2864 6973 7461 6e63 655f 6d61  ting(distance_ma
-000047a0: 7472 6978 293a 0a20 2020 2022 2222 4669  trix):.    """Fi
-000047b0: 6e64 7320 7468 6520 7061 6972 206f 6620  nds the pair of 
-000047c0: 7365 7175 656e 6365 7320 7769 7468 2074  sequences with t
-000047d0: 6865 206c 6172 6765 7374 2070 6169 7277  he largest pairw
-000047e0: 6973 6520 6469 7374 616e 6365 2074 6f20  ise distance to 
-000047f0: 7365 7420 6173 2061 6e20 6f75 7467 726f  set as an outgro
-00004800: 7570 2074 6f20 666f 726d 2074 6865 2072  up to form the r
-00004810: 6f6f 7420 6f66 2061 6e20 756e 726f 6f74  oot of an unroot
-00004820: 6564 2074 7265 650a 2020 2020 3a70 6172  ed tree.    :par
-00004830: 616d 2070 616e 6461 7320 6461 7461 6672  am pandas datafr
-00004840: 616d 6520 6469 7374 616e 6365 5f6d 6174  ame distance_mat
-00004850: 7269 783a 2063 6f6e 7461 696e 7320 7061  rix: contains pa
-00004860: 6972 7769 7365 2064 6973 7461 6e63 6573  irwise distances
-00004870: 2061 6372 6f73 7320 7468 6520 7365 7175   across the sequ
-00004880: 656e 6365 7320 696e 2074 6865 2074 7265  ences in the tre
-00004890: 6522 2222 0a0a 2020 2020 736f 7274 6564  e"""..    sorted
-000048a0: 5f64 6973 7461 6e63 655f 6d61 7472 6978  _distance_matrix
-000048b0: 203d 2064 6973 7461 6e63 655f 6d61 7472   = distance_matr
-000048c0: 6978 5b64 6973 7461 6e63 655f 6d61 7472  ix[distance_matr
-000048d0: 6978 2e67 7428 3029 5d2e 7374 6163 6b28  ix.gt(0)].stack(
-000048e0: 292e 736f 7274 5f76 616c 7565 7328 292e  ).sort_values().
-000048f0: 746f 5f66 7261 6d65 2829 0a20 2020 2073  to_frame().    s
-00004900: 6f72 7465 645f 6469 7374 616e 6365 5f6d  orted_distance_m
-00004910: 6174 7269 782e 7265 7365 745f 696e 6465  atrix.reset_inde
-00004920: 7828 6c65 7665 6c3d 302c 2069 6e70 6c61  x(level=0, inpla
-00004930: 6365 3d54 7275 6529 0a20 2020 2073 6f72  ce=True).    sor
-00004940: 7465 645f 6469 7374 616e 6365 5f6d 6174  ted_distance_mat
-00004950: 7269 782e 636f 6c75 6d6e 7320 3d20 5b22  rix.columns = ["
-00004960: 5365 7175 656e 6365 5f30 222c 2022 4469  Sequence_0", "Di
-00004970: 7374 616e 6365 225d 0a20 2020 2073 6f72  stance"].    sor
-00004980: 7465 645f 6469 7374 616e 6365 5f6d 6174  ted_distance_mat
-00004990: 7269 785b 2753 6571 7565 6e63 655f 3127  rix['Sequence_1'
-000049a0: 5d20 3d20 736f 7274 6564 5f64 6973 7461  ] = sorted_dista
-000049b0: 6e63 655f 6d61 7472 6978 2e69 6e64 6578  nce_matrix.index
-000049c0: 0a20 2020 2073 6f72 7465 645f 6469 7374  .    sorted_dist
-000049d0: 616e 6365 5f6d 6174 7269 7820 3d20 736f  ance_matrix = so
-000049e0: 7274 6564 5f64 6973 7461 6e63 655f 6d61  rted_distance_ma
-000049f0: 7472 6978 5b5b 2253 6571 7565 6e63 655f  trix[["Sequence_
-00004a00: 3122 2c20 2253 6571 7565 6e63 655f 3022  1", "Sequence_0"
-00004a10: 2c20 2244 6973 7461 6e63 6522 5d5d 0a20  , "Distance"]]. 
-00004a20: 2020 2073 6f72 7465 645f 6469 7374 616e     sorted_distan
-00004a30: 6365 5f6d 6174 7269 782e 7265 7365 745f  ce_matrix.reset_
-00004a40: 696e 6465 7828 696e 706c 6163 653d 5472  index(inplace=Tr
-00004a50: 7565 290a 2020 2020 736f 7274 6564 5f64  ue).    sorted_d
-00004a60: 6973 7461 6e63 655f 6d61 7472 6978 2e64  istance_matrix.d
-00004a70: 726f 7028 5b22 696e 6465 7822 5d2c 2069  rop(["index"], i
-00004a80: 6e70 6c61 6365 3d54 7275 652c 2061 7869  nplace=True, axi
-00004a90: 733d 3129 0a20 2020 2023 2073 6f72 7465  s=1).    # sorte
-00004aa0: 645f 6469 7374 616e 6365 5f6d 6174 7269  d_distance_matri
-00004ab0: 782e 6472 6f70 5f64 7570 6c69 6361 7465  x.drop_duplicate
-00004ac0: 7328 7375 6273 6574 3d5b 2744 6973 7461  s(subset=['Dista
-00004ad0: 6e63 6527 5d29 0a20 2020 2073 6f72 7465  nce']).    sorte
-00004ae0: 645f 6469 7374 616e 6365 5f6d 6174 7269  d_distance_matri
-00004af0: 7820 3d20 736f 7274 6564 5f64 6973 7461  x = sorted_dista
-00004b00: 6e63 655f 6d61 7472 6978 5b7e 736f 7274  nce_matrix[~sort
-00004b10: 6564 5f64 6973 7461 6e63 655f 6d61 7472  ed_distance_matr
-00004b20: 6978 5b5b 2753 6571 7565 6e63 655f 3127  ix[['Sequence_1'
-00004b30: 2c20 2753 6571 7565 6e63 655f 3027 5d5d  , 'Sequence_0']]
-00004b40: 2e61 7070 6c79 2866 726f 7a65 6e73 6574  .apply(frozenset
-00004b50: 2c61 7869 733d 3129 2e64 7570 6c69 6361  ,axis=1).duplica
-00004b60: 7465 6428 295d 2020 2320 5265 6d6f 7665  ted()]  # Remove
-00004b70: 2072 6570 6561 7465 6420 636f 6d62 696e   repeated combin
-00004b80: 6174 696f 6e73 206f 6620 7365 7175 656e  ations of sequen
-00004b90: 6365 730a 2020 2020 736f 7274 6564 5f64  ces.    sorted_d
-00004ba0: 6973 7461 6e63 655f 6d61 7472 6978 203d  istance_matrix =
-00004bb0: 2073 6f72 7465 645f 6469 7374 616e 6365   sorted_distance
-00004bc0: 5f6d 6174 7269 782e 7265 7365 745f 696e  _matrix.reset_in
-00004bd0: 6465 7828 6472 6f70 3d54 7275 6529 0a20  dex(drop=True). 
-00004be0: 2020 2072 616e 6b20 3d20 736f 7274 6564     rank = sorted
-00004bf0: 5f64 6973 7461 6e63 655f 6d61 7472 6978  _distance_matrix
-00004c00: 2e73 6861 7065 5b30 5d20 2d31 0a20 2020  .shape[0] -1.   
-00004c10: 2073 6571 7565 6e63 655f 3020 3d20 736f   sequence_0 = so
-00004c20: 7274 6564 5f64 6973 7461 6e63 655f 6d61  rted_distance_ma
-00004c30: 7472 6978 2e6c 6f63 5b72 616e 6b5d 2e53  trix.loc[rank].S
-00004c40: 6571 7565 6e63 655f 300a 2020 2020 7365  equence_0.    se
-00004c50: 7175 656e 6365 5f31 203d 2073 6f72 7465  quence_1 = sorte
-00004c60: 645f 6469 7374 616e 6365 5f6d 6174 7269  d_distance_matri
-00004c70: 782e 6c6f 635b 7261 6e6b 5d2e 5365 7175  x.loc[rank].Sequ
-00004c80: 656e 6365 5f31 0a20 2020 2072 6574 7572  ence_1.    retur
-00004c90: 6e20 7365 7175 656e 6365 5f30 2c73 6571  n sequence_0,seq
-00004ca0: 7565 6e63 655f 310a 6465 6620 646f 776e  uence_1.def down
-00004cb0: 6c6f 6164 5f50 4442 2866 696c 6573 5f6c  load_PDB(files_l
-00004cc0: 6973 742c 2064 6972 6563 746f 7279 293a  ist, directory):
-00004cd0: 0a20 2020 2022 2222 5265 6164 7320 6120  .    """Reads a 
-00004ce0: 6c69 7374 206f 6620 5044 425f 6669 6c65  list of PDB_file
-00004cf0: 7320 616e 6420 646f 776e 6c6f 6164 7320  s and downloads 
-00004d00: 7468 656d 2074 6f20 6120 666f 6c64 6572  them to a folder
-00004d10: 0a20 2020 203a 7061 7261 6d20 6c69 7374  .    :param list
-00004d20: 2066 696c 6573 5f6c 6973 743a 206c 6973   files_list: lis
-00004d30: 7420 6f66 2050 4442 2066 696c 6573 206e  t of PDB files n
-00004d40: 616d 6573 2074 6f20 646f 776e 6c6f 6164  ames to download
-00004d50: 0a20 2020 203a 7061 7261 6d20 7374 7220  .    :param str 
-00004d60: 6469 7265 6374 6f72 793a 2066 6f6c 6465  directory: folde
-00004d70: 7220 6f72 2064 6972 6563 746f 7279 2077  r or directory w
-00004d80: 6865 7265 2074 6f20 646f 776e 6c6f 6164  here to download
-00004d90: 2074 6865 2066 696c 6573 2222 220a 2020   the files""".  
-00004da0: 2020 7064 626c 203d 2050 4442 4c69 7374    pdbl = PDBList
-00004db0: 2829 0a20 2020 2066 6f72 2069 2c20 6669  ().    for i, fi
-00004dc0: 6c65 2069 6e20 656e 756d 6572 6174 6528  le in enumerate(
-00004dd0: 6669 6c65 735f 6c69 7374 293a 0a20 2020  files_list):.   
-00004de0: 2020 2020 2070 6462 6c2e 7265 7472 6965       pdbl.retrie
-00004df0: 7665 5f70 6462 5f66 696c 6528 6669 6c65  ve_pdb_file(file
-00004e00: 2c20 7064 6972 3d64 6972 6563 746f 7279  , pdir=directory
-00004e10: 2c20 6f62 736f 6c65 7465 3d46 616c 7365  , obsolete=False
-00004e20: 2c20 6669 6c65 5f66 6f72 6d61 743d 2270  , file_format="p
-00004e30: 6462 2229 0a64 6566 2063 6f6e 7665 7274  db").def convert
-00004e40: 5f74 6f5f 7061 6e64 6173 2864 6973 7461  _to_pandas(dista
-00004e50: 6e63 655f 6d61 7472 6978 293a 0a20 2020  nce_matrix):.   
-00004e60: 2022 2222 436f 6e76 6572 7473 2061 2062   """Converts a b
-00004e70: 696f 7079 7468 6f6e 2070 6169 7277 6973  iopython pairwis
-00004e80: 6520 6469 7374 616e 6365 206d 6174 7269  e distance matri
-00004e90: 7820 696e 746f 2061 2073 796d 6d65 7472  x into a symmetr
-00004ea0: 6963 2070 616e 6461 7320 6461 7461 6672  ic pandas datafr
-00004eb0: 616d 650a 2020 2020 3a70 6172 616d 2062  ame.    :param b
-00004ec0: 696f 7079 7468 6f6e 2063 616c 6375 6c61  iopython calcula
-00004ed0: 746f 7220 6d61 7472 6978 3a20 7061 6972  tor matrix: pair
-00004ee0: 7769 7365 2064 6973 7461 6e63 6520 6d61  wise distance ma
-00004ef0: 7472 6978 0a20 2020 2022 2222 0a20 2020  trix.    """.   
-00004f00: 2062 203d 206e 702e 7a65 726f 7328 5b6c   b = np.zeros([l
-00004f10: 656e 2864 6973 7461 6e63 655f 6d61 7472  en(distance_matr
-00004f20: 6978 2e6d 6174 7269 7829 2c20 6c65 6e28  ix.matrix), len(
-00004f30: 6d61 7828 6469 7374 616e 6365 5f6d 6174  max(distance_mat
-00004f40: 7269 782e 6d61 7472 6978 2c20 6b65 793d  rix.matrix, key=
-00004f50: 6c61 6d62 6461 2078 3a20 6c65 6e28 7829  lambda x: len(x)
-00004f60: 2929 5d29 0a20 2020 2066 6f72 2069 2c20  ))]).    for i, 
-00004f70: 6a20 696e 2065 6e75 6d65 7261 7465 2864  j in enumerate(d
-00004f80: 6973 7461 6e63 655f 6d61 7472 6978 2e6d  istance_matrix.m
-00004f90: 6174 7269 7829 3a0a 2020 2020 2020 2020  atrix):.        
-00004fa0: 625b 695d 5b30 3a6c 656e 286a 295d 203d  b[i][0:len(j)] =
-00004fb0: 206a 0a20 2020 2023 6466 203d 2070 642e   j.    #df = pd.
-00004fc0: 4461 7461 4672 616d 6528 622c 2069 6e64  DataFrame(b, ind
-00004fd0: 6578 3d44 6973 7461 6e63 654d 6174 7269  ex=DistanceMatri
-00004fe0: 782e 6e61 6d65 732c 2063 6f6c 756d 6e73  x.names, columns
-00004ff0: 3d44 6973 7461 6e63 654d 6174 7269 782e  =DistanceMatrix.
-00005000: 6e61 6d65 7329 0a20 2020 2062 5f74 7261  names).    b_tra
-00005010: 6e73 706f 7365 203d 206e 702e 7472 616e  nspose = np.tran
-00005020: 7370 6f73 6528 6229 0a20 2020 2062 203d  spose(b).    b =
-00005030: 2062 202b 2062 5f74 7261 6e73 706f 7365   b + b_transpose
-00005040: 0a20 2020 2064 6620 3d20 7064 2e44 6174  .    df = pd.Dat
-00005050: 6146 7261 6d65 2862 2c20 696e 6465 783d  aFrame(b, index=
-00005060: 6469 7374 616e 6365 5f6d 6174 7269 782e  distance_matrix.
-00005070: 6e61 6d65 732c 2063 6f6c 756d 6e73 3d64  names, columns=d
-00005080: 6973 7461 6e63 655f 6d61 7472 6978 2e6e  istance_matrix.n
-00005090: 616d 6573 290a 2020 2020 7265 7475 726e  ames).    return
-000050a0: 2020 6466 0a64 6566 2069 6e66 6572 5f74    df.def infer_t
-000050b0: 7265 6528 616c 6967 6e6d 656e 742c 2061  ree(alignment, a
-000050c0: 6c69 676e 6d65 6e74 5f66 696c 655f 6e61  lignment_file_na
-000050d0: 6d65 2c6e 616d 652c 6d65 7468 6f64 3d4e  me,name,method=N
-000050e0: 6f6e 652c 7472 6565 5f66 696c 655f 6e61  one,tree_file_na
-000050f0: 6d65 3d4e 6f6e 652c 7472 6565 5f66 696c  me=None,tree_fil
-00005100: 653d 4e6f 6e65 2c73 746f 7261 6765 5f66  e=None,storage_f
-00005110: 6f6c 6465 723d 2222 293a 0a20 2020 2022  older=""):.    "
-00005120: 2222 2050 6572 666f 726d 7320 7472 6565  "" Performs tree
-00005130: 2069 6e66 6572 656e 6365 206f 7220 7265   inference or re
-00005140: 6164 7320 616e 2069 6e70 7574 2067 6976  ads an input giv
-00005150: 656e 2074 7265 652c 2072 6574 7572 6e73  en tree, returns
-00005160: 2061 6e20 6574 6533 2074 7265 6520 666f   an ete3 tree fo
-00005170: 726d 6174 6564 2074 7265 650a 2020 2020  rmated tree.    
-00005180: 3a70 6172 616d 2062 696f 7079 7468 6f6e  :param biopython
-00005190: 2061 6c69 676e 6d65 6e74 2061 6c69 676e   alignment align
-000051a0: 6d65 6e74 3a20 6269 6f70 7974 686f 6e20  ment: biopython 
-000051b0: 616c 6967 6e6d 656e 7420 636c 6173 730a  alignment class.
-000051c0: 2020 2020 3a70 6172 616d 2073 7472 2061      :param str a
-000051d0: 6c69 676e 6d65 6e74 5f66 696c 655f 6e61  lignment_file_na
-000051e0: 6d65 3a20 7061 7468 2074 6f20 616c 6967  me: path to alig
-000051f0: 6e6d 656e 7420 6669 6c65 0a20 2020 203a  nment file.    :
-00005200: 7061 7261 6d20 7374 7220 6e61 6d65 3a20  param str name: 
-00005210: 6461 7461 7365 7420 7072 6f6a 6563 7420  dataset project 
-00005220: 6e61 6d65 0a20 2020 203a 7061 7261 6d20  name.    :param 
-00005230: 7374 7220 6d65 7468 6f64 3a20 7472 6565  str method: tree
-00005240: 2069 6e66 6572 656e 6365 206d 6574 686f   inference metho
-00005250: 6420 2869 6620 7472 6565 2069 7320 6e6f  d (if tree is no
-00005260: 7420 6769 7665 6e29 0a20 2020 203a 7061  t given).    :pa
-00005270: 7261 6d20 7374 7220 7472 6565 5f66 696c  ram str tree_fil
-00005280: 655f 6e61 6d65 3a20 6e61 6d65 2074 6f20  e_name: name to 
-00005290: 6769 7665 2074 6f20 7468 6520 7472 6565  give to the tree
-000052a0: 2066 696c 650a 2020 2020 2e70 6172 616d   file.    .param
-000052b0: 2073 7472 2074 7265 655f 6669 6c65 3a70   str tree_file:p
-000052c0: 6174 6820 746f 2061 2070 6f73 7369 626c  ath to a possibl
-000052d0: 6520 6769 7665 6e20 7472 6565 2069 6e20  e given tree in 
-000052e0: 6e65 7769 636b 2066 6f72 6d61 7420 310a  newick format 1.
-000052f0: 2020 2020 3a70 6172 616d 2073 7472 2073      :param str s
-00005300: 746f 7261 6765 5f66 6f6c 6465 723a 2066  torage_folder: f
-00005310: 6f6c 6465 7220 7768 6572 6520 746f 2073  older where to s
-00005320: 746f 7265 2074 6865 2072 6573 756c 7473  tore the results
-00005330: 206f 6620 7468 6520 7472 6565 2069 6e66   of the tree inf
-00005340: 6572 656e 6365 0a0a 2020 2020 2222 220a  erence..    """.
-00005350: 2020 2020 6966 2074 7265 655f 6669 6c65      if tree_file
-00005360: 3a0a 2020 2020 2020 2020 7072 696e 7428  :.        print(
-00005370: 2255 7369 6e67 2067 6976 656e 2074 7265  "Using given tre
-00005380: 6520 6669 6c65 2e2e 2e22 290a 2020 2020  e file...").    
-00005390: 2020 2020 7472 6565 203d 2054 7265 6545      tree = TreeE
-000053a0: 7465 3328 7472 6565 5f66 696c 652c 666f  te3(tree_file,fo
-000053b0: 726d 6174 3d31 2c71 756f 7465 645f 6e6f  rmat=1,quoted_no
-000053c0: 6465 5f6e 616d 6573 3d54 7275 6529 0a20  de_names=True). 
-000053d0: 2020 2020 2020 2072 6574 7572 6e20 7472         return tr
-000053e0: 6565 0a0a 2020 2020 656c 7365 3a0a 2020  ee..    else:.  
-000053f0: 2020 2020 2020 2320 5061 6972 7769 7365        # Pairwise
-00005400: 2064 6973 7461 6e63 6520 6d61 7472 6978   distance matrix
-00005410: 0a20 2020 2020 2020 2070 7269 6e74 2822  .        print("
-00005420: 4275 696c 6469 6e67 2064 6973 7461 6e63  Building distanc
-00005430: 6520 6d61 7472 6963 6573 2061 6e64 207b  e matrices and {
-00005440: 7d20 7472 6565 2e2e 2e22 2e66 6f72 6d61  } tree...".forma
-00005450: 7428 6d65 7468 6f64 2929 0a20 2020 2020  t(method)).     
-00005460: 2020 2069 6620 6c65 6e28 616c 6967 6e6d     if len(alignm
-00005470: 656e 7429 203c 2032 3030 2061 6e64 206d  ent) < 200 and m
-00005480: 6574 686f 6420 696e 205b 226e 6a22 2c22  ethod in ["nj","
-00005490: 6e6a 5f72 6f6f 7465 6422 2c22 7570 676d  nj_rooted","upgm
-000054a0: 6122 5d3a 0a20 2020 2020 2020 2020 2020  a"]:.           
-000054b0: 2063 616c 6375 6c61 746f 7220 3d20 4469   calculator = Di
-000054c0: 7374 616e 6365 4361 6c63 756c 6174 6f72  stanceCalculator
-000054d0: 2827 626c 6f73 756d 3632 2729 2020 2320  ('blosum62')  # 
-000054e0: 444e 4120 2d2d 2d3e 2049 6465 6e74 6974  DNA ---> Identit
-000054f0: 792f 2f20 5072 6f74 6569 6e20 2d2d 2d3e  y// Protein --->
-00005500: 2062 6c6f 7375 6d36 320a 2020 2020 2020   blosum62.      
-00005510: 2020 2020 2020 6469 7374 616e 6365 5f6d        distance_m
-00005520: 6174 7269 785f 6361 6c20 3d20 6361 6c63  atrix_cal = calc
-00005530: 756c 6174 6f72 2e67 6574 5f64 6973 7461  ulator.get_dista
-00005540: 6e63 6528 616c 6967 6e6d 656e 7429 0a20  nce(alignment). 
-00005550: 2020 2020 2020 2020 2020 2064 6973 7461             dista
-00005560: 6e63 655f 6d61 7472 6978 5f63 616c 5f70  nce_matrix_cal_p
-00005570: 616e 6461 7320 3d20 636f 6e76 6572 745f  andas = convert_
-00005580: 746f 5f70 616e 6461 7328 6469 7374 616e  to_pandas(distan
-00005590: 6365 5f6d 6174 7269 785f 6361 6c29 0a20  ce_matrix_cal). 
-000055a0: 2020 2020 2020 2020 2020 2064 6973 7461             dista
-000055b0: 6e63 655f 6d61 7472 6978 5f63 616c 5f70  nce_matrix_cal_p
-000055c0: 616e 6461 732e 746f 5f63 7376 2822 7b7d  andas.to_csv("{}
-000055d0: 2f7b 7d5f 6469 7374 616e 6365 5f6d 6174  /{}_distance_mat
-000055e0: 7269 782e 6373 7622 2e66 6f72 6d61 7428  rix.csv".format(
-000055f0: 7374 6f72 6167 655f 666f 6c64 6572 2c6e  storage_folder,n
-00005600: 616d 6529 290a 2020 2020 2020 2020 2020  ame)).          
-00005610: 2020 2368 7474 7073 3a2f 2f73 7461 636b    #https://stack
-00005620: 6f76 6572 666c 6f77 2e63 6f6d 2f71 7565  overflow.com/que
-00005630: 7374 696f 6e73 2f33 3032 3437 3335 392f  stions/30247359/
-00005640: 686f 772d 646f 6573 2d62 696f 7079 7468  how-does-biopyth
-00005650: 6f6e 2d64 6574 6572 6d69 6e65 2d74 6865  on-determine-the
-00005660: 2d72 6f6f 742d 6f66 2d61 2d70 6879 6c6f  -root-of-a-phylo
-00005670: 6765 6e65 7469 632d 7472 6565 0a20 2020  genetic-tree.   
-00005680: 2020 2020 2020 2020 2069 6620 6d65 7468           if meth
-00005690: 6f64 203d 3d20 226e 6a22 3a0a 2020 2020  od == "nj":.    
-000056a0: 2020 2020 2020 2020 2020 2020 7072 696e              prin
-000056b0: 7428 2254 7265 6520 696e 6665 7265 6e63  t("Tree inferenc
-000056c0: 6520 7669 6120 4e65 6967 6862 6f75 7220  e via Neighbour 
-000056d0: 4a6f 696e 696e 6720 4e4f 5420 726f 6f74  Joining NOT root
-000056e0: 6564 206d 6574 686f 642e 2e2e 2229 0a20  ed method..."). 
-000056f0: 2020 2020 2020 2020 2020 2020 2020 2063                 c
-00005700: 6f6e 7374 7275 6374 6f72 203d 2044 6973  onstructor = Dis
-00005710: 7461 6e63 6554 7265 6543 6f6e 7374 7275  tanceTreeConstru
-00005720: 6374 6f72 286d 6574 686f 643d 226e 6a22  ctor(method="nj"
-00005730: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
-00005740: 2020 7472 6565 203d 2063 6f6e 7374 7275    tree = constru
-00005750: 6374 6f72 2e6e 6a28 6469 7374 616e 6365  ctor.nj(distance
-00005760: 5f6d 6174 7269 785f 6361 6c29 0a20 2020  _matrix_cal).   
-00005770: 2020 2020 2020 2020 2020 2020 2074 7265               tre
-00005780: 6520 3d20 746f 5f65 7465 3328 7472 6565  e = to_ete3(tree
-00005790: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
-000057a0: 2020 7265 7475 726e 2074 7265 650a 2020    return tree.  
-000057b0: 2020 2020 2020 2020 2020 656c 6966 206d            elif m
-000057c0: 6574 686f 6420 3d3d 2022 6e6a 5f72 6f6f  ethod == "nj_roo
-000057d0: 7465 6422 3a0a 2020 2020 2020 2020 2020  ted":.          
-000057e0: 2020 2020 2020 7072 696e 7428 2254 7265        print("Tre
-000057f0: 6520 696e 6665 7265 6e63 6520 7669 6120  e inference via 
-00005800: 4e65 6967 6862 6f75 7220 4a6f 696e 696e  Neighbour Joinin
-00005810: 6720 7769 7468 2061 6464 6974 696f 6e61  g with additiona
-00005820: 6c20 726f 6f74 696e 6720 6d65 7468 6f64  l rooting method
-00005830: 2e2e 2e22 290a 2020 2020 2020 2020 2020  ...").          
-00005840: 2020 2020 2020 636f 6e73 7472 7563 746f        constructo
-00005850: 7220 3d20 4469 7374 616e 6365 5472 6565  r = DistanceTree
-00005860: 436f 6e73 7472 7563 746f 7228 6d65 7468  Constructor(meth
-00005870: 6f64 3d22 6e6a 2229 0a20 2020 2020 2020  od="nj").       
-00005880: 2020 2020 2020 2020 2074 7265 6520 3d20           tree = 
-00005890: 636f 6e73 7472 7563 746f 722e 6e6a 2864  constructor.nj(d
-000058a0: 6973 7461 6e63 655f 6d61 7472 6978 5f63  istance_matrix_c
-000058b0: 616c 290a 2020 2020 2020 2020 2020 2020  al).            
-000058c0: 2020 2020 7472 6565 203d 2074 6f5f 6574      tree = to_et
-000058d0: 6533 2874 7265 6529 0a20 2020 2020 2020  e3(tree).       
-000058e0: 2020 2020 2020 2020 2023 204d 616b 696e           # Makin
-000058f0: 6720 6120 726f 6f74 3a0a 2020 2020 2020  g a root:.      
-00005900: 2020 2020 2020 2020 2020 7365 7175 656e            sequen
-00005910: 6365 5f30 2c20 7365 7175 656e 6365 5f31  ce_0, sequence_1
-00005920: 203d 2074 7265 655f 7061 6972 5f66 6f72   = tree_pair_for
-00005930: 5f72 6f6f 7469 6e67 2864 6973 7461 6e63  _rooting(distanc
-00005940: 655f 6d61 7472 6978 5f63 616c 5f70 616e  e_matrix_cal_pan
-00005950: 6461 7329 0a20 2020 2020 2020 2020 2020  das).           
-00005960: 2020 2020 2074 7265 652e 7365 745f 6f75       tree.set_ou
-00005970: 7467 726f 7570 2874 7265 6520 2620 7365  tgroup(tree & se
-00005980: 7175 656e 6365 5f30 290a 2020 2020 2020  quence_0).      
-00005990: 2020 2020 2020 2020 2020 2320 616e 6365            # ance
-000059a0: 7374 6f72 203d 2074 7265 652e 6765 745f  stor = tree.get_
-000059b0: 636f 6d6d 6f6e 5f61 6e63 6573 746f 7228  common_ancestor(
-000059c0: 7365 7175 656e 6365 5f30 2c20 7365 7175  sequence_0, sequ
-000059d0: 656e 6365 5f31 290a 2020 2020 2020 2020  ence_1).        
-000059e0: 2020 2020 2020 2020 2320 7472 6565 2e73          # tree.s
-000059f0: 6574 5f6f 7574 6772 6f75 7028 616e 6365  et_outgroup(ance
-00005a00: 7374 6f72 290a 2020 2020 2020 2020 2020  stor).          
-00005a10: 2020 2020 2020 7265 7475 726e 2074 7265        return tre
-00005a20: 650a 2020 2020 2020 2020 2020 2020 656c  e.            el
-00005a30: 6966 206d 6574 686f 6420 3d3d 2022 7570  if method == "up
-00005a40: 676d 6122 3a0a 2020 2020 2020 2020 2020  gma":.          
-00005a50: 2020 2020 2020 7072 696e 7428 2254 7265        print("Tre
-00005a60: 6520 696e 6665 7265 6e63 6520 7669 6120  e inference via 
-00005a70: 5570 676d 6120 726f 6f74 6564 206d 6574  Upgma rooted met
-00005a80: 686f 642e 2e2e 2229 0a20 2020 2020 2020  hod...").       
-00005a90: 2020 2020 2020 2020 2063 6f6e 7374 7275           constru
-00005aa0: 6374 6f72 203d 2044 6973 7461 6e63 6554  ctor = DistanceT
-00005ab0: 7265 6543 6f6e 7374 7275 6374 6f72 286d  reeConstructor(m
-00005ac0: 6574 686f 643d 2275 7067 6d61 2229 2023  ethod="upgma") #
-00005ad0: 206e 6a20 6d65 7468 6f64 2069 7320 756e   nj method is un
-00005ae0: 726f 6f74 6564 2069 6e20 6269 6f70 7974  rooted in biopyt
-00005af0: 686f 6e2e 2075 7067 6d61 2069 7320 726f  hon. upgma is ro
-00005b00: 6f74 6564 0a20 2020 2020 2020 2020 2020  oted.           
-00005b10: 2020 2020 2074 7265 6520 3d20 636f 6e73       tree = cons
-00005b20: 7472 7563 746f 722e 7570 676d 6128 6469  tructor.upgma(di
-00005b30: 7374 616e 6365 5f6d 6174 7269 785f 6361  stance_matrix_ca
-00005b40: 6c29 0a20 2020 2020 2020 2020 2020 2020  l).             
-00005b50: 2020 2074 7265 6520 3d20 746f 5f65 7465     tree = to_ete
-00005b60: 3328 7472 6565 290a 2020 2020 2020 2020  3(tree).        
-00005b70: 2020 2020 2020 2020 7265 7475 726e 2074          return t
-00005b80: 7265 650a 2020 2020 2020 2020 656c 6966  ree.        elif
-00005b90: 206d 6574 686f 6420 3d3d 2022 6971 7472   method == "iqtr
-00005ba0: 6565 223a 0a20 2020 2020 2020 2020 2020  ee":.           
-00005bb0: 2070 7269 6e74 2822 4971 7472 6565 204d   print("Iqtree M
-00005bc0: 4c20 6d65 7468 6f64 2e2e 2e22 290a 2020  L method...").  
-00005bd0: 2020 2020 2020 2020 2020 616c 6967 6e6d            alignm
-00005be0: 656e 745f 6620 3d20 5b61 6c69 676e 6d65  ent_f = [alignme
-00005bf0: 6e74 5f66 696c 655f 6e61 6d65 2069 6620  nt_file_name if 
-00005c00: 616c 6967 6e6d 656e 745f 6669 6c65 5f6e  alignment_file_n
-00005c10: 616d 6520 656c 7365 2022 7b7d 2f7b 7d2e  ame else "{}/{}.
-00005c20: 6d61 6666 7422 2e66 6f72 6d61 7428 7374  mafft".format(st
-00005c30: 6f72 6167 655f 666f 6c64 6572 2c6e 616d  orage_folder,nam
-00005c40: 6529 5d5b 305d 0a20 2020 2020 2020 2020  e)][0].         
-00005c50: 2020 2074 7265 655f 6669 6c65 5f6e 616d     tree_file_nam
-00005c60: 6520 3d20 616c 6967 6e6d 656e 745f 662e  e = alignment_f.
-00005c70: 7370 6c69 7428 222e 2229 5b30 5d20 2b20  split(".")[0] + 
-00005c80: 222e 7472 6565 6669 6c65 220a 0a20 2020  ".treefile"..   
-00005c90: 2020 2020 2020 2020 2069 6620 6e6f 7420           if not 
-00005ca0: 6f73 2e70 6174 682e 6578 6973 7473 2874  os.path.exists(t
-00005cb0: 7265 655f 6669 6c65 5f6e 616d 6529 3a0a  ree_file_name):.
-00005cc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00005cd0: 232d 6f09 5370 6563 6966 7920 616e 206f  #-o.Specify an o
-00005ce0: 7574 6772 6f75 7020 7461 786f 6e20 6e61  utgroup taxon na
-00005cf0: 6d65 2074 6f20 726f 6f74 2074 6865 2074  me to root the t
-00005d00: 7265 652e 2054 6865 206f 7574 7075 7420  ree. The output 
-00005d10: 7472 6565 2069 6e20 2e74 7265 6566 696c  tree in .treefil
-00005d20: 6520 7769 6c6c 2062 6520 726f 6f74 6564  e will be rooted
-00005d30: 2061 6363 6f72 6469 6e67 6c79 2e20 4445   accordingly. DE
-00005d40: 4641 554c 543a 2066 6972 7374 2074 6178  FAULT: first tax
-00005d50: 6f6e 2069 6e20 616c 6967 6e6d 656e 740a  on in alignment.
-00005d60: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00005d70: 7461 786f 6e5f 726f 6f74 203d 2046 616c  taxon_root = Fal
-00005d80: 7365 0a20 2020 2020 2020 2020 2020 2020  se.             
-00005d90: 2020 2069 6620 7461 786f 6e5f 726f 6f74     if taxon_root
-00005da0: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
-00005db0: 2020 2020 2020 726f 6f74 3d32 310a 2020        root=21.  
+000040f0: 2020 206e 616d 6573 3d5b 2250 4442 5f31     names=["PDB_1
+00004100: 222c 2022 4368 6169 6e22 2c20 2245 6d70  ", "Chain", "Emp
+00004110: 7479 222c 2022 4675 6e63 7469 6f6e 222c  ty", "Function",
+00004120: 2022 4661 6d69 6c79 222c 2022 556e 6970   "Family", "Unip
+00004130: 726f 7422 2c20 2252 6573 6964 7565 7322  rot", "Residues"
+00004140: 5d29 2023 544f 444f 3a20 6d6f 7665 2074  ]) #TODO: move t
+00004150: 6f20 6461 7461 0a20 2020 2066 616d 696c  o data.    famil
+00004160: 7920 3d20 6466 5b64 665b 2746 616d 696c  y = df[df['Famil
+00004170: 7927 5d2e 7374 722e 636f 6e74 6169 6e73  y'].str.contains
+00004180: 2866 616d 696c 795f 6e61 6d65 295d 0a20  (family_name)]. 
+00004190: 2020 2069 6620 6669 7273 745f 6d61 7463     if first_matc
+000041a0: 683a 2023 3131 3520 236f 6e6c 7920 7461  h: #115 #only ta
+000041b0: 6b65 7320 7468 6520 6669 7273 7420 7064  kes the first pd
+000041c0: 6220 6669 6c65 2066 6f75 6e64 2066 6f72  b file found for
+000041d0: 2065 6163 6820 7365 7175 656e 6365 0a20   each sequence. 
+000041e0: 2020 2020 2020 2066 6972 7374 5f75 6e69         first_uni
+000041f0: 7175 655f 6d61 7463 6820 3d20 6661 6d69  que_match = fami
+00004200: 6c79 2e67 726f 7570 6279 2827 556e 6970  ly.groupby('Unip
+00004210: 726f 7427 292e 6865 6164 2831 292e 7265  rot').head(1).re
+00004220: 7365 745f 696e 6465 7828 6472 6f70 3d54  set_index(drop=T
+00004230: 7275 6529 0a20 2020 2020 2020 2066 6972  rue).        fir
+00004240: 7374 5f75 6e69 7175 655f 6d61 7463 685b  st_unique_match[
+00004250: 2250 4442 5f31 225d 203d 2066 6972 7374  "PDB_1"] = first
+00004260: 5f75 6e69 7175 655f 6d61 7463 685b 2750  _unique_match['P
+00004270: 4442 5f31 275d 2e73 7472 2e72 6570 6c61  DB_1'].str.repla
+00004280: 6365 2872 273b 272c 2027 2729 0a20 2020  ce(r';', '').   
+00004290: 2020 2020 2070 6462 5f6c 6973 7420 3d20       pdb_list = 
+000042a0: 6669 7273 745f 756e 6971 7565 5f6d 6174  first_unique_mat
+000042b0: 6368 5b22 5044 425f 3122 5d2e 746f 6c69  ch["PDB_1"].toli
+000042c0: 7374 2829 0a20 2020 2020 2020 2072 6573  st().        res
+000042d0: 203d 205b 2252 6573 6964 7565 7322 5d20   = ["Residues"] 
+000042e0: 2a20 6c65 6e28 7064 625f 6c69 7374 290a  * len(pdb_list).
+000042f0: 2020 2020 2020 2020 7265 7369 6475 6573          residues
+00004300: 5f6c 6973 7420 3d20 6669 7273 745f 756e  _list = first_un
+00004310: 6971 7565 5f6d 6174 6368 5b22 5265 7369  ique_match["Resi
+00004320: 6475 6573 225d 2e73 7472 2e72 6570 6c61  dues"].str.repla
+00004330: 6365 2872 273b 272c 2027 2729 2e74 6f6c  ce(r';', '').tol
+00004340: 6973 7428 290a 2020 2020 2020 2020 6368  ist().        ch
+00004350: 6169 6e20 3d20 5b22 4368 6169 6e22 5d20  ain = ["Chain"] 
+00004360: 2a20 6c65 6e28 7064 625f 6c69 7374 290a  * len(pdb_list).
+00004370: 2020 2020 2020 2020 6368 6169 6e73 5f6c          chains_l
+00004380: 6973 7420 3d20 6669 7273 745f 756e 6971  ist = first_uniq
+00004390: 7565 5f6d 6174 6368 5b22 4368 6169 6e22  ue_match["Chain"
+000043a0: 5d2e 7374 722e 7265 706c 6163 6528 7227  ].str.replace(r'
+000043b0: 3b27 2c20 2727 292e 746f 6c69 7374 2829  ;', '').tolist()
+000043c0: 0a20 2020 2020 2020 2070 6661 6d5f 6469  .        pfam_di
+000043d0: 6374 203d 207b 613a 207b 623a 2063 2c20  ct = {a: {b: c, 
+000043e0: 643a 2065 7d20 666f 7220 612c 2062 2c20  d: e} for a, b, 
+000043f0: 632c 2064 2c20 6520 696e 207a 6970 2870  c, d, e in zip(p
+00004400: 6462 5f6c 6973 742c 2072 6573 2c20 7265  db_list, res, re
+00004410: 7369 6475 6573 5f6c 6973 742c 2063 6861  sidues_list, cha
+00004420: 696e 2c20 6368 6169 6e73 5f6c 6973 7429  in, chains_list)
+00004430: 7d0a 2020 2020 656c 7365 3a20 2336 3431  }.    else: #641
+00004440: 2023 636f 6e74 6169 6e73 2061 6c6c 2070   #contains all p
+00004450: 6462 2066 696c 6573 2028 6475 706c 6963  db files (duplic
+00004460: 6174 6573 2920 666f 7220 6561 6368 2073  ates) for each s
+00004470: 6571 7565 6e63 650a 2020 2020 2020 2020  equence.        
+00004480: 6661 6d69 6c79 5f67 726f 7570 7320 3d20  family_groups = 
+00004490: 6661 6d69 6c79 2e67 726f 7570 6279 2827  family.groupby('
+000044a0: 556e 6970 726f 7427 295b 5b22 5044 425f  Uniprot')[["PDB_
+000044b0: 3122 2c20 2243 6861 696e 222c 2022 4675  1", "Chain", "Fu
+000044c0: 6e63 7469 6f6e 222c 2022 4661 6d69 6c79  nction", "Family
+000044d0: 222c 2022 5265 7369 6475 6573 225d 5d2e  ", "Residues"]].
+000044e0: 6167 6728 6c61 6d62 6461 2078 3a20 6c69  agg(lambda x: li
+000044f0: 7374 2878 2929 0a20 2020 2020 2020 2070  st(x)).        p
+00004500: 6462 5f6c 6973 7420 3d73 756d 2866 616d  db_list =sum(fam
+00004510: 696c 795f 6772 6f75 7073 5b22 5044 425f  ily_groups["PDB_
+00004520: 3122 5d2e 746f 6c69 7374 2829 2c5b 5d29  1"].tolist(),[])
+00004530: 2023 6d61 7028 7374 722c 2076 616c 7565   #map(str, value
+00004540: 5f6c 6973 7429 2023 2e73 7472 2e72 6570  _list) #.str.rep
+00004550: 6c61 6365 2872 273b 272c 2027 2729 0a20  lace(r';', ''). 
+00004560: 2020 2020 2020 2070 6462 5f6c 6973 7420         pdb_list 
+00004570: 3d20 5b70 6462 2e72 6570 6c61 6365 2872  = [pdb.replace(r
+00004580: 273b 272c 2027 2729 2066 6f72 2070 6462  ';', '') for pdb
+00004590: 2069 6e20 7064 625f 6c69 7374 5d0a 2020   in pdb_list].  
+000045a0: 2020 2020 2020 7265 7320 3d20 5b22 5265        res = ["Re
+000045b0: 7369 6475 6573 225d 2a6c 656e 2870 6462  sidues"]*len(pdb
+000045c0: 5f6c 6973 7429 0a20 2020 2020 2020 2072  _list).        r
+000045d0: 6573 6964 7565 735f 6c69 7374 3d73 756d  esidues_list=sum
+000045e0: 2866 616d 696c 795f 6772 6f75 7073 5b22  (family_groups["
+000045f0: 5265 7369 6475 6573 225d 2e74 6f6c 6973  Residues"].tolis
+00004600: 7428 292c 5b5d 290a 2020 2020 2020 2020  t(),[]).        
+00004610: 7265 7369 6475 6573 5f6c 6973 7420 3d20  residues_list = 
+00004620: 5b72 6573 2e72 6570 6c61 6365 2872 273b  [res.replace(r';
+00004630: 272c 2027 2729 2066 6f72 2072 6573 2069  ', '') for res i
+00004640: 6e20 7265 7369 6475 6573 5f6c 6973 745d  n residues_list]
+00004650: 0a20 2020 2020 2020 2063 6861 696e 203d  .        chain =
+00004660: 205b 2243 6861 696e 225d 202a 206c 656e   ["Chain"] * len
+00004670: 2870 6462 5f6c 6973 7429 0a20 2020 2020  (pdb_list).     
+00004680: 2020 2063 6861 696e 735f 6c69 7374 203d     chains_list =
+00004690: 2073 756d 2866 616d 696c 795f 6772 6f75   sum(family_grou
+000046a0: 7073 5b22 4368 6169 6e22 5d2e 746f 6c69  ps["Chain"].toli
+000046b0: 7374 2829 2c5b 5d29 0a20 2020 2020 2020  st(),[]).       
+000046c0: 2063 6861 696e 735f 6c69 7374 203d 205b   chains_list = [
+000046d0: 6368 6169 6e5b 2d32 5d20 666f 7220 6368  chain[-2] for ch
+000046e0: 6169 6e20 696e 2063 6861 696e 735f 6c69  ain in chains_li
+000046f0: 7374 5d0a 2020 2020 2020 2020 7066 616d  st].        pfam
+00004700: 5f64 6963 7420 3d20 7b61 3a7b 623a 632c  _dict = {a:{b:c,
+00004710: 643a 657d 2066 6f72 2061 2c20 622c 2063  d:e} for a, b, c
+00004720: 2c64 2c65 2069 6e20 7a69 7028 7064 625f  ,d,e in zip(pdb_
+00004730: 6c69 7374 2c20 7265 732c 7265 7369 6475  list, res,residu
+00004740: 6573 5f6c 6973 742c 6368 6169 6e2c 2063  es_list,chain, c
+00004750: 6861 696e 735f 6c69 7374 297d 0a20 2020  hains_list)}.   
+00004760: 2072 6574 7572 6e20 7066 616d 5f64 6963   return pfam_dic
+00004770: 742c 2070 6462 5f6c 6973 740a 6465 6620  t, pdb_list.def 
+00004780: 7472 6565 5f70 6169 725f 666f 725f 726f  tree_pair_for_ro
+00004790: 6f74 696e 6728 6469 7374 616e 6365 5f6d  oting(distance_m
+000047a0: 6174 7269 7829 3a0a 2020 2020 2222 2246  atrix):.    """F
+000047b0: 696e 6473 2074 6865 2070 6169 7220 6f66  inds the pair of
+000047c0: 2073 6571 7565 6e63 6573 2077 6974 6820   sequences with 
+000047d0: 7468 6520 6c61 7267 6573 7420 7061 6972  the largest pair
+000047e0: 7769 7365 2064 6973 7461 6e63 6520 746f  wise distance to
+000047f0: 2073 6574 2061 7320 616e 206f 7574 6772   set as an outgr
+00004800: 6f75 7020 746f 2066 6f72 6d20 7468 6520  oup to form the 
+00004810: 726f 6f74 206f 6620 616e 2075 6e72 6f6f  root of an unroo
+00004820: 7465 6420 7472 6565 0a20 2020 203a 7061  ted tree.    :pa
+00004830: 7261 6d20 7061 6e64 6173 2064 6174 6166  ram pandas dataf
+00004840: 7261 6d65 2064 6973 7461 6e63 655f 6d61  rame distance_ma
+00004850: 7472 6978 3a20 636f 6e74 6169 6e73 2070  trix: contains p
+00004860: 6169 7277 6973 6520 6469 7374 616e 6365  airwise distance
+00004870: 7320 6163 726f 7373 2074 6865 2073 6571  s across the seq
+00004880: 7565 6e63 6573 2069 6e20 7468 6520 7472  uences in the tr
+00004890: 6565 2222 220a 0a20 2020 2073 6f72 7465  ee"""..    sorte
+000048a0: 645f 6469 7374 616e 6365 5f6d 6174 7269  d_distance_matri
+000048b0: 7820 3d20 6469 7374 616e 6365 5f6d 6174  x = distance_mat
+000048c0: 7269 785b 6469 7374 616e 6365 5f6d 6174  rix[distance_mat
+000048d0: 7269 782e 6774 2830 295d 2e73 7461 636b  rix.gt(0)].stack
+000048e0: 2829 2e73 6f72 745f 7661 6c75 6573 2829  ().sort_values()
+000048f0: 2e74 6f5f 6672 616d 6528 290a 2020 2020  .to_frame().    
+00004900: 736f 7274 6564 5f64 6973 7461 6e63 655f  sorted_distance_
+00004910: 6d61 7472 6978 2e72 6573 6574 5f69 6e64  matrix.reset_ind
+00004920: 6578 286c 6576 656c 3d30 2c20 696e 706c  ex(level=0, inpl
+00004930: 6163 653d 5472 7565 290a 2020 2020 736f  ace=True).    so
+00004940: 7274 6564 5f64 6973 7461 6e63 655f 6d61  rted_distance_ma
+00004950: 7472 6978 2e63 6f6c 756d 6e73 203d 205b  trix.columns = [
+00004960: 2253 6571 7565 6e63 655f 3022 2c20 2244  "Sequence_0", "D
+00004970: 6973 7461 6e63 6522 5d0a 2020 2020 736f  istance"].    so
+00004980: 7274 6564 5f64 6973 7461 6e63 655f 6d61  rted_distance_ma
+00004990: 7472 6978 5b27 5365 7175 656e 6365 5f31  trix['Sequence_1
+000049a0: 275d 203d 2073 6f72 7465 645f 6469 7374  '] = sorted_dist
+000049b0: 616e 6365 5f6d 6174 7269 782e 696e 6465  ance_matrix.inde
+000049c0: 780a 2020 2020 736f 7274 6564 5f64 6973  x.    sorted_dis
+000049d0: 7461 6e63 655f 6d61 7472 6978 203d 2073  tance_matrix = s
+000049e0: 6f72 7465 645f 6469 7374 616e 6365 5f6d  orted_distance_m
+000049f0: 6174 7269 785b 5b22 5365 7175 656e 6365  atrix[["Sequence
+00004a00: 5f31 222c 2022 5365 7175 656e 6365 5f30  _1", "Sequence_0
+00004a10: 222c 2022 4469 7374 616e 6365 225d 5d0a  ", "Distance"]].
+00004a20: 2020 2020 736f 7274 6564 5f64 6973 7461      sorted_dista
+00004a30: 6e63 655f 6d61 7472 6978 2e72 6573 6574  nce_matrix.reset
+00004a40: 5f69 6e64 6578 2869 6e70 6c61 6365 3d54  _index(inplace=T
+00004a50: 7275 6529 0a20 2020 2073 6f72 7465 645f  rue).    sorted_
+00004a60: 6469 7374 616e 6365 5f6d 6174 7269 782e  distance_matrix.
+00004a70: 6472 6f70 285b 2269 6e64 6578 225d 2c20  drop(["index"], 
+00004a80: 696e 706c 6163 653d 5472 7565 2c20 6178  inplace=True, ax
+00004a90: 6973 3d31 290a 2020 2020 2320 736f 7274  is=1).    # sort
+00004aa0: 6564 5f64 6973 7461 6e63 655f 6d61 7472  ed_distance_matr
+00004ab0: 6978 2e64 726f 705f 6475 706c 6963 6174  ix.drop_duplicat
+00004ac0: 6573 2873 7562 7365 743d 5b27 4469 7374  es(subset=['Dist
+00004ad0: 616e 6365 275d 290a 2020 2020 736f 7274  ance']).    sort
+00004ae0: 6564 5f64 6973 7461 6e63 655f 6d61 7472  ed_distance_matr
+00004af0: 6978 203d 2073 6f72 7465 645f 6469 7374  ix = sorted_dist
+00004b00: 616e 6365 5f6d 6174 7269 785b 7e73 6f72  ance_matrix[~sor
+00004b10: 7465 645f 6469 7374 616e 6365 5f6d 6174  ted_distance_mat
+00004b20: 7269 785b 5b27 5365 7175 656e 6365 5f31  rix[['Sequence_1
+00004b30: 272c 2027 5365 7175 656e 6365 5f30 275d  ', 'Sequence_0']
+00004b40: 5d2e 6170 706c 7928 6672 6f7a 656e 7365  ].apply(frozense
+00004b50: 742c 6178 6973 3d31 292e 6475 706c 6963  t,axis=1).duplic
+00004b60: 6174 6564 2829 5d20 2023 2052 656d 6f76  ated()]  # Remov
+00004b70: 6520 7265 7065 6174 6564 2063 6f6d 6269  e repeated combi
+00004b80: 6e61 7469 6f6e 7320 6f66 2073 6571 7565  nations of seque
+00004b90: 6e63 6573 0a20 2020 2073 6f72 7465 645f  nces.    sorted_
+00004ba0: 6469 7374 616e 6365 5f6d 6174 7269 7820  distance_matrix 
+00004bb0: 3d20 736f 7274 6564 5f64 6973 7461 6e63  = sorted_distanc
+00004bc0: 655f 6d61 7472 6978 2e72 6573 6574 5f69  e_matrix.reset_i
+00004bd0: 6e64 6578 2864 726f 703d 5472 7565 290a  ndex(drop=True).
+00004be0: 2020 2020 7261 6e6b 203d 2073 6f72 7465      rank = sorte
+00004bf0: 645f 6469 7374 616e 6365 5f6d 6174 7269  d_distance_matri
+00004c00: 782e 7368 6170 655b 305d 202d 310a 2020  x.shape[0] -1.  
+00004c10: 2020 7365 7175 656e 6365 5f30 203d 2073    sequence_0 = s
+00004c20: 6f72 7465 645f 6469 7374 616e 6365 5f6d  orted_distance_m
+00004c30: 6174 7269 782e 6c6f 635b 7261 6e6b 5d2e  atrix.loc[rank].
+00004c40: 5365 7175 656e 6365 5f30 0a20 2020 2073  Sequence_0.    s
+00004c50: 6571 7565 6e63 655f 3120 3d20 736f 7274  equence_1 = sort
+00004c60: 6564 5f64 6973 7461 6e63 655f 6d61 7472  ed_distance_matr
+00004c70: 6978 2e6c 6f63 5b72 616e 6b5d 2e53 6571  ix.loc[rank].Seq
+00004c80: 7565 6e63 655f 310a 2020 2020 7265 7475  uence_1.    retu
+00004c90: 726e 2073 6571 7565 6e63 655f 302c 7365  rn sequence_0,se
+00004ca0: 7175 656e 6365 5f31 0a64 6566 2064 6f77  quence_1.def dow
+00004cb0: 6e6c 6f61 645f 5044 4228 6669 6c65 735f  nload_PDB(files_
+00004cc0: 6c69 7374 2c20 6469 7265 6374 6f72 7929  list, directory)
+00004cd0: 3a0a 2020 2020 2222 2252 6561 6473 2061  :.    """Reads a
+00004ce0: 206c 6973 7420 6f66 2050 4442 5f66 696c   list of PDB_fil
+00004cf0: 6573 2061 6e64 2064 6f77 6e6c 6f61 6473  es and downloads
+00004d00: 2074 6865 6d20 746f 2061 2066 6f6c 6465   them to a folde
+00004d10: 720a 2020 2020 3a70 6172 616d 206c 6973  r.    :param lis
+00004d20: 7420 6669 6c65 735f 6c69 7374 3a20 6c69  t files_list: li
+00004d30: 7374 206f 6620 5044 4220 6669 6c65 7320  st of PDB files 
+00004d40: 6e61 6d65 7320 746f 2064 6f77 6e6c 6f61  names to downloa
+00004d50: 640a 2020 2020 3a70 6172 616d 2073 7472  d.    :param str
+00004d60: 2064 6972 6563 746f 7279 3a20 666f 6c64   directory: fold
+00004d70: 6572 206f 7220 6469 7265 6374 6f72 7920  er or directory 
+00004d80: 7768 6572 6520 746f 2064 6f77 6e6c 6f61  where to downloa
+00004d90: 6420 7468 6520 6669 6c65 7322 2222 0a20  d the files""". 
+00004da0: 2020 2070 6462 6c20 3d20 5044 424c 6973     pdbl = PDBLis
+00004db0: 7428 290a 2020 2020 666f 7220 692c 2066  t().    for i, f
+00004dc0: 696c 6520 696e 2065 6e75 6d65 7261 7465  ile in enumerate
+00004dd0: 2866 696c 6573 5f6c 6973 7429 3a0a 2020  (files_list):.  
+00004de0: 2020 2020 2020 7064 626c 2e72 6574 7269        pdbl.retri
+00004df0: 6576 655f 7064 625f 6669 6c65 2866 696c  eve_pdb_file(fil
+00004e00: 652c 2070 6469 723d 6469 7265 6374 6f72  e, pdir=director
+00004e10: 792c 206f 6273 6f6c 6574 653d 4661 6c73  y, obsolete=Fals
+00004e20: 652c 2066 696c 655f 666f 726d 6174 3d22  e, file_format="
+00004e30: 7064 6222 290a 6465 6620 636f 6e76 6572  pdb").def conver
+00004e40: 745f 746f 5f70 616e 6461 7328 6469 7374  t_to_pandas(dist
+00004e50: 616e 6365 5f6d 6174 7269 7829 3a0a 2020  ance_matrix):.  
+00004e60: 2020 2222 2243 6f6e 7665 7274 7320 6120    """Converts a 
+00004e70: 6269 6f70 7974 686f 6e20 7061 6972 7769  biopython pairwi
+00004e80: 7365 2064 6973 7461 6e63 6520 6d61 7472  se distance matr
+00004e90: 6978 2069 6e74 6f20 6120 7379 6d6d 6574  ix into a symmet
+00004ea0: 7269 6320 7061 6e64 6173 2064 6174 6166  ric pandas dataf
+00004eb0: 7261 6d65 0a20 2020 203a 7061 7261 6d20  rame.    :param 
+00004ec0: 6269 6f70 7974 686f 6e20 6361 6c63 756c  biopython calcul
+00004ed0: 6174 6f72 206d 6174 7269 783a 2070 6169  ator matrix: pai
+00004ee0: 7277 6973 6520 6469 7374 616e 6365 206d  rwise distance m
+00004ef0: 6174 7269 780a 2020 2020 2222 220a 2020  atrix.    """.  
+00004f00: 2020 6220 3d20 6e70 2e7a 6572 6f73 285b    b = np.zeros([
+00004f10: 6c65 6e28 6469 7374 616e 6365 5f6d 6174  len(distance_mat
+00004f20: 7269 782e 6d61 7472 6978 292c 206c 656e  rix.matrix), len
+00004f30: 286d 6178 2864 6973 7461 6e63 655f 6d61  (max(distance_ma
+00004f40: 7472 6978 2e6d 6174 7269 782c 206b 6579  trix.matrix, key
+00004f50: 3d6c 616d 6264 6120 783a 206c 656e 2878  =lambda x: len(x
+00004f60: 2929 295d 290a 2020 2020 666f 7220 692c  )))]).    for i,
+00004f70: 206a 2069 6e20 656e 756d 6572 6174 6528   j in enumerate(
+00004f80: 6469 7374 616e 6365 5f6d 6174 7269 782e  distance_matrix.
+00004f90: 6d61 7472 6978 293a 0a20 2020 2020 2020  matrix):.       
+00004fa0: 2062 5b69 5d5b 303a 6c65 6e28 6a29 5d20   b[i][0:len(j)] 
+00004fb0: 3d20 6a0a 2020 2020 2364 6620 3d20 7064  = j.    #df = pd
+00004fc0: 2e44 6174 6146 7261 6d65 2862 2c20 696e  .DataFrame(b, in
+00004fd0: 6465 783d 4469 7374 616e 6365 4d61 7472  dex=DistanceMatr
+00004fe0: 6978 2e6e 616d 6573 2c20 636f 6c75 6d6e  ix.names, column
+00004ff0: 733d 4469 7374 616e 6365 4d61 7472 6978  s=DistanceMatrix
+00005000: 2e6e 616d 6573 290a 2020 2020 625f 7472  .names).    b_tr
+00005010: 616e 7370 6f73 6520 3d20 6e70 2e74 7261  anspose = np.tra
+00005020: 6e73 706f 7365 2862 290a 2020 2020 6220  nspose(b).    b 
+00005030: 3d20 6220 2b20 625f 7472 616e 7370 6f73  = b + b_transpos
+00005040: 650a 2020 2020 6466 203d 2070 642e 4461  e.    df = pd.Da
+00005050: 7461 4672 616d 6528 622c 2069 6e64 6578  taFrame(b, index
+00005060: 3d64 6973 7461 6e63 655f 6d61 7472 6978  =distance_matrix
+00005070: 2e6e 616d 6573 2c20 636f 6c75 6d6e 733d  .names, columns=
+00005080: 6469 7374 616e 6365 5f6d 6174 7269 782e  distance_matrix.
+00005090: 6e61 6d65 7329 0a20 2020 2072 6574 7572  names).    retur
+000050a0: 6e20 2064 660a 6465 6620 696e 6665 725f  n  df.def infer_
+000050b0: 7472 6565 2861 6c69 676e 6d65 6e74 2c20  tree(alignment, 
+000050c0: 616c 6967 6e6d 656e 745f 6669 6c65 5f6e  alignment_file_n
+000050d0: 616d 652c 6e61 6d65 2c6d 6574 686f 643d  ame,name,method=
+000050e0: 4e6f 6e65 2c74 7265 655f 6669 6c65 5f6e  None,tree_file_n
+000050f0: 616d 653d 4e6f 6e65 2c74 7265 655f 6669  ame=None,tree_fi
+00005100: 6c65 3d4e 6f6e 652c 7374 6f72 6167 655f  le=None,storage_
+00005110: 666f 6c64 6572 3d22 2229 3a0a 2020 2020  folder=""):.    
+00005120: 2222 2220 5065 7266 6f72 6d73 2074 7265  """ Performs tre
+00005130: 6520 696e 6665 7265 6e63 6520 6f72 2072  e inference or r
+00005140: 6561 6473 2061 6e20 696e 7075 7420 6769  eads an input gi
+00005150: 7665 6e20 7472 6565 2c20 7265 7475 726e  ven tree, return
+00005160: 7320 616e 2065 7465 3320 7472 6565 2066  s an ete3 tree f
+00005170: 6f72 6d61 7465 6420 7472 6565 0a20 2020  ormated tree.   
+00005180: 203a 7061 7261 6d20 6269 6f70 7974 686f   :param biopytho
+00005190: 6e20 616c 6967 6e6d 656e 7420 616c 6967  n alignment alig
+000051a0: 6e6d 656e 743a 2062 696f 7079 7468 6f6e  nment: biopython
+000051b0: 2061 6c69 676e 6d65 6e74 2063 6c61 7373   alignment class
+000051c0: 0a20 2020 203a 7061 7261 6d20 7374 7220  .    :param str 
+000051d0: 616c 6967 6e6d 656e 745f 6669 6c65 5f6e  alignment_file_n
+000051e0: 616d 653a 2070 6174 6820 746f 2061 6c69  ame: path to ali
+000051f0: 676e 6d65 6e74 2066 696c 650a 2020 2020  gnment file.    
+00005200: 3a70 6172 616d 2073 7472 206e 616d 653a  :param str name:
+00005210: 2064 6174 6173 6574 2070 726f 6a65 6374   dataset project
+00005220: 206e 616d 650a 2020 2020 3a70 6172 616d   name.    :param
+00005230: 2073 7472 206d 6574 686f 643a 2074 7265   str method: tre
+00005240: 6520 696e 6665 7265 6e63 6520 6d65 7468  e inference meth
+00005250: 6f64 2028 6966 2074 7265 6520 6973 206e  od (if tree is n
+00005260: 6f74 2067 6976 656e 290a 2020 2020 3a70  ot given).    :p
+00005270: 6172 616d 2073 7472 2074 7265 655f 6669  aram str tree_fi
+00005280: 6c65 5f6e 616d 653a 206e 616d 6520 746f  le_name: name to
+00005290: 2067 6976 6520 746f 2074 6865 2074 7265   give to the tre
+000052a0: 6520 6669 6c65 0a20 2020 202e 7061 7261  e file.    .para
+000052b0: 6d20 7374 7220 7472 6565 5f66 696c 653a  m str tree_file:
+000052c0: 7061 7468 2074 6f20 6120 706f 7373 6962  path to a possib
+000052d0: 6c65 2067 6976 656e 2074 7265 6520 696e  le given tree in
+000052e0: 206e 6577 6963 6b20 666f 726d 6174 2031   newick format 1
+000052f0: 0a20 2020 203a 7061 7261 6d20 7374 7220  .    :param str 
+00005300: 7374 6f72 6167 655f 666f 6c64 6572 3a20  storage_folder: 
+00005310: 666f 6c64 6572 2077 6865 7265 2074 6f20  folder where to 
+00005320: 7374 6f72 6520 7468 6520 7265 7375 6c74  store the result
+00005330: 7320 6f66 2074 6865 2074 7265 6520 696e  s of the tree in
+00005340: 6665 7265 6e63 650a 0a20 2020 2022 2222  ference..    """
+00005350: 0a20 2020 2069 6620 7472 6565 5f66 696c  .    if tree_fil
+00005360: 653a 0a20 2020 2020 2020 2070 7269 6e74  e:.        print
+00005370: 2822 5573 696e 6720 6769 7665 6e20 7472  ("Using given tr
+00005380: 6565 2066 696c 652e 2e2e 2229 0a20 2020  ee file...").   
+00005390: 2020 2020 2074 7265 6520 3d20 5472 6565       tree = Tree
+000053a0: 4574 6533 2874 7265 655f 6669 6c65 2c66  Ete3(tree_file,f
+000053b0: 6f72 6d61 743d 312c 7175 6f74 6564 5f6e  ormat=1,quoted_n
+000053c0: 6f64 655f 6e61 6d65 733d 5472 7565 290a  ode_names=True).
+000053d0: 2020 2020 2020 2020 7265 7475 726e 2074          return t
+000053e0: 7265 650a 0a20 2020 2065 6c73 653a 0a20  ree..    else:. 
+000053f0: 2020 2020 2020 2023 2050 6169 7277 6973         # Pairwis
+00005400: 6520 6469 7374 616e 6365 206d 6174 7269  e distance matri
+00005410: 780a 2020 2020 2020 2020 7072 696e 7428  x.        print(
+00005420: 2242 7569 6c64 696e 6720 6469 7374 616e  "Building distan
+00005430: 6365 206d 6174 7269 6365 7320 616e 6420  ce matrices and 
+00005440: 7b7d 2074 7265 652e 2e2e 222e 666f 726d  {} tree...".form
+00005450: 6174 286d 6574 686f 6429 290a 2020 2020  at(method)).    
+00005460: 2020 2020 6966 206c 656e 2861 6c69 676e      if len(align
+00005470: 6d65 6e74 2920 3c20 3230 3020 616e 6420  ment) < 200 and 
+00005480: 6d65 7468 6f64 2069 6e20 5b22 6e6a 222c  method in ["nj",
+00005490: 226e 6a5f 726f 6f74 6564 222c 2275 7067  "nj_rooted","upg
+000054a0: 6d61 225d 3a0a 2020 2020 2020 2020 2020  ma"]:.          
+000054b0: 2020 6361 6c63 756c 6174 6f72 203d 2044    calculator = D
+000054c0: 6973 7461 6e63 6543 616c 6375 6c61 746f  istanceCalculato
+000054d0: 7228 2762 6c6f 7375 6d36 3227 2920 2023  r('blosum62')  #
+000054e0: 2044 4e41 202d 2d2d 3e20 4964 656e 7469   DNA ---> Identi
+000054f0: 7479 2f2f 2050 726f 7465 696e 202d 2d2d  ty// Protein ---
+00005500: 3e20 626c 6f73 756d 3632 0a20 2020 2020  > blosum62.     
+00005510: 2020 2020 2020 2064 6973 7461 6e63 655f         distance_
+00005520: 6d61 7472 6978 5f63 616c 203d 2063 616c  matrix_cal = cal
+00005530: 6375 6c61 746f 722e 6765 745f 6469 7374  culator.get_dist
+00005540: 616e 6365 2861 6c69 676e 6d65 6e74 290a  ance(alignment).
+00005550: 2020 2020 2020 2020 2020 2020 6469 7374              dist
+00005560: 616e 6365 5f6d 6174 7269 785f 6361 6c5f  ance_matrix_cal_
+00005570: 7061 6e64 6173 203d 2063 6f6e 7665 7274  pandas = convert
+00005580: 5f74 6f5f 7061 6e64 6173 2864 6973 7461  _to_pandas(dista
+00005590: 6e63 655f 6d61 7472 6978 5f63 616c 290a  nce_matrix_cal).
+000055a0: 2020 2020 2020 2020 2020 2020 6469 7374              dist
+000055b0: 616e 6365 5f6d 6174 7269 785f 6361 6c5f  ance_matrix_cal_
+000055c0: 7061 6e64 6173 2e74 6f5f 6373 7628 227b  pandas.to_csv("{
+000055d0: 7d2f 7b7d 5f64 6973 7461 6e63 655f 6d61  }/{}_distance_ma
+000055e0: 7472 6978 2e63 7376 222e 666f 726d 6174  trix.csv".format
+000055f0: 2873 746f 7261 6765 5f66 6f6c 6465 722c  (storage_folder,
+00005600: 6e61 6d65 2929 0a20 2020 2020 2020 2020  name)).         
+00005610: 2020 2023 6874 7470 733a 2f2f 7374 6163     #https://stac
+00005620: 6b6f 7665 7266 6c6f 772e 636f 6d2f 7175  koverflow.com/qu
+00005630: 6573 7469 6f6e 732f 3330 3234 3733 3539  estions/30247359
+00005640: 2f68 6f77 2d64 6f65 732d 6269 6f70 7974  /how-does-biopyt
+00005650: 686f 6e2d 6465 7465 726d 696e 652d 7468  hon-determine-th
+00005660: 652d 726f 6f74 2d6f 662d 612d 7068 796c  e-root-of-a-phyl
+00005670: 6f67 656e 6574 6963 2d74 7265 650a 2020  ogenetic-tree.  
+00005680: 2020 2020 2020 2020 2020 6966 206d 6574            if met
+00005690: 686f 6420 3d3d 2022 6e6a 223a 0a20 2020  hod == "nj":.   
+000056a0: 2020 2020 2020 2020 2020 2020 2070 7269               pri
+000056b0: 6e74 2822 5472 6565 2069 6e66 6572 656e  nt("Tree inferen
+000056c0: 6365 2076 6961 204e 6569 6768 626f 7572  ce via Neighbour
+000056d0: 204a 6f69 6e69 6e67 204e 4f54 2072 6f6f   Joining NOT roo
+000056e0: 7465 6420 6d65 7468 6f64 2e2e 2e22 290a  ted method...").
+000056f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00005700: 636f 6e73 7472 7563 746f 7220 3d20 4469  constructor = Di
+00005710: 7374 616e 6365 5472 6565 436f 6e73 7472  stanceTreeConstr
+00005720: 7563 746f 7228 6d65 7468 6f64 3d22 6e6a  uctor(method="nj
+00005730: 2229 0a20 2020 2020 2020 2020 2020 2020  ").             
+00005740: 2020 2074 7265 6520 3d20 636f 6e73 7472     tree = constr
+00005750: 7563 746f 722e 6e6a 2864 6973 7461 6e63  uctor.nj(distanc
+00005760: 655f 6d61 7472 6978 5f63 616c 290a 2020  e_matrix_cal).  
+00005770: 2020 2020 2020 2020 2020 2020 2020 7472                tr
+00005780: 6565 203d 2074 6f5f 6574 6533 2874 7265  ee = to_ete3(tre
+00005790: 6529 0a20 2020 2020 2020 2020 2020 2020  e).             
+000057a0: 2020 2072 6574 7572 6e20 7472 6565 0a20     return tree. 
+000057b0: 2020 2020 2020 2020 2020 2065 6c69 6620             elif 
+000057c0: 6d65 7468 6f64 203d 3d20 226e 6a5f 726f  method == "nj_ro
+000057d0: 6f74 6564 223a 0a20 2020 2020 2020 2020  oted":.         
+000057e0: 2020 2020 2020 2070 7269 6e74 2822 5472         print("Tr
+000057f0: 6565 2069 6e66 6572 656e 6365 2076 6961  ee inference via
+00005800: 204e 6569 6768 626f 7572 204a 6f69 6e69   Neighbour Joini
+00005810: 6e67 2077 6974 6820 6164 6469 7469 6f6e  ng with addition
+00005820: 616c 2072 6f6f 7469 6e67 206d 6574 686f  al rooting metho
+00005830: 642e 2e2e 2229 0a20 2020 2020 2020 2020  d...").         
+00005840: 2020 2020 2020 2063 6f6e 7374 7275 6374         construct
+00005850: 6f72 203d 2044 6973 7461 6e63 6554 7265  or = DistanceTre
+00005860: 6543 6f6e 7374 7275 6374 6f72 286d 6574  eConstructor(met
+00005870: 686f 643d 226e 6a22 290a 2020 2020 2020  hod="nj").      
+00005880: 2020 2020 2020 2020 2020 7472 6565 203d            tree =
+00005890: 2063 6f6e 7374 7275 6374 6f72 2e6e 6a28   constructor.nj(
+000058a0: 6469 7374 616e 6365 5f6d 6174 7269 785f  distance_matrix_
+000058b0: 6361 6c29 0a20 2020 2020 2020 2020 2020  cal).           
+000058c0: 2020 2020 2074 7265 6520 3d20 746f 5f65       tree = to_e
+000058d0: 7465 3328 7472 6565 290a 2020 2020 2020  te3(tree).      
+000058e0: 2020 2020 2020 2020 2020 2320 4d61 6b69            # Maki
+000058f0: 6e67 2061 2072 6f6f 743a 0a20 2020 2020  ng a root:.     
+00005900: 2020 2020 2020 2020 2020 2073 6571 7565             seque
+00005910: 6e63 655f 302c 2073 6571 7565 6e63 655f  nce_0, sequence_
+00005920: 3120 3d20 7472 6565 5f70 6169 725f 666f  1 = tree_pair_fo
+00005930: 725f 726f 6f74 696e 6728 6469 7374 616e  r_rooting(distan
+00005940: 6365 5f6d 6174 7269 785f 6361 6c5f 7061  ce_matrix_cal_pa
+00005950: 6e64 6173 290a 2020 2020 2020 2020 2020  ndas).          
+00005960: 2020 2020 2020 7472 6565 2e73 6574 5f6f        tree.set_o
+00005970: 7574 6772 6f75 7028 7472 6565 2026 2073  utgroup(tree & s
+00005980: 6571 7565 6e63 655f 3029 0a20 2020 2020  equence_0).     
+00005990: 2020 2020 2020 2020 2020 2023 2061 6e63             # anc
+000059a0: 6573 746f 7220 3d20 7472 6565 2e67 6574  estor = tree.get
+000059b0: 5f63 6f6d 6d6f 6e5f 616e 6365 7374 6f72  _common_ancestor
+000059c0: 2873 6571 7565 6e63 655f 302c 2073 6571  (sequence_0, seq
+000059d0: 7565 6e63 655f 3129 0a20 2020 2020 2020  uence_1).       
+000059e0: 2020 2020 2020 2020 2023 2074 7265 652e           # tree.
+000059f0: 7365 745f 6f75 7467 726f 7570 2861 6e63  set_outgroup(anc
+00005a00: 6573 746f 7229 0a20 2020 2020 2020 2020  estor).         
+00005a10: 2020 2020 2020 2072 6574 7572 6e20 7472         return tr
+00005a20: 6565 0a20 2020 2020 2020 2020 2020 2065  ee.            e
+00005a30: 6c69 6620 6d65 7468 6f64 203d 3d20 2275  lif method == "u
+00005a40: 7067 6d61 223a 0a20 2020 2020 2020 2020  pgma":.         
+00005a50: 2020 2020 2020 2070 7269 6e74 2822 5472         print("Tr
+00005a60: 6565 2069 6e66 6572 656e 6365 2076 6961  ee inference via
+00005a70: 2055 7067 6d61 2072 6f6f 7465 6420 6d65   Upgma rooted me
+00005a80: 7468 6f64 2e2e 2e22 290a 2020 2020 2020  thod...").      
+00005a90: 2020 2020 2020 2020 2020 636f 6e73 7472            constr
+00005aa0: 7563 746f 7220 3d20 4469 7374 616e 6365  uctor = Distance
+00005ab0: 5472 6565 436f 6e73 7472 7563 746f 7228  TreeConstructor(
+00005ac0: 6d65 7468 6f64 3d22 7570 676d 6122 2920  method="upgma") 
+00005ad0: 2320 6e6a 206d 6574 686f 6420 6973 2075  # nj method is u
+00005ae0: 6e72 6f6f 7465 6420 696e 2062 696f 7079  nrooted in biopy
+00005af0: 7468 6f6e 2e20 7570 676d 6120 6973 2072  thon. upgma is r
+00005b00: 6f6f 7465 640a 2020 2020 2020 2020 2020  ooted.          
+00005b10: 2020 2020 2020 7472 6565 203d 2063 6f6e        tree = con
+00005b20: 7374 7275 6374 6f72 2e75 7067 6d61 2864  structor.upgma(d
+00005b30: 6973 7461 6e63 655f 6d61 7472 6978 5f63  istance_matrix_c
+00005b40: 616c 290a 2020 2020 2020 2020 2020 2020  al).            
+00005b50: 2020 2020 7472 6565 203d 2074 6f5f 6574      tree = to_et
+00005b60: 6533 2874 7265 6529 0a20 2020 2020 2020  e3(tree).       
+00005b70: 2020 2020 2020 2020 2072 6574 7572 6e20           return 
+00005b80: 7472 6565 0a20 2020 2020 2020 2065 6c69  tree.        eli
+00005b90: 6620 6d65 7468 6f64 203d 3d20 2269 7174  f method == "iqt
+00005ba0: 7265 6522 3a0a 2020 2020 2020 2020 2020  ree":.          
+00005bb0: 2020 7072 696e 7428 2249 7174 7265 6520    print("Iqtree 
+00005bc0: 4d4c 206d 6574 686f 642e 2e2e 2229 0a20  ML method..."). 
+00005bd0: 2020 2020 2020 2020 2020 2061 6c69 676e             align
+00005be0: 6d65 6e74 5f66 203d 205b 616c 6967 6e6d  ment_f = [alignm
+00005bf0: 656e 745f 6669 6c65 5f6e 616d 6520 6966  ent_file_name if
+00005c00: 2061 6c69 676e 6d65 6e74 5f66 696c 655f   alignment_file_
+00005c10: 6e61 6d65 2065 6c73 6520 227b 7d2f 7b7d  name else "{}/{}
+00005c20: 2e6d 6166 6674 222e 666f 726d 6174 2873  .mafft".format(s
+00005c30: 746f 7261 6765 5f66 6f6c 6465 722c 6e61  torage_folder,na
+00005c40: 6d65 295d 5b30 5d0a 2020 2020 2020 2020  me)][0].        
+00005c50: 2020 2020 7472 6565 5f66 696c 655f 6e61      tree_file_na
+00005c60: 6d65 203d 2061 6c69 676e 6d65 6e74 5f66  me = alignment_f
+00005c70: 2e73 706c 6974 2822 2e22 295b 305d 202b  .split(".")[0] +
+00005c80: 2022 2e74 7265 6566 696c 6522 0a0a 2020   ".treefile"..  
+00005c90: 2020 2020 2020 2020 2020 6966 206e 6f74            if not
+00005ca0: 206f 732e 7061 7468 2e65 7869 7374 7328   os.path.exists(
+00005cb0: 7472 6565 5f66 696c 655f 6e61 6d65 293a  tree_file_name):
+00005cc0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00005cd0: 2023 2d6f 0953 7065 6369 6679 2061 6e20   #-o.Specify an 
+00005ce0: 6f75 7467 726f 7570 2074 6178 6f6e 206e  outgroup taxon n
+00005cf0: 616d 6520 746f 2072 6f6f 7420 7468 6520  ame to root the 
+00005d00: 7472 6565 2e20 5468 6520 6f75 7470 7574  tree. The output
+00005d10: 2074 7265 6520 696e 202e 7472 6565 6669   tree in .treefi
+00005d20: 6c65 2077 696c 6c20 6265 2072 6f6f 7465  le will be roote
+00005d30: 6420 6163 636f 7264 696e 676c 792e 2044  d accordingly. D
+00005d40: 4546 4155 4c54 3a20 6669 7273 7420 7461  EFAULT: first ta
+00005d50: 786f 6e20 696e 2061 6c69 676e 6d65 6e74  xon in alignment
+00005d60: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00005d70: 2074 6178 6f6e 5f72 6f6f 7420 3d20 4661   taxon_root = Fa
+00005d80: 6c73 650a 2020 2020 2020 2020 2020 2020  lse.            
+00005d90: 2020 2020 6966 2074 6178 6f6e 5f72 6f6f      if taxon_roo
+00005da0: 743a 0a20 2020 2020 2020 2020 2020 2020  t:.             
+00005db0: 2020 2020 2020 2072 6f6f 743d 3231 0a20         root=21. 
 00005dc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00005dd0: 2020 7375 6270 726f 6365 7373 2e72 756e    subprocess.run
-00005de0: 2861 7267 733d 5b22 6971 7472 6565 222c  (args=["iqtree",
-00005df0: 222d 7322 2c61 6c69 676e 6d65 6e74 5f66  "-s",alignment_f
-00005e00: 2e73 706c 6974 2822 2e22 295b 305d 2c22  .split(".")[0],"
-00005e10: 2d2d 616c 6e22 2c61 6c69 676e 6d65 6e74  --aln",alignment
-00005e20: 5f66 2c22 2d6e 7422 2c22 4155 544f 222c  _f,"-nt","AUTO",
-00005e30: 222d 6f22 2c72 6f6f 745d 2c73 7464 6572  "-o",root],stder
-00005e40: 723d 7379 732e 7374 6465 7272 2c20 7374  r=sys.stderr, st
-00005e50: 646f 7574 3d73 7973 2e73 7464 6f75 7429  dout=sys.stdout)
-00005e60: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00005e70: 2065 6c73 653a 0a20 2020 2020 2020 2020   else:.         
-00005e80: 2020 2020 2020 2020 2020 2073 7562 7072             subpr
-00005e90: 6f63 6573 732e 7275 6e28 6172 6773 3d5b  ocess.run(args=[
-00005ea0: 2269 7174 7265 6522 2c22 2d73 222c 616c  "iqtree","-s",al
-00005eb0: 6967 6e6d 656e 745f 662e 7370 6c69 7428  ignment_f.split(
-00005ec0: 222e 2229 5b30 5d2c 222d 2d61 6c6e 222c  ".")[0],"--aln",
-00005ed0: 616c 6967 6e6d 656e 745f 662c 222d 6e74  alignment_f,"-nt
-00005ee0: 222c 2241 5554 4f22 5d2c 7374 6465 7272  ","AUTO"],stderr
-00005ef0: 3d73 7973 2e73 7464 6572 722c 2073 7464  =sys.stderr, std
-00005f00: 6f75 743d 7379 732e 7374 646f 7574 290a  out=sys.stdout).
-00005f10: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00005f20: 6f73 2e72 656d 6f76 6528 616c 6967 6e6d  os.remove(alignm
-00005f30: 656e 745f 6620 2b20 222e 6c6f 6722 290a  ent_f + ".log").
-00005f40: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00005f50: 6f73 2e72 656d 6f76 6528 616c 6967 6e6d  os.remove(alignm
-00005f60: 656e 745f 6620 2b20 222e 6269 6f6e 6a22  ent_f + ".bionj"
-00005f70: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
-00005f80: 2020 6f73 2e72 656d 6f76 6528 616c 6967    os.remove(alig
-00005f90: 6e6d 656e 745f 6620 2b20 222e 636b 702e  nment_f + ".ckp.
-00005fa0: 677a 2229 0a20 2020 2020 2020 2020 2020  gz").           
-00005fb0: 2020 2020 206f 732e 7265 6d6f 7665 2861       os.remove(a
-00005fc0: 6c69 676e 6d65 6e74 5f66 202b 2022 2e6d  lignment_f + ".m
-00005fd0: 6f64 656c 2e67 7a22 290a 2020 2020 2020  odel.gz").      
-00005fe0: 2020 2020 2020 6469 7374 616e 6365 5f6d        distance_m
-00005ff0: 6174 7269 785f 6361 6c20 3d20 7064 2e72  atrix_cal = pd.r
-00006000: 6561 645f 6373 7628 616c 6967 6e6d 656e  ead_csv(alignmen
-00006010: 745f 662b 222e 6d6c 6469 7374 222c 2073  t_f+".mldist", s
-00006020: 6570 3d22 5c5c 732b 222c 2073 6b69 7072  ep="\\s+", skipr
-00006030: 6f77 733d 312c 2068 6561 6465 723d 4e6f  ows=1, header=No
-00006040: 6e65 290a 2020 2020 2020 2020 2020 2020  ne).            
-00006050: 6469 7374 616e 6365 5f6d 6174 7269 785f  distance_matrix_
-00006060: 6361 6c2e 636f 6c75 6d6e 7320 3d20 5b22  cal.columns = ["
-00006070: 726f 7773 225d 202b 2064 6973 7461 6e63  rows"] + distanc
-00006080: 655f 6d61 7472 6978 5f63 616c 2e69 6c6f  e_matrix_cal.ilo
-00006090: 635b 3a2c 305d 2e74 6f5f 6c69 7374 2829  c[:,0].to_list()
-000060a0: 0a20 2020 2020 2020 2020 2020 2064 6973  .            dis
-000060b0: 7461 6e63 655f 6d61 7472 6978 5f63 616c  tance_matrix_cal
-000060c0: 2e73 6574 5f69 6e64 6578 2822 726f 7773  .set_index("rows
-000060d0: 222c 2069 6e70 6c61 6365 3d54 7275 6529  ", inplace=True)
-000060e0: 0a20 2020 2020 2020 2020 2020 2064 6973  .            dis
-000060f0: 7461 6e63 655f 6d61 7472 6978 5f63 616c  tance_matrix_cal
-00006100: 2e69 6e64 6578 2e6e 616d 6520 3d20 2222  .index.name = ""
-00006110: 0a20 2020 2020 2020 2020 2020 2064 6973  .            dis
-00006120: 7461 6e63 655f 6d61 7472 6978 5f63 616c  tance_matrix_cal
-00006130: 2e74 6f5f 6373 7628 227b 7d2f 7b7d 5f64  .to_csv("{}/{}_d
-00006140: 6973 7461 6e63 655f 6d61 7472 6978 2e63  istance_matrix.c
-00006150: 7376 222e 666f 726d 6174 2873 746f 7261  sv".format(stora
-00006160: 6765 5f66 6f6c 6465 722c 6e61 6d65 2929  ge_folder,name))
-00006170: 0a20 2020 2020 2020 2020 2020 2074 7265  .            tre
-00006180: 6520 3d20 5472 6565 4574 6533 2861 6c69  e = TreeEte3(ali
-00006190: 676e 6d65 6e74 5f66 2b20 222e 7472 6565  gnment_f+ ".tree
-000061a0: 6669 6c65 2229 0a20 2020 2020 2020 2020  file").         
-000061b0: 2020 2072 6574 7572 6e20 7472 6565 0a20     return tree. 
-000061c0: 2020 2020 2020 2065 6c69 6620 6d65 7468         elif meth
-000061d0: 6f64 203d 3d20 2272 6170 6964 6e6a 223a  od == "rapidnj":
-000061e0: 0a20 2020 2020 2020 2020 2020 2070 7269  .            pri
-000061f0: 6e74 2822 5573 696e 6720 5261 7069 646e  nt("Using Rapidn
-00006200: 6a20 746f 2062 7569 6c64 204e 4f54 2072  j to build NOT r
-00006210: 6f6f 7465 6420 7472 6565 2e2e 2e22 290a  ooted tree...").
-00006220: 2020 2020 2020 2020 2020 2020 7472 6565              tree
-00006230: 5f66 696c 655f 6e61 6d65 203d 205b 227b  _file_name = ["{
-00006240: 7d2f 7b7d 2e74 7265 6522 2e66 6f72 6d61  }/{}.tree".forma
-00006250: 7428 7374 6f72 6167 655f 666f 6c64 6572  t(storage_folder
-00006260: 2c6e 616d 6529 2069 6620 6e6f 7420 7472  ,name) if not tr
-00006270: 6565 5f66 696c 655f 6e61 6d65 2065 6c73  ee_file_name els
-00006280: 6520 7472 6565 5f66 696c 655f 6e61 6d65  e tree_file_name
-00006290: 5d5b 305d 0a20 2020 2020 2020 2020 2020  ][0].           
-000062a0: 2061 6c69 676e 6d65 6e74 5f66 203d 205b   alignment_f = [
-000062b0: 616c 6967 6e6d 656e 745f 6669 6c65 5f6e  alignment_file_n
-000062c0: 616d 6520 6966 2061 6c69 676e 6d65 6e74  ame if alignment
-000062d0: 5f66 696c 655f 6e61 6d65 2065 6c73 6520  _file_name else 
-000062e0: 227b 7d2f 7b7d 2e6d 6166 6674 222e 666f  "{}/{}.mafft".fo
-000062f0: 726d 6174 2873 746f 7261 6765 5f66 6f6c  rmat(storage_fol
-00006300: 6465 722c 6e61 6d65 295d 5b30 5d0a 2020  der,name)][0].  
-00006310: 2020 2020 2020 2020 2020 7769 7468 206f            with o
-00006320: 7065 6e28 7472 6565 5f66 696c 655f 6e61  pen(tree_file_na
-00006330: 6d65 2c20 2277 2229 2061 7320 7472 6565  me, "w") as tree
-00006340: 5f66 696c 655f 6f75 743a 0a20 2020 2020  _file_out:.     
-00006350: 2020 2020 2020 2020 2020 2073 7562 7072             subpr
-00006360: 6f63 6573 732e 7275 6e28 6172 6773 3d5b  ocess.run(args=[
-00006370: 2272 6170 6964 6e6a 222c 616c 6967 6e6d  "rapidnj",alignm
-00006380: 656e 745f 662c 2022 2d69 222c 2022 6661  ent_f, "-i", "fa
-00006390: 225d 2c20 7374 646f 7574 3d74 7265 655f  "], stdout=tree_
-000063a0: 6669 6c65 5f6f 7574 290a 2020 2020 2020  file_out).      
-000063b0: 2020 2020 2020 7472 6565 5f66 696c 655f        tree_file_
-000063c0: 6f75 742e 636c 6f73 6528 290a 2020 2020  out.close().    
-000063d0: 2020 2020 2020 2020 7472 6565 203d 2054          tree = T
-000063e0: 7265 6545 7465 3328 7472 6565 5f66 696c  reeEte3(tree_fil
-000063f0: 655f 6e61 6d65 290a 2020 2020 2020 2020  e_name).        
-00006400: 2020 2020 7265 7475 726e 2074 7265 650a      return tree.
-00006410: 6465 6620 696e 6665 725f 616c 6967 6e6d  def infer_alignm
-00006420: 656e 7428 616c 6967 6e6d 656e 745f 6669  ent(alignment_fi
-00006430: 6c65 2c69 6e70 7574 5f6e 616d 655f 6669  le,input_name_fi
-00006440: 6c65 2c6f 7574 7075 745f 6e61 6d65 5f66  le,output_name_f
-00006450: 696c 6529 3a0a 2020 2020 2222 220a 2020  ile):.    """.  
-00006460: 2020 5265 6164 7320 616e 6420 616c 6967    Reads and alig
-00006470: 6e6d 656e 7420 6f72 2070 6572 666f 726d  nment or perform
-00006480: 7320 616c 6967 6e6d 656e 7420 7573 696e  s alignment usin
-00006490: 6720 4d41 4646 5420 5b4d 4146 4654 206d  g MAFFT [MAFFT m
-000064a0: 756c 7469 706c 6520 7365 7175 656e 6365  ultiple sequence
-000064b0: 2061 6c69 676e 6d65 6e74 2073 6f66 7477   alignment softw
-000064c0: 6172 6520 7665 7273 696f 6e20 373a 2069  are version 7: i
-000064d0: 6d70 726f 7665 6d65 6e74 7320 696e 2070  mprovements in p
-000064e0: 6572 666f 726d 616e 6365 2061 6e64 2075  erformance and u
-000064f0: 7361 6269 6c69 7479 5d2e 2052 6574 7572  sability]. Retur
-00006500: 6e73 2061 2064 6963 7469 6f6e 6172 7920  ns a dictionary 
-00006510: 7769 7468 2074 6865 2073 6571 7565 6e63  with the sequenc
-00006520: 6520 6e61 6d65 0a20 2020 2061 6e64 2074  e name.    and t
-00006530: 6865 2073 6571 7565 6e63 6520 616e 6420  he sequence and 
-00006540: 6120 6269 6f70 7974 686f 6e20 616c 6967  a biopython alig
-00006550: 6e6d 656e 7420 6f62 6a65 6374 0a20 2020  nment object.   
-00006560: 203a 7061 7261 6d20 7374 7220 6f72 204e   :param str or N
-00006570: 6f6e 6520 616c 6967 6e6d 656e 745f 6669  one alignment_fi
-00006580: 6c65 3a20 7061 7468 2074 6f20 7072 652d  le: path to pre-
-00006590: 636f 6d70 7574 6564 2061 6c69 676e 6d65  computed alignme
-000065a0: 6e74 2074 6f20 7265 6164 0a20 2020 203a  nt to read.    :
-000065b0: 7061 7261 6d20 7374 7220 696e 7075 745f  param str input_
-000065c0: 6669 6c65 5f6e 616d 653a 2070 6174 6820  file_name: path 
-000065d0: 746f 2074 6865 2066 696c 6520 636f 6e74  to the file cont
-000065e0: 6169 6e69 6e67 2075 6e61 6c69 676e 6564  aining unaligned
-000065f0: 2073 6571 7565 6e63 6573 2c69 6e20 6661   sequences,in fa
-00006600: 7374 6120 666f 726d 6174 0a20 2020 203a  sta format.    :
-00006610: 7061 7261 6d20 7374 7220 6f75 7470 7574  param str output
-00006620: 5f66 696c 655f 6e61 6d65 3a20 6e61 6d65  _file_name: name
-00006630: 206f 6620 7468 6520 6669 6c65 2074 6861   of the file tha
-00006640: 7420 7769 6c6c 2063 6f6e 7461 696e 2074  t will contain t
-00006650: 6865 2061 6c69 676e 6564 2073 6571 7565  he aligned seque
-00006660: 6e63 6573 2222 220a 2020 2020 2320 416c  nces""".    # Al
-00006670: 6967 6e20 7468 6520 706f 6c79 7065 7074  ign the polypept
-00006680: 6964 6573 2f73 6571 7565 6e63 6573 2061  ides/sequences a
-00006690: 6e64 2077 7269 7465 2074 6f20 6120 6661  nd write to a fa
-000066a0: 7374 6120 6669 6c65 0a20 2020 2070 7269  sta file.    pri
-000066b0: 6e74 2822 416e 616c 797a 696e 6720 616c  nt("Analyzing al
-000066c0: 6967 6e6d 656e 742e 2e2e 2229 0a20 2020  ignment...").   
-000066d0: 2069 6620 616c 6967 6e6d 656e 745f 6669   if alignment_fi
-000066e0: 6c65 3a20 2354 6865 2061 6c69 676e 6d65  le: #The alignme
-000066f0: 6e74 2066 696c 6520 7368 6f75 6c64 2063  nt file should c
-00006700: 6f6e 7461 696e 2074 6865 2070 6f6c 7970  ontain the polyp
-00006710: 6570 7469 6465 7320 6f66 2074 6865 2050  eptides of the P
-00006720: 4442 2073 7472 7563 7475 7265 7320 616e  DB structures an
-00006730: 6420 7365 7175 656e 6365 7320 7769 7468  d sequences with
-00006740: 6f75 7420 7374 7275 6374 7572 6573 0a20  out structures. 
-00006750: 2020 2020 2020 2070 7269 6e74 2822 5265         print("Re
-00006760: 6164 696e 6720 6769 7665 6e20 616c 6967  ading given alig
-00006770: 6e6d 656e 7420 6669 6c65 202e 2e2e 2229  nment file ...")
-00006780: 0a20 2020 2020 2020 2023 2052 6561 6420  .        # Read 
-00006790: 7468 6520 616c 6967 6e65 6420 7365 7175  the aligned sequ
-000067a0: 656e 6365 730a 2020 2020 2020 2020 616c  ences.        al
-000067b0: 6967 6e6d 656e 7420 3d20 416c 6967 6e49  ignment = AlignI
-000067c0: 4f2e 7265 6164 2822 7b7d 222e 666f 726d  O.read("{}".form
-000067d0: 6174 2861 6c69 676e 6d65 6e74 5f66 696c  at(alignment_fil
-000067e0: 6529 2c20 2266 6173 7461 2229 0a20 2020  e), "fasta").   
-000067f0: 2020 2020 2061 6c69 676e 6d65 6e74 5f69       alignment_i
-00006800: 6473 203d 205b 5d0a 2020 2020 2020 2020  ds = [].        
-00006810: 616c 6967 6e6d 656e 745f 7365 7173 203d  alignment_seqs =
-00006820: 205b 5d0a 2020 2020 2020 2020 666f 7220   [].        for 
-00006830: 692c 616c 6967 6e65 6420 696e 2065 6e75  i,aligned in enu
-00006840: 6d65 7261 7465 2861 6c69 676e 6d65 6e74  merate(alignment
-00006850: 293a 0a20 2020 2020 2020 2020 2020 2061  ):.            a
-00006860: 6c69 676e 6d65 6e74 5f69 6473 2e61 7070  lignment_ids.app
-00006870: 656e 6428 616c 6967 6e6d 656e 745b 695d  end(alignment[i]
-00006880: 2e69 6429 0a20 2020 2020 2020 2020 2020  .id).           
-00006890: 2061 6c69 676e 6d65 6e74 5f73 6571 732e   alignment_seqs.
-000068a0: 6170 7065 6e64 2861 6c69 676e 6d65 6e74  append(alignment
-000068b0: 5b69 5d2e 7365 712e 7374 7269 7028 222a  [i].seq.strip("*
-000068c0: 2229 2920 2348 6967 686c 6967 6874 3a20  ")) #Highlight: 
-000068d0: 5265 6d6f 7665 2073 746f 7020 636f 646f  Remove stop codo
-000068e0: 6e73 0a20 2020 2020 2020 2064 6963 745f  ns.        dict_
-000068f0: 616c 6967 6e6d 656e 7420 3d20 6469 6374  alignment = dict
-00006900: 287a 6970 2861 6c69 676e 6d65 6e74 5f69  (zip(alignment_i
-00006910: 6473 2c20 616c 6967 6e6d 656e 745f 7365  ds, alignment_se
-00006920: 7173 2929 0a20 2020 2020 2020 2072 6574  qs)).        ret
-00006930: 7572 6e20 6469 6374 5f61 6c69 676e 6d65  urn dict_alignme
-00006940: 6e74 2c20 616c 6967 6e6d 656e 740a 2020  nt, alignment.  
-00006950: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
-00006960: 7072 696e 7428 2255 7369 6e67 206d 6166  print("Using maf
-00006970: 6674 2074 6f20 616c 6967 6e2e 2e2e 2229  ft to align...")
-00006980: 0a20 2020 2020 2020 206d 6166 6674 5f63  .        mafft_c
-00006990: 6c69 6e65 203d 204d 6166 6674 436f 6d6d  line = MafftComm
-000069a0: 616e 646c 696e 6528 696e 7075 743d 696e  andline(input=in
-000069b0: 7075 745f 6e61 6d65 5f66 696c 6529 0a20  put_name_file). 
-000069c0: 2020 2020 2020 2073 7464 6f75 742c 2073         stdout, s
-000069d0: 7464 6572 7220 3d20 6d61 6666 745f 636c  tderr = mafft_cl
-000069e0: 696e 6528 290a 2020 2020 2020 2020 7769  ine().        wi
-000069f0: 7468 206f 7065 6e28 6f75 7470 7574 5f6e  th open(output_n
-00006a00: 616d 655f 6669 6c65 2c20 2277 2229 2061  ame_file, "w") a
-00006a10: 7320 6861 6e64 6c65 3a0a 2020 2020 2020  s handle:.      
-00006a20: 2020 2020 2020 6861 6e64 6c65 2e77 7269        handle.wri
-00006a30: 7465 2873 7464 6f75 7429 0a20 2020 2020  te(stdout).     
-00006a40: 2020 2068 616e 646c 652e 636c 6f73 6528     handle.close(
-00006a50: 290a 0a20 2020 2020 2020 2023 2052 6561  )..        # Rea
-00006a60: 6420 7468 6520 616c 6967 6e65 6420 7365  d the aligned se
-00006a70: 7175 656e 6365 730a 2020 2020 2020 2020  quences.        
-00006a80: 616c 6967 6e6d 656e 7420 3d20 416c 6967  alignment = Alig
-00006a90: 6e49 4f2e 7265 6164 286f 7574 7075 745f  nIO.read(output_
-00006aa0: 6e61 6d65 5f66 696c 652c 2022 6661 7374  name_file, "fast
-00006ab0: 6122 290a 2020 2020 2020 2020 616c 6967  a").        alig
-00006ac0: 6e6d 656e 745f 6964 7320 3d20 5b61 6c69  nment_ids = [ali
-00006ad0: 676e 6d65 6e74 5b69 5d2e 6964 2066 6f72  gnment[i].id for
-00006ae0: 2069 2c20 616c 6967 6e65 6420 696e 2065   i, aligned in e
-00006af0: 6e75 6d65 7261 7465 2861 6c69 676e 6d65  numerate(alignme
-00006b00: 6e74 295d 0a20 2020 2020 2020 2061 6c69  nt)].        ali
-00006b10: 676e 6d65 6e74 5f73 6571 7320 3d20 5b61  gnment_seqs = [a
-00006b20: 6c69 676e 6d65 6e74 5b69 5d2e 7365 7120  lignment[i].seq 
-00006b30: 666f 7220 692c 2061 6c69 676e 6564 2069  for i, aligned i
-00006b40: 6e20 656e 756d 6572 6174 6528 616c 6967  n enumerate(alig
-00006b50: 6e6d 656e 7429 5d0a 2020 2020 2020 2020  nment)].        
-00006b60: 6469 6374 5f61 6c69 676e 6d65 6e74 203d  dict_alignment =
-00006b70: 2064 6963 7428 7a69 7028 616c 6967 6e6d   dict(zip(alignm
-00006b80: 656e 745f 6964 732c 2061 6c69 676e 6d65  ent_ids, alignme
-00006b90: 6e74 5f73 6571 7329 290a 2020 2020 2020  nt_seqs)).      
-00006ba0: 2020 7265 7475 726e 2064 6963 745f 616c    return dict_al
-00006bb0: 6967 6e6d 656e 742c 2061 6c69 676e 6d65  ignment, alignme
-00006bc0: 6e74 0a64 6566 2063 616c 6375 6c61 7465  nt.def calculate
-00006bd0: 5f70 6169 7277 6973 655f 6469 7374 616e  _pairwise_distan
-00006be0: 6365 286e 616d 652c 616c 6967 6e6d 656e  ce(name,alignmen
-00006bf0: 742c 7374 6f72 6167 655f 666f 6c64 6572  t,storage_folder
-00006c00: 293a 0a20 2020 2022 2222 4361 6c63 756c  ):.    """Calcul
-00006c10: 6174 6573 2074 6865 2070 6169 7277 6973  ates the pairwis
-00006c20: 6520 6469 7374 616e 6365 206d 6174 7269  e distance matri
-00006c30: 7820 6163 6372 6f73 7320 7468 6520 7365  x accross the se
-00006c40: 7175 656e 6365 7320 696e 2074 6865 2061  quences in the a
-00006c50: 6c69 676e 6d65 6e74 0a20 2020 203a 7061  lignment.    :pa
-00006c60: 7261 6d20 7374 7220 6e61 6d65 3a20 6461  ram str name: da
-00006c70: 7461 2073 6574 2070 726f 6a65 6374 206e  ta set project n
-00006c80: 616d 650a 2020 2020 3a70 6172 616d 2062  ame.    :param b
-00006c90: 696f 7079 7468 6f6e 2061 6c69 676e 6d65  iopython alignme
-00006ca0: 6e74 206f 626a 6563 743a 206f 626a 6563  nt object: objec
-00006cb0: 7420 636f 6e74 6169 6e69 6e67 2061 6c69  t containing ali
-00006cc0: 676e 6564 2073 6571 7565 6e63 6573 0a20  gned sequences. 
-00006cd0: 2020 203a 7061 7261 6d20 7374 7220 7374     :param str st
-00006ce0: 6f72 6167 655f 666f 6c64 6572 3a20 7061  orage_folder: pa
-00006cf0: 7468 2077 6865 7265 2074 6f20 7374 6f72  th where to stor
-00006d00: 6520 7468 6520 6361 6c63 756c 6174 6564  e the calculated
-00006d10: 2070 6169 7277 6973 6520 6d61 7472 6978   pairwise matrix
-00006d20: 2222 220a 2020 2020 7072 696e 7428 2242  """.    print("B
-00006d30: 7569 6c64 696e 6720 7061 6972 7769 7365  uilding pairwise
-00006d40: 2064 6973 7461 6e63 6520 6d61 7472 6978   distance matrix
-00006d50: 202e 2e2e 2229 0a20 2020 2069 6620 6c65   ...").    if le
-00006d60: 6e28 616c 6967 6e6d 656e 7429 203c 3d20  n(alignment) <= 
-00006d70: 3230 303a 2023 7665 7279 2073 6c6f 7720  200: #very slow 
-00006d80: 6d65 7468 6f64 0a20 2020 2020 2020 2063  method.        c
-00006d90: 616c 6375 6c61 746f 7220 3d20 4469 7374  alculator = Dist
-00006da0: 616e 6365 4361 6c63 756c 6174 6f72 2827  anceCalculator('
-00006db0: 6964 656e 7469 7479 2729 0a20 2020 2020  identity').     
-00006dc0: 2020 2064 6973 7461 6e63 655f 6d61 7472     distance_matr
-00006dd0: 6978 5f62 696f 7079 7468 6f6e 203d 2063  ix_biopython = c
-00006de0: 616c 6375 6c61 746f 722e 6765 745f 6469  alculator.get_di
-00006df0: 7374 616e 6365 2861 6c69 676e 6d65 6e74  stance(alignment
-00006e00: 290a 2020 2020 2020 2020 6469 7374 616e  ).        distan
-00006e10: 6365 5f64 6620 3d20 7064 2e44 6174 6146  ce_df = pd.DataF
-00006e20: 7261 6d65 2869 6e64 6578 3d64 6973 7461  rame(index=dista
-00006e30: 6e63 655f 6d61 7472 6978 5f62 696f 7079  nce_matrix_biopy
-00006e40: 7468 6f6e 2e6e 616d 6573 2c20 636f 6c75  thon.names, colu
-00006e50: 6d6e 733d 6469 7374 616e 6365 5f6d 6174  mns=distance_mat
-00006e60: 7269 785f 6269 6f70 7974 686f 6e2e 6e61  rix_biopython.na
-00006e70: 6d65 7329 0a20 2020 2020 2020 2064 6973  mes).        dis
-00006e80: 7461 6e63 655f 6466 203d 2064 6973 7461  tance_df = dista
-00006e90: 6e63 655f 6466 2e66 696c 6c6e 6128 3029  nce_df.fillna(0)
-00006ea0: 0a20 2020 2020 2020 2066 6f72 2069 2c20  .        for i, 
-00006eb0: 7431 2069 6e20 656e 756d 6572 6174 6528  t1 in enumerate(
-00006ec0: 6469 7374 616e 6365 5f6d 6174 7269 785f  distance_matrix_
-00006ed0: 6269 6f70 7974 686f 6e2e 6e61 6d65 7329  biopython.names)
-00006ee0: 3a0a 2020 2020 2020 2020 2020 2020 666f  :.            fo
-00006ef0: 7220 6a2c 2074 3220 696e 2065 6e75 6d65  r j, t2 in enume
-00006f00: 7261 7465 286c 6973 7428 6469 7374 616e  rate(list(distan
-00006f10: 6365 5f6d 6174 7269 785f 6269 6f70 7974  ce_matrix_biopyt
-00006f20: 686f 6e2e 6e61 6d65 7329 5b69 202b 2031  hon.names)[i + 1
-00006f30: 3a5d 293a 0a20 2020 2020 2020 2020 2020  :]):.           
-00006f40: 2020 2020 2064 6973 7461 6e63 655f 6466       distance_df
-00006f50: 2e6c 6f63 5b5b 7431 5d2c 205b 7432 5d5d  .loc[[t1], [t2]]
-00006f60: 203d 2064 6973 7461 6e63 655f 6d61 7472   = distance_matr
-00006f70: 6978 5f62 696f 7079 7468 6f6e 5b74 312c  ix_biopython[t1,
-00006f80: 7432 5d0a 2020 2020 2020 2020 2020 2020  t2].            
-00006f90: 2020 2020 6469 7374 616e 6365 5f64 662e      distance_df.
-00006fa0: 6c6f 635b 5b74 325d 2c20 5b74 315d 5d20  loc[[t2], [t1]] 
-00006fb0: 3d20 6469 7374 616e 6365 5f6d 6174 7269  = distance_matri
-00006fc0: 785f 6269 6f70 7974 686f 6e5b 7431 2c74  x_biopython[t1,t
-00006fd0: 325d 0a20 2020 2020 2020 2064 6973 7461  2].        dista
-00006fe0: 6e63 655f 6466 2e74 6f5f 6373 7628 227b  nce_df.to_csv("{
-00006ff0: 7d2f 7b7d 5f70 6169 7277 6973 655f 6469  }/{}_pairwise_di
-00007000: 7374 616e 6365 5f6d 6174 7269 782e 6373  stance_matrix.cs
-00007010: 7622 2e66 6f72 6d61 7428 7374 6f72 6167  v".format(storag
-00007020: 655f 666f 6c64 6572 2c6e 616d 6529 290a  e_folder,name)).
-00007030: 0a20 2020 2065 6c73 653a 2023 544f 444f  .    else: #TODO
-00007040: 3a20 6661 7374 6572 2069 6d70 6c65 6d65  : faster impleme
-00007050: 6e74 6174 696f 6e2d 2d2d 3e20 6974 2773  ntation---> it's
-00007060: 2064 6f6e 6520 656c 7365 7768 6572 650a   done elsewhere.
-00007070: 2020 2020 2020 2020 7072 696e 7428 2246          print("F
-00007080: 696e 6973 6820 696d 706c 656d 656e 7469  inish implementi
-00007090: 6e67 2066 6f72 206c 6172 6765 7220 6461  ng for larger da
-000070a0: 7461 7365 7473 2229 0a20 2020 2020 2020  tasets").       
-000070b0: 2023 4869 6768 6c69 6768 743a 2054 7572   #Highlight: Tur
-000070c0: 6e20 616c 6967 6e6d 656e 7420 696e 746f  n alignment into
-000070d0: 206e 756d 7079 2061 7272 6179 2c20 7665   numpy array, ve
-000070e0: 6374 6f72 697a 6520 746f 206e 756d 6265  ctorize to numbe
-000070f0: 7273 2c20 636f 6d70 7574 6572 2070 6169  rs, computer pai
-00007100: 7277 6973 6520 696e 2066 6173 7420 6d61  rwise in fast ma
-00007110: 6e6e 6572 0a20 2020 2020 2020 2070 6173  nner.        pas
-00007120: 730a 6465 6620 6361 6c63 756c 6174 655f  s.def calculate_
-00007130: 7061 7472 6973 7469 635f 6469 7374 616e  patristic_distan
-00007140: 6365 286e 616d 655f 6669 6c65 2c63 6f6d  ce(name_file,com
-00007150: 6269 6e65 645f 6469 6374 2c6e 6f64 6573  bined_dict,nodes
-00007160: 5f61 6e64 5f6c 6561 6673 5f6e 616d 6573  _and_leafs_names
-00007170: 2c74 7265 652c 7472 6565 5f66 696c 652c  ,tree,tree_file,
-00007180: 2073 746f 7261 6765 5f66 6f6c 6465 7229   storage_folder)
-00007190: 3a0a 2020 2020 2222 2243 616c 6375 6c61  :.    """Calcula
-000071a0: 7465 7320 7468 6520 7061 7472 6973 7469  tes the patristi
-000071b0: 6320 6469 7374 616e 6365 7320 6f72 2062  c distances or b
-000071c0: 7261 6e63 6820 6c65 6e67 7468 7320 6163  ranch lengths ac
-000071d0: 726f 7373 2074 6865 206e 6f64 6573 2069  ross the nodes i
-000071e0: 6e20 6120 7472 6565 2e20 4974 2061 6c73  n a tree. It als
-000071f0: 6f20 7361 7665 7320 7468 6520 7472 6565  o saves the tree
-00007200: 2069 6e20 6469 6666 6572 656e 7420 666f   in different fo
-00007210: 726d 6174 7320 6e65 6564 6564 2066 6f72  rmats needed for
-00007220: 2062 656e 6368 6d61 726b 696e 6720 6574   benchmarking et
-00007230: 630a 2020 2020 3a70 6172 616d 2073 7472  c.    :param str
-00007240: 206e 616d 655f 6669 6c65 3a20 6461 7461   name_file: data
-00007250: 2073 6574 2070 726f 6a65 6374 206e 616d   set project nam
-00007260: 650a 2020 2020 3a70 6172 616d 2064 6963  e.    :param dic
-00007270: 7420 636f 6d62 696e 6564 5f64 6963 7420  t combined_dict 
-00007280: 2354 4f44 4f3a 2052 656d 6f76 653f 0a20  #TODO: Remove?. 
-00007290: 2020 203a 7061 7261 6d20 6c69 7374 206e     :param list n
-000072a0: 6f64 6573 5f61 6e64 5f6c 6561 6673 5f6e  odes_and_leafs_n
-000072b0: 616d 6573 3a20 7472 6565 206e 6f64 6573  ames: tree nodes
-000072c0: 2069 6e20 7472 6565 2d6c 6576 656c 206f   in tree-level o
-000072d0: 7264 6572 2073 746f 7265 6420 696e 2061  rder stored in a
-000072e0: 206c 6973 740a 2020 2020 3a70 6172 616d   list.    :param
-000072f0: 2065 7465 332d 7472 6565 2074 7265 653a   ete3-tree tree:
-00007300: 2065 7465 3320 6f62 6a65 6374 2063 6f6e   ete3 object con
-00007310: 7461 696e 696e 6720 7472 6565 0a20 2020  taining tree.   
-00007320: 203a 7061 7261 6d20 7374 7220 7472 6565   :param str tree
-00007330: 5f66 696c 653a 2070 6174 6820 746f 2074  _file: path to t
-00007340: 6865 2073 746f 7265 6420 7472 6565 2066  he stored tree f
-00007350: 696c 650a 2020 2020 3a70 6172 616d 2073  ile.    :param s
-00007360: 7472 2073 746f 7261 6765 5f66 6f6c 6465  tr storage_folde
-00007370: 723a 2066 6f6c 6465 7220 7768 6572 6520  r: folder where 
-00007380: 746f 2073 746f 7265 2074 6865 2072 6573  to store the res
-00007390: 756c 7473 0a20 2020 2022 2222 0a0a 2020  ults.    """..  
-000073a0: 2020 6e5f 7365 7173 203d 206c 656e 2863    n_seqs = len(c
-000073b0: 6f6d 6269 6e65 645f 6469 6374 290a 2020  ombined_dict).  
-000073c0: 2020 2377 6f72 6b5f 6469 7220 3d20 6f73    #work_dir = os
-000073d0: 2e70 6174 682e 6469 726e 616d 6528 6f73  .path.dirname(os
-000073e0: 2e70 6174 682e 6162 7370 6174 6828 5f5f  .path.abspath(__
-000073f0: 6669 6c65 5f5f 2929 0a20 2020 2077 6f72  file__)).    wor
-00007400: 6b5f 6469 7220 3d20 2222 0a20 2020 2069  k_dir = "".    i
-00007410: 6620 6e5f 7365 7173 203e 2032 3030 3a0a  f n_seqs > 200:.
-00007420: 2020 2020 2020 2020 7072 696e 7428 2244          print("D
-00007430: 6174 6173 6574 206c 6172 6765 7220 7468  ataset larger th
-00007440: 616e 2032 3030 2073 6571 7565 6e63 6573  an 200 sequences
-00007450: 3a20 5573 696e 6720 5220 7363 7269 7074  : Using R script
-00007460: 2066 6f72 2070 6174 7269 7374 6963 2064   for patristic d
-00007470: 6973 7461 6e63 6573 2028 636c 6164 6973  istances (cladis
-00007480: 7469 6320 6d61 7472 6978 2069 7320 4e4f  tic matrix is NO
-00007490: 5420 6176 6169 6c61 626c 6529 2122 290a  T available)!").
-000074a0: 2020 2020 2020 2020 7761 726e 696e 6773          warnings
-000074b0: 2e77 6172 6e28 2244 6174 6173 6574 206c  .warn("Dataset l
-000074c0: 6172 6765 7220 7468 616e 2032 3030 2073  arger than 200 s
-000074d0: 6571 7565 6e63 6573 3a20 5265 7175 6972  equences: Requir
-000074e0: 6573 2052 2061 6e64 2074 6865 2061 7065  es R and the ape
-000074f0: 206c 6962 7261 7279 2229 0a20 2020 2020   library").     
-00007500: 2020 2063 6f6d 6d61 6e64 203d 2027 5273     command = 'Rs
-00007510: 6372 6970 7427 0a20 2020 2020 2020 2023  cript'.        #
-00007520: 7061 7468 3273 6372 6970 7420 3d20 272f  path2script = '/
-00007530: 686f 6d65 2f6c 7973 2f44 726f 7062 6f78  home/lys/Dropbox
-00007540: 2f50 6844 2f44 5241 5550 4e49 522f 4361  /PhD/DRAUPNIR/Ca
-00007550: 6c63 756c 6174 655f 5061 7472 6973 7469  lculate_Patristi
-00007560: 632e 5227 0a20 2020 2020 2020 2070 6174  c.R'.        pat
-00007570: 6832 7363 7269 7074 203d 2022 4361 6c63  h2script = "Calc
-00007580: 756c 6174 655f 5061 7472 6973 7469 632e  ulate_Patristic.
-00007590: 5222 0a20 2020 2020 2020 2069 6620 7472  R".        if tr
-000075a0: 6565 5f66 696c 653a 0a20 2020 2020 2020  ee_file:.       
-000075b0: 2020 2020 206e 6577 5f74 7265 6520 3d20       new_tree = 
-000075c0: 776f 726b 5f64 6972 202b 7472 6565 5f66  work_dir +tree_f
-000075d0: 696c 652e 7370 6c69 7428 222e 2229 5b30  ile.split(".")[0
-000075e0: 5d2b 222e 6e65 7769 636b 220a 2020 2020  ]+".newick".    
-000075f0: 2020 2020 2020 2020 6e65 775f 7472 6565          new_tree
-00007600: 5f66 6f72 6d61 7438 203d 2077 6f72 6b5f  _format8 = work_
-00007610: 6469 7220 202b 2074 7265 655f 6669 6c65  dir  + tree_file
-00007620: 2e73 706c 6974 2822 2e22 295b 305d 202b  .split(".")[0] +
-00007630: 2022 2e66 6f72 6d61 7438 6e65 7769 636b   ".format8newick
-00007640: 220a 2020 2020 2020 2020 2020 2020 6e65  ".            ne
-00007650: 775f 7472 6565 5f66 6f72 6d61 7436 203d  w_tree_format6 =
-00007660: 2077 6f72 6b5f 6469 7220 202b 2074 7265   work_dir  + tre
-00007670: 655f 6669 6c65 2e73 706c 6974 2822 2e22  e_file.split("."
-00007680: 295b 305d 202b 2022 2e66 6f72 6d61 7436  )[0] + ".format6
-00007690: 6e65 7769 636b 220a 2020 2020 2020 2020  newick".        
-000076a0: 2020 2020 6e65 775f 7472 6565 5f66 6f72      new_tree_for
-000076b0: 6d61 7437 203d 2077 6f72 6b5f 6469 7220  mat7 = work_dir 
-000076c0: 202b 2074 7265 655f 6669 6c65 2e73 706c   + tree_file.spl
-000076d0: 6974 2822 2e22 295b 305d 202b 2022 2e66  it(".")[0] + ".f
-000076e0: 6f72 6d61 7437 6e65 7769 636b 220a 2020  ormat7newick".  
-000076f0: 2020 2020 2020 656c 7365 3a0a 2020 2020        else:.    
-00007700: 2020 2020 2020 2020 6e65 775f 7472 6565          new_tree
-00007710: 203d 2077 6f72 6b5f 6469 7220 2b20 227b   = work_dir + "{
-00007720: 7d2f 7b7d 2e6e 6577 6963 6b22 2e66 6f72  }/{}.newick".for
-00007730: 6d61 7428 7374 6f72 6167 655f 666f 6c64  mat(storage_fold
-00007740: 6572 2c6e 616d 655f 6669 6c65 290a 2020  er,name_file).  
-00007750: 2020 2020 2020 2020 2020 6e65 775f 7472            new_tr
-00007760: 6565 5f66 6f72 6d61 7438 203d 2077 6f72  ee_format8 = wor
-00007770: 6b5f 6469 7220 2b20 227b 7d2f 7b7d 2e66  k_dir + "{}/{}.f
-00007780: 6f72 6d61 7438 6e65 7769 636b 222e 666f  ormat8newick".fo
-00007790: 726d 6174 2873 746f 7261 6765 5f66 6f6c  rmat(storage_fol
-000077a0: 6465 722c 6e61 6d65 5f66 696c 6529 0a20  der,name_file). 
-000077b0: 2020 2020 2020 2020 2020 206e 6577 5f74             new_t
-000077c0: 7265 655f 666f 726d 6174 3620 3d20 776f  ree_format6 = wo
-000077d0: 726b 5f64 6972 202b 2022 7b7d 2f7b 7d2e  rk_dir + "{}/{}.
-000077e0: 666f 726d 6174 366e 6577 6963 6b22 2e66  format6newick".f
-000077f0: 6f72 6d61 7428 7374 6f72 6167 655f 666f  ormat(storage_fo
-00007800: 6c64 6572 2c6e 616d 655f 6669 6c65 290a  lder,name_file).
-00007810: 2020 2020 2020 2020 2020 2020 6e65 775f              new_
-00007820: 7472 6565 5f66 6f72 6d61 7437 203d 2077  tree_format7 = w
-00007830: 6f72 6b5f 6469 7220 2b20 227b 7d2f 7b7d  ork_dir + "{}/{}
-00007840: 2e66 6f72 6d61 7437 6e65 7769 636b 222e  .format7newick".
-00007850: 666f 726d 6174 2873 746f 7261 6765 5f66  format(storage_f
-00007860: 6f6c 6465 722c 6e61 6d65 5f66 696c 6529  older,name_file)
-00007870: 0a0a 2020 2020 2020 2020 7472 6565 2e77  ..        tree.w
-00007880: 7269 7465 286f 7574 6669 6c65 3d6e 6577  rite(outfile=new
-00007890: 5f74 7265 655f 666f 726d 6174 382c 2066  _tree_format8, f
-000078a0: 6f72 6d61 743d 382c 666f 726d 6174 5f72  ormat=8,format_r
-000078b0: 6f6f 745f 6e6f 6465 3d54 7275 6529 2023  oot_node=True) #
-000078c0: 2066 6f72 6d61 7420 3820 616c 6c20 6e6f   format 8 all no
-000078d0: 6465 7320 6e61 6d65 730a 2020 2020 2020  des names.      
-000078e0: 2020 7472 6565 2e77 7269 7465 286f 7574    tree.write(out
-000078f0: 6669 6c65 3d6e 6577 5f74 7265 655f 666f  file=new_tree_fo
-00007900: 726d 6174 362c 2066 6f72 6d61 743d 362c  rmat6, format=6,
-00007910: 666f 726d 6174 5f72 6f6f 745f 6e6f 6465  format_root_node
-00007920: 3d54 7275 6529 0a20 2020 2020 2020 2074  =True).        t
-00007930: 7265 652e 7772 6974 6528 6f75 7466 696c  ree.write(outfil
-00007940: 653d 6e65 775f 7472 6565 5f66 6f72 6d61  e=new_tree_forma
-00007950: 7437 2c20 666f 726d 6174 3d37 2c66 6f72  t7, format=7,for
-00007960: 6d61 745f 726f 6f74 5f6e 6f64 653d 5472  mat_root_node=Tr
-00007970: 7565 2920 2361 6c6c 206e 6f64 6573 206e  ue) #all nodes n
-00007980: 616d 6573 202b 2062 7261 6e63 6820 6c65  ames + branch le
-00007990: 6e67 7468 730a 2020 2020 2020 2020 7472  ngths.        tr
-000079a0: 6565 2e77 7269 7465 286f 7574 6669 6c65  ee.write(outfile
-000079b0: 3d6e 6577 5f74 7265 652c 666f 726d 6174  =new_tree,format
-000079c0: 3d31 2920 2373 6176 6520 7468 6520 7265  =1) #save the re
-000079d0: 6e61 6d65 6420 7472 6565 2c20 666f 726d  named tree, form
-000079e0: 6174 2039 2074 6f20 6e6f 7420 7361 7665  at 9 to not save
-000079f0: 2074 6865 2069 6e74 6572 6e61 6c20 6e6f   the internal no
-00007a00: 6465 7320 6e61 6d65 730a 2020 2020 2020  des names.      
-00007a10: 2020 7061 7472 6973 7469 635f 6669 6c65    patristic_file
-00007a20: 203d 2022 7b7d 2f7b 7d5f 7061 7472 6973   = "{}/{}_patris
-00007a30: 7469 635f 6469 7374 616e 6365 5f6d 6174  tic_distance_mat
-00007a40: 7269 782e 6373 7622 2e66 6f72 6d61 7428  rix.csv".format(
-00007a50: 7374 6f72 6167 655f 666f 6c64 6572 2c6e  storage_folder,n
-00007a60: 616d 655f 6669 6c65 290a 2020 2020 2020  ame_file).      
-00007a70: 2020 6966 206e 6f74 206f 732e 7061 7468    if not os.path
-00007a80: 2e65 7869 7374 7328 7061 7472 6973 7469  .exists(patristi
-00007a90: 635f 6669 6c65 293a 0a20 2020 2020 2020  c_file):.       
-00007aa0: 2020 2020 2023 2042 7569 6c64 2073 7562       # Build sub
-00007ab0: 7072 6f63 6573 7320 636f 6d6d 616e 640a  process command.
-00007ac0: 2020 2020 2020 2020 2020 2020 7375 6270              subp
-00007ad0: 726f 6365 7373 2e63 6865 636b 5f63 616c  rocess.check_cal
-00007ae0: 6c28 5b63 6f6d 6d61 6e64 2c70 6174 6832  l([command,path2
-00007af0: 7363 7269 7074 2c6e 6577 5f74 7265 652c  script,new_tree,
-00007b00: 7061 7472 6973 7469 635f 6669 6c65 5d29  patristic_file])
-00007b10: 0a20 2020 2020 2020 2065 6c73 653a 0a20  .        else:. 
-00007b20: 2020 2020 2020 2020 2020 2070 7269 6e74             print
-00007b30: 2822 5061 7472 6973 7469 6320 6d61 7472  ("Patristic matr
-00007b40: 6978 2061 6c72 6561 6479 2065 7869 7374  ix already exist
-00007b50: 7320 6174 207b 7d2c 206e 6f74 2063 616c  s at {}, not cal
-00007b60: 6375 6c61 7465 642e 2044 656c 6574 6520  culated. Delete 
-00007b70: 6974 206f 7468 6572 7769 7365 222e 666f  it otherwise".fo
-00007b80: 726d 6174 2870 6174 7269 7374 6963 5f66  rmat(patristic_f
-00007b90: 696c 6529 290a 2020 2020 2020 2020 2348  ile)).        #H
-00007ba0: 6967 686c 6967 6874 3a20 5048 594c 4f43  ighlight: PHYLOC
-00007bb0: 4f4d 0a20 2020 2020 2020 2023 2023 2023  OM.        # # #
-00007bc0: 4869 6768 6c69 6768 743a 5472 616e 7366  Highlight:Transf
-00007bd0: 6f72 6d20 7468 6520 6669 6c65 2074 6f20  orm the file to 
-00007be0: 4e65 7875 7320 666f 726d 6174 0a20 2020  Nexus format.   
-00007bf0: 2020 2020 2023 206e 6577 5f74 7265 6520       # new_tree 
-00007c00: 3d20 7472 6565 5f66 696c 652e 7370 6c69  = tree_file.spli
-00007c10: 7428 222e 2229 5b30 5d2b 222e 6e65 7769  t(".")[0]+".newi
-00007c20: 636b 220a 2020 2020 2020 2020 2320 6e65  ck".        # ne
-00007c30: 775f 7472 6565 5f66 696c 656e 616d 6520  w_tree_filename 
-00007c40: 3d20 6e74 7061 7468 2e62 6173 656e 616d  = ntpath.basenam
-00007c50: 6528 6e65 775f 7472 6565 290a 2020 2020  e(new_tree).    
-00007c60: 2020 2020 2320 776f 726b 696e 675f 6469      # working_di
-00007c70: 7265 6374 6f72 7920 3d20 6f73 2e70 6174  rectory = os.pat
-00007c80: 682e 6469 726e 616d 6528 6f73 2e70 6174  h.dirname(os.pat
-00007c90: 682e 6162 7370 6174 6828 6e65 775f 7472  h.abspath(new_tr
-00007ca0: 6565 2929 0a20 2020 2020 2020 2023 2074  ee)).        # t
-00007cb0: 7265 652e 7772 6974 6528 6f75 7466 696c  ree.write(outfil
-00007cc0: 653d 6e65 775f 7472 6565 2c66 6f72 6d61  e=new_tree,forma
-00007cd0: 743d 3129 2023 7361 7665 2074 6865 2072  t=1) #save the r
-00007ce0: 656e 616d 6564 2074 7265 652c 2066 6f72  enamed tree, for
-00007cf0: 6d61 7420 3920 746f 206e 6f74 2073 6176  mat 9 to not sav
-00007d00: 6520 7468 6520 696e 7465 726e 616c 206e  e the internal n
-00007d10: 6f64 6573 206e 616d 6573 0a20 2020 2020  odes names.     
-00007d20: 2020 2023 2070 6174 7269 7374 6963 5f66     # patristic_f
-00007d30: 696c 6520 3d20 2244 6174 6173 6574 735f  ile = "Datasets_
-00007d40: 466f 6c64 6572 2f50 6174 7269 7374 6963  Folder/Patristic
-00007d50: 5f64 6973 7461 6e63 655f 6d61 7472 6978  _distance_matrix
-00007d60: 5f7b 7d2e 7478 7422 2e66 6f72 6d61 7428  _{}.txt".format(
-00007d70: 6e61 6d65 5f66 696c 6529 0a20 2020 2020  name_file).     
-00007d80: 2020 2023 2077 6974 6820 6f70 656e 2870     # with open(p
-00007d90: 6174 7269 7374 6963 5f66 696c 652c 2022  atristic_file, "
-00007da0: 7722 2920 6173 2070 6174 7269 7374 6963  w") as patristic
-00007db0: 5f64 6973 745f 6f75 743a 0a20 2020 2020  _dist_out:.     
-00007dc0: 2020 2023 2020 2020 2073 7562 7072 6f63     #     subproc
-00007dd0: 6573 732e 7275 6e28 6172 6773 3d5b 2270  ess.run(args=["p
-00007de0: 6879 6c6f 636f 6d22 2c22 7068 7964 6973  hylocom","phydis
-00007df0: 7422 2c20 222d 6622 2c20 6e65 775f 7472  t", "-f", new_tr
-00007e00: 6565 5f66 696c 656e 616d 655d 2c73 7464  ee_filename],std
-00007e10: 6572 723d 7379 732e 7374 6465 7272 2c20  err=sys.stderr, 
-00007e20: 7374 646f 7574 3d70 6174 7269 7374 6963  stdout=patristic
-00007e30: 5f64 6973 745f 6f75 742c 6377 643d 776f  _dist_out,cwd=wo
-00007e40: 726b 696e 675f 6469 7265 6374 6f72 7929  rking_directory)
-00007e50: 0a20 2020 2020 2020 2023 2070 6174 7269  .        # patri
-00007e60: 7374 6963 5f6d 6174 7269 7820 3d20 7064  stic_matrix = pd
-00007e70: 2e72 6561 645f 6373 7628 7061 7472 6973  .read_csv(patris
-00007e80: 7469 635f 6669 6c65 2c73 6570 3d22 5c74  tic_file,sep="\t
-00007e90: 222c 696e 6465 785f 636f 6c3d 3029 0a20  ",index_col=0). 
-00007ea0: 2020 2020 2020 2023 2070 6174 7269 7374         # patrist
-00007eb0: 6963 5f6d 6174 7269 782e 746f 5f63 7376  ic_matrix.to_csv
-00007ec0: 2822 4461 7461 7365 7473 5f46 6f6c 6465  ("Datasets_Folde
-00007ed0: 722f 5061 7472 6973 7469 635f 6469 7374  r/Patristic_dist
-00007ee0: 616e 6365 5f6d 6174 7269 785f 7b7d 2e63  ance_matrix_{}.c
-00007ef0: 7376 222e 666f 726d 6174 286e 616d 655f  sv".format(name_
-00007f00: 6669 6c65 292c 696e 6465 785f 6c61 6265  file),index_labe
-00007f10: 6c3d 2272 6f77 7322 290a 0a20 2020 2065  l="rows")..    e
-00007f20: 6c73 653a 0a20 2020 2020 2020 2069 6620  lse:.        if 
-00007f30: 7472 6565 5f66 696c 653a 0a20 2020 2020  tree_file:.     
-00007f40: 2020 2020 2020 206e 6577 5f74 7265 6520         new_tree 
-00007f50: 3d20 776f 726b 5f64 6972 202b 2022 2f22  = work_dir + "/"
-00007f60: 202b 2074 7265 655f 6669 6c65 2e73 706c   + tree_file.spl
-00007f70: 6974 2822 2e22 295b 305d 202b 222e 6e65  it(".")[0] +".ne
-00007f80: 7769 636b 220a 2020 2020 2020 2020 2020  wick".          
-00007f90: 2020 6e65 775f 7472 6565 5f66 6f72 6d61    new_tree_forma
-00007fa0: 7438 203d 2077 6f72 6b5f 6469 7220 2b20  t8 = work_dir + 
-00007fb0: 222f 2220 2b20 7472 6565 5f66 696c 652e  "/" + tree_file.
-00007fc0: 7370 6c69 7428 222e 2229 5b30 5d20 2b20  split(".")[0] + 
-00007fd0: 222e 666f 726d 6174 386e 6577 6963 6b22  ".format8newick"
-00007fe0: 0a20 2020 2020 2020 2020 2020 206e 6577  .            new
-00007ff0: 5f74 7265 655f 666f 726d 6174 3620 3d20  _tree_format6 = 
-00008000: 776f 726b 5f64 6972 202b 2022 2f22 202b  work_dir + "/" +
-00008010: 2074 7265 655f 6669 6c65 2e73 706c 6974   tree_file.split
-00008020: 2822 2e22 295b 305d 202b 2022 2e66 6f72  (".")[0] + ".for
-00008030: 6d61 7436 6e65 7769 636b 220a 2020 2020  mat6newick".    
-00008040: 2020 2020 2020 2020 6e65 775f 7472 6565          new_tree
-00008050: 5f66 6f72 6d61 7437 203d 2077 6f72 6b5f  _format7 = work_
-00008060: 6469 7220 2b20 222f 2220 2b20 7472 6565  dir + "/" + tree
-00008070: 5f66 696c 652e 7370 6c69 7428 222e 2229  _file.split(".")
-00008080: 5b30 5d20 2b20 222e 666f 726d 6174 376e  [0] + ".format7n
-00008090: 6577 6963 6b22 0a20 2020 2020 2020 2065  ewick".        e
-000080a0: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
-000080b0: 206e 6577 5f74 7265 6520 3d20 776f 726b   new_tree = work
-000080c0: 5f64 6972 202b 2022 2f7b 7d2f 7b7d 2e6e  _dir + "/{}/{}.n
-000080d0: 6577 6963 6b22 2e66 6f72 6d61 7428 7374  ewick".format(st
-000080e0: 6f72 6167 655f 666f 6c64 6572 2c6e 616d  orage_folder,nam
-000080f0: 655f 6669 6c65 290a 2020 2020 2020 2020  e_file).        
-00008100: 2020 2020 6e65 775f 7472 6565 5f66 6f72      new_tree_for
-00008110: 6d61 7438 203d 2077 6f72 6b5f 6469 7220  mat8 = work_dir 
-00008120: 2b20 222f 7b7d 2f7b 7d2e 666f 726d 6174  + "/{}/{}.format
-00008130: 386e 6577 6963 6b22 2e66 6f72 6d61 7428  8newick".format(
-00008140: 7374 6f72 6167 655f 666f 6c64 6572 2c6e  storage_folder,n
-00008150: 616d 655f 6669 6c65 290a 2020 2020 2020  ame_file).      
-00008160: 2020 2020 2020 6e65 775f 7472 6565 5f66        new_tree_f
-00008170: 6f72 6d61 7436 203d 2077 6f72 6b5f 6469  ormat6 = work_di
-00008180: 7220 2b20 222f 7b7d 2f7b 7d2e 666f 726d  r + "/{}/{}.form
-00008190: 6174 366e 6577 6963 6b22 2e66 6f72 6d61  at6newick".forma
-000081a0: 7428 7374 6f72 6167 655f 666f 6c64 6572  t(storage_folder
-000081b0: 2c6e 616d 655f 6669 6c65 290a 2020 2020  ,name_file).    
-000081c0: 2020 2020 2020 2020 6e65 775f 7472 6565          new_tree
-000081d0: 5f66 6f72 6d61 7437 203d 2077 6f72 6b5f  _format7 = work_
-000081e0: 6469 7220 2b20 222f 7b7d 2f7b 7d2e 666f  dir + "/{}/{}.fo
-000081f0: 726d 6174 376e 6577 6963 6b22 2e66 6f72  rmat7newick".for
-00008200: 6d61 7428 7374 6f72 6167 655f 666f 6c64  mat(storage_fold
-00008210: 6572 2c6e 616d 655f 6669 6c65 290a 0a0a  er,name_file)...
-00008220: 2020 2020 2020 2020 7472 6565 2e77 7269          tree.wri
-00008230: 7465 286f 7574 6669 6c65 3d6e 6577 5f74  te(outfile=new_t
-00008240: 7265 655f 666f 726d 6174 382c 2066 6f72  ree_format8, for
-00008250: 6d61 743d 382c 666f 726d 6174 5f72 6f6f  mat=8,format_roo
-00008260: 745f 6e6f 6465 3d54 7275 6529 0a20 2020  t_node=True).   
-00008270: 2020 2020 2074 7265 652e 7772 6974 6528       tree.write(
-00008280: 6f75 7466 696c 653d 6e65 775f 7472 6565  outfile=new_tree
-00008290: 5f66 6f72 6d61 7436 2c20 666f 726d 6174  _format6, format
-000082a0: 3d36 2c66 6f72 6d61 745f 726f 6f74 5f6e  =6,format_root_n
-000082b0: 6f64 653d 5472 7565 290a 2020 2020 2020  ode=True).      
-000082c0: 2020 7472 6565 2e77 7269 7465 286f 7574    tree.write(out
-000082d0: 6669 6c65 3d6e 6577 5f74 7265 655f 666f  file=new_tree_fo
-000082e0: 726d 6174 372c 2066 6f72 6d61 743d 372c  rmat7, format=7,
-000082f0: 666f 726d 6174 5f72 6f6f 745f 6e6f 6465  format_root_node
-00008300: 3d54 7275 6529 0a20 2020 2020 2020 2074  =True).        t
-00008310: 7265 652e 7772 6974 6528 6f75 7466 696c  ree.write(outfil
-00008320: 653d 6e65 775f 7472 6565 2c66 6f72 6d61  e=new_tree,forma
-00008330: 743d 3129 2020 2320 7361 7665 2074 6865  t=1)  # save the
-00008340: 2072 656e 616d 6564 2074 7265 652c 2066   renamed tree, f
-00008350: 6f72 6d61 7420 3920 746f 206e 6f74 2073  ormat 9 to not s
-00008360: 6176 6520 7468 6520 696e 7465 726e 616c  ave the internal
-00008370: 206e 6f64 6573 206e 616d 6573 2e20 666f   nodes names. fo
-00008380: 726d 6174 2038 2061 6c6c 206e 6f64 6573  rmat 8 all nodes
-00008390: 206e 616d 6573 0a0a 2020 2020 2020 2020   names..        
-000083a0: 6e5f 656c 656d 656e 7473 203d 206c 656e  n_elements = len
-000083b0: 286e 6f64 6573 5f61 6e64 5f6c 6561 6673  (nodes_and_leafs
-000083c0: 5f6e 616d 6573 290a 2020 2020 2020 2020  _names).        
-000083d0: 4920 3d20 7064 2e49 6e64 6578 286e 6f64  I = pd.Index(nod
-000083e0: 6573 5f61 6e64 5f6c 6561 6673 5f6e 616d  es_and_leafs_nam
-000083f0: 6573 2c20 6e61 6d65 3d22 726f 7773 2229  es, name="rows")
-00008400: 0a20 2020 2020 2020 2043 203d 2070 642e  .        C = pd.
-00008410: 496e 6465 7828 6e6f 6465 735f 616e 645f  Index(nodes_and_
-00008420: 6c65 6166 735f 6e61 6d65 732c 206e 616d  leafs_names, nam
-00008430: 653d 2263 6f6c 756d 6e73 2229 0a20 2020  e="columns").   
-00008440: 2020 2020 2070 6174 7269 7374 6963 5f6d       patristic_m
-00008450: 6174 7269 7820 3d20 7064 2e44 6174 6146  atrix = pd.DataF
-00008460: 7261 6d65 2864 6174 613d 6e70 2e7a 6572  rame(data=np.zer
-00008470: 6f73 2828 6e5f 656c 656d 656e 7473 2c20  os((n_elements, 
-00008480: 6e5f 656c 656d 656e 7473 2929 2c20 696e  n_elements)), in
-00008490: 6465 783d 492c 2063 6f6c 756d 6e73 3d43  dex=I, columns=C
-000084a0: 290a 2020 2020 2020 2020 636c 6164 6973  ).        cladis
-000084b0: 7469 635f 6d61 7472 6978 203d 2070 642e  tic_matrix = pd.
-000084c0: 4461 7461 4672 616d 6528 6461 7461 3d6e  DataFrame(data=n
-000084d0: 702e 7a65 726f 7328 286e 5f65 6c65 6d65  p.zeros((n_eleme
-000084e0: 6e74 732c 206e 5f65 6c65 6d65 6e74 7329  nts, n_elements)
-000084f0: 292c 2069 6e64 6578 3d49 2c20 636f 6c75  ), index=I, colu
-00008500: 6d6e 733d 4329 0a20 2020 2020 2020 2069  mns=C).        i
-00008510: 6620 6e6f 7420 6f73 2e70 6174 682e 6578  f not os.path.ex
-00008520: 6973 7473 2822 7b7d 2f7b 7d5f 7061 7472  ists("{}/{}_patr
-00008530: 6973 7469 635f 6469 7374 616e 6365 5f6d  istic_distance_m
-00008540: 6174 7269 782e 6373 7622 2e66 6f72 6d61  atrix.csv".forma
-00008550: 7428 7374 6f72 6167 655f 666f 6c64 6572  t(storage_folder
-00008560: 2c6e 616d 655f 6669 6c65 2929 3a0a 2020  ,name_file)):.  
-00008570: 2020 2020 2020 2020 2020 666f 7220 692c            for i,
-00008580: 2074 3120 696e 2065 6e75 6d65 7261 7465   t1 in enumerate
-00008590: 286e 6f64 6573 5f61 6e64 5f6c 6561 6673  (nodes_and_leafs
-000085a0: 5f6e 616d 6573 293a 0a20 2020 2020 2020  _names):.       
-000085b0: 2020 2020 2020 2020 2066 6f72 206a 2c20           for j, 
-000085c0: 7432 2069 6e20 656e 756d 6572 6174 6528  t2 in enumerate(
-000085d0: 6c69 7374 286e 6f64 6573 5f61 6e64 5f6c  list(nodes_and_l
-000085e0: 6561 6673 5f6e 616d 6573 295b 6920 2b20  eafs_names)[i + 
-000085f0: 313a 5d29 3a0a 2020 2020 2020 2020 2020  1:]):.          
-00008600: 2020 2020 2020 2020 2020 636c 6164 6973            cladis
-00008610: 7469 635f 6d61 7472 6978 2e6c 6f63 5b5b  tic_matrix.loc[[
-00008620: 7431 5d2c 205b 7432 5d5d 203d 2074 7265  t1], [t2]] = tre
-00008630: 652e 6765 745f 6469 7374 616e 6365 2874  e.get_distance(t
-00008640: 312c 2074 322c 2074 6f70 6f6c 6f67 795f  1, t2, topology_
-00008650: 6f6e 6c79 3d54 7275 6529 0a20 2020 2020  only=True).     
-00008660: 2020 2020 2020 2020 2020 2020 2020 2070                 p
-00008670: 6174 7269 7374 6963 5f6d 6174 7269 782e  atristic_matrix.
-00008680: 6c6f 635b 5b74 315d 2c20 5b74 325d 5d20  loc[[t1], [t2]] 
-00008690: 3d20 7472 6565 2e67 6574 5f64 6973 7461  = tree.get_dista
-000086a0: 6e63 6528 7431 2c20 7432 2c20 746f 706f  nce(t1, t2, topo
-000086b0: 6c6f 6779 5f6f 6e6c 793d 4661 6c73 6529  logy_only=False)
-000086c0: 0a20 2020 2020 2020 2020 2020 2063 6c61  .            cla
-000086d0: 6469 7374 6963 5f6d 6174 7269 782e 746f  distic_matrix.to
-000086e0: 5f63 7376 2822 7b7d 2f7b 7d5f 636c 6164  _csv("{}/{}_clad
-000086f0: 6973 7469 635f 6469 7374 616e 6365 5f6d  istic_distance_m
-00008700: 6174 7269 782e 6373 7622 2e66 6f72 6d61  atrix.csv".forma
-00008710: 7428 7374 6f72 6167 655f 666f 6c64 6572  t(storage_folder
-00008720: 2c6e 616d 655f 6669 6c65 2929 0a20 2020  ,name_file)).   
-00008730: 2020 2020 2020 2020 2070 6174 7269 7374           patrist
-00008740: 6963 5f6d 6174 7269 782e 746f 5f63 7376  ic_matrix.to_csv
-00008750: 2822 7b7d 2f7b 7d5f 7061 7472 6973 7469  ("{}/{}_patristi
-00008760: 635f 6469 7374 616e 6365 5f6d 6174 7269  c_distance_matri
-00008770: 782e 6373 7622 2e66 6f72 6d61 7428 7374  x.csv".format(st
-00008780: 6f72 6167 655f 666f 6c64 6572 2c6e 616d  orage_folder,nam
-00008790: 655f 6669 6c65 2929 0a20 2020 2020 2020  e_file)).       
-000087a0: 2065 6c73 653a 0a20 2020 2020 2020 2020   else:.         
-000087b0: 2020 2070 7269 6e74 2822 5061 7472 6973     print("Patris
-000087c0: 7469 6320 6d61 7472 6978 2066 696c 6520  tic matrix file 
-000087d0: 616c 7265 6164 7920 6578 6973 7473 2c20  already exists, 
-000087e0: 6e6f 7420 6361 6c63 756c 6174 6564 2229  not calculated")
-000087f0: 0a64 6566 206d 795f 6c61 796f 7574 286e  .def my_layout(n
-00008800: 6f64 6529 3a0a 2020 2020 2222 2245 7465  ode):.    """Ete
-00008810: 3320 6c61 796f 7574 2074 6861 7420 6164  3 layout that ad
-00008820: 6473 2074 6865 2069 6e74 6572 6e61 6c20  ds the internal 
-00008830: 6e6f 6465 7320 6e61 6d65 732e 2049 7420  nodes names. It 
-00008840: 6973 2061 2070 6c75 672d 696e 2066 6f72  is a plug-in for
-00008850: 2072 656e 6465 7269 6e67 2074 7265 6520   rendering tree 
-00008860: 696d 6167 6573 0a20 2020 203a 7061 7261  images.    :para
-00008870: 6d20 6574 6533 2d6e 6f64 6520 6e6f 6465  m ete3-node node
-00008880: 3a20 6e6f 6465 2066 726f 6d20 616e 2065  : node from an e
-00008890: 7465 3320 7472 6565 2222 220a 2020 2020  te3 tree""".    
-000088a0: 6966 206e 6f64 652e 6973 5f6c 6561 6628  if node.is_leaf(
-000088b0: 293a 0a20 2020 2020 2020 2023 2049 6620  ):.        # If 
-000088c0: 7465 726d 696e 616c 206e 6f64 652c 2064  terminal node, d
-000088d0: 7261 7773 2069 7473 206e 616d 650a 2020  raws its name.  
-000088e0: 2020 2020 2020 6e61 6d65 5f66 6163 6520        name_face 
-000088f0: 3d20 4174 7472 4661 6365 2822 6e61 6d65  = AttrFace("name
-00008900: 222c 6673 697a 653d 382c 6667 636f 6c6f  ",fsize=8,fgcolo
-00008910: 723d 2262 6c75 6522 290a 2020 2020 656c  r="blue").    el
-00008920: 7365 3a0a 2020 2020 2020 2020 2320 4966  se:.        # If
-00008930: 2069 6e74 6572 6e61 6c20 6e6f 6465 2c20   internal node, 
-00008940: 6472 6177 7320 6c61 6265 6c20 7769 7468  draws label with
-00008950: 2073 6d61 6c6c 6572 2066 6f6e 7420 7369   smaller font si
-00008960: 7a65 0a20 2020 2020 2020 206e 616d 655f  ze.        name_
-00008970: 6661 6365 203d 2041 7474 7246 6163 6528  face = AttrFace(
-00008980: 226e 616d 6522 2c20 6673 697a 653d 382c  "name", fsize=8,
-00008990: 6667 636f 6c6f 723d 2272 6564 2229 0a20  fgcolor="red"). 
-000089a0: 2020 2023 2041 6464 7320 7468 6520 6e61     # Adds the na
-000089b0: 6d65 2066 6163 6520 746f 2074 6865 2069  me face to the i
-000089c0: 6d61 6765 2061 7420 7468 6520 7072 6566  mage at the pref
-000089d0: 6572 7265 6420 706f 7369 7469 6f6e 0a20  erred position. 
-000089e0: 2020 2066 6163 6573 2e61 6464 5f66 6163     faces.add_fac
-000089f0: 655f 746f 5f6e 6f64 6528 6e61 6d65 5f66  e_to_node(name_f
-00008a00: 6163 652c 206e 6f64 652c 2063 6f6c 756d  ace, node, colum
-00008a10: 6e3d 302c 2070 6f73 6974 696f 6e3d 2262  n=0, position="b
-00008a20: 7261 6e63 682d 7269 6768 7422 290a 6465  ranch-right").de
-00008a30: 6620 7265 6e64 6572 5f74 7265 6528 7472  f render_tree(tr
-00008a40: 6565 2c73 746f 7261 6765 5f66 6f6c 6465  ee,storage_folde
-00008a50: 722c 6e61 6d65 5f66 696c 6529 3a0a 2020  r,name_file):.  
-00008a60: 2020 2222 2246 756e 6374 696f 6e20 746f    """Function to
-00008a70: 2072 656e 6465 7220 616e 2065 7465 3320   render an ete3 
-00008a80: 7472 6565 2069 6e74 6f20 616e 2069 6d61  tree into an ima
-00008a90: 6765 0a20 2020 203a 7061 7261 6d20 6574  ge.    :param et
-00008aa0: 6533 2d74 7265 6520 7472 6565 3a20 4574  e3-tree tree: Et
-00008ab0: 6533 2074 7265 6520 6f62 6a65 6374 0a20  e3 tree object. 
-00008ac0: 2020 203a 7061 7261 6d20 7374 7220 7374     :param str st
-00008ad0: 6f72 6167 655f 666f 6c64 6572 3a20 7061  orage_folder: pa
-00008ae0: 7468 2074 6f20 666f 6c64 6572 2077 6865  th to folder whe
-00008af0: 7265 2074 6f20 7374 6f72 6520 7468 6520  re to store the 
-00008b00: 7265 7375 6c74 730a 2020 2020 3a70 6172  results.    :par
-00008b10: 616d 206e 616d 655f 6669 6c65 3a20 6461  am name_file: da
-00008b20: 7461 2073 6574 2070 726f 6a65 6374 206e  ta set project n
-00008b30: 616d 6522 2222 0a20 2020 2074 7320 3d20  ame""".    ts = 
-00008b40: 5472 6565 5374 796c 6528 290a 2020 2020  TreeStyle().    
-00008b50: 6e73 203d 204e 6f64 6553 7479 6c65 2829  ns = NodeStyle()
-00008b60: 0a20 2020 2023 4d61 6b65 2074 6869 636b  .    #Make thick
-00008b70: 6572 206c 696e 6573 0a20 2020 206e 735b  er lines.    ns[
-00008b80: 2276 745f 6c69 6e65 5f77 6964 7468 225d  "vt_line_width"]
-00008b90: 203d 2035 0a20 2020 206e 735b 2268 7a5f   = 5.    ns["hz_
-00008ba0: 6c69 6e65 5f77 6964 7468 225d 203d 2035  line_width"] = 5
-00008bb0: 0a20 2020 2023 2044 6f20 6e6f 7420 6164  .    # Do not ad
-00008bc0: 6420 6c65 6166 206e 616d 6573 2061 7574  d leaf names aut
-00008bd0: 6f6d 6174 6963 616c 6c79 0a20 2020 2074  omatically.    t
-00008be0: 732e 7368 6f77 5f6c 6561 665f 6e61 6d65  s.show_leaf_name
-00008bf0: 203d 2046 616c 7365 0a20 2020 2023 2055   = False.    # U
-00008c00: 7365 206d 7920 6375 7374 6f6d 206c 6179  se my custom lay
-00008c10: 6f75 740a 2020 2020 7473 2e6c 6179 6f75  out.    ts.layou
-00008c20: 745f 666e 203d 206d 795f 6c61 796f 7574  t_fn = my_layout
-00008c30: 0a20 2020 2023 7072 696e 7420 7468 6520  .    #print the 
-00008c40: 6272 616e 6368 206c 656e 6774 6873 0a20  branch lengths. 
-00008c50: 2020 2074 732e 7368 6f77 5f62 7261 6e63     ts.show_branc
-00008c60: 685f 6c65 6e67 7468 203d 2054 7275 650a  h_length = True.
-00008c70: 2020 2020 666f 7220 6e20 696e 2074 7265      for n in tre
-00008c80: 652e 7472 6176 6572 7365 2829 3a0a 2020  e.traverse():.  
-00008c90: 2020 2020 2020 2020 2020 6e2e 7365 745f            n.set_
-00008ca0: 7374 796c 6528 6e73 290a 2020 2020 7472  style(ns).    tr
-00008cb0: 793a 0a20 2020 2020 2020 2074 7265 652e  y:.        tree.
-00008cc0: 7265 6e64 6572 2822 7b7d 2f72 6574 7572  render("{}/retur
-00008cd0: 6e5f 7b7d 2e70 6e67 222e 666f 726d 6174  n_{}.png".format
-00008ce0: 2873 746f 7261 6765 5f66 6f6c 6465 722c  (storage_folder,
-00008cf0: 6e61 6d65 5f66 696c 6529 2c77 3d31 3030  name_file),w=100
-00008d00: 302c 2075 6e69 7473 3d22 6d6d 222c 7472  0, units="mm",tr
-00008d10: 6565 5f73 7479 6c65 3d74 7329 0a20 2020  ee_style=ts).   
-00008d20: 2065 7863 6570 743a 0a20 2020 2020 2020   except:.       
-00008d30: 2074 7265 652e 7265 6e64 6572 2822 7b7d   tree.render("{}
-00008d40: 2f72 6574 7572 6e5f 7b7d 2e70 6e67 222e  /return_{}.png".
-00008d50: 666f 726d 6174 2873 746f 7261 6765 5f66  format(storage_f
-00008d60: 6f6c 6465 722c 6e61 6d65 5f66 696c 6529  older,name_file)
-00008d70: 2c20 773d 3130 3030 2c20 756e 6974 733d  , w=1000, units=
-00008d80: 226d 6d22 290a 6465 6620 7265 6e61 6d65  "mm").def rename
-00008d90: 5f74 7265 655f 696e 7465 726e 616c 5f6e  _tree_internal_n
-00008da0: 6f64 6573 5f73 696d 756c 6174 696f 6e73  odes_simulations
-00008db0: 2874 7265 652c 7769 7468 5f69 6e64 6578  (tree,with_index
-00008dc0: 6573 3d46 616c 7365 293a 0a20 2020 2022  es=False):.    "
-00008dd0: 2222 5265 6e61 6d65 2074 6865 2069 6e74  ""Rename the int
-00008de0: 6572 6e61 6c20 6e6f 6465 7320 6f66 2061  ernal nodes of a
-00008df0: 6e20 6574 6533 2074 7265 6520 746f 2061  n ete3 tree to a
-00008e00: 206c 6162 656c 2049 202b 206e 756d 6265   label I + numbe
-00008e10: 722c 2069 6e20 7369 6d75 6c61 7469 6f6e  r, in simulation
-00008e20: 7320 7468 6520 6c65 6176 6573 2068 6176  s the leaves hav
-00008e30: 6520 7468 6520 7072 6566 6978 2041 2e20  e the prefix A. 
-00008e40: 5769 7468 2069 6e64 6578 6573 2073 686f  With indexes sho
-00008e50: 7773 2074 6865 206e 616d 6573 2075 7365  ws the names use
-00008e60: 6420 7768 656e 2074 7261 6e73 6665 7272  d when transferr
-00008e70: 696e 6720 746f 2061 6e20 6172 7261 790a  ing to an array.
-00008e80: 2020 2020 3a70 6172 616d 2065 7465 332d      :param ete3-
-00008e90: 7472 6565 2074 7265 653a 2045 7465 3320  tree tree: Ete3 
-00008ea0: 7472 6565 0a20 2020 203a 7061 7261 6d20  tree.    :param 
-00008eb0: 626f 6f6c 2077 6974 685f 696e 6465 7865  bool with_indexe
-00008ec0: 733a 2054 7275 6520 2d2d 3e20 6164 6473  s: True --> adds
-00008ed0: 2074 6865 2074 7265 6520 6c65 7665 6c20   the tree level 
-00008ee0: 6f72 6465 7220 696e 6465 7865 7320 746f  order indexes to
-00008ef0: 2074 6865 206e 616d 650a 2020 2020 2222   the name.    ""
-00008f00: 220a 2020 2020 7072 696e 7428 2252 656e  ".    print("Ren
-00008f10: 616d 696e 6720 7472 6565 2066 726f 6d20  aming tree from 
-00008f20: 7369 6d75 6c61 7469 6f6e 7322 290a 2020  simulations").  
-00008f30: 2020 6c65 6166 735f 6e61 6d65 7320 3d20    leafs_names = 
-00008f40: 7472 6565 2e67 6574 5f6c 6561 665f 6e61  tree.get_leaf_na
-00008f50: 6d65 7328 290a 2020 2020 6564 6765 203d  mes().    edge =
-00008f60: 206c 656e 286c 6561 6673 5f6e 616d 6573   len(leafs_names
-00008f70: 290a 2020 2020 696e 7465 726e 616c 5f6e  ).    internal_n
-00008f80: 6f64 6573 5f6e 616d 6573 203d 205b 5d0a  odes_names = [].
-00008f90: 2020 2020 6966 2077 6974 685f 696e 6465      if with_inde
-00008fa0: 7865 733a 0a20 2020 2020 2020 2066 6f72  xes:.        for
-00008fb0: 2069 6478 2c6e 6f64 6520 696e 2065 6e75   idx,node in enu
-00008fc0: 6d65 7261 7465 2874 7265 652e 7472 6176  merate(tree.trav
-00008fd0: 6572 7365 2829 293a 2020 2320 6c65 7665  erse()):  # leve
-00008fe0: 6c6f 7264 6572 2028 6e6f 6465 7320 6172  lorder (nodes ar
-00008ff0: 6520 7669 7369 7465 6420 696e 207a 6967  e visited in zig
-00009000: 207a 6167 206f 7264 6572 2066 726f 6d20   zag order from 
-00009010: 726f 6f74 2074 6f20 6c65 6176 6573 290a  root to leaves).
-00009020: 2020 2020 2020 2020 2020 2020 6966 206e              if n
-00009030: 6f74 206e 6f64 652e 6973 5f6c 6561 6628  ot node.is_leaf(
-00009040: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
-00009050: 2020 206e 6f64 652e 6e61 6d65 203d 2022     node.name = "
-00009060: 4922 202b 206e 6f64 652e 6e61 6d65 202b  I" + node.name +
-00009070: 2022 2f7b 7d22 2e66 6f72 6d61 7428 6964   "/{}".format(id
-00009080: 7829 0a20 2020 2020 2020 2020 2020 2020  x).             
-00009090: 2020 2069 6e74 6572 6e61 6c5f 6e6f 6465     internal_node
-000090a0: 735f 6e61 6d65 732e 6170 7065 6e64 286e  s_names.append(n
-000090b0: 6f64 652e 6e61 6d65 290a 2020 2020 2020  ode.name).      
-000090c0: 2020 2020 2020 2020 2020 6564 6765 202b            edge +
-000090d0: 3d20 310a 2020 2020 2020 2020 2020 2020  = 1.            
-000090e0: 656c 7365 3a0a 2020 2020 2020 2020 2020  else:.          
-000090f0: 2020 2020 2020 6e6f 6465 2e6e 616d 6520        node.name 
-00009100: 3d20 6e6f 6465 2e6e 616d 6520 2b20 222f  = node.name + "/
-00009110: 7b7d 222e 666f 726d 6174 2869 6478 290a  {}".format(idx).
-00009120: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00009130: 2365 6467 6520 2b3d 2031 0a20 2020 2065  #edge += 1.    e
-00009140: 6c73 653a 0a20 2020 2020 2020 2066 6f72  lse:.        for
-00009150: 206e 6f64 6520 696e 2074 7265 652e 7472   node in tree.tr
-00009160: 6176 6572 7365 2829 3a20 2023 206c 6576  averse():  # lev
-00009170: 656c 6f72 6465 7220 286e 6f64 6573 2061  elorder (nodes a
-00009180: 7265 2076 6973 6974 6564 2069 6e20 7a69  re visited in zi
-00009190: 6720 7a61 6720 6f72 6465 7220 6672 6f6d  g zag order from
-000091a0: 2072 6f6f 7420 746f 206c 6561 7665 7329   root to leaves)
-000091b0: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
-000091c0: 6e6f 7420 6e6f 6465 2e69 735f 6c65 6166  not node.is_leaf
-000091d0: 2829 3a0a 2020 2020 2020 2020 2020 2020  ():.            
-000091e0: 2020 2020 6e6f 6465 2e6e 616d 6520 3d20      node.name = 
-000091f0: 2249 2220 2b20 6e6f 6465 2e6e 616d 650a  "I" + node.name.
-00009200: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00009210: 696e 7465 726e 616c 5f6e 6f64 6573 5f6e  internal_nodes_n
-00009220: 616d 6573 2e61 7070 656e 6428 6e6f 6465  ames.append(node
-00009230: 2e6e 616d 6529 0a20 2020 2020 2020 2020  .name).         
-00009240: 2020 2020 2020 2065 6467 6520 2b3d 2031         edge += 1
-00009250: 0a20 2020 2072 6574 7572 6e20 7472 6565  .    return tree
-00009260: 0a64 6566 2072 656e 616d 655f 7472 6565  .def rename_tree
-00009270: 5f69 6e74 6572 6e61 6c5f 6e6f 6465 7328  _internal_nodes(
-00009280: 7472 6565 293a 0a20 2020 2022 2222 5265  tree):.    """Re
-00009290: 6e61 6d65 2074 6865 2069 6e74 6572 6e61  name the interna
-000092a0: 6c20 6e6f 6465 7320 6f66 2061 2074 7265  l nodes of a tre
-000092b0: 6520 746f 2061 206c 6162 656c 2041 202b  e to a label A +
-000092c0: 206e 756d 6265 722c 2075 6e6c 6573 7320   number, unless 
-000092d0: 7468 6520 6769 7665 6e20 6e65 7769 636b  the given newick
-000092e0: 2066 696c 6520 616c 7265 6164 7920 6861   file already ha
-000092f0: 7320 7468 6520 6e61 6d65 7320 6f6e 2069  s the names on i
-00009300: 740a 2020 2020 3a70 6172 616d 2065 7465  t.    :param ete
-00009310: 332d 7472 6565 2074 7265 653a 2045 7465  3-tree tree: Ete
-00009320: 3320 7472 6565 2063 6f6e 7374 7275 6374  3 tree construct
-00009330: 6f72 2222 220a 0a20 2020 2023 5265 6e61  or"""..    #Rena
-00009340: 6d65 2074 6865 2069 6e74 6572 6e61 6c20  me the internal 
-00009350: 6e6f 6465 730a 2020 2020 6c65 6166 735f  nodes.    leafs_
-00009360: 6e61 6d65 7320 3d20 7472 6565 2e67 6574  names = tree.get
-00009370: 5f6c 6561 665f 6e61 6d65 7328 290a 2020  _leaf_names().  
-00009380: 2020 6564 6765 203d 206c 656e 286c 6561    edge = len(lea
-00009390: 6673 5f6e 616d 6573 290a 2020 2020 696e  fs_names).    in
-000093a0: 7465 726e 616c 5f6e 6f64 6573 5f6e 616d  ternal_nodes_nam
-000093b0: 6573 203d 205b 5d0a 2020 2020 666f 7220  es = [].    for 
-000093c0: 6e6f 6465 2069 6e20 7472 6565 2e74 7261  node in tree.tra
-000093d0: 7665 7273 6528 293a 2023 6c65 7665 6c6f  verse(): #levelo
-000093e0: 7264 6572 2028 6e6f 6465 7320 6172 6520  rder (nodes are 
-000093f0: 7669 7369 7465 6420 696e 207a 6967 207a  visited in zig z
-00009400: 6167 206f 7264 6572 2066 726f 6d20 726f  ag order from ro
-00009410: 6f74 2074 6f20 6c65 6176 6573 290a 2020  ot to leaves).  
-00009420: 2020 2020 2020 6966 206e 6f74 206e 6f64        if not nod
-00009430: 652e 6973 5f6c 6561 6628 293a 0a20 2020  e.is_leaf():.   
-00009440: 2020 2020 2020 2020 206e 6f64 652e 6e61           node.na
-00009450: 6d65 203d 2022 4125 6422 2025 2065 6467  me = "A%d" % edg
-00009460: 650a 2020 2020 2020 2020 2020 2020 696e  e.            in
-00009470: 7465 726e 616c 5f6e 6f64 6573 5f6e 616d  ternal_nodes_nam
-00009480: 6573 2e61 7070 656e 6428 6e6f 6465 2e6e  es.append(node.n
-00009490: 616d 6529 0a20 2020 2020 2020 2020 2020  ame).           
-000094a0: 2065 6467 6520 2b3d 2031 0a20 2020 2072   edge += 1.    r
-000094b0: 6574 7572 6e20 7472 6565 0a64 6566 2070  eturn tree.def p
-000094c0: 726f 6365 7373 5f70 6462 5f66 696c 6573  rocess_pdb_files
-000094d0: 2850 4442 5f66 6f6c 6465 722c 6161 5f70  (PDB_folder,aa_p
-000094e0: 726f 6273 2c70 6661 6d5f 6469 6374 2c6f  robs,pfam_dict,o
-000094f0: 6e65 5f68 6f74 5f65 6e63 6f64 696e 672c  ne_hot_encoding,
-00009500: 6d69 6e5f 6c65 6e29 3a0a 2020 2020 2222  min_len):.    ""
-00009510: 2222 4578 7472 6163 7420 696e 666f 726d  ""Extract inform
-00009520: 6174 696f 6e20 6672 6f6d 2074 6865 2050  ation from the P
-00009530: 4442 2066 696c 6573 0a20 2020 203a 7061  DB files.    :pa
-00009540: 7261 6d20 7374 7220 5044 425f 666f 6c64  ram str PDB_fold
-00009550: 6572 3a20 7061 7468 2074 6f20 666f 6c64  er: path to fold
-00009560: 6572 2063 6f6e 7461 696e 696e 6720 7468  er containing th
-00009570: 6520 5046 4220 6669 6c65 730a 2020 2020  e PFB files.    
-00009580: 3a70 6172 616d 2069 6e74 2061 615f 7072  :param int aa_pr
-00009590: 6f62 733a 2061 6d69 6e6f 2061 6369 6420  obs: amino acid 
-000095a0: 7072 6f62 6162 696c 6974 6965 730a 2020  probabilities.  
-000095b0: 2020 3a70 6172 616d 2064 6963 7420 7066    :param dict pf
-000095c0: 616d 5f64 6963 743a 2064 6963 7469 6f6e  am_dict: diction
-000095d0: 6172 7920 636f 6e74 6169 6e69 6e67 2074  ary containing t
-000095e0: 6865 2069 6e66 6f72 6d61 7469 6f6e 206f  he information o
-000095f0: 6e20 686f 7720 746f 2065 7874 7261 6374  n how to extract
-00009600: 2074 6865 2073 6571 7565 6e63 6573 2066   the sequences f
-00009610: 726f 6d20 7468 6520 7064 6220 6669 6c65  rom the pdb file
-00009620: 0a20 2020 203a 7061 7261 6d20 626f 6f6c  .    :param bool
-00009630: 206f 6e65 5f68 6f74 5f65 6e63 6f64 696e   one_hot_encodin
-00009640: 670a 2020 2020 3a70 6172 616d 2069 6e74  g.    :param int
-00009650: 206d 696e 5f6c 656e 3a20 4c6f 7765 7220   min_len: Lower 
-00009660: 626f 756e 6420 6f6e 2074 6865 2073 6571  bound on the seq
-00009670: 7565 6e63 6520 6c65 6e67 7468 0a20 2020  uence length.   
-00009680: 2022 2222 0a20 2020 2070 6172 7365 7220   """.    parser 
-00009690: 3d20 5044 422e 5044 4250 6172 7365 7228  = PDB.PDBParser(
-000096a0: 290a 2020 2020 2370 7062 203d 2050 5042  ).    #ppb = PPB
-000096b0: 7569 6c64 6572 2829 0a20 2020 2063 6170  uilder().    cap
-000096c0: 7020 3d20 4361 5050 4275 696c 6465 7228  p = CaPPBuilder(
-000096d0: 290a 2020 2020 7072 6f74 5f69 6e66 6f5f  ).    prot_info_
-000096e0: 6469 6374 203d 207b 7d0a 2020 2020 7072  dict = {}.    pr
-000096f0: 6f74 5f61 615f 6469 6374 203d 207b 7d0a  ot_aa_dict = {}.
-00009700: 2020 2020 616d 696e 6f61 6369 645f 6e61      aminoacid_na
-00009710: 6d65 7320 3d20 616d 696e 6f61 6369 645f  mes = aminoacid_
-00009720: 6e61 6d65 735f 6469 6374 2861 615f 7072  names_dict(aa_pr
-00009730: 6f62 7329 2020 2320 544f 444f 3a20 6765  obs)  # TODO: ge
-00009740: 6e65 7261 6c69 7a65 2074 6f20 696e 636c  neralize to incl
-00009750: 7564 6520 616c 6c20 7479 7065 7320 6f66  ude all types of
-00009760: 2061 6d69 6e6f 2061 6369 6473 2e20 466f   amino acids. Fo
-00009770: 7220 6f6e 6568 6f74 2065 6e63 6f64 696e  r onehot encodin
-00009780: 6773 2073 696d 706c 7920 7573 6520 6161  gs simply use aa
-00009790: 5f6e 616d 6573 5f64 6963 742e 6b65 7973  _names_dict.keys
-000097a0: 2829 0a20 2020 2066 696c 6573 5f6c 6973  ().    files_lis
-000097b0: 7420 3d20 5b66 2066 6f72 2066 2069 6e20  t = [f for f in 
-000097c0: 6c69 7374 6469 7228 5044 425f 666f 6c64  listdir(PDB_fold
-000097d0: 6572 2920 6966 2069 7366 696c 6528 6a6f  er) if isfile(jo
-000097e0: 696e 2850 4442 5f66 6f6c 6465 722c 2066  in(PDB_folder, f
-000097f0: 2929 5d0a 2020 2020 6475 706c 6963 6174  ))].    duplicat
-00009800: 6573 203d 205b 5d0a 2020 2020 666f 7220  es = [].    for 
-00009810: 692c 2050 4442 5f66 696c 6520 696e 2065  i, PDB_file in e
-00009820: 6e75 6d65 7261 7465 2866 696c 6573 5f6c  numerate(files_l
-00009830: 6973 7429 3a0a 2020 2020 2020 2020 7374  ist):.        st
-00009840: 7275 6374 7572 6520 3d20 7061 7273 6572  ructure = parser
-00009850: 2e67 6574 5f73 7472 7563 7475 7265 2827  .get_structure('
-00009860: 7b7d 272e 666f 726d 6174 2850 4442 5f66  {}'.format(PDB_f
-00009870: 696c 6529 2c20 6a6f 696e 2850 4442 5f66  ile), join(PDB_f
-00009880: 6f6c 6465 722c 2050 4442 5f66 696c 6529  older, PDB_file)
-00009890: 290a 2020 2020 2020 2020 6966 2070 6661  ).        if pfa
-000098a0: 6d5f 6469 6374 3a20 2023 2063 6f6e 7461  m_dict:  # conta
-000098b0: 696e 7320 696e 666f 726d 6174 696f 6e20  ins information 
-000098c0: 6f6e 2077 6869 6368 2063 6861 696e 2074  on which chain t
-000098d0: 6f20 7461 6b65 2061 6e64 2077 6869 6368  o take and which
-000098e0: 2072 6573 6964 7565 7320 746f 2073 656c   residues to sel
-000098f0: 6563 7420 6672 6f6d 2074 6865 2050 4442  ect from the PDB
-00009900: 2066 696c 650a 2020 2020 2020 2020 2020   file.          
-00009910: 2020 6368 6169 6e5f 6e61 6d65 203d 2070    chain_name = p
-00009920: 6661 6d5f 6469 6374 5b50 4442 5f66 696c  fam_dict[PDB_fil
-00009930: 655b 333a 375d 2e75 7070 6572 2829 5d5b  e[3:7].upper()][
-00009940: 2243 6861 696e 225d 0a20 2020 2020 2020  "Chain"].       
-00009950: 2020 2020 2043 6861 696e 5f30 203d 2073       Chain_0 = s
-00009960: 7472 7563 7475 7265 5b30 5d5b 6368 6169  tructure[0][chai
-00009970: 6e5f 6e61 6d65 5d0a 2020 2020 2020 2020  n_name].        
-00009980: 2020 2020 7374 6172 745f 7265 7369 6475      start_residu
-00009990: 652c 2065 6e64 5f72 6573 6964 7565 203d  e, end_residue =
-000099a0: 2070 6661 6d5f 6469 6374 5b50 4442 5f66   pfam_dict[PDB_f
-000099b0: 696c 655b 333a 375d 2e75 7070 6572 2829  ile[3:7].upper()
-000099c0: 5d5b 2252 6573 6964 7565 7322 5d2e 7370  ]["Residues"].sp
-000099d0: 6c69 7428 222d 2229 0a20 2020 2020 2020  lit("-").       
-000099e0: 2020 2020 206c 6973 745f 7265 7369 6475       list_residu
-000099f0: 6573 203d 206c 6973 7428 7261 6e67 6528  es = list(range(
-00009a00: 696e 7428 7374 6172 745f 7265 7369 6475  int(start_residu
-00009a10: 6529 2c20 696e 7428 656e 645f 7265 7369  e), int(end_resi
-00009a20: 6475 6529 202b 2031 2929 0a20 2020 2020  due) + 1)).     
-00009a30: 2020 2065 6c73 653a 2020 2320 7768 656e     else:  # when
-00009a40: 2074 6865 7265 2069 7320 6e6f 7420 496e   there is not In
-00009a50: 666f 726d 6174 696f 6e20 6e6f 7420 6176  formation not av
-00009a60: 6169 6c61 626c 6520 6f6e 2077 6869 6368  ailable on which
-00009a70: 2063 6861 696e 7320 746f 2070 6963 6b0a   chains to pick.
-00009a80: 2020 2020 2020 2020 2020 2020 6368 6169              chai
-00009a90: 6e73 203d 205b 6368 6169 6e20 666f 7220  ns = [chain for 
-00009aa0: 6368 6169 6e20 696e 2073 7472 7563 7475  chain in structu
-00009ab0: 7265 5b30 5d5d 0a20 2020 2020 2020 2020  re[0]].         
-00009ac0: 2020 2043 6861 696e 5f30 203d 2063 6861     Chain_0 = cha
-00009ad0: 696e 735b 305d 0a20 2020 2020 2020 2070  ins[0].        p
-00009ae0: 6f6c 7970 6570 7469 6465 7320 3d20 6361  olypeptides = ca
-00009af0: 7070 2e62 7569 6c64 5f70 6570 7469 6465  pp.build_peptide
-00009b00: 7328 4368 6169 6e5f 3029 2020 2320 435f  s(Chain_0)  # C_
-00009b10: 616c 7068 612d 432d 616c 7068 6120 706f  alpha-C-alpha po
-00009b20: 6c79 7065 7074 6964 650a 2020 2020 2020  lypeptide.      
-00009b30: 2020 616e 676c 6573 5f6c 6973 7420 3d20    angles_list = 
-00009b40: 5b5d 0a20 2020 2020 2020 2061 615f 6c69  [].        aa_li
-00009b50: 7374 5f65 6d62 6564 6465 6420 3d20 5b5d  st_embedded = []
-00009b60: 0a20 2020 2020 2020 2061 615f 6c69 7374  .        aa_list
-00009b70: 203d 205b 5d0a 2020 2020 2020 2020 2320   = [].        # 
-00009b80: 636f 6f72 6469 6e61 7465 735f 6c69 7374  coordinates_list
-00009b90: 203d 205b 5d0a 2020 2020 2020 2020 666f   = [].        fo
-00009ba0: 7220 706f 6c79 5f69 6e64 6578 2c20 706f  r poly_index, po
-00009bb0: 6c79 2069 6e20 656e 756d 6572 6174 6528  ly in enumerate(
-00009bc0: 706f 6c79 7065 7074 6964 6573 293a 0a20  polypeptides):. 
-00009bd0: 2020 2020 2020 2020 2020 2069 6620 6e6f             if no
-00009be0: 7420 7066 616d 5f64 6963 743a 0a20 2020  t pfam_dict:.   
-00009bf0: 2020 2020 2020 2020 2020 2020 206c 6973               lis
-00009c00: 745f 7265 7369 6475 6573 203d 205b 7265  t_residues = [re
-00009c10: 7369 6475 652e 6765 745f 6964 2829 5b31  sidue.get_id()[1
-00009c20: 5d20 666f 7220 7265 7369 6475 6520 696e  ] for residue in
-00009c30: 2070 6f6c 795d 2020 2320 544f 444f 3a20   poly]  # TODO: 
-00009c40: 4368 616e 6765 2074 6f20 2e67 6574 5f72  Change to .get_r
-00009c50: 6573 6e61 6d65 2829 3f0a 2020 2020 2020  esname()?.      
-00009c60: 2020 2020 2020 636f 7272 6573 706f 6e64        correspond
-00009c70: 656e 6365 5f69 6e64 6578 203d 205b 696e  ence_index = [in
-00009c80: 6465 7820 666f 7220 696e 6465 782c 2072  dex for index, r
-00009c90: 6573 6964 7565 2069 6e20 656e 756d 6572  esidue in enumer
-00009ca0: 6174 6528 706f 6c79 2920 6966 2072 6573  ate(poly) if res
-00009cb0: 6964 7565 2e67 6574 5f69 6428 295b 315d  idue.get_id()[1]
-00009cc0: 2069 6e20 6c69 7374 5f72 6573 6964 7565   in list_residue
-00009cd0: 735d 0a20 2020 2020 2020 2020 2020 2070  s].            p
-00009ce0: 6869 7073 695f 6c69 7374 203d 2070 6f6c  hipsi_list = pol
-00009cf0: 792e 6765 745f 7068 695f 7073 695f 6c69  y.get_phi_psi_li
-00009d00: 7374 2829 0a20 2020 2020 2020 2020 2020  st().           
-00009d10: 2023 2047 6f74 7461 2063 686f 7020 616c   # Gotta chop al
-00009d20: 736f 2074 6865 2061 6e67 6c65 7320 6c69  so the angles li
-00009d30: 7374 2061 6363 6f72 6469 6e67 2074 6f20  st according to 
-00009d40: 7468 6520 7261 6e67 6520 6f66 2064 6573  the range of des
-00009d50: 6972 6564 2072 6573 6964 7565 730a 2020  ired residues.  
-00009d60: 2020 2020 2020 2020 2020 6966 2063 6f72            if cor
-00009d70: 7265 7370 6f6e 6465 6e63 655f 696e 6465  respondence_inde
-00009d80: 7820 616e 6420 7066 616d 5f64 6963 743a  x and pfam_dict:
-00009d90: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00009da0: 2061 6e67 6c65 735f 6c69 7374 202b 3d20   angles_list += 
-00009db0: 7068 6970 7369 5f6c 6973 745b 636f 7272  phipsi_list[corr
-00009dc0: 6573 706f 6e64 656e 6365 5f69 6e64 6578  espondence_index
-00009dd0: 5b30 5d3a 636f 7272 6573 706f 6e64 656e  [0]:corresponden
-00009de0: 6365 5f69 6e64 6578 5b0a 2020 2020 2020  ce_index[.      
+00005dd0: 2020 2073 7562 7072 6f63 6573 732e 7275     subprocess.ru
+00005de0: 6e28 6172 6773 3d5b 2269 7174 7265 6522  n(args=["iqtree"
+00005df0: 2c22 2d73 222c 616c 6967 6e6d 656e 745f  ,"-s",alignment_
+00005e00: 662e 7370 6c69 7428 222e 2229 5b30 5d2c  f.split(".")[0],
+00005e10: 222d 2d61 6c6e 222c 616c 6967 6e6d 656e  "--aln",alignmen
+00005e20: 745f 662c 222d 6e74 222c 2241 5554 4f22  t_f,"-nt","AUTO"
+00005e30: 2c22 2d6f 222c 726f 6f74 5d2c 7374 6465  ,"-o",root],stde
+00005e40: 7272 3d73 7973 2e73 7464 6572 722c 2073  rr=sys.stderr, s
+00005e50: 7464 6f75 743d 7379 732e 7374 646f 7574  tdout=sys.stdout
+00005e60: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+00005e70: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
+00005e80: 2020 2020 2020 2020 2020 2020 7375 6270              subp
+00005e90: 726f 6365 7373 2e72 756e 2861 7267 733d  rocess.run(args=
+00005ea0: 5b22 6971 7472 6565 222c 222d 7322 2c61  ["iqtree","-s",a
+00005eb0: 6c69 676e 6d65 6e74 5f66 2e73 706c 6974  lignment_f.split
+00005ec0: 2822 2e22 295b 305d 2c22 2d2d 616c 6e22  (".")[0],"--aln"
+00005ed0: 2c61 6c69 676e 6d65 6e74 5f66 2c22 2d6e  ,alignment_f,"-n
+00005ee0: 7422 2c22 4155 544f 225d 2c73 7464 6572  t","AUTO"],stder
+00005ef0: 723d 7379 732e 7374 6465 7272 2c20 7374  r=sys.stderr, st
+00005f00: 646f 7574 3d73 7973 2e73 7464 6f75 7429  dout=sys.stdout)
+00005f10: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00005f20: 206f 732e 7265 6d6f 7665 2861 6c69 676e   os.remove(align
+00005f30: 6d65 6e74 5f66 202b 2022 2e6c 6f67 2229  ment_f + ".log")
+00005f40: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00005f50: 206f 732e 7265 6d6f 7665 2861 6c69 676e   os.remove(align
+00005f60: 6d65 6e74 5f66 202b 2022 2e62 696f 6e6a  ment_f + ".bionj
+00005f70: 2229 0a20 2020 2020 2020 2020 2020 2020  ").             
+00005f80: 2020 206f 732e 7265 6d6f 7665 2861 6c69     os.remove(ali
+00005f90: 676e 6d65 6e74 5f66 202b 2022 2e63 6b70  gnment_f + ".ckp
+00005fa0: 2e67 7a22 290a 2020 2020 2020 2020 2020  .gz").          
+00005fb0: 2020 2020 2020 6f73 2e72 656d 6f76 6528        os.remove(
+00005fc0: 616c 6967 6e6d 656e 745f 6620 2b20 222e  alignment_f + ".
+00005fd0: 6d6f 6465 6c2e 677a 2229 0a20 2020 2020  model.gz").     
+00005fe0: 2020 2020 2020 2064 6973 7461 6e63 655f         distance_
+00005ff0: 6d61 7472 6978 5f63 616c 203d 2070 642e  matrix_cal = pd.
+00006000: 7265 6164 5f63 7376 2861 6c69 676e 6d65  read_csv(alignme
+00006010: 6e74 5f66 2b22 2e6d 6c64 6973 7422 2c20  nt_f+".mldist", 
+00006020: 7365 703d 225c 5c73 2b22 2c20 736b 6970  sep="\\s+", skip
+00006030: 726f 7773 3d31 2c20 6865 6164 6572 3d4e  rows=1, header=N
+00006040: 6f6e 6529 0a20 2020 2020 2020 2020 2020  one).           
+00006050: 2064 6973 7461 6e63 655f 6d61 7472 6978   distance_matrix
+00006060: 5f63 616c 2e63 6f6c 756d 6e73 203d 205b  _cal.columns = [
+00006070: 2272 6f77 7322 5d20 2b20 6469 7374 616e  "rows"] + distan
+00006080: 6365 5f6d 6174 7269 785f 6361 6c2e 696c  ce_matrix_cal.il
+00006090: 6f63 5b3a 2c30 5d2e 746f 5f6c 6973 7428  oc[:,0].to_list(
+000060a0: 290a 2020 2020 2020 2020 2020 2020 6469  ).            di
+000060b0: 7374 616e 6365 5f6d 6174 7269 785f 6361  stance_matrix_ca
+000060c0: 6c2e 7365 745f 696e 6465 7828 2272 6f77  l.set_index("row
+000060d0: 7322 2c20 696e 706c 6163 653d 5472 7565  s", inplace=True
+000060e0: 290a 2020 2020 2020 2020 2020 2020 6469  ).            di
+000060f0: 7374 616e 6365 5f6d 6174 7269 785f 6361  stance_matrix_ca
+00006100: 6c2e 696e 6465 782e 6e61 6d65 203d 2022  l.index.name = "
+00006110: 220a 2020 2020 2020 2020 2020 2020 6469  ".            di
+00006120: 7374 616e 6365 5f6d 6174 7269 785f 6361  stance_matrix_ca
+00006130: 6c2e 746f 5f63 7376 2822 7b7d 2f7b 7d5f  l.to_csv("{}/{}_
+00006140: 6469 7374 616e 6365 5f6d 6174 7269 782e  distance_matrix.
+00006150: 6373 7622 2e66 6f72 6d61 7428 7374 6f72  csv".format(stor
+00006160: 6167 655f 666f 6c64 6572 2c6e 616d 6529  age_folder,name)
+00006170: 290a 2020 2020 2020 2020 2020 2020 7472  ).            tr
+00006180: 6565 203d 2054 7265 6545 7465 3328 616c  ee = TreeEte3(al
+00006190: 6967 6e6d 656e 745f 662b 2022 2e74 7265  ignment_f+ ".tre
+000061a0: 6566 696c 6522 290a 2020 2020 2020 2020  efile").        
+000061b0: 2020 2020 7265 7475 726e 2074 7265 650a      return tree.
+000061c0: 2020 2020 2020 2020 656c 6966 206d 6574          elif met
+000061d0: 686f 6420 3d3d 2022 7261 7069 646e 6a22  hod == "rapidnj"
+000061e0: 3a0a 2020 2020 2020 2020 2020 2020 7072  :.            pr
+000061f0: 696e 7428 2255 7369 6e67 2052 6170 6964  int("Using Rapid
+00006200: 6e6a 2074 6f20 6275 696c 6420 4e4f 5420  nj to build NOT 
+00006210: 726f 6f74 6564 2074 7265 652e 2e2e 2229  rooted tree...")
+00006220: 0a20 2020 2020 2020 2020 2020 2074 7265  .            tre
+00006230: 655f 6669 6c65 5f6e 616d 6520 3d20 5b22  e_file_name = ["
+00006240: 7b7d 2f7b 7d2e 7472 6565 222e 666f 726d  {}/{}.tree".form
+00006250: 6174 2873 746f 7261 6765 5f66 6f6c 6465  at(storage_folde
+00006260: 722c 6e61 6d65 2920 6966 206e 6f74 2074  r,name) if not t
+00006270: 7265 655f 6669 6c65 5f6e 616d 6520 656c  ree_file_name el
+00006280: 7365 2074 7265 655f 6669 6c65 5f6e 616d  se tree_file_nam
+00006290: 655d 5b30 5d0a 2020 2020 2020 2020 2020  e][0].          
+000062a0: 2020 616c 6967 6e6d 656e 745f 6620 3d20    alignment_f = 
+000062b0: 5b61 6c69 676e 6d65 6e74 5f66 696c 655f  [alignment_file_
+000062c0: 6e61 6d65 2069 6620 616c 6967 6e6d 656e  name if alignmen
+000062d0: 745f 6669 6c65 5f6e 616d 6520 656c 7365  t_file_name else
+000062e0: 2022 7b7d 2f7b 7d2e 6d61 6666 7422 2e66   "{}/{}.mafft".f
+000062f0: 6f72 6d61 7428 7374 6f72 6167 655f 666f  ormat(storage_fo
+00006300: 6c64 6572 2c6e 616d 6529 5d5b 305d 0a20  lder,name)][0]. 
+00006310: 2020 2020 2020 2020 2020 2077 6974 6820             with 
+00006320: 6f70 656e 2874 7265 655f 6669 6c65 5f6e  open(tree_file_n
+00006330: 616d 652c 2022 7722 2920 6173 2074 7265  ame, "w") as tre
+00006340: 655f 6669 6c65 5f6f 7574 3a0a 2020 2020  e_file_out:.    
+00006350: 2020 2020 2020 2020 2020 2020 7375 6270              subp
+00006360: 726f 6365 7373 2e72 756e 2861 7267 733d  rocess.run(args=
+00006370: 5b22 7261 7069 646e 6a22 2c61 6c69 676e  ["rapidnj",align
+00006380: 6d65 6e74 5f66 2c20 222d 6922 2c20 2266  ment_f, "-i", "f
+00006390: 6122 5d2c 2073 7464 6f75 743d 7472 6565  a"], stdout=tree
+000063a0: 5f66 696c 655f 6f75 7429 0a20 2020 2020  _file_out).     
+000063b0: 2020 2020 2020 2074 7265 655f 6669 6c65         tree_file
+000063c0: 5f6f 7574 2e63 6c6f 7365 2829 0a20 2020  _out.close().   
+000063d0: 2020 2020 2020 2020 2074 7265 6520 3d20           tree = 
+000063e0: 5472 6565 4574 6533 2874 7265 655f 6669  TreeEte3(tree_fi
+000063f0: 6c65 5f6e 616d 6529 0a20 2020 2020 2020  le_name).       
+00006400: 2020 2020 2072 6574 7572 6e20 7472 6565       return tree
+00006410: 0a64 6566 2069 6e66 6572 5f61 6c69 676e  .def infer_align
+00006420: 6d65 6e74 2861 6c69 676e 6d65 6e74 5f66  ment(alignment_f
+00006430: 696c 652c 696e 7075 745f 6e61 6d65 5f66  ile,input_name_f
+00006440: 696c 652c 6f75 7470 7574 5f6e 616d 655f  ile,output_name_
+00006450: 6669 6c65 293a 0a20 2020 2022 2222 0a20  file):.    """. 
+00006460: 2020 2052 6561 6473 2061 6e64 2061 6c69     Reads and ali
+00006470: 676e 6d65 6e74 206f 7220 7065 7266 6f72  gnment or perfor
+00006480: 6d73 2061 6c69 676e 6d65 6e74 2075 7369  ms alignment usi
+00006490: 6e67 204d 4146 4654 205b 4d41 4646 5420  ng MAFFT [MAFFT 
+000064a0: 6d75 6c74 6970 6c65 2073 6571 7565 6e63  multiple sequenc
+000064b0: 6520 616c 6967 6e6d 656e 7420 736f 6674  e alignment soft
+000064c0: 7761 7265 2076 6572 7369 6f6e 2037 3a20  ware version 7: 
+000064d0: 696d 7072 6f76 656d 656e 7473 2069 6e20  improvements in 
+000064e0: 7065 7266 6f72 6d61 6e63 6520 616e 6420  performance and 
+000064f0: 7573 6162 696c 6974 795d 2e20 5265 7475  usability]. Retu
+00006500: 726e 7320 6120 6469 6374 696f 6e61 7279  rns a dictionary
+00006510: 2077 6974 6820 7468 6520 7365 7175 656e   with the sequen
+00006520: 6365 206e 616d 650a 2020 2020 616e 6420  ce name.    and 
+00006530: 7468 6520 7365 7175 656e 6365 2061 6e64  the sequence and
+00006540: 2061 2062 696f 7079 7468 6f6e 2061 6c69   a biopython ali
+00006550: 676e 6d65 6e74 206f 626a 6563 740a 2020  gnment object.  
+00006560: 2020 3a70 6172 616d 2073 7472 206f 7220    :param str or 
+00006570: 4e6f 6e65 2061 6c69 676e 6d65 6e74 5f66  None alignment_f
+00006580: 696c 653a 2070 6174 6820 746f 2070 7265  ile: path to pre
+00006590: 2d63 6f6d 7075 7465 6420 616c 6967 6e6d  -computed alignm
+000065a0: 656e 7420 746f 2072 6561 640a 2020 2020  ent to read.    
+000065b0: 3a70 6172 616d 2073 7472 2069 6e70 7574  :param str input
+000065c0: 5f66 696c 655f 6e61 6d65 3a20 7061 7468  _file_name: path
+000065d0: 2074 6f20 7468 6520 6669 6c65 2063 6f6e   to the file con
+000065e0: 7461 696e 696e 6720 756e 616c 6967 6e65  taining unaligne
+000065f0: 6420 7365 7175 656e 6365 732c 696e 2066  d sequences,in f
+00006600: 6173 7461 2066 6f72 6d61 740a 2020 2020  asta format.    
+00006610: 3a70 6172 616d 2073 7472 206f 7574 7075  :param str outpu
+00006620: 745f 6669 6c65 5f6e 616d 653a 206e 616d  t_file_name: nam
+00006630: 6520 6f66 2074 6865 2066 696c 6520 7468  e of the file th
+00006640: 6174 2077 696c 6c20 636f 6e74 6169 6e20  at will contain 
+00006650: 7468 6520 616c 6967 6e65 6420 7365 7175  the aligned sequ
+00006660: 656e 6365 7322 2222 0a20 2020 2023 2041  ences""".    # A
+00006670: 6c69 676e 2074 6865 2070 6f6c 7970 6570  lign the polypep
+00006680: 7469 6465 732f 7365 7175 656e 6365 7320  tides/sequences 
+00006690: 616e 6420 7772 6974 6520 746f 2061 2066  and write to a f
+000066a0: 6173 7461 2066 696c 650a 2020 2020 7072  asta file.    pr
+000066b0: 696e 7428 2241 6e61 6c79 7a69 6e67 2061  int("Analyzing a
+000066c0: 6c69 676e 6d65 6e74 2e2e 2e22 290a 2020  lignment...").  
+000066d0: 2020 6966 2061 6c69 676e 6d65 6e74 5f66    if alignment_f
+000066e0: 696c 653a 2023 5468 6520 616c 6967 6e6d  ile: #The alignm
+000066f0: 656e 7420 6669 6c65 2073 686f 756c 6420  ent file should 
+00006700: 636f 6e74 6169 6e20 7468 6520 706f 6c79  contain the poly
+00006710: 7065 7074 6964 6573 206f 6620 7468 6520  peptides of the 
+00006720: 5044 4220 7374 7275 6374 7572 6573 2061  PDB structures a
+00006730: 6e64 2073 6571 7565 6e63 6573 2077 6974  nd sequences wit
+00006740: 686f 7574 2073 7472 7563 7475 7265 730a  hout structures.
+00006750: 2020 2020 2020 2020 7072 696e 7428 2252          print("R
+00006760: 6561 6469 6e67 2067 6976 656e 2061 6c69  eading given ali
+00006770: 676e 6d65 6e74 2066 696c 6520 2e2e 2e22  gnment file ..."
+00006780: 290a 2020 2020 2020 2020 2320 5265 6164  ).        # Read
+00006790: 2074 6865 2061 6c69 676e 6564 2073 6571   the aligned seq
+000067a0: 7565 6e63 6573 0a20 2020 2020 2020 2061  uences.        a
+000067b0: 6c69 676e 6d65 6e74 203d 2041 6c69 676e  lignment = Align
+000067c0: 494f 2e72 6561 6428 227b 7d22 2e66 6f72  IO.read("{}".for
+000067d0: 6d61 7428 616c 6967 6e6d 656e 745f 6669  mat(alignment_fi
+000067e0: 6c65 292c 2022 6661 7374 6122 290a 2020  le), "fasta").  
+000067f0: 2020 2020 2020 616c 6967 6e6d 656e 745f        alignment_
+00006800: 6964 7320 3d20 5b5d 0a20 2020 2020 2020  ids = [].       
+00006810: 2061 6c69 676e 6d65 6e74 5f73 6571 7320   alignment_seqs 
+00006820: 3d20 5b5d 0a20 2020 2020 2020 2066 6f72  = [].        for
+00006830: 2069 2c61 6c69 676e 6564 2069 6e20 656e   i,aligned in en
+00006840: 756d 6572 6174 6528 616c 6967 6e6d 656e  umerate(alignmen
+00006850: 7429 3a0a 2020 2020 2020 2020 2020 2020  t):.            
+00006860: 616c 6967 6e6d 656e 745f 6964 732e 6170  alignment_ids.ap
+00006870: 7065 6e64 2861 6c69 676e 6d65 6e74 5b69  pend(alignment[i
+00006880: 5d2e 6964 290a 2020 2020 2020 2020 2020  ].id).          
+00006890: 2020 616c 6967 6e6d 656e 745f 7365 7173    alignment_seqs
+000068a0: 2e61 7070 656e 6428 616c 6967 6e6d 656e  .append(alignmen
+000068b0: 745b 695d 2e73 6571 2e73 7472 6970 2822  t[i].seq.strip("
+000068c0: 2a22 2929 2023 4869 6768 6c69 6768 743a  *")) #Highlight:
+000068d0: 2052 656d 6f76 6520 7374 6f70 2063 6f64   Remove stop cod
+000068e0: 6f6e 730a 2020 2020 2020 2020 6469 6374  ons.        dict
+000068f0: 5f61 6c69 676e 6d65 6e74 203d 2064 6963  _alignment = dic
+00006900: 7428 7a69 7028 616c 6967 6e6d 656e 745f  t(zip(alignment_
+00006910: 6964 732c 2061 6c69 676e 6d65 6e74 5f73  ids, alignment_s
+00006920: 6571 7329 290a 2020 2020 2020 2020 7265  eqs)).        re
+00006930: 7475 726e 2064 6963 745f 616c 6967 6e6d  turn dict_alignm
+00006940: 656e 742c 2061 6c69 676e 6d65 6e74 0a20  ent, alignment. 
+00006950: 2020 2065 6c73 653a 0a20 2020 2020 2020     else:.       
+00006960: 2070 7269 6e74 2822 5573 696e 6720 6d61   print("Using ma
+00006970: 6666 7420 746f 2061 6c69 676e 2e2e 2e22  fft to align..."
+00006980: 290a 2020 2020 2020 2020 6d61 6666 745f  ).        mafft_
+00006990: 636c 696e 6520 3d20 4d61 6666 7443 6f6d  cline = MafftCom
+000069a0: 6d61 6e64 6c69 6e65 2869 6e70 7574 3d69  mandline(input=i
+000069b0: 6e70 7574 5f6e 616d 655f 6669 6c65 290a  nput_name_file).
+000069c0: 2020 2020 2020 2020 7374 646f 7574 2c20          stdout, 
+000069d0: 7374 6465 7272 203d 206d 6166 6674 5f63  stderr = mafft_c
+000069e0: 6c69 6e65 2829 0a20 2020 2020 2020 2077  line().        w
+000069f0: 6974 6820 6f70 656e 286f 7574 7075 745f  ith open(output_
+00006a00: 6e61 6d65 5f66 696c 652c 2022 7722 2920  name_file, "w") 
+00006a10: 6173 2068 616e 646c 653a 0a20 2020 2020  as handle:.     
+00006a20: 2020 2020 2020 2068 616e 646c 652e 7772         handle.wr
+00006a30: 6974 6528 7374 646f 7574 290a 2020 2020  ite(stdout).    
+00006a40: 2020 2020 6861 6e64 6c65 2e63 6c6f 7365      handle.close
+00006a50: 2829 0a0a 2020 2020 2020 2020 2320 5265  ()..        # Re
+00006a60: 6164 2074 6865 2061 6c69 676e 6564 2073  ad the aligned s
+00006a70: 6571 7565 6e63 6573 0a20 2020 2020 2020  equences.       
+00006a80: 2061 6c69 676e 6d65 6e74 203d 2041 6c69   alignment = Ali
+00006a90: 676e 494f 2e72 6561 6428 6f75 7470 7574  gnIO.read(output
+00006aa0: 5f6e 616d 655f 6669 6c65 2c20 2266 6173  _name_file, "fas
+00006ab0: 7461 2229 0a20 2020 2020 2020 2061 6c69  ta").        ali
+00006ac0: 676e 6d65 6e74 5f69 6473 203d 205b 616c  gnment_ids = [al
+00006ad0: 6967 6e6d 656e 745b 695d 2e69 6420 666f  ignment[i].id fo
+00006ae0: 7220 692c 2061 6c69 676e 6564 2069 6e20  r i, aligned in 
+00006af0: 656e 756d 6572 6174 6528 616c 6967 6e6d  enumerate(alignm
+00006b00: 656e 7429 5d0a 2020 2020 2020 2020 616c  ent)].        al
+00006b10: 6967 6e6d 656e 745f 7365 7173 203d 205b  ignment_seqs = [
+00006b20: 616c 6967 6e6d 656e 745b 695d 2e73 6571  alignment[i].seq
+00006b30: 2066 6f72 2069 2c20 616c 6967 6e65 6420   for i, aligned 
+00006b40: 696e 2065 6e75 6d65 7261 7465 2861 6c69  in enumerate(ali
+00006b50: 676e 6d65 6e74 295d 0a20 2020 2020 2020  gnment)].       
+00006b60: 2064 6963 745f 616c 6967 6e6d 656e 7420   dict_alignment 
+00006b70: 3d20 6469 6374 287a 6970 2861 6c69 676e  = dict(zip(align
+00006b80: 6d65 6e74 5f69 6473 2c20 616c 6967 6e6d  ment_ids, alignm
+00006b90: 656e 745f 7365 7173 2929 0a20 2020 2020  ent_seqs)).     
+00006ba0: 2020 2072 6574 7572 6e20 6469 6374 5f61     return dict_a
+00006bb0: 6c69 676e 6d65 6e74 2c20 616c 6967 6e6d  lignment, alignm
+00006bc0: 656e 740a 6465 6620 6361 6c63 756c 6174  ent.def calculat
+00006bd0: 655f 7061 6972 7769 7365 5f64 6973 7461  e_pairwise_dista
+00006be0: 6e63 6528 6e61 6d65 2c61 6c69 676e 6d65  nce(name,alignme
+00006bf0: 6e74 2c73 746f 7261 6765 5f66 6f6c 6465  nt,storage_folde
+00006c00: 7229 3a0a 2020 2020 2222 2243 616c 6375  r):.    """Calcu
+00006c10: 6c61 7465 7320 7468 6520 7061 6972 7769  lates the pairwi
+00006c20: 7365 2064 6973 7461 6e63 6520 6d61 7472  se distance matr
+00006c30: 6978 2061 6363 726f 7373 2074 6865 2073  ix accross the s
+00006c40: 6571 7565 6e63 6573 2069 6e20 7468 6520  equences in the 
+00006c50: 616c 6967 6e6d 656e 740a 2020 2020 3a70  alignment.    :p
+00006c60: 6172 616d 2073 7472 206e 616d 653a 2064  aram str name: d
+00006c70: 6174 6120 7365 7420 7072 6f6a 6563 7420  ata set project 
+00006c80: 6e61 6d65 0a20 2020 203a 7061 7261 6d20  name.    :param 
+00006c90: 6269 6f70 7974 686f 6e20 616c 6967 6e6d  biopython alignm
+00006ca0: 656e 7420 6f62 6a65 6374 3a20 6f62 6a65  ent object: obje
+00006cb0: 6374 2063 6f6e 7461 696e 696e 6720 616c  ct containing al
+00006cc0: 6967 6e65 6420 7365 7175 656e 6365 730a  igned sequences.
+00006cd0: 2020 2020 3a70 6172 616d 2073 7472 2073      :param str s
+00006ce0: 746f 7261 6765 5f66 6f6c 6465 723a 2070  torage_folder: p
+00006cf0: 6174 6820 7768 6572 6520 746f 2073 746f  ath where to sto
+00006d00: 7265 2074 6865 2063 616c 6375 6c61 7465  re the calculate
+00006d10: 6420 7061 6972 7769 7365 206d 6174 7269  d pairwise matri
+00006d20: 7822 2222 0a20 2020 2070 7269 6e74 2822  x""".    print("
+00006d30: 4275 696c 6469 6e67 2070 6169 7277 6973  Building pairwis
+00006d40: 6520 6469 7374 616e 6365 206d 6174 7269  e distance matri
+00006d50: 7820 2e2e 2e22 290a 2020 2020 6966 206c  x ...").    if l
+00006d60: 656e 2861 6c69 676e 6d65 6e74 2920 3c3d  en(alignment) <=
+00006d70: 2032 3030 3a20 2376 6572 7920 736c 6f77   200: #very slow
+00006d80: 206d 6574 686f 640a 2020 2020 2020 2020   method.        
+00006d90: 6361 6c63 756c 6174 6f72 203d 2044 6973  calculator = Dis
+00006da0: 7461 6e63 6543 616c 6375 6c61 746f 7228  tanceCalculator(
+00006db0: 2769 6465 6e74 6974 7927 290a 2020 2020  'identity').    
+00006dc0: 2020 2020 6469 7374 616e 6365 5f6d 6174      distance_mat
+00006dd0: 7269 785f 6269 6f70 7974 686f 6e20 3d20  rix_biopython = 
+00006de0: 6361 6c63 756c 6174 6f72 2e67 6574 5f64  calculator.get_d
+00006df0: 6973 7461 6e63 6528 616c 6967 6e6d 656e  istance(alignmen
+00006e00: 7429 0a20 2020 2020 2020 2064 6973 7461  t).        dista
+00006e10: 6e63 655f 6466 203d 2070 642e 4461 7461  nce_df = pd.Data
+00006e20: 4672 616d 6528 696e 6465 783d 6469 7374  Frame(index=dist
+00006e30: 616e 6365 5f6d 6174 7269 785f 6269 6f70  ance_matrix_biop
+00006e40: 7974 686f 6e2e 6e61 6d65 732c 2063 6f6c  ython.names, col
+00006e50: 756d 6e73 3d64 6973 7461 6e63 655f 6d61  umns=distance_ma
+00006e60: 7472 6978 5f62 696f 7079 7468 6f6e 2e6e  trix_biopython.n
+00006e70: 616d 6573 290a 2020 2020 2020 2020 6469  ames).        di
+00006e80: 7374 616e 6365 5f64 6620 3d20 6469 7374  stance_df = dist
+00006e90: 616e 6365 5f64 662e 6669 6c6c 6e61 2830  ance_df.fillna(0
+00006ea0: 290a 2020 2020 2020 2020 666f 7220 692c  ).        for i,
+00006eb0: 2074 3120 696e 2065 6e75 6d65 7261 7465   t1 in enumerate
+00006ec0: 2864 6973 7461 6e63 655f 6d61 7472 6978  (distance_matrix
+00006ed0: 5f62 696f 7079 7468 6f6e 2e6e 616d 6573  _biopython.names
+00006ee0: 293a 0a20 2020 2020 2020 2020 2020 2066  ):.            f
+00006ef0: 6f72 206a 2c20 7432 2069 6e20 656e 756d  or j, t2 in enum
+00006f00: 6572 6174 6528 6c69 7374 2864 6973 7461  erate(list(dista
+00006f10: 6e63 655f 6d61 7472 6978 5f62 696f 7079  nce_matrix_biopy
+00006f20: 7468 6f6e 2e6e 616d 6573 295b 6920 2b20  thon.names)[i + 
+00006f30: 313a 5d29 3a0a 2020 2020 2020 2020 2020  1:]):.          
+00006f40: 2020 2020 2020 6469 7374 616e 6365 5f64        distance_d
+00006f50: 662e 6c6f 635b 5b74 315d 2c20 5b74 325d  f.loc[[t1], [t2]
+00006f60: 5d20 3d20 6469 7374 616e 6365 5f6d 6174  ] = distance_mat
+00006f70: 7269 785f 6269 6f70 7974 686f 6e5b 7431  rix_biopython[t1
+00006f80: 2c74 325d 0a20 2020 2020 2020 2020 2020  ,t2].           
+00006f90: 2020 2020 2064 6973 7461 6e63 655f 6466       distance_df
+00006fa0: 2e6c 6f63 5b5b 7432 5d2c 205b 7431 5d5d  .loc[[t2], [t1]]
+00006fb0: 203d 2064 6973 7461 6e63 655f 6d61 7472   = distance_matr
+00006fc0: 6978 5f62 696f 7079 7468 6f6e 5b74 312c  ix_biopython[t1,
+00006fd0: 7432 5d0a 2020 2020 2020 2020 6469 7374  t2].        dist
+00006fe0: 616e 6365 5f64 662e 746f 5f63 7376 2822  ance_df.to_csv("
+00006ff0: 7b7d 2f7b 7d5f 7061 6972 7769 7365 5f64  {}/{}_pairwise_d
+00007000: 6973 7461 6e63 655f 6d61 7472 6978 2e63  istance_matrix.c
+00007010: 7376 222e 666f 726d 6174 2873 746f 7261  sv".format(stora
+00007020: 6765 5f66 6f6c 6465 722c 6e61 6d65 2929  ge_folder,name))
+00007030: 0a0a 2020 2020 656c 7365 3a20 2354 4f44  ..    else: #TOD
+00007040: 4f3a 2066 6173 7465 7220 696d 706c 656d  O: faster implem
+00007050: 656e 7461 7469 6f6e 2d2d 2d3e 2069 7427  entation---> it'
+00007060: 7320 646f 6e65 2065 6c73 6577 6865 7265  s done elsewhere
+00007070: 0a20 2020 2020 2020 2070 7269 6e74 2822  .        print("
+00007080: 4669 6e69 7368 2069 6d70 6c65 6d65 6e74  Finish implement
+00007090: 696e 6720 666f 7220 6c61 7267 6572 2064  ing for larger d
+000070a0: 6174 6173 6574 7322 290a 2020 2020 2020  atasets").      
+000070b0: 2020 2348 6967 686c 6967 6874 3a20 5475    #Highlight: Tu
+000070c0: 726e 2061 6c69 676e 6d65 6e74 2069 6e74  rn alignment int
+000070d0: 6f20 6e75 6d70 7920 6172 7261 792c 2076  o numpy array, v
+000070e0: 6563 746f 7269 7a65 2074 6f20 6e75 6d62  ectorize to numb
+000070f0: 6572 732c 2063 6f6d 7075 7465 7220 7061  ers, computer pa
+00007100: 6972 7769 7365 2069 6e20 6661 7374 206d  irwise in fast m
+00007110: 616e 6e65 720a 2020 2020 2020 2020 7061  anner.        pa
+00007120: 7373 0a64 6566 2063 616c 6375 6c61 7465  ss.def calculate
+00007130: 5f70 6174 7269 7374 6963 5f64 6973 7461  _patristic_dista
+00007140: 6e63 6528 6e61 6d65 5f66 696c 652c 636f  nce(name_file,co
+00007150: 6d62 696e 6564 5f64 6963 742c 6e6f 6465  mbined_dict,node
+00007160: 735f 616e 645f 6c65 6166 735f 6e61 6d65  s_and_leafs_name
+00007170: 732c 7472 6565 2c74 7265 655f 6669 6c65  s,tree,tree_file
+00007180: 2c20 7374 6f72 6167 655f 666f 6c64 6572  , storage_folder
+00007190: 293a 0a20 2020 2022 2222 4361 6c63 756c  ):.    """Calcul
+000071a0: 6174 6573 2074 6865 2070 6174 7269 7374  ates the patrist
+000071b0: 6963 2064 6973 7461 6e63 6573 206f 7220  ic distances or 
+000071c0: 6272 616e 6368 206c 656e 6774 6873 2061  branch lengths a
+000071d0: 6372 6f73 7320 7468 6520 6e6f 6465 7320  cross the nodes 
+000071e0: 696e 2061 2074 7265 652e 2049 7420 616c  in a tree. It al
+000071f0: 736f 2073 6176 6573 2074 6865 2074 7265  so saves the tre
+00007200: 6520 696e 2064 6966 6665 7265 6e74 2066  e in different f
+00007210: 6f72 6d61 7473 206e 6565 6465 6420 666f  ormats needed fo
+00007220: 7220 6265 6e63 686d 6172 6b69 6e67 2065  r benchmarking e
+00007230: 7463 0a20 2020 203a 7061 7261 6d20 7374  tc.    :param st
+00007240: 7220 6e61 6d65 5f66 696c 653a 2064 6174  r name_file: dat
+00007250: 6120 7365 7420 7072 6f6a 6563 7420 6e61  a set project na
+00007260: 6d65 0a20 2020 203a 7061 7261 6d20 6469  me.    :param di
+00007270: 6374 2063 6f6d 6269 6e65 645f 6469 6374  ct combined_dict
+00007280: 2023 544f 444f 3a20 5265 6d6f 7665 3f0a   #TODO: Remove?.
+00007290: 2020 2020 3a70 6172 616d 206c 6973 7420      :param list 
+000072a0: 6e6f 6465 735f 616e 645f 6c65 6166 735f  nodes_and_leafs_
+000072b0: 6e61 6d65 733a 2074 7265 6520 6e6f 6465  names: tree node
+000072c0: 7320 696e 2074 7265 652d 6c65 7665 6c20  s in tree-level 
+000072d0: 6f72 6465 7220 7374 6f72 6564 2069 6e20  order stored in 
+000072e0: 6120 6c69 7374 0a20 2020 203a 7061 7261  a list.    :para
+000072f0: 6d20 6574 6533 2d74 7265 6520 7472 6565  m ete3-tree tree
+00007300: 3a20 6574 6533 206f 626a 6563 7420 636f  : ete3 object co
+00007310: 6e74 6169 6e69 6e67 2074 7265 650a 2020  ntaining tree.  
+00007320: 2020 3a70 6172 616d 2073 7472 2074 7265    :param str tre
+00007330: 655f 6669 6c65 3a20 7061 7468 2074 6f20  e_file: path to 
+00007340: 7468 6520 7374 6f72 6564 2074 7265 6520  the stored tree 
+00007350: 6669 6c65 0a20 2020 203a 7061 7261 6d20  file.    :param 
+00007360: 7374 7220 7374 6f72 6167 655f 666f 6c64  str storage_fold
+00007370: 6572 3a20 666f 6c64 6572 2077 6865 7265  er: folder where
+00007380: 2074 6f20 7374 6f72 6520 7468 6520 7265   to store the re
+00007390: 7375 6c74 730a 2020 2020 2222 220a 0a20  sults.    """.. 
+000073a0: 2020 206e 5f73 6571 7320 3d20 6c65 6e28     n_seqs = len(
+000073b0: 636f 6d62 696e 6564 5f64 6963 7429 0a20  combined_dict). 
+000073c0: 2020 2023 776f 726b 5f64 6972 203d 206f     #work_dir = o
+000073d0: 732e 7061 7468 2e64 6972 6e61 6d65 286f  s.path.dirname(o
+000073e0: 732e 7061 7468 2e61 6273 7061 7468 285f  s.path.abspath(_
+000073f0: 5f66 696c 655f 5f29 290a 2020 2020 776f  _file__)).    wo
+00007400: 726b 5f64 6972 203d 2022 220a 2020 2020  rk_dir = "".    
+00007410: 6966 206e 5f73 6571 7320 3e20 3230 303a  if n_seqs > 200:
+00007420: 0a20 2020 2020 2020 2070 7269 6e74 2822  .        print("
+00007430: 4461 7461 7365 7420 6c61 7267 6572 2074  Dataset larger t
+00007440: 6861 6e20 3230 3020 7365 7175 656e 6365  han 200 sequence
+00007450: 733a 2055 7369 6e67 2052 2073 6372 6970  s: Using R scrip
+00007460: 7420 666f 7220 7061 7472 6973 7469 6320  t for patristic 
+00007470: 6469 7374 616e 6365 7320 2863 6c61 6469  distances (cladi
+00007480: 7374 6963 206d 6174 7269 7820 6973 204e  stic matrix is N
+00007490: 4f54 2061 7661 696c 6162 6c65 2921 2229  OT available)!")
+000074a0: 0a20 2020 2020 2020 2077 6172 6e69 6e67  .        warning
+000074b0: 732e 7761 726e 2822 4461 7461 7365 7420  s.warn("Dataset 
+000074c0: 6c61 7267 6572 2074 6861 6e20 3230 3020  larger than 200 
+000074d0: 7365 7175 656e 6365 733a 2052 6571 7569  sequences: Requi
+000074e0: 7265 7320 5220 616e 6420 7468 6520 6170  res R and the ap
+000074f0: 6520 6c69 6272 6172 7922 290a 2020 2020  e library").    
+00007500: 2020 2020 636f 6d6d 616e 6420 3d20 2752      command = 'R
+00007510: 7363 7269 7074 270a 2020 2020 2020 2020  script'.        
+00007520: 2370 6174 6832 7363 7269 7074 203d 2027  #path2script = '
+00007530: 2f68 6f6d 652f 6c79 732f 4472 6f70 626f  /home/lys/Dropbo
+00007540: 782f 5068 442f 4452 4155 504e 4952 2f43  x/PhD/DRAUPNIR/C
+00007550: 616c 6375 6c61 7465 5f50 6174 7269 7374  alculate_Patrist
+00007560: 6963 2e52 270a 2020 2020 2020 2020 7061  ic.R'.        pa
+00007570: 7468 3273 6372 6970 7420 3d20 2243 616c  th2script = "Cal
+00007580: 6375 6c61 7465 5f50 6174 7269 7374 6963  culate_Patristic
+00007590: 2e52 220a 2020 2020 2020 2020 6966 2074  .R".        if t
+000075a0: 7265 655f 6669 6c65 3a0a 2020 2020 2020  ree_file:.      
+000075b0: 2020 2020 2020 6e65 775f 7472 6565 203d        new_tree =
+000075c0: 2077 6f72 6b5f 6469 7220 2b74 7265 655f   work_dir +tree_
+000075d0: 6669 6c65 2e73 706c 6974 2822 2e22 295b  file.split(".")[
+000075e0: 305d 2b22 2e6e 6577 6963 6b22 0a20 2020  0]+".newick".   
+000075f0: 2020 2020 2020 2020 206e 6577 5f74 7265           new_tre
+00007600: 655f 666f 726d 6174 3820 3d20 776f 726b  e_format8 = work
+00007610: 5f64 6972 2020 2b20 7472 6565 5f66 696c  _dir  + tree_fil
+00007620: 652e 7370 6c69 7428 222e 2229 5b30 5d20  e.split(".")[0] 
+00007630: 2b20 222e 666f 726d 6174 386e 6577 6963  + ".format8newic
+00007640: 6b22 0a20 2020 2020 2020 2020 2020 206e  k".            n
+00007650: 6577 5f74 7265 655f 666f 726d 6174 3620  ew_tree_format6 
+00007660: 3d20 776f 726b 5f64 6972 2020 2b20 7472  = work_dir  + tr
+00007670: 6565 5f66 696c 652e 7370 6c69 7428 222e  ee_file.split(".
+00007680: 2229 5b30 5d20 2b20 222e 666f 726d 6174  ")[0] + ".format
+00007690: 366e 6577 6963 6b22 0a20 2020 2020 2020  6newick".       
+000076a0: 2020 2020 206e 6577 5f74 7265 655f 666f       new_tree_fo
+000076b0: 726d 6174 3720 3d20 776f 726b 5f64 6972  rmat7 = work_dir
+000076c0: 2020 2b20 7472 6565 5f66 696c 652e 7370    + tree_file.sp
+000076d0: 6c69 7428 222e 2229 5b30 5d20 2b20 222e  lit(".")[0] + ".
+000076e0: 666f 726d 6174 376e 6577 6963 6b22 0a20  format7newick". 
+000076f0: 2020 2020 2020 2065 6c73 653a 0a20 2020         else:.   
+00007700: 2020 2020 2020 2020 206e 6577 5f74 7265           new_tre
+00007710: 6520 3d20 776f 726b 5f64 6972 202b 2022  e = work_dir + "
+00007720: 7b7d 2f7b 7d2e 6e65 7769 636b 222e 666f  {}/{}.newick".fo
+00007730: 726d 6174 2873 746f 7261 6765 5f66 6f6c  rmat(storage_fol
+00007740: 6465 722c 6e61 6d65 5f66 696c 6529 0a20  der,name_file). 
+00007750: 2020 2020 2020 2020 2020 206e 6577 5f74             new_t
+00007760: 7265 655f 666f 726d 6174 3820 3d20 776f  ree_format8 = wo
+00007770: 726b 5f64 6972 202b 2022 7b7d 2f7b 7d2e  rk_dir + "{}/{}.
+00007780: 666f 726d 6174 386e 6577 6963 6b22 2e66  format8newick".f
+00007790: 6f72 6d61 7428 7374 6f72 6167 655f 666f  ormat(storage_fo
+000077a0: 6c64 6572 2c6e 616d 655f 6669 6c65 290a  lder,name_file).
+000077b0: 2020 2020 2020 2020 2020 2020 6e65 775f              new_
+000077c0: 7472 6565 5f66 6f72 6d61 7436 203d 2077  tree_format6 = w
+000077d0: 6f72 6b5f 6469 7220 2b20 227b 7d2f 7b7d  ork_dir + "{}/{}
+000077e0: 2e66 6f72 6d61 7436 6e65 7769 636b 222e  .format6newick".
+000077f0: 666f 726d 6174 2873 746f 7261 6765 5f66  format(storage_f
+00007800: 6f6c 6465 722c 6e61 6d65 5f66 696c 6529  older,name_file)
+00007810: 0a20 2020 2020 2020 2020 2020 206e 6577  .            new
+00007820: 5f74 7265 655f 666f 726d 6174 3720 3d20  _tree_format7 = 
+00007830: 776f 726b 5f64 6972 202b 2022 7b7d 2f7b  work_dir + "{}/{
+00007840: 7d2e 666f 726d 6174 376e 6577 6963 6b22  }.format7newick"
+00007850: 2e66 6f72 6d61 7428 7374 6f72 6167 655f  .format(storage_
+00007860: 666f 6c64 6572 2c6e 616d 655f 6669 6c65  folder,name_file
+00007870: 290a 0a20 2020 2020 2020 2074 7265 652e  )..        tree.
+00007880: 7772 6974 6528 6f75 7466 696c 653d 6e65  write(outfile=ne
+00007890: 775f 7472 6565 5f66 6f72 6d61 7438 2c20  w_tree_format8, 
+000078a0: 666f 726d 6174 3d38 2c66 6f72 6d61 745f  format=8,format_
+000078b0: 726f 6f74 5f6e 6f64 653d 5472 7565 2920  root_node=True) 
+000078c0: 2320 666f 726d 6174 2038 2061 6c6c 206e  # format 8 all n
+000078d0: 6f64 6573 206e 616d 6573 0a20 2020 2020  odes names.     
+000078e0: 2020 2074 7265 652e 7772 6974 6528 6f75     tree.write(ou
+000078f0: 7466 696c 653d 6e65 775f 7472 6565 5f66  tfile=new_tree_f
+00007900: 6f72 6d61 7436 2c20 666f 726d 6174 3d36  ormat6, format=6
+00007910: 2c66 6f72 6d61 745f 726f 6f74 5f6e 6f64  ,format_root_nod
+00007920: 653d 5472 7565 290a 2020 2020 2020 2020  e=True).        
+00007930: 7472 6565 2e77 7269 7465 286f 7574 6669  tree.write(outfi
+00007940: 6c65 3d6e 6577 5f74 7265 655f 666f 726d  le=new_tree_form
+00007950: 6174 372c 2066 6f72 6d61 743d 372c 666f  at7, format=7,fo
+00007960: 726d 6174 5f72 6f6f 745f 6e6f 6465 3d54  rmat_root_node=T
+00007970: 7275 6529 2023 616c 6c20 6e6f 6465 7320  rue) #all nodes 
+00007980: 6e61 6d65 7320 2b20 6272 616e 6368 206c  names + branch l
+00007990: 656e 6774 6873 0a20 2020 2020 2020 2074  engths.        t
+000079a0: 7265 652e 7772 6974 6528 6f75 7466 696c  ree.write(outfil
+000079b0: 653d 6e65 775f 7472 6565 2c66 6f72 6d61  e=new_tree,forma
+000079c0: 743d 3129 2023 7361 7665 2074 6865 2072  t=1) #save the r
+000079d0: 656e 616d 6564 2074 7265 652c 2066 6f72  enamed tree, for
+000079e0: 6d61 7420 3920 746f 206e 6f74 2073 6176  mat 9 to not sav
+000079f0: 6520 7468 6520 696e 7465 726e 616c 206e  e the internal n
+00007a00: 6f64 6573 206e 616d 6573 0a20 2020 2020  odes names.     
+00007a10: 2020 2070 6174 7269 7374 6963 5f66 696c     patristic_fil
+00007a20: 6520 3d20 227b 7d2f 7b7d 5f70 6174 7269  e = "{}/{}_patri
+00007a30: 7374 6963 5f64 6973 7461 6e63 655f 6d61  stic_distance_ma
+00007a40: 7472 6978 2e63 7376 222e 666f 726d 6174  trix.csv".format
+00007a50: 2873 746f 7261 6765 5f66 6f6c 6465 722c  (storage_folder,
+00007a60: 6e61 6d65 5f66 696c 6529 0a20 2020 2020  name_file).     
+00007a70: 2020 2069 6620 6e6f 7420 6f73 2e70 6174     if not os.pat
+00007a80: 682e 6578 6973 7473 2870 6174 7269 7374  h.exists(patrist
+00007a90: 6963 5f66 696c 6529 3a0a 2020 2020 2020  ic_file):.      
+00007aa0: 2020 2020 2020 2320 4275 696c 6420 7375        # Build su
+00007ab0: 6270 726f 6365 7373 2063 6f6d 6d61 6e64  bprocess command
+00007ac0: 0a20 2020 2020 2020 2020 2020 2073 7562  .            sub
+00007ad0: 7072 6f63 6573 732e 6368 6563 6b5f 6361  process.check_ca
+00007ae0: 6c6c 285b 636f 6d6d 616e 642c 7061 7468  ll([command,path
+00007af0: 3273 6372 6970 742c 6e65 775f 7472 6565  2script,new_tree
+00007b00: 2c70 6174 7269 7374 6963 5f66 696c 655d  ,patristic_file]
+00007b10: 290a 2020 2020 2020 2020 656c 7365 3a0a  ).        else:.
+00007b20: 2020 2020 2020 2020 2020 2020 7072 696e              prin
+00007b30: 7428 2250 6174 7269 7374 6963 206d 6174  t("Patristic mat
+00007b40: 7269 7820 616c 7265 6164 7920 6578 6973  rix already exis
+00007b50: 7473 2061 7420 7b7d 2c20 6e6f 7420 6361  ts at {}, not ca
+00007b60: 6c63 756c 6174 6564 2e20 4465 6c65 7465  lculated. Delete
+00007b70: 2069 7420 6f74 6865 7277 6973 6522 2e66   it otherwise".f
+00007b80: 6f72 6d61 7428 7061 7472 6973 7469 635f  ormat(patristic_
+00007b90: 6669 6c65 2929 0a20 2020 2020 2020 2023  file)).        #
+00007ba0: 4869 6768 6c69 6768 743a 2050 4859 4c4f  Highlight: PHYLO
+00007bb0: 434f 4d0a 2020 2020 2020 2020 2320 2320  COM.        # # 
+00007bc0: 2348 6967 686c 6967 6874 3a54 7261 6e73  #Highlight:Trans
+00007bd0: 666f 726d 2074 6865 2066 696c 6520 746f  form the file to
+00007be0: 204e 6578 7573 2066 6f72 6d61 740a 2020   Nexus format.  
+00007bf0: 2020 2020 2020 2320 6e65 775f 7472 6565        # new_tree
+00007c00: 203d 2074 7265 655f 6669 6c65 2e73 706c   = tree_file.spl
+00007c10: 6974 2822 2e22 295b 305d 2b22 2e6e 6577  it(".")[0]+".new
+00007c20: 6963 6b22 0a20 2020 2020 2020 2023 206e  ick".        # n
+00007c30: 6577 5f74 7265 655f 6669 6c65 6e61 6d65  ew_tree_filename
+00007c40: 203d 206e 7470 6174 682e 6261 7365 6e61   = ntpath.basena
+00007c50: 6d65 286e 6577 5f74 7265 6529 0a20 2020  me(new_tree).   
+00007c60: 2020 2020 2023 2077 6f72 6b69 6e67 5f64       # working_d
+00007c70: 6972 6563 746f 7279 203d 206f 732e 7061  irectory = os.pa
+00007c80: 7468 2e64 6972 6e61 6d65 286f 732e 7061  th.dirname(os.pa
+00007c90: 7468 2e61 6273 7061 7468 286e 6577 5f74  th.abspath(new_t
+00007ca0: 7265 6529 290a 2020 2020 2020 2020 2320  ree)).        # 
+00007cb0: 7472 6565 2e77 7269 7465 286f 7574 6669  tree.write(outfi
+00007cc0: 6c65 3d6e 6577 5f74 7265 652c 666f 726d  le=new_tree,form
+00007cd0: 6174 3d31 2920 2373 6176 6520 7468 6520  at=1) #save the 
+00007ce0: 7265 6e61 6d65 6420 7472 6565 2c20 666f  renamed tree, fo
+00007cf0: 726d 6174 2039 2074 6f20 6e6f 7420 7361  rmat 9 to not sa
+00007d00: 7665 2074 6865 2069 6e74 6572 6e61 6c20  ve the internal 
+00007d10: 6e6f 6465 7320 6e61 6d65 730a 2020 2020  nodes names.    
+00007d20: 2020 2020 2320 7061 7472 6973 7469 635f      # patristic_
+00007d30: 6669 6c65 203d 2022 4461 7461 7365 7473  file = "Datasets
+00007d40: 5f46 6f6c 6465 722f 5061 7472 6973 7469  _Folder/Patristi
+00007d50: 635f 6469 7374 616e 6365 5f6d 6174 7269  c_distance_matri
+00007d60: 785f 7b7d 2e74 7874 222e 666f 726d 6174  x_{}.txt".format
+00007d70: 286e 616d 655f 6669 6c65 290a 2020 2020  (name_file).    
+00007d80: 2020 2020 2320 7769 7468 206f 7065 6e28      # with open(
+00007d90: 7061 7472 6973 7469 635f 6669 6c65 2c20  patristic_file, 
+00007da0: 2277 2229 2061 7320 7061 7472 6973 7469  "w") as patristi
+00007db0: 635f 6469 7374 5f6f 7574 3a0a 2020 2020  c_dist_out:.    
+00007dc0: 2020 2020 2320 2020 2020 7375 6270 726f      #     subpro
+00007dd0: 6365 7373 2e72 756e 2861 7267 733d 5b22  cess.run(args=["
+00007de0: 7068 796c 6f63 6f6d 222c 2270 6879 6469  phylocom","phydi
+00007df0: 7374 222c 2022 2d66 222c 206e 6577 5f74  st", "-f", new_t
+00007e00: 7265 655f 6669 6c65 6e61 6d65 5d2c 7374  ree_filename],st
+00007e10: 6465 7272 3d73 7973 2e73 7464 6572 722c  derr=sys.stderr,
+00007e20: 2073 7464 6f75 743d 7061 7472 6973 7469   stdout=patristi
+00007e30: 635f 6469 7374 5f6f 7574 2c63 7764 3d77  c_dist_out,cwd=w
+00007e40: 6f72 6b69 6e67 5f64 6972 6563 746f 7279  orking_directory
+00007e50: 290a 2020 2020 2020 2020 2320 7061 7472  ).        # patr
+00007e60: 6973 7469 635f 6d61 7472 6978 203d 2070  istic_matrix = p
+00007e70: 642e 7265 6164 5f63 7376 2870 6174 7269  d.read_csv(patri
+00007e80: 7374 6963 5f66 696c 652c 7365 703d 225c  stic_file,sep="\
+00007e90: 7422 2c69 6e64 6578 5f63 6f6c 3d30 290a  t",index_col=0).
+00007ea0: 2020 2020 2020 2020 2320 7061 7472 6973          # patris
+00007eb0: 7469 635f 6d61 7472 6978 2e74 6f5f 6373  tic_matrix.to_cs
+00007ec0: 7628 2244 6174 6173 6574 735f 466f 6c64  v("Datasets_Fold
+00007ed0: 6572 2f50 6174 7269 7374 6963 5f64 6973  er/Patristic_dis
+00007ee0: 7461 6e63 655f 6d61 7472 6978 5f7b 7d2e  tance_matrix_{}.
+00007ef0: 6373 7622 2e66 6f72 6d61 7428 6e61 6d65  csv".format(name
+00007f00: 5f66 696c 6529 2c69 6e64 6578 5f6c 6162  _file),index_lab
+00007f10: 656c 3d22 726f 7773 2229 0a0a 2020 2020  el="rows")..    
+00007f20: 656c 7365 3a0a 2020 2020 2020 2020 6966  else:.        if
+00007f30: 2074 7265 655f 6669 6c65 3a0a 2020 2020   tree_file:.    
+00007f40: 2020 2020 2020 2020 6e65 775f 7472 6565          new_tree
+00007f50: 203d 2077 6f72 6b5f 6469 7220 2b20 222f   = work_dir + "/
+00007f60: 2220 2b20 7472 6565 5f66 696c 652e 7370  " + tree_file.sp
+00007f70: 6c69 7428 222e 2229 5b30 5d20 2b22 2e6e  lit(".")[0] +".n
+00007f80: 6577 6963 6b22 0a20 2020 2020 2020 2020  ewick".         
+00007f90: 2020 206e 6577 5f74 7265 655f 666f 726d     new_tree_form
+00007fa0: 6174 3820 3d20 776f 726b 5f64 6972 202b  at8 = work_dir +
+00007fb0: 2022 2f22 202b 2074 7265 655f 6669 6c65   "/" + tree_file
+00007fc0: 2e73 706c 6974 2822 2e22 295b 305d 202b  .split(".")[0] +
+00007fd0: 2022 2e66 6f72 6d61 7438 6e65 7769 636b   ".format8newick
+00007fe0: 220a 2020 2020 2020 2020 2020 2020 6e65  ".            ne
+00007ff0: 775f 7472 6565 5f66 6f72 6d61 7436 203d  w_tree_format6 =
+00008000: 2077 6f72 6b5f 6469 7220 2b20 222f 2220   work_dir + "/" 
+00008010: 2b20 7472 6565 5f66 696c 652e 7370 6c69  + tree_file.spli
+00008020: 7428 222e 2229 5b30 5d20 2b20 222e 666f  t(".")[0] + ".fo
+00008030: 726d 6174 366e 6577 6963 6b22 0a20 2020  rmat6newick".   
+00008040: 2020 2020 2020 2020 206e 6577 5f74 7265           new_tre
+00008050: 655f 666f 726d 6174 3720 3d20 776f 726b  e_format7 = work
+00008060: 5f64 6972 202b 2022 2f22 202b 2074 7265  _dir + "/" + tre
+00008070: 655f 6669 6c65 2e73 706c 6974 2822 2e22  e_file.split("."
+00008080: 295b 305d 202b 2022 2e66 6f72 6d61 7437  )[0] + ".format7
+00008090: 6e65 7769 636b 220a 2020 2020 2020 2020  newick".        
+000080a0: 656c 7365 3a0a 2020 2020 2020 2020 2020  else:.          
+000080b0: 2020 6e65 775f 7472 6565 203d 2077 6f72    new_tree = wor
+000080c0: 6b5f 6469 7220 2b20 222f 7b7d 2f7b 7d2e  k_dir + "/{}/{}.
+000080d0: 6e65 7769 636b 222e 666f 726d 6174 2873  newick".format(s
+000080e0: 746f 7261 6765 5f66 6f6c 6465 722c 6e61  torage_folder,na
+000080f0: 6d65 5f66 696c 6529 0a20 2020 2020 2020  me_file).       
+00008100: 2020 2020 206e 6577 5f74 7265 655f 666f       new_tree_fo
+00008110: 726d 6174 3820 3d20 776f 726b 5f64 6972  rmat8 = work_dir
+00008120: 202b 2022 2f7b 7d2f 7b7d 2e66 6f72 6d61   + "/{}/{}.forma
+00008130: 7438 6e65 7769 636b 222e 666f 726d 6174  t8newick".format
+00008140: 2873 746f 7261 6765 5f66 6f6c 6465 722c  (storage_folder,
+00008150: 6e61 6d65 5f66 696c 6529 0a20 2020 2020  name_file).     
+00008160: 2020 2020 2020 206e 6577 5f74 7265 655f         new_tree_
+00008170: 666f 726d 6174 3620 3d20 776f 726b 5f64  format6 = work_d
+00008180: 6972 202b 2022 2f7b 7d2f 7b7d 2e66 6f72  ir + "/{}/{}.for
+00008190: 6d61 7436 6e65 7769 636b 222e 666f 726d  mat6newick".form
+000081a0: 6174 2873 746f 7261 6765 5f66 6f6c 6465  at(storage_folde
+000081b0: 722c 6e61 6d65 5f66 696c 6529 0a20 2020  r,name_file).   
+000081c0: 2020 2020 2020 2020 206e 6577 5f74 7265           new_tre
+000081d0: 655f 666f 726d 6174 3720 3d20 776f 726b  e_format7 = work
+000081e0: 5f64 6972 202b 2022 2f7b 7d2f 7b7d 2e66  _dir + "/{}/{}.f
+000081f0: 6f72 6d61 7437 6e65 7769 636b 222e 666f  ormat7newick".fo
+00008200: 726d 6174 2873 746f 7261 6765 5f66 6f6c  rmat(storage_fol
+00008210: 6465 722c 6e61 6d65 5f66 696c 6529 0a0a  der,name_file)..
+00008220: 0a20 2020 2020 2020 2074 7265 652e 7772  .        tree.wr
+00008230: 6974 6528 6f75 7466 696c 653d 6e65 775f  ite(outfile=new_
+00008240: 7472 6565 5f66 6f72 6d61 7438 2c20 666f  tree_format8, fo
+00008250: 726d 6174 3d38 2c66 6f72 6d61 745f 726f  rmat=8,format_ro
+00008260: 6f74 5f6e 6f64 653d 5472 7565 290a 2020  ot_node=True).  
+00008270: 2020 2020 2020 7472 6565 2e77 7269 7465        tree.write
+00008280: 286f 7574 6669 6c65 3d6e 6577 5f74 7265  (outfile=new_tre
+00008290: 655f 666f 726d 6174 362c 2066 6f72 6d61  e_format6, forma
+000082a0: 743d 362c 666f 726d 6174 5f72 6f6f 745f  t=6,format_root_
+000082b0: 6e6f 6465 3d54 7275 6529 0a20 2020 2020  node=True).     
+000082c0: 2020 2074 7265 652e 7772 6974 6528 6f75     tree.write(ou
+000082d0: 7466 696c 653d 6e65 775f 7472 6565 5f66  tfile=new_tree_f
+000082e0: 6f72 6d61 7437 2c20 666f 726d 6174 3d37  ormat7, format=7
+000082f0: 2c66 6f72 6d61 745f 726f 6f74 5f6e 6f64  ,format_root_nod
+00008300: 653d 5472 7565 290a 2020 2020 2020 2020  e=True).        
+00008310: 7472 6565 2e77 7269 7465 286f 7574 6669  tree.write(outfi
+00008320: 6c65 3d6e 6577 5f74 7265 652c 666f 726d  le=new_tree,form
+00008330: 6174 3d31 2920 2023 2073 6176 6520 7468  at=1)  # save th
+00008340: 6520 7265 6e61 6d65 6420 7472 6565 2c20  e renamed tree, 
+00008350: 666f 726d 6174 2039 2074 6f20 6e6f 7420  format 9 to not 
+00008360: 7361 7665 2074 6865 2069 6e74 6572 6e61  save the interna
+00008370: 6c20 6e6f 6465 7320 6e61 6d65 732e 2066  l nodes names. f
+00008380: 6f72 6d61 7420 3820 616c 6c20 6e6f 6465  ormat 8 all node
+00008390: 7320 6e61 6d65 730a 0a20 2020 2020 2020  s names..       
+000083a0: 206e 5f65 6c65 6d65 6e74 7320 3d20 6c65   n_elements = le
+000083b0: 6e28 6e6f 6465 735f 616e 645f 6c65 6166  n(nodes_and_leaf
+000083c0: 735f 6e61 6d65 7329 0a20 2020 2020 2020  s_names).       
+000083d0: 2049 203d 2070 642e 496e 6465 7828 6e6f   I = pd.Index(no
+000083e0: 6465 735f 616e 645f 6c65 6166 735f 6e61  des_and_leafs_na
+000083f0: 6d65 732c 206e 616d 653d 2272 6f77 7322  mes, name="rows"
+00008400: 290a 2020 2020 2020 2020 4320 3d20 7064  ).        C = pd
+00008410: 2e49 6e64 6578 286e 6f64 6573 5f61 6e64  .Index(nodes_and
+00008420: 5f6c 6561 6673 5f6e 616d 6573 2c20 6e61  _leafs_names, na
+00008430: 6d65 3d22 636f 6c75 6d6e 7322 290a 2020  me="columns").  
+00008440: 2020 2020 2020 7061 7472 6973 7469 635f        patristic_
+00008450: 6d61 7472 6978 203d 2070 642e 4461 7461  matrix = pd.Data
+00008460: 4672 616d 6528 6461 7461 3d6e 702e 7a65  Frame(data=np.ze
+00008470: 726f 7328 286e 5f65 6c65 6d65 6e74 732c  ros((n_elements,
+00008480: 206e 5f65 6c65 6d65 6e74 7329 292c 2069   n_elements)), i
+00008490: 6e64 6578 3d49 2c20 636f 6c75 6d6e 733d  ndex=I, columns=
+000084a0: 4329 0a20 2020 2020 2020 2063 6c61 6469  C).        cladi
+000084b0: 7374 6963 5f6d 6174 7269 7820 3d20 7064  stic_matrix = pd
+000084c0: 2e44 6174 6146 7261 6d65 2864 6174 613d  .DataFrame(data=
+000084d0: 6e70 2e7a 6572 6f73 2828 6e5f 656c 656d  np.zeros((n_elem
+000084e0: 656e 7473 2c20 6e5f 656c 656d 656e 7473  ents, n_elements
+000084f0: 2929 2c20 696e 6465 783d 492c 2063 6f6c  )), index=I, col
+00008500: 756d 6e73 3d43 290a 2020 2020 2020 2020  umns=C).        
+00008510: 6966 206e 6f74 206f 732e 7061 7468 2e65  if not os.path.e
+00008520: 7869 7374 7328 227b 7d2f 7b7d 5f70 6174  xists("{}/{}_pat
+00008530: 7269 7374 6963 5f64 6973 7461 6e63 655f  ristic_distance_
+00008540: 6d61 7472 6978 2e63 7376 222e 666f 726d  matrix.csv".form
+00008550: 6174 2873 746f 7261 6765 5f66 6f6c 6465  at(storage_folde
+00008560: 722c 6e61 6d65 5f66 696c 6529 293a 0a20  r,name_file)):. 
+00008570: 2020 2020 2020 2020 2020 2066 6f72 2069             for i
+00008580: 2c20 7431 2069 6e20 656e 756d 6572 6174  , t1 in enumerat
+00008590: 6528 6e6f 6465 735f 616e 645f 6c65 6166  e(nodes_and_leaf
+000085a0: 735f 6e61 6d65 7329 3a0a 2020 2020 2020  s_names):.      
+000085b0: 2020 2020 2020 2020 2020 666f 7220 6a2c            for j,
+000085c0: 2074 3220 696e 2065 6e75 6d65 7261 7465   t2 in enumerate
+000085d0: 286c 6973 7428 6e6f 6465 735f 616e 645f  (list(nodes_and_
+000085e0: 6c65 6166 735f 6e61 6d65 7329 5b69 202b  leafs_names)[i +
+000085f0: 2031 3a5d 293a 0a20 2020 2020 2020 2020   1:]):.         
+00008600: 2020 2020 2020 2020 2020 2063 6c61 6469             cladi
+00008610: 7374 6963 5f6d 6174 7269 782e 6c6f 635b  stic_matrix.loc[
+00008620: 5b74 315d 2c20 5b74 325d 5d20 3d20 7472  [t1], [t2]] = tr
+00008630: 6565 2e67 6574 5f64 6973 7461 6e63 6528  ee.get_distance(
+00008640: 7431 2c20 7432 2c20 746f 706f 6c6f 6779  t1, t2, topology
+00008650: 5f6f 6e6c 793d 5472 7565 290a 2020 2020  _only=True).    
+00008660: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008670: 7061 7472 6973 7469 635f 6d61 7472 6978  patristic_matrix
+00008680: 2e6c 6f63 5b5b 7431 5d2c 205b 7432 5d5d  .loc[[t1], [t2]]
+00008690: 203d 2074 7265 652e 6765 745f 6469 7374   = tree.get_dist
+000086a0: 616e 6365 2874 312c 2074 322c 2074 6f70  ance(t1, t2, top
+000086b0: 6f6c 6f67 795f 6f6e 6c79 3d46 616c 7365  ology_only=False
+000086c0: 290a 2020 2020 2020 2020 2020 2020 636c  ).            cl
+000086d0: 6164 6973 7469 635f 6d61 7472 6978 2e74  adistic_matrix.t
+000086e0: 6f5f 6373 7628 227b 7d2f 7b7d 5f63 6c61  o_csv("{}/{}_cla
+000086f0: 6469 7374 6963 5f64 6973 7461 6e63 655f  distic_distance_
+00008700: 6d61 7472 6978 2e63 7376 222e 666f 726d  matrix.csv".form
+00008710: 6174 2873 746f 7261 6765 5f66 6f6c 6465  at(storage_folde
+00008720: 722c 6e61 6d65 5f66 696c 6529 290a 2020  r,name_file)).  
+00008730: 2020 2020 2020 2020 2020 7061 7472 6973            patris
+00008740: 7469 635f 6d61 7472 6978 2e74 6f5f 6373  tic_matrix.to_cs
+00008750: 7628 227b 7d2f 7b7d 5f70 6174 7269 7374  v("{}/{}_patrist
+00008760: 6963 5f64 6973 7461 6e63 655f 6d61 7472  ic_distance_matr
+00008770: 6978 2e63 7376 222e 666f 726d 6174 2873  ix.csv".format(s
+00008780: 746f 7261 6765 5f66 6f6c 6465 722c 6e61  torage_folder,na
+00008790: 6d65 5f66 696c 6529 290a 2020 2020 2020  me_file)).      
+000087a0: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
+000087b0: 2020 2020 7072 696e 7428 2250 6174 7269      print("Patri
+000087c0: 7374 6963 206d 6174 7269 7820 6669 6c65  stic matrix file
+000087d0: 2061 6c72 6561 6479 2065 7869 7374 732c   already exists,
+000087e0: 206e 6f74 2063 616c 6375 6c61 7465 6422   not calculated"
+000087f0: 290a 6465 6620 6d79 5f6c 6179 6f75 7428  ).def my_layout(
+00008800: 6e6f 6465 293a 0a20 2020 2022 2222 4574  node):.    """Et
+00008810: 6533 206c 6179 6f75 7420 7468 6174 2061  e3 layout that a
+00008820: 6464 7320 7468 6520 696e 7465 726e 616c  dds the internal
+00008830: 206e 6f64 6573 206e 616d 6573 2e20 4974   nodes names. It
+00008840: 2069 7320 6120 706c 7567 2d69 6e20 666f   is a plug-in fo
+00008850: 7220 7265 6e64 6572 696e 6720 7472 6565  r rendering tree
+00008860: 2069 6d61 6765 730a 2020 2020 3a70 6172   images.    :par
+00008870: 616d 2065 7465 332d 6e6f 6465 206e 6f64  am ete3-node nod
+00008880: 653a 206e 6f64 6520 6672 6f6d 2061 6e20  e: node from an 
+00008890: 6574 6533 2074 7265 6522 2222 0a20 2020  ete3 tree""".   
+000088a0: 2069 6620 6e6f 6465 2e69 735f 6c65 6166   if node.is_leaf
+000088b0: 2829 3a0a 2020 2020 2020 2020 2320 4966  ():.        # If
+000088c0: 2074 6572 6d69 6e61 6c20 6e6f 6465 2c20   terminal node, 
+000088d0: 6472 6177 7320 6974 7320 6e61 6d65 0a20  draws its name. 
+000088e0: 2020 2020 2020 206e 616d 655f 6661 6365         name_face
+000088f0: 203d 2041 7474 7246 6163 6528 226e 616d   = AttrFace("nam
+00008900: 6522 2c66 7369 7a65 3d38 2c66 6763 6f6c  e",fsize=8,fgcol
+00008910: 6f72 3d22 626c 7565 2229 0a20 2020 2065  or="blue").    e
+00008920: 6c73 653a 0a20 2020 2020 2020 2023 2049  lse:.        # I
+00008930: 6620 696e 7465 726e 616c 206e 6f64 652c  f internal node,
+00008940: 2064 7261 7773 206c 6162 656c 2077 6974   draws label wit
+00008950: 6820 736d 616c 6c65 7220 666f 6e74 2073  h smaller font s
+00008960: 697a 650a 2020 2020 2020 2020 6e61 6d65  ize.        name
+00008970: 5f66 6163 6520 3d20 4174 7472 4661 6365  _face = AttrFace
+00008980: 2822 6e61 6d65 222c 2066 7369 7a65 3d38  ("name", fsize=8
+00008990: 2c66 6763 6f6c 6f72 3d22 7265 6422 290a  ,fgcolor="red").
+000089a0: 2020 2020 2320 4164 6473 2074 6865 206e      # Adds the n
+000089b0: 616d 6520 6661 6365 2074 6f20 7468 6520  ame face to the 
+000089c0: 696d 6167 6520 6174 2074 6865 2070 7265  image at the pre
+000089d0: 6665 7272 6564 2070 6f73 6974 696f 6e0a  ferred position.
+000089e0: 2020 2020 6661 6365 732e 6164 645f 6661      faces.add_fa
+000089f0: 6365 5f74 6f5f 6e6f 6465 286e 616d 655f  ce_to_node(name_
+00008a00: 6661 6365 2c20 6e6f 6465 2c20 636f 6c75  face, node, colu
+00008a10: 6d6e 3d30 2c20 706f 7369 7469 6f6e 3d22  mn=0, position="
+00008a20: 6272 616e 6368 2d72 6967 6874 2229 0a64  branch-right").d
+00008a30: 6566 2072 656e 6465 725f 7472 6565 2874  ef render_tree(t
+00008a40: 7265 652c 7374 6f72 6167 655f 666f 6c64  ree,storage_fold
+00008a50: 6572 2c6e 616d 655f 6669 6c65 293a 0a20  er,name_file):. 
+00008a60: 2020 2022 2222 4675 6e63 7469 6f6e 2074     """Function t
+00008a70: 6f20 7265 6e64 6572 2061 6e20 6574 6533  o render an ete3
+00008a80: 2074 7265 6520 696e 746f 2061 6e20 696d   tree into an im
+00008a90: 6167 650a 2020 2020 3a70 6172 616d 2065  age.    :param e
+00008aa0: 7465 332d 7472 6565 2074 7265 653a 2045  te3-tree tree: E
+00008ab0: 7465 3320 7472 6565 206f 626a 6563 740a  te3 tree object.
+00008ac0: 2020 2020 3a70 6172 616d 2073 7472 2073      :param str s
+00008ad0: 746f 7261 6765 5f66 6f6c 6465 723a 2070  torage_folder: p
+00008ae0: 6174 6820 746f 2066 6f6c 6465 7220 7768  ath to folder wh
+00008af0: 6572 6520 746f 2073 746f 7265 2074 6865  ere to store the
+00008b00: 2072 6573 756c 7473 0a20 2020 203a 7061   results.    :pa
+00008b10: 7261 6d20 6e61 6d65 5f66 696c 653a 2064  ram name_file: d
+00008b20: 6174 6120 7365 7420 7072 6f6a 6563 7420  ata set project 
+00008b30: 6e61 6d65 2222 220a 2020 2020 7473 203d  name""".    ts =
+00008b40: 2054 7265 6553 7479 6c65 2829 0a20 2020   TreeStyle().   
+00008b50: 206e 7320 3d20 4e6f 6465 5374 796c 6528   ns = NodeStyle(
+00008b60: 290a 2020 2020 234d 616b 6520 7468 6963  ).    #Make thic
+00008b70: 6b65 7220 6c69 6e65 730a 2020 2020 6e73  ker lines.    ns
+00008b80: 5b22 7674 5f6c 696e 655f 7769 6474 6822  ["vt_line_width"
+00008b90: 5d20 3d20 350a 2020 2020 6e73 5b22 687a  ] = 5.    ns["hz
+00008ba0: 5f6c 696e 655f 7769 6474 6822 5d20 3d20  _line_width"] = 
+00008bb0: 350a 2020 2020 2320 446f 206e 6f74 2061  5.    # Do not a
+00008bc0: 6464 206c 6561 6620 6e61 6d65 7320 6175  dd leaf names au
+00008bd0: 746f 6d61 7469 6361 6c6c 790a 2020 2020  tomatically.    
+00008be0: 7473 2e73 686f 775f 6c65 6166 5f6e 616d  ts.show_leaf_nam
+00008bf0: 6520 3d20 4661 6c73 650a 2020 2020 2320  e = False.    # 
+00008c00: 5573 6520 6d79 2063 7573 746f 6d20 6c61  Use my custom la
+00008c10: 796f 7574 0a20 2020 2074 732e 6c61 796f  yout.    ts.layo
+00008c20: 7574 5f66 6e20 3d20 6d79 5f6c 6179 6f75  ut_fn = my_layou
+00008c30: 740a 2020 2020 2370 7269 6e74 2074 6865  t.    #print the
+00008c40: 2062 7261 6e63 6820 6c65 6e67 7468 730a   branch lengths.
+00008c50: 2020 2020 7473 2e73 686f 775f 6272 616e      ts.show_bran
+00008c60: 6368 5f6c 656e 6774 6820 3d20 5472 7565  ch_length = True
+00008c70: 0a20 2020 2066 6f72 206e 2069 6e20 7472  .    for n in tr
+00008c80: 6565 2e74 7261 7665 7273 6528 293a 0a20  ee.traverse():. 
+00008c90: 2020 2020 2020 2020 2020 206e 2e73 6574             n.set
+00008ca0: 5f73 7479 6c65 286e 7329 0a20 2020 2074  _style(ns).    t
+00008cb0: 7279 3a0a 2020 2020 2020 2020 7472 6565  ry:.        tree
+00008cc0: 2e72 656e 6465 7228 227b 7d2f 7265 7475  .render("{}/retu
+00008cd0: 726e 5f7b 7d2e 706e 6722 2e66 6f72 6d61  rn_{}.png".forma
+00008ce0: 7428 7374 6f72 6167 655f 666f 6c64 6572  t(storage_folder
+00008cf0: 2c6e 616d 655f 6669 6c65 292c 773d 3130  ,name_file),w=10
+00008d00: 3030 2c20 756e 6974 733d 226d 6d22 2c74  00, units="mm",t
+00008d10: 7265 655f 7374 796c 653d 7473 290a 2020  ree_style=ts).  
+00008d20: 2020 6578 6365 7074 3a0a 2020 2020 2020    except:.      
+00008d30: 2020 7472 6565 2e72 656e 6465 7228 227b    tree.render("{
+00008d40: 7d2f 7265 7475 726e 5f7b 7d2e 706e 6722  }/return_{}.png"
+00008d50: 2e66 6f72 6d61 7428 7374 6f72 6167 655f  .format(storage_
+00008d60: 666f 6c64 6572 2c6e 616d 655f 6669 6c65  folder,name_file
+00008d70: 292c 2077 3d31 3030 302c 2075 6e69 7473  ), w=1000, units
+00008d80: 3d22 6d6d 2229 0a64 6566 2072 656e 616d  ="mm").def renam
+00008d90: 655f 7472 6565 5f69 6e74 6572 6e61 6c5f  e_tree_internal_
+00008da0: 6e6f 6465 735f 7369 6d75 6c61 7469 6f6e  nodes_simulation
+00008db0: 7328 7472 6565 2c77 6974 685f 696e 6465  s(tree,with_inde
+00008dc0: 7865 733d 4661 6c73 6529 3a0a 2020 2020  xes=False):.    
+00008dd0: 2222 2252 656e 616d 6520 7468 6520 696e  """Rename the in
+00008de0: 7465 726e 616c 206e 6f64 6573 206f 6620  ternal nodes of 
+00008df0: 616e 2065 7465 3320 7472 6565 2074 6f20  an ete3 tree to 
+00008e00: 6120 6c61 6265 6c20 4920 2b20 6e75 6d62  a label I + numb
+00008e10: 6572 2c20 696e 2073 696d 756c 6174 696f  er, in simulatio
+00008e20: 6e73 2074 6865 206c 6561 7665 7320 6861  ns the leaves ha
+00008e30: 7665 2074 6865 2070 7265 6669 7820 412e  ve the prefix A.
+00008e40: 2057 6974 6820 696e 6465 7865 7320 7368   With indexes sh
+00008e50: 6f77 7320 7468 6520 6e61 6d65 7320 7573  ows the names us
+00008e60: 6564 2077 6865 6e20 7472 616e 7366 6572  ed when transfer
+00008e70: 7269 6e67 2074 6f20 616e 2061 7272 6179  ring to an array
+00008e80: 0a20 2020 203a 7061 7261 6d20 6574 6533  .    :param ete3
+00008e90: 2d74 7265 6520 7472 6565 3a20 4574 6533  -tree tree: Ete3
+00008ea0: 2074 7265 650a 2020 2020 3a70 6172 616d   tree.    :param
+00008eb0: 2062 6f6f 6c20 7769 7468 5f69 6e64 6578   bool with_index
+00008ec0: 6573 3a20 5472 7565 202d 2d3e 2061 6464  es: True --> add
+00008ed0: 7320 7468 6520 7472 6565 206c 6576 656c  s the tree level
+00008ee0: 206f 7264 6572 2069 6e64 6578 6573 2074   order indexes t
+00008ef0: 6f20 7468 6520 6e61 6d65 0a20 2020 2022  o the name.    "
+00008f00: 2222 0a20 2020 2070 7269 6e74 2822 5265  "".    print("Re
+00008f10: 6e61 6d69 6e67 2074 7265 6520 6672 6f6d  naming tree from
+00008f20: 2073 696d 756c 6174 696f 6e73 2229 0a20   simulations"). 
+00008f30: 2020 206c 6561 6673 5f6e 616d 6573 203d     leafs_names =
+00008f40: 2074 7265 652e 6765 745f 6c65 6166 5f6e   tree.get_leaf_n
+00008f50: 616d 6573 2829 0a20 2020 2065 6467 6520  ames().    edge 
+00008f60: 3d20 6c65 6e28 6c65 6166 735f 6e61 6d65  = len(leafs_name
+00008f70: 7329 0a20 2020 2069 6e74 6572 6e61 6c5f  s).    internal_
+00008f80: 6e6f 6465 735f 6e61 6d65 7320 3d20 5b5d  nodes_names = []
+00008f90: 0a20 2020 2069 6620 7769 7468 5f69 6e64  .    if with_ind
+00008fa0: 6578 6573 3a0a 2020 2020 2020 2020 666f  exes:.        fo
+00008fb0: 7220 6964 782c 6e6f 6465 2069 6e20 656e  r idx,node in en
+00008fc0: 756d 6572 6174 6528 7472 6565 2e74 7261  umerate(tree.tra
+00008fd0: 7665 7273 6528 2929 3a20 2023 206c 6576  verse()):  # lev
+00008fe0: 656c 6f72 6465 7220 286e 6f64 6573 2061  elorder (nodes a
+00008ff0: 7265 2076 6973 6974 6564 2069 6e20 7a69  re visited in zi
+00009000: 6720 7a61 6720 6f72 6465 7220 6672 6f6d  g zag order from
+00009010: 2072 6f6f 7420 746f 206c 6561 7665 7329   root to leaves)
+00009020: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
+00009030: 6e6f 7420 6e6f 6465 2e69 735f 6c65 6166  not node.is_leaf
+00009040: 2829 3a0a 2020 2020 2020 2020 2020 2020  ():.            
+00009050: 2020 2020 6e6f 6465 2e6e 616d 6520 3d20      node.name = 
+00009060: 2249 2220 2b20 6e6f 6465 2e6e 616d 6520  "I" + node.name 
+00009070: 2b20 222f 7b7d 222e 666f 726d 6174 2869  + "/{}".format(i
+00009080: 6478 290a 2020 2020 2020 2020 2020 2020  dx).            
+00009090: 2020 2020 696e 7465 726e 616c 5f6e 6f64      internal_nod
+000090a0: 6573 5f6e 616d 6573 2e61 7070 656e 6428  es_names.append(
+000090b0: 6e6f 6465 2e6e 616d 6529 0a20 2020 2020  node.name).     
+000090c0: 2020 2020 2020 2020 2020 2065 6467 6520             edge 
+000090d0: 2b3d 2031 0a20 2020 2020 2020 2020 2020  += 1.           
+000090e0: 2065 6c73 653a 0a20 2020 2020 2020 2020   else:.         
+000090f0: 2020 2020 2020 206e 6f64 652e 6e61 6d65         node.name
+00009100: 203d 206e 6f64 652e 6e61 6d65 202b 2022   = node.name + "
+00009110: 2f7b 7d22 2e66 6f72 6d61 7428 6964 7829  /{}".format(idx)
+00009120: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00009130: 2023 6564 6765 202b 3d20 310a 2020 2020   #edge += 1.    
+00009140: 656c 7365 3a0a 2020 2020 2020 2020 666f  else:.        fo
+00009150: 7220 6e6f 6465 2069 6e20 7472 6565 2e74  r node in tree.t
+00009160: 7261 7665 7273 6528 293a 2020 2320 6c65  raverse():  # le
+00009170: 7665 6c6f 7264 6572 2028 6e6f 6465 7320  velorder (nodes 
+00009180: 6172 6520 7669 7369 7465 6420 696e 207a  are visited in z
+00009190: 6967 207a 6167 206f 7264 6572 2066 726f  ig zag order fro
+000091a0: 6d20 726f 6f74 2074 6f20 6c65 6176 6573  m root to leaves
+000091b0: 290a 2020 2020 2020 2020 2020 2020 6966  ).            if
+000091c0: 206e 6f74 206e 6f64 652e 6973 5f6c 6561   not node.is_lea
+000091d0: 6628 293a 0a20 2020 2020 2020 2020 2020  f():.           
+000091e0: 2020 2020 206e 6f64 652e 6e61 6d65 203d       node.name =
+000091f0: 2022 4922 202b 206e 6f64 652e 6e61 6d65   "I" + node.name
+00009200: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00009210: 2069 6e74 6572 6e61 6c5f 6e6f 6465 735f   internal_nodes_
+00009220: 6e61 6d65 732e 6170 7065 6e64 286e 6f64  names.append(nod
+00009230: 652e 6e61 6d65 290a 2020 2020 2020 2020  e.name).        
+00009240: 2020 2020 2020 2020 6564 6765 202b 3d20          edge += 
+00009250: 310a 2020 2020 7265 7475 726e 2074 7265  1.    return tre
+00009260: 650a 6465 6620 7265 6e61 6d65 5f74 7265  e.def rename_tre
+00009270: 655f 696e 7465 726e 616c 5f6e 6f64 6573  e_internal_nodes
+00009280: 2874 7265 6529 3a0a 2020 2020 2222 2252  (tree):.    """R
+00009290: 656e 616d 6520 7468 6520 696e 7465 726e  ename the intern
+000092a0: 616c 206e 6f64 6573 206f 6620 6120 7472  al nodes of a tr
+000092b0: 6565 2074 6f20 6120 6c61 6265 6c20 4120  ee to a label A 
+000092c0: 2b20 6e75 6d62 6572 2c20 756e 6c65 7373  + number, unless
+000092d0: 2074 6865 2067 6976 656e 206e 6577 6963   the given newic
+000092e0: 6b20 6669 6c65 2061 6c72 6561 6479 2068  k file already h
+000092f0: 6173 2074 6865 206e 616d 6573 206f 6e20  as the names on 
+00009300: 6974 0a20 2020 203a 7061 7261 6d20 6574  it.    :param et
+00009310: 6533 2d74 7265 6520 7472 6565 3a20 4574  e3-tree tree: Et
+00009320: 6533 2074 7265 6520 636f 6e73 7472 7563  e3 tree construc
+00009330: 746f 7222 2222 0a0a 2020 2020 2352 656e  tor"""..    #Ren
+00009340: 616d 6520 7468 6520 696e 7465 726e 616c  ame the internal
+00009350: 206e 6f64 6573 0a20 2020 206c 6561 6673   nodes.    leafs
+00009360: 5f6e 616d 6573 203d 2074 7265 652e 6765  _names = tree.ge
+00009370: 745f 6c65 6166 5f6e 616d 6573 2829 0a20  t_leaf_names(). 
+00009380: 2020 2065 6467 6520 3d20 6c65 6e28 6c65     edge = len(le
+00009390: 6166 735f 6e61 6d65 7329 0a20 2020 2069  afs_names).    i
+000093a0: 6e74 6572 6e61 6c5f 6e6f 6465 735f 6e61  nternal_nodes_na
+000093b0: 6d65 7320 3d20 5b5d 0a20 2020 2066 6f72  mes = [].    for
+000093c0: 206e 6f64 6520 696e 2074 7265 652e 7472   node in tree.tr
+000093d0: 6176 6572 7365 2829 3a20 236c 6576 656c  averse(): #level
+000093e0: 6f72 6465 7220 286e 6f64 6573 2061 7265  order (nodes are
+000093f0: 2076 6973 6974 6564 2069 6e20 7a69 6720   visited in zig 
+00009400: 7a61 6720 6f72 6465 7220 6672 6f6d 2072  zag order from r
+00009410: 6f6f 7420 746f 206c 6561 7665 7329 0a20  oot to leaves). 
+00009420: 2020 2020 2020 2069 6620 6e6f 7420 6e6f         if not no
+00009430: 6465 2e69 735f 6c65 6166 2829 3a0a 2020  de.is_leaf():.  
+00009440: 2020 2020 2020 2020 2020 6e6f 6465 2e6e            node.n
+00009450: 616d 6520 3d20 2241 2564 2220 2520 6564  ame = "A%d" % ed
+00009460: 6765 0a20 2020 2020 2020 2020 2020 2069  ge.            i
+00009470: 6e74 6572 6e61 6c5f 6e6f 6465 735f 6e61  nternal_nodes_na
+00009480: 6d65 732e 6170 7065 6e64 286e 6f64 652e  mes.append(node.
+00009490: 6e61 6d65 290a 2020 2020 2020 2020 2020  name).          
+000094a0: 2020 6564 6765 202b 3d20 310a 2020 2020    edge += 1.    
+000094b0: 7265 7475 726e 2074 7265 650a 6465 6620  return tree.def 
+000094c0: 7072 6f63 6573 735f 7064 625f 6669 6c65  process_pdb_file
+000094d0: 7328 5044 425f 666f 6c64 6572 2c61 615f  s(PDB_folder,aa_
+000094e0: 7072 6f62 732c 7066 616d 5f64 6963 742c  probs,pfam_dict,
+000094f0: 6f6e 655f 686f 745f 656e 636f 6469 6e67  one_hot_encoding
+00009500: 2c6d 696e 5f6c 656e 293a 0a20 2020 2022  ,min_len):.    "
+00009510: 2222 2245 7874 7261 6374 2069 6e66 6f72  """Extract infor
+00009520: 6d61 7469 6f6e 2066 726f 6d20 7468 6520  mation from the 
+00009530: 5044 4220 6669 6c65 730a 2020 2020 3a70  PDB files.    :p
+00009540: 6172 616d 2073 7472 2050 4442 5f66 6f6c  aram str PDB_fol
+00009550: 6465 723a 2070 6174 6820 746f 2066 6f6c  der: path to fol
+00009560: 6465 7220 636f 6e74 6169 6e69 6e67 2074  der containing t
+00009570: 6865 2050 4642 2066 696c 6573 0a20 2020  he PFB files.   
+00009580: 203a 7061 7261 6d20 696e 7420 6161 5f70   :param int aa_p
+00009590: 726f 6273 3a20 616d 696e 6f20 6163 6964  robs: amino acid
+000095a0: 2070 726f 6261 6269 6c69 7469 6573 0a20   probabilities. 
+000095b0: 2020 203a 7061 7261 6d20 6469 6374 2070     :param dict p
+000095c0: 6661 6d5f 6469 6374 3a20 6469 6374 696f  fam_dict: dictio
+000095d0: 6e61 7279 2063 6f6e 7461 696e 696e 6720  nary containing 
+000095e0: 7468 6520 696e 666f 726d 6174 696f 6e20  the information 
+000095f0: 6f6e 2068 6f77 2074 6f20 6578 7472 6163  on how to extrac
+00009600: 7420 7468 6520 7365 7175 656e 6365 7320  t the sequences 
+00009610: 6672 6f6d 2074 6865 2070 6462 2066 696c  from the pdb fil
+00009620: 650a 2020 2020 3a70 6172 616d 2062 6f6f  e.    :param boo
+00009630: 6c20 6f6e 655f 686f 745f 656e 636f 6469  l one_hot_encodi
+00009640: 6e67 0a20 2020 203a 7061 7261 6d20 696e  ng.    :param in
+00009650: 7420 6d69 6e5f 6c65 6e3a 204c 6f77 6572  t min_len: Lower
+00009660: 2062 6f75 6e64 206f 6e20 7468 6520 7365   bound on the se
+00009670: 7175 656e 6365 206c 656e 6774 680a 2020  quence length.  
+00009680: 2020 2222 220a 2020 2020 7061 7273 6572    """.    parser
+00009690: 203d 2050 4442 2e50 4442 5061 7273 6572   = PDB.PDBParser
+000096a0: 2829 0a20 2020 2023 7070 6220 3d20 5050  ().    #ppb = PP
+000096b0: 4275 696c 6465 7228 290a 2020 2020 6361  Builder().    ca
+000096c0: 7070 203d 2043 6150 5042 7569 6c64 6572  pp = CaPPBuilder
+000096d0: 2829 0a20 2020 2070 726f 745f 696e 666f  ().    prot_info
+000096e0: 5f64 6963 7420 3d20 7b7d 0a20 2020 2070  _dict = {}.    p
+000096f0: 726f 745f 6161 5f64 6963 7420 3d20 7b7d  rot_aa_dict = {}
+00009700: 0a20 2020 2061 6d69 6e6f 6163 6964 5f6e  .    aminoacid_n
+00009710: 616d 6573 203d 2061 6d69 6e6f 6163 6964  ames = aminoacid
+00009720: 5f6e 616d 6573 5f64 6963 7428 6161 5f70  _names_dict(aa_p
+00009730: 726f 6273 2920 2023 2054 4f44 4f3a 2067  robs)  # TODO: g
+00009740: 656e 6572 616c 697a 6520 746f 2069 6e63  eneralize to inc
+00009750: 6c75 6465 2061 6c6c 2074 7970 6573 206f  lude all types o
+00009760: 6620 616d 696e 6f20 6163 6964 732e 2046  f amino acids. F
+00009770: 6f72 206f 6e65 686f 7420 656e 636f 6469  or onehot encodi
+00009780: 6e67 7320 7369 6d70 6c79 2075 7365 2061  ngs simply use a
+00009790: 615f 6e61 6d65 735f 6469 6374 2e6b 6579  a_names_dict.key
+000097a0: 7328 290a 2020 2020 6669 6c65 735f 6c69  s().    files_li
+000097b0: 7374 203d 205b 6620 666f 7220 6620 696e  st = [f for f in
+000097c0: 206c 6973 7464 6972 2850 4442 5f66 6f6c   listdir(PDB_fol
+000097d0: 6465 7229 2069 6620 6973 6669 6c65 286a  der) if isfile(j
+000097e0: 6f69 6e28 5044 425f 666f 6c64 6572 2c20  oin(PDB_folder, 
+000097f0: 6629 295d 0a20 2020 2064 7570 6c69 6361  f))].    duplica
+00009800: 7465 7320 3d20 5b5d 0a20 2020 2066 6f72  tes = [].    for
+00009810: 2069 2c20 5044 425f 6669 6c65 2069 6e20   i, PDB_file in 
+00009820: 656e 756d 6572 6174 6528 6669 6c65 735f  enumerate(files_
+00009830: 6c69 7374 293a 0a20 2020 2020 2020 2073  list):.        s
+00009840: 7472 7563 7475 7265 203d 2070 6172 7365  tructure = parse
+00009850: 722e 6765 745f 7374 7275 6374 7572 6528  r.get_structure(
+00009860: 277b 7d27 2e66 6f72 6d61 7428 5044 425f  '{}'.format(PDB_
+00009870: 6669 6c65 292c 206a 6f69 6e28 5044 425f  file), join(PDB_
+00009880: 666f 6c64 6572 2c20 5044 425f 6669 6c65  folder, PDB_file
+00009890: 2929 0a20 2020 2020 2020 2069 6620 7066  )).        if pf
+000098a0: 616d 5f64 6963 743a 2020 2320 636f 6e74  am_dict:  # cont
+000098b0: 6169 6e73 2069 6e66 6f72 6d61 7469 6f6e  ains information
+000098c0: 206f 6e20 7768 6963 6820 6368 6169 6e20   on which chain 
+000098d0: 746f 2074 616b 6520 616e 6420 7768 6963  to take and whic
+000098e0: 6820 7265 7369 6475 6573 2074 6f20 7365  h residues to se
+000098f0: 6c65 6374 2066 726f 6d20 7468 6520 5044  lect from the PD
+00009900: 4220 6669 6c65 0a20 2020 2020 2020 2020  B file.         
+00009910: 2020 2063 6861 696e 5f6e 616d 6520 3d20     chain_name = 
+00009920: 7066 616d 5f64 6963 745b 5044 425f 6669  pfam_dict[PDB_fi
+00009930: 6c65 5b33 3a37 5d2e 7570 7065 7228 295d  le[3:7].upper()]
+00009940: 5b22 4368 6169 6e22 5d0a 2020 2020 2020  ["Chain"].      
+00009950: 2020 2020 2020 4368 6169 6e5f 3020 3d20        Chain_0 = 
+00009960: 7374 7275 6374 7572 655b 305d 5b63 6861  structure[0][cha
+00009970: 696e 5f6e 616d 655d 0a20 2020 2020 2020  in_name].       
+00009980: 2020 2020 2073 7461 7274 5f72 6573 6964       start_resid
+00009990: 7565 2c20 656e 645f 7265 7369 6475 6520  ue, end_residue 
+000099a0: 3d20 7066 616d 5f64 6963 745b 5044 425f  = pfam_dict[PDB_
+000099b0: 6669 6c65 5b33 3a37 5d2e 7570 7065 7228  file[3:7].upper(
+000099c0: 295d 5b22 5265 7369 6475 6573 225d 2e73  )]["Residues"].s
+000099d0: 706c 6974 2822 2d22 290a 2020 2020 2020  plit("-").      
+000099e0: 2020 2020 2020 6c69 7374 5f72 6573 6964        list_resid
+000099f0: 7565 7320 3d20 6c69 7374 2872 616e 6765  ues = list(range
+00009a00: 2869 6e74 2873 7461 7274 5f72 6573 6964  (int(start_resid
+00009a10: 7565 292c 2069 6e74 2865 6e64 5f72 6573  ue), int(end_res
+00009a20: 6964 7565 2920 2b20 3129 290a 2020 2020  idue) + 1)).    
+00009a30: 2020 2020 656c 7365 3a20 2023 2077 6865      else:  # whe
+00009a40: 6e20 7468 6572 6520 6973 206e 6f74 2049  n there is not I
+00009a50: 6e66 6f72 6d61 7469 6f6e 206e 6f74 2061  nformation not a
+00009a60: 7661 696c 6162 6c65 206f 6e20 7768 6963  vailable on whic
+00009a70: 6820 6368 6169 6e73 2074 6f20 7069 636b  h chains to pick
+00009a80: 0a20 2020 2020 2020 2020 2020 2063 6861  .            cha
+00009a90: 696e 7320 3d20 5b63 6861 696e 2066 6f72  ins = [chain for
+00009aa0: 2063 6861 696e 2069 6e20 7374 7275 6374   chain in struct
+00009ab0: 7572 655b 305d 5d0a 2020 2020 2020 2020  ure[0]].        
+00009ac0: 2020 2020 4368 6169 6e5f 3020 3d20 6368      Chain_0 = ch
+00009ad0: 6169 6e73 5b30 5d0a 2020 2020 2020 2020  ains[0].        
+00009ae0: 706f 6c79 7065 7074 6964 6573 203d 2063  polypeptides = c
+00009af0: 6170 702e 6275 696c 645f 7065 7074 6964  app.build_peptid
+00009b00: 6573 2843 6861 696e 5f30 2920 2023 2043  es(Chain_0)  # C
+00009b10: 5f61 6c70 6861 2d43 2d61 6c70 6861 2070  _alpha-C-alpha p
+00009b20: 6f6c 7970 6570 7469 6465 0a20 2020 2020  olypeptide.     
+00009b30: 2020 2061 6e67 6c65 735f 6c69 7374 203d     angles_list =
+00009b40: 205b 5d0a 2020 2020 2020 2020 6161 5f6c   [].        aa_l
+00009b50: 6973 745f 656d 6265 6464 6564 203d 205b  ist_embedded = [
+00009b60: 5d0a 2020 2020 2020 2020 6161 5f6c 6973  ].        aa_lis
+00009b70: 7420 3d20 5b5d 0a20 2020 2020 2020 2023  t = [].        #
+00009b80: 2063 6f6f 7264 696e 6174 6573 5f6c 6973   coordinates_lis
+00009b90: 7420 3d20 5b5d 0a20 2020 2020 2020 2066  t = [].        f
+00009ba0: 6f72 2070 6f6c 795f 696e 6465 782c 2070  or poly_index, p
+00009bb0: 6f6c 7920 696e 2065 6e75 6d65 7261 7465  oly in enumerate
+00009bc0: 2870 6f6c 7970 6570 7469 6465 7329 3a0a  (polypeptides):.
+00009bd0: 2020 2020 2020 2020 2020 2020 6966 206e              if n
+00009be0: 6f74 2070 6661 6d5f 6469 6374 3a0a 2020  ot pfam_dict:.  
+00009bf0: 2020 2020 2020 2020 2020 2020 2020 6c69                li
+00009c00: 7374 5f72 6573 6964 7565 7320 3d20 5b72  st_residues = [r
+00009c10: 6573 6964 7565 2e67 6574 5f69 6428 295b  esidue.get_id()[
+00009c20: 315d 2066 6f72 2072 6573 6964 7565 2069  1] for residue i
+00009c30: 6e20 706f 6c79 5d20 2023 2054 4f44 4f3a  n poly]  # TODO:
+00009c40: 2043 6861 6e67 6520 746f 202e 6765 745f   Change to .get_
+00009c50: 7265 736e 616d 6528 293f 0a20 2020 2020  resname()?.     
+00009c60: 2020 2020 2020 2063 6f72 7265 7370 6f6e         correspon
+00009c70: 6465 6e63 655f 696e 6465 7820 3d20 5b69  dence_index = [i
+00009c80: 6e64 6578 2066 6f72 2069 6e64 6578 2c20  ndex for index, 
+00009c90: 7265 7369 6475 6520 696e 2065 6e75 6d65  residue in enume
+00009ca0: 7261 7465 2870 6f6c 7929 2069 6620 7265  rate(poly) if re
+00009cb0: 7369 6475 652e 6765 745f 6964 2829 5b31  sidue.get_id()[1
+00009cc0: 5d20 696e 206c 6973 745f 7265 7369 6475  ] in list_residu
+00009cd0: 6573 5d0a 2020 2020 2020 2020 2020 2020  es].            
+00009ce0: 7068 6970 7369 5f6c 6973 7420 3d20 706f  phipsi_list = po
+00009cf0: 6c79 2e67 6574 5f70 6869 5f70 7369 5f6c  ly.get_phi_psi_l
+00009d00: 6973 7428 290a 2020 2020 2020 2020 2020  ist().          
+00009d10: 2020 2320 476f 7474 6120 6368 6f70 2061    # Gotta chop a
+00009d20: 6c73 6f20 7468 6520 616e 676c 6573 206c  lso the angles l
+00009d30: 6973 7420 6163 636f 7264 696e 6720 746f  ist according to
+00009d40: 2074 6865 2072 616e 6765 206f 6620 6465   the range of de
+00009d50: 7369 7265 6420 7265 7369 6475 6573 0a20  sired residues. 
+00009d60: 2020 2020 2020 2020 2020 2069 6620 636f             if co
+00009d70: 7272 6573 706f 6e64 656e 6365 5f69 6e64  rrespondence_ind
+00009d80: 6578 2061 6e64 2070 6661 6d5f 6469 6374  ex and pfam_dict
+00009d90: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+00009da0: 2020 616e 676c 6573 5f6c 6973 7420 2b3d    angles_list +=
+00009db0: 2070 6869 7073 695f 6c69 7374 5b63 6f72   phipsi_list[cor
+00009dc0: 7265 7370 6f6e 6465 6e63 655f 696e 6465  respondence_inde
+00009dd0: 785b 305d 3a63 6f72 7265 7370 6f6e 6465  x[0]:corresponde
+00009de0: 6e63 655f 696e 6465 785b 0a20 2020 2020  nce_index[.     
 00009df0: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00009e00: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00009e10: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00009e20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00009e30: 202d 315d 202b 2031 5d20 2023 2063 686f   -1] + 1]  # cho
-00009e40: 7020 616c 736f 2074 6865 2061 6e67 6c65  p also the angle
-00009e50: 735f 6c69 7374 2061 6363 6f72 6469 6e67  s_list according
-00009e60: 2074 6f20 7468 6520 7365 6c65 6374 6564   to the selected
-00009e70: 2072 6573 6964 7565 730a 2020 2020 2020   residues.      
-00009e80: 2020 2020 2020 656c 6966 206e 6f74 2063        elif not c
-00009e90: 6f72 7265 7370 6f6e 6465 6e63 655f 696e  orrespondence_in
-00009ea0: 6465 7820 616e 6420 7066 616d 5f64 6963  dex and pfam_dic
-00009eb0: 743a 2020 2320 6966 2074 6865 7265 2069  t:  # if there i
-00009ec0: 7320 6e6f 7420 6120 636f 7272 6573 706f  s not a correspo
-00009ed0: 6e64 696e 6720 7265 7369 6475 6520 646f  nding residue do
-00009ee0: 206e 6f74 2061 7070 656e 6420 616e 7974   not append anyt
-00009ef0: 6869 6e67 0a20 2020 2020 2020 2020 2020  hing.           
-00009f00: 2020 2020 2061 6e67 6c65 735f 6c69 7374       angles_list
-00009f10: 202b 3d20 5b5d 0a20 2020 2020 2020 2020   += [].         
-00009f20: 2020 2065 6c73 653a 2020 2320 6b65 6570     else:  # keep
-00009f30: 2061 7070 656e 6469 6e67 2061 6c6c 2072   appending all r
-00009f40: 6573 6964 7565 7320 696e 2074 6865 2070  esidues in the p
-00009f50: 6f6c 7970 6570 7469 6465 0a20 2020 2020  olypeptide.     
-00009f60: 2020 2020 2020 2020 2020 2061 6e67 6c65             angle
-00009f70: 735f 6c69 7374 202b 3d20 7068 6970 7369  s_list += phipsi
-00009f80: 5f6c 6973 740a 2020 2020 2020 2020 2020  _list.          
-00009f90: 2020 666f 7220 7265 7369 6475 6520 696e    for residue in
-00009fa0: 2070 6f6c 793a 0a20 2020 2020 2020 2020   poly:.         
-00009fb0: 2020 2020 2020 2072 6573 6964 7565 5f70         residue_p
-00009fc0: 6f73 6974 696f 6e20 3d20 7265 7369 6475  osition = residu
-00009fd0: 652e 6765 745f 6964 2829 5b31 5d0a 2020  e.get_id()[1].  
-00009fe0: 2020 2020 2020 2020 2020 2020 2020 6966                if
-00009ff0: 2072 6573 6964 7565 5f70 6f73 6974 696f   residue_positio
-0000a000: 6e20 696e 206c 6973 745f 7265 7369 6475  n in list_residu
-0000a010: 6573 3a20 2023 204f 6e6c 7920 6765 7420  es:  # Only get 
-0000a020: 7468 6520 6465 7369 7265 6420 7265 7369  the desired resi
-0000a030: 6475 6573 2028 7468 6f75 6768 2c20 6966  dues (though, if
-0000a040: 2070 6661 6d5f 6469 6374 2069 7320 6162   pfam_dict is ab
-0000a050: 7365 6e74 2069 7420 7769 6c6c 2061 7070  sent it will app
-0000a060: 656e 6420 7468 6520 656e 7469 7265 2070  end the entire p
-0000a070: 6f6c 7970 6570 7469 6465 290a 2020 2020  olypeptide).    
+00009e30: 2020 2d31 5d20 2b20 315d 2020 2320 6368    -1] + 1]  # ch
+00009e40: 6f70 2061 6c73 6f20 7468 6520 616e 676c  op also the angl
+00009e50: 6573 5f6c 6973 7420 6163 636f 7264 696e  es_list accordin
+00009e60: 6720 746f 2074 6865 2073 656c 6563 7465  g to the selecte
+00009e70: 6420 7265 7369 6475 6573 0a20 2020 2020  d residues.     
+00009e80: 2020 2020 2020 2065 6c69 6620 6e6f 7420         elif not 
+00009e90: 636f 7272 6573 706f 6e64 656e 6365 5f69  correspondence_i
+00009ea0: 6e64 6578 2061 6e64 2070 6661 6d5f 6469  ndex and pfam_di
+00009eb0: 6374 3a20 2023 2069 6620 7468 6572 6520  ct:  # if there 
+00009ec0: 6973 206e 6f74 2061 2063 6f72 7265 7370  is not a corresp
+00009ed0: 6f6e 6469 6e67 2072 6573 6964 7565 2064  onding residue d
+00009ee0: 6f20 6e6f 7420 6170 7065 6e64 2061 6e79  o not append any
+00009ef0: 7468 696e 670a 2020 2020 2020 2020 2020  thing.          
+00009f00: 2020 2020 2020 616e 676c 6573 5f6c 6973        angles_lis
+00009f10: 7420 2b3d 205b 5d0a 2020 2020 2020 2020  t += [].        
+00009f20: 2020 2020 656c 7365 3a20 2023 206b 6565      else:  # kee
+00009f30: 7020 6170 7065 6e64 696e 6720 616c 6c20  p appending all 
+00009f40: 7265 7369 6475 6573 2069 6e20 7468 6520  residues in the 
+00009f50: 706f 6c79 7065 7074 6964 650a 2020 2020  polypeptide.    
+00009f60: 2020 2020 2020 2020 2020 2020 616e 676c              angl
+00009f70: 6573 5f6c 6973 7420 2b3d 2070 6869 7073  es_list += phips
+00009f80: 695f 6c69 7374 0a20 2020 2020 2020 2020  i_list.         
+00009f90: 2020 2066 6f72 2072 6573 6964 7565 2069     for residue i
+00009fa0: 6e20 706f 6c79 3a0a 2020 2020 2020 2020  n poly:.        
+00009fb0: 2020 2020 2020 2020 7265 7369 6475 655f          residue_
+00009fc0: 706f 7369 7469 6f6e 203d 2072 6573 6964  position = resid
+00009fd0: 7565 2e67 6574 5f69 6428 295b 315d 0a20  ue.get_id()[1]. 
+00009fe0: 2020 2020 2020 2020 2020 2020 2020 2069                 i
+00009ff0: 6620 7265 7369 6475 655f 706f 7369 7469  f residue_positi
+0000a000: 6f6e 2069 6e20 6c69 7374 5f72 6573 6964  on in list_resid
+0000a010: 7565 733a 2020 2320 4f6e 6c79 2067 6574  ues:  # Only get
+0000a020: 2074 6865 2064 6573 6972 6564 2072 6573   the desired res
+0000a030: 6964 7565 7320 2874 686f 7567 682c 2069  idues (though, i
+0000a040: 6620 7066 616d 5f64 6963 7420 6973 2061  f pfam_dict is a
+0000a050: 6273 656e 7420 6974 2077 696c 6c20 6170  bsent it will ap
+0000a060: 7065 6e64 2074 6865 2065 6e74 6972 6520  pend the entire 
+0000a070: 706f 6c79 7065 7074 6964 6529 0a20 2020  polypeptide).   
 0000a080: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a090: 6161 5f6e 616d 6520 3d20 7072 6f74 6569  aa_name = protei
-0000a0a0: 6e5f 6c65 7474 6572 735f 3374 6f31 5b72  n_letters_3to1[r
-0000a0b0: 6573 6964 7565 2e67 6574 5f72 6573 6e61  esidue.get_resna
-0000a0c0: 6d65 2829 5d0a 2020 2020 2020 2020 2020  me()].          
-0000a0d0: 2020 2020 2020 2020 2020 6966 2061 615f            if aa_
-0000a0e0: 6e61 6d65 206e 6f74 2069 6e20 616d 696e  name not in amin
-0000a0f0: 6f61 6369 645f 6e61 6d65 732e 6b65 7973  oacid_names.keys
-0000a100: 2829 3a0a 2020 2020 2020 2020 2020 2020  ():.            
-0000a110: 2020 2020 2020 2020 2020 2020 7261 6973              rais
-0000a120: 6520 5661 6c75 6545 7272 6f72 2822 506c  e ValueError("Pl
-0000a130: 6561 7365 2073 656c 6563 7420 6161 5f70  ease select aa_p
-0000a140: 726f 6273 203e 2032 3120 2c20 696e 206f  robs > 21 , in o
-0000a150: 7264 6572 2074 6f20 616c 6c6f 7720 7573  rder to allow us
-0000a160: 696e 6720 7370 6563 6961 6c20 616d 696e  ing special amin
-0000a170: 6f20 6163 6964 7322 290a 2020 2020 2020  o acids").      
-0000a180: 2020 2020 2020 2020 2020 2020 2020 6966                if
-0000a190: 206f 6e65 5f68 6f74 5f65 6e63 6f64 696e   one_hot_encodin
-0000a1a0: 673a 0a20 2020 2020 2020 2020 2020 2020  g:.             
-0000a1b0: 2020 2020 2020 2020 2020 2061 615f 6e61             aa_na
-0000a1c0: 6d65 5f69 6e64 6578 203d 2061 6d69 6e6f  me_index = amino
-0000a1d0: 6163 6964 5f6e 616d 6573 5b61 615f 6e61  acid_names[aa_na
-0000a1e0: 6d65 5d20 2330 2d69 6e64 6578 0a20 2020  me] #0-index.   
+0000a090: 2061 615f 6e61 6d65 203d 2070 726f 7465   aa_name = prote
+0000a0a0: 696e 5f6c 6574 7465 7273 5f33 746f 315b  in_letters_3to1[
+0000a0b0: 7265 7369 6475 652e 6765 745f 7265 736e  residue.get_resn
+0000a0c0: 616d 6528 295d 0a20 2020 2020 2020 2020  ame()].         
+0000a0d0: 2020 2020 2020 2020 2020 2069 6620 6161             if aa
+0000a0e0: 5f6e 616d 6520 6e6f 7420 696e 2061 6d69  _name not in ami
+0000a0f0: 6e6f 6163 6964 5f6e 616d 6573 2e6b 6579  noacid_names.key
+0000a100: 7328 293a 0a20 2020 2020 2020 2020 2020  s():.           
+0000a110: 2020 2020 2020 2020 2020 2020 2072 6169               rai
+0000a120: 7365 2056 616c 7565 4572 726f 7228 2250  se ValueError("P
+0000a130: 6c65 6173 6520 7365 6c65 6374 2061 615f  lease select aa_
+0000a140: 7072 6f62 7320 3e20 3231 202c 2069 6e20  probs > 21 , in 
+0000a150: 6f72 6465 7220 746f 2061 6c6c 6f77 2075  order to allow u
+0000a160: 7369 6e67 2073 7065 6369 616c 2061 6d69  sing special ami
+0000a170: 6e6f 2061 6369 6473 2229 0a20 2020 2020  no acids").     
+0000a180: 2020 2020 2020 2020 2020 2020 2020 2069                 i
+0000a190: 6620 6f6e 655f 686f 745f 656e 636f 6469  f one_hot_encodi
+0000a1a0: 6e67 3a0a 2020 2020 2020 2020 2020 2020  ng:.            
+0000a1b0: 2020 2020 2020 2020 2020 2020 6161 5f6e              aa_n
+0000a1c0: 616d 655f 696e 6465 7820 3d20 616d 696e  ame_index = amin
+0000a1d0: 6f61 6369 645f 6e61 6d65 735b 6161 5f6e  oacid_names[aa_n
+0000a1e0: 616d 655d 2023 302d 696e 6465 780a 2020  ame] #0-index.  
 0000a1f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a200: 2020 2020 206f 6e65 5f68 6f74 203d 206e       one_hot = n
-0000a210: 702e 7a65 726f 7328 6161 5f70 726f 6273  p.zeros(aa_probs
-0000a220: 2920 2023 206f 6e65 2070 6f73 6974 696f  )  # one positio
-0000a230: 6e2c 206f 6e65 2063 6861 7261 6374 6572  n, one character
-0000a240: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0000a250: 2020 2020 2020 2020 206f 6e65 5f68 6f74           one_hot
-0000a260: 5b61 615f 6e61 6d65 5f69 6e64 6578 5d20  [aa_name_index] 
-0000a270: 3d20 312e 300a 2020 2020 2020 2020 2020  = 1.0.          
-0000a280: 2020 2020 2020 2020 2020 2020 2020 6161                aa
-0000a290: 5f6c 6973 745f 656d 6265 6464 6564 2e61  _list_embedded.a
-0000a2a0: 7070 656e 6428 6f6e 655f 686f 7429 0a20  ppend(one_hot). 
+0000a200: 2020 2020 2020 6f6e 655f 686f 7420 3d20        one_hot = 
+0000a210: 6e70 2e7a 6572 6f73 2861 615f 7072 6f62  np.zeros(aa_prob
+0000a220: 7329 2020 2320 6f6e 6520 706f 7369 7469  s)  # one positi
+0000a230: 6f6e 2c20 6f6e 6520 6368 6172 6163 7465  on, one characte
+0000a240: 720a 2020 2020 2020 2020 2020 2020 2020  r.              
+0000a250: 2020 2020 2020 2020 2020 6f6e 655f 686f            one_ho
+0000a260: 745b 6161 5f6e 616d 655f 696e 6465 785d  t[aa_name_index]
+0000a270: 203d 2031 2e30 0a20 2020 2020 2020 2020   = 1.0.         
+0000a280: 2020 2020 2020 2020 2020 2020 2020 2061                 a
+0000a290: 615f 6c69 7374 5f65 6d62 6564 6465 642e  a_list_embedded.
+0000a2a0: 6170 7065 6e64 286f 6e65 5f68 6f74 290a  append(one_hot).
 0000a2b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a2c0: 2020 2065 6c73 653a 0a20 2020 2020 2020     else:.       
+0000a2c0: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
 0000a2d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a2e0: 2061 615f 6e61 6d65 5f69 6e64 6578 203d   aa_name_index =
-0000a2f0: 2061 6d69 6e6f 6163 6964 5f6e 616d 6573   aminoacid_names
-0000a300: 5b61 615f 6e61 6d65 5d0a 2020 2020 2020  [aa_name].      
+0000a2e0: 2020 6161 5f6e 616d 655f 696e 6465 7820    aa_name_index 
+0000a2f0: 3d20 616d 696e 6f61 6369 645f 6e61 6d65  = aminoacid_name
+0000a300: 735b 6161 5f6e 616d 655d 0a20 2020 2020  s[aa_name].     
 0000a310: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a320: 2020 6161 5f6c 6973 745f 656d 6265 6464    aa_list_embedd
-0000a330: 6564 2e61 7070 656e 6428 6161 5f6e 616d  ed.append(aa_nam
-0000a340: 655f 696e 6465 7829 0a20 2020 2020 2020  e_index).       
-0000a350: 2020 2020 2020 2020 2020 2020 2061 615f               aa_
-0000a360: 6c69 7374 2e61 7070 656e 6428 6161 5f6e  list.append(aa_n
-0000a370: 616d 6529 0a20 2020 2020 2020 2023 2052  ame).        # R
-0000a380: 6570 6c61 6365 204e 6f6e 6520 7661 6c75  eplace None valu
-0000a390: 6573 2066 726f 6d20 4e54 2061 6e64 2043  es from NT and C
-0000a3a0: 5420 616e 676c 6573 2062 7920 6e70 722e  T angles by npr.
-0000a3b0: 6e6f 726d 616c 286e 702e 7069 2c30 2e31  normal(np.pi,0.1
-0000a3c0: 292d 2d2d 2d3e 2061 6374 7561 6c6c 792c  )----> actually,
-0000a3d0: 2070 6572 6861 7073 206a 7573 7420 6368   perhaps just ch
-0000a3e0: 6f70 2074 6865 6d2c 2074 6f6f 206d 7563  op them, too muc
-0000a3f0: 6820 6d69 7373 696e 666f 726d 6174 696f  h missinformatio
-0000a400: 6e0a 2020 2020 2020 2020 616e 676c 6573  n.        angles
-0000a410: 5f6c 6973 745f 6669 6c6c 6564 203d 205b  _list_filled = [
-0000a420: 5d0a 2020 2020 2020 2020 666f 7220 7470  ].        for tp
-0000a430: 6c20 696e 2061 6e67 6c65 735f 6c69 7374  l in angles_list
-0000a440: 3a0a 2020 2020 2020 2020 2020 2020 7470  :.            tp
-0000a450: 6c20 3d20 6c69 7374 2874 706c 290a 2020  l = list(tpl).  
-0000a460: 2020 2020 2020 2020 2020 6966 204e 6f6e            if Non
-0000a470: 6520 696e 2074 706c 3a0a 2020 2020 2020  e in tpl:.      
-0000a480: 2020 2020 2020 2020 2020 7470 6c5b 6e70            tpl[np
-0000a490: 2e77 6865 7265 286e 702e 6172 7261 7928  .where(np.array(
-0000a4a0: 7470 6c29 203d 3d20 4e6f 6e65 295b 305d  tpl) == None)[0]
-0000a4b0: 5b30 5d5d 203d 206e 7072 2e6e 6f72 6d61  [0]] = npr.norma
-0000a4c0: 6c28 6e70 2e70 692c 2030 2e31 290a 2020  l(np.pi, 0.1).  
-0000a4d0: 2020 2020 2020 2020 2020 616e 676c 6573            angles
-0000a4e0: 5f6c 6973 745f 6669 6c6c 6564 2e61 7070  _list_filled.app
-0000a4f0: 656e 6428 7470 6c29 0a20 2020 2020 2020  end(tpl).       
-0000a500: 2073 6571 5f6c 656e 203d 206c 656e 2861   seq_len = len(a
-0000a510: 6e67 6c65 735f 6c69 7374 5f66 696c 6c65  ngles_list_fille
-0000a520: 6429 0a20 2020 2020 2020 2061 615f 696e  d).        aa_in
-0000a530: 666f 203d 206e 702e 7a65 726f 7328 2873  fo = np.zeros((s
-0000a540: 6571 5f6c 656e 202b 2032 2c20 3330 2929  eq_len + 2, 30))
-0000a550: 2020 2320 7769 6c6c 2063 6f6e 7461 696e    # will contain
-0000a560: 2069 6e66 6f72 6d61 7469 6f6e 2061 626f   information abo
-0000a570: 7574 2061 6c6c 2074 6865 2061 6d69 6e6f  ut all the amino
-0000a580: 6163 6964 7320 696e 2074 6869 7320 6375  acids in this cu
-0000a590: 7272 656e 7420 7365 7175 656e 6365 0a20  rrent sequence. 
-0000a5a0: 2020 2020 2020 2061 615f 696e 666f 5b30         aa_info[0
-0000a5b0: 5d20 3d20 5b73 6571 5f6c 656e 5d20 2b20  ] = [seq_len] + 
-0000a5c0: 5b30 5d20 2a20 3239 0a20 2020 2020 2020  [0] * 29.       
-0000a5d0: 2066 6f72 2069 6e64 6578 2069 6e20 7261   for index in ra
-0000a5e0: 6e67 6528 322c 7365 715f 6c65 6e20 2b20  nge(2,seq_len + 
-0000a5f0: 3229 3a20 2023 2046 6972 7374 2064 696d  2):  # First dim
-0000a600: 656e 7369 6f6e 2063 6f6e 7461 696e 7320  ension contains 
-0000a610: 736f 6d65 2063 7573 746f 6d20 696e 666f  some custom info
-0000a620: 726d 6174 696f 6e20 2869 2e65 206e 616d  rmation (i.e nam
-0000a630: 6529 2c20 7365 636f 6e64 2069 7320 7468  e), second is th
-0000a640: 6520 6769 7420 7665 6374 6f72 2061 6e64  e git vector and
-0000a650: 2074 6865 2072 6573 7420 7368 6f75 6c64   the rest should
-0000a660: 2068 6176 6520 7468 6520 6161 2074 7970   have the aa typ
-0000a670: 6520 616e 6420 7468 6520 616e 676c 6573  e and the angles
-0000a680: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
-0000a690: 6f6e 655f 686f 745f 656e 636f 6469 6e67  one_hot_encoding
-0000a6a0: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
-0000a6b0: 2020 6161 5f69 6e66 6f5b 696e 6465 785d    aa_info[index]
-0000a6c0: 203d 206e 702e 6873 7461 636b 285b 6161   = np.hstack([aa
-0000a6d0: 5f6c 6973 745f 656d 6265 6464 6564 5b69  _list_embedded[i
-0000a6e0: 6e64 6578 202d 2032 5d2c 2061 6e67 6c65  ndex - 2], angle
-0000a6f0: 735f 6c69 7374 5f66 696c 6c65 645b 696e  s_list_filled[in
-0000a700: 6465 7820 2d20 325d 2c20 5b30 5d20 2a20  dex - 2], [0] * 
-0000a710: 375d 290a 2020 2020 2020 2020 2020 2020  7]).            
-0000a720: 656c 7365 3a0a 2020 2020 2020 2020 2020  else:.          
-0000a730: 2020 2020 2020 6161 5f69 6e66 6f5b 696e        aa_info[in
-0000a740: 6465 785d 203d 206e 702e 6873 7461 636b  dex] = np.hstack
-0000a750: 285b 6161 5f6c 6973 745f 656d 6265 6464  ([aa_list_embedd
-0000a760: 6564 5b69 6e64 6578 202d 2032 5d2c 2061  ed[index - 2], a
-0000a770: 6e67 6c65 735f 6c69 7374 5f66 696c 6c65  ngles_list_fille
-0000a780: 645b 696e 6465 7820 2d20 325d 2c20 5b30  d[index - 2], [0
-0000a790: 5d20 2a20 3237 5d29 0a20 2020 2020 2020  ] * 27]).       
-0000a7a0: 2069 6620 7365 715f 6c65 6e20 3e20 6d69   if seq_len > mi
-0000a7b0: 6e5f 6c65 6e20 616e 6420 6e6f 7420 616c  n_len and not al
-0000a7c0: 6c28 7620 3d3d 2028 4e6f 6e65 2c20 4e6f  l(v == (None, No
-0000a7d0: 6e65 2920 666f 7220 7620 696e 2061 6e67  ne) for v in ang
-0000a7e0: 6c65 735f 6c69 7374 293a 2020 2320 736b  les_list):  # sk
-0000a7f0: 6970 2074 6865 2070 726f 7465 696e 7320  ip the proteins 
-0000a800: 7468 6174 2061 7265 2065 6d70 7479 206f  that are empty o
-0000a810: 7220 746f 6f20 736d 616c 6c20 7c20 736b  r too small | sk
-0000a820: 6970 2074 6865 2070 726f 7465 696e 7320  ip the proteins 
-0000a830: 7769 7468 2061 6c6c 204e 6f6e 6520 7661  with all None va
-0000a840: 6c75 6573 2069 6e20 7468 6520 616e 676c  lues in the angl
-0000a850: 6573 0a20 2020 2020 2020 2020 2020 2069  es.            i
-0000a860: 6620 2222 2e6a 6f69 6e28 6161 5f6c 6973  f "".join(aa_lis
-0000a870: 7429 206e 6f74 2069 6e20 6475 706c 6963  t) not in duplic
-0000a880: 6174 6573 3a0a 2020 2020 2020 2020 2020  ates:.          
-0000a890: 2020 2020 2020 7072 6f74 5f69 6e66 6f5f        prot_info_
-0000a8a0: 6469 6374 5b66 696c 6573 5f6c 6973 745b  dict[files_list[
-0000a8b0: 695d 5b33 3a37 5d5d 203d 2061 615f 696e  i][3:7]] = aa_in
-0000a8c0: 666f 0a20 2020 2020 2020 2020 2020 2020  fo.             
-0000a8d0: 2020 2070 726f 745f 6161 5f64 6963 745b     prot_aa_dict[
-0000a8e0: 6669 6c65 735f 6c69 7374 5b69 5d5b 333a  files_list[i][3:
-0000a8f0: 375d 5d20 3d20 6161 5f6c 6973 740a 2020  7]] = aa_list.  
-0000a900: 2020 2020 2020 2020 2020 2020 2020 6475                du
-0000a910: 706c 6963 6174 6573 2e61 7070 656e 6428  plicates.append(
-0000a920: 2222 2e6a 6f69 6e28 6161 5f6c 6973 7429  "".join(aa_list)
-0000a930: 290a 2020 2020 7265 7475 726e 2070 726f  ).    return pro
-0000a940: 745f 6161 5f64 6963 742c 7072 6f74 5f69  t_aa_dict,prot_i
-0000a950: 6e66 6f5f 6469 6374 0a64 6566 2063 7265  nfo_dict.def cre
-0000a960: 6174 655f 6461 7461 7365 7428 6e61 6d65  ate_dataset(name
-0000a970: 5f66 696c 652c 0a20 2020 2020 2020 2020  _file,.         
-0000a980: 2020 2020 2020 2020 2020 6f6e 655f 686f            one_ho
-0000a990: 745f 656e 636f 6469 6e67 2c0a 2020 2020  t_encoding,.    
-0000a9a0: 2020 2020 2020 2020 2020 2020 2020 206d                 m
-0000a9b0: 696e 5f6c 656e 3d33 302c 0a20 2020 2020  in_len=30,.     
-0000a9c0: 2020 2020 2020 2020 2020 2020 2020 6661                fa
-0000a9d0: 7374 615f 6669 6c65 3d4e 6f6e 652c 0a20  sta_file=None,. 
+0000a320: 2020 2061 615f 6c69 7374 5f65 6d62 6564     aa_list_embed
+0000a330: 6465 642e 6170 7065 6e64 2861 615f 6e61  ded.append(aa_na
+0000a340: 6d65 5f69 6e64 6578 290a 2020 2020 2020  me_index).      
+0000a350: 2020 2020 2020 2020 2020 2020 2020 6161                aa
+0000a360: 5f6c 6973 742e 6170 7065 6e64 2861 615f  _list.append(aa_
+0000a370: 6e61 6d65 290a 2020 2020 2020 2020 2320  name).        # 
+0000a380: 5265 706c 6163 6520 4e6f 6e65 2076 616c  Replace None val
+0000a390: 7565 7320 6672 6f6d 204e 5420 616e 6420  ues from NT and 
+0000a3a0: 4354 2061 6e67 6c65 7320 6279 206e 7072  CT angles by npr
+0000a3b0: 2e6e 6f72 6d61 6c28 6e70 2e70 692c 302e  .normal(np.pi,0.
+0000a3c0: 3129 2d2d 2d2d 3e20 6163 7475 616c 6c79  1)----> actually
+0000a3d0: 2c20 7065 7268 6170 7320 6a75 7374 2063  , perhaps just c
+0000a3e0: 686f 7020 7468 656d 2c20 746f 6f20 6d75  hop them, too mu
+0000a3f0: 6368 206d 6973 7369 6e66 6f72 6d61 7469  ch missinformati
+0000a400: 6f6e 0a20 2020 2020 2020 2061 6e67 6c65  on.        angle
+0000a410: 735f 6c69 7374 5f66 696c 6c65 6420 3d20  s_list_filled = 
+0000a420: 5b5d 0a20 2020 2020 2020 2066 6f72 2074  [].        for t
+0000a430: 706c 2069 6e20 616e 676c 6573 5f6c 6973  pl in angles_lis
+0000a440: 743a 0a20 2020 2020 2020 2020 2020 2074  t:.            t
+0000a450: 706c 203d 206c 6973 7428 7470 6c29 0a20  pl = list(tpl). 
+0000a460: 2020 2020 2020 2020 2020 2069 6620 4e6f             if No
+0000a470: 6e65 2069 6e20 7470 6c3a 0a20 2020 2020  ne in tpl:.     
+0000a480: 2020 2020 2020 2020 2020 2074 706c 5b6e             tpl[n
+0000a490: 702e 7768 6572 6528 6e70 2e61 7272 6179  p.where(np.array
+0000a4a0: 2874 706c 2920 3d3d 204e 6f6e 6529 5b30  (tpl) == None)[0
+0000a4b0: 5d5b 305d 5d20 3d20 6e70 722e 6e6f 726d  ][0]] = npr.norm
+0000a4c0: 616c 286e 702e 7069 2c20 302e 3129 0a20  al(np.pi, 0.1). 
+0000a4d0: 2020 2020 2020 2020 2020 2061 6e67 6c65             angle
+0000a4e0: 735f 6c69 7374 5f66 696c 6c65 642e 6170  s_list_filled.ap
+0000a4f0: 7065 6e64 2874 706c 290a 2020 2020 2020  pend(tpl).      
+0000a500: 2020 7365 715f 6c65 6e20 3d20 6c65 6e28    seq_len = len(
+0000a510: 616e 676c 6573 5f6c 6973 745f 6669 6c6c  angles_list_fill
+0000a520: 6564 290a 2020 2020 2020 2020 6161 5f69  ed).        aa_i
+0000a530: 6e66 6f20 3d20 6e70 2e7a 6572 6f73 2828  nfo = np.zeros((
+0000a540: 7365 715f 6c65 6e20 2b20 322c 2033 3029  seq_len + 2, 30)
+0000a550: 2920 2023 2077 696c 6c20 636f 6e74 6169  )  # will contai
+0000a560: 6e20 696e 666f 726d 6174 696f 6e20 6162  n information ab
+0000a570: 6f75 7420 616c 6c20 7468 6520 616d 696e  out all the amin
+0000a580: 6f61 6369 6473 2069 6e20 7468 6973 2063  oacids in this c
+0000a590: 7572 7265 6e74 2073 6571 7565 6e63 650a  urrent sequence.
+0000a5a0: 2020 2020 2020 2020 6161 5f69 6e66 6f5b          aa_info[
+0000a5b0: 305d 203d 205b 7365 715f 6c65 6e5d 202b  0] = [seq_len] +
+0000a5c0: 205b 305d 202a 2032 390a 2020 2020 2020   [0] * 29.      
+0000a5d0: 2020 666f 7220 696e 6465 7820 696e 2072    for index in r
+0000a5e0: 616e 6765 2832 2c73 6571 5f6c 656e 202b  ange(2,seq_len +
+0000a5f0: 2032 293a 2020 2320 4669 7273 7420 6469   2):  # First di
+0000a600: 6d65 6e73 696f 6e20 636f 6e74 6169 6e73  mension contains
+0000a610: 2073 6f6d 6520 6375 7374 6f6d 2069 6e66   some custom inf
+0000a620: 6f72 6d61 7469 6f6e 2028 692e 6520 6e61  ormation (i.e na
+0000a630: 6d65 292c 2073 6563 6f6e 6420 6973 2074  me), second is t
+0000a640: 6865 2067 6974 2076 6563 746f 7220 616e  he git vector an
+0000a650: 6420 7468 6520 7265 7374 2073 686f 756c  d the rest shoul
+0000a660: 6420 6861 7665 2074 6865 2061 6120 7479  d have the aa ty
+0000a670: 7065 2061 6e64 2074 6865 2061 6e67 6c65  pe and the angle
+0000a680: 730a 2020 2020 2020 2020 2020 2020 6966  s.            if
+0000a690: 206f 6e65 5f68 6f74 5f65 6e63 6f64 696e   one_hot_encodin
+0000a6a0: 673a 0a20 2020 2020 2020 2020 2020 2020  g:.             
+0000a6b0: 2020 2061 615f 696e 666f 5b69 6e64 6578     aa_info[index
+0000a6c0: 5d20 3d20 6e70 2e68 7374 6163 6b28 5b61  ] = np.hstack([a
+0000a6d0: 615f 6c69 7374 5f65 6d62 6564 6465 645b  a_list_embedded[
+0000a6e0: 696e 6465 7820 2d20 325d 2c20 616e 676c  index - 2], angl
+0000a6f0: 6573 5f6c 6973 745f 6669 6c6c 6564 5b69  es_list_filled[i
+0000a700: 6e64 6578 202d 2032 5d2c 205b 305d 202a  ndex - 2], [0] *
+0000a710: 2037 5d29 0a20 2020 2020 2020 2020 2020   7]).           
+0000a720: 2065 6c73 653a 0a20 2020 2020 2020 2020   else:.         
+0000a730: 2020 2020 2020 2061 615f 696e 666f 5b69         aa_info[i
+0000a740: 6e64 6578 5d20 3d20 6e70 2e68 7374 6163  ndex] = np.hstac
+0000a750: 6b28 5b61 615f 6c69 7374 5f65 6d62 6564  k([aa_list_embed
+0000a760: 6465 645b 696e 6465 7820 2d20 325d 2c20  ded[index - 2], 
+0000a770: 616e 676c 6573 5f6c 6973 745f 6669 6c6c  angles_list_fill
+0000a780: 6564 5b69 6e64 6578 202d 2032 5d2c 205b  ed[index - 2], [
+0000a790: 305d 202a 2032 375d 290a 2020 2020 2020  0] * 27]).      
+0000a7a0: 2020 6966 2073 6571 5f6c 656e 203e 206d    if seq_len > m
+0000a7b0: 696e 5f6c 656e 2061 6e64 206e 6f74 2061  in_len and not a
+0000a7c0: 6c6c 2876 203d 3d20 284e 6f6e 652c 204e  ll(v == (None, N
+0000a7d0: 6f6e 6529 2066 6f72 2076 2069 6e20 616e  one) for v in an
+0000a7e0: 676c 6573 5f6c 6973 7429 3a20 2023 2073  gles_list):  # s
+0000a7f0: 6b69 7020 7468 6520 7072 6f74 6569 6e73  kip the proteins
+0000a800: 2074 6861 7420 6172 6520 656d 7074 7920   that are empty 
+0000a810: 6f72 2074 6f6f 2073 6d61 6c6c 207c 2073  or too small | s
+0000a820: 6b69 7020 7468 6520 7072 6f74 6569 6e73  kip the proteins
+0000a830: 2077 6974 6820 616c 6c20 4e6f 6e65 2076   with all None v
+0000a840: 616c 7565 7320 696e 2074 6865 2061 6e67  alues in the ang
+0000a850: 6c65 730a 2020 2020 2020 2020 2020 2020  les.            
+0000a860: 6966 2022 222e 6a6f 696e 2861 615f 6c69  if "".join(aa_li
+0000a870: 7374 2920 6e6f 7420 696e 2064 7570 6c69  st) not in dupli
+0000a880: 6361 7465 733a 0a20 2020 2020 2020 2020  cates:.         
+0000a890: 2020 2020 2020 2070 726f 745f 696e 666f         prot_info
+0000a8a0: 5f64 6963 745b 6669 6c65 735f 6c69 7374  _dict[files_list
+0000a8b0: 5b69 5d5b 333a 375d 5d20 3d20 6161 5f69  [i][3:7]] = aa_i
+0000a8c0: 6e66 6f0a 2020 2020 2020 2020 2020 2020  nfo.            
+0000a8d0: 2020 2020 7072 6f74 5f61 615f 6469 6374      prot_aa_dict
+0000a8e0: 5b66 696c 6573 5f6c 6973 745b 695d 5b33  [files_list[i][3
+0000a8f0: 3a37 5d5d 203d 2061 615f 6c69 7374 0a20  :7]] = aa_list. 
+0000a900: 2020 2020 2020 2020 2020 2020 2020 2064                 d
+0000a910: 7570 6c69 6361 7465 732e 6170 7065 6e64  uplicates.append
+0000a920: 2822 222e 6a6f 696e 2861 615f 6c69 7374  ("".join(aa_list
+0000a930: 2929 0a20 2020 2072 6574 7572 6e20 7072  )).    return pr
+0000a940: 6f74 5f61 615f 6469 6374 2c70 726f 745f  ot_aa_dict,prot_
+0000a950: 696e 666f 5f64 6963 740a 6465 6620 6372  info_dict.def cr
+0000a960: 6561 7465 5f64 6174 6173 6574 286e 616d  eate_dataset(nam
+0000a970: 655f 6669 6c65 2c0a 2020 2020 2020 2020  e_file,.        
+0000a980: 2020 2020 2020 2020 2020 206f 6e65 5f68             one_h
+0000a990: 6f74 5f65 6e63 6f64 696e 672c 0a20 2020  ot_encoding,.   
+0000a9a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000a9b0: 6d69 6e5f 6c65 6e3d 3330 2c0a 2020 2020  min_len=30,.    
+0000a9c0: 2020 2020 2020 2020 2020 2020 2020 2066                 f
+0000a9d0: 6173 7461 5f66 696c 653d 4e6f 6e65 2c0a  asta_file=None,.
 0000a9e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000a9f0: 2020 5044 425f 666f 6c64 6572 3d4e 6f6e    PDB_folder=Non
-0000aa00: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
-0000aa10: 2020 2020 2020 616c 6967 6e6d 656e 745f        alignment_
-0000aa20: 6669 6c65 3d4e 6f6e 652c 0a20 2020 2020  file=None,.     
-0000aa30: 2020 2020 2020 2020 2020 2020 2020 7472                tr
-0000aa40: 6565 5f66 696c 6520 3d20 4e6f 6e65 2c0a  ee_file = None,.
-0000aa50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000aa60: 2020 2070 6661 6d5f 6469 6374 3d20 4e6f     pfam_dict= No
-0000aa70: 6e65 2c0a 2020 2020 2020 2020 2020 2020  ne,.            
-0000aa80: 2020 2020 2020 206d 6574 686f 643d 2269         method="i
-0000aa90: 7174 7265 6522 2c0a 2020 2020 2020 2020  qtree",.        
-0000aaa0: 2020 2020 2020 2020 2020 2061 615f 7072             aa_pr
-0000aab0: 6f62 733d 3231 2c0a 2020 2020 2020 2020  obs=21,.        
-0000aac0: 2020 2020 2020 2020 2020 2072 656e 616d             renam
-0000aad0: 655f 696e 7465 726e 616c 5f6e 6f64 6573  e_internal_nodes
-0000aae0: 3d46 616c 7365 2c0a 2020 2020 2020 2020  =False,.        
-0000aaf0: 2020 2020 2020 2020 2020 2073 746f 7261             stora
-0000ab00: 6765 5f66 6f6c 6465 723d 2222 293a 0a20  ge_folder=""):. 
-0000ab10: 2020 2022 2222 2043 6f6d 706c 6578 2066     """ Complex f
-0000ab20: 756e 6374 696f 6e20 746f 2063 7265 6174  unction to creat
-0000ab30: 6520 7468 6520 6461 7461 7365 7420 616e  e the dataset an
-0000ab40: 6420 6164 6469 7469 6f6e 616c 2066 696c  d additional fil
-0000ab50: 6573 2028 692e 6520 6469 6374 696f 6e61  es (i.e dictiona
-0000ab60: 7269 6573 2920 7468 6174 2044 7261 7570  ries) that Draup
-0000ab70: 6e69 7220 7573 6573 2066 6f72 2069 6e66  nir uses for inf
-0000ab80: 6572 656e 6365 0a20 2020 2069 6e3a 0a20  erence.    in:. 
-0000ab90: 2020 2020 2020 203a 7061 7261 6d20 7374         :param st
-0000aba0: 7220 6e61 6d65 5f66 696c 6520 3a20 6461  r name_file : da
-0000abb0: 7461 7365 7420 6e61 6d65 0a20 2020 2020  taset name.     
-0000abc0: 2020 203a 7061 7261 6d20 626f 6f6c 206f     :param bool o
-0000abd0: 6e65 5f68 6f74 5f65 6e63 6f64 696e 673a  ne_hot_encoding:
-0000abe0: 207b 5472 7565 2c46 616c 7365 7d20 5741   {True,False} WA
-0000abf0: 524e 494e 473a 204f 6e65 2068 6f74 2065  RNING: One hot e
-0000ac00: 6e63 6f64 696e 6720 6973 2066 6175 6c74  ncoding is fault
-0000ac10: 792c 206e 6565 6473 2074 6f20 6265 2066  y, needs to be f
-0000ac20: 6978 6564 2c20 444f 204e 4f54 2055 5345  ixed, DO NOT USE
-0000ac30: 0a20 2020 2020 2020 203a 7061 7261 6d20  .        :param 
-0000ac40: 696e 7420 6d69 6e5f 6c65 6e3a 206d 696e  int min_len: min
-0000ac50: 696d 756d 206c 656e 6774 6820 6f66 2074  imum length of t
-0000ac60: 6865 2073 6571 7565 6e63 652c 2064 726f  he sequence, dro
-0000ac70: 7073 206f 7574 2073 6571 7565 6e63 6573  ps out sequences
-0000ac80: 2073 6d61 6c6c 6572 2074 6861 6e20 7468   smaller than th
-0000ac90: 6973 0a20 2020 2020 2020 203a 7061 7261  is.        :para
-0000aca0: 6d20 7374 7220 6661 7374 615f 6669 6c65  m str fasta_file
-0000acb0: 3a20 7061 7468 2074 6f20 6661 7374 6120  : path to fasta 
-0000acc0: 7769 7468 2075 6e61 6c69 676e 6564 2073  with unaligned s
-0000acd0: 6571 7565 6e63 6573 0a20 2020 2020 2020  equences.       
-0000ace0: 203a 7061 7261 6d20 7374 7220 6f72 204e   :param str or N
-0000acf0: 6f6e 6520 5044 425f 666f 6c64 6572 3a20  one PDB_folder: 
-0000ad00: 466f 6c64 6572 2077 6974 6820 5044 4220  Folder with PDB 
-0000ad10: 6669 6c65 7320 6672 6f6d 2077 6865 7265  files from where
-0000ad20: 2074 6f20 6578 7472 6163 7420 7365 7175   to extract sequ
-0000ad30: 656e 6365 7320 616e 6420 616e 676c 6573  ences and angles
-0000ad40: 0a20 2020 2020 2020 203a 7061 7261 6d20  .        :param 
-0000ad50: 7374 7220 6f72 204e 6f6e 6520 616c 6967  str or None alig
-0000ad60: 6e6d 656e 745f 6669 6c65 3a20 7061 7468  nment_file: path
-0000ad70: 2074 6f20 6661 7374 6120 7769 7468 2061   to fasta with a
-0000ad80: 6c69 676e 6564 2073 6571 7565 6e63 6573  ligned sequences
-0000ad90: 0a20 2020 2020 2020 203a 7061 7261 6d20  .        :param 
-0000ada0: 7374 7220 6f72 204e 6f6e 6520 7472 6565  str or None tree
-0000adb0: 5f66 696c 653a 2070 6174 6820 746f 206e  _file: path to n
-0000adc0: 6577 6963 6b20 7472 6565 2c20 666f 726d  ewick tree, form
-0000add0: 6174 2031 2028 6574 6533 206e 6f6d 656e  at 1 (ete3 nomen
-0000ade0: 636c 6174 7572 6529 0a20 2020 2020 2020  clature).       
-0000adf0: 203a 7061 7261 6d20 6469 6374 206f 7220   :param dict or 
-0000ae00: 4e6f 6e65 2070 6661 6d5f 6469 6374 3a20  None pfam_dict: 
-0000ae10: 6469 6374 696f 6e61 7279 2077 6974 6820  dictionary with 
-0000ae20: 5044 4220 6669 6c65 7320 6e61 6d65 730a  PDB files names.
-0000ae30: 2020 2020 2020 2020 3a70 6172 616d 2073          :param s
-0000ae40: 7472 206d 6574 686f 643a 2074 7265 6520  tr method: tree 
-0000ae50: 696e 6665 7265 6e63 6520 6d65 7468 6f64  inference method
-0000ae60: 6f6c 6f67 792c 0a20 2020 2020 2020 2020  ology,.         
+0000a9f0: 2020 2050 4442 5f66 6f6c 6465 723d 4e6f     PDB_folder=No
+0000aa00: 6e65 2c0a 2020 2020 2020 2020 2020 2020  ne,.            
+0000aa10: 2020 2020 2020 2061 6c69 676e 6d65 6e74         alignment
+0000aa20: 5f66 696c 653d 4e6f 6e65 2c0a 2020 2020  _file=None,.    
+0000aa30: 2020 2020 2020 2020 2020 2020 2020 2074                 t
+0000aa40: 7265 655f 6669 6c65 203d 204e 6f6e 652c  ree_file = None,
+0000aa50: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000aa60: 2020 2020 7066 616d 5f64 6963 743d 204e      pfam_dict= N
+0000aa70: 6f6e 652c 0a20 2020 2020 2020 2020 2020  one,.           
+0000aa80: 2020 2020 2020 2020 6d65 7468 6f64 3d22          method="
+0000aa90: 6971 7472 6565 222c 0a20 2020 2020 2020  iqtree",.       
+0000aaa0: 2020 2020 2020 2020 2020 2020 6161 5f70              aa_p
+0000aab0: 726f 6273 3d32 312c 0a20 2020 2020 2020  robs=21,.       
+0000aac0: 2020 2020 2020 2020 2020 2020 7265 6e61              rena
+0000aad0: 6d65 5f69 6e74 6572 6e61 6c5f 6e6f 6465  me_internal_node
+0000aae0: 733d 4661 6c73 652c 0a20 2020 2020 2020  s=False,.       
+0000aaf0: 2020 2020 2020 2020 2020 2020 7374 6f72              stor
+0000ab00: 6167 655f 666f 6c64 6572 3d22 2229 3a0a  age_folder=""):.
+0000ab10: 2020 2020 2222 2220 436f 6d70 6c65 7820      """ Complex 
+0000ab20: 6675 6e63 7469 6f6e 2074 6f20 6372 6561  function to crea
+0000ab30: 7465 2074 6865 2064 6174 6173 6574 2061  te the dataset a
+0000ab40: 6e64 2061 6464 6974 696f 6e61 6c20 6669  nd additional fi
+0000ab50: 6c65 7320 2869 2e65 2064 6963 7469 6f6e  les (i.e diction
+0000ab60: 6172 6965 7329 2074 6861 7420 4472 6175  aries) that Drau
+0000ab70: 706e 6972 2075 7365 7320 666f 7220 696e  pnir uses for in
+0000ab80: 6665 7265 6e63 650a 2020 2020 696e 3a0a  ference.    in:.
+0000ab90: 2020 2020 2020 2020 3a70 6172 616d 2073          :param s
+0000aba0: 7472 206e 616d 655f 6669 6c65 203a 2064  tr name_file : d
+0000abb0: 6174 6173 6574 206e 616d 650a 2020 2020  ataset name.    
+0000abc0: 2020 2020 3a70 6172 616d 2062 6f6f 6c20      :param bool 
+0000abd0: 6f6e 655f 686f 745f 656e 636f 6469 6e67  one_hot_encoding
+0000abe0: 3a20 7b54 7275 652c 4661 6c73 657d 2057  : {True,False} W
+0000abf0: 4152 4e49 4e47 3a20 4f6e 6520 686f 7420  ARNING: One hot 
+0000ac00: 656e 636f 6469 6e67 2069 7320 6661 756c  encoding is faul
+0000ac10: 7479 2c20 6e65 6564 7320 746f 2062 6520  ty, needs to be 
+0000ac20: 6669 7865 642c 2044 4f20 4e4f 5420 5553  fixed, DO NOT US
+0000ac30: 450a 2020 2020 2020 2020 3a70 6172 616d  E.        :param
+0000ac40: 2069 6e74 206d 696e 5f6c 656e 3a20 6d69   int min_len: mi
+0000ac50: 6e69 6d75 6d20 6c65 6e67 7468 206f 6620  nimum length of 
+0000ac60: 7468 6520 7365 7175 656e 6365 2c20 6472  the sequence, dr
+0000ac70: 6f70 7320 6f75 7420 7365 7175 656e 6365  ops out sequence
+0000ac80: 7320 736d 616c 6c65 7220 7468 616e 2074  s smaller than t
+0000ac90: 6869 730a 2020 2020 2020 2020 3a70 6172  his.        :par
+0000aca0: 616d 2073 7472 2066 6173 7461 5f66 696c  am str fasta_fil
+0000acb0: 653a 2070 6174 6820 746f 2066 6173 7461  e: path to fasta
+0000acc0: 2077 6974 6820 756e 616c 6967 6e65 6420   with unaligned 
+0000acd0: 7365 7175 656e 6365 730a 2020 2020 2020  sequences.      
+0000ace0: 2020 3a70 6172 616d 2073 7472 206f 7220    :param str or 
+0000acf0: 4e6f 6e65 2050 4442 5f66 6f6c 6465 723a  None PDB_folder:
+0000ad00: 2046 6f6c 6465 7220 7769 7468 2050 4442   Folder with PDB
+0000ad10: 2066 696c 6573 2066 726f 6d20 7768 6572   files from wher
+0000ad20: 6520 746f 2065 7874 7261 6374 2073 6571  e to extract seq
+0000ad30: 7565 6e63 6573 2061 6e64 2061 6e67 6c65  uences and angle
+0000ad40: 730a 2020 2020 2020 2020 3a70 6172 616d  s.        :param
+0000ad50: 2073 7472 206f 7220 4e6f 6e65 2061 6c69   str or None ali
+0000ad60: 676e 6d65 6e74 5f66 696c 653a 2070 6174  gnment_file: pat
+0000ad70: 6820 746f 2066 6173 7461 2077 6974 6820  h to fasta with 
+0000ad80: 616c 6967 6e65 6420 7365 7175 656e 6365  aligned sequence
+0000ad90: 730a 2020 2020 2020 2020 3a70 6172 616d  s.        :param
+0000ada0: 2073 7472 206f 7220 4e6f 6e65 2074 7265   str or None tre
+0000adb0: 655f 6669 6c65 3a20 7061 7468 2074 6f20  e_file: path to 
+0000adc0: 6e65 7769 636b 2074 7265 652c 2066 6f72  newick tree, for
+0000add0: 6d61 7420 3120 2865 7465 3320 6e6f 6d65  mat 1 (ete3 nome
+0000ade0: 6e63 6c61 7475 7265 290a 2020 2020 2020  nclature).      
+0000adf0: 2020 3a70 6172 616d 2064 6963 7420 6f72    :param dict or
+0000ae00: 204e 6f6e 6520 7066 616d 5f64 6963 743a   None pfam_dict:
+0000ae10: 2064 6963 7469 6f6e 6172 7920 7769 7468   dictionary with
+0000ae20: 2050 4442 2066 696c 6573 206e 616d 6573   PDB files names
+0000ae30: 0a20 2020 2020 2020 203a 7061 7261 6d20  .        :param 
+0000ae40: 7374 7220 6d65 7468 6f64 3a20 7472 6565  str method: tree
+0000ae50: 2069 6e66 6572 656e 6365 206d 6574 686f   inference metho
+0000ae60: 646f 6c6f 6779 2c0a 2020 2020 2020 2020  dology,.        
 0000ae70: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000ae80: 2022 6971 7472 6565 223a 2066 6f72 204d   "iqtree": for M
-0000ae90: 4c20 7472 6565 2069 6e66 6572 656e 6365  L tree inference
-0000aea0: 2062 7920 4951 7472 6565 2028 6d61 6b65   by IQtree (make
-0000aeb0: 2073 7572 6520 6973 2069 6e73 7461 6c6c   sure is install
-0000aec0: 6564 2067 6c6f 6261 6c6c 7929 2c0a 2020  ed globally),.  
+0000ae80: 2020 2269 7174 7265 6522 3a20 666f 7220    "iqtree": for 
+0000ae90: 4d4c 2074 7265 6520 696e 6665 7265 6e63  ML tree inferenc
+0000aea0: 6520 6279 2049 5174 7265 6520 286d 616b  e by IQtree (mak
+0000aeb0: 6520 7375 7265 2069 7320 696e 7374 616c  e sure is instal
+0000aec0: 6c65 6420 676c 6f62 616c 6c79 292c 0a20  led globally),. 
 0000aed0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000aee0: 2020 2020 2020 2020 226e 6a22 3a20 666f          "nj": fo
-0000aef0: 7220 6e65 6967 6862 6f75 7220 6a6f 696e  r neighbour join
-0000af00: 696e 6720 756e 726f 6f74 6564 2074 7265  ing unrooted tre
-0000af10: 6520 696e 6665 7265 6e63 6520 2862 696f  e inference (bio
-0000af20: 7079 7468 6f6e 292c 0a20 2020 2020 2020  python),.       
+0000aee0: 2020 2020 2020 2020 2022 6e6a 223a 2066           "nj": f
+0000aef0: 6f72 206e 6569 6768 626f 7572 206a 6f69  or neighbour joi
+0000af00: 6e69 6e67 2075 6e72 6f6f 7465 6420 7472  ning unrooted tr
+0000af10: 6565 2069 6e66 6572 656e 6365 2028 6269  ee inference (bi
+0000af20: 6f70 7974 686f 6e29 2c0a 2020 2020 2020  opython),.      
 0000af30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000af40: 2020 2022 6e6a 5f72 6f6f 7465 6422 3a20     "nj_rooted": 
-0000af50: 666f 7220 4e4a 2072 6f6f 7465 6420 7472  for NJ rooted tr
-0000af60: 6565 2069 6e66 6572 656e 6365 2028 7365  ee inference (se
-0000af70: 6c65 6374 7320 6120 726f 6f74 2062 6173  lects a root bas
-0000af80: 6564 206f 6e20 7468 6520 6469 7374 616e  ed on the distan
-0000af90: 6365 7320 6265 6574 7765 6e20 6e6f 6465  ces beetwen node
-0000afa0: 7329 2028 6269 6f70 7974 686f 6e29 2c0a  s) (biopython),.
-0000afb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000afc0: 2020 2020 2020 2020 2020 2275 7067 6d61            "upgma
-0000afd0: 223a 2055 5047 4d41 2028 6269 6f70 7974  ": UPGMA (biopyt
-0000afe0: 686f 6e29 2c0a 2020 2020 2020 2020 2020  hon),.          
+0000af40: 2020 2020 226e 6a5f 726f 6f74 6564 223a      "nj_rooted":
+0000af50: 2066 6f72 204e 4a20 726f 6f74 6564 2074   for NJ rooted t
+0000af60: 7265 6520 696e 6665 7265 6e63 6520 2873  ree inference (s
+0000af70: 656c 6563 7473 2061 2072 6f6f 7420 6261  elects a root ba
+0000af80: 7365 6420 6f6e 2074 6865 2064 6973 7461  sed on the dista
+0000af90: 6e63 6573 2062 6565 7477 656e 206e 6f64  nces beetwen nod
+0000afa0: 6573 2920 2862 696f 7079 7468 6f6e 292c  es) (biopython),
+0000afb0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000afc0: 2020 2020 2020 2020 2020 2022 7570 676d             "upgm
+0000afd0: 6122 3a20 5550 474d 4120 2862 696f 7079  a": UPGMA (biopy
+0000afe0: 7468 6f6e 292c 0a20 2020 2020 2020 2020  thon),.         
 0000aff0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000b000: 2272 6170 6964 6e6a 223a 2069 6e66 6572  "rapidnj": infer
-0000b010: 656e 6365 206f 6620 4661 7374 204e 4a20  ence of Fast NJ 
-0000b020: 756e 726f 6f74 6564 2069 6e66 6572 656e  unrooted inferen
-0000b030: 6365 2028 6d61 6b65 2073 7572 6520 6973  ce (make sure is
-0000b040: 2069 6e73 7461 6c6c 6564 2067 6c6f 6261   installed globa
-0000b050: 6c6c 7929 2c0a 2020 2020 2020 2020 6161  lly),.        aa
-0000b060: 5f70 726f 6273 3a20 616d 696e 6f20 6163  _probs: amino ac
-0000b070: 6964 2070 726f 6261 6269 6c69 7469 6573  id probabilities
-0000b080: 0a20 2020 2020 2020 2072 656e 616d 655f  .        rename_
-0000b090: 696e 7465 726e 616c 5f6e 6f64 6573 3a20  internal_nodes: 
-0000b0a0: 7b54 7275 652c 4661 6c73 657d 2075 7365  {True,False} use
-0000b0b0: 2064 6966 6665 7265 6e74 206e 616d 6573   different names
-0000b0c0: 2066 6f72 2074 6865 2069 6e74 6572 6e61   for the interna
-0000b0d0: 6c2f 616e 6365 7374 7261 6c20 6e6f 6465  l/ancestral node
-0000b0e0: 7320 6672 6f6d 2074 6865 206f 6e65 7320  s from the ones 
-0000b0f0: 6769 7665 6e20 696e 2074 6865 2074 7265  given in the tre
-0000b100: 650a 2020 2020 2020 2020 7374 6f72 6167  e.        storag
-0000b110: 655f 666f 6c64 6572 3a20 2264 7261 7570  e_folder: "draup
-0000b120: 6e69 722f 7372 632f 6472 6175 706e 6972  nir/src/draupnir
-0000b130: 2f64 6174 612f 6461 7461 7365 745f 6e61  /data/dataset_na
-0000b140: 6d65 2220 6f72 2066 6f6c 6465 7220 7768  me" or folder wh
-0000b150: 6572 6520 7468 6520 6661 7374 6120 6669  ere the fasta fi
-0000b160: 6c65 2069 7320 6c6f 6361 7465 6420 2872  le is located (r
-0000b170: 6563 6f6d 6d65 6e64 6564 2074 6f20 7075  ecommended to pu
-0000b180: 7420 696e 2061 2073 7065 6369 6669 6320  t in a specific 
-0000b190: 666f 6c64 6572 290a 2020 2020 6f75 743a  folder).    out:
-0000b1a0: 0a20 2020 2020 2020 2069 6620 6f6e 655f  .        if one_
-0000b1b0: 686f 745f 656e 636f 6469 6e67 3a20 7768  hot_encoding: wh
-0000b1c0: 6572 6520 6761 7020 6973 205b 312c 302c  ere gap is [1,0,
-0000b1d0: 302c 302c 302c 302c 302c 302c 302c 302c  0,0,0,0,0,0,0,0,
-0000b1e0: 302c 302c 302c 302c 302c 302c 302c 302c  0,0,0,0,0,0,0,0,
-0000b1f0: 302c 305d 0a20 2020 2020 2020 2020 2020  0,0].           
-0000b200: 2020 2020 5465 6e73 6f72 2077 6974 6820      Tensor with 
-0000b210: 7369 7a65 3a20 5b4e 7365 7175 656e 6365  size: [Nsequence
-0000b220: 735d 785b 416c 6967 6e6d 656e 7420 6c65  s]x[Alignment le
-0000b230: 6e67 7468 202b 2032 5d78 5b33 305d 202d  ngth + 2]x[30] -
-0000b240: 2d3e 205b 5b6c 656e 6774 682c 7472 6565  -> [[length,tree
-0000b250: 5f70 6f73 6974 696f 6e2c 6469 7374 5f74  _position,dist_t
-0000b260: 6f5f 726f 6f74 2c3f 2c30 2c30 2c30 2e2e  o_root,?,0,0,0..
-0000b270: 2e5d 2c5b 4749 5420 7665 6374 6f72 5d2c  .],[GIT vector],
-0000b280: 5b61 6131 206f 6e65 2068 6f74 2c20 7068  [aa1 one hot, ph
-0000b290: 692c 2070 7369 5d2c 5b61 6132 206f 6e65  i, psi],[aa2 one
-0000b2a0: 2068 6f74 2c20 7068 692c 2070 7369 5d2c   hot, phi, psi],
-0000b2b0: 2e2e 2e2e 5b5d 5d0a 2020 2020 2020 2020  ....[]].        
-0000b2c0: 656c 7365 3a20 416d 696e 6f20 6163 6964  else: Amino acid
-0000b2d0: 7320 6172 6520 6173 7369 676e 6564 206e  s are assigned n
-0000b2e0: 756d 6265 7273 2066 726f 6d20 312d 3230  umbers from 1-20
-0000b2f0: 2c20 3020 6d65 616e 7320 6761 700a 2020  , 0 means gap.  
-0000b300: 2020 2020 2020 2020 2020 2020 2054 656e               Ten
-0000b310: 736f 7220 7769 7468 2073 697a 653a 205b  sor with size: [
-0000b320: 4e73 6571 7565 6e63 6573 5d78 5b41 6c69  Nsequences]x[Ali
-0000b330: 676e 6d65 6e74 206c 656e 6774 6820 2b20  gnment length + 
-0000b340: 335d 785b 3330 5d20 2d2d 3e20 5b5b 6c65  3]x[30] --> [[le
-0000b350: 6e67 7468 2c74 7265 655f 706f 7369 7469  ngth,tree_positi
-0000b360: 6f6e 2c64 6973 745f 746f 5f72 6f6f 742c  on,dist_to_root,
-0000b370: 302c 302c 302c 302e 2e2e 5d2c 5b47 4954  0,0,0,0...],[GIT
-0000b380: 2076 6563 746f 725d 2c5b 6161 3120 6e75   vector],[aa1 nu
-0000b390: 6d62 6572 2c20 7068 692c 2070 7369 5d2c  mber, phi, psi],
-0000b3a0: 5b61 6132 206e 756d 6265 722c 2070 6869  [aa2 number, phi
-0000b3b0: 2c20 7073 695d 2c2e 2e2e 2e5b 5d5d 0a20  , psi],....[]]. 
-0000b3c0: 2020 2022 2222 0a0a 2020 2020 7761 726e     """..    warn
-0000b3d0: 696e 6773 2e73 696d 706c 6566 696c 7465  ings.simplefilte
-0000b3e0: 7228 2769 676e 6f72 6527 2c20 4269 6f70  r('ignore', Biop
-0000b3f0: 7974 686f 6e57 6172 6e69 6e67 290a 2020  ythonWarning).  
-0000b400: 2020 6f6e 655f 686f 745f 6c61 6265 6c3d    one_hot_label=
-0000b410: 205b 226f 6e65 686f 7422 2069 6620 6f6e   ["onehot" if on
-0000b420: 655f 686f 745f 656e 636f 6469 6e67 2065  e_hot_encoding e
-0000b430: 6c73 6520 2269 6e74 6567 6572 7322 5d0a  lse "integers"].
-0000b440: 0a20 2020 2070 726f 745f 696e 666f 5f64  .    prot_info_d
-0000b450: 6963 7420 3d20 7b7d 0a20 2020 2070 726f  ict = {}.    pro
-0000b460: 745f 6161 5f64 6963 7420 3d20 7b7d 0a20  t_aa_dict = {}. 
-0000b470: 2020 2069 6620 5044 425f 666f 6c64 6572     if PDB_folder
-0000b480: 3a23 2061 6e64 206e 6f74 2061 6c69 676e  :# and not align
-0000b490: 6d65 6e74 5f66 696c 653a 2d2d 2d3e 2057  ment_file:---> W
-0000b4a0: 6520 616c 6c6f 7720 746f 2068 6176 6520  e allow to have 
-0000b4b0: 7365 7175 656e 6365 7320 7468 6174 2068  sequences that h
-0000b4c0: 6176 6520 616e 6420 646f 6e27 7420 6861  ave and don't ha
-0000b4d0: 7665 2033 4420 7374 7275 6374 7572 650a  ve 3D structure.
-0000b4e0: 2020 2020 2020 2020 7072 696e 7428 2243          print("C
-0000b4f0: 7265 6174 696e 6720 6461 7461 7365 7420  reating dataset 
-0000b500: 6672 6f6d 2050 4442 2066 696c 6573 2e2e  from PDB files..
-0000b510: 2e22 290a 2020 2020 2020 2020 7072 6f74  .").        prot
-0000b520: 5f61 615f 6469 6374 2c70 726f 745f 696e  _aa_dict,prot_in
-0000b530: 666f 5f64 6963 7420 3d20 7072 6f63 6573  fo_dict = proces
-0000b540: 735f 7064 625f 6669 6c65 7328 5044 425f  s_pdb_files(PDB_
-0000b550: 666f 6c64 6572 2c61 615f 7072 6f62 732c  folder,aa_probs,
-0000b560: 7066 616d 5f64 6963 742c 6f6e 655f 686f  pfam_dict,one_ho
-0000b570: 745f 656e 636f 6469 6e67 2c6d 696e 5f6c  t_encoding,min_l
-0000b580: 656e 290a 2020 2020 2352 656d 6f76 6520  en).    #Remove 
-0000b590: 6475 706c 6963 6174 6564 2073 6571 7565  duplicated seque
-0000b5a0: 6e63 6573 2069 6e20 626f 7468 2064 6963  nces in both dic
-0000b5b0: 7469 6f6e 6172 6965 730a 2020 2020 6966  tionaries.    if
-0000b5c0: 2070 726f 745f 6161 5f64 6963 743a 0a20   prot_aa_dict:. 
-0000b5d0: 2020 2020 2020 2066 6173 7461 5f66 696c         fasta_fil
-0000b5e0: 6520 3d20 227b 7d2f 7b7d 2f7b 7d2e 6661  e = "{}/{}/{}.fa
-0000b5f0: 7374 6122 2e66 6f72 6d61 7428 7374 6f72  sta".format(stor
-0000b600: 6167 655f 666f 6c64 6572 2c6e 616d 655f  age_folder,name_
-0000b610: 6669 6c65 2c6e 616d 655f 6669 6c65 290a  file,name_file).
-0000b620: 2020 2020 2020 2020 7072 696e 7428 2257          print("W
-0000b630: 7269 7469 6e67 2070 6f6c 7970 6570 7469  riting polypepti
-0000b640: 6465 7320 746f 2066 6173 7461 2066 696c  des to fasta fil
-0000b650: 6520 746f 207b 7d22 2e66 6f72 6d61 7428  e to {}".format(
-0000b660: 6661 7374 615f 6669 6c65 2929 0a20 2020  fasta_file)).   
-0000b670: 2020 2020 2077 6974 6820 6f70 656e 2866       with open(f
-0000b680: 6173 7461 5f66 696c 652c 2022 7722 2920  asta_file, "w") 
-0000b690: 6173 206f 7574 7075 745f 6861 6e64 6c65  as output_handle
-0000b6a0: 3a0a 2020 2020 2020 2020 2020 2020 7064  :.            pd
-0000b6b0: 625f 7265 636f 7264 7320 3d5b 5d0a 2020  b_records =[].  
-0000b6c0: 2020 2020 2020 2020 2020 666f 7220 6964            for id
-0000b6d0: 2c73 6571 7565 6e63 6520 696e 2070 726f  ,sequence in pro
-0000b6e0: 745f 6161 5f64 6963 742e 6974 656d 7328  t_aa_dict.items(
-0000b6f0: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
-0000b700: 2020 2070 6462 5f72 6563 6f72 6420 3d20     pdb_record = 
-0000b710: 5365 7152 6563 6f72 6428 5365 7128 2727  SeqRecord(Seq(''
-0000b720: 2e6a 6f69 6e28 7365 7175 656e 6365 2929  .join(sequence))
-0000b730: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
+0000b000: 2022 7261 7069 646e 6a22 3a20 696e 6665   "rapidnj": infe
+0000b010: 7265 6e63 6520 6f66 2046 6173 7420 4e4a  rence of Fast NJ
+0000b020: 2075 6e72 6f6f 7465 6420 696e 6665 7265   unrooted infere
+0000b030: 6e63 6520 286d 616b 6520 7375 7265 2069  nce (make sure i
+0000b040: 7320 696e 7374 616c 6c65 6420 676c 6f62  s installed glob
+0000b050: 616c 6c79 292c 0a20 2020 2020 2020 2061  ally),.        a
+0000b060: 615f 7072 6f62 733a 2061 6d69 6e6f 2061  a_probs: amino a
+0000b070: 6369 6420 7072 6f62 6162 696c 6974 6965  cid probabilitie
+0000b080: 730a 2020 2020 2020 2020 7265 6e61 6d65  s.        rename
+0000b090: 5f69 6e74 6572 6e61 6c5f 6e6f 6465 733a  _internal_nodes:
+0000b0a0: 207b 5472 7565 2c46 616c 7365 7d20 7573   {True,False} us
+0000b0b0: 6520 6469 6666 6572 656e 7420 6e61 6d65  e different name
+0000b0c0: 7320 666f 7220 7468 6520 696e 7465 726e  s for the intern
+0000b0d0: 616c 2f61 6e63 6573 7472 616c 206e 6f64  al/ancestral nod
+0000b0e0: 6573 2066 726f 6d20 7468 6520 6f6e 6573  es from the ones
+0000b0f0: 2067 6976 656e 2069 6e20 7468 6520 7472   given in the tr
+0000b100: 6565 0a20 2020 2020 2020 2073 746f 7261  ee.        stora
+0000b110: 6765 5f66 6f6c 6465 723a 2022 6472 6175  ge_folder: "drau
+0000b120: 706e 6972 2f73 7263 2f64 7261 7570 6e69  pnir/src/draupni
+0000b130: 722f 6461 7461 2f64 6174 6173 6574 5f6e  r/data/dataset_n
+0000b140: 616d 6522 206f 7220 666f 6c64 6572 2077  ame" or folder w
+0000b150: 6865 7265 2074 6865 2066 6173 7461 2066  here the fasta f
+0000b160: 696c 6520 6973 206c 6f63 6174 6564 2028  ile is located (
+0000b170: 7265 636f 6d6d 656e 6465 6420 746f 2070  recommended to p
+0000b180: 7574 2069 6e20 6120 7370 6563 6966 6963  ut in a specific
+0000b190: 2066 6f6c 6465 7229 0a20 2020 206f 7574   folder).    out
+0000b1a0: 3a0a 2020 2020 2020 2020 6966 206f 6e65  :.        if one
+0000b1b0: 5f68 6f74 5f65 6e63 6f64 696e 673a 2077  _hot_encoding: w
+0000b1c0: 6865 7265 2067 6170 2069 7320 5b31 2c30  here gap is [1,0
+0000b1d0: 2c30 2c30 2c30 2c30 2c30 2c30 2c30 2c30  ,0,0,0,0,0,0,0,0
+0000b1e0: 2c30 2c30 2c30 2c30 2c30 2c30 2c30 2c30  ,0,0,0,0,0,0,0,0
+0000b1f0: 2c30 2c30 5d0a 2020 2020 2020 2020 2020  ,0,0].          
+0000b200: 2020 2020 2054 656e 736f 7220 7769 7468       Tensor with
+0000b210: 2073 697a 653a 205b 4e73 6571 7565 6e63   size: [Nsequenc
+0000b220: 6573 5d78 5b41 6c69 676e 6d65 6e74 206c  es]x[Alignment l
+0000b230: 656e 6774 6820 2b20 325d 785b 3330 5d20  ength + 2]x[30] 
+0000b240: 2d2d 3e20 5b5b 6c65 6e67 7468 2c74 7265  --> [[length,tre
+0000b250: 655f 706f 7369 7469 6f6e 2c64 6973 745f  e_position,dist_
+0000b260: 746f 5f72 6f6f 742c 3f2c 302c 302c 302e  to_root,?,0,0,0.
+0000b270: 2e2e 5d2c 5b47 4954 2076 6563 746f 725d  ..],[GIT vector]
+0000b280: 2c5b 6161 3120 6f6e 6520 686f 742c 2070  ,[aa1 one hot, p
+0000b290: 6869 2c20 7073 695d 2c5b 6161 3220 6f6e  hi, psi],[aa2 on
+0000b2a0: 6520 686f 742c 2070 6869 2c20 7073 695d  e hot, phi, psi]
+0000b2b0: 2c2e 2e2e 2e5b 5d5d 0a20 2020 2020 2020  ,....[]].       
+0000b2c0: 2065 6c73 653a 2041 6d69 6e6f 2061 6369   else: Amino aci
+0000b2d0: 6473 2061 7265 2061 7373 6967 6e65 6420  ds are assigned 
+0000b2e0: 6e75 6d62 6572 7320 6672 6f6d 2031 2d32  numbers from 1-2
+0000b2f0: 302c 2030 206d 6561 6e73 2067 6170 0a20  0, 0 means gap. 
+0000b300: 2020 2020 2020 2020 2020 2020 2020 5465                Te
+0000b310: 6e73 6f72 2077 6974 6820 7369 7a65 3a20  nsor with size: 
+0000b320: 5b4e 7365 7175 656e 6365 735d 785b 416c  [Nsequences]x[Al
+0000b330: 6967 6e6d 656e 7420 6c65 6e67 7468 202b  ignment length +
+0000b340: 2033 5d78 5b33 305d 202d 2d3e 205b 5b6c   3]x[30] --> [[l
+0000b350: 656e 6774 682c 7472 6565 5f70 6f73 6974  ength,tree_posit
+0000b360: 696f 6e2c 6469 7374 5f74 6f5f 726f 6f74  ion,dist_to_root
+0000b370: 2c30 2c30 2c30 2c30 2e2e 2e5d 2c5b 4749  ,0,0,0,0...],[GI
+0000b380: 5420 7665 6374 6f72 5d2c 5b61 6131 206e  T vector],[aa1 n
+0000b390: 756d 6265 722c 2070 6869 2c20 7073 695d  umber, phi, psi]
+0000b3a0: 2c5b 6161 3220 6e75 6d62 6572 2c20 7068  ,[aa2 number, ph
+0000b3b0: 692c 2070 7369 5d2c 2e2e 2e2e 5b5d 5d0a  i, psi],....[]].
+0000b3c0: 2020 2020 2222 220a 0a20 2020 2077 6172      """..    war
+0000b3d0: 6e69 6e67 732e 7369 6d70 6c65 6669 6c74  nings.simplefilt
+0000b3e0: 6572 2827 6967 6e6f 7265 272c 2042 696f  er('ignore', Bio
+0000b3f0: 7079 7468 6f6e 5761 726e 696e 6729 0a20  pythonWarning). 
+0000b400: 2020 206f 6e65 5f68 6f74 5f6c 6162 656c     one_hot_label
+0000b410: 3d20 5b22 6f6e 6568 6f74 2220 6966 206f  = ["onehot" if o
+0000b420: 6e65 5f68 6f74 5f65 6e63 6f64 696e 6720  ne_hot_encoding 
+0000b430: 656c 7365 2022 696e 7465 6765 7273 225d  else "integers"]
+0000b440: 0a0a 2020 2020 7072 6f74 5f69 6e66 6f5f  ..    prot_info_
+0000b450: 6469 6374 203d 207b 7d0a 2020 2020 7072  dict = {}.    pr
+0000b460: 6f74 5f61 615f 6469 6374 203d 207b 7d0a  ot_aa_dict = {}.
+0000b470: 2020 2020 6966 2050 4442 5f66 6f6c 6465      if PDB_folde
+0000b480: 723a 2320 616e 6420 6e6f 7420 616c 6967  r:# and not alig
+0000b490: 6e6d 656e 745f 6669 6c65 3a2d 2d2d 3e20  nment_file:---> 
+0000b4a0: 5765 2061 6c6c 6f77 2074 6f20 6861 7665  We allow to have
+0000b4b0: 2073 6571 7565 6e63 6573 2074 6861 7420   sequences that 
+0000b4c0: 6861 7665 2061 6e64 2064 6f6e 2774 2068  have and don't h
+0000b4d0: 6176 6520 3344 2073 7472 7563 7475 7265  ave 3D structure
+0000b4e0: 0a20 2020 2020 2020 2070 7269 6e74 2822  .        print("
+0000b4f0: 4372 6561 7469 6e67 2064 6174 6173 6574  Creating dataset
+0000b500: 2066 726f 6d20 5044 4220 6669 6c65 732e   from PDB files.
+0000b510: 2e2e 2229 0a20 2020 2020 2020 2070 726f  ..").        pro
+0000b520: 745f 6161 5f64 6963 742c 7072 6f74 5f69  t_aa_dict,prot_i
+0000b530: 6e66 6f5f 6469 6374 203d 2070 726f 6365  nfo_dict = proce
+0000b540: 7373 5f70 6462 5f66 696c 6573 2850 4442  ss_pdb_files(PDB
+0000b550: 5f66 6f6c 6465 722c 6161 5f70 726f 6273  _folder,aa_probs
+0000b560: 2c70 6661 6d5f 6469 6374 2c6f 6e65 5f68  ,pfam_dict,one_h
+0000b570: 6f74 5f65 6e63 6f64 696e 672c 6d69 6e5f  ot_encoding,min_
+0000b580: 6c65 6e29 0a20 2020 2023 5265 6d6f 7665  len).    #Remove
+0000b590: 2064 7570 6c69 6361 7465 6420 7365 7175   duplicated sequ
+0000b5a0: 656e 6365 7320 696e 2062 6f74 6820 6469  ences in both di
+0000b5b0: 6374 696f 6e61 7269 6573 0a20 2020 2069  ctionaries.    i
+0000b5c0: 6620 7072 6f74 5f61 615f 6469 6374 3a0a  f prot_aa_dict:.
+0000b5d0: 2020 2020 2020 2020 6661 7374 615f 6669          fasta_fi
+0000b5e0: 6c65 203d 2022 7b7d 2f7b 7d2f 7b7d 2e66  le = "{}/{}/{}.f
+0000b5f0: 6173 7461 222e 666f 726d 6174 2873 746f  asta".format(sto
+0000b600: 7261 6765 5f66 6f6c 6465 722c 6e61 6d65  rage_folder,name
+0000b610: 5f66 696c 652c 6e61 6d65 5f66 696c 6529  _file,name_file)
+0000b620: 0a20 2020 2020 2020 2070 7269 6e74 2822  .        print("
+0000b630: 5772 6974 696e 6720 706f 6c79 7065 7074  Writing polypept
+0000b640: 6964 6573 2074 6f20 6661 7374 6120 6669  ides to fasta fi
+0000b650: 6c65 2074 6f20 7b7d 222e 666f 726d 6174  le to {}".format
+0000b660: 2866 6173 7461 5f66 696c 6529 290a 2020  (fasta_file)).  
+0000b670: 2020 2020 2020 7769 7468 206f 7065 6e28        with open(
+0000b680: 6661 7374 615f 6669 6c65 2c20 2277 2229  fasta_file, "w")
+0000b690: 2061 7320 6f75 7470 7574 5f68 616e 646c   as output_handl
+0000b6a0: 653a 0a20 2020 2020 2020 2020 2020 2070  e:.            p
+0000b6b0: 6462 5f72 6563 6f72 6473 203d 5b5d 0a20  db_records =[]. 
+0000b6c0: 2020 2020 2020 2020 2020 2066 6f72 2069             for i
+0000b6d0: 642c 7365 7175 656e 6365 2069 6e20 7072  d,sequence in pr
+0000b6e0: 6f74 5f61 615f 6469 6374 2e69 7465 6d73  ot_aa_dict.items
+0000b6f0: 2829 3a0a 2020 2020 2020 2020 2020 2020  ():.            
+0000b700: 2020 2020 7064 625f 7265 636f 7264 203d      pdb_record =
+0000b710: 2053 6571 5265 636f 7264 2853 6571 2827   SeqRecord(Seq('
+0000b720: 272e 6a6f 696e 2873 6571 7565 6e63 6529  '.join(sequence)
+0000b730: 292c 0a20 2020 2020 2020 2020 2020 2020  ),.             
 0000b740: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000b750: 2020 2020 2069 643d 7374 7228 6964 292c       id=str(id),
-0000b760: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+0000b750: 2020 2020 2020 6964 3d73 7472 2869 6429        id=str(id)
+0000b760: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
 0000b770: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000b780: 2020 2020 6465 7363 7269 7074 696f 6e3d      description=
-0000b790: 2222 2c0a 2020 2020 2020 2020 2020 2020  "",.            
+0000b780: 2020 2020 2064 6573 6372 6970 7469 6f6e       description
+0000b790: 3d22 222c 0a20 2020 2020 2020 2020 2020  ="",.           
 0000b7a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000b7b0: 2020 2020 2020 2061 6e6e 6f74 6174 696f         annotatio
-0000b7c0: 6e73 3d7b 226d 6f6c 6563 756c 655f 7479  ns={"molecule_ty
-0000b7d0: 7065 223a 2022 7072 6f74 6569 6e22 7d29  pe": "protein"})
-0000b7e0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0000b7f0: 2070 6462 5f72 6563 6f72 6473 2e61 7070   pdb_records.app
-0000b800: 656e 6428 7064 625f 7265 636f 7264 290a  end(pdb_record).
-0000b810: 2020 2020 2020 2020 2020 2020 5365 7149              SeqI
-0000b820: 4f2e 7772 6974 6528 7064 625f 7265 636f  O.write(pdb_reco
-0000b830: 7264 732c 206f 7574 7075 745f 6861 6e64  rds, output_hand
-0000b840: 6c65 2c20 2266 6173 7461 2229 0a20 2020  le, "fasta").   
-0000b850: 2020 2020 206f 7574 7075 745f 6861 6e64       output_hand
-0000b860: 6c65 2e63 6c6f 7365 2829 0a20 2020 2023  le.close().    #
-0000b870: 2048 6967 686c 6967 6874 3a20 416c 6967   Highlight: Alig
-0000b880: 6e20 7468 6520 706f 6c79 7065 7074 6964  n the polypeptid
-0000b890: 6573 2f73 6571 7565 6e63 6573 2061 6e64  es/sequences and
-0000b8a0: 2077 7269 7465 2074 6f20 6120 6661 7374   write to a fast
-0000b8b0: 6120 616e 676c 6573 5f6c 6973 7420 6669  a angles_list fi
-0000b8c0: 6c65 0a20 2020 2064 6963 745f 616c 6967  le.    dict_alig
-0000b8d0: 6e6d 656e 742c 616c 6967 6e6d 656e 7420  nment,alignment 
-0000b8e0: 3d20 696e 6665 725f 616c 6967 6e6d 656e  = infer_alignmen
-0000b8f0: 7428 616c 6967 6e6d 656e 745f 6669 6c65  t(alignment_file
-0000b900: 2c69 6e70 7574 5f6e 616d 655f 6669 6c65  ,input_name_file
-0000b910: 3d66 6173 7461 5f66 696c 652c 6f75 7470  =fasta_file,outp
-0000b920: 7574 5f6e 616d 655f 6669 6c65 3d22 7b7d  ut_name_file="{}
-0000b930: 2f7b 7d2f 7b7d 2e6d 6166 6674 222e 666f  /{}/{}.mafft".fo
-0000b940: 726d 6174 2873 746f 7261 6765 5f66 6f6c  rmat(storage_fol
-0000b950: 6465 722c 6e61 6d65 5f66 696c 652c 6e61  der,name_file,na
-0000b960: 6d65 5f66 696c 6529 290a 2020 2020 2363  me_file)).    #c
-0000b970: 616c 6375 6c61 7465 5f70 6169 7277 6973  alculate_pairwis
-0000b980: 655f 6469 7374 616e 6365 286e 616d 655f  e_distance(name_
-0000b990: 6669 6c65 2c61 6c69 676e 6d65 6e74 290a  file,alignment).
-0000b9a0: 2020 2020 616c 6967 6e6d 656e 745f 6669      alignment_fi
-0000b9b0: 6c65 203d 205b 616c 6967 6e6d 656e 745f  le = [alignment_
-0000b9c0: 6669 6c65 2069 6620 616c 6967 6e6d 656e  file if alignmen
-0000b9d0: 745f 6669 6c65 2065 6c73 6520 227b 7d2f  t_file else "{}/
-0000b9e0: 7b7d 2f7b 7d2e 6d61 6666 7422 2e66 6f72  {}/{}.mafft".for
-0000b9f0: 6d61 7428 7374 6f72 6167 655f 666f 6c64  mat(storage_fold
-0000ba00: 6572 2c6e 616d 655f 6669 6c65 2c6e 616d  er,name_file,nam
-0000ba10: 655f 6669 6c65 295d 5b30 5d0a 2020 2020  e_file)][0].    
-0000ba20: 2348 6967 686c 6967 6874 3a20 6368 6563  #Highlight: chec
-0000ba30: 6b69 6e67 2074 6861 7420 7468 6520 7365  king that the se
-0000ba40: 6c65 6374 6564 206e 756d 6265 7220 6f66  lected number of
-0000ba50: 2070 726f 6261 6269 6c69 7469 6573 2069   probabilities i
-0000ba60: 7320 636f 7272 6563 740a 2020 2020 7375  s correct.    su
-0000ba70: 6d6d 6172 795f 6161 5f70 726f 6273 203d  mmary_aa_probs =
-0000ba80: 205b 7661 6c69 6461 7465 5f73 6571 7565   [validate_seque
-0000ba90: 6e63 655f 616c 7068 6162 6574 2876 616c  nce_alphabet(val
-0000baa0: 7565 2920 666f 7220 6b65 792c 7661 6c75  ue) for key,valu
-0000bab0: 6520 696e 2064 6963 745f 616c 6967 6e6d  e in dict_alignm
-0000bac0: 656e 742e 6974 656d 7328 295d 2023 6669  ent.items()] #fi
-0000bad0: 6e64 7320 7468 6520 616c 7068 6162 6574  nds the alphabet
-0000bae0: 7320 6f66 2065 6163 6820 6f66 2074 6865  s of each of the
-0000baf0: 2073 6571 7565 6e63 6573 2069 6e20 7468   sequences in th
-0000bb00: 6520 616c 6967 6e6d 656e 742c 2063 6865  e alignment, che
-0000bb10: 636b 7320 666f 7220 646e 610a 2020 2020  cks for dna.    
-0000bb20: 6161 5f70 726f 6273 203d 206d 6178 2861  aa_probs = max(a
-0000bb30: 615f 7072 6f62 732c 6d61 7828 7375 6d6d  a_probs,max(summ
-0000bb40: 6172 795f 6161 5f70 726f 6273 2929 2023  ary_aa_probs)) #
-0000bb50: 6966 2074 6865 2069 6e70 7574 2061 615f  if the input aa_
-0000bb60: 7072 6f62 7320 6973 2064 6966 6665 7265  probs is differe
-0000bb70: 6e74 2066 726f 6d20 7468 6f73 6520 666f  nt from those fo
-0000bb80: 756e 642c 2074 6865 2061 615f 7072 6f62  und, the aa_prob
-0000bb90: 7320 6368 616e 6765 2e20 416e 6420 616c  s change. And al
-0000bba0: 736f 2074 6865 2061 6120 7375 6273 7469  so the aa substi
-0000bbb0: 7475 7469 6f6e 2020 6d61 7472 6978 0a20  tution  matrix. 
-0000bbc0: 2020 2061 615f 6e61 6d65 735f 6469 6374     aa_names_dict
-0000bbd0: 203d 2061 6d69 6e6f 6163 6964 5f6e 616d   = aminoacid_nam
-0000bbe0: 6573 5f64 6963 7428 6161 5f70 726f 6273  es_dict(aa_probs
-0000bbf0: 290a 2020 2020 2348 6967 686c 6967 6874  ).    #Highlight
-0000bc00: 3a20 4966 2074 6865 2061 6120 7365 7175  : If the aa sequ
-0000bc10: 656e 6365 7320 646f 206e 6f74 2063 6f6d  ences do not com
-0000bc20: 6520 6672 6f6d 2020 5044 4220 6669 6c65  e from  PDB file
-0000bc30: 732c 2074 6865 7920 636f 6d65 2066 726f  s, they come fro
-0000bc40: 6d20 616e 2061 6c69 676e 6d65 6e74 2066  m an alignment f
-0000bc50: 696c 6520 7468 6174 206e 6565 6473 2074  ile that needs t
-0000bc60: 6f20 6265 2070 726f 6365 7373 6564 0a20  o be processed. 
-0000bc70: 2020 2023 6469 6374 5f61 6c69 676e 6d65     #dict_alignme
-0000bc80: 6e74 5f32 203d 2064 6963 742e 6672 6f6d  nt_2 = dict.from
-0000bc90: 6b65 7973 2864 6963 745f 616c 6967 6e6d  keys(dict_alignm
-0000bca0: 656e 742e 6b65 7973 2829 290a 2020 2020  ent.keys()).    
-0000bcb0: 6e6f 745f 616c 6967 6e65 645f 7365 7173  not_aligned_seqs
-0000bcc0: 5f66 726f 6d5f 616c 6967 6e6d 656e 745f  _from_alignment_
-0000bcd0: 6669 6c65 203d 7b7d 0a20 2020 2066 6f72  file ={}.    for
-0000bce0: 206b 6579 2c76 616c 7565 2069 6e20 6469   key,value in di
-0000bcf0: 6374 5f61 6c69 676e 6d65 6e74 2e69 7465  ct_alignment.ite
-0000bd00: 6d73 2829 3a0a 2020 2020 2020 2020 616c  ms():.        al
-0000bd10: 6967 6e65 645f 7365 7120 3d20 6c69 7374  igned_seq = list
-0000bd20: 2864 6963 745f 616c 6967 6e6d 656e 745b  (dict_alignment[
-0000bd30: 6b65 795d 290a 2020 2020 2020 2020 236e  key]).        #n
-0000bd40: 6f5f 6761 705f 696e 6465 7865 7320 3d20  o_gap_indexes = 
-0000bd50: 6e70 2e77 6865 7265 286e 702e 6172 7261  np.where(np.arra
-0000bd60: 7928 616c 6967 6e65 645f 7365 7129 2021  y(aligned_seq) !
-0000bd70: 3d20 222d 2229 5b30 5d20 2b20 3220 2023  = "-")[0] + 2  #
-0000bd80: 2070 6c75 7320 3220 696e 206f 7264 6572   plus 2 in order
-0000bd90: 2074 6f20 6d61 6b65 2074 6865 2069 6e64   to make the ind
-0000bda0: 6578 6573 2066 6974 2069 6e20 7468 6520  exes fit in the 
-0000bdb0: 6669 6e61 6c20 6461 7461 6672 616d 650a  final dataframe.
-0000bdc0: 2020 2020 2020 2020 6e6f 745f 616c 6967          not_alig
-0000bdd0: 6e65 645f 7365 7120 3d6c 6973 7428 6669  ned_seq =list(fi
-0000bde0: 6c74 6572 286c 616d 6264 6120 613a 2061  lter(lambda a: a
-0000bdf0: 2021 3d20 222d 222c 2061 6c69 676e 6564   != "-", aligned
-0000be00: 5f73 6571 2929 0a20 2020 2020 2020 2073  _seq)).        s
-0000be10: 6571 5f6c 656e 203d 206c 656e 286e 6f74  eq_len = len(not
-0000be20: 5f61 6c69 676e 6564 5f73 6571 290a 2020  _aligned_seq).  
-0000be30: 2020 2020 2020 6769 745f 7665 6374 6f72        git_vector
-0000be40: 203d 206e 702e 7a65 726f 7328 3330 2920   = np.zeros(30) 
-0000be50: 2366 616b 6520 6769 7420 7665 6374 6f72  #fake git vector
-0000be60: 0a20 2020 2020 2020 2061 615f 696e 666f  .        aa_info
-0000be70: 203d 206e 702e 7a65 726f 7328 2873 6571   = np.zeros((seq
-0000be80: 5f6c 656e 202b 2032 2c20 3330 2929 0a20  _len + 2, 30)). 
-0000be90: 2020 2020 2020 2061 615f 696e 666f 5b30         aa_info[0
-0000bea0: 5d20 3d20 6e70 2e68 7374 6163 6b28 5b73  ] = np.hstack([s
-0000beb0: 6571 5f6c 656e 2c5b 305d 2a32 395d 2920  eq_len,[0]*29]) 
-0000bec0: 2366 6972 7374 2072 6f77 2063 6f6e 7461  #first row conta
-0000bed0: 696e 7320 736f 6d65 2073 6571 7565 6e63  ins some sequenc
-0000bee0: 6520 6c65 6e67 7468 2069 6e66 6f72 6d61  e length informa
-0000bef0: 7469 6f6e 0a20 2020 2020 2020 2061 615f  tion.        aa_
-0000bf00: 696e 666f 5b31 5d20 3d20 6769 745f 7665  info[1] = git_ve
-0000bf10: 6374 6f72 2023 7365 636f 6e64 2072 6f77  ctor #second row
-0000bf20: 2063 6f6e 7461 696e 7320 6120 6769 7420   contains a git 
-0000bf30: 7665 6374 6f72 2028 6e65 7665 7220 7573  vector (never us
-0000bf40: 6564 2062 7574 2077 6520 636f 756c 6420  ed but we could 
-0000bf50: 7573 6520 6974 2066 6f72 2073 6f6d 6574  use it for somet
-0000bf60: 6869 6e67 2065 6c73 6529 0a20 2020 2020  hing else).     
-0000bf70: 2020 2069 6620 6f6e 655f 686f 745f 656e     if one_hot_en
-0000bf80: 636f 6469 6e67 3a0a 2020 2020 2020 2020  coding:.        
-0000bf90: 2020 2020 666f 7220 696e 6465 785f 612c      for index_a,
-0000bfa0: 2061 615f 6e61 6d65 2069 6e20 656e 756d   aa_name in enum
-0000bfb0: 6572 6174 6528 6e6f 745f 616c 6967 6e65  erate(not_aligne
-0000bfc0: 645f 7365 7129 3a0a 2020 2020 2020 2020  d_seq):.        
-0000bfd0: 2020 2020 2020 2020 6f6e 655f 686f 7420          one_hot 
-0000bfe0: 3d20 6e70 2e7a 6572 6f73 2861 615f 7072  = np.zeros(aa_pr
-0000bff0: 6f62 7329 0a20 2020 2020 2020 2020 2020  obs).           
-0000c000: 2020 2020 2069 6e64 6578 203d 2061 615f       index = aa_
-0000c010: 6e61 6d65 735f 6469 6374 5b61 615f 6e61  names_dict[aa_na
-0000c020: 6d65 5d0a 2020 2020 2020 2020 2020 2020  me].            
-0000c030: 2020 2020 6f6e 655f 686f 745b 696e 6465      one_hot[inde
-0000c040: 785d 203d 2031 0a20 2020 2020 2020 2020  x] = 1.         
-0000c050: 2020 2020 2020 2065 7874 7261 5f73 7061         extra_spa
-0000c060: 6365 203d 2033 302d 6161 5f70 726f 6273  ce = 30-aa_probs
-0000c070: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0000c080: 2061 615f 696e 666f 5b69 6e64 6578 5f61   aa_info[index_a
-0000c090: 2b32 5d20 3d20 6e70 2e68 7374 6163 6b28  +2] = np.hstack(
-0000c0a0: 5b6f 6e65 5f68 6f74 2c5b 305d 2a65 7874  [one_hot,[0]*ext
-0000c0b0: 7261 5f73 7061 6365 5d29 2023 6578 7472  ra_space]) #extr
-0000c0c0: 615f 7370 6163 6520 3d20 3130 2066 6f72  a_space = 10 for
-0000c0d0: 2032 3020 6161 5f70 726f 6273 2c20 3920   20 aa_probs, 9 
-0000c0e0: 666f 7220 3231 2061 615f 7072 6f62 732c  for 21 aa_probs,
-0000c0f0: 2038 2066 6f72 2032 3220 6161 5f70 726f   8 for 22 aa_pro
-0000c100: 6273 0a20 2020 2020 2020 2065 6c73 653a  bs.        else:
-0000c110: 0a20 2020 2020 2020 2020 2020 2066 6f72  .            for
-0000c120: 2069 6e64 6578 5f61 612c 2061 615f 6e61   index_aa, aa_na
-0000c130: 6d65 2069 6e20 656e 756d 6572 6174 6528  me in enumerate(
-0000c140: 6e6f 745f 616c 6967 6e65 645f 7365 7129  not_aligned_seq)
-0000c150: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
-0000c160: 2020 696e 6465 7820 3d20 6161 5f6e 616d    index = aa_nam
-0000c170: 6573 5f64 6963 745b 6161 5f6e 616d 655d  es_dict[aa_name]
-0000c180: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-0000c190: 2061 615f 696e 666f 5b69 6e64 6578 5f61   aa_info[index_a
-0000c1a0: 612b 325d 203d 6e70 2e68 7374 6163 6b28  a+2] =np.hstack(
-0000c1b0: 5b69 6e64 6578 2c20 5b30 5d2a 3239 5d29  [index, [0]*29])
-0000c1c0: 0a20 2020 2020 2020 206e 6f74 5f61 6c69  .        not_ali
-0000c1d0: 676e 6564 5f73 6571 735f 6672 6f6d 5f61  gned_seqs_from_a
-0000c1e0: 6c69 676e 6d65 6e74 5f66 696c 655b 6b65  lignment_file[ke
-0000c1f0: 795d 203d 2061 615f 696e 666f 0a20 2020  y] = aa_info.   
-0000c200: 2020 2020 2023 6469 6374 5f61 6c69 676e       #dict_align
-0000c210: 6d65 6e74 5f32 5b6b 6579 5d20 3d20 6161  ment_2[key] = aa
-0000c220: 5f69 6e66 6f0a 0a20 2020 2074 7265 6520  _info..    tree 
-0000c230: 3d20 696e 6665 725f 7472 6565 2861 6c69  = infer_tree(ali
-0000c240: 676e 6d65 6e74 3d61 6c69 676e 6d65 6e74  gnment=alignment
-0000c250: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-0000c260: 2020 2020 2020 2020 616c 6967 6e6d 656e          alignmen
-0000c270: 745f 6669 6c65 5f6e 616d 653d 616c 6967  t_file_name=alig
-0000c280: 6e6d 656e 745f 6669 6c65 2c0a 2020 2020  nment_file,.    
+0000b7b0: 2020 2020 2020 2020 616e 6e6f 7461 7469          annotati
+0000b7c0: 6f6e 733d 7b22 6d6f 6c65 6375 6c65 5f74  ons={"molecule_t
+0000b7d0: 7970 6522 3a20 2270 726f 7465 696e 227d  ype": "protein"}
+0000b7e0: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+0000b7f0: 2020 7064 625f 7265 636f 7264 732e 6170    pdb_records.ap
+0000b800: 7065 6e64 2870 6462 5f72 6563 6f72 6429  pend(pdb_record)
+0000b810: 0a20 2020 2020 2020 2020 2020 2053 6571  .            Seq
+0000b820: 494f 2e77 7269 7465 2870 6462 5f72 6563  IO.write(pdb_rec
+0000b830: 6f72 6473 2c20 6f75 7470 7574 5f68 616e  ords, output_han
+0000b840: 646c 652c 2022 6661 7374 6122 290a 2020  dle, "fasta").  
+0000b850: 2020 2020 2020 6f75 7470 7574 5f68 616e        output_han
+0000b860: 646c 652e 636c 6f73 6528 290a 2020 2020  dle.close().    
+0000b870: 2320 4869 6768 6c69 6768 743a 2041 6c69  # Highlight: Ali
+0000b880: 676e 2074 6865 2070 6f6c 7970 6570 7469  gn the polypepti
+0000b890: 6465 732f 7365 7175 656e 6365 7320 616e  des/sequences an
+0000b8a0: 6420 7772 6974 6520 746f 2061 2066 6173  d write to a fas
+0000b8b0: 7461 2061 6e67 6c65 735f 6c69 7374 2066  ta angles_list f
+0000b8c0: 696c 650a 2020 2020 6469 6374 5f61 6c69  ile.    dict_ali
+0000b8d0: 676e 6d65 6e74 2c61 6c69 676e 6d65 6e74  gnment,alignment
+0000b8e0: 203d 2069 6e66 6572 5f61 6c69 676e 6d65   = infer_alignme
+0000b8f0: 6e74 2861 6c69 676e 6d65 6e74 5f66 696c  nt(alignment_fil
+0000b900: 652c 696e 7075 745f 6e61 6d65 5f66 696c  e,input_name_fil
+0000b910: 653d 6661 7374 615f 6669 6c65 2c6f 7574  e=fasta_file,out
+0000b920: 7075 745f 6e61 6d65 5f66 696c 653d 227b  put_name_file="{
+0000b930: 7d2f 7b7d 2f7b 7d2e 6d61 6666 7422 2e66  }/{}/{}.mafft".f
+0000b940: 6f72 6d61 7428 7374 6f72 6167 655f 666f  ormat(storage_fo
+0000b950: 6c64 6572 2c6e 616d 655f 6669 6c65 2c6e  lder,name_file,n
+0000b960: 616d 655f 6669 6c65 2929 0a20 2020 2023  ame_file)).    #
+0000b970: 6361 6c63 756c 6174 655f 7061 6972 7769  calculate_pairwi
+0000b980: 7365 5f64 6973 7461 6e63 6528 6e61 6d65  se_distance(name
+0000b990: 5f66 696c 652c 616c 6967 6e6d 656e 7429  _file,alignment)
+0000b9a0: 0a20 2020 2061 6c69 676e 6d65 6e74 5f66  .    alignment_f
+0000b9b0: 696c 6520 3d20 5b61 6c69 676e 6d65 6e74  ile = [alignment
+0000b9c0: 5f66 696c 6520 6966 2061 6c69 676e 6d65  _file if alignme
+0000b9d0: 6e74 5f66 696c 6520 656c 7365 2022 7b7d  nt_file else "{}
+0000b9e0: 2f7b 7d2f 7b7d 2e6d 6166 6674 222e 666f  /{}/{}.mafft".fo
+0000b9f0: 726d 6174 2873 746f 7261 6765 5f66 6f6c  rmat(storage_fol
+0000ba00: 6465 722c 6e61 6d65 5f66 696c 652c 6e61  der,name_file,na
+0000ba10: 6d65 5f66 696c 6529 5d5b 305d 0a20 2020  me_file)][0].   
+0000ba20: 2023 4869 6768 6c69 6768 743a 2063 6865   #Highlight: che
+0000ba30: 636b 696e 6720 7468 6174 2074 6865 2073  cking that the s
+0000ba40: 656c 6563 7465 6420 6e75 6d62 6572 206f  elected number o
+0000ba50: 6620 7072 6f62 6162 696c 6974 6965 7320  f probabilities 
+0000ba60: 6973 2063 6f72 7265 6374 0a20 2020 2073  is correct.    s
+0000ba70: 756d 6d61 7279 5f61 615f 7072 6f62 7320  ummary_aa_probs 
+0000ba80: 3d20 5b76 616c 6964 6174 655f 7365 7175  = [validate_sequ
+0000ba90: 656e 6365 5f61 6c70 6861 6265 7428 7661  ence_alphabet(va
+0000baa0: 6c75 6529 2066 6f72 206b 6579 2c76 616c  lue) for key,val
+0000bab0: 7565 2069 6e20 6469 6374 5f61 6c69 676e  ue in dict_align
+0000bac0: 6d65 6e74 2e69 7465 6d73 2829 5d20 2366  ment.items()] #f
+0000bad0: 696e 6473 2074 6865 2061 6c70 6861 6265  inds the alphabe
+0000bae0: 7473 206f 6620 6561 6368 206f 6620 7468  ts of each of th
+0000baf0: 6520 7365 7175 656e 6365 7320 696e 2074  e sequences in t
+0000bb00: 6865 2061 6c69 676e 6d65 6e74 2c20 6368  he alignment, ch
+0000bb10: 6563 6b73 2066 6f72 2064 6e61 0a20 2020  ecks for dna.   
+0000bb20: 2061 615f 7072 6f62 7320 3d20 6d61 7828   aa_probs = max(
+0000bb30: 6161 5f70 726f 6273 2c6d 6178 2873 756d  aa_probs,max(sum
+0000bb40: 6d61 7279 5f61 615f 7072 6f62 7329 2920  mary_aa_probs)) 
+0000bb50: 2369 6620 7468 6520 696e 7075 7420 6161  #if the input aa
+0000bb60: 5f70 726f 6273 2069 7320 6469 6666 6572  _probs is differ
+0000bb70: 656e 7420 6672 6f6d 2074 686f 7365 2066  ent from those f
+0000bb80: 6f75 6e64 2c20 7468 6520 6161 5f70 726f  ound, the aa_pro
+0000bb90: 6273 2063 6861 6e67 652e 2041 6e64 2061  bs change. And a
+0000bba0: 6c73 6f20 7468 6520 6161 2073 7562 7374  lso the aa subst
+0000bbb0: 6974 7574 696f 6e20 206d 6174 7269 780a  itution  matrix.
+0000bbc0: 2020 2020 6161 5f6e 616d 6573 5f64 6963      aa_names_dic
+0000bbd0: 7420 3d20 616d 696e 6f61 6369 645f 6e61  t = aminoacid_na
+0000bbe0: 6d65 735f 6469 6374 2861 615f 7072 6f62  mes_dict(aa_prob
+0000bbf0: 7329 0a20 2020 2023 4869 6768 6c69 6768  s).    #Highligh
+0000bc00: 743a 2049 6620 7468 6520 6161 2073 6571  t: If the aa seq
+0000bc10: 7565 6e63 6573 2064 6f20 6e6f 7420 636f  uences do not co
+0000bc20: 6d65 2066 726f 6d20 2050 4442 2066 696c  me from  PDB fil
+0000bc30: 6573 2c20 7468 6579 2063 6f6d 6520 6672  es, they come fr
+0000bc40: 6f6d 2061 6e20 616c 6967 6e6d 656e 7420  om an alignment 
+0000bc50: 6669 6c65 2074 6861 7420 6e65 6564 7320  file that needs 
+0000bc60: 746f 2062 6520 7072 6f63 6573 7365 640a  to be processed.
+0000bc70: 2020 2020 2364 6963 745f 616c 6967 6e6d      #dict_alignm
+0000bc80: 656e 745f 3220 3d20 6469 6374 2e66 726f  ent_2 = dict.fro
+0000bc90: 6d6b 6579 7328 6469 6374 5f61 6c69 676e  mkeys(dict_align
+0000bca0: 6d65 6e74 2e6b 6579 7328 2929 0a20 2020  ment.keys()).   
+0000bcb0: 206e 6f74 5f61 6c69 676e 6564 5f73 6571   not_aligned_seq
+0000bcc0: 735f 6672 6f6d 5f61 6c69 676e 6d65 6e74  s_from_alignment
+0000bcd0: 5f66 696c 6520 3d7b 7d0a 2020 2020 666f  _file ={}.    fo
+0000bce0: 7220 6b65 792c 7661 6c75 6520 696e 2064  r key,value in d
+0000bcf0: 6963 745f 616c 6967 6e6d 656e 742e 6974  ict_alignment.it
+0000bd00: 656d 7328 293a 0a20 2020 2020 2020 2061  ems():.        a
+0000bd10: 6c69 676e 6564 5f73 6571 203d 206c 6973  ligned_seq = lis
+0000bd20: 7428 6469 6374 5f61 6c69 676e 6d65 6e74  t(dict_alignment
+0000bd30: 5b6b 6579 5d29 0a20 2020 2020 2020 2023  [key]).        #
+0000bd40: 6e6f 5f67 6170 5f69 6e64 6578 6573 203d  no_gap_indexes =
+0000bd50: 206e 702e 7768 6572 6528 6e70 2e61 7272   np.where(np.arr
+0000bd60: 6179 2861 6c69 676e 6564 5f73 6571 2920  ay(aligned_seq) 
+0000bd70: 213d 2022 2d22 295b 305d 202b 2032 2020  != "-")[0] + 2  
+0000bd80: 2320 706c 7573 2032 2069 6e20 6f72 6465  # plus 2 in orde
+0000bd90: 7220 746f 206d 616b 6520 7468 6520 696e  r to make the in
+0000bda0: 6465 7865 7320 6669 7420 696e 2074 6865  dexes fit in the
+0000bdb0: 2066 696e 616c 2064 6174 6166 7261 6d65   final dataframe
+0000bdc0: 0a20 2020 2020 2020 206e 6f74 5f61 6c69  .        not_ali
+0000bdd0: 676e 6564 5f73 6571 203d 6c69 7374 2866  gned_seq =list(f
+0000bde0: 696c 7465 7228 6c61 6d62 6461 2061 3a20  ilter(lambda a: 
+0000bdf0: 6120 213d 2022 2d22 2c20 616c 6967 6e65  a != "-", aligne
+0000be00: 645f 7365 7129 290a 2020 2020 2020 2020  d_seq)).        
+0000be10: 7365 715f 6c65 6e20 3d20 6c65 6e28 6e6f  seq_len = len(no
+0000be20: 745f 616c 6967 6e65 645f 7365 7129 0a20  t_aligned_seq). 
+0000be30: 2020 2020 2020 2067 6974 5f76 6563 746f         git_vecto
+0000be40: 7220 3d20 6e70 2e7a 6572 6f73 2833 3029  r = np.zeros(30)
+0000be50: 2023 6661 6b65 2067 6974 2076 6563 746f   #fake git vecto
+0000be60: 720a 2020 2020 2020 2020 6161 5f69 6e66  r.        aa_inf
+0000be70: 6f20 3d20 6e70 2e7a 6572 6f73 2828 7365  o = np.zeros((se
+0000be80: 715f 6c65 6e20 2b20 322c 2033 3029 290a  q_len + 2, 30)).
+0000be90: 2020 2020 2020 2020 6161 5f69 6e66 6f5b          aa_info[
+0000bea0: 305d 203d 206e 702e 6873 7461 636b 285b  0] = np.hstack([
+0000beb0: 7365 715f 6c65 6e2c 5b30 5d2a 3239 5d29  seq_len,[0]*29])
+0000bec0: 2023 6669 7273 7420 726f 7720 636f 6e74   #first row cont
+0000bed0: 6169 6e73 2073 6f6d 6520 7365 7175 656e  ains some sequen
+0000bee0: 6365 206c 656e 6774 6820 696e 666f 726d  ce length inform
+0000bef0: 6174 696f 6e0a 2020 2020 2020 2020 6161  ation.        aa
+0000bf00: 5f69 6e66 6f5b 315d 203d 2067 6974 5f76  _info[1] = git_v
+0000bf10: 6563 746f 7220 2373 6563 6f6e 6420 726f  ector #second ro
+0000bf20: 7720 636f 6e74 6169 6e73 2061 2067 6974  w contains a git
+0000bf30: 2076 6563 746f 7220 286e 6576 6572 2075   vector (never u
+0000bf40: 7365 6420 6275 7420 7765 2063 6f75 6c64  sed but we could
+0000bf50: 2075 7365 2069 7420 666f 7220 736f 6d65   use it for some
+0000bf60: 7468 696e 6720 656c 7365 290a 2020 2020  thing else).    
+0000bf70: 2020 2020 6966 206f 6e65 5f68 6f74 5f65      if one_hot_e
+0000bf80: 6e63 6f64 696e 673a 0a20 2020 2020 2020  ncoding:.       
+0000bf90: 2020 2020 2066 6f72 2069 6e64 6578 5f61       for index_a
+0000bfa0: 2c20 6161 5f6e 616d 6520 696e 2065 6e75  , aa_name in enu
+0000bfb0: 6d65 7261 7465 286e 6f74 5f61 6c69 676e  merate(not_align
+0000bfc0: 6564 5f73 6571 293a 0a20 2020 2020 2020  ed_seq):.       
+0000bfd0: 2020 2020 2020 2020 206f 6e65 5f68 6f74           one_hot
+0000bfe0: 203d 206e 702e 7a65 726f 7328 6161 5f70   = np.zeros(aa_p
+0000bff0: 726f 6273 290a 2020 2020 2020 2020 2020  robs).          
+0000c000: 2020 2020 2020 696e 6465 7820 3d20 6161        index = aa
+0000c010: 5f6e 616d 6573 5f64 6963 745b 6161 5f6e  _names_dict[aa_n
+0000c020: 616d 655d 0a20 2020 2020 2020 2020 2020  ame].           
+0000c030: 2020 2020 206f 6e65 5f68 6f74 5b69 6e64       one_hot[ind
+0000c040: 6578 5d20 3d20 310a 2020 2020 2020 2020  ex] = 1.        
+0000c050: 2020 2020 2020 2020 6578 7472 615f 7370          extra_sp
+0000c060: 6163 6520 3d20 3330 2d61 615f 7072 6f62  ace = 30-aa_prob
+0000c070: 730a 2020 2020 2020 2020 2020 2020 2020  s.              
+0000c080: 2020 6161 5f69 6e66 6f5b 696e 6465 785f    aa_info[index_
+0000c090: 612b 325d 203d 206e 702e 6873 7461 636b  a+2] = np.hstack
+0000c0a0: 285b 6f6e 655f 686f 742c 5b30 5d2a 6578  ([one_hot,[0]*ex
+0000c0b0: 7472 615f 7370 6163 655d 2920 2365 7874  tra_space]) #ext
+0000c0c0: 7261 5f73 7061 6365 203d 2031 3020 666f  ra_space = 10 fo
+0000c0d0: 7220 3230 2061 615f 7072 6f62 732c 2039  r 20 aa_probs, 9
+0000c0e0: 2066 6f72 2032 3120 6161 5f70 726f 6273   for 21 aa_probs
+0000c0f0: 2c20 3820 666f 7220 3232 2061 615f 7072  , 8 for 22 aa_pr
+0000c100: 6f62 730a 2020 2020 2020 2020 656c 7365  obs.        else
+0000c110: 3a0a 2020 2020 2020 2020 2020 2020 666f  :.            fo
+0000c120: 7220 696e 6465 785f 6161 2c20 6161 5f6e  r index_aa, aa_n
+0000c130: 616d 6520 696e 2065 6e75 6d65 7261 7465  ame in enumerate
+0000c140: 286e 6f74 5f61 6c69 676e 6564 5f73 6571  (not_aligned_seq
+0000c150: 293a 0a20 2020 2020 2020 2020 2020 2020  ):.             
+0000c160: 2020 2069 6e64 6578 203d 2061 615f 6e61     index = aa_na
+0000c170: 6d65 735f 6469 6374 5b61 615f 6e61 6d65  mes_dict[aa_name
+0000c180: 5d0a 2020 2020 2020 2020 2020 2020 2020  ].              
+0000c190: 2020 6161 5f69 6e66 6f5b 696e 6465 785f    aa_info[index_
+0000c1a0: 6161 2b32 5d20 3d6e 702e 6873 7461 636b  aa+2] =np.hstack
+0000c1b0: 285b 696e 6465 782c 205b 305d 2a32 395d  ([index, [0]*29]
+0000c1c0: 290a 2020 2020 2020 2020 6e6f 745f 616c  ).        not_al
+0000c1d0: 6967 6e65 645f 7365 7173 5f66 726f 6d5f  igned_seqs_from_
+0000c1e0: 616c 6967 6e6d 656e 745f 6669 6c65 5b6b  alignment_file[k
+0000c1f0: 6579 5d20 3d20 6161 5f69 6e66 6f0a 2020  ey] = aa_info.  
+0000c200: 2020 2020 2020 2364 6963 745f 616c 6967        #dict_alig
+0000c210: 6e6d 656e 745f 325b 6b65 795d 203d 2061  nment_2[key] = a
+0000c220: 615f 696e 666f 0a0a 2020 2020 7472 6565  a_info..    tree
+0000c230: 203d 2069 6e66 6572 5f74 7265 6528 616c   = infer_tree(al
+0000c240: 6967 6e6d 656e 743d 616c 6967 6e6d 656e  ignment=alignmen
+0000c250: 742c 0a20 2020 2020 2020 2020 2020 2020  t,.             
+0000c260: 2020 2020 2020 2020 2061 6c69 676e 6d65           alignme
+0000c270: 6e74 5f66 696c 655f 6e61 6d65 3d61 6c69  nt_file_name=ali
+0000c280: 676e 6d65 6e74 5f66 696c 652c 0a20 2020  gnment_file,.   
 0000c290: 2020 2020 2020 2020 2020 2020 2020 2020                  
-0000c2a0: 2020 6e61 6d65 3d6e 616d 655f 6669 6c65    name=name_file
-0000c2b0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-0000c2c0: 2020 2020 2020 2020 6d65 7468 6f64 3d6d          method=m
-0000c2d0: 6574 686f 642c 0a20 2020 2020 2020 2020  ethod,.         
-0000c2e0: 2020 2020 2020 2020 2020 2020 2074 7265               tre
-0000c2f0: 655f 6669 6c65 5f6e 616d 653d 227b 7d2f  e_file_name="{}/
-0000c300: 7b7d 2f7b 7d2e 7472 6565 222e 666f 726d  {}/{}.tree".form
-0000c310: 6174 2873 746f 7261 6765 5f66 6f6c 6465  at(storage_folde
-0000c320: 722c 6e61 6d65 5f66 696c 652c 6e61 6d65  r,name_file,name
-0000c330: 5f66 696c 6529 2c0a 2020 2020 2020 2020  _file),.        
-0000c340: 2020 2020 2020 2020 2020 2020 2020 7472                tr
-0000c350: 6565 5f66 696c 653d 7472 6565 5f66 696c  ee_file=tree_fil
-0000c360: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
-0000c370: 2020 2020 2020 2020 2073 746f 7261 6765           storage
-0000c380: 5f66 6f6c 6465 723d 227b 7d2f 7b7d 222e  _folder="{}/{}".
-0000c390: 666f 726d 6174 2873 746f 7261 6765 5f66  format(storage_f
-0000c3a0: 6f6c 6465 722c 6e61 6d65 5f66 696c 6529  older,name_file)
-0000c3b0: 290a 2020 2020 6d61 785f 6c65 6e67 6874  ).    max_lenght
-0000c3c0: 203d 2061 6c69 676e 6d65 6e74 2e67 6574   = alignment.get
-0000c3d0: 5f61 6c69 676e 6d65 6e74 5f6c 656e 6774  _alignment_lengt
-0000c3e0: 6828 290a 0a20 2020 2023 4869 6768 6c69  h()..    #Highli
-0000c3f0: 6768 743a 2043 6f6d 6269 6e69 6e67 2073  ght: Combining s
-0000c400: 6571 7565 6e63 6573 2069 6e20 7468 6520  equences in the 
-0000c410: 616c 6967 6e6d 656e 7420 7468 6174 2068  alignment that h
-0000c420: 6176 6520 6120 5044 4220 7374 7275 6374  ave a PDB struct
-0000c430: 7572 6520 616e 6420 7468 6f73 6520 7768  ure and those wh
-0000c440: 6f20 646f 6e27 740a 2020 2020 6966 2070  o don't.    if p
-0000c450: 726f 745f 696e 666f 5f64 6963 743a 2023  rot_info_dict: #
-0000c460: 4f74 6865 7277 6973 6520 6974 2064 6f65  Otherwise it doe
-0000c470: 7320 6e6f 7420 6c6f 6f70 206f 7665 7220  s not loop over 
-0000c480: 656d 7074 7920 6469 6374 696f 6e61 7269  empty dictionari
-0000c490: 6573 2061 6e64 2064 6f65 7320 6e6f 7468  es and does noth
-0000c4a0: 696e 670a 2020 2020 2020 2020 6e6f 745f  ing.        not_
-0000c4b0: 616c 6967 6e65 645f 7365 7173 5f66 726f  aligned_seqs_fro
-0000c4c0: 6d5f 616c 6967 6e6d 656e 745f 6669 6c65  m_alignment_file
-0000c4d0: 2e75 7064 6174 6528 286b 2c20 7072 6f74  .update((k, prot
-0000c4e0: 5f69 6e66 6f5f 6469 6374 5b6b 5d29 2066  _info_dict[k]) f
-0000c4f0: 6f72 206b 2c20 7620 696e 206e 6f74 5f61  or k, v in not_a
-0000c500: 6c69 676e 6564 5f73 6571 735f 6672 6f6d  ligned_seqs_from
-0000c510: 5f61 6c69 676e 6d65 6e74 5f66 696c 652e  _alignment_file.
-0000c520: 6974 656d 7328 2920 6966 206b 2069 6e20  items() if k in 
-0000c530: 7072 6f74 5f69 6e66 6f5f 6469 6374 2e6b  prot_info_dict.k
-0000c540: 6579 7328 2929 2020 2375 7064 6174 6520  eys())  #update 
-0000c550: 7468 6520 616c 6967 6e6d 656e 7420 6b65  the alignment ke
-0000c560: 7973 2077 6974 6820 7468 6569 7220 686f  ys with their ho
-0000c570: 6d6f 6c6f 6720 7769 7468 2070 6462 2069  molog with pdb i
-0000c580: 6e66 6f72 6d61 7469 6f6e 2e20 5570 6461  nformation. Upda
-0000c590: 7465 206f 6e6c 7920 7468 6f73 6520 7365  te only those se
-0000c5a0: 7175 656e 6365 7320 696e 2074 6865 2061  quences in the a
-0000c5b0: 6c69 676e 6d65 6e74 2e20 4d61 6666 7420  lignment. Mafft 
-0000c5c0: 616e 6420 7468 6520 4951 2074 7265 6520  and the IQ tree 
-0000c5d0: 7072 6f67 7261 6d20 6d69 6768 7420 6469  program might di
-0000c5e0: 7363 6172 6420 6469 6666 6572 656e 7420  scard different 
-0000c5f0: 7072 6f74 6569 6e73 2028 666f 7220 6578  proteins (for ex
-0000c600: 616d 706c 6520 7468 6579 2064 726f 7020  ample they drop 
-0000c610: 6469 6666 6572 656e 7420 6964 656e 7469  different identi
-0000c620: 6361 6c20 7072 6f74 6569 6e73 290a 2020  cal proteins).  
-0000c630: 2020 2020 2020 436f 6d62 696e 6564 5f64        Combined_d
-0000c640: 6963 7420 3d20 6e6f 745f 616c 6967 6e65  ict = not_aligne
-0000c650: 645f 7365 7173 5f66 726f 6d5f 616c 6967  d_seqs_from_alig
-0000c660: 6e6d 656e 745f 6669 6c65 0a20 2020 2065  nment_file.    e
-0000c670: 6c73 653a 0a20 2020 2020 2020 2043 6f6d  lse:.        Com
-0000c680: 6269 6e65 645f 6469 6374 203d 206e 6f74  bined_dict = not
-0000c690: 5f61 6c69 676e 6564 5f73 6571 735f 6672  _aligned_seqs_fr
-0000c6a0: 6f6d 5f61 6c69 676e 6d65 6e74 5f66 696c  om_alignment_fil
-0000c6b0: 650a 0a0a 2020 2020 6966 2072 656e 616d  e...    if renam
-0000c6c0: 655f 696e 7465 726e 616c 5f6e 6f64 6573  e_internal_nodes
-0000c6d0: 3a0a 2020 2020 2020 2020 6966 206e 616d  :.        if nam
-0000c6e0: 655f 6669 6c65 2e73 7461 7274 7377 6974  e_file.startswit
-0000c6f0: 6828 2273 696d 756c 6174 696f 6e73 2229  h("simulations")
-0000c700: 3a0a 2020 2020 2020 2020 2020 2074 7265  :.           tre
-0000c710: 6520 3d20 7265 6e61 6d65 5f74 7265 655f  e = rename_tree_
-0000c720: 696e 7465 726e 616c 5f6e 6f64 6573 5f73  internal_nodes_s
-0000c730: 696d 756c 6174 696f 6e73 2874 7265 652c  imulations(tree,
-0000c740: 7769 7468 5f69 6e64 6578 6573 3d46 616c  with_indexes=Fal
-0000c750: 7365 290a 2020 2020 2020 2020 656c 7365  se).        else
-0000c760: 3a0a 2020 2020 2020 2020 2020 2074 7265  :.           tre
-0000c770: 6520 3d20 7265 6e61 6d65 5f74 7265 655f  e = rename_tree_
-0000c780: 696e 7465 726e 616c 5f6e 6f64 6573 2874  internal_nodes(t
-0000c790: 7265 6529 0a0a 2020 2020 6c65 6166 735f  ree)..    leafs_
-0000c7a0: 6e61 6d65 7320 3d20 7472 6565 2e67 6574  names = tree.get
-0000c7b0: 5f6c 6561 665f 6e61 6d65 7328 290a 2020  _leaf_names().  
-0000c7c0: 2020 7069 636b 6c65 2e64 756d 7028 6c65    pickle.dump(le
-0000c7d0: 6166 735f 6e61 6d65 732c 6f70 656e 2827  afs_names,open('
-0000c7e0: 7b7d 2f7b 7d2f 7b7d 5f4c 6561 6673 5f6e  {}/{}/{}_Leafs_n
-0000c7f0: 616d 6573 5f6c 6973 742e 7027 2e66 6f72  ames_list.p'.for
-0000c800: 6d61 7428 7374 6f72 6167 655f 666f 6c64  mat(storage_fold
-0000c810: 6572 2c6e 616d 655f 6669 6c65 2c6e 616d  er,name_file,nam
-0000c820: 655f 6669 6c65 292c 2027 7762 2729 290a  e_file), 'wb')).
-0000c830: 2020 2020 6966 206c 656e 286c 6561 6673      if len(leafs
-0000c840: 5f6e 616d 6573 2920 3c3d 2032 3030 3a0a  _names) <= 200:.
-0000c850: 2020 2020 2020 2020 7072 696e 7428 2252          print("R
-0000c860: 656e 6465 7269 6e67 2074 7265 652e 2e2e  endering tree...
-0000c870: 2229 0a20 2020 2020 2020 2072 656e 6465  ").        rende
-0000c880: 725f 7472 6565 2874 7265 652c 2022 7b7d  r_tree(tree, "{}
-0000c890: 2f7b 7d22 2e66 6f72 6d61 7428 7374 6f72  /{}".format(stor
-0000c8a0: 6167 655f 666f 6c64 6572 2c6e 616d 655f  age_folder,name_
-0000c8b0: 6669 6c65 292c 206e 616d 655f 6669 6c65  file), name_file
-0000c8c0: 290a 2020 2020 696e 7465 726e 616c 5f6e  ).    internal_n
-0000c8d0: 6f64 6573 5f6e 616d 6573 203d 205b 6e6f  odes_names = [no
-0000c8e0: 6465 2e6e 616d 6520 666f 7220 6e6f 6465  de.name for node
-0000c8f0: 2069 6e20 7472 6565 2e74 7261 7665 7273   in tree.travers
-0000c900: 6528 2920 6966 206e 6f74 206e 6f64 652e  e() if not node.
-0000c910: 6973 5f6c 6561 6628 295d 0a0a 2020 2020  is_leaf()]..    
-0000c920: 616e 6365 7374 6f72 735f 616c 6c20 3d5b  ancestors_all =[
-0000c930: 5d0a 2020 2020 666f 7220 6e6f 6465 2069  ].    for node i
-0000c940: 6e20 7472 6565 2e74 7261 7665 7273 6528  n tree.traverse(
-0000c950: 293a 0a20 2020 2020 2020 2061 6e63 6573  ):.        ances
-0000c960: 746f 7273 5f6e 6f64 6520 3d5b 6e6f 6465  tors_node =[node
-0000c970: 2e6e 616d 652e 7265 706c 6163 6528 2227  .name.replace("'
-0000c980: 222c 2222 295d 2b5b 6e6f 6465 2e64 6973  ","")]+[node.dis
-0000c990: 745d 202b 5b61 6e63 6573 746f 722e 6e61  t] +[ancestor.na
-0000c9a0: 6d65 2e72 6570 6c61 6365 2822 2722 2c22  me.replace("'","
-0000c9b0: 2229 2066 6f72 2061 6e63 6573 746f 7220  ") for ancestor 
-0000c9c0: 696e 206e 6f64 652e 6765 745f 616e 6365  in node.get_ance
-0000c9d0: 7374 6f72 7328 295d 0a20 2020 2020 2020  stors()].       
-0000c9e0: 2061 6e63 6573 746f 7273 5f61 6c6c 2e61   ancestors_all.a
-0000c9f0: 7070 656e 6428 616e 6365 7374 6f72 735f  ppend(ancestors_
-0000ca00: 6e6f 6465 290a 2020 2020 6c65 6e67 7468  node).    length
-0000ca10: 203d 206d 6178 286d 6170 286c 656e 2c20   = max(map(len, 
-0000ca20: 616e 6365 7374 6f72 735f 616c 6c29 290a  ancestors_all)).
-0000ca30: 2020 2020 616e 6365 7374 6f72 735f 696e      ancestors_in
-0000ca40: 666f 203d 206e 702e 6172 7261 7928 5b78  fo = np.array([x
-0000ca50: 6920 2b20 5b4e 6f6e 655d 202a 2028 6c65  i + [None] * (le
-0000ca60: 6e67 7468 202d 206c 656e 2878 6929 2920  ngth - len(xi)) 
-0000ca70: 666f 7220 7869 2069 6e20 616e 6365 7374  for xi in ancest
-0000ca80: 6f72 735f 616c 6c5d 290a 0a20 2020 2074  ors_all])..    t
-0000ca90: 7265 655f 6c65 7665 6c6f 7264 6572 5f6e  ree_levelorder_n
-0000caa0: 616d 6573 203d 206e 702e 6173 6172 7261  ames = np.asarra
-0000cab0: 7928 5b6e 6f64 652e 6e61 6d65 2e72 6570  y([node.name.rep
-0000cac0: 6c61 6365 2822 2722 2c22 2229 2066 6f72  lace("'","") for
-0000cad0: 206e 6f64 6520 696e 2074 7265 652e 7472   node in tree.tr
-0000cae0: 6176 6572 7365 2829 5d29 0a0a 2020 2020  averse()])..    
-0000caf0: 7472 6565 5f6c 6576 656c 6f72 6465 725f  tree_levelorder_
-0000cb00: 6469 7374 203d 6e70 2e61 7361 7272 6179  dist =np.asarray
-0000cb10: 285b 6e6f 6465 2e64 6973 7420 666f 7220  ([node.dist for 
-0000cb20: 6e6f 6465 2069 6e20 7472 6565 2e74 7261  node in tree.tra
-0000cb30: 7665 7273 6528 295d 290a 2020 2020 2341  verse()]).    #A
-0000cb40: 6464 2074 6865 2069 6e64 6578 206f 6620  dd the index of 
-0000cb50: 7468 6520 7365 7175 656e 6365 2069 6e20  the sequence in 
-0000cb60: 7468 6520 7472 6565 2074 6f20 7468 6520  the tree to the 
-0000cb70: 7365 7120 6c65 6e67 7468 2061 7272 6179  seq length array
-0000cb80: 2069 6e66 6f0a 2020 2020 4461 7461 7365   info.    Datase
-0000cb90: 7420 3d20 6e70 2e7a 6572 6f73 2828 6c65  t = np.zeros((le
-0000cba0: 6e28 436f 6d62 696e 6564 5f64 6963 7429  n(Combined_dict)
-0000cbb0: 2c20 6d61 785f 6c65 6e67 6874 202b 2031  , max_lenght + 1
-0000cbc0: 202b 2031 202b 312c 2033 3029 2c64 7479   + 1 +1, 30),dty
-0000cbd0: 7065 3d6f 626a 6563 7429 2020 2320 3330  pe=object)  # 30
-0000cbe0: 2064 696d 2074 6f20 6163 636f 6d6f 6461   dim to accomoda
-0000cbf0: 7465 2067 6974 2076 6563 746f 7273 2e20  te git vectors. 
-0000cc00: 4361 7265 6675 6c20 7769 7468 2074 6865  Careful with the
-0000cc10: 202b 3320 2874 6f20 696e 636c 7564 6520   +3 (to include 
-0000cc20: 6769 742c 2073 6571 6c65 6e20 616e 6420  git, seqlen and 
-0000cc30: 726f 772f 7072 6f74 6569 6e20 6e61 6d65  row/protein name
-0000cc40: 290a 2020 2020 666f 7220 692c 2028 6b65  ).    for i, (ke
-0000cc50: 792c 7661 6c29 2069 6e20 656e 756d 6572  y,val) in enumer
-0000cc60: 6174 6528 436f 6d62 696e 6564 5f64 6963  ate(Combined_dic
-0000cc70: 742e 6974 656d 7328 2929 3a0a 2020 2020  t.items()):.    
-0000cc80: 2020 2020 616c 6967 6e65 645f 7365 7120      aligned_seq 
-0000cc90: 3d20 6c69 7374 2864 6963 745f 616c 6967  = list(dict_alig
-0000cca0: 6e6d 656e 745b 6b65 795d 2e73 7472 6970  nment[key].strip
-0000ccb0: 2822 2c22 2929 0a20 2020 2020 2020 206e  (",")).        n
-0000ccc0: 6f5f 6761 705f 696e 6465 7865 7320 3d20  o_gap_indexes = 
-0000ccd0: 6e70 2e77 6865 7265 286e 702e 6172 7261  np.where(np.arra
-0000cce0: 7928 616c 6967 6e65 645f 7365 7129 2021  y(aligned_seq) !
-0000ccf0: 3d20 222d 2229 5b30 5d20 2b20 3320 2023  = "-")[0] + 3  #
-0000cd00: 2070 6c75 7320 3320 696e 206f 7264 6572   plus 3 in order
-0000cd10: 2074 6f20 6d61 6b65 2074 6865 2069 6e64   to make the ind
-0000cd20: 6578 6573 2066 6974 2069 6e20 7468 6520  exes fit in the 
-0000cd30: 6669 6e61 6c20 6461 7461 6672 616d 650a  final dataframe.
-0000cd40: 2020 2020 2020 2020 4461 7461 7365 745b          Dataset[
-0000cd50: 692c 302c 305d 203d 206b 6579 2e72 6570  i,0,0] = key.rep
-0000cd60: 6c61 6365 2822 2722 2c22 2229 2023 726f  lace("'","") #ro
-0000cd70: 7720 6e61 6d65 2f73 6571 7565 6e63 6520  w name/sequence 
-0000cd80: 6e61 6d65 0a20 2020 2020 2020 2044 6174  name.        Dat
-0000cd90: 6173 6574 5b69 2c20 313a 335d 203d 2043  aset[i, 1:3] = C
-0000cda0: 6f6d 6269 6e65 645f 6469 6374 5b6b 6579  ombined_dict[key
-0000cdb0: 5d5b 3a32 5d20 2349 6e73 6572 7420 7365  ][:2] #Insert se
-0000cdc0: 7120 6c65 6e20 616e 6420 6769 7420 7665  q len and git ve
-0000cdd0: 6374 6f72 0a20 2020 2020 2020 2069 6620  ctor.        if 
-0000cde0: 6e61 6d65 5f66 696c 6520 696e 205b 2262  name_file in ["b
-0000cdf0: 656e 6368 6d61 726b 5f72 616e 6461 6c6c  enchmark_randall
-0000ce00: 222c 2262 656e 6368 6d61 726b 5f72 616e  ","benchmark_ran
-0000ce10: 6461 6c6c 5f6f 7269 6769 6e61 6c22 2c22  dall_original","
-0000ce20: 6265 6e63 686d 6172 6b5f 7261 6e64 616c  benchmark_randal
-0000ce30: 6c5f 6f72 6967 696e 616c 5f6e 616d 696e  l_original_namin
-0000ce40: 6722 5d3a 2374 6865 6972 206c 6561 7665  g"]:#their leave
-0000ce50: 7320 6861 7665 206e 756d 6265 7220 6e61  s have number na
-0000ce60: 6d65 732c 2073 6f20 7765 206b 6565 7020  mes, so we keep 
-0000ce70: 7468 656d 2069 6e73 7465 6164 206f 6620  them instead of 
-0000ce80: 6368 616e 6769 6e67 2074 6865 6d20 666f  changing them fo
-0000ce90: 7220 7468 6520 7472 6565 206c 6576 656c  r the tree level
-0000cea0: 206f 7264 6572 206f 6e65 730a 2020 2020   order ones.    
-0000ceb0: 2020 2020 2020 2020 4461 7461 7365 745b          Dataset[
-0000cec0: 692c 2031 2c20 315d 203d 2069 6e74 286b  i, 1, 1] = int(k
-0000ced0: 6579 2e72 6570 6c61 6365 2822 2722 2c20  ey.replace("'", 
-0000cee0: 2222 2929 0a20 2020 2020 2020 2065 6c69  "")).        eli
-0000cef0: 6620 6e61 6d65 5f66 696c 6520 696e 205b  f name_file in [
-0000cf00: 2250 4630 3130 3338 5f6c 6970 6374 695f  "PF01038_lipcti_
-0000cf10: 6d73 615f 6675 6e67 6922 5d3a 0a20 2020  msa_fungi"]:.   
-0000cf20: 2020 2020 2020 2020 2044 6174 6173 6574           Dataset
-0000cf30: 5b69 2c20 312c 2031 5d20 3d20 6e70 2e77  [i, 1, 1] = np.w
-0000cf40: 6865 7265 2874 7265 655f 6c65 7665 6c6f  here(tree_levelo
-0000cf50: 7264 6572 5f6e 616d 6573 203d 3d20 6b65  rder_names == ke
-0000cf60: 792e 7265 706c 6163 6528 223a 222c 225f  y.replace(":","_
-0000cf70: 2229 295b 305d 5b30 5d20 2023 2074 6865  "))[0][0]  # the
-0000cf80: 206e 6f64 6520 6e61 6d65 2077 696c 6c20   node name will 
-0000cf90: 6265 2069 7473 2070 6f73 6974 696f 6e20  be its position 
-0000cfa0: 696e 2074 6865 2074 7265 650a 2020 2020  in the tree.    
-0000cfb0: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
-0000cfc0: 2020 2020 2020 4461 7461 7365 745b 692c        Dataset[i,
-0000cfd0: 312c 315d 203d 206e 702e 7768 6572 6528  1,1] = np.where(
-0000cfe0: 7472 6565 5f6c 6576 656c 6f72 6465 725f  tree_levelorder_
-0000cff0: 6e61 6d65 7320 3d3d 206b 6579 2e72 6570  names == key.rep
-0000d000: 6c61 6365 2822 2722 2c22 2229 295b 305d  lace("'",""))[0]
-0000d010: 5b30 5d20 2374 6865 206e 6f64 6520 6e61  [0] #the node na
-0000d020: 6d65 2077 696c 6c20 6265 2069 7473 2070  me will be its p
-0000d030: 6f73 6974 696f 6e20 696e 2074 6865 2074  osition in the t
-0000d040: 7265 650a 2020 2020 2020 2020 4461 7461  ree.        Data
-0000d050: 7365 745b 692c 2031 2c20 325d 203d 2020  set[i, 1, 2] =  
-0000d060: 7472 6565 5f6c 6576 656c 6f72 6465 725f  tree_levelorder_
-0000d070: 6469 7374 5b44 6174 6173 6574 5b69 2c31  dist[Dataset[i,1
-0000d080: 2c31 5d5d 2023 6469 7374 616e 6365 2074  ,1]] #distance t
-0000d090: 6f20 7468 6520 726f 6f74 3f20 7468 6174  o the root? that
-0000d0a0: 2773 2061 6363 6f72 6469 6e67 2074 6f20  's according to 
-0000d0b0: 7468 6520 646f 6375 6d65 6e74 6174 696f  the documentatio
-0000d0c0: 6e20 7965 732c 2062 7574 2069 7320 6469  n yes, but is di
-0000d0d0: 6666 6572 656e 7420 6672 6f6d 2074 6865  fferent from the
-0000d0e0: 2070 6174 7269 7374 6963 2064 6973 7461   patristic dista
-0000d0f0: 6e63 6573 0a20 2020 2020 2020 2044 6174  nces.        Dat
-0000d100: 6173 6574 5b69 2c20 6e6f 5f67 6170 5f69  aset[i, no_gap_i
-0000d110: 6e64 6578 6573 5d20 3d20 436f 6d62 696e  ndexes] = Combin
-0000d120: 6564 5f64 6963 745b 6b65 795d 5b32 3a5d  ed_dict[key][2:]
-0000d130: 2020 2320 4173 7369 676e 2074 6865 2061    # Assign the a
-0000d140: 6120 696e 666f 2028 696e 636c 7564 696e  a info (includin
-0000d150: 6720 616e 676c 6573 2920 746f 2074 686f  g angles) to tho
-0000d160: 7365 2070 6f73 6974 696f 6e73 2077 6865  se positions whe
-0000d170: 7265 2074 6865 7265 2069 7320 6e6f 7420  re there is not 
-0000d180: 6120 6761 700a 2020 2020 2020 2020 6966  a gap.        if
-0000d190: 206f 6e65 5f68 6f74 5f65 6e63 6f64 696e   one_hot_encodin
-0000d1a0: 673a 0a20 2020 2020 2020 2020 2020 2023  g:.            #
-0000d1b0: 4869 6768 6c69 6768 743a 206e 6f5f 6761  Highlight: no_ga
-0000d1c0: 7020 696e 6465 7865 7320 6973 206e 6f74  p indexes is not
-0000d1d0: 2061 2062 6f6f 6c65 616e 2c20 7765 2068   a boolean, we h
-0000d1e0: 6176 6520 746f 2063 6f6e 7665 7274 2069  ave to convert i
-0000d1f0: 740a 2020 2020 2020 2020 2020 2020 6761  t.            ga
-0000d200: 7073 5f6d 6173 6b20 3d20 6e70 2e6f 6e65  ps_mask = np.one
-0000d210: 7328 6d61 785f 6c65 6e67 6874 2b33 2c6e  s(max_lenght+3,n
-0000d220: 702e 626f 6f6c 290a 2020 2020 2020 2020  p.bool).        
-0000d230: 2020 2020 6761 7073 5f6d 6173 6b5b 6e6f      gaps_mask[no
-0000d240: 5f67 6170 5f69 6e64 6578 6573 5d20 3d20  _gap_indexes] = 
-0000d250: 4661 6c73 6520 2364 6f20 6e6f 7420 6b65  False #do not ke
-0000d260: 6570 2074 6865 2070 6f73 6974 696f 6e73  ep the positions
-0000d270: 2077 6865 7265 2074 6865 7265 2069 7320   where there is 
-0000d280: 6e6f 7420 6120 6761 700a 2020 2020 2020  not a gap.      
-0000d290: 2020 2020 2020 6761 7073 5f6d 6173 6b5b        gaps_mask[
-0000d2a0: 305d 203d 2046 616c 7365 2023 6e6f 7420  0] = False #not 
-0000d2b0: 7468 6520 6669 7273 7420 7477 6f20 726f  the first two ro
-0000d2c0: 7773 0a20 2020 2020 2020 2020 2020 2067  ws.            g
-0000d2d0: 6170 735f 6d61 736b 5b31 5d20 3d20 4661  aps_mask[1] = Fa
-0000d2e0: 6c73 650a 2020 2020 2020 2020 2020 2020  lse.            
-0000d2f0: 6761 7073 5f6d 6173 6b5b 325d 203d 2046  gaps_mask[2] = F
-0000d300: 616c 7365 0a20 2020 2020 2020 2020 2020  alse.           
-0000d310: 2044 6174 6173 6574 5b69 2c67 6170 735f   Dataset[i,gaps_
-0000d320: 6d61 736b 2c30 5d20 3d20 312e 2023 6966  mask,0] = 1. #if
-0000d330: 206f 6e65 2068 6f74 2065 6e63 6f64 696e   one hot encodin
-0000d340: 6720 7468 6520 6761 7073 2077 6974 6820  g the gaps with 
-0000d350: 6265 2061 7373 6967 6e65 6420 7468 6520  be assigned the 
-0000d360: 6669 7273 7420 706f 7369 7469 6f6e 2069  first position i
-0000d370: 6e20 6f6e 6520 686f 7420 656e 636f 6469  n one hot encodi
-0000d380: 6e67 0a0a 2020 2020 2320 2052 6563 6f6e  ng..    #  Recon
-0000d390: 7374 7275 6374 2074 6865 2074 7265 6520  struct the tree 
-0000d3a0: 6672 6f6d 2074 6865 2064 6973 7461 6e63  from the distanc
-0000d3b0: 6520 6d61 7472 6978 0a20 2020 2070 7269  e matrix.    pri
-0000d3c0: 6e74 2822 4275 696c 6469 6e67 2070 6174  nt("Building pat
-0000d3d0: 7269 7374 6963 2061 6e64 2063 6c61 6469  ristic and cladi
-0000d3e0: 7374 6963 206d 6174 7269 6365 7320 2e2e  stic matrices ..
-0000d3f0: 2e22 290a 2020 2020 7472 6565 5f73 6176  .").    tree_sav
-0000d400: 6520 3d20 7064 2e44 6174 6146 7261 6d65  e = pd.DataFrame
-0000d410: 2861 6e63 6573 746f 7273 5f69 6e66 6f29  (ancestors_info)
-0000d420: 0a20 2020 2023 7472 6565 5f73 6176 6520  .    #tree_save 
-0000d430: 3d20 7064 2e44 6174 6146 7261 6d65 287b  = pd.DataFrame({
-0000d440: 224e 6f64 6573 5f4e 616d 6573 223a 7472  "Nodes_Names":tr
-0000d450: 6565 5f6c 6576 656c 6f72 6465 725f 6e61  ee_levelorder_na
-0000d460: 6d65 732e 746f 6c69 7374 2829 2c22 4469  mes.tolist(),"Di
-0000d470: 7374 616e 6365 5f74 6f5f 726f 6f74 223a  stance_to_root":
-0000d480: 7472 6565 5f6c 6576 656c 6f72 6465 725f  tree_levelorder_
-0000d490: 6469 7374 2e74 6f6c 6973 7428 297d 290a  dist.tolist()}).
-0000d4a0: 2020 2020 7472 6565 5f73 6176 652e 746f      tree_save.to
-0000d4b0: 5f63 7376 2822 7b7d 2f7b 7d2f 7b7d 5f74  _csv("{}/{}/{}_t
-0000d4c0: 7265 655f 6c65 7665 6c6f 7264 6572 5f69  ree_levelorder_i
-0000d4d0: 6e66 6f2e 6373 7622 2e66 6f72 6d61 7428  nfo.csv".format(
-0000d4e0: 7374 6f72 6167 655f 666f 6c64 6572 2c6e  storage_folder,n
-0000d4f0: 616d 655f 6669 6c65 2c6e 616d 655f 6669  ame_file,name_fi
-0000d500: 6c65 292c 7365 703d 225c 7422 290a 2020  le),sep="\t").  
-0000d510: 2020 6e6f 6465 735f 616e 645f 6c65 6166    nodes_and_leaf
-0000d520: 735f 6e61 6d65 7320 3d20 696e 7465 726e  s_names = intern
-0000d530: 616c 5f6e 6f64 6573 5f6e 616d 6573 202b  al_nodes_names +
-0000d540: 206c 6561 6673 5f6e 616d 6573 0a20 2020   leafs_names.   
-0000d550: 2063 616c 6375 6c61 7465 5f70 6174 7269   calculate_patri
-0000d560: 7374 6963 5f64 6973 7461 6e63 6528 6e61  stic_distance(na
-0000d570: 6d65 5f66 696c 652c 436f 6d62 696e 6564  me_file,Combined
-0000d580: 5f64 6963 742c 6e6f 6465 735f 616e 645f  _dict,nodes_and_
-0000d590: 6c65 6166 735f 6e61 6d65 732c 7472 6565  leafs_names,tree
-0000d5a0: 2c74 7265 655f 6669 6c65 2c22 7b7d 2f7b  ,tree_file,"{}/{
-0000d5b0: 7d22 2e66 6f72 6d61 7428 7374 6f72 6167  }".format(storag
-0000d5c0: 655f 666f 6c64 6572 2c6e 616d 655f 6669  e_folder,name_fi
-0000d5d0: 6c65 2929 0a20 2020 2063 616c 6375 6c61  le)).    calcula
-0000d5e0: 7465 5f63 6c6f 7365 7374 5f6c 6561 7665  te_closest_leave
-0000d5f0: 7328 6e61 6d65 5f66 696c 652c 7472 6565  s(name_file,tree
-0000d600: 2c22 7b7d 2f7b 7d22 2e66 6f72 6d61 7428  ,"{}/{}".format(
-0000d610: 7374 6f72 6167 655f 666f 6c64 6572 2c6e  storage_folder,n
-0000d620: 616d 655f 6669 6c65 2929 0a20 2020 2063  ame_file)).    c
-0000d630: 616c 6375 6c61 7465 5f64 6972 6563 746c  alculate_directl
-0000d640: 795f 6c69 6e6b 6564 5f6e 6f64 6573 286e  y_linked_nodes(n
-0000d650: 616d 655f 6669 6c65 2c20 7472 6565 2c22  ame_file, tree,"
-0000d660: 7b7d 2f7b 7d22 2e66 6f72 6d61 7428 7374  {}/{}".format(st
-0000d670: 6f72 6167 655f 666f 6c64 6572 2c6e 616d  orage_folder,nam
-0000d680: 655f 6669 6c65 2929 0a20 2020 2063 616c  e_file)).    cal
-0000d690: 6375 6c61 7465 5f64 6573 6365 6e64 616e  culate_descendan
-0000d6a0: 7473 286e 616d 655f 6669 6c65 2c74 7265  ts(name_file,tre
-0000d6b0: 652c 227b 7d2f 7b7d 222e 666f 726d 6174  e,"{}/{}".format
-0000d6c0: 2873 746f 7261 6765 5f66 6f6c 6465 722c  (storage_folder,
-0000d6d0: 6e61 6d65 5f66 696c 6529 290a 2020 2020  name_file)).    
-0000d6e0: 7072 696e 7428 2252 6561 6479 2061 6e64  print("Ready and
-0000d6f0: 2073 6176 6564 2122 290a 2020 2020 7072   saved!").    pr
-0000d700: 696e 7428 2242 7569 6c64 696e 6720 636c  int("Building cl
-0000d710: 6164 6573 2028 7761 726e 696e 673a 2063  ades (warning: c
-0000d720: 6f6c 6c61 7073 6573 2074 6865 206f 7269  ollapses the ori
-0000d730: 6769 6e61 6c20 7472 6565 2129 2229 0a20  ginal tree!)"). 
-0000d740: 2020 2064 6976 6964 655f 696e 746f 5f6d     divide_into_m
-0000d750: 6f6e 6f70 6879 6c65 7469 635f 636c 6164  onophyletic_clad
-0000d760: 6573 2874 7265 652c 227b 7d2f 7b7d 222e  es(tree,"{}/{}".
-0000d770: 666f 726d 6174 2873 746f 7261 6765 5f66  format(storage_f
-0000d780: 6f6c 6465 722c 6e61 6d65 5f66 696c 6529  older,name_file)
-0000d790: 2c6e 616d 655f 6669 6c65 290a 2020 2020  ,name_file).    
-0000d7a0: 6e70 2e73 6176 6528 227b 7d2f 7b7d 2f7b  np.save("{}/{}/{
-0000d7b0: 7d5f 6461 7461 7365 745f 6e75 6d70 795f  }_dataset_numpy_
-0000d7c0: 616c 6967 6e65 645f 7b7d 2e6e 7079 222e  aligned_{}.npy".
-0000d7d0: 666f 726d 6174 2873 746f 7261 6765 5f66  format(storage_f
-0000d7e0: 6f6c 6465 722c 6e61 6d65 5f66 696c 652c  older,name_file,
-0000d7f0: 6e61 6d65 5f66 696c 652c 6f6e 655f 686f  name_file,one_ho
-0000d800: 745f 6c61 6265 6c5b 305d 292c 2044 6174  t_label[0]), Dat
-0000d810: 6173 6574 290a 2020 2020 6d61 785f 6c65  aset).    max_le
-0000d820: 6e67 6874 5f6e 6f74 5f61 6c69 676e 6564  nght_not_aligned
-0000d830: 203d 206d 6178 285b 696e 7428 7365 7175   = max([int(sequ
-0000d840: 656e 6365 5b30 5d5b 305d 2920 666f 7220  ence[0][0]) for 
-0000d850: 6964 782c 7365 7175 656e 6365 2069 6e20  idx,sequence in 
-0000d860: 436f 6d62 696e 6564 5f64 6963 742e 6974  Combined_dict.it
-0000d870: 656d 7328 295d 2920 2346 696e 6420 7468  ems()]) #Find th
-0000d880: 6520 6c61 7267 6573 7420 7365 7175 656e  e largest sequen
-0000d890: 6365 2077 6974 686f 7574 2062 6569 6e67  ce without being
-0000d8a0: 2061 6c69 676e 6564 0a20 2020 2070 7269   aligned.    pri
-0000d8b0: 6e74 2822 4372 6561 7469 6e67 206e 6f74  nt("Creating not
-0000d8c0: 2061 6c69 676e 6564 2064 6174 6173 6574   aligned dataset
-0000d8d0: 2e2e 2e22 290a 2020 2020 4461 7461 7365  ...").    Datase
-0000d8e0: 745f 6e6f 745f 616c 6967 6e65 6420 3d20  t_not_aligned = 
-0000d8f0: 6e70 2e7a 6572 6f73 2828 6c65 6e28 436f  np.zeros((len(Co
-0000d900: 6d62 696e 6564 5f64 6963 7429 2c20 6d61  mbined_dict), ma
-0000d910: 785f 6c65 6e67 6874 5f6e 6f74 5f61 6c69  x_lenght_not_ali
-0000d920: 676e 6564 202b 332c 2033 3029 2c20 6474  gned +3, 30), dt
-0000d930: 7970 653d 6f62 6a65 6374 2920 2023 2033  ype=object)  # 3
-0000d940: 3020 666f 7220 6675 7475 7265 2067 6974  0 for future git
-0000d950: 2076 6563 746f 7273 2e20 4361 7265 6675   vectors. Carefu
-0000d960: 6c20 7769 7468 2074 6865 202b 330a 2020  l with the +3.  
-0000d970: 2020 666f 7220 692c 286b 6579 2c76 616c    for i,(key,val
-0000d980: 7565 2920 696e 2065 6e75 6d65 7261 7465  ue) in enumerate
-0000d990: 2843 6f6d 6269 6e65 645f 6469 6374 2e69  (Combined_dict.i
-0000d9a0: 7465 6d73 2829 293a 0a20 2020 2020 2020  tems()):.       
-0000d9b0: 2044 6174 6173 6574 5f6e 6f74 5f61 6c69   Dataset_not_ali
-0000d9c0: 676e 6564 5b69 2c30 2c30 5d20 3d20 6b65  gned[i,0,0] = ke
-0000d9d0: 790a 2020 2020 2020 2020 4461 7461 7365  y.        Datase
-0000d9e0: 745f 6e6f 745f 616c 6967 6e65 645b 692c  t_not_aligned[i,
-0000d9f0: 2031 3a33 5d20 3d20 436f 6d62 696e 6564   1:3] = Combined
-0000da00: 5f64 6963 745b 6b65 795d 5b3a 325d 2023  _dict[key][:2] #
-0000da10: 4669 6c6c 2069 6e20 7468 6520 7365 7175  Fill in the sequ
-0000da20: 656e 6365 206c 656e 6768 7420 616e 6420  ence lenght and 
-0000da30: 7468 6520 6769 7420 7665 6374 6f72 0a20  the git vector. 
-0000da40: 2020 2020 2020 2069 6620 6e61 6d65 5f66         if name_f
-0000da50: 696c 6520 696e 205b 2262 656e 6368 6d61  ile in ["benchma
-0000da60: 726b 5f72 616e 6461 6c6c 5f6f 7269 6769  rk_randall_origi
-0000da70: 6e61 6c5f 6e61 6d69 6e67 225d 3a0a 2020  nal_naming"]:.  
-0000da80: 2020 2020 2020 2020 2020 4461 7461 7365            Datase
-0000da90: 745b 692c 2031 2c20 315d 203d 2069 6e74  t[i, 1, 1] = int
-0000daa0: 286b 6579 2e72 6570 6c61 6365 2822 2722  (key.replace("'"
-0000dab0: 2c20 2222 2929 0a20 2020 2020 2020 2065  , "")).        e
-0000dac0: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
-0000dad0: 2044 6174 6173 6574 5b69 2c31 2c31 5d20   Dataset[i,1,1] 
-0000dae0: 3d20 6e70 2e77 6865 7265 2874 7265 655f  = np.where(tree_
-0000daf0: 6c65 7665 6c6f 7264 6572 5f6e 616d 6573  levelorder_names
-0000db00: 203d 3d20 6b65 792e 7265 706c 6163 6528   == key.replace(
-0000db10: 2227 222c 2222 2929 5b30 5d5b 305d 2023  "'",""))[0][0] #
-0000db20: 706f 7369 7469 6f6e 2069 6e20 7468 6520  position in the 
-0000db30: 7472 6565 0a20 2020 2020 2020 2044 6174  tree.        Dat
-0000db40: 6173 6574 5f6e 6f74 5f61 6c69 676e 6564  aset_not_aligned
-0000db50: 5b69 2c20 312c 2032 5d20 3d20 2074 7265  [i, 1, 2] =  tre
-0000db60: 655f 6c65 7665 6c6f 7264 6572 5f64 6973  e_levelorder_dis
-0000db70: 745b 4461 7461 7365 745b 692c 312c 315d  t[Dataset[i,1,1]
-0000db80: 5d0a 2020 2020 2020 2020 4461 7461 7365  ].        Datase
-0000db90: 745f 6e6f 745f 616c 6967 6e65 645b 692c  t_not_aligned[i,
-0000dba0: 2033 3a69 6e74 2843 6f6d 6269 6e65 645f   3:int(Combined_
-0000dbb0: 6469 6374 5b6b 6579 5d5b 305d 5b30 5d29  dict[key][0][0])
-0000dbc0: 202b 335d 203d 2043 6f6d 6269 6e65 645f   +3] = Combined_
-0000dbd0: 6469 6374 5b6b 6579 5d5b 323a 5d20 2346  dict[key][2:] #F
-0000dbe0: 696c 6c20 696e 2074 6865 2061 6d69 6e6f  ill in the amino
-0000dbf0: 2061 6369 6473 2022 6c65 7474 6572 7322   acids "letters"
-0000dc00: 2f22 6e75 6d62 6572 7322 2061 6e64 2074  /"numbers" and t
-0000dc10: 6865 6972 2061 6e67 6c65 730a 2020 2020  heir angles.    
-0000dc20: 2020 2020 6966 206f 6e65 5f68 6f74 5f65      if one_hot_e
-0000dc30: 6e63 6f64 696e 673a 0a20 2020 2020 2020  ncoding:.       
-0000dc40: 2020 2020 2023 2023 4869 6768 6c69 6768       # #Highligh
-0000dc50: 743a 206e 6f5f 6761 7020 696e 6465 7865  t: no_gap indexe
-0000dc60: 7320 6973 206e 6f74 2061 2062 6f6f 6c65  s is not a boole
-0000dc70: 616e 2c20 7765 2068 6176 6520 746f 2063  an, we have to c
-0000dc80: 6f6e 7665 7274 2069 740a 2020 2020 2020  onvert it.      
-0000dc90: 2020 2020 2020 2320 6761 7073 5f6d 6173        # gaps_mas
-0000dca0: 6b20 3d20 6e70 2e6f 6e65 7328 6d61 785f  k = np.ones(max_
-0000dcb0: 6c65 6e67 6874 2b33 2c6e 702e 626f 6f6c  lenght+3,np.bool
-0000dcc0: 290a 2020 2020 2020 2020 2020 2020 2320  ).            # 
-0000dcd0: 6761 7073 5f6d 6173 6b5b 6e6f 5f67 6170  gaps_mask[no_gap
-0000dce0: 5f69 6e64 6578 6573 5d20 3d20 4661 6c73  _indexes] = Fals
-0000dcf0: 6520 2364 6f20 6e6f 7420 6b65 6570 2074  e #do not keep t
-0000dd00: 6865 2070 6f73 6974 696f 6e73 2077 6865  he positions whe
-0000dd10: 7265 2074 6865 7265 2069 7320 6e6f 7420  re there is not 
-0000dd20: 6120 6761 700a 2020 2020 2020 2020 2020  a gap.          
-0000dd30: 2020 2320 6761 7073 5f6d 6173 6b5b 305d    # gaps_mask[0]
-0000dd40: 203d 2046 616c 7365 2023 6e6f 7420 7468   = False #not th
-0000dd50: 6520 6669 7273 7420 7477 6f20 726f 7773  e first two rows
-0000dd60: 0a20 2020 2020 2020 2020 2020 2023 2067  .            # g
-0000dd70: 6170 735f 6d61 736b 5b31 5d20 3d20 4661  aps_mask[1] = Fa
-0000dd80: 6c73 650a 2020 2020 2020 2020 2020 2020  lse.            
-0000dd90: 2320 6761 7073 5f6d 6173 6b5b 325d 203d  # gaps_mask[2] =
-0000dda0: 2046 616c 7365 0a20 2020 2020 2020 2020   False.         
-0000ddb0: 2020 2023 2044 6174 6173 6574 5b69 2c67     # Dataset[i,g
-0000ddc0: 6170 735f 6d61 736b 2c30 5d20 3d20 312e  aps_mask,0] = 1.
-0000ddd0: 2023 6966 206f 6e65 2068 6f74 2065 6e63   #if one hot enc
-0000dde0: 6f64 696e 6720 7468 6520 6761 7073 2077  oding the gaps w
-0000ddf0: 6974 6820 6265 2061 7373 6967 6e65 6420  ith be assigned 
-0000de00: 7468 6520 6669 7273 7420 706f 7369 7469  the first positi
-0000de10: 6f6e 2069 6e20 6f6e 6520 686f 7420 656e  on in one hot en
-0000de20: 636f 6469 6e67 0a20 2020 2020 2020 2020  coding.         
-0000de30: 2020 2044 6174 6173 6574 5f6e 6f74 5f61     Dataset_not_a
-0000de40: 6c69 676e 6564 5b69 2c20 2869 6e74 2843  ligned[i, (int(C
-0000de50: 6f6d 6269 6e65 645f 6469 6374 5b6b 6579  ombined_dict[key
-0000de60: 5d5b 305d 5b30 5d29 202b 2033 293a 2c30  ][0][0]) + 3):,0
-0000de70: 5d20 3d20 312e 0a20 2020 206e 702e 7361  ] = 1..    np.sa
-0000de80: 7665 2822 7b7d 2f7b 7d2f 7b7d 5f64 6174  ve("{}/{}/{}_dat
-0000de90: 6173 6574 5f6e 756d 7079 5f4e 4f54 5f61  aset_numpy_NOT_a
-0000dea0: 6c69 676e 6564 5f7b 7d2e 6e70 7922 2e66  ligned_{}.npy".f
-0000deb0: 6f72 6d61 7428 7374 6f72 6167 655f 666f  ormat(storage_fo
-0000dec0: 6c64 6572 2c6e 616d 655f 6669 6c65 2c6e  lder,name_file,n
-0000ded0: 616d 655f 6669 6c65 2c6f 6e65 5f68 6f74  ame_file,one_hot
-0000dee0: 5f6c 6162 656c 5b30 5d29 2c20 4461 7461  _label[0]), Data
-0000def0: 7365 745f 6e6f 745f 616c 6967 6e65 6429  set_not_aligned)
-0000df00: 0a0a 2020 2020 7265 7475 726e 2074 7265  ..    return tre
-0000df10: 655f 6669 6c65 0a0a 6465 6620 7379 6d6d  e_file..def symm
-0000df20: 6574 7269 7a65 5f61 6e64 5f63 6c65 616e  etrize_and_clean
-0000df30: 286d 6174 7269 782c 616e 6365 7374 7261  (matrix,ancestra
-0000df40: 6c3d 5472 7565 293a 0a20 2020 2022 2222  l=True):.    """
-0000df50: 5265 6d6f 7665 2c20 6966 206e 6563 6573  Remove, if neces
-0000df60: 7361 7279 2074 6865 2061 6e63 6573 7472  sary the ancestr
-0000df70: 616c 206e 6f64 6573 2066 726f 6d20 7468  al nodes from th
-0000df80: 6520 7472 6169 6e20 6d61 7472 6978 0a20  e train matrix. 
-0000df90: 2020 203a 7061 7261 6d20 7061 6e64 6173     :param pandas
-0000dfa0: 2d61 7272 6179 206d 6174 7269 783a 2070  -array matrix: p
-0000dfb0: 6174 7269 7374 6963 206f 7220 636c 6164  atristic or clad
-0000dfc0: 6973 7469 6320 6d61 7472 6963 6573 2077  istic matrices w
-0000dfd0: 6974 6820 636f 6c75 6d6e 7320 616e 6420  ith columns and 
-0000dfe0: 696e 6465 7865 7320 6173 206e 6f64 6520  indexes as node 
-0000dff0: 6e61 6d65 7320 2222 220a 2020 2020 6966  names """.    if
-0000e000: 206e 6f74 2061 6e63 6573 7472 616c 3a23   not ancestral:#
-0000e010: 4472 6f70 2074 6865 2061 6e63 6573 7472  Drop the ancestr
-0000e020: 616c 206e 6f64 6573 2069 6e66 6f72 6d61  al nodes informa
-0000e030: 7469 6f6e 0a20 2020 2020 2020 206d 6174  tion.        mat
-0000e040: 7269 7820 3d20 6d61 7472 6978 5b7e 6d61  rix = matrix[~ma
-0000e050: 7472 6978 2e69 6e64 6578 2e73 7472 2e63  trix.index.str.c
-0000e060: 6f6e 7461 696e 7328 275e 417b 317d 5b30  ontains('^A{1}[0
-0000e070: 2d39 5d2b 283f 215b 412d 5a5d 292b 2729  -9]+(?![A-Z])+')
-0000e080: 5d20 2023 2072 6570 6561 7420 7769 7468  ]  # repeat with
-0000e090: 205b 612d 7a5d 2069 6620 7072 6f62 6c65   [a-z] if proble
-0000e0a0: 6d73 0a20 2020 2020 2020 206d 6174 7269  ms.        matri
-0000e0b0: 7820 3d20 6d61 7472 6978 2e6c 6f63 5b3a  x = matrix.loc[:
-0000e0c0: 2c7e 6d61 7472 6978 2e63 6f6c 756d 6e73  ,~matrix.columns
-0000e0d0: 2e73 7472 2e63 6f6e 7461 696e 7328 275e  .str.contains('^
-0000e0e0: 417b 317d 5b30 2d39 5d2b 283f 215b 412d  A{1}[0-9]+(?![A-
-0000e0f0: 5a5d 292b 2729 5d0a 2020 2020 6d61 7472  Z])+')].    matr
-0000e100: 6978 203d 2073 796d 6d65 7472 697a 6528  ix = symmetrize(
-0000e110: 6d61 7472 6978 290a 2020 2020 7265 7475  matrix).    retu
-0000e120: 726e 206d 6174 7269 780a 6465 6620 7265  rn matrix.def re
-0000e130: 6e61 6d65 5f61 7869 7328 6d61 7472 6978  name_axis(matrix
-0000e140: 2c6e 6f64 6573 2c6e 616d 655f 6669 6c65  ,nodes,name_file
-0000e150: 203d 204e 6f6e 6529 3a0a 2020 2020 2222   = None):.    ""
-0000e160: 220a 2020 2020 5265 6e61 6d65 2074 6865  ".    Rename the
-0000e170: 206e 6f64 6573 206e 616d 6573 2074 6f20   nodes names to 
-0000e180: 7468 6569 7220 7472 6565 206c 6576 656c  their tree level
-0000e190: 2069 6e64 6578 0a20 2020 206e 6f64 6573   index.    nodes
-0000e1a0: 3a20 7472 6565 206c 6576 656c 206f 7264  : tree level ord
-0000e1b0: 6572 206e 6f64 6520 6e61 6d65 7322 2222  er node names"""
-0000e1c0: 0a20 2020 2069 6620 6c65 6e28 6e6f 6465  .    if len(node
-0000e1d0: 7329 2021 3d20 3020 616e 6420 6e61 6d65  s) != 0 and name
-0000e1e0: 5f66 696c 6520 6e6f 7420 696e 205b 2262  _file not in ["b
-0000e1f0: 656e 6368 6d61 726b 5f72 616e 6461 6c6c  enchmark_randall
-0000e200: 222c 2262 656e 6368 6d61 726b 5f72 616e  ","benchmark_ran
-0000e210: 6461 6c6c 5f6f 7269 6769 6e61 6c22 2c22  dall_original","
-0000e220: 6265 6e63 686d 6172 6b5f 7261 6e64 616c  benchmark_randal
-0000e230: 6c5f 6f72 6967 696e 616c 5f6e 616d 696e  l_original_namin
-0000e240: 6722 5d3a 2023 616e 6420 6e6f 7420 6e61  g"]: #and not na
-0000e250: 6d65 2e73 7461 7274 7377 6974 6828 2273  me.startswith("s
-0000e260: 696d 756c 6174 696f 6e73 2229 3a20 2375  imulations"): #u
-0000e270: 7365 2074 6865 206c 6576 656c 206f 7264  se the level ord
-0000e280: 6572 2074 7261 6e73 7665 7273 616c 2074  er transversal t
-0000e290: 7265 6520 696e 666f 726d 6174 696f 6e0a  ree information.
-0000e2a0: 2020 2020 2020 2020 2348 6967 686c 6967          #Highlig
-0000e2b0: 6874 3a20 4966 206e 616e 2028 7064 2e69  ht: If nan (pd.i
-0000e2c0: 736e 756c 6c29 2069 7320 666f 756e 642c  snull) is found,
-0000e2d0: 2069 7320 6265 6361 7573 6520 7468 6520   is because the 
-0000e2e0: 726f 6f74 206e 616d 6520 6973 206d 6573  root name is mes
-0000e2f0: 7365 6420 7570 2061 6e64 206d 6973 7369  sed up and missi
-0000e300: 6e67 2c20 6a75 7374 2077 7269 7465 2069  ng, just write i
-0000e310: 7420 646f 776e 2069 6e20 7468 6520 7061  t down in the pa
-0000e320: 7472 6973 7469 6320 6d61 7472 6978 2066  tristic matrix f
-0000e330: 696c 6521 2121 2120 4f72 2074 6865 206e  ile!!!! Or the n
-0000e340: 616d 6573 2069 6e20 7468 6520 6d61 7472  ames in the matr
-0000e350: 6978 2021 3d20 6e61 6d65 7320 696e 2074  ix != names in t
-0000e360: 7265 7665 6c20 6f72 6465 720a 0a20 2020  revel order..   
-0000e370: 2020 2020 206d 6174 7269 782e 696e 6465       matrix.inde
-0000e380: 7820 3d20 5b6e 702e 7768 6572 6528 6e6f  x = [np.where(no
-0000e390: 6465 7320 3d3d 206e 6f64 655f 6e61 6d65  des == node_name
-0000e3a0: 295b 305d 5b30 5d20 666f 7220 6e6f 6465  )[0][0] for node
-0000e3b0: 5f6e 616d 6520 696e 206d 6174 7269 782e  _name in matrix.
-0000e3c0: 696e 6465 785d 2023 544f 444f 3b20 7768  index] #TODO; wh
-0000e3d0: 7920 616d 2049 2064 6f69 6e67 2074 6869  y am I doing thi
-0000e3e0: 7320 7477 6963 653f 0a20 2020 2020 2020  s twice?.       
-0000e3f0: 206d 6174 7269 782e 636f 6c75 6d6e 7320   matrix.columns 
-0000e400: 3d20 5b6e 702e 7768 6572 6528 6e6f 6465  = [np.where(node
-0000e410: 7320 3d3d 206e 6f64 655f 6e61 6d65 295b  s == node_name)[
-0000e420: 305d 5b30 5d20 666f 7220 6e6f 6465 5f6e  0][0] for node_n
-0000e430: 616d 6520 696e 206d 6174 7269 782e 636f  ame in matrix.co
-0000e440: 6c75 6d6e 735d 0a20 2020 2020 2020 2072  lumns].        r
-0000e450: 6574 7572 6e20 6d61 7472 6978 0a20 2020  eturn matrix.   
-0000e460: 2065 6c69 6620 206e 616d 655f 6669 6c65   elif  name_file
-0000e470: 2069 6e20 5b22 6265 6e63 686d 6172 6b5f   in ["benchmark_
-0000e480: 7261 6e64 616c 6c22 2c22 6265 6e63 686d  randall","benchm
-0000e490: 6172 6b5f 7261 6e64 616c 6c5f 6f72 6967  ark_randall_orig
-0000e4a0: 696e 616c 222c 2262 656e 6368 6d61 726b  inal","benchmark
-0000e4b0: 5f72 616e 6461 6c6c 5f6f 7269 6769 6e61  _randall_origina
-0000e4c0: 6c5f 6e61 6d69 6e67 225d 203a 0a20 2020  l_naming"] :.   
-0000e4d0: 2020 2020 206d 6174 7269 782e 696e 6465       matrix.inde
-0000e4e0: 7820 3d20 6d61 7472 6978 2e69 6e64 6578  x = matrix.index
-0000e4f0: 2e73 7472 2e72 6570 6c61 6365 2822 4122  .str.replace("A"
-0000e500: 2c22 2229 2e61 7374 7970 6528 2269 6e74  ,"").astype("int
-0000e510: 2229 0a20 2020 2020 2020 206d 6174 7269  ").        matri
-0000e520: 782e 636f 6c75 6d6e 7320 3d20 6d61 7472  x.columns = matr
-0000e530: 6978 2e63 6f6c 756d 6e73 2e73 7472 2e72  ix.columns.str.r
-0000e540: 6570 6c61 6365 2822 4122 2c20 2222 292e  eplace("A", "").
-0000e550: 6173 7479 7065 2822 696e 7422 290a 2020  astype("int").  
-0000e560: 2020 2020 2020 7265 7475 726e 206d 6174        return mat
-0000e570: 7269 780a 0a20 2020 2065 6c73 653a 0a20  rix..    else:. 
-0000e580: 2020 2020 2020 2072 6574 7572 6e20 6d61         return ma
-0000e590: 7472 6978 0a64 6566 2073 756d 5f6d 6174  trix.def sum_mat
-0000e5a0: 7269 6365 7328 6d61 7472 6978 312c 6d61  rices(matrix1,ma
-0000e5b0: 7472 6978 3229 3a0a 2020 2020 2222 2253  trix2):.    """S
-0000e5c0: 756d 7320 636c 6164 6973 7469 6320 616e  ums cladistic an
-0000e5d0: 6420 7061 7472 6973 7469 6320 6d61 7472  d patristic matr
-0000e5e0: 6963 6573 2074 6f20 666f 726d 2061 6e20  ices to form an 
-0000e5f0: 6576 6f6c 7574 696f 6e61 7279 206d 6174  evolutionary mat
-0000e600: 7269 780a 2020 2020 3a70 6172 616d 3a20  rix.    :param: 
-0000e610: 7061 6e64 6173 2d61 7272 6179 206d 6174  pandas-array mat
-0000e620: 7269 7831 0a20 2020 203a 7061 7261 6d3a  rix1.    :param:
-0000e630: 2070 616e 6461 732d 6172 7261 7920 6d61   pandas-array ma
-0000e640: 7472 6978 320a 2020 2020 2222 220a 2020  trix2.    """.  
-0000e650: 2020 636f 6c75 6d6e 5f6e 616d 6573 203d    column_names =
-0000e660: 206d 6174 7269 7831 2e63 6f6c 756d 6e73   matrix1.columns
-0000e670: 2e76 616c 7565 7320 2023 2038 3220 2b20  .values  # 82 + 
-0000e680: 310a 2020 2020 636f 6c75 6d6e 5f6e 616d  1.    column_nam
-0000e690: 6573 203d 206e 702e 636f 6e63 6174 656e  es = np.concaten
-0000e6a0: 6174 6528 286e 702e 6172 7261 7928 5b66  ate((np.array([f
-0000e6b0: 6c6f 6174 2822 2d69 6e66 2229 5d29 2c20  loat("-inf")]), 
-0000e6c0: 636f 6c75 6d6e 5f6e 616d 6573 2929 2020  column_names))  
-0000e6d0: 2320 3832 202b 2031 2028 746f 2066 6974  # 82 + 1 (to fit
-0000e6e0: 2074 6865 206e 756d 7079 2061 7272 6179   the numpy array
-0000e6f0: 0a20 2020 2072 6f77 735f 6e61 6d65 7320  .    rows_names 
-0000e700: 3d20 6d61 7472 6978 312e 696e 6465 782e  = matrix1.index.
-0000e710: 7661 6c75 6573 2020 2320 3832 0a20 2020  values  # 82.   
-0000e720: 206d 6174 7269 7831 203d 206d 6174 7269   matrix1 = matri
-0000e730: 7831 2e74 6f5f 6e75 6d70 7928 290a 2020  x1.to_numpy().  
-0000e740: 2020 6d61 7472 6978 3220 3d20 6d61 7472    matrix2 = matr
-0000e750: 6978 322e 746f 5f6e 756d 7079 2829 0a20  ix2.to_numpy(). 
-0000e760: 2020 206d 6174 7269 7820 3d20 6d61 7472     matrix = matr
-0000e770: 6978 3120 2b20 6d61 7472 6978 320a 2020  ix1 + matrix2.  
-0000e780: 2020 6d61 7472 6978 203d 206e 6f72 6d61    matrix = norma
-0000e790: 6c69 7a65 5f73 7461 6e64 6172 697a 6528  lize_standarize(
-0000e7a0: 6d61 7472 6978 290a 2020 2020 6d61 7472  matrix).    matr
-0000e7b0: 6978 203d 206e 702e 636f 6c75 6d6e 5f73  ix = np.column_s
-0000e7c0: 7461 636b 2828 726f 7773 5f6e 616d 6573  tack((rows_names
-0000e7d0: 2c20 6d61 7472 6978 2929 0a20 2020 206d  , matrix)).    m
-0000e7e0: 6174 7269 7820 3d20 6e70 2e72 6f77 5f73  atrix = np.row_s
-0000e7f0: 7461 636b 2828 636f 6c75 6d6e 5f6e 616d  tack((column_nam
-0000e800: 6573 2c20 6d61 7472 6978 2929 0a20 2020  es, matrix)).   
-0000e810: 2072 6574 7572 6e20 6d61 7472 6978 0a64   return matrix.d
-0000e820: 6566 2070 616e 6461 735f 746f 5f6e 756d  ef pandas_to_num
-0000e830: 7079 286d 6174 7269 7829 3a0a 2020 2020  py(matrix):.    
-0000e840: 2222 2243 6f6e 7665 7274 7320 7061 6e64  """Converts pand
-0000e850: 6173 2061 7272 6179 2074 6f20 6e75 6d70  as array to nump
-0000e860: 7920 6172 7261 792c 2069 6e20 7468 6973  y array, in this
-0000e870: 2063 6173 6520 6279 2074 7261 6e73 666f   case by transfo
-0000e880: 726d 696e 6720 7468 6520 6e6f 6465 7320  rming the nodes 
-0000e890: 6e61 6d65 730a 2020 2020 3a70 6172 616d  names.    :param
-0000e8a0: 206d 6174 7269 7822 2222 0a20 2020 2063   matrix""".    c
-0000e8b0: 6f6c 756d 6e5f 6e61 6d65 7320 3d20 6d61  olumn_names = ma
-0000e8c0: 7472 6978 2e63 6f6c 756d 6e73 2e76 616c  trix.columns.val
-0000e8d0: 7565 732e 6173 7479 7065 2822 696e 7422  ues.astype("int"
-0000e8e0: 2920 2023 2038 3220 2b20 310a 2020 2020  )  # 82 + 1.    
-0000e8f0: 636f 6c75 6d6e 5f6e 616d 6573 203d 206e  column_names = n
-0000e900: 702e 636f 6e63 6174 656e 6174 6528 286e  p.concatenate((n
-0000e910: 702e 6172 7261 7928 5b66 6c6f 6174 2822  p.array([float("
-0000e920: 2d69 6e66 2229 5d29 2c20 636f 6c75 6d6e  -inf")]), column
-0000e930: 5f6e 616d 6573 2929 2020 2320 3832 202b  _names))  # 82 +
-0000e940: 2031 2028 746f 2066 6974 2074 6865 206e   1 (to fit the n
-0000e950: 756d 7079 2061 7272 6179 0a20 2020 2072  umpy array.    r
-0000e960: 6f77 735f 6e61 6d65 7320 3d20 6d61 7472  ows_names = matr
-0000e970: 6978 2e69 6e64 6578 2e76 616c 7565 732e  ix.index.values.
-0000e980: 6173 7479 7065 2822 696e 7422 2920 2023  astype("int")  #
-0000e990: 2038 320a 2020 2020 6d61 7472 6978 203d   82.    matrix =
-0000e9a0: 206e 702e 636f 6c75 6d6e 5f73 7461 636b   np.column_stack
-0000e9b0: 2828 726f 7773 5f6e 616d 6573 2c20 6d61  ((rows_names, ma
-0000e9c0: 7472 6978 2929 0a20 2020 206d 6174 7269  trix)).    matri
-0000e9d0: 7820 3d20 6e70 2e72 6f77 5f73 7461 636b  x = np.row_stack
-0000e9e0: 2828 636f 6c75 6d6e 5f6e 616d 6573 2c20  ((column_names, 
-0000e9f0: 6d61 7472 6978 2929 0a20 2020 2072 6574  matrix)).    ret
-0000ea00: 7572 6e20 6d61 7472 6978 0a64 6566 2063  urn matrix.def c
-0000ea10: 6f6e 7665 7274 5f74 6f5f 6c65 7474 6572  onvert_to_letter
-0000ea20: 7328 7365 712c 6161 5f70 726f 6273 293a  s(seq,aa_probs):
-0000ea30: 0a20 2020 2022 2222 436f 6e76 6572 7473  .    """Converts
-0000ea40: 2062 6163 6b20 7468 6520 696e 7465 6765   back the intege
-0000ea50: 7273 2061 7373 6967 6e65 6420 746f 2074  rs assigned to t
-0000ea60: 6865 2061 6d69 6e6f 2061 6369 6473 2074  he amino acids t
-0000ea70: 6f20 6c65 7474 6572 730a 2020 2020 3a70  o letters.    :p
-0000ea80: 6172 616d 2073 6571 3a20 7365 710a 2020  aram seq: seq.  
-0000ea90: 2020 3a70 6172 616d 2069 6e74 2061 615f    :param int aa_
-0000eaa0: 7072 6f62 733a 2061 6d69 6e6f 2061 6369  probs: amino aci
-0000eab0: 6420 7072 6f62 6162 696c 6974 6965 7322  d probabilities"
-0000eac0: 2222 0a0a 2020 2020 6161 5f6e 616d 6573  ""..    aa_names
-0000ead0: 5f64 6963 7420 3d20 616d 696e 6f61 6369  _dict = aminoaci
-0000eae0: 645f 6e61 6d65 735f 6469 6374 2861 615f  d_names_dict(aa_
-0000eaf0: 7072 6f62 7329 0a20 2020 2061 615f 6e61  probs).    aa_na
-0000eb00: 6d65 735f 6469 6374 5f72 6576 6572 7365  mes_dict_reverse
-0000eb10: 203d 207b 7661 6c3a 6b65 7920 666f 7220   = {val:key for 
-0000eb20: 6b65 792c 7661 6c20 696e 2061 615f 6e61  key,val in aa_na
-0000eb30: 6d65 735f 6469 6374 2e69 7465 6d73 2829  mes_dict.items()
-0000eb40: 7d0a 2020 2020 6966 206e 6f74 2069 7369  }.    if not isi
-0000eb50: 6e73 7461 6e63 6528 7365 715b 305d 2c66  nstance(seq[0],f
-0000eb60: 6c6f 6174 293a 0a20 2020 2020 2020 2073  loat):.        s
-0000eb70: 6571 5f6c 6574 7465 7273 203d 205b 6161  eq_letters = [aa
-0000eb80: 5f6e 616d 6573 5f64 6963 745f 7265 7665  _names_dict_reve
-0000eb90: 7273 655b 706f 7369 7469 6f6e 2e69 7465  rse[position.ite
-0000eba0: 6d28 295d 2066 6f72 2070 6f73 6974 696f  m()] for positio
-0000ebb0: 6e20 696e 2073 6571 2069 6620 706f 7369  n in seq if posi
-0000ebc0: 7469 6f6e 2e69 7465 6d28 2920 696e 2061  tion.item() in a
-0000ebd0: 615f 6e61 6d65 735f 6469 6374 5f72 6576  a_names_dict_rev
-0000ebe0: 6572 7365 5d0a 0a20 2020 2065 6c73 653a  erse]..    else:
-0000ebf0: 0a20 2020 2020 2020 2073 6571 5f6c 6574  .        seq_let
-0000ec00: 7465 7273 203d 205b 6161 5f6e 616d 6573  ters = [aa_names
-0000ec10: 5f64 6963 745f 7265 7665 7273 655b 706f  _dict_reverse[po
-0000ec20: 7369 7469 6f6e 5d20 666f 7220 706f 7369  sition] for posi
-0000ec30: 7469 6f6e 2069 6e20 7365 7120 6966 2070  tion in seq if p
-0000ec40: 6f73 6974 696f 6e20 696e 2061 615f 6e61  osition in aa_na
-0000ec50: 6d65 735f 6469 6374 5f72 6576 6572 7365  mes_dict_reverse
-0000ec60: 5d0a 0a20 2020 2072 6574 7572 6e20 2727  ]..    return ''
-0000ec70: 2e6a 6f69 6e28 7365 715f 6c65 7474 6572  .join(seq_letter
-0000ec80: 7329 0a64 6566 2073 636f 7265 5f6d 6174  s).def score_mat
-0000ec90: 6368 2870 6169 722c 206d 6174 7269 7829  ch(pair, matrix)
-0000eca0: 3a0a 2020 2020 2222 2252 6574 7572 6e73  :.    """Returns
-0000ecb0: 2074 6865 2063 6f72 7265 7370 6f6e 6469   the correspondi
-0000ecc0: 6e67 2062 6c6f 7375 6d20 7363 6f72 6573  ng blosum scores
-0000ecd0: 2062 6574 7765 656e 2074 6865 2070 6169   between the pai
-0000ece0: 7220 6f66 2061 6d69 6e6f 2061 6369 6473  r of amino acids
-0000ecf0: 0a20 2020 203a 7061 7261 6d20 7475 706c  .    :param tupl
-0000ed00: 6520 7061 6972 3a20 7061 6972 206f 6620  e pair: pair of 
-0000ed10: 616d 696e 6f20 6163 6964 7320 746f 2063  amino acids to c
-0000ed20: 6f6d 7061 7265 0a20 2020 203a 7061 7261  ompare.    :para
-0000ed30: 6d20 6d61 7472 6978 3a20 426c 6f73 756d  m matrix: Blosum
-0000ed40: 206d 6174 7269 7820 2222 220a 2020 2020   matrix """.    
-0000ed50: 6966 2070 6169 7220 6e6f 7420 696e 206d  if pair not in m
-0000ed60: 6174 7269 783a 0a20 2020 2020 2020 2072  atrix:.        r
-0000ed70: 6574 7572 6e20 6d61 7472 6978 5b28 7475  eturn matrix[(tu
-0000ed80: 706c 6528 7265 7665 7273 6564 2870 6169  ple(reversed(pai
-0000ed90: 7229 2929 5d0a 2020 2020 656c 7365 3a0a  r)))].    else:.
-0000eda0: 2020 2020 2020 2020 7265 7475 726e 206d          return m
-0000edb0: 6174 7269 785b 7061 6972 5d0a 6465 6620  atrix[pair].def 
-0000edc0: 7363 6f72 655f 7061 6972 7769 7365 2873  score_pairwise(s
-0000edd0: 6571 312c 2073 6571 322c 206d 6174 7269  eq1, seq2, matri
-0000ede0: 782c 2067 6170 5f73 2c20 6761 705f 6529  x, gap_s, gap_e)
-0000edf0: 3a0a 2020 2020 2354 4f44 4f3a 2068 7474  :.    #TODO: htt
-0000ee00: 7073 3a2f 2f73 7461 636b 6f76 6572 666c  ps://stackoverfl
-0000ee10: 6f77 2e63 6f6d 2f71 7565 7374 696f 6e73  ow.com/questions
-0000ee20: 2f35 3638 3632 3131 2f69 732d 7468 6572  /5686211/is-ther
-0000ee30: 652d 612d 6675 6e63 7469 6f6e 2d74 6861  e-a-function-tha
-0000ee40: 742d 6361 6e2d 6361 6c63 756c 6174 652d  t-can-calculate-
-0000ee50: 612d 7363 6f72 652d 666f 722d 616c 6967  a-score-for-alig
-0000ee60: 6e65 642d 7365 7175 656e 6365 732d 6769  ned-sequences-gi
-0000ee70: 7665 6e2d 7468 652d 610a 2020 2020 2222  ven-the-a.    ""
-0000ee80: 220a 2020 2020 4361 6c63 756c 6174 6573  ".    Calculates
-0000ee90: 2074 6865 2062 6c6f 7375 6d20 7363 6f72   the blosum scor
-0000eea0: 6520 6f66 2074 6865 2074 7275 6520 7365  e of the true se
-0000eeb0: 7175 656e 6365 2061 6761 696e 7374 2074  quence against t
-0000eec0: 6865 2070 7265 6469 6374 696f 6e73 0a20  he predictions. 
-0000eed0: 2020 203a 7061 7261 6d20 6d61 7472 6978     :param matrix
-0000eee0: 206d 6174 7269 783a 2042 6c6f 7375 6d20   matrix: Blosum 
-0000eef0: 6d61 7472 6978 2063 6f6e 7461 696e 696e  matrix containin
-0000ef00: 6720 7468 6520 6c6f 672d 6f64 6473 2073  g the log-odds s
-0000ef10: 636f 7265 7320 2868 6520 6c6f 6761 7269  cores (he logari
-0000ef20: 7468 6d20 666f 7220 7468 6520 7261 7469  thm for the rati
-0000ef30: 6f20 6f66 2074 6865 0a20 2020 2020 2020  o of the.       
-0000ef40: 6c69 6b65 6c69 686f 6f64 206f 6620 7477  likelihood of tw
-0000ef50: 6f20 616d 696e 6f20 6163 6964 7320 6170  o amino acids ap
-0000ef60: 7065 6172 696e 6720 7769 7468 2061 2062  pearing with a b
-0000ef70: 696f 6c6f 6769 6361 6c20 7365 6e73 6520  iological sense 
-0000ef80: 616e 6420 7468 6520 6c69 6b65 6c69 686f  and the likeliho
-0000ef90: 6f64 206f 6620 7468 6520 7361 6d65 2061  od of the same a
-0000efa0: 6d69 6e6f 2061 6369 6473 2061 7070 6561  mino acids appea
-0000efb0: 7269 6e67 2062 7920 6368 616e 6365 290a  ring by chance).
-0000efc0: 2020 2020 2020 2020 2874 6865 2068 6967          (the hig
-0000efd0: 6865 7220 7468 6520 7363 6f72 652c 2074  her the score, t
-0000efe0: 6865 206d 6f72 6520 6c69 6b65 6c79 2074  he more likely t
-0000eff0: 6865 2063 6f72 7265 7370 6f6e 6469 6e67  he corresponding
-0000f000: 2061 6d69 6e6f 2d61 6369 6420 7375 6273   amino-acid subs
-0000f010: 7469 7475 7469 6f6e 2069 7329 0a20 2020  titution is).   
-0000f020: 203a 7061 7261 6d20 696e 7420 6761 705f   :param int gap_
-0000f030: 733a 2067 6170 2070 656e 616c 7479 0a20  s: gap penalty. 
-0000f040: 2020 203a 7061 7261 6d20 696e 7420 6761     :param int ga
-0000f050: 705f 6520 3a20 6d69 736d 6174 6368 2070  p_e : mismatch p
-0000f060: 656e 616c 7479 0a20 2020 2022 2222 0a20  enalty.    """. 
-0000f070: 2020 2073 636f 7265 203d 2030 0a20 2020     score = 0.   
-0000f080: 2067 6170 203d 2046 616c 7365 0a20 2020   gap = False.   
-0000f090: 2066 6f72 2069 2069 6e20 7261 6e67 6528   for i in range(
-0000f0a0: 6c65 6e28 7365 7131 2929 3a0a 2020 2020  len(seq1)):.    
-0000f0b0: 2020 2020 7061 6972 203d 2028 7365 7131      pair = (seq1
-0000f0c0: 5b69 5d2c 2073 6571 325b 695d 290a 2020  [i], seq2[i]).  
-0000f0d0: 2020 2020 2020 6966 206e 6f74 2067 6170        if not gap
-0000f0e0: 3a0a 2020 2020 2020 2020 2020 2020 6966  :.            if
-0000f0f0: 2027 2d27 2069 6e20 7061 6972 3a0a 2020   '-' in pair:.  
-0000f100: 2020 2020 2020 2020 2020 2020 2020 6761                ga
-0000f110: 7020 3d20 5472 7565 0a20 2020 2020 2020  p = True.       
-0000f120: 2020 2020 2020 2020 2073 636f 7265 202b           score +
-0000f130: 3d20 6761 705f 730a 2020 2020 2020 2020  = gap_s.        
-0000f140: 2020 2020 2320 656c 6966 2022 2a22 2069      # elif "*" i
-0000f150: 6e20 7061 6972 3a20 2354 4f44 4f3a 204b  n pair: #TODO: K
-0000f160: 6565 703f 0a20 2020 2020 2020 2020 2020  eep?.           
-0000f170: 2023 2020 2020 2073 636f 7265 202b 3d67   #     score +=g
-0000f180: 6170 5f65 0a20 2020 2020 2020 2020 2020  ap_e.           
-0000f190: 2065 6c73 653a 0a20 2020 2020 2020 2020   else:.         
-0000f1a0: 2020 2020 2020 2073 636f 7265 202b 3d20         score += 
-0000f1b0: 7363 6f72 655f 6d61 7463 6828 7061 6972  score_match(pair
-0000f1c0: 2c20 6d61 7472 6978 290a 2020 2020 2020  , matrix).      
-0000f1d0: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
-0000f1e0: 2020 2020 6966 2027 2d27 206e 6f74 2069      if '-' not i
-0000f1f0: 6e20 7061 6972 3a23 2061 6e64 2022 2a22  n pair:# and "*"
-0000f200: 206e 6f74 2069 6e20 7061 6972 3a0a 2020   not in pair:.  
-0000f210: 2020 2020 2020 2020 2020 2020 2020 6761                ga
-0000f220: 7020 3d20 4661 6c73 650a 2020 2020 2020  p = False.      
-0000f230: 2020 2020 2020 2020 2020 7363 6f72 6520            score 
-0000f240: 2b3d 2073 636f 7265 5f6d 6174 6368 2870  += score_match(p
-0000f250: 6169 722c 206d 6174 7269 7829 0a20 2020  air, matrix).   
-0000f260: 2020 2020 2020 2020 2065 6c73 653a 0a20           else:. 
-0000f270: 2020 2020 2020 2020 2020 2020 2020 2073                 s
-0000f280: 636f 7265 202b 3d20 6761 705f 650a 2020  core += gap_e.  
-0000f290: 2020 7265 7475 726e 2073 636f 7265 0a64    return score.d
-0000f2a0: 6566 2073 636f 7265 5f70 6169 7277 6973  ef score_pairwis
-0000f2b0: 655f 3228 7365 7131 2c20 7365 7132 2c20  e_2(seq1, seq2, 
-0000f2c0: 6d61 7472 6978 2c20 6761 705f 732c 2067  matrix, gap_s, g
-0000f2d0: 6170 5f65 2c20 6761 7020 3d20 5472 7565  ap_e, gap = True
-0000f2e0: 293a 0a20 2020 2022 2222 416c 7465 726e  ):.    """Altern
-0000f2f0: 6174 6976 6520 6d65 7468 6f64 2074 6f20  ative method to 
-0000f300: 6361 6c63 756c 6174 6520 7468 6520 626c  calculate the bl
-0000f310: 6f73 756d 2073 636f 7265 2062 6574 7765  osum score betwe
-0000f320: 656e 2032 2073 6571 7565 6e63 6573 2222  en 2 sequences""
-0000f330: 220a 2020 2020 666f 7220 412c 4220 696e  ".    for A,B in
-0000f340: 207a 6970 2873 6571 312c 2073 6571 3229   zip(seq1, seq2)
-0000f350: 3a0a 2020 2020 2020 2020 6469 6167 203d  :.        diag =
-0000f360: 2028 272d 273d 3d41 2920 6f72 2028 272d   ('-'==A) or ('-
-0000f370: 273d 3d42 290a 2020 2020 2020 2020 7969  '==B).        yi
-0000f380: 656c 6420 2867 6170 5f65 2069 6620 6761  eld (gap_e if ga
-0000f390: 7020 656c 7365 2067 6170 5f73 2920 6966  p else gap_s) if
-0000f3a0: 2064 6961 6720 656c 7365 206d 6174 7269   diag else matri
-0000f3b0: 785b 2841 2c42 295d 0a20 2020 2020 2020  x[(A,B)].       
-0000f3c0: 2067 6170 203d 2064 6961 670a 6465 6620   gap = diag.def 
-0000f3d0: 6e6f 726d 616c 697a 655f 7374 616e 6461  normalize_standa
-0000f3e0: 7269 7a65 2878 293a 0a20 2020 2022 4e6f  rize(x):.    "No
-0000f3f0: 726d 616c 697a 6573 2061 6e64 2073 7461  rmalizes and sta
-0000f400: 6e64 6172 697a 6573 2061 6e20 696e 7075  ndarizes an inpu
-0000f410: 7420 6d61 7472 6978 220a 2020 2020 6e6f  t matrix".    no
-0000f420: 726d 203d 206e 702e 6c69 6e61 6c67 2e6e  rm = np.linalg.n
-0000f430: 6f72 6d28 7829 0a20 2020 206e 6f72 6d61  orm(x).    norma
-0000f440: 6c5f 6172 7261 7920 3d20 7820 2f20 6e6f  l_array = x / no
-0000f450: 726d 0a20 2020 2072 6574 7572 6e20 6e6f  rm.    return no
-0000f460: 726d 616c 5f61 7272 6179 0a64 6566 2066  rmal_array.def f
-0000f470: 6f6c 6465 7273 2866 6f6c 6465 725f 6e61  olders(folder_na
-0000f480: 6d65 2c62 6173 6570 6174 6829 3a0a 2020  me,basepath):.  
-0000f490: 2020 2222 2220 4372 6561 7465 7320 6120    """ Creates a 
-0000f4a0: 666f 6c64 6572 2061 7420 7468 6520 696e  folder at the in
-0000f4b0: 6469 6361 7465 6420 6c6f 6361 7469 6f6e  dicated location
-0000f4c0: 2e20 4974 2072 6577 7269 7465 7320 666f  . It rewrites fo
-0000f4d0: 6c64 6572 7320 7769 7468 2074 6865 2073  lders with the s
-0000f4e0: 616d 6520 6e61 6d65 0a20 2020 203a 7061  ame name.    :pa
-0000f4f0: 7261 6d20 7374 7220 666f 6c64 6572 5f6e  ram str folder_n
-0000f500: 616d 653a 206e 616d 6520 6f66 2074 6865  ame: name of the
-0000f510: 2066 6f6c 6465 720a 2020 2020 3a70 6172   folder.    :par
-0000f520: 616d 2073 7472 2062 6173 6570 6174 683a  am str basepath:
-0000f530: 2069 6e64 6963 6174 6573 2074 6865 2070   indicates the p
-0000f540: 6c61 6365 2077 6865 7265 2074 6f20 6372  lace where to cr
-0000f550: 6561 7465 2074 6865 2066 6f6c 6465 720a  eate the folder.
-0000f560: 2020 2020 2222 220a 2020 2020 2362 6173      """.    #bas
-0000f570: 6570 6174 6820 3d20 6f73 2e67 6574 6377  epath = os.getcw
-0000f580: 6428 290a 0a20 2020 2069 6620 6e6f 7420  d()..    if not 
-0000f590: 6261 7365 7061 7468 3a0a 2020 2020 2020  basepath:.      
-0000f5a0: 2020 6e65 7770 6174 6820 3d20 666f 6c64    newpath = fold
-0000f5b0: 6572 5f6e 616d 650a 2020 2020 656c 7365  er_name.    else
-0000f5c0: 3a0a 2020 2020 2020 2020 6e65 7770 6174  :.        newpat
-0000f5d0: 6820 3d20 6261 7365 7061 7468 202b 2022  h = basepath + "
-0000f5e0: 2f25 7322 2025 2066 6f6c 6465 725f 6e61  /%s" % folder_na
-0000f5f0: 6d65 0a0a 2020 2020 6966 206e 6f74 206f  me..    if not o
-0000f600: 732e 7061 7468 2e65 7869 7374 7328 6e65  s.path.exists(ne
-0000f610: 7770 6174 6829 3a0a 0a20 2020 2020 2020  wpath):..       
-0000f620: 2074 7279 3a0a 2020 2020 2020 2020 2020   try:.          
-0000f630: 2020 6f72 6967 696e 616c 5f75 6d61 736b    original_umask
-0000f640: 203d 206f 732e 756d 6173 6b28 3029 0a20   = os.umask(0). 
-0000f650: 2020 2020 2020 2020 2020 206f 732e 6d61             os.ma
-0000f660: 6b65 6469 7273 286e 6577 7061 7468 2c20  kedirs(newpath, 
-0000f670: 306f 3737 3729 0a20 2020 2020 2020 2066  0o777).        f
-0000f680: 696e 616c 6c79 3a0a 2020 2020 2020 2020  inally:.        
-0000f690: 2020 2020 6f73 2e75 6d61 736b 286f 7269      os.umask(ori
-0000f6a0: 6769 6e61 6c5f 756d 6173 6b29 0a20 2020  ginal_umask).   
-0000f6b0: 2065 6c73 653a 0a20 2020 2020 2020 2073   else:.        s
-0000f6c0: 6875 7469 6c2e 726d 7472 6565 286e 6577  hutil.rmtree(new
-0000f6d0: 7061 7468 2920 2023 2072 656d 6f76 6573  path)  # removes
-0000f6e0: 2061 6c6c 2074 6865 2073 7562 6469 7265   all the subdire
-0000f6f0: 6374 6f72 6965 7321 0a20 2020 2020 2020  ctories!.       
-0000f700: 206f 732e 6d61 6b65 6469 7273 286e 6577   os.makedirs(new
-0000f710: 7061 7468 2c30 6f37 3737 290a 6465 6620  path,0o777).def 
-0000f720: 6469 7669 6465 5f62 6174 6368 6573 2844  divide_batches(D
-0000f730: 6174 6173 6574 2c6e 756d 6265 725f 7370  ataset,number_sp
-0000f740: 6c69 7473 293a 0a20 2020 2022 2222 2044  lits):.    """ D
-0000f750: 6976 6964 6573 2074 6865 2074 7261 696e  ivides the train
-0000f760: 696e 6720 6461 7461 7365 7420 696e 2061  ing dataset in a
-0000f770: 2067 6976 656e 206e 756d 6265 7220 6f66   given number of
-0000f780: 2073 706c 6974 7322 2222 0a20 2020 2044   splits""".    D
-0000f790: 6174 6173 6574 203d 2044 6174 6173 6574  ataset = Dataset
-0000f7a0: 5b44 6174 6173 6574 5b3a 2c30 2c30 5d2e  [Dataset[:,0,0].
-0000f7b0: 6172 6773 6f72 7428 295d 0a20 2020 2044  argsort()].    D
-0000f7c0: 6174 6173 6574 5f73 706c 6974 7320 3d6e  ataset_splits =n
-0000f7d0: 702e 6172 7261 795f 7370 6c69 7428 4461  p.array_split(Da
-0000f7e0: 7461 7365 742c 6e75 6d62 6572 5f73 706c  taset,number_spl
-0000f7f0: 6974 7329 0a20 2020 2072 6574 7572 6e20  its).    return 
-0000f800: 4461 7461 7365 745f 7370 6c69 7473 0a64  Dataset_splits.d
-0000f810: 6566 2070 6572 635f 6964 656e 7469 7479  ef perc_identity
-0000f820: 5f61 6c69 676e 6d65 6e74 2861 6c6e 293a  _alignment(aln):
-0000f830: 0a20 2020 2022 2222 4361 6c63 756c 6174  .    """Calculat
-0000f840: 6573 2074 6865 2025 4944 206f 6620 616e  es the %ID of an
-0000f850: 2061 6c69 676e 6d65 6e74 0a20 2020 203a   alignment.    :
-0000f860: 7061 7261 6d20 616c 6e3a 2062 696f 7079  param aln: biopy
-0000f870: 7468 6f6e 2061 6c69 676e 6d65 6e74 206f  thon alignment o
-0000f880: 626a 6563 7422 2222 0a20 2020 2069 203d  bject""".    i =
-0000f890: 2030 0a20 2020 2066 6f72 2061 2069 6e20   0.    for a in 
-0000f8a0: 7261 6e67 6528 302c 6c65 6e28 616c 6e5b  range(0,len(aln[
-0000f8b0: 305d 2929 3a0a 2020 2020 2020 2020 7320  0])):.        s 
-0000f8c0: 3d20 616c 6e5b 3a2c 615d 0a20 2020 2020  = aln[:,a].     
-0000f8d0: 2020 2069 6620 7320 3d3d 206c 656e 2873     if s == len(s
-0000f8e0: 2920 2a20 735b 305d 3a0a 2020 2020 2020  ) * s[0]:.      
-0000f8f0: 2020 2020 2020 6920 2b3d 2031 0a20 2020        i += 1.   
-0000f900: 2072 6574 7572 6e20 3130 302a 692f 666c   return 100*i/fl
-0000f910: 6f61 7428 6c65 6e28 616c 6e5b 305d 2929  oat(len(aln[0]))
-0000f920: 0a64 6566 2070 6572 635f 6964 656e 7469  .def perc_identi
-0000f930: 7479 5f70 6169 725f 7365 7128 7365 7131  ty_pair_seq(seq1
-0000f940: 2c73 6571 3229 3a0a 2020 2020 2222 2243  ,seq2):.    """C
-0000f950: 616c 6375 6c61 7465 7320 7468 6520 7065  alculates the pe
-0000f960: 7263 656e 7420 6964 656e 7469 7479 2061  rcent identity a
-0000f970: 6d6f 6e67 2061 2070 6169 7220 6f66 2073  mong a pair of s
-0000f980: 6571 7565 6e63 6573 0a20 2020 203a 7061  equences.    :pa
-0000f990: 7261 6d20 7374 7220 7365 7131 0a20 2020  ram str seq1.   
-0000f9a0: 203a 7061 7261 6d20 7374 7220 7365 7132   :param str seq2
-0000f9b0: 2222 220a 2020 2020 6920 3d20 300a 2020  """.    i = 0.  
-0000f9c0: 2020 7365 7131 203d 6c69 7374 2873 6571    seq1 =list(seq
-0000f9d0: 3129 0a20 2020 2073 6571 323d 6c69 7374  1).    seq2=list
-0000f9e0: 2873 6571 3229 0a0a 2020 2020 616c 6e20  (seq2)..    aln 
-0000f9f0: 3d20 6e70 2e72 6f77 5f73 7461 636b 285b  = np.row_stack([
-0000fa00: 7365 7131 2c73 6571 325d 290a 2020 2020  seq1,seq2]).    
-0000fa10: 666f 7220 6120 696e 2072 616e 6765 2830  for a in range(0
-0000fa20: 2c20 616c 6e2e 7368 6170 655b 315d 293a  , aln.shape[1]):
-0000fa30: 0a20 2020 2020 2020 2073 203d 2061 6c6e  .        s = aln
-0000fa40: 5b3a 2c61 5d2e 746f 6c69 7374 2829 0a20  [:,a].tolist(). 
-0000fa50: 2020 2020 2020 2073 203d 2022 222e 6a6f         s = "".jo
-0000fa60: 696e 2873 290a 2020 2020 2020 2020 6966  in(s).        if
-0000fa70: 2073 203d 3d20 6c65 6e28 7329 202a 2073   s == len(s) * s
-0000fa80: 5b30 5d3a 0a20 2020 2020 2020 2020 2020  [0]:.           
-0000fa90: 2069 202b 3d20 310a 2020 2020 7265 7475   i += 1.    retu
-0000faa0: 726e 2031 3030 2a69 2f66 6c6f 6174 286c  rn 100*i/float(l
-0000fab0: 656e 2861 6c6e 5b30 5d29 290a 6465 6620  en(aln[0])).def 
-0000fac0: 696e 636f 7272 6563 746c 795f 7072 6564  incorrectly_pred
-0000fad0: 6963 7465 645f 6161 2873 6571 312c 7365  icted_aa(seq1,se
-0000fae0: 7132 293a 0a20 2020 2022 2222 2222 220a  q2):.    """""".
-0000faf0: 2020 2020 6920 3d20 300a 2020 2020 7365      i = 0.    se
-0000fb00: 7131 203d 6c69 7374 2873 6571 3129 0a20  q1 =list(seq1). 
-0000fb10: 2020 2073 6571 323d 6c69 7374 2873 6571     seq2=list(seq
-0000fb20: 3229 0a0a 2020 2020 616c 6e20 3d20 6e70  2)..    aln = np
-0000fb30: 2e72 6f77 5f73 7461 636b 285b 7365 7131  .row_stack([seq1
-0000fb40: 2c73 6571 325d 290a 2020 2020 666f 7220  ,seq2]).    for 
-0000fb50: 6120 696e 2072 616e 6765 2830 2c20 616c  a in range(0, al
-0000fb60: 6e2e 7368 6170 655b 315d 293a 0a20 2020  n.shape[1]):.   
-0000fb70: 2020 2020 2073 203d 2061 6c6e 5b3a 2c61       s = aln[:,a
-0000fb80: 5d2e 746f 6c69 7374 2829 0a20 2020 2020  ].tolist().     
-0000fb90: 2020 2073 203d 2022 222e 6a6f 696e 2873     s = "".join(s
-0000fba0: 290a 2020 2020 2020 2020 6966 2073 203d  ).        if s =
-0000fbb0: 3d20 6c65 6e28 7329 202a 2073 5b30 5d3a  = len(s) * s[0]:
-0000fbc0: 0a20 2020 2020 2020 2020 2020 2069 202b  .            i +
-0000fbd0: 3d20 3120 2373 616d 6520 6161 0a20 2020  = 1 #same aa.   
-0000fbe0: 2072 6574 7572 6e20 6c65 6e28 616c 6e5b   return len(aln[
-0000fbf0: 305d 292d 690a 6465 6620 746f 5f65 7465  0])-i.def to_ete
-0000fc00: 3328 7472 6565 293a 0a20 2020 2069 6d70  3(tree):.    imp
-0000fc10: 6f72 7420 7465 6d70 6669 6c65 0a20 2020  ort tempfile.   
-0000fc20: 2066 726f 6d20 6574 6533 2069 6d70 6f72   from ete3 impor
-0000fc30: 7420 5472 6565 2061 7320 4574 6554 7265  t Tree as EteTre
-0000fc40: 650a 2020 2020 7769 7468 2074 656d 7066  e.    with tempf
-0000fc50: 696c 652e 4e61 6d65 6454 656d 706f 7261  ile.NamedTempora
-0000fc60: 7279 4669 6c65 286d 6f64 653d 2277 2229  ryFile(mode="w")
-0000fc70: 2061 7320 746d 703a 0a20 2020 2020 2020   as tmp:.       
-0000fc80: 2050 6879 6c6f 2e77 7269 7465 2874 7265   Phylo.write(tre
-0000fc90: 652c 2074 6d70 2c20 276e 6577 6963 6b27  e, tmp, 'newick'
-0000fca0: 290a 2020 2020 2020 2020 746d 702e 666c  ).        tmp.fl
-0000fcb0: 7573 6828 290a 2020 2020 2020 2020 7265  ush().        re
-0000fcc0: 7475 726e 2045 7465 5472 6565 2874 6d70  turn EteTree(tmp
-0000fcd0: 2e6e 616d 652c 666f 726d 6174 3d31 290a  .name,format=1).
-0000fce0: 6465 6620 746f 5f62 696f 7079 7468 6f6e  def to_biopython
-0000fcf0: 5472 6565 2874 7265 6529 3a0a 2020 2020  Tree(tree):.    
-0000fd00: 696d 706f 7274 2074 656d 7066 696c 650a  import tempfile.
-0000fd10: 2020 2020 6672 6f6d 2065 7465 3320 696d      from ete3 im
-0000fd20: 706f 7274 2054 7265 6520 6173 2045 7465  port Tree as Ete
-0000fd30: 5472 6565 0a20 2020 2066 726f 6d20 4269  Tree.    from Bi
-0000fd40: 6f20 696d 706f 7274 2050 6879 6c6f 0a20  o import Phylo. 
-0000fd50: 2020 2077 6974 6820 7465 6d70 6669 6c65     with tempfile
-0000fd60: 2e4e 616d 6564 5465 6d70 6f72 6172 7946  .NamedTemporaryF
-0000fd70: 696c 6528 6d6f 6465 3d22 772b 2229 2061  ile(mode="w+") a
-0000fd80: 7320 746d 703a 0a20 2020 2020 2020 2074  s tmp:.        t
-0000fd90: 7265 652e 7772 6974 6528 6f75 7466 696c  ree.write(outfil
-0000fda0: 653d 746d 702c 666f 726d 6174 3d31 2c66  e=tmp,format=1,f
-0000fdb0: 6f72 6d61 745f 726f 6f74 5f6e 6f64 653d  ormat_root_node=
-0000fdc0: 5472 7565 290a 2020 2020 2020 2020 746d  True).        tm
-0000fdd0: 702e 666c 7573 6828 290a 2020 2020 2020  p.flush().      
-0000fde0: 2020 7265 7475 726e 2050 6879 6c6f 2e72    return Phylo.r
-0000fdf0: 6561 6428 746d 702e 6e61 6d65 2c20 276e  ead(tmp.name, 'n
-0000fe00: 6577 6963 6b27 290a 6465 6620 6c6f 6164  ewick').def load
-0000fe10: 5f6f 626a 286e 616d 6529 3a0a 2020 2020  _obj(name):.    
-0000fe20: 2320 7769 7468 206f 7065 6e28 6e61 6d65  # with open(name
-0000fe30: 2c20 2772 6227 2920 6173 2066 3a0a 2020  , 'rb') as f:.  
-0000fe40: 2020 2320 2020 2020 7265 7475 726e 2070    #     return p
-0000fe50: 6963 6b6c 652e 6c6f 6164 2866 290a 2020  ickle.load(f).  
-0000fe60: 2020 6461 7461 203d 2062 7a32 2e42 5a32    data = bz2.BZ2
-0000fe70: 4669 6c65 286e 616d 652c 2022 7262 2229  File(name, "rb")
-0000fe80: 0a20 2020 2064 6174 6120 3d20 6350 6963  .    data = cPic
-0000fe90: 6b6c 652e 6c6f 6164 2864 6174 6129 0a20  kle.load(data). 
-0000fea0: 2020 2023 2077 6974 6820 6f70 656e 286e     # with open(n
-0000feb0: 616d 652c 2272 6222 2920 6173 206a 736f  ame,"rb") as jso
-0000fec0: 6e5f 6669 6c65 3a0a 2020 2020 2320 2020  n_file:.    #   
-0000fed0: 2020 6461 7461 203d 206a 736f 6e2e 6c6f    data = json.lo
-0000fee0: 6164 286a 736f 6e5f 6669 6c65 2c20 656e  ad(json_file, en
-0000fef0: 636f 6469 6e67 3d22 7574 662d 3822 290a  coding="utf-8").
-0000ff00: 2020 2020 2320 6461 7461 5f61 7272 6179      # data_array
-0000ff10: 7320 3d20 7b7d 0a20 2020 2023 2066 6f72  s = {}.    # for
-0000ff20: 206b 6579 2c76 616c 7565 2069 6e20 6461   key,value in da
-0000ff30: 7461 2e69 7465 6d73 2829 3a0a 2020 2020  ta.items():.    
-0000ff40: 2320 2020 2020 6461 7461 5f61 7272 6179  #     data_array
-0000ff50: 735b 6b65 795d 203d 206e 705f 6a61 782e  s[key] = np_jax.
-0000ff60: 6172 7261 7928 7661 6c75 6529 0a20 2020  array(value).   
-0000ff70: 2023 2072 6574 7572 6e20 6461 7461 5f61   # return data_a
-0000ff80: 7272 6179 730a 2020 2020 7265 7475 726e  rrays.    return
-0000ff90: 2064 6174 610a 6465 6620 7361 7665 5f6f   data.def save_o
-0000ffa0: 626a 286f 626a 2c20 6e61 6d65 293a 0a20  bj(obj, name):. 
-0000ffb0: 2020 2023 2077 6974 6820 6f70 656e 286e     # with open(n
-0000ffc0: 616d 652c 2027 7762 2729 2061 7320 663a  ame, 'wb') as f:
-0000ffd0: 0a20 2020 2023 2020 2020 2070 6963 6b6c  .    #     pickl
-0000ffe0: 652e 6475 6d70 286f 626a 2c20 662c 2070  e.dump(obj, f, p
-0000fff0: 6963 6b6c 652e 4849 4748 4553 545f 5052  ickle.HIGHEST_PR
-00010000: 4f54 4f43 4f4c 290a 2020 2020 7769 7468  OTOCOL).    with
-00010010: 2062 7a32 2e42 5a32 4669 6c65 286e 616d   bz2.BZ2File(nam
-00010020: 652c 2022 7762 2229 2061 7320 663a 0a20  e, "wb") as f:. 
-00010030: 2020 2020 2020 2063 5069 636b 6c65 2e64         cPickle.d
-00010040: 756d 7028 6f62 6a2c 2066 290a 2020 2020  ump(obj, f).    
-00010050: 2320 6f62 6a5f 7365 7269 616c 697a 6564  # obj_serialized
-00010060: 3d7b 7d0a 2020 2020 2320 666f 7220 6b65  ={}.    # for ke
-00010070: 792c 7661 6c75 6520 696e 206f 626a 2e69  y,value in obj.i
-00010080: 7465 6d73 2829 3a0a 2020 2020 2320 2020  tems():.    #   
-00010090: 2020 6f62 6a5f 7365 7269 616c 697a 6564    obj_serialized
-000100a0: 5b6b 6579 5d20 3d20 7661 6c75 652e 746f  [key] = value.to
-000100b0: 6c69 7374 2829 0a20 2020 2023 0a20 2020  list().    #.   
-000100c0: 2023 2077 6974 6820 6f70 656e 286e 616d   # with open(nam
-000100d0: 652c 2027 7727 2c65 6e63 6f64 696e 673d  e, 'w',encoding=
-000100e0: 2775 7466 2d38 2729 2061 7320 6f75 7466  'utf-8') as outf
-000100f0: 696c 653a 0a20 2020 2023 2020 2020 206a  ile:.    #     j
-00010100: 736f 6e2e 6475 6d70 286f 626a 5f73 6572  son.dump(obj_ser
-00010110: 6961 6c69 7a65 642c 206f 7574 6669 6c65  ialized, outfile
-00010120: 2c65 6e73 7572 655f 6173 6369 693d 4661  ,ensure_ascii=Fa
-00010130: 6c73 652c 2069 6e64 656e 743d 3429 0a64  lse, indent=4).d
-00010140: 6566 2061 7574 6f6c 6162 656c 2872 6563  ef autolabel(rec
-00010150: 7473 2c20 6178 2c20 626c 6f73 756d 5f64  ts, ax, blosum_d
-00010160: 6963 743d 4e6f 6e65 2c70 6572 6365 6e74  ict=None,percent
-00010170: 5f69 645f 6469 6374 3d4e 6f6e 652c 626c  _id_dict=None,bl
-00010180: 6f73 756d 5f74 7275 655f 6469 6374 3d4e  osum_true_dict=N
-00010190: 6f6e 6529 3a0a 2020 2020 2222 2248 656c  one):.    """Hel
-000101a0: 7065 7220 6675 6e63 7469 6f6e 2074 6f20  per function to 
-000101b0: 6173 7369 676e 2061 2069 6e74 6567 6572  assign a integer
-000101c0: 2076 616c 7565 206f 6e20 746f 7020 6f66   value on top of
-000101d0: 2074 6865 2068 6973 746f 6772 616d 2062   the histogram b
-000101e0: 6172 0a20 2020 203a 7061 7261 6d20 7265  ar.    :param re
-000101f0: 6374 733a 206d 6174 706c 6f74 6c69 6220  cts: matplotlib 
-00010200: 7265 6374 616e 676c 6573 0a20 2020 203a  rectangles.    :
-00010210: 7061 7261 6d20 6178 3a20 6d61 7470 6c6f  param ax: matplo
-00010220: 746c 6962 2061 7869 730a 2020 2020 3a70  tlib axis.    :p
-00010230: 6172 616d 2062 6c6f 7375 6d5f 6469 6374  aram blosum_dict
-00010240: 3a20 6469 6374 696f 6e61 7279 2063 6f6e  : dictionary con
-00010250: 7461 696e 696e 6720 7468 6520 626c 6f73  taining the blos
-00010260: 756d 2073 636f 7265 7320 6f66 2074 6865  um scores of the
-00010270: 2074 7275 6520 7673 2074 6865 2070 7265   true vs the pre
-00010280: 6469 6374 696f 6e20 7365 7175 656e 6365  diction sequence
-00010290: 730a 2020 2020 3a70 6172 616d 2062 6c6f  s.    :param blo
-000102a0: 7375 6d5f 7472 7565 5f64 6963 743a 2064  sum_true_dict: d
-000102b0: 6963 7469 6f6e 6172 7920 636f 6e74 6169  ictionary contai
-000102c0: 6e69 6e67 2074 6865 2062 6c6f 7375 6d20  ning the blosum 
-000102d0: 7363 6f72 6573 206f 6620 7468 6520 7472  scores of the tr
-000102e0: 7565 2076 7320 7472 7565 2073 6571 7565  ue vs true seque
-000102f0: 6e63 6573 2222 220a 2020 2020 2320 4765  nces""".    # Ge
-00010300: 7420 792d 6178 6973 2068 6569 6768 7420  t y-axis height 
-00010310: 746f 2063 616c 6375 6c61 7465 206c 6162  to calculate lab
-00010320: 656c 2070 6f73 6974 696f 6e20 6672 6f6d  el position from
-00010330: 2e0a 2020 2020 2879 5f62 6f74 746f 6d2c  ..    (y_bottom,
-00010340: 2079 5f74 6f70 2920 3d20 6178 2e67 6574   y_top) = ax.get
-00010350: 5f79 6c69 6d28 290a 2020 2020 795f 6865  _ylim().    y_he
-00010360: 6967 6874 203d 2079 5f74 6f70 202d 2079  ight = y_top - y
-00010370: 5f62 6f74 746f 6d0a 2020 2020 6966 2062  _bottom.    if b
-00010380: 6c6f 7375 6d5f 6469 6374 2069 7320 6e6f  losum_dict is no
-00010390: 7420 4e6f 6e65 3a0a 2020 2020 2020 2020  t None:.        
-000103a0: 666f 7220 7265 6374 2c62 6c6f 7375 6d5f  for rect,blosum_
-000103b0: 7363 6f72 652c 7069 642c 626c 6f73 756d  score,pid,blosum
-000103c0: 5f73 636f 7265 5f74 7275 6520 696e 207a  _score_true in z
-000103d0: 6970 2872 6563 7473 2c62 6c6f 7375 6d5f  ip(rects,blosum_
-000103e0: 6469 6374 2e76 616c 7565 7328 292c 7065  dict.values(),pe
-000103f0: 7263 656e 745f 6964 5f64 6963 742e 7661  rcent_id_dict.va
-00010400: 6c75 6573 2829 2c62 6c6f 7375 6d5f 7472  lues(),blosum_tr
-00010410: 7565 5f64 6963 742e 7661 6c75 6573 2829  ue_dict.values()
-00010420: 293a 0a20 2020 2020 2020 2020 2020 2068  ):.            h
-00010430: 6569 6768 7420 3d20 7265 6374 2e67 6574  eight = rect.get
-00010440: 5f68 6569 6768 7428 290a 2020 2020 2020  _height().      
-00010450: 2020 2020 2020 2320 4672 6163 7469 6f6e        # Fraction
-00010460: 206f 6620 6178 6973 2068 6569 6768 7420   of axis height 
-00010470: 7461 6b65 6e20 7570 2062 7920 7468 6973  taken up by this
-00010480: 2072 6563 7461 6e67 6c65 0a20 2020 2020   rectangle.     
-00010490: 2020 2020 2020 2070 5f68 6569 6768 7420         p_height 
-000104a0: 3d20 2868 6569 6768 7420 2f20 795f 6865  = (height / y_he
-000104b0: 6967 6874 290a 2020 2020 2020 2020 2020  ight).          
-000104c0: 2020 2320 4966 2077 6520 6361 6e20 6669    # If we can fi
-000104d0: 7420 7468 6520 6c61 6265 6c20 6162 6f76  t the label abov
-000104e0: 6520 7468 6520 636f 6c75 6d6e 2c20 646f  e the column, do
-000104f0: 2074 6861 743b 0a20 2020 2020 2020 2020   that;.         
-00010500: 2020 2023 206f 7468 6572 7769 7365 2c20     # otherwise, 
-00010510: 7075 7420 6974 2069 6e73 6964 6520 7468  put it inside th
-00010520: 6520 636f 6c75 6d6e 2e0a 2020 2020 2020  e column..      
-00010530: 2020 2020 2020 6966 2070 5f68 6569 6768        if p_heigh
-00010540: 7420 3e20 302e 3935 3a20 2023 2061 7262  t > 0.95:  # arb
-00010550: 6974 7261 7279 3b20 3935 2520 6c6f 6f6b  itrary; 95% look
-00010560: 6564 2067 6f6f 6420 746f 206d 652e 0a20  ed good to me.. 
-00010570: 2020 2020 2020 2020 2020 2020 2020 206c                 l
-00010580: 6162 656c 5f70 6f73 6974 696f 6e20 3d20  abel_position = 
-00010590: 6865 6967 6874 202d 2028 795f 6865 6967  height - (y_heig
-000105a0: 6874 202a 2030 2e30 3529 0a20 2020 2020  ht * 0.05).     
-000105b0: 2020 2020 2020 2065 6c73 653a 0a20 2020         else:.   
-000105c0: 2020 2020 2020 2020 2020 2020 206c 6162               lab
-000105d0: 656c 5f70 6f73 6974 696f 6e20 3d20 6865  el_position = he
-000105e0: 6967 6874 202b 2028 795f 6865 6967 6874  ight + (y_height
-000105f0: 202a 2030 2e30 3129 0a0a 2020 2020 2020   * 0.01)..      
-00010600: 2020 2020 2020 6178 2e74 6578 7428 7265        ax.text(re
-00010610: 6374 2e67 6574 5f78 2829 202b 2072 6563  ct.get_x() + rec
-00010620: 742e 6765 745f 7769 6474 6828 2920 2f20  t.get_width() / 
-00010630: 322e 2c20 6c61 6265 6c5f 706f 7369 7469  2., label_positi
-00010640: 6f6e 2c0a 2020 2020 2020 2020 2020 2020  on,.            
-00010650: 2020 2020 2020 2020 2749 6e41 413a 2564          'InAA:%d
-00010660: 205c 6e20 4273 3a25 6420 7c20 4273 5472   \n Bs:%d | BsTr
-00010670: 7565 3a25 6420 5c6e 2050 4944 3a25 3166  ue:%d \n PID:%1f
-00010680: 2720 2520 2869 6e74 2868 6569 6768 7429  ' % (int(height)
-00010690: 2c62 6c6f 7375 6d5f 7363 6f72 652c 626c  ,blosum_score,bl
-000106a0: 6f73 756d 5f73 636f 7265 5f74 7275 652c  osum_score_true,
-000106b0: 7069 6429 2c0a 2020 2020 2020 2020 2020  pid),.          
-000106c0: 2020 2020 2020 2020 2020 6861 3d27 6365            ha='ce
-000106d0: 6e74 6572 272c 2076 613d 2762 6f74 746f  nter', va='botto
-000106e0: 6d27 290a 2020 2020 656c 7365 3a0a 2020  m').    else:.  
-000106f0: 2020 2020 2020 666f 7220 7265 6374 2069        for rect i
-00010700: 6e20 7265 6374 733a 0a20 2020 2020 2020  n rects:.       
-00010710: 2020 2020 2068 6569 6768 7420 3d20 7265       height = re
-00010720: 6374 2e67 6574 5f68 6569 6768 7428 290a  ct.get_height().
-00010730: 2020 2020 2020 2020 2020 2020 2320 4672              # Fr
-00010740: 6163 7469 6f6e 206f 6620 6178 6973 2068  action of axis h
-00010750: 6569 6768 7420 7461 6b65 6e20 7570 2062  eight taken up b
-00010760: 7920 7468 6973 2072 6563 7461 6e67 6c65  y this rectangle
-00010770: 0a20 2020 2020 2020 2020 2020 2070 5f68  .            p_h
-00010780: 6569 6768 7420 3d20 2868 6569 6768 7420  eight = (height 
-00010790: 2f20 795f 6865 6967 6874 290a 2020 2020  / y_height).    
-000107a0: 2020 2020 2020 2020 2320 4966 2077 6520          # If we 
-000107b0: 6361 6e20 6669 7420 7468 6520 6c61 6265  can fit the labe
-000107c0: 6c20 6162 6f76 6520 7468 6520 636f 6c75  l above the colu
-000107d0: 6d6e 2c20 646f 2074 6861 743b 0a20 2020  mn, do that;.   
-000107e0: 2020 2020 2020 2020 2023 206f 7468 6572           # other
-000107f0: 7769 7365 2c20 7075 7420 6974 2069 6e73  wise, put it ins
-00010800: 6964 6520 7468 6520 636f 6c75 6d6e 2e0a  ide the column..
-00010810: 2020 2020 2020 2020 2020 2020 6966 2070              if p
-00010820: 5f68 6569 6768 7420 3e20 302e 3935 3a20  _height > 0.95: 
-00010830: 2023 2061 7262 6974 7261 7279 3b20 3935   # arbitrary; 95
-00010840: 2520 6c6f 6f6b 6564 2067 6f6f 6420 746f  % looked good to
-00010850: 206d 652e 0a20 2020 2020 2020 2020 2020   me..           
-00010860: 2020 2020 206c 6162 656c 5f70 6f73 6974       label_posit
-00010870: 696f 6e20 3d20 6865 6967 6874 202d 2028  ion = height - (
-00010880: 795f 6865 6967 6874 202a 2030 2e30 3529  y_height * 0.05)
-00010890: 0a20 2020 2020 2020 2020 2020 2065 6c73  .            els
-000108a0: 653a 0a20 2020 2020 2020 2020 2020 2020  e:.             
-000108b0: 2020 206c 6162 656c 5f70 6f73 6974 696f     label_positio
-000108c0: 6e20 3d20 6865 6967 6874 202b 2028 795f  n = height + (y_
-000108d0: 6865 6967 6874 202a 2030 2e30 3129 0a0a  height * 0.01)..
-000108e0: 2020 2020 2020 2020 2020 2020 6178 2e74              ax.t
-000108f0: 6578 7428 7265 6374 2e67 6574 5f78 2829  ext(rect.get_x()
-00010900: 202b 2072 6563 742e 6765 745f 7769 6474   + rect.get_widt
-00010910: 6828 2920 2f20 322e 2c20 6c61 6265 6c5f  h() / 2., label_
-00010920: 706f 7369 7469 6f6e 2c0a 2020 2020 2020  position,.      
-00010930: 2020 2020 2020 2020 2020 2020 2020 2725                '%
-00010940: 6427 2025 2069 6e74 2868 6569 6768 7429  d' % int(height)
-00010950: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-00010960: 2020 2020 2020 6861 3d27 6365 6e74 6572        ha='center
-00010970: 272c 2076 613d 2762 6f74 746f 6d27 290a  ', va='bottom').
-00010980: 6465 6620 7379 6d6d 6574 7269 7a65 2861  def symmetrize(a
-00010990: 293a 0a20 2020 2022 4d61 6b65 2061 2074  ):.    "Make a t
-000109a0: 7269 616e 6775 6c61 7220 7570 7065 7220  riangular upper 
-000109b0: 6d61 7472 6978 2073 796d 6d65 7472 6963  matrix symmetric
-000109c0: 2061 6c6f 6e67 2074 6865 2064 6961 676f   along the diago
-000109d0: 6e61 6c2e 2056 616c 6964 2066 6f72 206e  nal. Valid for n
-000109e0: 756d 7079 2061 7272 6179 7320 616e 6420  umpy arrays and 
-000109f0: 7061 6e64 6173 2064 6174 6166 7261 6d65  pandas dataframe
-00010a00: 7322 0a20 2020 2061 203d 206e 702e 6d61  s".    a = np.ma
-00010a10: 7869 6d75 6d28 612c 2061 2e74 7261 6e73  ximum(a, a.trans
-00010a20: 706f 7365 2829 290a 2020 2020 7265 7475  pose()).    retu
-00010a30: 726e 2061 0a64 6566 2062 7569 6c64 5f70  rn a.def build_p
-00010a40: 7265 6469 6374 6564 5f74 7265 6528 696e  redicted_tree(in
-00010a50: 6465 782c 7361 6d70 6c65 645f 7365 7175  dex,sampled_sequ
-00010a60: 656e 6365 732c 6c65 6166 5f6e 616d 6573  ences,leaf_names
-00010a70: 2c6e 616d 652c 7265 7375 6c74 735f 6469  ,name,results_di
-00010a80: 7265 6374 6f72 792c 6d65 7468 6f64 293a  rectory,method):
-00010a90: 2023 544f 444f 3a20 436f 7272 6563 740a   #TODO: Correct.
-00010aa0: 2020 2020 2222 2242 7569 6c64 7320 6120      """Builds a 
-00010ab0: 7472 6565 2062 6574 7765 656e 2074 6865  tree between the
-00010ac0: 2074 7261 696e 2028 6c65 6176 6573 2920   train (leaves) 
-00010ad0: 6461 7461 2061 6e64 2074 6865 2070 7265  data and the pre
-00010ae0: 6469 6374 6564 2061 6e63 6573 746f 7273  dicted ancestors
-00010af0: 0a20 2020 203a 7061 7261 6d20 696e 6465  .    :param inde
-00010b00: 783a 206e 6f64 6520 696e 6465 780a 2020  x: node index.  
-00010b10: 2020 3a70 6172 616d 2073 616d 706c 6564    :param sampled
-00010b20: 5f73 6571 7565 6e63 6573 3a20 6461 7461  _sequences: data
-00010b30: 7365 745f 7465 7374 2077 6974 6820 7361  set_test with sa
-00010b40: 6d70 6c65 6420 7365 7175 656e 6365 730a  mpled sequences.
-00010b50: 2020 2020 3a70 6172 616d 206c 6561 665f      :param leaf_
-00010b60: 6e61 6d65 733a 206c 6561 6620 6e6f 6465  names: leaf node
-00010b70: 206e 616d 6573 0a20 2020 203a 7061 7261   names.    :para
-00010b80: 6d20 6e61 6d65 3a20 6461 7461 7365 7470  m name: datasetp
-00010b90: 726f 6a65 6374 206e 616d 650a 2020 2020  roject name.    
-00010ba0: 3a70 6172 616d 2072 6573 756c 7473 5f64  :param results_d
-00010bb0: 6972 6563 746f 7279 0a20 2020 203a 7061  irectory.    :pa
-00010bc0: 7261 6d20 6d65 7468 6f64 3a20 7472 6565  ram method: tree
-00010bd0: 2069 6e66 6572 656e 6365 2065 6e67 696e   inference engin
-00010be0: 6522 2222 0a20 2020 2074 7261 696e 696e  e""".    trainin
-00010bf0: 675f 7365 7173 203d 2053 6571 494f 2e70  g_seqs = SeqIO.p
-00010c00: 6172 7365 2822 7b7d 2f7b 7d5f 7472 6169  arse("{}/{}_trai
-00010c10: 6e69 6e67 2e66 6173 7461 222e 666f 726d  ning.fasta".form
-00010c20: 6174 2872 6573 756c 7473 5f64 6972 6563  at(results_direc
-00010c30: 746f 7279 2c20 6e61 6d65 292c 2022 6661  tory, name), "fa
-00010c40: 7374 6122 290a 2020 2020 7769 7468 206f  sta").    with o
-00010c50: 7065 6e28 227b 7d2f 5472 6565 5f41 6c69  pen("{}/Tree_Ali
-00010c60: 676e 6d65 6e74 5f53 616d 706c 6564 2f7b  gnment_Sampled/{
-00010c70: 7d5f 636f 6d62 696e 6564 5f73 616d 706c  }_combined_sampl
-00010c80: 655f 696e 6465 785f 7b7d 2e66 6173 7461  e_index_{}.fasta
-00010c90: 222e 666f 726d 6174 2872 6573 756c 7473  ".format(results
-00010ca0: 5f64 6972 6563 746f 7279 2c20 6e61 6d65  _directory, name
-00010cb0: 2c69 6e64 6578 292c 2027 7727 2920 6173  ,index), 'w') as
-00010cc0: 2077 5f66 696c 653a 0a20 2020 2020 2020   w_file:.       
-00010cd0: 2053 6571 494f 2e77 7269 7465 2874 7261   SeqIO.write(tra
-00010ce0: 696e 696e 675f 7365 7173 2c77 5f66 696c  ining_seqs,w_fil
-00010cf0: 652c 2266 6173 7461 2229 0a20 2020 2020  e,"fasta").     
-00010d00: 2020 2072 6563 6f72 6473 3d5b 5d0a 2020     records=[].  
-00010d10: 2020 2020 2020 666f 7220 7361 6d70 6c65        for sample
-00010d20: 645f 7365 712c 6c65 6166 5f6e 616d 6520  d_seq,leaf_name 
-00010d30: 696e 207a 6970 2873 616d 706c 6564 5f73  in zip(sampled_s
-00010d40: 6571 7565 6e63 6573 2c6c 6561 665f 6e61  equences,leaf_na
-00010d50: 6d65 7329 3a0a 2020 2020 2020 2020 2020  mes):.          
-00010d60: 2020 7265 636f 7264 203d 2053 6571 5265    record = SeqRe
-00010d70: 636f 7264 2e53 6571 5265 636f 7264 2853  cord.SeqRecord(S
-00010d80: 6571 2827 272e 6a6f 696e 2873 616d 706c  eq(''.join(sampl
-00010d90: 6564 5f73 6571 292e 7265 706c 6163 6528  ed_seq).replace(
-00010da0: 222d 222c 2022 2229 292c 616e 6e6f 7461  "-", "")),annota
-00010db0: 7469 6f6e 733d 7b22 6d6f 6c65 6375 6c65  tions={"molecule
-00010dc0: 5f74 7970 6522 3a20 2270 726f 7465 696e  _type": "protein
-00010dd0: 227d 2c69 643d 6c65 6166 5f6e 616d 652c  "},id=leaf_name,
-00010de0: 2064 6573 6372 6970 7469 6f6e 3d22 2229   description="")
-00010df0: 0a20 2020 2020 2020 2020 2020 2072 6563  .            rec
-00010e00: 6f72 6473 2e61 7070 656e 6428 7265 636f  ords.append(reco
-00010e10: 7264 290a 2020 2020 2020 2020 5365 7149  rd).        SeqI
-00010e20: 4f2e 7772 6974 6528 7265 636f 7264 732c  O.write(records,
-00010e30: 775f 6669 6c65 2c22 6661 7374 6122 290a  w_file,"fasta").
-00010e40: 2020 2020 6469 6374 5f61 6c69 676e 2c61      dict_align,a
-00010e50: 6c69 676e 6d65 6e74 203d 2069 6e66 6572  lignment = infer
-00010e60: 5f61 6c69 676e 6d65 6e74 2869 6e70 7574  _alignment(input
-00010e70: 5f6e 616d 655f 6669 6c65 3d22 7b7d 2f54  _name_file="{}/T
-00010e80: 7265 655f 416c 6967 6e6d 656e 745f 5361  ree_Alignment_Sa
-00010e90: 6d70 6c65 642f 7b7d 5f63 6f6d 6269 6e65  mpled/{}_combine
-00010ea0: 645f 7361 6d70 6c65 5f69 6e64 6578 5f7b  d_sample_index_{
-00010eb0: 7d2e 6661 7374 6122 2e66 6f72 6d61 7428  }.fasta".format(
-00010ec0: 7265 7375 6c74 735f 6469 7265 6374 6f72  results_director
-00010ed0: 792c 206e 616d 652c 696e 6465 7829 2c0a  y, name,index),.
-00010ee0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000c2a0: 2020 206e 616d 653d 6e61 6d65 5f66 696c     name=name_fil
+0000c2b0: 652c 0a20 2020 2020 2020 2020 2020 2020  e,.             
+0000c2c0: 2020 2020 2020 2020 206d 6574 686f 643d           method=
+0000c2d0: 6d65 7468 6f64 2c0a 2020 2020 2020 2020  method,.        
+0000c2e0: 2020 2020 2020 2020 2020 2020 2020 7472                tr
+0000c2f0: 6565 5f66 696c 655f 6e61 6d65 3d22 7b7d  ee_file_name="{}
+0000c300: 2f7b 7d2f 7b7d 2e74 7265 6522 2e66 6f72  /{}/{}.tree".for
+0000c310: 6d61 7428 7374 6f72 6167 655f 666f 6c64  mat(storage_fold
+0000c320: 6572 2c6e 616d 655f 6669 6c65 2c6e 616d  er,name_file,nam
+0000c330: 655f 6669 6c65 292c 0a20 2020 2020 2020  e_file),.       
+0000c340: 2020 2020 2020 2020 2020 2020 2020 2074                 t
+0000c350: 7265 655f 6669 6c65 3d74 7265 655f 6669  ree_file=tree_fi
+0000c360: 6c65 2c0a 2020 2020 2020 2020 2020 2020  le,.            
+0000c370: 2020 2020 2020 2020 2020 7374 6f72 6167            storag
+0000c380: 655f 666f 6c64 6572 3d22 7b7d 2f7b 7d22  e_folder="{}/{}"
+0000c390: 2e66 6f72 6d61 7428 7374 6f72 6167 655f  .format(storage_
+0000c3a0: 666f 6c64 6572 2c6e 616d 655f 6669 6c65  folder,name_file
+0000c3b0: 2929 0a20 2020 206d 6178 5f6c 656e 6768  )).    max_lengh
+0000c3c0: 7420 3d20 616c 6967 6e6d 656e 742e 6765  t = alignment.ge
+0000c3d0: 745f 616c 6967 6e6d 656e 745f 6c65 6e67  t_alignment_leng
+0000c3e0: 7468 2829 0a0a 2020 2020 2348 6967 686c  th()..    #Highl
+0000c3f0: 6967 6874 3a20 436f 6d62 696e 696e 6720  ight: Combining 
+0000c400: 7365 7175 656e 6365 7320 696e 2074 6865  sequences in the
+0000c410: 2061 6c69 676e 6d65 6e74 2074 6861 7420   alignment that 
+0000c420: 6861 7665 2061 2050 4442 2073 7472 7563  have a PDB struc
+0000c430: 7475 7265 2061 6e64 2074 686f 7365 2077  ture and those w
+0000c440: 686f 2064 6f6e 2774 0a20 2020 2069 6620  ho don't.    if 
+0000c450: 7072 6f74 5f69 6e66 6f5f 6469 6374 3a20  prot_info_dict: 
+0000c460: 234f 7468 6572 7769 7365 2069 7420 646f  #Otherwise it do
+0000c470: 6573 206e 6f74 206c 6f6f 7020 6f76 6572  es not loop over
+0000c480: 2065 6d70 7479 2064 6963 7469 6f6e 6172   empty dictionar
+0000c490: 6965 7320 616e 6420 646f 6573 206e 6f74  ies and does not
+0000c4a0: 6869 6e67 0a20 2020 2020 2020 206e 6f74  hing.        not
+0000c4b0: 5f61 6c69 676e 6564 5f73 6571 735f 6672  _aligned_seqs_fr
+0000c4c0: 6f6d 5f61 6c69 676e 6d65 6e74 5f66 696c  om_alignment_fil
+0000c4d0: 652e 7570 6461 7465 2828 6b2c 2070 726f  e.update((k, pro
+0000c4e0: 745f 696e 666f 5f64 6963 745b 6b5d 2920  t_info_dict[k]) 
+0000c4f0: 666f 7220 6b2c 2076 2069 6e20 6e6f 745f  for k, v in not_
+0000c500: 616c 6967 6e65 645f 7365 7173 5f66 726f  aligned_seqs_fro
+0000c510: 6d5f 616c 6967 6e6d 656e 745f 6669 6c65  m_alignment_file
+0000c520: 2e69 7465 6d73 2829 2069 6620 6b20 696e  .items() if k in
+0000c530: 2070 726f 745f 696e 666f 5f64 6963 742e   prot_info_dict.
+0000c540: 6b65 7973 2829 2920 2023 7570 6461 7465  keys())  #update
+0000c550: 2074 6865 2061 6c69 676e 6d65 6e74 206b   the alignment k
+0000c560: 6579 7320 7769 7468 2074 6865 6972 2068  eys with their h
+0000c570: 6f6d 6f6c 6f67 2077 6974 6820 7064 6220  omolog with pdb 
+0000c580: 696e 666f 726d 6174 696f 6e2e 2055 7064  information. Upd
+0000c590: 6174 6520 6f6e 6c79 2074 686f 7365 2073  ate only those s
+0000c5a0: 6571 7565 6e63 6573 2069 6e20 7468 6520  equences in the 
+0000c5b0: 616c 6967 6e6d 656e 742e 204d 6166 6674  alignment. Mafft
+0000c5c0: 2061 6e64 2074 6865 2049 5120 7472 6565   and the IQ tree
+0000c5d0: 2070 726f 6772 616d 206d 6967 6874 2064   program might d
+0000c5e0: 6973 6361 7264 2064 6966 6665 7265 6e74  iscard different
+0000c5f0: 2070 726f 7465 696e 7320 2866 6f72 2065   proteins (for e
+0000c600: 7861 6d70 6c65 2074 6865 7920 6472 6f70  xample they drop
+0000c610: 2064 6966 6665 7265 6e74 2069 6465 6e74   different ident
+0000c620: 6963 616c 2070 726f 7465 696e 7329 0a20  ical proteins). 
+0000c630: 2020 2020 2020 2043 6f6d 6269 6e65 645f         Combined_
+0000c640: 6469 6374 203d 206e 6f74 5f61 6c69 676e  dict = not_align
+0000c650: 6564 5f73 6571 735f 6672 6f6d 5f61 6c69  ed_seqs_from_ali
+0000c660: 676e 6d65 6e74 5f66 696c 650a 2020 2020  gnment_file.    
+0000c670: 656c 7365 3a0a 2020 2020 2020 2020 436f  else:.        Co
+0000c680: 6d62 696e 6564 5f64 6963 7420 3d20 6e6f  mbined_dict = no
+0000c690: 745f 616c 6967 6e65 645f 7365 7173 5f66  t_aligned_seqs_f
+0000c6a0: 726f 6d5f 616c 6967 6e6d 656e 745f 6669  rom_alignment_fi
+0000c6b0: 6c65 0a0a 0a20 2020 2069 6620 7265 6e61  le...    if rena
+0000c6c0: 6d65 5f69 6e74 6572 6e61 6c5f 6e6f 6465  me_internal_node
+0000c6d0: 733a 0a20 2020 2020 2020 2069 6620 6e61  s:.        if na
+0000c6e0: 6d65 5f66 696c 652e 7374 6172 7473 7769  me_file.startswi
+0000c6f0: 7468 2822 7369 6d75 6c61 7469 6f6e 7322  th("simulations"
+0000c700: 293a 0a20 2020 2020 2020 2020 2020 7472  ):.           tr
+0000c710: 6565 203d 2072 656e 616d 655f 7472 6565  ee = rename_tree
+0000c720: 5f69 6e74 6572 6e61 6c5f 6e6f 6465 735f  _internal_nodes_
+0000c730: 7369 6d75 6c61 7469 6f6e 7328 7472 6565  simulations(tree
+0000c740: 2c77 6974 685f 696e 6465 7865 733d 4661  ,with_indexes=Fa
+0000c750: 6c73 6529 0a20 2020 2020 2020 2065 6c73  lse).        els
+0000c760: 653a 0a20 2020 2020 2020 2020 2020 7472  e:.           tr
+0000c770: 6565 203d 2072 656e 616d 655f 7472 6565  ee = rename_tree
+0000c780: 5f69 6e74 6572 6e61 6c5f 6e6f 6465 7328  _internal_nodes(
+0000c790: 7472 6565 290a 0a20 2020 206c 6561 6673  tree)..    leafs
+0000c7a0: 5f6e 616d 6573 203d 2074 7265 652e 6765  _names = tree.ge
+0000c7b0: 745f 6c65 6166 5f6e 616d 6573 2829 0a20  t_leaf_names(). 
+0000c7c0: 2020 2070 6963 6b6c 652e 6475 6d70 286c     pickle.dump(l
+0000c7d0: 6561 6673 5f6e 616d 6573 2c6f 7065 6e28  eafs_names,open(
+0000c7e0: 277b 7d2f 7b7d 2f7b 7d5f 4c65 6166 735f  '{}/{}/{}_Leafs_
+0000c7f0: 6e61 6d65 735f 6c69 7374 2e70 272e 666f  names_list.p'.fo
+0000c800: 726d 6174 2873 746f 7261 6765 5f66 6f6c  rmat(storage_fol
+0000c810: 6465 722c 6e61 6d65 5f66 696c 652c 6e61  der,name_file,na
+0000c820: 6d65 5f66 696c 6529 2c20 2777 6227 2929  me_file), 'wb'))
+0000c830: 0a20 2020 2069 6620 6c65 6e28 6c65 6166  .    if len(leaf
+0000c840: 735f 6e61 6d65 7329 203c 3d20 3230 303a  s_names) <= 200:
+0000c850: 0a20 2020 2020 2020 2070 7269 6e74 2822  .        print("
+0000c860: 5265 6e64 6572 696e 6720 7472 6565 2e2e  Rendering tree..
+0000c870: 2e22 290a 2020 2020 2020 2020 7265 6e64  .").        rend
+0000c880: 6572 5f74 7265 6528 7472 6565 2c20 227b  er_tree(tree, "{
+0000c890: 7d2f 7b7d 222e 666f 726d 6174 2873 746f  }/{}".format(sto
+0000c8a0: 7261 6765 5f66 6f6c 6465 722c 6e61 6d65  rage_folder,name
+0000c8b0: 5f66 696c 6529 2c20 6e61 6d65 5f66 696c  _file), name_fil
+0000c8c0: 6529 0a20 2020 2069 6e74 6572 6e61 6c5f  e).    internal_
+0000c8d0: 6e6f 6465 735f 6e61 6d65 7320 3d20 5b6e  nodes_names = [n
+0000c8e0: 6f64 652e 6e61 6d65 2066 6f72 206e 6f64  ode.name for nod
+0000c8f0: 6520 696e 2074 7265 652e 7472 6176 6572  e in tree.traver
+0000c900: 7365 2829 2069 6620 6e6f 7420 6e6f 6465  se() if not node
+0000c910: 2e69 735f 6c65 6166 2829 5d0a 0a20 2020  .is_leaf()]..   
+0000c920: 2061 6e63 6573 746f 7273 5f61 6c6c 203d   ancestors_all =
+0000c930: 5b5d 0a20 2020 2066 6f72 206e 6f64 6520  [].    for node 
+0000c940: 696e 2074 7265 652e 7472 6176 6572 7365  in tree.traverse
+0000c950: 2829 3a0a 2020 2020 2020 2020 616e 6365  ():.        ance
+0000c960: 7374 6f72 735f 6e6f 6465 203d 5b6e 6f64  stors_node =[nod
+0000c970: 652e 6e61 6d65 2e72 6570 6c61 6365 2822  e.name.replace("
+0000c980: 2722 2c22 2229 5d2b 5b6e 6f64 652e 6469  '","")]+[node.di
+0000c990: 7374 5d20 2b5b 616e 6365 7374 6f72 2e6e  st] +[ancestor.n
+0000c9a0: 616d 652e 7265 706c 6163 6528 2227 222c  ame.replace("'",
+0000c9b0: 2222 2920 666f 7220 616e 6365 7374 6f72  "") for ancestor
+0000c9c0: 2069 6e20 6e6f 6465 2e67 6574 5f61 6e63   in node.get_anc
+0000c9d0: 6573 746f 7273 2829 5d0a 2020 2020 2020  estors()].      
+0000c9e0: 2020 616e 6365 7374 6f72 735f 616c 6c2e    ancestors_all.
+0000c9f0: 6170 7065 6e64 2861 6e63 6573 746f 7273  append(ancestors
+0000ca00: 5f6e 6f64 6529 0a20 2020 206c 656e 6774  _node).    lengt
+0000ca10: 6820 3d20 6d61 7828 6d61 7028 6c65 6e2c  h = max(map(len,
+0000ca20: 2061 6e63 6573 746f 7273 5f61 6c6c 2929   ancestors_all))
+0000ca30: 0a20 2020 2061 6e63 6573 746f 7273 5f69  .    ancestors_i
+0000ca40: 6e66 6f20 3d20 6e70 2e61 7272 6179 285b  nfo = np.array([
+0000ca50: 7869 202b 205b 4e6f 6e65 5d20 2a20 286c  xi + [None] * (l
+0000ca60: 656e 6774 6820 2d20 6c65 6e28 7869 2929  ength - len(xi))
+0000ca70: 2066 6f72 2078 6920 696e 2061 6e63 6573   for xi in ances
+0000ca80: 746f 7273 5f61 6c6c 5d29 0a0a 2020 2020  tors_all])..    
+0000ca90: 7472 6565 5f6c 6576 656c 6f72 6465 725f  tree_levelorder_
+0000caa0: 6e61 6d65 7320 3d20 6e70 2e61 7361 7272  names = np.asarr
+0000cab0: 6179 285b 6e6f 6465 2e6e 616d 652e 7265  ay([node.name.re
+0000cac0: 706c 6163 6528 2227 222c 2222 2920 666f  place("'","") fo
+0000cad0: 7220 6e6f 6465 2069 6e20 7472 6565 2e74  r node in tree.t
+0000cae0: 7261 7665 7273 6528 295d 290a 0a20 2020  raverse()])..   
+0000caf0: 2074 7265 655f 6c65 7665 6c6f 7264 6572   tree_levelorder
+0000cb00: 5f64 6973 7420 3d6e 702e 6173 6172 7261  _dist =np.asarra
+0000cb10: 7928 5b6e 6f64 652e 6469 7374 2066 6f72  y([node.dist for
+0000cb20: 206e 6f64 6520 696e 2074 7265 652e 7472   node in tree.tr
+0000cb30: 6176 6572 7365 2829 5d29 0a20 2020 2023  averse()]).    #
+0000cb40: 4164 6420 7468 6520 696e 6465 7820 6f66  Add the index of
+0000cb50: 2074 6865 2073 6571 7565 6e63 6520 696e   the sequence in
+0000cb60: 2074 6865 2074 7265 6520 746f 2074 6865   the tree to the
+0000cb70: 2073 6571 206c 656e 6774 6820 6172 7261   seq length arra
+0000cb80: 7920 696e 666f 0a20 2020 2044 6174 6173  y info.    Datas
+0000cb90: 6574 203d 206e 702e 7a65 726f 7328 286c  et = np.zeros((l
+0000cba0: 656e 2843 6f6d 6269 6e65 645f 6469 6374  en(Combined_dict
+0000cbb0: 292c 206d 6178 5f6c 656e 6768 7420 2b20  ), max_lenght + 
+0000cbc0: 3120 2b20 3120 2b31 2c20 3330 292c 6474  1 + 1 +1, 30),dt
+0000cbd0: 7970 653d 6f62 6a65 6374 2920 2023 2033  ype=object)  # 3
+0000cbe0: 3020 6469 6d20 746f 2061 6363 6f6d 6f64  0 dim to accomod
+0000cbf0: 6174 6520 6769 7420 7665 6374 6f72 732e  ate git vectors.
+0000cc00: 2043 6172 6566 756c 2077 6974 6820 7468   Careful with th
+0000cc10: 6520 2b33 2028 746f 2069 6e63 6c75 6465  e +3 (to include
+0000cc20: 2067 6974 2c20 7365 716c 656e 2061 6e64   git, seqlen and
+0000cc30: 2072 6f77 2f70 726f 7465 696e 206e 616d   row/protein nam
+0000cc40: 6529 0a20 2020 2066 6f72 2069 2c20 286b  e).    for i, (k
+0000cc50: 6579 2c76 616c 2920 696e 2065 6e75 6d65  ey,val) in enume
+0000cc60: 7261 7465 2843 6f6d 6269 6e65 645f 6469  rate(Combined_di
+0000cc70: 6374 2e69 7465 6d73 2829 293a 0a20 2020  ct.items()):.   
+0000cc80: 2020 2020 2061 6c69 676e 6564 5f73 6571       aligned_seq
+0000cc90: 203d 206c 6973 7428 6469 6374 5f61 6c69   = list(dict_ali
+0000cca0: 676e 6d65 6e74 5b6b 6579 5d2e 7374 7269  gnment[key].stri
+0000ccb0: 7028 222c 2229 290a 2020 2020 2020 2020  p(",")).        
+0000ccc0: 6e6f 5f67 6170 5f69 6e64 6578 6573 203d  no_gap_indexes =
+0000ccd0: 206e 702e 7768 6572 6528 6e70 2e61 7272   np.where(np.arr
+0000cce0: 6179 2861 6c69 676e 6564 5f73 6571 2920  ay(aligned_seq) 
+0000ccf0: 213d 2022 2d22 295b 305d 202b 2033 2020  != "-")[0] + 3  
+0000cd00: 2320 706c 7573 2033 2069 6e20 6f72 6465  # plus 3 in orde
+0000cd10: 7220 746f 206d 616b 6520 7468 6520 696e  r to make the in
+0000cd20: 6465 7865 7320 6669 7420 696e 2074 6865  dexes fit in the
+0000cd30: 2066 696e 616c 2064 6174 6166 7261 6d65   final dataframe
+0000cd40: 0a20 2020 2020 2020 2044 6174 6173 6574  .        Dataset
+0000cd50: 5b69 2c30 2c30 5d20 3d20 6b65 792e 7265  [i,0,0] = key.re
+0000cd60: 706c 6163 6528 2227 222c 2222 2920 2372  place("'","") #r
+0000cd70: 6f77 206e 616d 652f 7365 7175 656e 6365  ow name/sequence
+0000cd80: 206e 616d 650a 2020 2020 2020 2020 4461   name.        Da
+0000cd90: 7461 7365 745b 692c 2031 3a33 5d20 3d20  taset[i, 1:3] = 
+0000cda0: 436f 6d62 696e 6564 5f64 6963 745b 6b65  Combined_dict[ke
+0000cdb0: 795d 5b3a 325d 2023 496e 7365 7274 2073  y][:2] #Insert s
+0000cdc0: 6571 206c 656e 2061 6e64 2067 6974 2076  eq len and git v
+0000cdd0: 6563 746f 720a 2020 2020 2020 2020 6966  ector.        if
+0000cde0: 206e 616d 655f 6669 6c65 2069 6e20 5b22   name_file in ["
+0000cdf0: 6265 6e63 686d 6172 6b5f 7261 6e64 616c  benchmark_randal
+0000ce00: 6c22 2c22 6265 6e63 686d 6172 6b5f 7261  l","benchmark_ra
+0000ce10: 6e64 616c 6c5f 6f72 6967 696e 616c 222c  ndall_original",
+0000ce20: 2262 656e 6368 6d61 726b 5f72 616e 6461  "benchmark_randa
+0000ce30: 6c6c 5f6f 7269 6769 6e61 6c5f 6e61 6d69  ll_original_nami
+0000ce40: 6e67 225d 3a23 7468 6569 7220 6c65 6176  ng"]:#their leav
+0000ce50: 6573 2068 6176 6520 6e75 6d62 6572 206e  es have number n
+0000ce60: 616d 6573 2c20 736f 2077 6520 6b65 6570  ames, so we keep
+0000ce70: 2074 6865 6d20 696e 7374 6561 6420 6f66   them instead of
+0000ce80: 2063 6861 6e67 696e 6720 7468 656d 2066   changing them f
+0000ce90: 6f72 2074 6865 2074 7265 6520 6c65 7665  or the tree leve
+0000cea0: 6c20 6f72 6465 7220 6f6e 6573 0a20 2020  l order ones.   
+0000ceb0: 2020 2020 2020 2020 2044 6174 6173 6574           Dataset
+0000cec0: 5b69 2c20 312c 2031 5d20 3d20 696e 7428  [i, 1, 1] = int(
+0000ced0: 6b65 792e 7265 706c 6163 6528 2227 222c  key.replace("'",
+0000cee0: 2022 2229 290a 2020 2020 2020 2020 656c   "")).        el
+0000cef0: 6966 206e 616d 655f 6669 6c65 2069 6e20  if name_file in 
+0000cf00: 5b22 5046 3031 3033 385f 6c69 7063 7469  ["PF01038_lipcti
+0000cf10: 5f6d 7361 5f66 756e 6769 225d 3a0a 2020  _msa_fungi"]:.  
+0000cf20: 2020 2020 2020 2020 2020 4461 7461 7365            Datase
+0000cf30: 745b 692c 2031 2c20 315d 203d 206e 702e  t[i, 1, 1] = np.
+0000cf40: 7768 6572 6528 7472 6565 5f6c 6576 656c  where(tree_level
+0000cf50: 6f72 6465 725f 6e61 6d65 7320 3d3d 206b  order_names == k
+0000cf60: 6579 2e72 6570 6c61 6365 2822 3a22 2c22  ey.replace(":","
+0000cf70: 5f22 2929 5b30 5d5b 305d 2020 2320 7468  _"))[0][0]  # th
+0000cf80: 6520 6e6f 6465 206e 616d 6520 7769 6c6c  e node name will
+0000cf90: 2062 6520 6974 7320 706f 7369 7469 6f6e   be its position
+0000cfa0: 2069 6e20 7468 6520 7472 6565 0a20 2020   in the tree.   
+0000cfb0: 2020 2020 2065 6c73 653a 0a20 2020 2020       else:.     
+0000cfc0: 2020 2020 2020 2044 6174 6173 6574 5b69         Dataset[i
+0000cfd0: 2c31 2c31 5d20 3d20 6e70 2e77 6865 7265  ,1,1] = np.where
+0000cfe0: 2874 7265 655f 6c65 7665 6c6f 7264 6572  (tree_levelorder
+0000cff0: 5f6e 616d 6573 203d 3d20 6b65 792e 7265  _names == key.re
+0000d000: 706c 6163 6528 2227 222c 2222 2929 5b30  place("'",""))[0
+0000d010: 5d5b 305d 2023 7468 6520 6e6f 6465 206e  ][0] #the node n
+0000d020: 616d 6520 7769 6c6c 2062 6520 6974 7320  ame will be its 
+0000d030: 706f 7369 7469 6f6e 2069 6e20 7468 6520  position in the 
+0000d040: 7472 6565 0a20 2020 2020 2020 2044 6174  tree.        Dat
+0000d050: 6173 6574 5b69 2c20 312c 2032 5d20 3d20  aset[i, 1, 2] = 
+0000d060: 2074 7265 655f 6c65 7665 6c6f 7264 6572   tree_levelorder
+0000d070: 5f64 6973 745b 4461 7461 7365 745b 692c  _dist[Dataset[i,
+0000d080: 312c 315d 5d20 2364 6973 7461 6e63 6520  1,1]] #distance 
+0000d090: 746f 2074 6865 2072 6f6f 743f 2074 6861  to the root? tha
+0000d0a0: 7427 7320 6163 636f 7264 696e 6720 746f  t's according to
+0000d0b0: 2074 6865 2064 6f63 756d 656e 7461 7469   the documentati
+0000d0c0: 6f6e 2079 6573 2c20 6275 7420 6973 2064  on yes, but is d
+0000d0d0: 6966 6665 7265 6e74 2066 726f 6d20 7468  ifferent from th
+0000d0e0: 6520 7061 7472 6973 7469 6320 6469 7374  e patristic dist
+0000d0f0: 616e 6365 730a 2020 2020 2020 2020 4461  ances.        Da
+0000d100: 7461 7365 745b 692c 206e 6f5f 6761 705f  taset[i, no_gap_
+0000d110: 696e 6465 7865 735d 203d 2043 6f6d 6269  indexes] = Combi
+0000d120: 6e65 645f 6469 6374 5b6b 6579 5d5b 323a  ned_dict[key][2:
+0000d130: 5d20 2023 2041 7373 6967 6e20 7468 6520  ]  # Assign the 
+0000d140: 6161 2069 6e66 6f20 2869 6e63 6c75 6469  aa info (includi
+0000d150: 6e67 2061 6e67 6c65 7329 2074 6f20 7468  ng angles) to th
+0000d160: 6f73 6520 706f 7369 7469 6f6e 7320 7768  ose positions wh
+0000d170: 6572 6520 7468 6572 6520 6973 206e 6f74  ere there is not
+0000d180: 2061 2067 6170 0a20 2020 2020 2020 2069   a gap.        i
+0000d190: 6620 6f6e 655f 686f 745f 656e 636f 6469  f one_hot_encodi
+0000d1a0: 6e67 3a0a 2020 2020 2020 2020 2020 2020  ng:.            
+0000d1b0: 2348 6967 686c 6967 6874 3a20 6e6f 5f67  #Highlight: no_g
+0000d1c0: 6170 2069 6e64 6578 6573 2069 7320 6e6f  ap indexes is no
+0000d1d0: 7420 6120 626f 6f6c 6561 6e2c 2077 6520  t a boolean, we 
+0000d1e0: 6861 7665 2074 6f20 636f 6e76 6572 7420  have to convert 
+0000d1f0: 6974 0a20 2020 2020 2020 2020 2020 2067  it.            g
+0000d200: 6170 735f 6d61 736b 203d 206e 702e 6f6e  aps_mask = np.on
+0000d210: 6573 286d 6178 5f6c 656e 6768 742b 332c  es(max_lenght+3,
+0000d220: 6e70 2e62 6f6f 6c29 0a20 2020 2020 2020  np.bool).       
+0000d230: 2020 2020 2067 6170 735f 6d61 736b 5b6e       gaps_mask[n
+0000d240: 6f5f 6761 705f 696e 6465 7865 735d 203d  o_gap_indexes] =
+0000d250: 2046 616c 7365 2023 646f 206e 6f74 206b   False #do not k
+0000d260: 6565 7020 7468 6520 706f 7369 7469 6f6e  eep the position
+0000d270: 7320 7768 6572 6520 7468 6572 6520 6973  s where there is
+0000d280: 206e 6f74 2061 2067 6170 0a20 2020 2020   not a gap.     
+0000d290: 2020 2020 2020 2067 6170 735f 6d61 736b         gaps_mask
+0000d2a0: 5b30 5d20 3d20 4661 6c73 6520 236e 6f74  [0] = False #not
+0000d2b0: 2074 6865 2066 6972 7374 2074 776f 2072   the first two r
+0000d2c0: 6f77 730a 2020 2020 2020 2020 2020 2020  ows.            
+0000d2d0: 6761 7073 5f6d 6173 6b5b 315d 203d 2046  gaps_mask[1] = F
+0000d2e0: 616c 7365 0a20 2020 2020 2020 2020 2020  alse.           
+0000d2f0: 2067 6170 735f 6d61 736b 5b32 5d20 3d20   gaps_mask[2] = 
+0000d300: 4661 6c73 650a 2020 2020 2020 2020 2020  False.          
+0000d310: 2020 4461 7461 7365 745b 692c 6761 7073    Dataset[i,gaps
+0000d320: 5f6d 6173 6b2c 305d 203d 2031 2e20 2369  _mask,0] = 1. #i
+0000d330: 6620 6f6e 6520 686f 7420 656e 636f 6469  f one hot encodi
+0000d340: 6e67 2074 6865 2067 6170 7320 7769 7468  ng the gaps with
+0000d350: 2062 6520 6173 7369 676e 6564 2074 6865   be assigned the
+0000d360: 2066 6972 7374 2070 6f73 6974 696f 6e20   first position 
+0000d370: 696e 206f 6e65 2068 6f74 2065 6e63 6f64  in one hot encod
+0000d380: 696e 670a 0a20 2020 2023 2020 5265 636f  ing..    #  Reco
+0000d390: 6e73 7472 7563 7420 7468 6520 7472 6565  nstruct the tree
+0000d3a0: 2066 726f 6d20 7468 6520 6469 7374 616e   from the distan
+0000d3b0: 6365 206d 6174 7269 780a 2020 2020 7072  ce matrix.    pr
+0000d3c0: 696e 7428 2242 7569 6c64 696e 6720 7061  int("Building pa
+0000d3d0: 7472 6973 7469 6320 616e 6420 636c 6164  tristic and clad
+0000d3e0: 6973 7469 6320 6d61 7472 6963 6573 202e  istic matrices .
+0000d3f0: 2e2e 2229 0a20 2020 2074 7265 655f 7361  ..").    tree_sa
+0000d400: 7665 203d 2070 642e 4461 7461 4672 616d  ve = pd.DataFram
+0000d410: 6528 616e 6365 7374 6f72 735f 696e 666f  e(ancestors_info
+0000d420: 290a 2020 2020 2374 7265 655f 7361 7665  ).    #tree_save
+0000d430: 203d 2070 642e 4461 7461 4672 616d 6528   = pd.DataFrame(
+0000d440: 7b22 4e6f 6465 735f 4e61 6d65 7322 3a74  {"Nodes_Names":t
+0000d450: 7265 655f 6c65 7665 6c6f 7264 6572 5f6e  ree_levelorder_n
+0000d460: 616d 6573 2e74 6f6c 6973 7428 292c 2244  ames.tolist(),"D
+0000d470: 6973 7461 6e63 655f 746f 5f72 6f6f 7422  istance_to_root"
+0000d480: 3a74 7265 655f 6c65 7665 6c6f 7264 6572  :tree_levelorder
+0000d490: 5f64 6973 742e 746f 6c69 7374 2829 7d29  _dist.tolist()})
+0000d4a0: 0a20 2020 2074 7265 655f 7361 7665 2e74  .    tree_save.t
+0000d4b0: 6f5f 6373 7628 227b 7d2f 7b7d 2f7b 7d5f  o_csv("{}/{}/{}_
+0000d4c0: 7472 6565 5f6c 6576 656c 6f72 6465 725f  tree_levelorder_
+0000d4d0: 696e 666f 2e63 7376 222e 666f 726d 6174  info.csv".format
+0000d4e0: 2873 746f 7261 6765 5f66 6f6c 6465 722c  (storage_folder,
+0000d4f0: 6e61 6d65 5f66 696c 652c 6e61 6d65 5f66  name_file,name_f
+0000d500: 696c 6529 2c73 6570 3d22 5c74 2229 0a20  ile),sep="\t"). 
+0000d510: 2020 206e 6f64 6573 5f61 6e64 5f6c 6561     nodes_and_lea
+0000d520: 6673 5f6e 616d 6573 203d 2069 6e74 6572  fs_names = inter
+0000d530: 6e61 6c5f 6e6f 6465 735f 6e61 6d65 7320  nal_nodes_names 
+0000d540: 2b20 6c65 6166 735f 6e61 6d65 730a 2020  + leafs_names.  
+0000d550: 2020 6361 6c63 756c 6174 655f 7061 7472    calculate_patr
+0000d560: 6973 7469 635f 6469 7374 616e 6365 286e  istic_distance(n
+0000d570: 616d 655f 6669 6c65 2c43 6f6d 6269 6e65  ame_file,Combine
+0000d580: 645f 6469 6374 2c6e 6f64 6573 5f61 6e64  d_dict,nodes_and
+0000d590: 5f6c 6561 6673 5f6e 616d 6573 2c74 7265  _leafs_names,tre
+0000d5a0: 652c 7472 6565 5f66 696c 652c 227b 7d2f  e,tree_file,"{}/
+0000d5b0: 7b7d 222e 666f 726d 6174 2873 746f 7261  {}".format(stora
+0000d5c0: 6765 5f66 6f6c 6465 722c 6e61 6d65 5f66  ge_folder,name_f
+0000d5d0: 696c 6529 290a 2020 2020 6361 6c63 756c  ile)).    calcul
+0000d5e0: 6174 655f 636c 6f73 6573 745f 6c65 6176  ate_closest_leav
+0000d5f0: 6573 286e 616d 655f 6669 6c65 2c74 7265  es(name_file,tre
+0000d600: 652c 227b 7d2f 7b7d 222e 666f 726d 6174  e,"{}/{}".format
+0000d610: 2873 746f 7261 6765 5f66 6f6c 6465 722c  (storage_folder,
+0000d620: 6e61 6d65 5f66 696c 6529 290a 2020 2020  name_file)).    
+0000d630: 6361 6c63 756c 6174 655f 6469 7265 6374  calculate_direct
+0000d640: 6c79 5f6c 696e 6b65 645f 6e6f 6465 7328  ly_linked_nodes(
+0000d650: 6e61 6d65 5f66 696c 652c 2074 7265 652c  name_file, tree,
+0000d660: 227b 7d2f 7b7d 222e 666f 726d 6174 2873  "{}/{}".format(s
+0000d670: 746f 7261 6765 5f66 6f6c 6465 722c 6e61  torage_folder,na
+0000d680: 6d65 5f66 696c 6529 290a 2020 2020 6361  me_file)).    ca
+0000d690: 6c63 756c 6174 655f 6465 7363 656e 6461  lculate_descenda
+0000d6a0: 6e74 7328 6e61 6d65 5f66 696c 652c 7472  nts(name_file,tr
+0000d6b0: 6565 2c22 7b7d 2f7b 7d22 2e66 6f72 6d61  ee,"{}/{}".forma
+0000d6c0: 7428 7374 6f72 6167 655f 666f 6c64 6572  t(storage_folder
+0000d6d0: 2c6e 616d 655f 6669 6c65 2929 0a20 2020  ,name_file)).   
+0000d6e0: 2070 7269 6e74 2822 5265 6164 7920 616e   print("Ready an
+0000d6f0: 6420 7361 7665 6421 2229 0a20 2020 2070  d saved!").    p
+0000d700: 7269 6e74 2822 4275 696c 6469 6e67 2063  rint("Building c
+0000d710: 6c61 6465 7320 2877 6172 6e69 6e67 3a20  lades (warning: 
+0000d720: 636f 6c6c 6170 7365 7320 7468 6520 6f72  collapses the or
+0000d730: 6967 696e 616c 2074 7265 6521 2922 290a  iginal tree!)").
+0000d740: 2020 2020 6469 7669 6465 5f69 6e74 6f5f      divide_into_
+0000d750: 6d6f 6e6f 7068 796c 6574 6963 5f63 6c61  monophyletic_cla
+0000d760: 6465 7328 7472 6565 2c22 7b7d 2f7b 7d22  des(tree,"{}/{}"
+0000d770: 2e66 6f72 6d61 7428 7374 6f72 6167 655f  .format(storage_
+0000d780: 666f 6c64 6572 2c6e 616d 655f 6669 6c65  folder,name_file
+0000d790: 292c 6e61 6d65 5f66 696c 6529 0a20 2020  ),name_file).   
+0000d7a0: 206e 702e 7361 7665 2822 7b7d 2f7b 7d2f   np.save("{}/{}/
+0000d7b0: 7b7d 5f64 6174 6173 6574 5f6e 756d 7079  {}_dataset_numpy
+0000d7c0: 5f61 6c69 676e 6564 5f7b 7d2e 6e70 7922  _aligned_{}.npy"
+0000d7d0: 2e66 6f72 6d61 7428 7374 6f72 6167 655f  .format(storage_
+0000d7e0: 666f 6c64 6572 2c6e 616d 655f 6669 6c65  folder,name_file
+0000d7f0: 2c6e 616d 655f 6669 6c65 2c6f 6e65 5f68  ,name_file,one_h
+0000d800: 6f74 5f6c 6162 656c 5b30 5d29 2c20 4461  ot_label[0]), Da
+0000d810: 7461 7365 7429 0a20 2020 206d 6178 5f6c  taset).    max_l
+0000d820: 656e 6768 745f 6e6f 745f 616c 6967 6e65  enght_not_aligne
+0000d830: 6420 3d20 6d61 7828 5b69 6e74 2873 6571  d = max([int(seq
+0000d840: 7565 6e63 655b 305d 5b30 5d29 2066 6f72  uence[0][0]) for
+0000d850: 2069 6478 2c73 6571 7565 6e63 6520 696e   idx,sequence in
+0000d860: 2043 6f6d 6269 6e65 645f 6469 6374 2e69   Combined_dict.i
+0000d870: 7465 6d73 2829 5d29 2023 4669 6e64 2074  tems()]) #Find t
+0000d880: 6865 206c 6172 6765 7374 2073 6571 7565  he largest seque
+0000d890: 6e63 6520 7769 7468 6f75 7420 6265 696e  nce without bein
+0000d8a0: 6720 616c 6967 6e65 640a 2020 2020 7072  g aligned.    pr
+0000d8b0: 696e 7428 2243 7265 6174 696e 6720 6e6f  int("Creating no
+0000d8c0: 7420 616c 6967 6e65 6420 6461 7461 7365  t aligned datase
+0000d8d0: 742e 2e2e 2229 0a20 2020 2044 6174 6173  t...").    Datas
+0000d8e0: 6574 5f6e 6f74 5f61 6c69 676e 6564 203d  et_not_aligned =
+0000d8f0: 206e 702e 7a65 726f 7328 286c 656e 2843   np.zeros((len(C
+0000d900: 6f6d 6269 6e65 645f 6469 6374 292c 206d  ombined_dict), m
+0000d910: 6178 5f6c 656e 6768 745f 6e6f 745f 616c  ax_lenght_not_al
+0000d920: 6967 6e65 6420 2b33 2c20 3330 292c 2064  igned +3, 30), d
+0000d930: 7479 7065 3d6f 626a 6563 7429 2020 2320  type=object)  # 
+0000d940: 3330 2066 6f72 2066 7574 7572 6520 6769  30 for future gi
+0000d950: 7420 7665 6374 6f72 732e 2043 6172 6566  t vectors. Caref
+0000d960: 756c 2077 6974 6820 7468 6520 2b33 0a20  ul with the +3. 
+0000d970: 2020 2066 6f72 2069 2c28 6b65 792c 7661     for i,(key,va
+0000d980: 6c75 6529 2069 6e20 656e 756d 6572 6174  lue) in enumerat
+0000d990: 6528 436f 6d62 696e 6564 5f64 6963 742e  e(Combined_dict.
+0000d9a0: 6974 656d 7328 2929 3a0a 2020 2020 2020  items()):.      
+0000d9b0: 2020 4461 7461 7365 745f 6e6f 745f 616c    Dataset_not_al
+0000d9c0: 6967 6e65 645b 692c 302c 305d 203d 206b  igned[i,0,0] = k
+0000d9d0: 6579 0a20 2020 2020 2020 2044 6174 6173  ey.        Datas
+0000d9e0: 6574 5f6e 6f74 5f61 6c69 676e 6564 5b69  et_not_aligned[i
+0000d9f0: 2c20 313a 335d 203d 2043 6f6d 6269 6e65  , 1:3] = Combine
+0000da00: 645f 6469 6374 5b6b 6579 5d5b 3a32 5d20  d_dict[key][:2] 
+0000da10: 2346 696c 6c20 696e 2074 6865 2073 6571  #Fill in the seq
+0000da20: 7565 6e63 6520 6c65 6e67 6874 2061 6e64  uence lenght and
+0000da30: 2074 6865 2067 6974 2076 6563 746f 720a   the git vector.
+0000da40: 2020 2020 2020 2020 6966 206e 616d 655f          if name_
+0000da50: 6669 6c65 2069 6e20 5b22 6265 6e63 686d  file in ["benchm
+0000da60: 6172 6b5f 7261 6e64 616c 6c5f 6f72 6967  ark_randall_orig
+0000da70: 696e 616c 5f6e 616d 696e 6722 5d3a 0a20  inal_naming"]:. 
+0000da80: 2020 2020 2020 2020 2020 2044 6174 6173             Datas
+0000da90: 6574 5b69 2c20 312c 2031 5d20 3d20 696e  et[i, 1, 1] = in
+0000daa0: 7428 6b65 792e 7265 706c 6163 6528 2227  t(key.replace("'
+0000dab0: 222c 2022 2229 290a 2020 2020 2020 2020  ", "")).        
+0000dac0: 656c 7365 3a0a 2020 2020 2020 2020 2020  else:.          
+0000dad0: 2020 4461 7461 7365 745b 692c 312c 315d    Dataset[i,1,1]
+0000dae0: 203d 206e 702e 7768 6572 6528 7472 6565   = np.where(tree
+0000daf0: 5f6c 6576 656c 6f72 6465 725f 6e61 6d65  _levelorder_name
+0000db00: 7320 3d3d 206b 6579 2e72 6570 6c61 6365  s == key.replace
+0000db10: 2822 2722 2c22 2229 295b 305d 5b30 5d20  ("'",""))[0][0] 
+0000db20: 2370 6f73 6974 696f 6e20 696e 2074 6865  #position in the
+0000db30: 2074 7265 650a 2020 2020 2020 2020 4461   tree.        Da
+0000db40: 7461 7365 745f 6e6f 745f 616c 6967 6e65  taset_not_aligne
+0000db50: 645b 692c 2031 2c20 325d 203d 2020 7472  d[i, 1, 2] =  tr
+0000db60: 6565 5f6c 6576 656c 6f72 6465 725f 6469  ee_levelorder_di
+0000db70: 7374 5b44 6174 6173 6574 5b69 2c31 2c31  st[Dataset[i,1,1
+0000db80: 5d5d 0a20 2020 2020 2020 2044 6174 6173  ]].        Datas
+0000db90: 6574 5f6e 6f74 5f61 6c69 676e 6564 5b69  et_not_aligned[i
+0000dba0: 2c20 333a 696e 7428 436f 6d62 696e 6564  , 3:int(Combined
+0000dbb0: 5f64 6963 745b 6b65 795d 5b30 5d5b 305d  _dict[key][0][0]
+0000dbc0: 2920 2b33 5d20 3d20 436f 6d62 696e 6564  ) +3] = Combined
+0000dbd0: 5f64 6963 745b 6b65 795d 5b32 3a5d 2023  _dict[key][2:] #
+0000dbe0: 4669 6c6c 2069 6e20 7468 6520 616d 696e  Fill in the amin
+0000dbf0: 6f20 6163 6964 7320 226c 6574 7465 7273  o acids "letters
+0000dc00: 222f 226e 756d 6265 7273 2220 616e 6420  "/"numbers" and 
+0000dc10: 7468 6569 7220 616e 676c 6573 0a20 2020  their angles.   
+0000dc20: 2020 2020 2069 6620 6f6e 655f 686f 745f       if one_hot_
+0000dc30: 656e 636f 6469 6e67 3a0a 2020 2020 2020  encoding:.      
+0000dc40: 2020 2020 2020 2320 2348 6967 686c 6967        # #Highlig
+0000dc50: 6874 3a20 6e6f 5f67 6170 2069 6e64 6578  ht: no_gap index
+0000dc60: 6573 2069 7320 6e6f 7420 6120 626f 6f6c  es is not a bool
+0000dc70: 6561 6e2c 2077 6520 6861 7665 2074 6f20  ean, we have to 
+0000dc80: 636f 6e76 6572 7420 6974 0a20 2020 2020  convert it.     
+0000dc90: 2020 2020 2020 2023 2067 6170 735f 6d61         # gaps_ma
+0000dca0: 736b 203d 206e 702e 6f6e 6573 286d 6178  sk = np.ones(max
+0000dcb0: 5f6c 656e 6768 742b 332c 6e70 2e62 6f6f  _lenght+3,np.boo
+0000dcc0: 6c29 0a20 2020 2020 2020 2020 2020 2023  l).            #
+0000dcd0: 2067 6170 735f 6d61 736b 5b6e 6f5f 6761   gaps_mask[no_ga
+0000dce0: 705f 696e 6465 7865 735d 203d 2046 616c  p_indexes] = Fal
+0000dcf0: 7365 2023 646f 206e 6f74 206b 6565 7020  se #do not keep 
+0000dd00: 7468 6520 706f 7369 7469 6f6e 7320 7768  the positions wh
+0000dd10: 6572 6520 7468 6572 6520 6973 206e 6f74  ere there is not
+0000dd20: 2061 2067 6170 0a20 2020 2020 2020 2020   a gap.         
+0000dd30: 2020 2023 2067 6170 735f 6d61 736b 5b30     # gaps_mask[0
+0000dd40: 5d20 3d20 4661 6c73 6520 236e 6f74 2074  ] = False #not t
+0000dd50: 6865 2066 6972 7374 2074 776f 2072 6f77  he first two row
+0000dd60: 730a 2020 2020 2020 2020 2020 2020 2320  s.            # 
+0000dd70: 6761 7073 5f6d 6173 6b5b 315d 203d 2046  gaps_mask[1] = F
+0000dd80: 616c 7365 0a20 2020 2020 2020 2020 2020  alse.           
+0000dd90: 2023 2067 6170 735f 6d61 736b 5b32 5d20   # gaps_mask[2] 
+0000dda0: 3d20 4661 6c73 650a 2020 2020 2020 2020  = False.        
+0000ddb0: 2020 2020 2320 4461 7461 7365 745b 692c      # Dataset[i,
+0000ddc0: 6761 7073 5f6d 6173 6b2c 305d 203d 2031  gaps_mask,0] = 1
+0000ddd0: 2e20 2369 6620 6f6e 6520 686f 7420 656e  . #if one hot en
+0000dde0: 636f 6469 6e67 2074 6865 2067 6170 7320  coding the gaps 
+0000ddf0: 7769 7468 2062 6520 6173 7369 676e 6564  with be assigned
+0000de00: 2074 6865 2066 6972 7374 2070 6f73 6974   the first posit
+0000de10: 696f 6e20 696e 206f 6e65 2068 6f74 2065  ion in one hot e
+0000de20: 6e63 6f64 696e 670a 2020 2020 2020 2020  ncoding.        
+0000de30: 2020 2020 4461 7461 7365 745f 6e6f 745f      Dataset_not_
+0000de40: 616c 6967 6e65 645b 692c 2028 696e 7428  aligned[i, (int(
+0000de50: 436f 6d62 696e 6564 5f64 6963 745b 6b65  Combined_dict[ke
+0000de60: 795d 5b30 5d5b 305d 2920 2b20 3329 3a2c  y][0][0]) + 3):,
+0000de70: 305d 203d 2031 2e0a 2020 2020 6e70 2e73  0] = 1..    np.s
+0000de80: 6176 6528 227b 7d2f 7b7d 2f7b 7d5f 6461  ave("{}/{}/{}_da
+0000de90: 7461 7365 745f 6e75 6d70 795f 4e4f 545f  taset_numpy_NOT_
+0000dea0: 616c 6967 6e65 645f 7b7d 2e6e 7079 222e  aligned_{}.npy".
+0000deb0: 666f 726d 6174 2873 746f 7261 6765 5f66  format(storage_f
+0000dec0: 6f6c 6465 722c 6e61 6d65 5f66 696c 652c  older,name_file,
+0000ded0: 6e61 6d65 5f66 696c 652c 6f6e 655f 686f  name_file,one_ho
+0000dee0: 745f 6c61 6265 6c5b 305d 292c 2044 6174  t_label[0]), Dat
+0000def0: 6173 6574 5f6e 6f74 5f61 6c69 676e 6564  aset_not_aligned
+0000df00: 290a 0a20 2020 2072 6574 7572 6e20 7472  )..    return tr
+0000df10: 6565 5f66 696c 650a 0a64 6566 2073 796d  ee_file..def sym
+0000df20: 6d65 7472 697a 655f 616e 645f 636c 6561  metrize_and_clea
+0000df30: 6e28 6d61 7472 6978 2c61 6e63 6573 7472  n(matrix,ancestr
+0000df40: 616c 3d54 7275 6529 3a0a 2020 2020 2222  al=True):.    ""
+0000df50: 2252 656d 6f76 652c 2069 6620 6e65 6365  "Remove, if nece
+0000df60: 7373 6172 7920 7468 6520 616e 6365 7374  ssary the ancest
+0000df70: 7261 6c20 6e6f 6465 7320 6672 6f6d 2074  ral nodes from t
+0000df80: 6865 2074 7261 696e 206d 6174 7269 780a  he train matrix.
+0000df90: 2020 2020 3a70 6172 616d 2070 616e 6461      :param panda
+0000dfa0: 732d 6172 7261 7920 6d61 7472 6978 3a20  s-array matrix: 
+0000dfb0: 7061 7472 6973 7469 6320 6f72 2063 6c61  patristic or cla
+0000dfc0: 6469 7374 6963 206d 6174 7269 6365 7320  distic matrices 
+0000dfd0: 7769 7468 2063 6f6c 756d 6e73 2061 6e64  with columns and
+0000dfe0: 2069 6e64 6578 6573 2061 7320 6e6f 6465   indexes as node
+0000dff0: 206e 616d 6573 2022 2222 0a20 2020 2069   names """.    i
+0000e000: 6620 6e6f 7420 616e 6365 7374 7261 6c3a  f not ancestral:
+0000e010: 2344 726f 7020 7468 6520 616e 6365 7374  #Drop the ancest
+0000e020: 7261 6c20 6e6f 6465 7320 696e 666f 726d  ral nodes inform
+0000e030: 6174 696f 6e0a 2020 2020 2020 2020 6d61  ation.        ma
+0000e040: 7472 6978 203d 206d 6174 7269 785b 7e6d  trix = matrix[~m
+0000e050: 6174 7269 782e 696e 6465 782e 7374 722e  atrix.index.str.
+0000e060: 636f 6e74 6169 6e73 2827 5e41 7b31 7d5b  contains('^A{1}[
+0000e070: 302d 395d 2b28 3f21 5b41 2d5a 5d29 2b27  0-9]+(?![A-Z])+'
+0000e080: 295d 2020 2320 7265 7065 6174 2077 6974  )]  # repeat wit
+0000e090: 6820 5b61 2d7a 5d20 6966 2070 726f 626c  h [a-z] if probl
+0000e0a0: 656d 730a 2020 2020 2020 2020 6d61 7472  ems.        matr
+0000e0b0: 6978 203d 206d 6174 7269 782e 6c6f 635b  ix = matrix.loc[
+0000e0c0: 3a2c 7e6d 6174 7269 782e 636f 6c75 6d6e  :,~matrix.column
+0000e0d0: 732e 7374 722e 636f 6e74 6169 6e73 2827  s.str.contains('
+0000e0e0: 5e41 7b31 7d5b 302d 395d 2b28 3f21 5b41  ^A{1}[0-9]+(?![A
+0000e0f0: 2d5a 5d29 2b27 295d 0a20 2020 206d 6174  -Z])+')].    mat
+0000e100: 7269 7820 3d20 7379 6d6d 6574 7269 7a65  rix = symmetrize
+0000e110: 286d 6174 7269 7829 0a20 2020 2072 6574  (matrix).    ret
+0000e120: 7572 6e20 6d61 7472 6978 0a64 6566 2072  urn matrix.def r
+0000e130: 656e 616d 655f 6178 6973 286d 6174 7269  ename_axis(matri
+0000e140: 782c 6e6f 6465 732c 6e61 6d65 5f66 696c  x,nodes,name_fil
+0000e150: 6520 3d20 4e6f 6e65 293a 0a20 2020 2022  e = None):.    "
+0000e160: 2222 0a20 2020 2052 656e 616d 6520 7468  "".    Rename th
+0000e170: 6520 6e6f 6465 7320 6e61 6d65 7320 746f  e nodes names to
+0000e180: 2074 6865 6972 2074 7265 6520 6c65 7665   their tree leve
+0000e190: 6c20 696e 6465 780a 2020 2020 6e6f 6465  l index.    node
+0000e1a0: 733a 2074 7265 6520 6c65 7665 6c20 6f72  s: tree level or
+0000e1b0: 6465 7220 6e6f 6465 206e 616d 6573 2222  der node names""
+0000e1c0: 220a 2020 2020 6966 206c 656e 286e 6f64  ".    if len(nod
+0000e1d0: 6573 2920 213d 2030 2061 6e64 206e 616d  es) != 0 and nam
+0000e1e0: 655f 6669 6c65 206e 6f74 2069 6e20 5b22  e_file not in ["
+0000e1f0: 6265 6e63 686d 6172 6b5f 7261 6e64 616c  benchmark_randal
+0000e200: 6c22 2c22 6265 6e63 686d 6172 6b5f 7261  l","benchmark_ra
+0000e210: 6e64 616c 6c5f 6f72 6967 696e 616c 222c  ndall_original",
+0000e220: 2262 656e 6368 6d61 726b 5f72 616e 6461  "benchmark_randa
+0000e230: 6c6c 5f6f 7269 6769 6e61 6c5f 6e61 6d69  ll_original_nami
+0000e240: 6e67 225d 3a20 2361 6e64 206e 6f74 206e  ng"]: #and not n
+0000e250: 616d 652e 7374 6172 7473 7769 7468 2822  ame.startswith("
+0000e260: 7369 6d75 6c61 7469 6f6e 7322 293a 2023  simulations"): #
+0000e270: 7573 6520 7468 6520 6c65 7665 6c20 6f72  use the level or
+0000e280: 6465 7220 7472 616e 7376 6572 7361 6c20  der transversal 
+0000e290: 7472 6565 2069 6e66 6f72 6d61 7469 6f6e  tree information
+0000e2a0: 0a20 2020 2020 2020 2023 4869 6768 6c69  .        #Highli
+0000e2b0: 6768 743a 2049 6620 6e61 6e20 2870 642e  ght: If nan (pd.
+0000e2c0: 6973 6e75 6c6c 2920 6973 2066 6f75 6e64  isnull) is found
+0000e2d0: 2c20 6973 2062 6563 6175 7365 2074 6865  , is because the
+0000e2e0: 2072 6f6f 7420 6e61 6d65 2069 7320 6d65   root name is me
+0000e2f0: 7373 6564 2075 7020 616e 6420 6d69 7373  ssed up and miss
+0000e300: 696e 672c 206a 7573 7420 7772 6974 6520  ing, just write 
+0000e310: 6974 2064 6f77 6e20 696e 2074 6865 2070  it down in the p
+0000e320: 6174 7269 7374 6963 206d 6174 7269 7820  atristic matrix 
+0000e330: 6669 6c65 2121 2121 204f 7220 7468 6520  file!!!! Or the 
+0000e340: 6e61 6d65 7320 696e 2074 6865 206d 6174  names in the mat
+0000e350: 7269 7820 213d 206e 616d 6573 2069 6e20  rix != names in 
+0000e360: 7472 6576 656c 206f 7264 6572 0a0a 2020  trevel order..  
+0000e370: 2020 2020 2020 6d61 7472 6978 2e69 6e64        matrix.ind
+0000e380: 6578 203d 205b 6e70 2e77 6865 7265 286e  ex = [np.where(n
+0000e390: 6f64 6573 203d 3d20 6e6f 6465 5f6e 616d  odes == node_nam
+0000e3a0: 6529 5b30 5d5b 305d 2066 6f72 206e 6f64  e)[0][0] for nod
+0000e3b0: 655f 6e61 6d65 2069 6e20 6d61 7472 6978  e_name in matrix
+0000e3c0: 2e69 6e64 6578 5d20 2354 4f44 4f3b 2077  .index] #TODO; w
+0000e3d0: 6879 2061 6d20 4920 646f 696e 6720 7468  hy am I doing th
+0000e3e0: 6973 2074 7769 6365 3f0a 2020 2020 2020  is twice?.      
+0000e3f0: 2020 6d61 7472 6978 2e63 6f6c 756d 6e73    matrix.columns
+0000e400: 203d 205b 6e70 2e77 6865 7265 286e 6f64   = [np.where(nod
+0000e410: 6573 203d 3d20 6e6f 6465 5f6e 616d 6529  es == node_name)
+0000e420: 5b30 5d5b 305d 2066 6f72 206e 6f64 655f  [0][0] for node_
+0000e430: 6e61 6d65 2069 6e20 6d61 7472 6978 2e63  name in matrix.c
+0000e440: 6f6c 756d 6e73 5d0a 2020 2020 2020 2020  olumns].        
+0000e450: 7265 7475 726e 206d 6174 7269 780a 2020  return matrix.  
+0000e460: 2020 656c 6966 2020 6e61 6d65 5f66 696c    elif  name_fil
+0000e470: 6520 696e 205b 2262 656e 6368 6d61 726b  e in ["benchmark
+0000e480: 5f72 616e 6461 6c6c 222c 2262 656e 6368  _randall","bench
+0000e490: 6d61 726b 5f72 616e 6461 6c6c 5f6f 7269  mark_randall_ori
+0000e4a0: 6769 6e61 6c22 2c22 6265 6e63 686d 6172  ginal","benchmar
+0000e4b0: 6b5f 7261 6e64 616c 6c5f 6f72 6967 696e  k_randall_origin
+0000e4c0: 616c 5f6e 616d 696e 6722 5d20 3a0a 2020  al_naming"] :.  
+0000e4d0: 2020 2020 2020 6d61 7472 6978 2e69 6e64        matrix.ind
+0000e4e0: 6578 203d 206d 6174 7269 782e 696e 6465  ex = matrix.inde
+0000e4f0: 782e 7374 722e 7265 706c 6163 6528 2241  x.str.replace("A
+0000e500: 222c 2222 292e 6173 7479 7065 2822 696e  ","").astype("in
+0000e510: 7422 290a 2020 2020 2020 2020 6d61 7472  t").        matr
+0000e520: 6978 2e63 6f6c 756d 6e73 203d 206d 6174  ix.columns = mat
+0000e530: 7269 782e 636f 6c75 6d6e 732e 7374 722e  rix.columns.str.
+0000e540: 7265 706c 6163 6528 2241 222c 2022 2229  replace("A", "")
+0000e550: 2e61 7374 7970 6528 2269 6e74 2229 0a20  .astype("int"). 
+0000e560: 2020 2020 2020 2072 6574 7572 6e20 6d61         return ma
+0000e570: 7472 6978 0a0a 2020 2020 656c 7365 3a0a  trix..    else:.
+0000e580: 2020 2020 2020 2020 7265 7475 726e 206d          return m
+0000e590: 6174 7269 780a 6465 6620 7375 6d5f 6d61  atrix.def sum_ma
+0000e5a0: 7472 6963 6573 286d 6174 7269 7831 2c6d  trices(matrix1,m
+0000e5b0: 6174 7269 7832 293a 0a20 2020 2022 2222  atrix2):.    """
+0000e5c0: 5375 6d73 2063 6c61 6469 7374 6963 2061  Sums cladistic a
+0000e5d0: 6e64 2070 6174 7269 7374 6963 206d 6174  nd patristic mat
+0000e5e0: 7269 6365 7320 746f 2066 6f72 6d20 616e  rices to form an
+0000e5f0: 2065 766f 6c75 7469 6f6e 6172 7920 6d61   evolutionary ma
+0000e600: 7472 6978 0a20 2020 203a 7061 7261 6d3a  trix.    :param:
+0000e610: 2070 616e 6461 732d 6172 7261 7920 6d61   pandas-array ma
+0000e620: 7472 6978 310a 2020 2020 3a70 6172 616d  trix1.    :param
+0000e630: 3a20 7061 6e64 6173 2d61 7272 6179 206d  : pandas-array m
+0000e640: 6174 7269 7832 0a20 2020 2022 2222 0a20  atrix2.    """. 
+0000e650: 2020 2063 6f6c 756d 6e5f 6e61 6d65 7320     column_names 
+0000e660: 3d20 6d61 7472 6978 312e 636f 6c75 6d6e  = matrix1.column
+0000e670: 732e 7661 6c75 6573 2020 2320 3832 202b  s.values  # 82 +
+0000e680: 2031 0a20 2020 2063 6f6c 756d 6e5f 6e61   1.    column_na
+0000e690: 6d65 7320 3d20 6e70 2e63 6f6e 6361 7465  mes = np.concate
+0000e6a0: 6e61 7465 2828 6e70 2e61 7272 6179 285b  nate((np.array([
+0000e6b0: 666c 6f61 7428 222d 696e 6622 295d 292c  float("-inf")]),
+0000e6c0: 2063 6f6c 756d 6e5f 6e61 6d65 7329 2920   column_names)) 
+0000e6d0: 2023 2038 3220 2b20 3120 2874 6f20 6669   # 82 + 1 (to fi
+0000e6e0: 7420 7468 6520 6e75 6d70 7920 6172 7261  t the numpy arra
+0000e6f0: 790a 2020 2020 726f 7773 5f6e 616d 6573  y.    rows_names
+0000e700: 203d 206d 6174 7269 7831 2e69 6e64 6578   = matrix1.index
+0000e710: 2e76 616c 7565 7320 2023 2038 320a 2020  .values  # 82.  
+0000e720: 2020 6d61 7472 6978 3120 3d20 6d61 7472    matrix1 = matr
+0000e730: 6978 312e 746f 5f6e 756d 7079 2829 0a20  ix1.to_numpy(). 
+0000e740: 2020 206d 6174 7269 7832 203d 206d 6174     matrix2 = mat
+0000e750: 7269 7832 2e74 6f5f 6e75 6d70 7928 290a  rix2.to_numpy().
+0000e760: 2020 2020 6d61 7472 6978 203d 206d 6174      matrix = mat
+0000e770: 7269 7831 202b 206d 6174 7269 7832 0a20  rix1 + matrix2. 
+0000e780: 2020 206d 6174 7269 7820 3d20 6e6f 726d     matrix = norm
+0000e790: 616c 697a 655f 7374 616e 6461 7269 7a65  alize_standarize
+0000e7a0: 286d 6174 7269 7829 0a20 2020 206d 6174  (matrix).    mat
+0000e7b0: 7269 7820 3d20 6e70 2e63 6f6c 756d 6e5f  rix = np.column_
+0000e7c0: 7374 6163 6b28 2872 6f77 735f 6e61 6d65  stack((rows_name
+0000e7d0: 732c 206d 6174 7269 7829 290a 2020 2020  s, matrix)).    
+0000e7e0: 6d61 7472 6978 203d 206e 702e 726f 775f  matrix = np.row_
+0000e7f0: 7374 6163 6b28 2863 6f6c 756d 6e5f 6e61  stack((column_na
+0000e800: 6d65 732c 206d 6174 7269 7829 290a 2020  mes, matrix)).  
+0000e810: 2020 7265 7475 726e 206d 6174 7269 780a    return matrix.
+0000e820: 6465 6620 7061 6e64 6173 5f74 6f5f 6e75  def pandas_to_nu
+0000e830: 6d70 7928 6d61 7472 6978 293a 0a20 2020  mpy(matrix):.   
+0000e840: 2022 2222 436f 6e76 6572 7473 2070 616e   """Converts pan
+0000e850: 6461 7320 6172 7261 7920 746f 206e 756d  das array to num
+0000e860: 7079 2061 7272 6179 2c20 696e 2074 6869  py array, in thi
+0000e870: 7320 6361 7365 2062 7920 7472 616e 7366  s case by transf
+0000e880: 6f72 6d69 6e67 2074 6865 206e 6f64 6573  orming the nodes
+0000e890: 206e 616d 6573 0a20 2020 203a 7061 7261   names.    :para
+0000e8a0: 6d20 6d61 7472 6978 2222 220a 2020 2020  m matrix""".    
+0000e8b0: 636f 6c75 6d6e 5f6e 616d 6573 203d 206d  column_names = m
+0000e8c0: 6174 7269 782e 636f 6c75 6d6e 732e 7661  atrix.columns.va
+0000e8d0: 6c75 6573 2e61 7374 7970 6528 2269 6e74  lues.astype("int
+0000e8e0: 2229 2020 2320 3832 202b 2031 0a20 2020  ")  # 82 + 1.   
+0000e8f0: 2063 6f6c 756d 6e5f 6e61 6d65 7320 3d20   column_names = 
+0000e900: 6e70 2e63 6f6e 6361 7465 6e61 7465 2828  np.concatenate((
+0000e910: 6e70 2e61 7272 6179 285b 666c 6f61 7428  np.array([float(
+0000e920: 222d 696e 6622 295d 292c 2063 6f6c 756d  "-inf")]), colum
+0000e930: 6e5f 6e61 6d65 7329 2920 2023 2038 3220  n_names))  # 82 
+0000e940: 2b20 3120 2874 6f20 6669 7420 7468 6520  + 1 (to fit the 
+0000e950: 6e75 6d70 7920 6172 7261 790a 2020 2020  numpy array.    
+0000e960: 726f 7773 5f6e 616d 6573 203d 206d 6174  rows_names = mat
+0000e970: 7269 782e 696e 6465 782e 7661 6c75 6573  rix.index.values
+0000e980: 2e61 7374 7970 6528 2269 6e74 2229 2020  .astype("int")  
+0000e990: 2320 3832 0a20 2020 206d 6174 7269 7820  # 82.    matrix 
+0000e9a0: 3d20 6e70 2e63 6f6c 756d 6e5f 7374 6163  = np.column_stac
+0000e9b0: 6b28 2872 6f77 735f 6e61 6d65 732c 206d  k((rows_names, m
+0000e9c0: 6174 7269 7829 290a 2020 2020 6d61 7472  atrix)).    matr
+0000e9d0: 6978 203d 206e 702e 726f 775f 7374 6163  ix = np.row_stac
+0000e9e0: 6b28 2863 6f6c 756d 6e5f 6e61 6d65 732c  k((column_names,
+0000e9f0: 206d 6174 7269 7829 290a 2020 2020 7265   matrix)).    re
+0000ea00: 7475 726e 206d 6174 7269 780a 6465 6620  turn matrix.def 
+0000ea10: 636f 6e76 6572 745f 746f 5f6c 6574 7465  convert_to_lette
+0000ea20: 7273 2873 6571 2c61 615f 7072 6f62 7329  rs(seq,aa_probs)
+0000ea30: 3a0a 2020 2020 2222 2243 6f6e 7665 7274  :.    """Convert
+0000ea40: 7320 6261 636b 2074 6865 2069 6e74 6567  s back the integ
+0000ea50: 6572 7320 6173 7369 676e 6564 2074 6f20  ers assigned to 
+0000ea60: 7468 6520 616d 696e 6f20 6163 6964 7320  the amino acids 
+0000ea70: 746f 206c 6574 7465 7273 0a20 2020 203a  to letters.    :
+0000ea80: 7061 7261 6d20 7365 713a 2073 6571 0a20  param seq: seq. 
+0000ea90: 2020 203a 7061 7261 6d20 696e 7420 6161     :param int aa
+0000eaa0: 5f70 726f 6273 3a20 616d 696e 6f20 6163  _probs: amino ac
+0000eab0: 6964 2070 726f 6261 6269 6c69 7469 6573  id probabilities
+0000eac0: 2222 220a 0a20 2020 2061 615f 6e61 6d65  """..    aa_name
+0000ead0: 735f 6469 6374 203d 2061 6d69 6e6f 6163  s_dict = aminoac
+0000eae0: 6964 5f6e 616d 6573 5f64 6963 7428 6161  id_names_dict(aa
+0000eaf0: 5f70 726f 6273 290a 2020 2020 6161 5f6e  _probs).    aa_n
+0000eb00: 616d 6573 5f64 6963 745f 7265 7665 7273  ames_dict_revers
+0000eb10: 6520 3d20 7b76 616c 3a6b 6579 2066 6f72  e = {val:key for
+0000eb20: 206b 6579 2c76 616c 2069 6e20 6161 5f6e   key,val in aa_n
+0000eb30: 616d 6573 5f64 6963 742e 6974 656d 7328  ames_dict.items(
+0000eb40: 297d 0a20 2020 2069 6620 6e6f 7420 6973  )}.    if not is
+0000eb50: 696e 7374 616e 6365 2873 6571 5b30 5d2c  instance(seq[0],
+0000eb60: 666c 6f61 7429 3a0a 2020 2020 2020 2020  float):.        
+0000eb70: 7365 715f 6c65 7474 6572 7320 3d20 5b61  seq_letters = [a
+0000eb80: 615f 6e61 6d65 735f 6469 6374 5f72 6576  a_names_dict_rev
+0000eb90: 6572 7365 5b70 6f73 6974 696f 6e2e 6974  erse[position.it
+0000eba0: 656d 2829 5d20 666f 7220 706f 7369 7469  em()] for positi
+0000ebb0: 6f6e 2069 6e20 7365 7120 6966 2070 6f73  on in seq if pos
+0000ebc0: 6974 696f 6e2e 6974 656d 2829 2069 6e20  ition.item() in 
+0000ebd0: 6161 5f6e 616d 6573 5f64 6963 745f 7265  aa_names_dict_re
+0000ebe0: 7665 7273 655d 0a0a 2020 2020 656c 7365  verse]..    else
+0000ebf0: 3a0a 2020 2020 2020 2020 7365 715f 6c65  :.        seq_le
+0000ec00: 7474 6572 7320 3d20 5b61 615f 6e61 6d65  tters = [aa_name
+0000ec10: 735f 6469 6374 5f72 6576 6572 7365 5b70  s_dict_reverse[p
+0000ec20: 6f73 6974 696f 6e5d 2066 6f72 2070 6f73  osition] for pos
+0000ec30: 6974 696f 6e20 696e 2073 6571 2069 6620  ition in seq if 
+0000ec40: 706f 7369 7469 6f6e 2069 6e20 6161 5f6e  position in aa_n
+0000ec50: 616d 6573 5f64 6963 745f 7265 7665 7273  ames_dict_revers
+0000ec60: 655d 0a0a 2020 2020 7265 7475 726e 2027  e]..    return '
+0000ec70: 272e 6a6f 696e 2873 6571 5f6c 6574 7465  '.join(seq_lette
+0000ec80: 7273 290a 6465 6620 7363 6f72 655f 6d61  rs).def score_ma
+0000ec90: 7463 6828 7061 6972 2c20 6d61 7472 6978  tch(pair, matrix
+0000eca0: 293a 0a20 2020 2022 2222 5265 7475 726e  ):.    """Return
+0000ecb0: 7320 7468 6520 636f 7272 6573 706f 6e64  s the correspond
+0000ecc0: 696e 6720 626c 6f73 756d 2073 636f 7265  ing blosum score
+0000ecd0: 7320 6265 7477 6565 6e20 7468 6520 7061  s between the pa
+0000ece0: 6972 206f 6620 616d 696e 6f20 6163 6964  ir of amino acid
+0000ecf0: 730a 2020 2020 3a70 6172 616d 2074 7570  s.    :param tup
+0000ed00: 6c65 2070 6169 723a 2070 6169 7220 6f66  le pair: pair of
+0000ed10: 2061 6d69 6e6f 2061 6369 6473 2074 6f20   amino acids to 
+0000ed20: 636f 6d70 6172 650a 2020 2020 3a70 6172  compare.    :par
+0000ed30: 616d 206d 6174 7269 783a 2042 6c6f 7375  am matrix: Blosu
+0000ed40: 6d20 6d61 7472 6978 2022 2222 0a20 2020  m matrix """.   
+0000ed50: 2069 6620 7061 6972 206e 6f74 2069 6e20   if pair not in 
+0000ed60: 6d61 7472 6978 3a0a 2020 2020 2020 2020  matrix:.        
+0000ed70: 7265 7475 726e 206d 6174 7269 785b 2874  return matrix[(t
+0000ed80: 7570 6c65 2872 6576 6572 7365 6428 7061  uple(reversed(pa
+0000ed90: 6972 2929 295d 0a20 2020 2065 6c73 653a  ir)))].    else:
+0000eda0: 0a20 2020 2020 2020 2072 6574 7572 6e20  .        return 
+0000edb0: 6d61 7472 6978 5b70 6169 725d 0a64 6566  matrix[pair].def
+0000edc0: 2073 636f 7265 5f70 6169 7277 6973 6528   score_pairwise(
+0000edd0: 7365 7131 2c20 7365 7132 2c20 6d61 7472  seq1, seq2, matr
+0000ede0: 6978 2c20 6761 705f 732c 2067 6170 5f65  ix, gap_s, gap_e
+0000edf0: 293a 0a20 2020 2023 544f 444f 3a20 6874  ):.    #TODO: ht
+0000ee00: 7470 733a 2f2f 7374 6163 6b6f 7665 7266  tps://stackoverf
+0000ee10: 6c6f 772e 636f 6d2f 7175 6573 7469 6f6e  low.com/question
+0000ee20: 732f 3536 3836 3231 312f 6973 2d74 6865  s/5686211/is-the
+0000ee30: 7265 2d61 2d66 756e 6374 696f 6e2d 7468  re-a-function-th
+0000ee40: 6174 2d63 616e 2d63 616c 6375 6c61 7465  at-can-calculate
+0000ee50: 2d61 2d73 636f 7265 2d66 6f72 2d61 6c69  -a-score-for-ali
+0000ee60: 676e 6564 2d73 6571 7565 6e63 6573 2d67  gned-sequences-g
+0000ee70: 6976 656e 2d74 6865 2d61 0a20 2020 2022  iven-the-a.    "
+0000ee80: 2222 0a20 2020 2043 616c 6375 6c61 7465  "".    Calculate
+0000ee90: 7320 7468 6520 626c 6f73 756d 2073 636f  s the blosum sco
+0000eea0: 7265 206f 6620 7468 6520 7472 7565 2073  re of the true s
+0000eeb0: 6571 7565 6e63 6520 6167 6169 6e73 7420  equence against 
+0000eec0: 7468 6520 7072 6564 6963 7469 6f6e 730a  the predictions.
+0000eed0: 2020 2020 3a70 6172 616d 206d 6174 7269      :param matri
+0000eee0: 7820 6d61 7472 6978 3a20 426c 6f73 756d  x matrix: Blosum
+0000eef0: 206d 6174 7269 7820 636f 6e74 6169 6e69   matrix containi
+0000ef00: 6e67 2074 6865 206c 6f67 2d6f 6464 7320  ng the log-odds 
+0000ef10: 7363 6f72 6573 2028 6865 206c 6f67 6172  scores (he logar
+0000ef20: 6974 686d 2066 6f72 2074 6865 2072 6174  ithm for the rat
+0000ef30: 696f 206f 6620 7468 650a 2020 2020 2020  io of the.      
+0000ef40: 206c 696b 656c 6968 6f6f 6420 6f66 2074   likelihood of t
+0000ef50: 776f 2061 6d69 6e6f 2061 6369 6473 2061  wo amino acids a
+0000ef60: 7070 6561 7269 6e67 2077 6974 6820 6120  ppearing with a 
+0000ef70: 6269 6f6c 6f67 6963 616c 2073 656e 7365  biological sense
+0000ef80: 2061 6e64 2074 6865 206c 696b 656c 6968   and the likelih
+0000ef90: 6f6f 6420 6f66 2074 6865 2073 616d 6520  ood of the same 
+0000efa0: 616d 696e 6f20 6163 6964 7320 6170 7065  amino acids appe
+0000efb0: 6172 696e 6720 6279 2063 6861 6e63 6529  aring by chance)
+0000efc0: 0a20 2020 2020 2020 2028 7468 6520 6869  .        (the hi
+0000efd0: 6768 6572 2074 6865 2073 636f 7265 2c20  gher the score, 
+0000efe0: 7468 6520 6d6f 7265 206c 696b 656c 7920  the more likely 
+0000eff0: 7468 6520 636f 7272 6573 706f 6e64 696e  the correspondin
+0000f000: 6720 616d 696e 6f2d 6163 6964 2073 7562  g amino-acid sub
+0000f010: 7374 6974 7574 696f 6e20 6973 290a 2020  stitution is).  
+0000f020: 2020 3a70 6172 616d 2069 6e74 2067 6170    :param int gap
+0000f030: 5f73 3a20 6761 7020 7065 6e61 6c74 790a  _s: gap penalty.
+0000f040: 2020 2020 3a70 6172 616d 2069 6e74 2067      :param int g
+0000f050: 6170 5f65 203a 206d 6973 6d61 7463 6820  ap_e : mismatch 
+0000f060: 7065 6e61 6c74 790a 2020 2020 2222 220a  penalty.    """.
+0000f070: 2020 2020 7363 6f72 6520 3d20 300a 2020      score = 0.  
+0000f080: 2020 6761 7020 3d20 4661 6c73 650a 2020    gap = False.  
+0000f090: 2020 666f 7220 6920 696e 2072 616e 6765    for i in range
+0000f0a0: 286c 656e 2873 6571 3129 293a 0a20 2020  (len(seq1)):.   
+0000f0b0: 2020 2020 2070 6169 7220 3d20 2873 6571       pair = (seq
+0000f0c0: 315b 695d 2c20 7365 7132 5b69 5d29 0a20  1[i], seq2[i]). 
+0000f0d0: 2020 2020 2020 2069 6620 6e6f 7420 6761         if not ga
+0000f0e0: 703a 0a20 2020 2020 2020 2020 2020 2069  p:.            i
+0000f0f0: 6620 272d 2720 696e 2070 6169 723a 0a20  f '-' in pair:. 
+0000f100: 2020 2020 2020 2020 2020 2020 2020 2067                 g
+0000f110: 6170 203d 2054 7275 650a 2020 2020 2020  ap = True.      
+0000f120: 2020 2020 2020 2020 2020 7363 6f72 6520            score 
+0000f130: 2b3d 2067 6170 5f73 0a20 2020 2020 2020  += gap_s.       
+0000f140: 2020 2020 2023 2065 6c69 6620 222a 2220       # elif "*" 
+0000f150: 696e 2070 6169 723a 2023 544f 444f 3a20  in pair: #TODO: 
+0000f160: 4b65 6570 3f0a 2020 2020 2020 2020 2020  Keep?.          
+0000f170: 2020 2320 2020 2020 7363 6f72 6520 2b3d    #     score +=
+0000f180: 6761 705f 650a 2020 2020 2020 2020 2020  gap_e.          
+0000f190: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
+0000f1a0: 2020 2020 2020 2020 7363 6f72 6520 2b3d          score +=
+0000f1b0: 2073 636f 7265 5f6d 6174 6368 2870 6169   score_match(pai
+0000f1c0: 722c 206d 6174 7269 7829 0a20 2020 2020  r, matrix).     
+0000f1d0: 2020 2065 6c73 653a 0a20 2020 2020 2020     else:.       
+0000f1e0: 2020 2020 2069 6620 272d 2720 6e6f 7420       if '-' not 
+0000f1f0: 696e 2070 6169 723a 2320 616e 6420 222a  in pair:# and "*
+0000f200: 2220 6e6f 7420 696e 2070 6169 723a 0a20  " not in pair:. 
+0000f210: 2020 2020 2020 2020 2020 2020 2020 2067                 g
+0000f220: 6170 203d 2046 616c 7365 0a20 2020 2020  ap = False.     
+0000f230: 2020 2020 2020 2020 2020 2073 636f 7265             score
+0000f240: 202b 3d20 7363 6f72 655f 6d61 7463 6828   += score_match(
+0000f250: 7061 6972 2c20 6d61 7472 6978 290a 2020  pair, matrix).  
+0000f260: 2020 2020 2020 2020 2020 656c 7365 3a0a            else:.
+0000f270: 2020 2020 2020 2020 2020 2020 2020 2020                  
+0000f280: 7363 6f72 6520 2b3d 2067 6170 5f65 0a20  score += gap_e. 
+0000f290: 2020 2072 6574 7572 6e20 7363 6f72 650a     return score.
+0000f2a0: 6465 6620 7363 6f72 655f 7061 6972 7769  def score_pairwi
+0000f2b0: 7365 5f32 2873 6571 312c 2073 6571 322c  se_2(seq1, seq2,
+0000f2c0: 206d 6174 7269 782c 2067 6170 5f73 2c20   matrix, gap_s, 
+0000f2d0: 6761 705f 652c 2067 6170 203d 2054 7275  gap_e, gap = Tru
+0000f2e0: 6529 3a0a 2020 2020 2222 2241 6c74 6572  e):.    """Alter
+0000f2f0: 6e61 7469 7665 206d 6574 686f 6420 746f  native method to
+0000f300: 2063 616c 6375 6c61 7465 2074 6865 2062   calculate the b
+0000f310: 6c6f 7375 6d20 7363 6f72 6520 6265 7477  losum score betw
+0000f320: 6565 6e20 3220 7365 7175 656e 6365 7322  een 2 sequences"
+0000f330: 2222 0a20 2020 2066 6f72 2041 2c42 2069  "".    for A,B i
+0000f340: 6e20 7a69 7028 7365 7131 2c20 7365 7132  n zip(seq1, seq2
+0000f350: 293a 0a20 2020 2020 2020 2064 6961 6720  ):.        diag 
+0000f360: 3d20 2827 2d27 3d3d 4129 206f 7220 2827  = ('-'==A) or ('
+0000f370: 2d27 3d3d 4229 0a20 2020 2020 2020 2079  -'==B).        y
+0000f380: 6965 6c64 2028 6761 705f 6520 6966 2067  ield (gap_e if g
+0000f390: 6170 2065 6c73 6520 6761 705f 7329 2069  ap else gap_s) i
+0000f3a0: 6620 6469 6167 2065 6c73 6520 6d61 7472  f diag else matr
+0000f3b0: 6978 5b28 412c 4229 5d0a 2020 2020 2020  ix[(A,B)].      
+0000f3c0: 2020 6761 7020 3d20 6469 6167 0a64 6566    gap = diag.def
+0000f3d0: 206e 6f72 6d61 6c69 7a65 5f73 7461 6e64   normalize_stand
+0000f3e0: 6172 697a 6528 7829 3a0a 2020 2020 224e  arize(x):.    "N
+0000f3f0: 6f72 6d61 6c69 7a65 7320 616e 6420 7374  ormalizes and st
+0000f400: 616e 6461 7269 7a65 7320 616e 2069 6e70  andarizes an inp
+0000f410: 7574 206d 6174 7269 7822 0a20 2020 206e  ut matrix".    n
+0000f420: 6f72 6d20 3d20 6e70 2e6c 696e 616c 672e  orm = np.linalg.
+0000f430: 6e6f 726d 2878 290a 2020 2020 6e6f 726d  norm(x).    norm
+0000f440: 616c 5f61 7272 6179 203d 2078 202f 206e  al_array = x / n
+0000f450: 6f72 6d0a 2020 2020 7265 7475 726e 206e  orm.    return n
+0000f460: 6f72 6d61 6c5f 6172 7261 790a 6465 6620  ormal_array.def 
+0000f470: 666f 6c64 6572 7328 666f 6c64 6572 5f6e  folders(folder_n
+0000f480: 616d 652c 6261 7365 7061 7468 293a 0a20  ame,basepath):. 
+0000f490: 2020 2022 2222 2043 7265 6174 6573 2061     """ Creates a
+0000f4a0: 2066 6f6c 6465 7220 6174 2074 6865 2069   folder at the i
+0000f4b0: 6e64 6963 6174 6564 206c 6f63 6174 696f  ndicated locatio
+0000f4c0: 6e2e 2049 7420 7265 7772 6974 6573 2066  n. It rewrites f
+0000f4d0: 6f6c 6465 7273 2077 6974 6820 7468 6520  olders with the 
+0000f4e0: 7361 6d65 206e 616d 650a 2020 2020 3a70  same name.    :p
+0000f4f0: 6172 616d 2073 7472 2066 6f6c 6465 725f  aram str folder_
+0000f500: 6e61 6d65 3a20 6e61 6d65 206f 6620 7468  name: name of th
+0000f510: 6520 666f 6c64 6572 0a20 2020 203a 7061  e folder.    :pa
+0000f520: 7261 6d20 7374 7220 6261 7365 7061 7468  ram str basepath
+0000f530: 3a20 696e 6469 6361 7465 7320 7468 6520  : indicates the 
+0000f540: 706c 6163 6520 7768 6572 6520 746f 2063  place where to c
+0000f550: 7265 6174 6520 7468 6520 666f 6c64 6572  reate the folder
+0000f560: 0a20 2020 2022 2222 0a20 2020 2023 6261  .    """.    #ba
+0000f570: 7365 7061 7468 203d 206f 732e 6765 7463  sepath = os.getc
+0000f580: 7764 2829 0a0a 2020 2020 6966 206e 6f74  wd()..    if not
+0000f590: 2062 6173 6570 6174 683a 0a20 2020 2020   basepath:.     
+0000f5a0: 2020 206e 6577 7061 7468 203d 2066 6f6c     newpath = fol
+0000f5b0: 6465 725f 6e61 6d65 0a20 2020 2065 6c73  der_name.    els
+0000f5c0: 653a 0a20 2020 2020 2020 206e 6577 7061  e:.        newpa
+0000f5d0: 7468 203d 2062 6173 6570 6174 6820 2b20  th = basepath + 
+0000f5e0: 222f 2573 2220 2520 666f 6c64 6572 5f6e  "/%s" % folder_n
+0000f5f0: 616d 650a 0a20 2020 2069 6620 6e6f 7420  ame..    if not 
+0000f600: 6f73 2e70 6174 682e 6578 6973 7473 286e  os.path.exists(n
+0000f610: 6577 7061 7468 293a 0a0a 2020 2020 2020  ewpath):..      
+0000f620: 2020 7472 793a 0a20 2020 2020 2020 2020    try:.         
+0000f630: 2020 206f 7269 6769 6e61 6c5f 756d 6173     original_umas
+0000f640: 6b20 3d20 6f73 2e75 6d61 736b 2830 290a  k = os.umask(0).
+0000f650: 2020 2020 2020 2020 2020 2020 6f73 2e6d              os.m
+0000f660: 616b 6564 6972 7328 6e65 7770 6174 682c  akedirs(newpath,
+0000f670: 2030 6f37 3737 290a 2020 2020 2020 2020   0o777).        
+0000f680: 6669 6e61 6c6c 793a 0a20 2020 2020 2020  finally:.       
+0000f690: 2020 2020 206f 732e 756d 6173 6b28 6f72       os.umask(or
+0000f6a0: 6967 696e 616c 5f75 6d61 736b 290a 2020  iginal_umask).  
+0000f6b0: 2020 656c 7365 3a0a 2020 2020 2020 2020    else:.        
+0000f6c0: 7368 7574 696c 2e72 6d74 7265 6528 6e65  shutil.rmtree(ne
+0000f6d0: 7770 6174 6829 2020 2320 7265 6d6f 7665  wpath)  # remove
+0000f6e0: 7320 616c 6c20 7468 6520 7375 6264 6972  s all the subdir
+0000f6f0: 6563 746f 7269 6573 210a 2020 2020 2020  ectories!.      
+0000f700: 2020 6f73 2e6d 616b 6564 6972 7328 6e65    os.makedirs(ne
+0000f710: 7770 6174 682c 306f 3737 3729 0a64 6566  wpath,0o777).def
+0000f720: 2064 6976 6964 655f 6261 7463 6865 7328   divide_batches(
+0000f730: 4461 7461 7365 742c 6e75 6d62 6572 5f73  Dataset,number_s
+0000f740: 706c 6974 7329 3a0a 2020 2020 2222 2220  plits):.    """ 
+0000f750: 4469 7669 6465 7320 7468 6520 7472 6169  Divides the trai
+0000f760: 6e69 6e67 2064 6174 6173 6574 2069 6e20  ning dataset in 
+0000f770: 6120 6769 7665 6e20 6e75 6d62 6572 206f  a given number o
+0000f780: 6620 7370 6c69 7473 2222 220a 2020 2020  f splits""".    
+0000f790: 4461 7461 7365 7420 3d20 4461 7461 7365  Dataset = Datase
+0000f7a0: 745b 4461 7461 7365 745b 3a2c 302c 305d  t[Dataset[:,0,0]
+0000f7b0: 2e61 7267 736f 7274 2829 5d0a 2020 2020  .argsort()].    
+0000f7c0: 4461 7461 7365 745f 7370 6c69 7473 203d  Dataset_splits =
+0000f7d0: 6e70 2e61 7272 6179 5f73 706c 6974 2844  np.array_split(D
+0000f7e0: 6174 6173 6574 2c6e 756d 6265 725f 7370  ataset,number_sp
+0000f7f0: 6c69 7473 290a 2020 2020 7265 7475 726e  lits).    return
+0000f800: 2044 6174 6173 6574 5f73 706c 6974 730a   Dataset_splits.
+0000f810: 6465 6620 7065 7263 5f69 6465 6e74 6974  def perc_identit
+0000f820: 795f 616c 6967 6e6d 656e 7428 616c 6e29  y_alignment(aln)
+0000f830: 3a0a 2020 2020 2222 2243 616c 6375 6c61  :.    """Calcula
+0000f840: 7465 7320 7468 6520 2549 4420 6f66 2061  tes the %ID of a
+0000f850: 6e20 616c 6967 6e6d 656e 740a 2020 2020  n alignment.    
+0000f860: 3a70 6172 616d 2061 6c6e 3a20 6269 6f70  :param aln: biop
+0000f870: 7974 686f 6e20 616c 6967 6e6d 656e 7420  ython alignment 
+0000f880: 6f62 6a65 6374 2222 220a 2020 2020 6920  object""".    i 
+0000f890: 3d20 300a 2020 2020 666f 7220 6120 696e  = 0.    for a in
+0000f8a0: 2072 616e 6765 2830 2c6c 656e 2861 6c6e   range(0,len(aln
+0000f8b0: 5b30 5d29 293a 0a20 2020 2020 2020 2073  [0])):.        s
+0000f8c0: 203d 2061 6c6e 5b3a 2c61 5d0a 2020 2020   = aln[:,a].    
+0000f8d0: 2020 2020 6966 2073 203d 3d20 6c65 6e28      if s == len(
+0000f8e0: 7329 202a 2073 5b30 5d3a 0a20 2020 2020  s) * s[0]:.     
+0000f8f0: 2020 2020 2020 2069 202b 3d20 310a 2020         i += 1.  
+0000f900: 2020 7265 7475 726e 2031 3030 2a69 2f66    return 100*i/f
+0000f910: 6c6f 6174 286c 656e 2861 6c6e 5b30 5d29  loat(len(aln[0])
+0000f920: 290a 6465 6620 7065 7263 5f69 6465 6e74  ).def perc_ident
+0000f930: 6974 795f 7061 6972 5f73 6571 2873 6571  ity_pair_seq(seq
+0000f940: 312c 7365 7132 293a 0a20 2020 2022 2222  1,seq2):.    """
+0000f950: 4361 6c63 756c 6174 6573 2074 6865 2070  Calculates the p
+0000f960: 6572 6365 6e74 2069 6465 6e74 6974 7920  ercent identity 
+0000f970: 616d 6f6e 6720 6120 7061 6972 206f 6620  among a pair of 
+0000f980: 7365 7175 656e 6365 730a 2020 2020 3a70  sequences.    :p
+0000f990: 6172 616d 2073 7472 2073 6571 310a 2020  aram str seq1.  
+0000f9a0: 2020 3a70 6172 616d 2073 7472 2073 6571    :param str seq
+0000f9b0: 3222 2222 0a20 2020 2069 203d 2030 0a20  2""".    i = 0. 
+0000f9c0: 2020 2073 6571 3120 3d6c 6973 7428 7365     seq1 =list(se
+0000f9d0: 7131 290a 2020 2020 7365 7132 3d6c 6973  q1).    seq2=lis
+0000f9e0: 7428 7365 7132 290a 0a20 2020 2061 6c6e  t(seq2)..    aln
+0000f9f0: 203d 206e 702e 726f 775f 7374 6163 6b28   = np.row_stack(
+0000fa00: 5b73 6571 312c 7365 7132 5d29 0a20 2020  [seq1,seq2]).   
+0000fa10: 2066 6f72 2061 2069 6e20 7261 6e67 6528   for a in range(
+0000fa20: 302c 2061 6c6e 2e73 6861 7065 5b31 5d29  0, aln.shape[1])
+0000fa30: 3a0a 2020 2020 2020 2020 7320 3d20 616c  :.        s = al
+0000fa40: 6e5b 3a2c 615d 2e74 6f6c 6973 7428 290a  n[:,a].tolist().
+0000fa50: 2020 2020 2020 2020 7320 3d20 2222 2e6a          s = "".j
+0000fa60: 6f69 6e28 7329 0a20 2020 2020 2020 2069  oin(s).        i
+0000fa70: 6620 7320 3d3d 206c 656e 2873 2920 2a20  f s == len(s) * 
+0000fa80: 735b 305d 3a0a 2020 2020 2020 2020 2020  s[0]:.          
+0000fa90: 2020 6920 2b3d 2031 0a20 2020 2072 6574    i += 1.    ret
+0000faa0: 7572 6e20 3130 302a 692f 666c 6f61 7428  urn 100*i/float(
+0000fab0: 6c65 6e28 616c 6e5b 305d 2929 0a64 6566  len(aln[0])).def
+0000fac0: 2069 6e63 6f72 7265 6374 6c79 5f70 7265   incorrectly_pre
+0000fad0: 6469 6374 6564 5f61 6128 7365 7131 2c73  dicted_aa(seq1,s
+0000fae0: 6571 3229 3a0a 2020 2020 2222 2222 2222  eq2):.    """"""
+0000faf0: 0a20 2020 2069 203d 2030 0a20 2020 2073  .    i = 0.    s
+0000fb00: 6571 3120 3d6c 6973 7428 7365 7131 290a  eq1 =list(seq1).
+0000fb10: 2020 2020 7365 7132 3d6c 6973 7428 7365      seq2=list(se
+0000fb20: 7132 290a 0a20 2020 2061 6c6e 203d 206e  q2)..    aln = n
+0000fb30: 702e 726f 775f 7374 6163 6b28 5b73 6571  p.row_stack([seq
+0000fb40: 312c 7365 7132 5d29 0a20 2020 2066 6f72  1,seq2]).    for
+0000fb50: 2061 2069 6e20 7261 6e67 6528 302c 2061   a in range(0, a
+0000fb60: 6c6e 2e73 6861 7065 5b31 5d29 3a0a 2020  ln.shape[1]):.  
+0000fb70: 2020 2020 2020 7320 3d20 616c 6e5b 3a2c        s = aln[:,
+0000fb80: 615d 2e74 6f6c 6973 7428 290a 2020 2020  a].tolist().    
+0000fb90: 2020 2020 7320 3d20 2222 2e6a 6f69 6e28      s = "".join(
+0000fba0: 7329 0a20 2020 2020 2020 2069 6620 7320  s).        if s 
+0000fbb0: 3d3d 206c 656e 2873 2920 2a20 735b 305d  == len(s) * s[0]
+0000fbc0: 3a0a 2020 2020 2020 2020 2020 2020 6920  :.            i 
+0000fbd0: 2b3d 2031 2023 7361 6d65 2061 610a 2020  += 1 #same aa.  
+0000fbe0: 2020 7265 7475 726e 206c 656e 2861 6c6e    return len(aln
+0000fbf0: 5b30 5d29 2d69 0a64 6566 2074 6f5f 6574  [0])-i.def to_et
+0000fc00: 6533 2874 7265 6529 3a0a 2020 2020 696d  e3(tree):.    im
+0000fc10: 706f 7274 2074 656d 7066 696c 650a 2020  port tempfile.  
+0000fc20: 2020 6672 6f6d 2065 7465 3320 696d 706f    from ete3 impo
+0000fc30: 7274 2054 7265 6520 6173 2045 7465 5472  rt Tree as EteTr
+0000fc40: 6565 0a20 2020 2077 6974 6820 7465 6d70  ee.    with temp
+0000fc50: 6669 6c65 2e4e 616d 6564 5465 6d70 6f72  file.NamedTempor
+0000fc60: 6172 7946 696c 6528 6d6f 6465 3d22 7722  aryFile(mode="w"
+0000fc70: 2920 6173 2074 6d70 3a0a 2020 2020 2020  ) as tmp:.      
+0000fc80: 2020 5068 796c 6f2e 7772 6974 6528 7472    Phylo.write(tr
+0000fc90: 6565 2c20 746d 702c 2027 6e65 7769 636b  ee, tmp, 'newick
+0000fca0: 2729 0a20 2020 2020 2020 2074 6d70 2e66  ').        tmp.f
+0000fcb0: 6c75 7368 2829 0a20 2020 2020 2020 2072  lush().        r
+0000fcc0: 6574 7572 6e20 4574 6554 7265 6528 746d  eturn EteTree(tm
+0000fcd0: 702e 6e61 6d65 2c66 6f72 6d61 743d 3129  p.name,format=1)
+0000fce0: 0a64 6566 2074 6f5f 6269 6f70 7974 686f  .def to_biopytho
+0000fcf0: 6e54 7265 6528 7472 6565 293a 0a20 2020  nTree(tree):.   
+0000fd00: 2069 6d70 6f72 7420 7465 6d70 6669 6c65   import tempfile
+0000fd10: 0a20 2020 2066 726f 6d20 6574 6533 2069  .    from ete3 i
+0000fd20: 6d70 6f72 7420 5472 6565 2061 7320 4574  mport Tree as Et
+0000fd30: 6554 7265 650a 2020 2020 6672 6f6d 2042  eTree.    from B
+0000fd40: 696f 2069 6d70 6f72 7420 5068 796c 6f0a  io import Phylo.
+0000fd50: 2020 2020 7769 7468 2074 656d 7066 696c      with tempfil
+0000fd60: 652e 4e61 6d65 6454 656d 706f 7261 7279  e.NamedTemporary
+0000fd70: 4669 6c65 286d 6f64 653d 2277 2b22 2920  File(mode="w+") 
+0000fd80: 6173 2074 6d70 3a0a 2020 2020 2020 2020  as tmp:.        
+0000fd90: 7472 6565 2e77 7269 7465 286f 7574 6669  tree.write(outfi
+0000fda0: 6c65 3d74 6d70 2c66 6f72 6d61 743d 312c  le=tmp,format=1,
+0000fdb0: 666f 726d 6174 5f72 6f6f 745f 6e6f 6465  format_root_node
+0000fdc0: 3d54 7275 6529 0a20 2020 2020 2020 2074  =True).        t
+0000fdd0: 6d70 2e66 6c75 7368 2829 0a20 2020 2020  mp.flush().     
+0000fde0: 2020 2072 6574 7572 6e20 5068 796c 6f2e     return Phylo.
+0000fdf0: 7265 6164 2874 6d70 2e6e 616d 652c 2027  read(tmp.name, '
+0000fe00: 6e65 7769 636b 2729 0a64 6566 206c 6f61  newick').def loa
+0000fe10: 645f 6f62 6a28 6e61 6d65 293a 0a20 2020  d_obj(name):.   
+0000fe20: 2023 2077 6974 6820 6f70 656e 286e 616d   # with open(nam
+0000fe30: 652c 2027 7262 2729 2061 7320 663a 0a20  e, 'rb') as f:. 
+0000fe40: 2020 2023 2020 2020 2072 6574 7572 6e20     #     return 
+0000fe50: 7069 636b 6c65 2e6c 6f61 6428 6629 0a20  pickle.load(f). 
+0000fe60: 2020 2064 6174 6120 3d20 627a 322e 425a     data = bz2.BZ
+0000fe70: 3246 696c 6528 6e61 6d65 2c20 2272 6222  2File(name, "rb"
+0000fe80: 290a 2020 2020 6461 7461 203d 2063 5069  ).    data = cPi
+0000fe90: 636b 6c65 2e6c 6f61 6428 6461 7461 290a  ckle.load(data).
+0000fea0: 2020 2020 2320 7769 7468 206f 7065 6e28      # with open(
+0000feb0: 6e61 6d65 2c22 7262 2229 2061 7320 6a73  name,"rb") as js
+0000fec0: 6f6e 5f66 696c 653a 0a20 2020 2023 2020  on_file:.    #  
+0000fed0: 2020 2064 6174 6120 3d20 6a73 6f6e 2e6c     data = json.l
+0000fee0: 6f61 6428 6a73 6f6e 5f66 696c 652c 2065  oad(json_file, e
+0000fef0: 6e63 6f64 696e 673d 2275 7466 2d38 2229  ncoding="utf-8")
+0000ff00: 0a20 2020 2023 2064 6174 615f 6172 7261  .    # data_arra
+0000ff10: 7973 203d 207b 7d0a 2020 2020 2320 666f  ys = {}.    # fo
+0000ff20: 7220 6b65 792c 7661 6c75 6520 696e 2064  r key,value in d
+0000ff30: 6174 612e 6974 656d 7328 293a 0a20 2020  ata.items():.   
+0000ff40: 2023 2020 2020 2064 6174 615f 6172 7261   #     data_arra
+0000ff50: 7973 5b6b 6579 5d20 3d20 6e70 5f6a 6178  ys[key] = np_jax
+0000ff60: 2e61 7272 6179 2876 616c 7565 290a 2020  .array(value).  
+0000ff70: 2020 2320 7265 7475 726e 2064 6174 615f    # return data_
+0000ff80: 6172 7261 7973 0a20 2020 2072 6574 7572  arrays.    retur
+0000ff90: 6e20 6461 7461 0a64 6566 2073 6176 655f  n data.def save_
+0000ffa0: 6f62 6a28 6f62 6a2c 206e 616d 6529 3a0a  obj(obj, name):.
+0000ffb0: 2020 2020 2320 7769 7468 206f 7065 6e28      # with open(
+0000ffc0: 6e61 6d65 2c20 2777 6227 2920 6173 2066  name, 'wb') as f
+0000ffd0: 3a0a 2020 2020 2320 2020 2020 7069 636b  :.    #     pick
+0000ffe0: 6c65 2e64 756d 7028 6f62 6a2c 2066 2c20  le.dump(obj, f, 
+0000fff0: 7069 636b 6c65 2e48 4947 4845 5354 5f50  pickle.HIGHEST_P
+00010000: 524f 544f 434f 4c29 0a20 2020 2077 6974  ROTOCOL).    wit
+00010010: 6820 627a 322e 425a 3246 696c 6528 6e61  h bz2.BZ2File(na
+00010020: 6d65 2c20 2277 6222 2920 6173 2066 3a0a  me, "wb") as f:.
+00010030: 2020 2020 2020 2020 6350 6963 6b6c 652e          cPickle.
+00010040: 6475 6d70 286f 626a 2c20 6629 0a20 2020  dump(obj, f).   
+00010050: 2023 206f 626a 5f73 6572 6961 6c69 7a65   # obj_serialize
+00010060: 643d 7b7d 0a20 2020 2023 2066 6f72 206b  d={}.    # for k
+00010070: 6579 2c76 616c 7565 2069 6e20 6f62 6a2e  ey,value in obj.
+00010080: 6974 656d 7328 293a 0a20 2020 2023 2020  items():.    #  
+00010090: 2020 206f 626a 5f73 6572 6961 6c69 7a65     obj_serialize
+000100a0: 645b 6b65 795d 203d 2076 616c 7565 2e74  d[key] = value.t
+000100b0: 6f6c 6973 7428 290a 2020 2020 230a 2020  olist().    #.  
+000100c0: 2020 2320 7769 7468 206f 7065 6e28 6e61    # with open(na
+000100d0: 6d65 2c20 2777 272c 656e 636f 6469 6e67  me, 'w',encoding
+000100e0: 3d27 7574 662d 3827 2920 6173 206f 7574  ='utf-8') as out
+000100f0: 6669 6c65 3a0a 2020 2020 2320 2020 2020  file:.    #     
+00010100: 6a73 6f6e 2e64 756d 7028 6f62 6a5f 7365  json.dump(obj_se
+00010110: 7269 616c 697a 6564 2c20 6f75 7466 696c  rialized, outfil
+00010120: 652c 656e 7375 7265 5f61 7363 6969 3d46  e,ensure_ascii=F
+00010130: 616c 7365 2c20 696e 6465 6e74 3d34 290a  alse, indent=4).
+00010140: 6465 6620 6175 746f 6c61 6265 6c28 7265  def autolabel(re
+00010150: 6374 732c 2061 782c 2062 6c6f 7375 6d5f  cts, ax, blosum_
+00010160: 6469 6374 3d4e 6f6e 652c 7065 7263 656e  dict=None,percen
+00010170: 745f 6964 5f64 6963 743d 4e6f 6e65 2c62  t_id_dict=None,b
+00010180: 6c6f 7375 6d5f 7472 7565 5f64 6963 743d  losum_true_dict=
+00010190: 4e6f 6e65 293a 0a20 2020 2022 2222 4865  None):.    """He
+000101a0: 6c70 6572 2066 756e 6374 696f 6e20 746f  lper function to
+000101b0: 2061 7373 6967 6e20 6120 696e 7465 6765   assign a intege
+000101c0: 7220 7661 6c75 6520 6f6e 2074 6f70 206f  r value on top o
+000101d0: 6620 7468 6520 6869 7374 6f67 7261 6d20  f the histogram 
+000101e0: 6261 720a 2020 2020 3a70 6172 616d 2072  bar.    :param r
+000101f0: 6563 7473 3a20 6d61 7470 6c6f 746c 6962  ects: matplotlib
+00010200: 2072 6563 7461 6e67 6c65 730a 2020 2020   rectangles.    
+00010210: 3a70 6172 616d 2061 783a 206d 6174 706c  :param ax: matpl
+00010220: 6f74 6c69 6220 6178 6973 0a20 2020 203a  otlib axis.    :
+00010230: 7061 7261 6d20 626c 6f73 756d 5f64 6963  param blosum_dic
+00010240: 743a 2064 6963 7469 6f6e 6172 7920 636f  t: dictionary co
+00010250: 6e74 6169 6e69 6e67 2074 6865 2062 6c6f  ntaining the blo
+00010260: 7375 6d20 7363 6f72 6573 206f 6620 7468  sum scores of th
+00010270: 6520 7472 7565 2076 7320 7468 6520 7072  e true vs the pr
+00010280: 6564 6963 7469 6f6e 2073 6571 7565 6e63  ediction sequenc
+00010290: 6573 0a20 2020 203a 7061 7261 6d20 626c  es.    :param bl
+000102a0: 6f73 756d 5f74 7275 655f 6469 6374 3a20  osum_true_dict: 
+000102b0: 6469 6374 696f 6e61 7279 2063 6f6e 7461  dictionary conta
+000102c0: 696e 696e 6720 7468 6520 626c 6f73 756d  ining the blosum
+000102d0: 2073 636f 7265 7320 6f66 2074 6865 2074   scores of the t
+000102e0: 7275 6520 7673 2074 7275 6520 7365 7175  rue vs true sequ
+000102f0: 656e 6365 7322 2222 0a20 2020 2023 2047  ences""".    # G
+00010300: 6574 2079 2d61 7869 7320 6865 6967 6874  et y-axis height
+00010310: 2074 6f20 6361 6c63 756c 6174 6520 6c61   to calculate la
+00010320: 6265 6c20 706f 7369 7469 6f6e 2066 726f  bel position fro
+00010330: 6d2e 0a20 2020 2028 795f 626f 7474 6f6d  m..    (y_bottom
+00010340: 2c20 795f 746f 7029 203d 2061 782e 6765  , y_top) = ax.ge
+00010350: 745f 796c 696d 2829 0a20 2020 2079 5f68  t_ylim().    y_h
+00010360: 6569 6768 7420 3d20 795f 746f 7020 2d20  eight = y_top - 
+00010370: 795f 626f 7474 6f6d 0a20 2020 2069 6620  y_bottom.    if 
+00010380: 626c 6f73 756d 5f64 6963 7420 6973 206e  blosum_dict is n
+00010390: 6f74 204e 6f6e 653a 0a20 2020 2020 2020  ot None:.       
+000103a0: 2066 6f72 2072 6563 742c 626c 6f73 756d   for rect,blosum
+000103b0: 5f73 636f 7265 2c70 6964 2c62 6c6f 7375  _score,pid,blosu
+000103c0: 6d5f 7363 6f72 655f 7472 7565 2069 6e20  m_score_true in 
+000103d0: 7a69 7028 7265 6374 732c 626c 6f73 756d  zip(rects,blosum
+000103e0: 5f64 6963 742e 7661 6c75 6573 2829 2c70  _dict.values(),p
+000103f0: 6572 6365 6e74 5f69 645f 6469 6374 2e76  ercent_id_dict.v
+00010400: 616c 7565 7328 292c 626c 6f73 756d 5f74  alues(),blosum_t
+00010410: 7275 655f 6469 6374 2e76 616c 7565 7328  rue_dict.values(
+00010420: 2929 3a0a 2020 2020 2020 2020 2020 2020  )):.            
+00010430: 6865 6967 6874 203d 2072 6563 742e 6765  height = rect.ge
+00010440: 745f 6865 6967 6874 2829 0a20 2020 2020  t_height().     
+00010450: 2020 2020 2020 2023 2046 7261 6374 696f         # Fractio
+00010460: 6e20 6f66 2061 7869 7320 6865 6967 6874  n of axis height
+00010470: 2074 616b 656e 2075 7020 6279 2074 6869   taken up by thi
+00010480: 7320 7265 6374 616e 676c 650a 2020 2020  s rectangle.    
+00010490: 2020 2020 2020 2020 705f 6865 6967 6874          p_height
+000104a0: 203d 2028 6865 6967 6874 202f 2079 5f68   = (height / y_h
+000104b0: 6569 6768 7429 0a20 2020 2020 2020 2020  eight).         
+000104c0: 2020 2023 2049 6620 7765 2063 616e 2066     # If we can f
+000104d0: 6974 2074 6865 206c 6162 656c 2061 626f  it the label abo
+000104e0: 7665 2074 6865 2063 6f6c 756d 6e2c 2064  ve the column, d
+000104f0: 6f20 7468 6174 3b0a 2020 2020 2020 2020  o that;.        
+00010500: 2020 2020 2320 6f74 6865 7277 6973 652c      # otherwise,
+00010510: 2070 7574 2069 7420 696e 7369 6465 2074   put it inside t
+00010520: 6865 2063 6f6c 756d 6e2e 0a20 2020 2020  he column..     
+00010530: 2020 2020 2020 2069 6620 705f 6865 6967         if p_heig
+00010540: 6874 203e 2030 2e39 353a 2020 2320 6172  ht > 0.95:  # ar
+00010550: 6269 7472 6172 793b 2039 3525 206c 6f6f  bitrary; 95% loo
+00010560: 6b65 6420 676f 6f64 2074 6f20 6d65 2e0a  ked good to me..
+00010570: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00010580: 6c61 6265 6c5f 706f 7369 7469 6f6e 203d  label_position =
+00010590: 2068 6569 6768 7420 2d20 2879 5f68 6569   height - (y_hei
+000105a0: 6768 7420 2a20 302e 3035 290a 2020 2020  ght * 0.05).    
+000105b0: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
+000105c0: 2020 2020 2020 2020 2020 2020 2020 6c61                la
+000105d0: 6265 6c5f 706f 7369 7469 6f6e 203d 2068  bel_position = h
+000105e0: 6569 6768 7420 2b20 2879 5f68 6569 6768  eight + (y_heigh
+000105f0: 7420 2a20 302e 3031 290a 0a20 2020 2020  t * 0.01)..     
+00010600: 2020 2020 2020 2061 782e 7465 7874 2872         ax.text(r
+00010610: 6563 742e 6765 745f 7828 2920 2b20 7265  ect.get_x() + re
+00010620: 6374 2e67 6574 5f77 6964 7468 2829 202f  ct.get_width() /
+00010630: 2032 2e2c 206c 6162 656c 5f70 6f73 6974   2., label_posit
+00010640: 696f 6e2c 0a20 2020 2020 2020 2020 2020  ion,.           
+00010650: 2020 2020 2020 2020 2027 496e 4141 3a25           'InAA:%
+00010660: 6420 5c6e 2042 733a 2564 207c 2042 7354  d \n Bs:%d | BsT
+00010670: 7275 653a 2564 205c 6e20 5049 443a 2531  rue:%d \n PID:%1
+00010680: 6627 2025 2028 696e 7428 6865 6967 6874  f' % (int(height
+00010690: 292c 626c 6f73 756d 5f73 636f 7265 2c62  ),blosum_score,b
+000106a0: 6c6f 7375 6d5f 7363 6f72 655f 7472 7565  losum_score_true
+000106b0: 2c70 6964 292c 0a20 2020 2020 2020 2020  ,pid),.         
+000106c0: 2020 2020 2020 2020 2020 2068 613d 2763             ha='c
+000106d0: 656e 7465 7227 2c20 7661 3d27 626f 7474  enter', va='bott
+000106e0: 6f6d 2729 0a20 2020 2065 6c73 653a 0a20  om').    else:. 
+000106f0: 2020 2020 2020 2066 6f72 2072 6563 7420         for rect 
+00010700: 696e 2072 6563 7473 3a0a 2020 2020 2020  in rects:.      
+00010710: 2020 2020 2020 6865 6967 6874 203d 2072        height = r
+00010720: 6563 742e 6765 745f 6865 6967 6874 2829  ect.get_height()
+00010730: 0a20 2020 2020 2020 2020 2020 2023 2046  .            # F
+00010740: 7261 6374 696f 6e20 6f66 2061 7869 7320  raction of axis 
+00010750: 6865 6967 6874 2074 616b 656e 2075 7020  height taken up 
+00010760: 6279 2074 6869 7320 7265 6374 616e 676c  by this rectangl
+00010770: 650a 2020 2020 2020 2020 2020 2020 705f  e.            p_
+00010780: 6865 6967 6874 203d 2028 6865 6967 6874  height = (height
+00010790: 202f 2079 5f68 6569 6768 7429 0a20 2020   / y_height).   
+000107a0: 2020 2020 2020 2020 2023 2049 6620 7765           # If we
+000107b0: 2063 616e 2066 6974 2074 6865 206c 6162   can fit the lab
+000107c0: 656c 2061 626f 7665 2074 6865 2063 6f6c  el above the col
+000107d0: 756d 6e2c 2064 6f20 7468 6174 3b0a 2020  umn, do that;.  
+000107e0: 2020 2020 2020 2020 2020 2320 6f74 6865            # othe
+000107f0: 7277 6973 652c 2070 7574 2069 7420 696e  rwise, put it in
+00010800: 7369 6465 2074 6865 2063 6f6c 756d 6e2e  side the column.
+00010810: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
+00010820: 705f 6865 6967 6874 203e 2030 2e39 353a  p_height > 0.95:
+00010830: 2020 2320 6172 6269 7472 6172 793b 2039    # arbitrary; 9
+00010840: 3525 206c 6f6f 6b65 6420 676f 6f64 2074  5% looked good t
+00010850: 6f20 6d65 2e0a 2020 2020 2020 2020 2020  o me..          
+00010860: 2020 2020 2020 6c61 6265 6c5f 706f 7369        label_posi
+00010870: 7469 6f6e 203d 2068 6569 6768 7420 2d20  tion = height - 
+00010880: 2879 5f68 6569 6768 7420 2a20 302e 3035  (y_height * 0.05
+00010890: 290a 2020 2020 2020 2020 2020 2020 656c  ).            el
+000108a0: 7365 3a0a 2020 2020 2020 2020 2020 2020  se:.            
+000108b0: 2020 2020 6c61 6265 6c5f 706f 7369 7469      label_positi
+000108c0: 6f6e 203d 2068 6569 6768 7420 2b20 2879  on = height + (y
+000108d0: 5f68 6569 6768 7420 2a20 302e 3031 290a  _height * 0.01).
+000108e0: 0a20 2020 2020 2020 2020 2020 2061 782e  .            ax.
+000108f0: 7465 7874 2872 6563 742e 6765 745f 7828  text(rect.get_x(
+00010900: 2920 2b20 7265 6374 2e67 6574 5f77 6964  ) + rect.get_wid
+00010910: 7468 2829 202f 2032 2e2c 206c 6162 656c  th() / 2., label
+00010920: 5f70 6f73 6974 696f 6e2c 0a20 2020 2020  _position,.     
+00010930: 2020 2020 2020 2020 2020 2020 2020 2027                 '
+00010940: 2564 2720 2520 696e 7428 6865 6967 6874  %d' % int(height
+00010950: 292c 0a20 2020 2020 2020 2020 2020 2020  ),.             
+00010960: 2020 2020 2020 2068 613d 2763 656e 7465         ha='cente
+00010970: 7227 2c20 7661 3d27 626f 7474 6f6d 2729  r', va='bottom')
+00010980: 0a64 6566 2073 796d 6d65 7472 697a 6528  .def symmetrize(
+00010990: 6129 3a0a 2020 2020 224d 616b 6520 6120  a):.    "Make a 
+000109a0: 7472 6961 6e67 756c 6172 2075 7070 6572  triangular upper
+000109b0: 206d 6174 7269 7820 7379 6d6d 6574 7269   matrix symmetri
+000109c0: 6320 616c 6f6e 6720 7468 6520 6469 6167  c along the diag
+000109d0: 6f6e 616c 2e20 5661 6c69 6420 666f 7220  onal. Valid for 
+000109e0: 6e75 6d70 7920 6172 7261 7973 2061 6e64  numpy arrays and
+000109f0: 2070 616e 6461 7320 6461 7461 6672 616d   pandas datafram
+00010a00: 6573 220a 2020 2020 6120 3d20 6e70 2e6d  es".    a = np.m
+00010a10: 6178 696d 756d 2861 2c20 612e 7472 616e  aximum(a, a.tran
+00010a20: 7370 6f73 6528 2929 0a20 2020 2072 6574  spose()).    ret
+00010a30: 7572 6e20 610a 6465 6620 6275 696c 645f  urn a.def build_
+00010a40: 7072 6564 6963 7465 645f 7472 6565 2869  predicted_tree(i
+00010a50: 6e64 6578 2c73 616d 706c 6564 5f73 6571  ndex,sampled_seq
+00010a60: 7565 6e63 6573 2c6c 6561 665f 6e61 6d65  uences,leaf_name
+00010a70: 732c 6e61 6d65 2c72 6573 756c 7473 5f64  s,name,results_d
+00010a80: 6972 6563 746f 7279 2c6d 6574 686f 6429  irectory,method)
+00010a90: 3a20 2354 4f44 4f3a 2043 6f72 7265 6374  : #TODO: Correct
+00010aa0: 0a20 2020 2022 2222 4275 696c 6473 2061  .    """Builds a
+00010ab0: 2074 7265 6520 6265 7477 6565 6e20 7468   tree between th
+00010ac0: 6520 7472 6169 6e20 286c 6561 7665 7329  e train (leaves)
+00010ad0: 2064 6174 6120 616e 6420 7468 6520 7072   data and the pr
+00010ae0: 6564 6963 7465 6420 616e 6365 7374 6f72  edicted ancestor
+00010af0: 730a 2020 2020 3a70 6172 616d 2069 6e64  s.    :param ind
+00010b00: 6578 3a20 6e6f 6465 2069 6e64 6578 0a20  ex: node index. 
+00010b10: 2020 203a 7061 7261 6d20 7361 6d70 6c65     :param sample
+00010b20: 645f 7365 7175 656e 6365 733a 2064 6174  d_sequences: dat
+00010b30: 6173 6574 5f74 6573 7420 7769 7468 2073  aset_test with s
+00010b40: 616d 706c 6564 2073 6571 7565 6e63 6573  ampled sequences
+00010b50: 0a20 2020 203a 7061 7261 6d20 6c65 6166  .    :param leaf
+00010b60: 5f6e 616d 6573 3a20 6c65 6166 206e 6f64  _names: leaf nod
+00010b70: 6520 6e61 6d65 730a 2020 2020 3a70 6172  e names.    :par
+00010b80: 616d 206e 616d 653a 2064 6174 6173 6574  am name: dataset
+00010b90: 7072 6f6a 6563 7420 6e61 6d65 0a20 2020  project name.   
+00010ba0: 203a 7061 7261 6d20 7265 7375 6c74 735f   :param results_
+00010bb0: 6469 7265 6374 6f72 790a 2020 2020 3a70  directory.    :p
+00010bc0: 6172 616d 206d 6574 686f 643a 2074 7265  aram method: tre
+00010bd0: 6520 696e 6665 7265 6e63 6520 656e 6769  e inference engi
+00010be0: 6e65 2222 220a 2020 2020 7472 6169 6e69  ne""".    traini
+00010bf0: 6e67 5f73 6571 7320 3d20 5365 7149 4f2e  ng_seqs = SeqIO.
+00010c00: 7061 7273 6528 227b 7d2f 7b7d 5f74 7261  parse("{}/{}_tra
+00010c10: 696e 696e 672e 6661 7374 6122 2e66 6f72  ining.fasta".for
+00010c20: 6d61 7428 7265 7375 6c74 735f 6469 7265  mat(results_dire
+00010c30: 6374 6f72 792c 206e 616d 6529 2c20 2266  ctory, name), "f
+00010c40: 6173 7461 2229 0a20 2020 2077 6974 6820  asta").    with 
+00010c50: 6f70 656e 2822 7b7d 2f54 7265 655f 416c  open("{}/Tree_Al
+00010c60: 6967 6e6d 656e 745f 5361 6d70 6c65 642f  ignment_Sampled/
+00010c70: 7b7d 5f63 6f6d 6269 6e65 645f 7361 6d70  {}_combined_samp
+00010c80: 6c65 5f69 6e64 6578 5f7b 7d2e 6661 7374  le_index_{}.fast
+00010c90: 6122 2e66 6f72 6d61 7428 7265 7375 6c74  a".format(result
+00010ca0: 735f 6469 7265 6374 6f72 792c 206e 616d  s_directory, nam
+00010cb0: 652c 696e 6465 7829 2c20 2777 2729 2061  e,index), 'w') a
+00010cc0: 7320 775f 6669 6c65 3a0a 2020 2020 2020  s w_file:.      
+00010cd0: 2020 5365 7149 4f2e 7772 6974 6528 7472    SeqIO.write(tr
+00010ce0: 6169 6e69 6e67 5f73 6571 732c 775f 6669  aining_seqs,w_fi
+00010cf0: 6c65 2c22 6661 7374 6122 290a 2020 2020  le,"fasta").    
+00010d00: 2020 2020 7265 636f 7264 733d 5b5d 0a20      records=[]. 
+00010d10: 2020 2020 2020 2066 6f72 2073 616d 706c         for sampl
+00010d20: 6564 5f73 6571 2c6c 6561 665f 6e61 6d65  ed_seq,leaf_name
+00010d30: 2069 6e20 7a69 7028 7361 6d70 6c65 645f   in zip(sampled_
+00010d40: 7365 7175 656e 6365 732c 6c65 6166 5f6e  sequences,leaf_n
+00010d50: 616d 6573 293a 0a20 2020 2020 2020 2020  ames):.         
+00010d60: 2020 2072 6563 6f72 6420 3d20 5365 7152     record = SeqR
+00010d70: 6563 6f72 642e 5365 7152 6563 6f72 6428  ecord.SeqRecord(
+00010d80: 5365 7128 2727 2e6a 6f69 6e28 7361 6d70  Seq(''.join(samp
+00010d90: 6c65 645f 7365 7129 2e72 6570 6c61 6365  led_seq).replace
+00010da0: 2822 2d22 2c20 2222 2929 2c61 6e6e 6f74  ("-", "")),annot
+00010db0: 6174 696f 6e73 3d7b 226d 6f6c 6563 756c  ations={"molecul
+00010dc0: 655f 7479 7065 223a 2022 7072 6f74 6569  e_type": "protei
+00010dd0: 6e22 7d2c 6964 3d6c 6561 665f 6e61 6d65  n"},id=leaf_name
+00010de0: 2c20 6465 7363 7269 7074 696f 6e3d 2222  , description=""
+00010df0: 290a 2020 2020 2020 2020 2020 2020 7265  ).            re
+00010e00: 636f 7264 732e 6170 7065 6e64 2872 6563  cords.append(rec
+00010e10: 6f72 6429 0a20 2020 2020 2020 2053 6571  ord).        Seq
+00010e20: 494f 2e77 7269 7465 2872 6563 6f72 6473  IO.write(records
+00010e30: 2c77 5f66 696c 652c 2266 6173 7461 2229  ,w_file,"fasta")
+00010e40: 0a20 2020 2064 6963 745f 616c 6967 6e2c  .    dict_align,
+00010e50: 616c 6967 6e6d 656e 7420 3d20 696e 6665  alignment = infe
+00010e60: 725f 616c 6967 6e6d 656e 7428 696e 7075  r_alignment(inpu
+00010e70: 745f 6e61 6d65 5f66 696c 653d 227b 7d2f  t_name_file="{}/
+00010e80: 5472 6565 5f41 6c69 676e 6d65 6e74 5f53  Tree_Alignment_S
+00010e90: 616d 706c 6564 2f7b 7d5f 636f 6d62 696e  ampled/{}_combin
+00010ea0: 6564 5f73 616d 706c 655f 696e 6465 785f  ed_sample_index_
+00010eb0: 7b7d 2e66 6173 7461 222e 666f 726d 6174  {}.fasta".format
+00010ec0: 2872 6573 756c 7473 5f64 6972 6563 746f  (results_directo
+00010ed0: 7279 2c20 6e61 6d65 2c69 6e64 6578 292c  ry, name,index),
+00010ee0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
 00010ef0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00010f00: 2020 2020 2020 2020 2020 206f 7574 7075             outpu
-00010f10: 745f 6e61 6d65 5f66 696c 653d 227b 7d2f  t_name_file="{}/
-00010f20: 5472 6565 5f41 6c69 676e 6d65 6e74 5f53  Tree_Alignment_S
-00010f30: 616d 706c 6564 2f7b 7d5f 636f 6d62 696e  ampled/{}_combin
-00010f40: 6564 5f73 616d 706c 655f 696e 6465 785f  ed_sample_index_
-00010f50: 7b7d 2e6d 6166 6674 222e 666f 726d 6174  {}.mafft".format
-00010f60: 2872 6573 756c 7473 5f64 6972 6563 746f  (results_directo
-00010f70: 7279 2c6e 616d 652c 696e 6465 7829 2c61  ry,name,index),a
-00010f80: 6c69 676e 6d65 6e74 5f66 696c 653d 4e6f  lignment_file=No
-00010f90: 6e65 290a 0a20 2020 2023 616c 6967 6e6d  ne)..    #alignm
-00010fa0: 656e 742c 2061 6c69 676e 6d65 6e74 5f66  ent, alignment_f
-00010fb0: 696c 655f 6e61 6d65 2c6e 616d 652c 6d65  ile_name,name,me
-00010fc0: 7468 6f64 3d4e 6f6e 652c 7472 6565 5f66  thod=None,tree_f
-00010fd0: 696c 655f 6e61 6d65 3d4e 6f6e 652c 7472  ile_name=None,tr
-00010fe0: 6565 5f66 696c 653d 4e6f 6e65 2c73 746f  ee_file=None,sto
-00010ff0: 7261 6765 5f66 6f6c 6465 723d 2222 0a20  rage_folder="". 
-00011000: 2020 2074 7265 6520 3d20 696e 6665 725f     tree = infer_
-00011010: 7472 6565 2861 6c69 676e 6d65 6e74 3d61  tree(alignment=a
-00011020: 6c69 676e 6d65 6e74 2c0a 2020 2020 2020  lignment,.      
+00010f00: 2020 2020 2020 2020 2020 2020 6f75 7470              outp
+00010f10: 7574 5f6e 616d 655f 6669 6c65 3d22 7b7d  ut_name_file="{}
+00010f20: 2f54 7265 655f 416c 6967 6e6d 656e 745f  /Tree_Alignment_
+00010f30: 5361 6d70 6c65 642f 7b7d 5f63 6f6d 6269  Sampled/{}_combi
+00010f40: 6e65 645f 7361 6d70 6c65 5f69 6e64 6578  ned_sample_index
+00010f50: 5f7b 7d2e 6d61 6666 7422 2e66 6f72 6d61  _{}.mafft".forma
+00010f60: 7428 7265 7375 6c74 735f 6469 7265 6374  t(results_direct
+00010f70: 6f72 792c 6e61 6d65 2c69 6e64 6578 292c  ory,name,index),
+00010f80: 616c 6967 6e6d 656e 745f 6669 6c65 3d4e  alignment_file=N
+00010f90: 6f6e 6529 0a0a 2020 2020 2361 6c69 676e  one)..    #align
+00010fa0: 6d65 6e74 2c20 616c 6967 6e6d 656e 745f  ment, alignment_
+00010fb0: 6669 6c65 5f6e 616d 652c 6e61 6d65 2c6d  file_name,name,m
+00010fc0: 6574 686f 643d 4e6f 6e65 2c74 7265 655f  ethod=None,tree_
+00010fd0: 6669 6c65 5f6e 616d 653d 4e6f 6e65 2c74  file_name=None,t
+00010fe0: 7265 655f 6669 6c65 3d4e 6f6e 652c 7374  ree_file=None,st
+00010ff0: 6f72 6167 655f 666f 6c64 6572 3d22 220a  orage_folder="".
+00011000: 2020 2020 7472 6565 203d 2069 6e66 6572      tree = infer
+00011010: 5f74 7265 6528 616c 6967 6e6d 656e 743d  _tree(alignment=
+00011020: 616c 6967 6e6d 656e 742c 0a20 2020 2020  alignment,.     
 00011030: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011040: 616c 6967 6e6d 656e 745f 6669 6c65 5f6e  alignment_file_n
-00011050: 616d 653d 227b 7d2f 5472 6565 5f41 6c69  ame="{}/Tree_Ali
-00011060: 676e 6d65 6e74 5f53 616d 706c 6564 2f7b  gnment_Sampled/{
-00011070: 7d5f 636f 6d62 696e 6564 5f73 616d 706c  }_combined_sampl
-00011080: 655f 696e 6465 785f 7b7d 2e6d 6166 6674  e_index_{}.mafft
-00011090: 222e 666f 726d 6174 2872 6573 756c 7473  ".format(results
-000110a0: 5f64 6972 6563 746f 7279 2c6e 616d 652c  _directory,name,
-000110b0: 696e 6465 7829 2c0a 2020 2020 2020 2020  index),.        
-000110c0: 2020 2020 2020 2020 2020 2020 2020 7472                tr
-000110d0: 6565 5f66 696c 655f 6e61 6d65 3d22 7b7d  ee_file_name="{}
-000110e0: 2f54 7265 655f 416c 6967 6e6d 656e 745f  /Tree_Alignment_
-000110f0: 5361 6d70 6c65 642f 7b7d 5f63 6f6d 6269  Sampled/{}_combi
-00011100: 6e65 645f 7361 6d70 6c65 5f69 6e64 6578  ned_sample_index
-00011110: 5f7b 7d2e 7472 6565 222e 666f 726d 6174  _{}.tree".format
-00011120: 2872 6573 756c 7473 5f64 6972 6563 746f  (results_directo
-00011130: 7279 2c6e 616d 652c 696e 6465 7829 2c0a  ry,name,index),.
-00011140: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00011150: 2020 2020 2020 6d65 7468 6f64 3d6d 6574        method=met
-00011160: 686f 642c 0a20 2020 2020 2020 2020 2020  hod,.           
-00011170: 2020 2020 2020 2020 2020 206e 616d 653d             name=
-00011180: 227b 7d5f 636f 6d62 696e 6564 5f73 616d  "{}_combined_sam
-00011190: 706c 655f 696e 6465 785f 7b7d 2e74 7265  ple_index_{}.tre
-000111a0: 6522 2e66 6f72 6d61 7428 6e61 6d65 2c69  e".format(name,i
-000111b0: 6e64 6578 2929 0a20 2020 2072 6574 7572  ndex)).    retur
-000111c0: 6e20 7472 6565 0a64 6566 2068 6561 746d  n tree.def heatm
-000111d0: 6170 7328 7072 6564 6963 7469 6f6e 735f  aps(predictions_
-000111e0: 7361 6d70 6c65 732c 6461 7461 7365 745f  samples,dataset_
-000111f0: 7465 7374 2c6e 616d 652c 206e 756d 5f73  test,name, num_s
-00011200: 616d 706c 6573 2c6e 6f64 6573 5f69 6e64  amples,nodes_ind
-00011210: 6578 6573 2c72 6573 756c 7473 5f64 6972  exes,results_dir
-00011220: 6563 746f 7279 2c61 615f 7072 6f62 732c  ectory,aa_probs,
-00011230: 6164 6469 7469 6f6e 616c 5f6c 6f61 642c  additional_load,
-00011240: 6164 6469 7469 6f6e 616c 5f69 6e66 6f2c  additional_info,
-00011250: 636f 7272 6573 706f 6e64 656e 6365 5f64  correspondence_d
-00011260: 6963 743d 4e6f 6e65 293a 0a20 2020 2022  ict=None):.    "
-00011270: 2222 4275 696c 6420 6865 6174 6d61 7073  ""Build heatmaps
-00011280: 2062 6574 7765 656e 2074 6865 2070 6572   between the per
-00011290: 6365 6e74 2069 6465 6e74 6974 7920 7472  cent identity tr
-000112a0: 7565 2073 6571 7565 6e63 6573 2028 6c65  ue sequences (le
-000112b0: 6176 6573 206f 7220 696e 7465 726e 616c  aves or internal
-000112c0: 206e 6f64 6573 2920 616e 6420 7468 6520   nodes) and the 
-000112d0: 7072 6564 6963 7465 6420 7365 7175 656e  predicted sequen
-000112e0: 6365 730a 2020 2020 3a70 6172 616d 2074  ces.    :param t
-000112f0: 656e 736f 7220 7072 6564 6963 7469 6f6e  ensor prediction
-00011300: 735f 7361 6d70 6c65 730a 2020 2020 3a70  s_samples.    :p
-00011310: 6172 616d 2064 6174 6173 6574 5f74 6573  aram dataset_tes
-00011320: 740a 2020 2020 3a70 6172 616d 2073 7472  t.    :param str
-00011330: 206e 616d 653a 2064 6174 6173 6574 2070   name: dataset p
-00011340: 726f 6a65 6374 206e 616d 650a 2020 2020  roject name.    
-00011350: 3a70 6172 616d 2069 6e74 206e 756d 5f73  :param int num_s
-00011360: 616d 706c 6573 0a20 2020 203a 7061 7261  amples.    :para
-00011370: 6d20 7465 6e73 6f72 206e 6f64 6573 5f69  m tensor nodes_i
-00011380: 6e64 6578 6573 0a20 2020 203a 7061 7261  ndexes.    :para
-00011390: 6d20 7374 7220 7265 7375 6c74 735f 6469  m str results_di
-000113a0: 7265 6374 6f72 790a 2020 2020 3a70 6172  rectory.    :par
-000113b0: 616d 2069 6e74 2061 615f 7072 6f62 730a  am int aa_probs.
-000113c0: 2020 2020 3a70 6172 616d 206e 616d 6564      :param named
-000113d0: 7475 706c 6520 6164 6469 7469 6f6e 616c  tuple additional
-000113e0: 5f6c 6f61 640a 2020 2020 3a70 6172 616d  _load.    :param
-000113f0: 206e 616d 6564 7475 706c 6520 6164 6469   namedtuple addi
-00011400: 7469 6f6e 616c 5f69 6e66 6f0a 2020 2020  tional_info.    
-00011410: 3a70 6172 616d 2063 6f72 7265 7370 6f6e  :param correspon
-00011420: 6465 6e63 655f 6469 6374 2222 220a 2020  dence_dict""".  
-00011430: 2020 7072 696e 7428 2242 7569 6c64 696e    print("Buildin
-00011440: 6720 2549 4420 616e 6420 426c 6f73 756d  g %ID and Blosum
-00011450: 2068 6561 746d 6170 7320 2620 496e 636f   heatmaps & Inco
-00011460: 7272 6563 7420 6161 2068 6973 746f 6772  rrect aa histogr
-00011470: 616d 2e2e 2e22 290a 2020 2020 2362 6c6f  am...").    #blo
-00011480: 7375 6d20 3d20 4d61 7472 6978 496e 666f  sum = MatrixInfo
-00011490: 2e62 6c6f 7375 6d36 320a 2020 2020 626c  .blosum62.    bl
-000114a0: 6f73 756d 203d 2061 6464 6974 696f 6e61  osum = additiona
-000114b0: 6c5f 696e 666f 2e62 6c6f 7375 6d5f 6469  l_info.blosum_di
-000114c0: 6374 0a20 2020 206e 6f64 6573 5f69 6e64  ct.    nodes_ind
-000114d0: 6578 6573 203d 206e 6f64 6573 5f69 6e64  exes = nodes_ind
-000114e0: 6578 6573 2e74 6f6c 6973 7428 290a 2020  exes.tolist().  
-000114f0: 2020 236e 6f64 655f 6e61 6d65 7320 3d20    #node_names = 
-00011500: 6461 7461 7365 745f 7465 7374 5b3a 2c20  dataset_test[:, 
-00011510: 302c 2031 5d0a 2020 2020 666f 6c64 6572  0, 1].    folder
-00011520: 203d 206f 732e 7061 7468 2e62 6173 656e   = os.path.basen
-00011530: 616d 6528 7265 7375 6c74 735f 6469 7265  ame(results_dire
-00011540: 6374 6f72 7929 2e73 706c 6974 2822 5f22  ctory).split("_"
-00011550: 295b 305d 0a20 2020 2064 6566 2050 6572  )[0].    def Per
-00011560: 6365 6e74 5f49 445f 5465 7374 5f53 414d  cent_ID_Test_SAM
-00011570: 504c 4544 5f53 414d 504c 4544 286d 756c  PLED_SAMPLED(mul
-00011580: 7469 696e 6465 783d 4661 6c73 6529 3a0a  tiindex=False):.
-00011590: 2020 2020 2020 2020 2247 656e 6572 6174          "Generat
-000115a0: 6520 7468 6520 4176 6572 6167 6520 616e  e the Average an
-000115b0: 6420 5354 4420 2549 4420 616d 6f6e 6720  d STD %ID among 
-000115c0: 7468 6520 7361 6d70 6c65 6420 7365 7173  the sampled seqs
-000115d0: 220a 2020 2020 2020 2020 2354 4f44 4f3a  ".        #TODO:
-000115e0: 2046 756e 6374 696f 6e20 666f 7220 7468   Function for th
-000115f0: 6520 6d75 6c74 6969 6e64 6578 2e20 536c  e multiindex. Sl
-00011600: 6963 6520 6f75 7420 7465 6e73 6f72 206f  ice out tensor o
-00011610: 7574 7369 6465 206f 6620 666f 7220 6c6f  utside of for lo
-00011620: 6f70 2e20 2253 746f 7265 2220 6e6f 6465  op. "Store" node
-00011630: 206e 616d 6573 206f 7574 7369 6465 206c   names outside l
-00011640: 6f6f 7020 286e 6f74 2077 6173 7465 2074  oop (not waste t
-00011650: 696d 6520 6c6f 6f6b 696e 6720 7570 206d  ime looking up m
-00011660: 756c 7469 6469 6d20 6172 7261 7973 290a  ultidim arrays).
-00011670: 2020 2020 2020 2020 7065 7263 656e 745f          percent_
-00011680: 6964 5f53 414d 504c 4544 5f53 414d 504c  id_SAMPLED_SAMPL
-00011690: 4544 203d 2064 6566 6175 6c74 6469 6374  ED = defaultdict
-000116a0: 286c 616d 6264 613a 2064 6566 6175 6c74  (lambda: default
-000116b0: 6469 6374 286c 616d 6264 613a 2064 6566  dict(lambda: def
-000116c0: 6175 6c74 6469 6374 2864 6963 7429 2929  aultdict(dict)))
-000116d0: 0a20 2020 2020 2020 2062 6c6f 7375 6d5f  .        blosum_
-000116e0: 5341 4d50 4c45 445f 5341 4d50 4c45 4420  SAMPLED_SAMPLED 
-000116f0: 3d20 6465 6661 756c 7464 6963 7428 6c61  = defaultdict(la
-00011700: 6d62 6461 3a20 6465 6661 756c 7464 6963  mbda: defaultdic
-00011710: 7428 6c61 6d62 6461 3a20 6465 6661 756c  t(lambda: defaul
-00011720: 7464 6963 7428 6469 6374 2929 290a 2020  tdict(dict))).  
-00011730: 2020 2020 2020 7072 6564 6963 7469 6f6e        prediction
-00011740: 7320 3d20 7072 6564 6963 7469 6f6e 735f  s = predictions_
-00011750: 7361 6d70 6c65 735b 3a2c 203a 2c20 333a  samples[:, :, 3:
-00011760: 5d0a 2020 2020 2020 2020 6966 206d 756c  ].        if mul
-00011770: 7469 696e 6465 783a 0a20 2020 2020 2020  tiindex:.       
-00011780: 2020 2020 2064 6566 2073 746f 7261 6765       def storage
-00011790: 2829 3a0a 2020 2020 2020 2020 2020 2020  ():.            
-000117a0: 2020 2020 7065 7263 656e 745f 6964 5f53      percent_id_S
-000117b0: 414d 504c 4544 5f53 414d 504c 4544 5b74  AMPLED_SAMPLED[t
-000117c0: 315d 5b74 325d 205b 2241 7665 7261 6765  1][t2] ["Average
-000117d0: 225d 203d 206e 702e 6d65 616e 286e 702e  "] = np.mean(np.
-000117e0: 6172 7261 7928 7065 7263 656e 745f 6964  array(percent_id
-000117f0: 5f69 292c 2061 7869 733d 3029 0a20 2020  _i), axis=0).   
-00011800: 2020 2020 2020 2020 2020 2020 2070 6572               per
-00011810: 6365 6e74 5f69 645f 5341 4d50 4c45 445f  cent_id_SAMPLED_
-00011820: 5341 4d50 4c45 445b 7431 5d5b 7432 5d20  SAMPLED[t1][t2] 
-00011830: 5b22 5374 6422 5d20 3d20 6e70 2e73 7464  ["Std"] = np.std
-00011840: 286e 702e 6172 7261 7928 7065 7263 656e  (np.array(percen
-00011850: 745f 6964 5f69 292c 2061 7869 733d 3029  t_id_i), axis=0)
-00011860: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00011870: 2062 6c6f 7375 6d5f 5341 4d50 4c45 445f   blosum_SAMPLED_
-00011880: 5341 4d50 4c45 445b 7431 5d5b 7432 5d20  SAMPLED[t1][t2] 
-00011890: 5b22 4176 6572 6167 6522 5d20 3d20 6e70  ["Average"] = np
-000118a0: 2e6d 6561 6e28 6e70 2e61 7272 6179 2862  .mean(np.array(b
-000118b0: 6c6f 7375 6d5f 7363 6f72 655f 6929 2c20  losum_score_i), 
-000118c0: 6178 6973 3d30 290a 2020 2020 2020 2020  axis=0).        
-000118d0: 2020 2020 2020 2020 626c 6f73 756d 5f53          blosum_S
-000118e0: 414d 504c 4544 5f53 414d 504c 4544 5b74  AMPLED_SAMPLED[t
-000118f0: 315d 5b74 325d 205b 2022 5374 6422 5d20  1][t2] [ "Std"] 
-00011900: 3d20 6e70 2e73 7464 286e 702e 6172 7261  = np.std(np.arra
-00011910: 7928 626c 6f73 756d 5f73 636f 7265 5f69  y(blosum_score_i
-00011920: 292c 2061 7869 733d 3029 0a20 2020 2020  ), axis=0).     
-00011930: 2020 2065 6c73 653a 0a20 2020 2020 2020     else:.       
-00011940: 2020 2020 2064 6566 2073 746f 7261 6765       def storage
-00011950: 2829 3a0a 2020 2020 2020 2020 2020 2020  ():.            
-00011960: 2020 2020 7065 7263 656e 745f 6964 5f53      percent_id_S
-00011970: 414d 504c 4544 5f53 414d 504c 4544 5b74  AMPLED_SAMPLED[t
-00011980: 315d 5b74 325d 2020 3d20 6e70 2e6d 6561  1][t2]  = np.mea
-00011990: 6e28 6e70 2e61 7272 6179 2870 6572 6365  n(np.array(perce
-000119a0: 6e74 5f69 645f 6929 2c20 6178 6973 3d30  nt_id_i), axis=0
-000119b0: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
-000119c0: 2020 626c 6f73 756d 5f53 414d 504c 4544    blosum_SAMPLED
-000119d0: 5f53 414d 504c 4544 5b74 315d 5b74 325d  _SAMPLED[t1][t2]
-000119e0: 2020 3d20 6e70 2e6d 6561 6e28 6e70 2e61    = np.mean(np.a
-000119f0: 7272 6179 2862 6c6f 7375 6d5f 7363 6f72  rray(blosum_scor
-00011a00: 655f 6929 2c20 6178 6973 3d30 290a 0a20  e_i), axis=0).. 
-00011a10: 2020 2020 2020 2066 6f72 2069 2c20 7431         for i, t1
-00011a20: 2069 6e20 656e 756d 6572 6174 6528 6e6f   in enumerate(no
-00011a30: 6465 735f 696e 6465 7865 7329 3a23 666f  des_indexes):#fo
-00011a40: 7220 7465 7374 206e 6f64 650a 2020 2020  r test node.    
-00011a50: 2020 2020 2020 2020 616c 6c5f 7361 6d70          all_samp
-00011a60: 6c65 645f 6920 3d20 7072 6564 6963 7469  led_i = predicti
-00011a70: 6f6e 735b 3a2c 695d 2020 2320 416c 6c20  ons[:,i]  # All 
-00011a80: 7361 6d70 6c65 7320 666f 7220 7468 6520  samples for the 
-00011a90: 7361 6d65 2074 6573 7420 7365 710a 2020  same test seq.  
-00011aa0: 2020 2020 2020 2020 2020 666f 7220 6a2c            for j,
-00011ab0: 2074 3220 696e 2065 6e75 6d65 7261 7465   t2 in enumerate
-00011ac0: 286e 6f64 6573 5f69 6e64 6578 6573 5b69  (nodes_indexes[i
-00011ad0: 3a5d 293a 0a20 2020 2020 2020 2020 2020  :]):.           
-00011ae0: 2020 2020 2061 6c6c 5f73 616d 706c 6564       all_sampled
-00011af0: 5f6a 203d 2070 7265 6469 6374 696f 6e73  _j = predictions
-00011b00: 5b3a 2c20 692b 6a5d 2020 2320 416c 6c20  [:, i+j]  # All 
-00011b10: 7361 6d70 6c65 7320 666f 7220 7468 6520  samples for the 
-00011b20: 7361 6d65 2074 6573 7420 7365 710a 2020  same test seq.  
-00011b30: 2020 2020 2020 2020 2020 2020 2020 7065                pe
-00011b40: 7263 656e 745f 6964 5f69 3d5b 5d20 2361  rcent_id_i=[] #a
-00011b50: 6c6c 2073 616d 706c 6573 2066 6f72 2074  ll samples for t
-00011b60: 6865 2073 616d 6520 7465 7374 2073 6571  he same test seq
-00011b70: 7565 6e63 650a 2020 2020 2020 2020 2020  uence.          
-00011b80: 2020 2020 2020 626c 6f73 756d 5f73 636f        blosum_sco
-00011b90: 7265 5f69 3d5b 5d0a 2020 2020 2020 2020  re_i=[].        
-00011ba0: 2020 2020 2020 2020 666f 7220 7720 696e          for w in
-00011bb0: 2072 616e 6765 286e 756d 5f73 616d 706c   range(num_sampl
-00011bc0: 6573 293a 0a20 2020 2020 2020 2020 2020  es):.           
-00011bd0: 2020 2020 2020 2020 2073 6571 5f73 616d           seq_sam
-00011be0: 706c 6564 5f69 203d 2063 6f6e 7665 7274  pled_i = convert
-00011bf0: 5f74 6f5f 6c65 7474 6572 7328 616c 6c5f  _to_letters(all_
-00011c00: 7361 6d70 6c65 645f 695b 775d 2c61 615f  sampled_i[w],aa_
-00011c10: 7072 6f62 7329 0a20 2020 2020 2020 2020  probs).         
-00011c20: 2020 2020 2020 2020 2020 2073 6571 5f73             seq_s
-00011c30: 616d 706c 6564 5f6a 203d 2063 6f6e 7665  ampled_j = conve
-00011c40: 7274 5f74 6f5f 6c65 7474 6572 7328 616c  rt_to_letters(al
-00011c50: 6c5f 7361 6d70 6c65 645f 6a5b 775d 2c61  l_sampled_j[w],a
-00011c60: 615f 7072 6f62 7329 0a20 2020 2020 2020  a_probs).       
-00011c70: 2020 2020 2020 2020 2020 2020 2070 6964               pid
-00011c80: 203d 2070 6572 635f 6964 656e 7469 7479   = perc_identity
-00011c90: 5f70 6169 725f 7365 7128 7365 715f 7361  _pair_seq(seq_sa
-00011ca0: 6d70 6c65 645f 692c 7365 715f 7361 6d70  mpled_i,seq_samp
-00011cb0: 6c65 645f 6a29 0a20 2020 2020 2020 2020  led_j).         
-00011cc0: 2020 2020 2020 2020 2020 2062 6c6f 7320             blos 
-00011cd0: 3d20 7363 6f72 655f 7061 6972 7769 7365  = score_pairwise
-00011ce0: 2873 6571 5f73 616d 706c 6564 5f69 2c73  (seq_sampled_i,s
-00011cf0: 6571 5f73 616d 706c 6564 5f6a 2c62 6c6f  eq_sampled_j,blo
-00011d00: 7375 6d2c 6761 705f 733d 2d31 312c 2067  sum,gap_s=-11, g
-00011d10: 6170 5f65 3d2d 3129 0a20 2020 2020 2020  ap_e=-1).       
-00011d20: 2020 2020 2020 2020 2020 2020 2070 6572               per
-00011d30: 6365 6e74 5f69 645f 692e 6170 7065 6e64  cent_id_i.append
-00011d40: 2870 6964 290a 2020 2020 2020 2020 2020  (pid).          
-00011d50: 2020 2020 2020 2020 2020 626c 6f73 756d            blosum
-00011d60: 5f73 636f 7265 5f69 2e61 7070 656e 6428  _score_i.append(
-00011d70: 626c 6f73 290a 2020 2020 2020 2020 2020  blos).          
-00011d80: 2020 2020 2020 7374 6f72 6167 6528 290a        storage().
-00011d90: 2020 2020 2020 2020 7265 7475 726e 2070          return p
-00011da0: 6572 6365 6e74 5f69 645f 5341 4d50 4c45  ercent_id_SAMPLE
-00011db0: 445f 5341 4d50 4c45 442c 626c 6f73 756d  D_SAMPLED,blosum
-00011dc0: 5f53 414d 504c 4544 5f53 414d 504c 4544  _SAMPLED_SAMPLED
-00011dd0: 0a0a 2020 2020 6465 6620 5065 7263 656e  ..    def Percen
-00011de0: 745f 4944 5f54 6573 745f 5341 4d50 4c45  t_ID_Test_SAMPLE
-00011df0: 445f 4f42 5345 5256 4544 286d 756c 7469  D_OBSERVED(multi
-00011e00: 696e 6465 783d 4661 6c73 6529 3a0a 2020  index=False):.  
-00011e10: 2020 2020 2020 2247 656e 6572 6174 6520        "Generate 
-00011e20: 7468 6520 4176 6572 6167 6520 616e 6420  the Average and 
-00011e30: 5354 4420 2549 4420 616e 6420 426c 6f73  STD %ID and Blos
-00011e40: 756d 2073 636f 7265 206f 6620 7468 6520  um score of the 
-00011e50: 7361 6d70 6c65 6420 7365 7173 2076 7320  sampled seqs vs 
-00011e60: 6f62 7320 7465 7374 220a 2020 2020 2020  obs test".      
-00011e70: 2020 7065 7263 656e 745f 6964 5f53 414d    percent_id_SAM
-00011e80: 504c 4544 5f4f 4253 4552 5645 4420 3d20  PLED_OBSERVED = 
-00011e90: 6465 6661 756c 7464 6963 7428 6c61 6d62  defaultdict(lamb
-00011ea0: 6461 3a20 6465 6661 756c 7464 6963 7428  da: defaultdict(
-00011eb0: 6c61 6d62 6461 3a20 6465 6661 756c 7464  lambda: defaultd
-00011ec0: 6963 7428 6469 6374 2929 290a 2020 2020  ict(dict))).    
-00011ed0: 2020 2020 626c 6f73 756d 5f53 414d 504c      blosum_SAMPL
-00011ee0: 4544 5f4f 4253 4552 5645 4420 3d20 6465  ED_OBSERVED = de
-00011ef0: 6661 756c 7464 6963 7428 6c61 6d62 6461  faultdict(lambda
-00011f00: 3a20 6465 6661 756c 7464 6963 7428 6c61  : defaultdict(la
-00011f10: 6d62 6461 3a20 6465 6661 756c 7464 6963  mbda: defaultdic
-00011f20: 7428 6469 6374 2929 290a 2020 2020 2020  t(dict))).      
-00011f30: 2020 696e 636f 7272 6563 745f 5341 4d50    incorrect_SAMP
-00011f40: 4c45 445f 4f42 5345 5256 4544 203d 2064  LED_OBSERVED = d
-00011f50: 6566 6175 6c74 6469 6374 286c 616d 6264  efaultdict(lambd
-00011f60: 613a 2064 6566 6175 6c74 6469 6374 286c  a: defaultdict(l
-00011f70: 616d 6264 613a 2064 6566 6175 6c74 6469  ambda: defaultdi
-00011f80: 6374 2864 6963 7429 2929 0a20 2020 2020  ct(dict))).     
-00011f90: 2020 2044 6174 6173 6574 5f74 6573 745f     Dataset_test_
-00011fa0: 736c 6963 6564 203d 2064 6174 6173 6574  sliced = dataset
-00011fb0: 5f74 6573 745b 3a2c 323a 2c30 5d0a 2020  _test[:,2:,0].  
-00011fc0: 2020 2020 2020 7072 6564 6963 7469 6f6e        prediction
-00011fd0: 735f 7361 6d70 6c65 735f 736c 6963 6564  s_samples_sliced
-00011fe0: 203d 2070 7265 6469 6374 696f 6e73 5f73   = predictions_s
-00011ff0: 616d 706c 6573 5b3a 2c3a 2c33 3a5d 0a20  amples[:,:,3:]. 
-00012000: 2020 2020 2020 2069 6620 6d75 6c74 6969         if multii
-00012010: 6e64 6578 3a0a 2020 2020 2020 2020 2020  ndex:.          
-00012020: 2020 6465 6620 7374 6f72 6167 6528 293a    def storage():
-00012030: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00012040: 2070 6572 6365 6e74 5f69 645f 5341 4d50   percent_id_SAMP
-00012050: 4c45 445f 4f42 5345 5256 4544 5b74 315d  LED_OBSERVED[t1]
-00012060: 5b74 325d 5b22 4176 6572 6167 6522 5d20  [t2]["Average"] 
-00012070: 3d20 6e70 2e6d 6561 6e28 6e70 2e61 7272  = np.mean(np.arr
-00012080: 6179 2870 6572 6365 6e74 5f69 645f 6929  ay(percent_id_i)
-00012090: 2c20 6178 6973 3d30 290a 2020 2020 2020  , axis=0).      
-000120a0: 2020 2020 2020 2020 2020 7065 7263 656e            percen
-000120b0: 745f 6964 5f53 414d 504c 4544 5f4f 4253  t_id_SAMPLED_OBS
-000120c0: 4552 5645 445b 7431 5d5b 7432 5d5b 2253  ERVED[t1][t2]["S
-000120d0: 7464 225d 203d 206e 702e 7374 6428 6e70  td"] = np.std(np
-000120e0: 2e61 7272 6179 2870 6572 6365 6e74 5f69  .array(percent_i
-000120f0: 645f 6929 2c20 6178 6973 3d30 290a 2020  d_i), axis=0).  
-00012100: 2020 2020 2020 2020 2020 2020 2020 626c                bl
-00012110: 6f73 756d 5f53 414d 504c 4544 5f4f 4253  osum_SAMPLED_OBS
-00012120: 4552 5645 445b 7431 5d5b 7432 5d5b 2241  ERVED[t1][t2]["A
-00012130: 7665 7261 6765 225d 203d 206e 702e 6d65  verage"] = np.me
-00012140: 616e 286e 702e 6172 7261 7928 626c 6f73  an(np.array(blos
-00012150: 756d 5f73 636f 7265 5f69 292c 2061 7869  um_score_i), axi
-00012160: 733d 3029 0a20 2020 2020 2020 2020 2020  s=0).           
-00012170: 2020 2020 2062 6c6f 7375 6d5f 5341 4d50       blosum_SAMP
-00012180: 4c45 445f 4f42 5345 5256 4544 5b74 315d  LED_OBSERVED[t1]
-00012190: 5b74 325d 5b22 5374 6422 5d20 3d20 6e70  [t2]["Std"] = np
-000121a0: 2e73 7464 286e 702e 6172 7261 7928 626c  .std(np.array(bl
-000121b0: 6f73 756d 5f73 636f 7265 5f69 292c 2061  osum_score_i), a
-000121c0: 7869 733d 3029 0a20 2020 2020 2020 2020  xis=0).         
-000121d0: 2020 2020 2020 2023 2069 6e63 6f72 7265         # incorre
-000121e0: 6374 5f53 414d 504c 4544 5f4f 4253 4552  ct_SAMPLED_OBSER
-000121f0: 5645 445b 6e6f 6465 5f6e 616d 6573 5b69  VED[node_names[i
-00012200: 5d2e 6974 656d 2829 5d5b 6e6f 6465 5f6e  ].item()][node_n
-00012210: 616d 6573 5b6a 5d2e 6974 656d 2829 5d20  ames[j].item()] 
-00012220: 5b22 4176 6572 6167 6522 5d20 3d20 6e70  ["Average"] = np
-00012230: 2e6d 6561 6e28 6e70 2e61 7272 6179 2869  .mean(np.array(i
-00012240: 6e63 6f72 7265 6374 5f61 615f 6929 2c20  ncorrect_aa_i), 
-00012250: 6178 6973 3d30 290a 2020 2020 2020 2020  axis=0).        
-00012260: 2020 2020 2020 2020 2320 696e 636f 7272          # incorr
-00012270: 6563 745f 5341 4d50 4c45 445f 4f42 5345  ect_SAMPLED_OBSE
-00012280: 5256 4544 5b6e 6f64 655f 6e61 6d65 735b  RVED[node_names[
-00012290: 695d 2e69 7465 6d28 295d 5b6e 6f64 655f  i].item()][node_
-000122a0: 6e61 6d65 735b 6a5d 2e69 7465 6d28 295d  names[j].item()]
-000122b0: 205b 2253 7464 225d 203d 206e 702e 7374   ["Std"] = np.st
-000122c0: 6428 6e70 2e61 7272 6179 2869 6e63 6f72  d(np.array(incor
-000122d0: 7265 6374 5f61 615f 6929 2c20 6178 6973  rect_aa_i), axis
-000122e0: 3d30 290a 2020 2020 2020 2020 656c 7365  =0).        else
-000122f0: 3a0a 2020 2020 2020 2020 2020 2020 6465  :.            de
-00012300: 6620 7374 6f72 6167 6528 293a 0a20 2020  f storage():.   
-00012310: 2020 2020 2020 2020 2020 2020 2070 6572               per
-00012320: 6365 6e74 5f69 645f 5341 4d50 4c45 445f  cent_id_SAMPLED_
-00012330: 4f42 5345 5256 4544 5b74 315d 5b74 325d  OBSERVED[t1][t2]
-00012340: 2020 3d20 6e70 2e6d 6561 6e28 6e70 2e61    = np.mean(np.a
-00012350: 7272 6179 2870 6572 6365 6e74 5f69 645f  rray(percent_id_
-00012360: 6929 2c20 6178 6973 3d30 290a 2020 2020  i), axis=0).    
-00012370: 2020 2020 2020 2020 2020 2020 626c 6f73              blos
-00012380: 756d 5f53 414d 504c 4544 5f4f 4253 4552  um_SAMPLED_OBSER
-00012390: 5645 445b 7431 5d5b 7432 5d20 203d 206e  VED[t1][t2]  = n
-000123a0: 702e 6d65 616e 286e 702e 6172 7261 7928  p.mean(np.array(
-000123b0: 626c 6f73 756d 5f73 636f 7265 5f69 292c  blosum_score_i),
-000123c0: 2061 7869 733d 3029 0a20 2020 2020 2020   axis=0).       
-000123d0: 2020 2020 2020 2020 2023 696e 636f 7272           #incorr
-000123e0: 6563 745f 5341 4d50 4c45 445f 4f42 5345  ect_SAMPLED_OBSE
-000123f0: 5256 4544 5b6e 6f64 655f 6e61 6d65 735b  RVED[node_names[
-00012400: 695d 2e69 7465 6d28 295d 5b6e 6f64 655f  i].item()][node_
-00012410: 6e61 6d65 735b 6a5d 2e69 7465 6d28 295d  names[j].item()]
-00012420: 2020 3d20 6e70 2e6d 6561 6e28 6e70 2e61    = np.mean(np.a
-00012430: 7272 6179 2869 6e63 6f72 7265 6374 5f61  rray(incorrect_a
-00012440: 615f 6929 2c20 6178 6973 3d30 290a 2020  a_i), axis=0).  
-00012450: 2020 2020 2020 666f 7220 692c 2074 3120        for i, t1 
-00012460: 696e 2065 6e75 6d65 7261 7465 286e 6f64  in enumerate(nod
-00012470: 6573 5f69 6e64 6578 6573 293a 0a20 2020  es_indexes):.   
-00012480: 2020 2020 2020 2020 2074 6573 745f 6f62           test_ob
-00012490: 735f 6920 3d20 636f 6e76 6572 745f 746f  s_i = convert_to
-000124a0: 5f6c 6574 7465 7273 2844 6174 6173 6574  _letters(Dataset
-000124b0: 5f74 6573 745f 736c 6963 6564 5b69 5d2c  _test_sliced[i],
-000124c0: 6161 5f70 726f 6273 290a 2020 2020 2020  aa_probs).      
-000124d0: 2020 2020 2020 666f 7220 6a2c 2074 3220        for j, t2 
-000124e0: 696e 2065 6e75 6d65 7261 7465 286e 6f64  in enumerate(nod
-000124f0: 6573 5f69 6e64 6578 6573 5b69 3a5d 293a  es_indexes[i:]):
-00012500: 2020 2320 666f 7220 7465 7374 206e 6f64    # for test nod
-00012510: 650a 2020 2020 2020 2020 2020 2020 2020  e.              
-00012520: 2020 616c 6c5f 7361 6d70 6c65 645f 7465    all_sampled_te
-00012530: 7374 203d 2070 7265 6469 6374 696f 6e73  st = predictions
-00012540: 5f73 616d 706c 6573 5f73 6c69 6365 645b  _samples_sliced[
-00012550: 3a2c 2069 2b6a 5d20 2023 2041 6c6c 2073  :, i+j]  # All s
-00012560: 616d 706c 6573 2066 6f72 2074 6865 2073  amples for the s
-00012570: 616d 6520 7465 7374 2073 6571 0a20 2020  ame test seq.   
-00012580: 2020 2020 2020 2020 2020 2020 2070 6572               per
-00012590: 6365 6e74 5f69 645f 6920 3d20 5b5d 0a20  cent_id_i = []. 
-000125a0: 2020 2020 2020 2020 2020 2020 2020 2062                 b
-000125b0: 6c6f 7375 6d5f 7363 6f72 655f 6920 3d20  losum_score_i = 
-000125c0: 5b5d 0a20 2020 2020 2020 2020 2020 2020  [].             
-000125d0: 2020 2023 696e 636f 7272 6563 745f 6161     #incorrect_aa
-000125e0: 5f69 203d 205b 5d0a 2020 2020 2020 2020  _i = [].        
-000125f0: 2020 2020 2020 2020 666f 7220 7720 696e          for w in
-00012600: 2072 616e 6765 286e 756d 5f73 616d 706c   range(num_sampl
-00012610: 6573 293a 0a20 2020 2020 2020 2020 2020  es):.           
-00012620: 2020 2020 2020 2020 2073 6571 5f73 616d           seq_sam
-00012630: 706c 6564 5f74 6573 7420 3d20 636f 6e76  pled_test = conv
-00012640: 6572 745f 746f 5f6c 6574 7465 7273 2861  ert_to_letters(a
-00012650: 6c6c 5f73 616d 706c 6564 5f74 6573 745b  ll_sampled_test[
-00012660: 775d 2c61 615f 7072 6f62 7329 0a20 2020  w],aa_probs).   
+00011040: 2061 6c69 676e 6d65 6e74 5f66 696c 655f   alignment_file_
+00011050: 6e61 6d65 3d22 7b7d 2f54 7265 655f 416c  name="{}/Tree_Al
+00011060: 6967 6e6d 656e 745f 5361 6d70 6c65 642f  ignment_Sampled/
+00011070: 7b7d 5f63 6f6d 6269 6e65 645f 7361 6d70  {}_combined_samp
+00011080: 6c65 5f69 6e64 6578 5f7b 7d2e 6d61 6666  le_index_{}.maff
+00011090: 7422 2e66 6f72 6d61 7428 7265 7375 6c74  t".format(result
+000110a0: 735f 6469 7265 6374 6f72 792c 6e61 6d65  s_directory,name
+000110b0: 2c69 6e64 6578 292c 0a20 2020 2020 2020  ,index),.       
+000110c0: 2020 2020 2020 2020 2020 2020 2020 2074                 t
+000110d0: 7265 655f 6669 6c65 5f6e 616d 653d 227b  ree_file_name="{
+000110e0: 7d2f 5472 6565 5f41 6c69 676e 6d65 6e74  }/Tree_Alignment
+000110f0: 5f53 616d 706c 6564 2f7b 7d5f 636f 6d62  _Sampled/{}_comb
+00011100: 696e 6564 5f73 616d 706c 655f 696e 6465  ined_sample_inde
+00011110: 785f 7b7d 2e74 7265 6522 2e66 6f72 6d61  x_{}.tree".forma
+00011120: 7428 7265 7375 6c74 735f 6469 7265 6374  t(results_direct
+00011130: 6f72 792c 6e61 6d65 2c69 6e64 6578 292c  ory,name,index),
+00011140: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00011150: 2020 2020 2020 206d 6574 686f 643d 6d65         method=me
+00011160: 7468 6f64 2c0a 2020 2020 2020 2020 2020  thod,.          
+00011170: 2020 2020 2020 2020 2020 2020 6e61 6d65              name
+00011180: 3d22 7b7d 5f63 6f6d 6269 6e65 645f 7361  ="{}_combined_sa
+00011190: 6d70 6c65 5f69 6e64 6578 5f7b 7d2e 7472  mple_index_{}.tr
+000111a0: 6565 222e 666f 726d 6174 286e 616d 652c  ee".format(name,
+000111b0: 696e 6465 7829 290a 2020 2020 7265 7475  index)).    retu
+000111c0: 726e 2074 7265 650a 6465 6620 6865 6174  rn tree.def heat
+000111d0: 6d61 7073 2870 7265 6469 6374 696f 6e73  maps(predictions
+000111e0: 5f73 616d 706c 6573 2c64 6174 6173 6574  _samples,dataset
+000111f0: 5f74 6573 742c 6e61 6d65 2c20 6e75 6d5f  _test,name, num_
+00011200: 7361 6d70 6c65 732c 6e6f 6465 735f 696e  samples,nodes_in
+00011210: 6465 7865 732c 7265 7375 6c74 735f 6469  dexes,results_di
+00011220: 7265 6374 6f72 792c 6161 5f70 726f 6273  rectory,aa_probs
+00011230: 2c61 6464 6974 696f 6e61 6c5f 6c6f 6164  ,additional_load
+00011240: 2c61 6464 6974 696f 6e61 6c5f 696e 666f  ,additional_info
+00011250: 2c63 6f72 7265 7370 6f6e 6465 6e63 655f  ,correspondence_
+00011260: 6469 6374 3d4e 6f6e 6529 3a0a 2020 2020  dict=None):.    
+00011270: 2222 2242 7569 6c64 2068 6561 746d 6170  """Build heatmap
+00011280: 7320 6265 7477 6565 6e20 7468 6520 7065  s between the pe
+00011290: 7263 656e 7420 6964 656e 7469 7479 2074  rcent identity t
+000112a0: 7275 6520 7365 7175 656e 6365 7320 286c  rue sequences (l
+000112b0: 6561 7665 7320 6f72 2069 6e74 6572 6e61  eaves or interna
+000112c0: 6c20 6e6f 6465 7329 2061 6e64 2074 6865  l nodes) and the
+000112d0: 2070 7265 6469 6374 6564 2073 6571 7565   predicted seque
+000112e0: 6e63 6573 0a20 2020 203a 7061 7261 6d20  nces.    :param 
+000112f0: 7465 6e73 6f72 2070 7265 6469 6374 696f  tensor predictio
+00011300: 6e73 5f73 616d 706c 6573 0a20 2020 203a  ns_samples.    :
+00011310: 7061 7261 6d20 6461 7461 7365 745f 7465  param dataset_te
+00011320: 7374 0a20 2020 203a 7061 7261 6d20 7374  st.    :param st
+00011330: 7220 6e61 6d65 3a20 6461 7461 7365 7420  r name: dataset 
+00011340: 7072 6f6a 6563 7420 6e61 6d65 0a20 2020  project name.   
+00011350: 203a 7061 7261 6d20 696e 7420 6e75 6d5f   :param int num_
+00011360: 7361 6d70 6c65 730a 2020 2020 3a70 6172  samples.    :par
+00011370: 616d 2074 656e 736f 7220 6e6f 6465 735f  am tensor nodes_
+00011380: 696e 6465 7865 730a 2020 2020 3a70 6172  indexes.    :par
+00011390: 616d 2073 7472 2072 6573 756c 7473 5f64  am str results_d
+000113a0: 6972 6563 746f 7279 0a20 2020 203a 7061  irectory.    :pa
+000113b0: 7261 6d20 696e 7420 6161 5f70 726f 6273  ram int aa_probs
+000113c0: 0a20 2020 203a 7061 7261 6d20 6e61 6d65  .    :param name
+000113d0: 6474 7570 6c65 2061 6464 6974 696f 6e61  dtuple additiona
+000113e0: 6c5f 6c6f 6164 0a20 2020 203a 7061 7261  l_load.    :para
+000113f0: 6d20 6e61 6d65 6474 7570 6c65 2061 6464  m namedtuple add
+00011400: 6974 696f 6e61 6c5f 696e 666f 0a20 2020  itional_info.   
+00011410: 203a 7061 7261 6d20 636f 7272 6573 706f   :param correspo
+00011420: 6e64 656e 6365 5f64 6963 7422 2222 0a20  ndence_dict""". 
+00011430: 2020 2070 7269 6e74 2822 4275 696c 6469     print("Buildi
+00011440: 6e67 2025 4944 2061 6e64 2042 6c6f 7375  ng %ID and Blosu
+00011450: 6d20 6865 6174 6d61 7073 2026 2049 6e63  m heatmaps & Inc
+00011460: 6f72 7265 6374 2061 6120 6869 7374 6f67  orrect aa histog
+00011470: 7261 6d2e 2e2e 2229 0a20 2020 2023 626c  ram...").    #bl
+00011480: 6f73 756d 203d 204d 6174 7269 7849 6e66  osum = MatrixInf
+00011490: 6f2e 626c 6f73 756d 3632 0a20 2020 2062  o.blosum62.    b
+000114a0: 6c6f 7375 6d20 3d20 6164 6469 7469 6f6e  losum = addition
+000114b0: 616c 5f69 6e66 6f2e 626c 6f73 756d 5f64  al_info.blosum_d
+000114c0: 6963 740a 2020 2020 6e6f 6465 735f 696e  ict.    nodes_in
+000114d0: 6465 7865 7320 3d20 6e6f 6465 735f 696e  dexes = nodes_in
+000114e0: 6465 7865 732e 746f 6c69 7374 2829 0a20  dexes.tolist(). 
+000114f0: 2020 2023 6e6f 6465 5f6e 616d 6573 203d     #node_names =
+00011500: 2064 6174 6173 6574 5f74 6573 745b 3a2c   dataset_test[:,
+00011510: 2030 2c20 315d 0a20 2020 2066 6f6c 6465   0, 1].    folde
+00011520: 7220 3d20 6f73 2e70 6174 682e 6261 7365  r = os.path.base
+00011530: 6e61 6d65 2872 6573 756c 7473 5f64 6972  name(results_dir
+00011540: 6563 746f 7279 292e 7370 6c69 7428 225f  ectory).split("_
+00011550: 2229 5b30 5d0a 2020 2020 6465 6620 5065  ")[0].    def Pe
+00011560: 7263 656e 745f 4944 5f54 6573 745f 5341  rcent_ID_Test_SA
+00011570: 4d50 4c45 445f 5341 4d50 4c45 4428 6d75  MPLED_SAMPLED(mu
+00011580: 6c74 6969 6e64 6578 3d46 616c 7365 293a  ltiindex=False):
+00011590: 0a20 2020 2020 2020 2022 4765 6e65 7261  .        "Genera
+000115a0: 7465 2074 6865 2041 7665 7261 6765 2061  te the Average a
+000115b0: 6e64 2053 5444 2025 4944 2061 6d6f 6e67  nd STD %ID among
+000115c0: 2074 6865 2073 616d 706c 6564 2073 6571   the sampled seq
+000115d0: 7322 0a20 2020 2020 2020 2023 544f 444f  s".        #TODO
+000115e0: 3a20 4675 6e63 7469 6f6e 2066 6f72 2074  : Function for t
+000115f0: 6865 206d 756c 7469 696e 6465 782e 2053  he multiindex. S
+00011600: 6c69 6365 206f 7574 2074 656e 736f 7220  lice out tensor 
+00011610: 6f75 7473 6964 6520 6f66 2066 6f72 206c  outside of for l
+00011620: 6f6f 702e 2022 5374 6f72 6522 206e 6f64  oop. "Store" nod
+00011630: 6520 6e61 6d65 7320 6f75 7473 6964 6520  e names outside 
+00011640: 6c6f 6f70 2028 6e6f 7420 7761 7374 6520  loop (not waste 
+00011650: 7469 6d65 206c 6f6f 6b69 6e67 2075 7020  time looking up 
+00011660: 6d75 6c74 6964 696d 2061 7272 6179 7329  multidim arrays)
+00011670: 0a20 2020 2020 2020 2070 6572 6365 6e74  .        percent
+00011680: 5f69 645f 5341 4d50 4c45 445f 5341 4d50  _id_SAMPLED_SAMP
+00011690: 4c45 4420 3d20 6465 6661 756c 7464 6963  LED = defaultdic
+000116a0: 7428 6c61 6d62 6461 3a20 6465 6661 756c  t(lambda: defaul
+000116b0: 7464 6963 7428 6c61 6d62 6461 3a20 6465  tdict(lambda: de
+000116c0: 6661 756c 7464 6963 7428 6469 6374 2929  faultdict(dict))
+000116d0: 290a 2020 2020 2020 2020 626c 6f73 756d  ).        blosum
+000116e0: 5f53 414d 504c 4544 5f53 414d 504c 4544  _SAMPLED_SAMPLED
+000116f0: 203d 2064 6566 6175 6c74 6469 6374 286c   = defaultdict(l
+00011700: 616d 6264 613a 2064 6566 6175 6c74 6469  ambda: defaultdi
+00011710: 6374 286c 616d 6264 613a 2064 6566 6175  ct(lambda: defau
+00011720: 6c74 6469 6374 2864 6963 7429 2929 0a20  ltdict(dict))). 
+00011730: 2020 2020 2020 2070 7265 6469 6374 696f         predictio
+00011740: 6e73 203d 2070 7265 6469 6374 696f 6e73  ns = predictions
+00011750: 5f73 616d 706c 6573 5b3a 2c20 3a2c 2033  _samples[:, :, 3
+00011760: 3a5d 0a20 2020 2020 2020 2069 6620 6d75  :].        if mu
+00011770: 6c74 6969 6e64 6578 3a0a 2020 2020 2020  ltiindex:.      
+00011780: 2020 2020 2020 6465 6620 7374 6f72 6167        def storag
+00011790: 6528 293a 0a20 2020 2020 2020 2020 2020  e():.           
+000117a0: 2020 2020 2070 6572 6365 6e74 5f69 645f       percent_id_
+000117b0: 5341 4d50 4c45 445f 5341 4d50 4c45 445b  SAMPLED_SAMPLED[
+000117c0: 7431 5d5b 7432 5d20 5b22 4176 6572 6167  t1][t2] ["Averag
+000117d0: 6522 5d20 3d20 6e70 2e6d 6561 6e28 6e70  e"] = np.mean(np
+000117e0: 2e61 7272 6179 2870 6572 6365 6e74 5f69  .array(percent_i
+000117f0: 645f 6929 2c20 6178 6973 3d30 290a 2020  d_i), axis=0).  
+00011800: 2020 2020 2020 2020 2020 2020 2020 7065                pe
+00011810: 7263 656e 745f 6964 5f53 414d 504c 4544  rcent_id_SAMPLED
+00011820: 5f53 414d 504c 4544 5b74 315d 5b74 325d  _SAMPLED[t1][t2]
+00011830: 205b 2253 7464 225d 203d 206e 702e 7374   ["Std"] = np.st
+00011840: 6428 6e70 2e61 7272 6179 2870 6572 6365  d(np.array(perce
+00011850: 6e74 5f69 645f 6929 2c20 6178 6973 3d30  nt_id_i), axis=0
+00011860: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+00011870: 2020 626c 6f73 756d 5f53 414d 504c 4544    blosum_SAMPLED
+00011880: 5f53 414d 504c 4544 5b74 315d 5b74 325d  _SAMPLED[t1][t2]
+00011890: 205b 2241 7665 7261 6765 225d 203d 206e   ["Average"] = n
+000118a0: 702e 6d65 616e 286e 702e 6172 7261 7928  p.mean(np.array(
+000118b0: 626c 6f73 756d 5f73 636f 7265 5f69 292c  blosum_score_i),
+000118c0: 2061 7869 733d 3029 0a20 2020 2020 2020   axis=0).       
+000118d0: 2020 2020 2020 2020 2062 6c6f 7375 6d5f           blosum_
+000118e0: 5341 4d50 4c45 445f 5341 4d50 4c45 445b  SAMPLED_SAMPLED[
+000118f0: 7431 5d5b 7432 5d20 5b20 2253 7464 225d  t1][t2] [ "Std"]
+00011900: 203d 206e 702e 7374 6428 6e70 2e61 7272   = np.std(np.arr
+00011910: 6179 2862 6c6f 7375 6d5f 7363 6f72 655f  ay(blosum_score_
+00011920: 6929 2c20 6178 6973 3d30 290a 2020 2020  i), axis=0).    
+00011930: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
+00011940: 2020 2020 2020 6465 6620 7374 6f72 6167        def storag
+00011950: 6528 293a 0a20 2020 2020 2020 2020 2020  e():.           
+00011960: 2020 2020 2070 6572 6365 6e74 5f69 645f       percent_id_
+00011970: 5341 4d50 4c45 445f 5341 4d50 4c45 445b  SAMPLED_SAMPLED[
+00011980: 7431 5d5b 7432 5d20 203d 206e 702e 6d65  t1][t2]  = np.me
+00011990: 616e 286e 702e 6172 7261 7928 7065 7263  an(np.array(perc
+000119a0: 656e 745f 6964 5f69 292c 2061 7869 733d  ent_id_i), axis=
+000119b0: 3029 0a20 2020 2020 2020 2020 2020 2020  0).             
+000119c0: 2020 2062 6c6f 7375 6d5f 5341 4d50 4c45     blosum_SAMPLE
+000119d0: 445f 5341 4d50 4c45 445b 7431 5d5b 7432  D_SAMPLED[t1][t2
+000119e0: 5d20 203d 206e 702e 6d65 616e 286e 702e  ]  = np.mean(np.
+000119f0: 6172 7261 7928 626c 6f73 756d 5f73 636f  array(blosum_sco
+00011a00: 7265 5f69 292c 2061 7869 733d 3029 0a0a  re_i), axis=0)..
+00011a10: 2020 2020 2020 2020 666f 7220 692c 2074          for i, t
+00011a20: 3120 696e 2065 6e75 6d65 7261 7465 286e  1 in enumerate(n
+00011a30: 6f64 6573 5f69 6e64 6578 6573 293a 2366  odes_indexes):#f
+00011a40: 6f72 2074 6573 7420 6e6f 6465 0a20 2020  or test node.   
+00011a50: 2020 2020 2020 2020 2061 6c6c 5f73 616d           all_sam
+00011a60: 706c 6564 5f69 203d 2070 7265 6469 6374  pled_i = predict
+00011a70: 696f 6e73 5b3a 2c69 5d20 2023 2041 6c6c  ions[:,i]  # All
+00011a80: 2073 616d 706c 6573 2066 6f72 2074 6865   samples for the
+00011a90: 2073 616d 6520 7465 7374 2073 6571 0a20   same test seq. 
+00011aa0: 2020 2020 2020 2020 2020 2066 6f72 206a             for j
+00011ab0: 2c20 7432 2069 6e20 656e 756d 6572 6174  , t2 in enumerat
+00011ac0: 6528 6e6f 6465 735f 696e 6465 7865 735b  e(nodes_indexes[
+00011ad0: 693a 5d29 3a0a 2020 2020 2020 2020 2020  i:]):.          
+00011ae0: 2020 2020 2020 616c 6c5f 7361 6d70 6c65        all_sample
+00011af0: 645f 6a20 3d20 7072 6564 6963 7469 6f6e  d_j = prediction
+00011b00: 735b 3a2c 2069 2b6a 5d20 2023 2041 6c6c  s[:, i+j]  # All
+00011b10: 2073 616d 706c 6573 2066 6f72 2074 6865   samples for the
+00011b20: 2073 616d 6520 7465 7374 2073 6571 0a20   same test seq. 
+00011b30: 2020 2020 2020 2020 2020 2020 2020 2070                 p
+00011b40: 6572 6365 6e74 5f69 645f 693d 5b5d 2023  ercent_id_i=[] #
+00011b50: 616c 6c20 7361 6d70 6c65 7320 666f 7220  all samples for 
+00011b60: 7468 6520 7361 6d65 2074 6573 7420 7365  the same test se
+00011b70: 7175 656e 6365 0a20 2020 2020 2020 2020  quence.         
+00011b80: 2020 2020 2020 2062 6c6f 7375 6d5f 7363         blosum_sc
+00011b90: 6f72 655f 693d 5b5d 0a20 2020 2020 2020  ore_i=[].       
+00011ba0: 2020 2020 2020 2020 2066 6f72 2077 2069           for w i
+00011bb0: 6e20 7261 6e67 6528 6e75 6d5f 7361 6d70  n range(num_samp
+00011bc0: 6c65 7329 3a0a 2020 2020 2020 2020 2020  les):.          
+00011bd0: 2020 2020 2020 2020 2020 7365 715f 7361            seq_sa
+00011be0: 6d70 6c65 645f 6920 3d20 636f 6e76 6572  mpled_i = conver
+00011bf0: 745f 746f 5f6c 6574 7465 7273 2861 6c6c  t_to_letters(all
+00011c00: 5f73 616d 706c 6564 5f69 5b77 5d2c 6161  _sampled_i[w],aa
+00011c10: 5f70 726f 6273 290a 2020 2020 2020 2020  _probs).        
+00011c20: 2020 2020 2020 2020 2020 2020 7365 715f              seq_
+00011c30: 7361 6d70 6c65 645f 6a20 3d20 636f 6e76  sampled_j = conv
+00011c40: 6572 745f 746f 5f6c 6574 7465 7273 2861  ert_to_letters(a
+00011c50: 6c6c 5f73 616d 706c 6564 5f6a 5b77 5d2c  ll_sampled_j[w],
+00011c60: 6161 5f70 726f 6273 290a 2020 2020 2020  aa_probs).      
+00011c70: 2020 2020 2020 2020 2020 2020 2020 7069                pi
+00011c80: 6420 3d20 7065 7263 5f69 6465 6e74 6974  d = perc_identit
+00011c90: 795f 7061 6972 5f73 6571 2873 6571 5f73  y_pair_seq(seq_s
+00011ca0: 616d 706c 6564 5f69 2c73 6571 5f73 616d  ampled_i,seq_sam
+00011cb0: 706c 6564 5f6a 290a 2020 2020 2020 2020  pled_j).        
+00011cc0: 2020 2020 2020 2020 2020 2020 626c 6f73              blos
+00011cd0: 203d 2073 636f 7265 5f70 6169 7277 6973   = score_pairwis
+00011ce0: 6528 7365 715f 7361 6d70 6c65 645f 692c  e(seq_sampled_i,
+00011cf0: 7365 715f 7361 6d70 6c65 645f 6a2c 626c  seq_sampled_j,bl
+00011d00: 6f73 756d 2c67 6170 5f73 3d2d 3131 2c20  osum,gap_s=-11, 
+00011d10: 6761 705f 653d 2d31 290a 2020 2020 2020  gap_e=-1).      
+00011d20: 2020 2020 2020 2020 2020 2020 2020 7065                pe
+00011d30: 7263 656e 745f 6964 5f69 2e61 7070 656e  rcent_id_i.appen
+00011d40: 6428 7069 6429 0a20 2020 2020 2020 2020  d(pid).         
+00011d50: 2020 2020 2020 2020 2020 2062 6c6f 7375             blosu
+00011d60: 6d5f 7363 6f72 655f 692e 6170 7065 6e64  m_score_i.append
+00011d70: 2862 6c6f 7329 0a20 2020 2020 2020 2020  (blos).         
+00011d80: 2020 2020 2020 2073 746f 7261 6765 2829         storage()
+00011d90: 0a20 2020 2020 2020 2072 6574 7572 6e20  .        return 
+00011da0: 7065 7263 656e 745f 6964 5f53 414d 504c  percent_id_SAMPL
+00011db0: 4544 5f53 414d 504c 4544 2c62 6c6f 7375  ED_SAMPLED,blosu
+00011dc0: 6d5f 5341 4d50 4c45 445f 5341 4d50 4c45  m_SAMPLED_SAMPLE
+00011dd0: 440a 0a20 2020 2064 6566 2050 6572 6365  D..    def Perce
+00011de0: 6e74 5f49 445f 5465 7374 5f53 414d 504c  nt_ID_Test_SAMPL
+00011df0: 4544 5f4f 4253 4552 5645 4428 6d75 6c74  ED_OBSERVED(mult
+00011e00: 6969 6e64 6578 3d46 616c 7365 293a 0a20  iindex=False):. 
+00011e10: 2020 2020 2020 2022 4765 6e65 7261 7465         "Generate
+00011e20: 2074 6865 2041 7665 7261 6765 2061 6e64   the Average and
+00011e30: 2053 5444 2025 4944 2061 6e64 2042 6c6f   STD %ID and Blo
+00011e40: 7375 6d20 7363 6f72 6520 6f66 2074 6865  sum score of the
+00011e50: 2073 616d 706c 6564 2073 6571 7320 7673   sampled seqs vs
+00011e60: 206f 6273 2074 6573 7422 0a20 2020 2020   obs test".     
+00011e70: 2020 2070 6572 6365 6e74 5f69 645f 5341     percent_id_SA
+00011e80: 4d50 4c45 445f 4f42 5345 5256 4544 203d  MPLED_OBSERVED =
+00011e90: 2064 6566 6175 6c74 6469 6374 286c 616d   defaultdict(lam
+00011ea0: 6264 613a 2064 6566 6175 6c74 6469 6374  bda: defaultdict
+00011eb0: 286c 616d 6264 613a 2064 6566 6175 6c74  (lambda: default
+00011ec0: 6469 6374 2864 6963 7429 2929 0a20 2020  dict(dict))).   
+00011ed0: 2020 2020 2062 6c6f 7375 6d5f 5341 4d50       blosum_SAMP
+00011ee0: 4c45 445f 4f42 5345 5256 4544 203d 2064  LED_OBSERVED = d
+00011ef0: 6566 6175 6c74 6469 6374 286c 616d 6264  efaultdict(lambd
+00011f00: 613a 2064 6566 6175 6c74 6469 6374 286c  a: defaultdict(l
+00011f10: 616d 6264 613a 2064 6566 6175 6c74 6469  ambda: defaultdi
+00011f20: 6374 2864 6963 7429 2929 0a20 2020 2020  ct(dict))).     
+00011f30: 2020 2069 6e63 6f72 7265 6374 5f53 414d     incorrect_SAM
+00011f40: 504c 4544 5f4f 4253 4552 5645 4420 3d20  PLED_OBSERVED = 
+00011f50: 6465 6661 756c 7464 6963 7428 6c61 6d62  defaultdict(lamb
+00011f60: 6461 3a20 6465 6661 756c 7464 6963 7428  da: defaultdict(
+00011f70: 6c61 6d62 6461 3a20 6465 6661 756c 7464  lambda: defaultd
+00011f80: 6963 7428 6469 6374 2929 290a 2020 2020  ict(dict))).    
+00011f90: 2020 2020 4461 7461 7365 745f 7465 7374      Dataset_test
+00011fa0: 5f73 6c69 6365 6420 3d20 6461 7461 7365  _sliced = datase
+00011fb0: 745f 7465 7374 5b3a 2c32 3a2c 305d 0a20  t_test[:,2:,0]. 
+00011fc0: 2020 2020 2020 2070 7265 6469 6374 696f         predictio
+00011fd0: 6e73 5f73 616d 706c 6573 5f73 6c69 6365  ns_samples_slice
+00011fe0: 6420 3d20 7072 6564 6963 7469 6f6e 735f  d = predictions_
+00011ff0: 7361 6d70 6c65 735b 3a2c 3a2c 333a 5d0a  samples[:,:,3:].
+00012000: 2020 2020 2020 2020 6966 206d 756c 7469          if multi
+00012010: 696e 6465 783a 0a20 2020 2020 2020 2020  index:.         
+00012020: 2020 2064 6566 2073 746f 7261 6765 2829     def storage()
+00012030: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+00012040: 2020 7065 7263 656e 745f 6964 5f53 414d    percent_id_SAM
+00012050: 504c 4544 5f4f 4253 4552 5645 445b 7431  PLED_OBSERVED[t1
+00012060: 5d5b 7432 5d5b 2241 7665 7261 6765 225d  ][t2]["Average"]
+00012070: 203d 206e 702e 6d65 616e 286e 702e 6172   = np.mean(np.ar
+00012080: 7261 7928 7065 7263 656e 745f 6964 5f69  ray(percent_id_i
+00012090: 292c 2061 7869 733d 3029 0a20 2020 2020  ), axis=0).     
+000120a0: 2020 2020 2020 2020 2020 2070 6572 6365             perce
+000120b0: 6e74 5f69 645f 5341 4d50 4c45 445f 4f42  nt_id_SAMPLED_OB
+000120c0: 5345 5256 4544 5b74 315d 5b74 325d 5b22  SERVED[t1][t2]["
+000120d0: 5374 6422 5d20 3d20 6e70 2e73 7464 286e  Std"] = np.std(n
+000120e0: 702e 6172 7261 7928 7065 7263 656e 745f  p.array(percent_
+000120f0: 6964 5f69 292c 2061 7869 733d 3029 0a20  id_i), axis=0). 
+00012100: 2020 2020 2020 2020 2020 2020 2020 2062                 b
+00012110: 6c6f 7375 6d5f 5341 4d50 4c45 445f 4f42  losum_SAMPLED_OB
+00012120: 5345 5256 4544 5b74 315d 5b74 325d 5b22  SERVED[t1][t2]["
+00012130: 4176 6572 6167 6522 5d20 3d20 6e70 2e6d  Average"] = np.m
+00012140: 6561 6e28 6e70 2e61 7272 6179 2862 6c6f  ean(np.array(blo
+00012150: 7375 6d5f 7363 6f72 655f 6929 2c20 6178  sum_score_i), ax
+00012160: 6973 3d30 290a 2020 2020 2020 2020 2020  is=0).          
+00012170: 2020 2020 2020 626c 6f73 756d 5f53 414d        blosum_SAM
+00012180: 504c 4544 5f4f 4253 4552 5645 445b 7431  PLED_OBSERVED[t1
+00012190: 5d5b 7432 5d5b 2253 7464 225d 203d 206e  ][t2]["Std"] = n
+000121a0: 702e 7374 6428 6e70 2e61 7272 6179 2862  p.std(np.array(b
+000121b0: 6c6f 7375 6d5f 7363 6f72 655f 6929 2c20  losum_score_i), 
+000121c0: 6178 6973 3d30 290a 2020 2020 2020 2020  axis=0).        
+000121d0: 2020 2020 2020 2020 2320 696e 636f 7272          # incorr
+000121e0: 6563 745f 5341 4d50 4c45 445f 4f42 5345  ect_SAMPLED_OBSE
+000121f0: 5256 4544 5b6e 6f64 655f 6e61 6d65 735b  RVED[node_names[
+00012200: 695d 2e69 7465 6d28 295d 5b6e 6f64 655f  i].item()][node_
+00012210: 6e61 6d65 735b 6a5d 2e69 7465 6d28 295d  names[j].item()]
+00012220: 205b 2241 7665 7261 6765 225d 203d 206e   ["Average"] = n
+00012230: 702e 6d65 616e 286e 702e 6172 7261 7928  p.mean(np.array(
+00012240: 696e 636f 7272 6563 745f 6161 5f69 292c  incorrect_aa_i),
+00012250: 2061 7869 733d 3029 0a20 2020 2020 2020   axis=0).       
+00012260: 2020 2020 2020 2020 2023 2069 6e63 6f72           # incor
+00012270: 7265 6374 5f53 414d 504c 4544 5f4f 4253  rect_SAMPLED_OBS
+00012280: 4552 5645 445b 6e6f 6465 5f6e 616d 6573  ERVED[node_names
+00012290: 5b69 5d2e 6974 656d 2829 5d5b 6e6f 6465  [i].item()][node
+000122a0: 5f6e 616d 6573 5b6a 5d2e 6974 656d 2829  _names[j].item()
+000122b0: 5d20 5b22 5374 6422 5d20 3d20 6e70 2e73  ] ["Std"] = np.s
+000122c0: 7464 286e 702e 6172 7261 7928 696e 636f  td(np.array(inco
+000122d0: 7272 6563 745f 6161 5f69 292c 2061 7869  rrect_aa_i), axi
+000122e0: 733d 3029 0a20 2020 2020 2020 2065 6c73  s=0).        els
+000122f0: 653a 0a20 2020 2020 2020 2020 2020 2064  e:.            d
+00012300: 6566 2073 746f 7261 6765 2829 3a0a 2020  ef storage():.  
+00012310: 2020 2020 2020 2020 2020 2020 2020 7065                pe
+00012320: 7263 656e 745f 6964 5f53 414d 504c 4544  rcent_id_SAMPLED
+00012330: 5f4f 4253 4552 5645 445b 7431 5d5b 7432  _OBSERVED[t1][t2
+00012340: 5d20 203d 206e 702e 6d65 616e 286e 702e  ]  = np.mean(np.
+00012350: 6172 7261 7928 7065 7263 656e 745f 6964  array(percent_id
+00012360: 5f69 292c 2061 7869 733d 3029 0a20 2020  _i), axis=0).   
+00012370: 2020 2020 2020 2020 2020 2020 2062 6c6f               blo
+00012380: 7375 6d5f 5341 4d50 4c45 445f 4f42 5345  sum_SAMPLED_OBSE
+00012390: 5256 4544 5b74 315d 5b74 325d 2020 3d20  RVED[t1][t2]  = 
+000123a0: 6e70 2e6d 6561 6e28 6e70 2e61 7272 6179  np.mean(np.array
+000123b0: 2862 6c6f 7375 6d5f 7363 6f72 655f 6929  (blosum_score_i)
+000123c0: 2c20 6178 6973 3d30 290a 2020 2020 2020  , axis=0).      
+000123d0: 2020 2020 2020 2020 2020 2369 6e63 6f72            #incor
+000123e0: 7265 6374 5f53 414d 504c 4544 5f4f 4253  rect_SAMPLED_OBS
+000123f0: 4552 5645 445b 6e6f 6465 5f6e 616d 6573  ERVED[node_names
+00012400: 5b69 5d2e 6974 656d 2829 5d5b 6e6f 6465  [i].item()][node
+00012410: 5f6e 616d 6573 5b6a 5d2e 6974 656d 2829  _names[j].item()
+00012420: 5d20 203d 206e 702e 6d65 616e 286e 702e  ]  = np.mean(np.
+00012430: 6172 7261 7928 696e 636f 7272 6563 745f  array(incorrect_
+00012440: 6161 5f69 292c 2061 7869 733d 3029 0a20  aa_i), axis=0). 
+00012450: 2020 2020 2020 2066 6f72 2069 2c20 7431         for i, t1
+00012460: 2069 6e20 656e 756d 6572 6174 6528 6e6f   in enumerate(no
+00012470: 6465 735f 696e 6465 7865 7329 3a0a 2020  des_indexes):.  
+00012480: 2020 2020 2020 2020 2020 7465 7374 5f6f            test_o
+00012490: 6273 5f69 203d 2063 6f6e 7665 7274 5f74  bs_i = convert_t
+000124a0: 6f5f 6c65 7474 6572 7328 4461 7461 7365  o_letters(Datase
+000124b0: 745f 7465 7374 5f73 6c69 6365 645b 695d  t_test_sliced[i]
+000124c0: 2c61 615f 7072 6f62 7329 0a20 2020 2020  ,aa_probs).     
+000124d0: 2020 2020 2020 2066 6f72 206a 2c20 7432         for j, t2
+000124e0: 2069 6e20 656e 756d 6572 6174 6528 6e6f   in enumerate(no
+000124f0: 6465 735f 696e 6465 7865 735b 693a 5d29  des_indexes[i:])
+00012500: 3a20 2023 2066 6f72 2074 6573 7420 6e6f  :  # for test no
+00012510: 6465 0a20 2020 2020 2020 2020 2020 2020  de.             
+00012520: 2020 2061 6c6c 5f73 616d 706c 6564 5f74     all_sampled_t
+00012530: 6573 7420 3d20 7072 6564 6963 7469 6f6e  est = prediction
+00012540: 735f 7361 6d70 6c65 735f 736c 6963 6564  s_samples_sliced
+00012550: 5b3a 2c20 692b 6a5d 2020 2320 416c 6c20  [:, i+j]  # All 
+00012560: 7361 6d70 6c65 7320 666f 7220 7468 6520  samples for the 
+00012570: 7361 6d65 2074 6573 7420 7365 710a 2020  same test seq.  
+00012580: 2020 2020 2020 2020 2020 2020 2020 7065                pe
+00012590: 7263 656e 745f 6964 5f69 203d 205b 5d0a  rcent_id_i = [].
+000125a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000125b0: 626c 6f73 756d 5f73 636f 7265 5f69 203d  blosum_score_i =
+000125c0: 205b 5d0a 2020 2020 2020 2020 2020 2020   [].            
+000125d0: 2020 2020 2369 6e63 6f72 7265 6374 5f61      #incorrect_a
+000125e0: 615f 6920 3d20 5b5d 0a20 2020 2020 2020  a_i = [].       
+000125f0: 2020 2020 2020 2020 2066 6f72 2077 2069           for w i
+00012600: 6e20 7261 6e67 6528 6e75 6d5f 7361 6d70  n range(num_samp
+00012610: 6c65 7329 3a0a 2020 2020 2020 2020 2020  les):.          
+00012620: 2020 2020 2020 2020 2020 7365 715f 7361            seq_sa
+00012630: 6d70 6c65 645f 7465 7374 203d 2063 6f6e  mpled_test = con
+00012640: 7665 7274 5f74 6f5f 6c65 7474 6572 7328  vert_to_letters(
+00012650: 616c 6c5f 7361 6d70 6c65 645f 7465 7374  all_sampled_test
+00012660: 5b77 5d2c 6161 5f70 726f 6273 290a 2020  [w],aa_probs).  
 00012670: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012680: 2070 6964 203d 2070 6572 635f 6964 656e   pid = perc_iden
-00012690: 7469 7479 5f70 6169 725f 7365 7128 7465  tity_pair_seq(te
-000126a0: 7374 5f6f 6273 5f69 2c20 7365 715f 7361  st_obs_i, seq_sa
-000126b0: 6d70 6c65 645f 7465 7374 290a 2020 2020  mpled_test).    
+00012680: 2020 7069 6420 3d20 7065 7263 5f69 6465    pid = perc_ide
+00012690: 6e74 6974 795f 7061 6972 5f73 6571 2874  ntity_pair_seq(t
+000126a0: 6573 745f 6f62 735f 692c 2073 6571 5f73  est_obs_i, seq_s
+000126b0: 616d 706c 6564 5f74 6573 7429 0a20 2020  ampled_test).   
 000126c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000126d0: 2377 726f 6e67 5f70 7265 6420 3d20 696e  #wrong_pred = in
-000126e0: 636f 7272 6563 746c 795f 7072 6564 6963  correctly_predic
-000126f0: 7465 645f 6161 2874 6573 745f 6f62 735f  ted_aa(test_obs_
-00012700: 692c 7365 715f 7361 6d70 6c65 645f 7465  i,seq_sampled_te
-00012710: 7374 290a 2020 2020 2020 2020 2020 2020  st).            
-00012720: 2020 2020 2020 2020 626c 6f73 203d 2073          blos = s
-00012730: 636f 7265 5f70 6169 7277 6973 6528 7465  core_pairwise(te
-00012740: 7374 5f6f 6273 5f69 2c20 7365 715f 7361  st_obs_i, seq_sa
-00012750: 6d70 6c65 645f 7465 7374 2c20 626c 6f73  mpled_test, blos
-00012760: 756d 2c20 6761 705f 733d 2d31 312c 2067  um, gap_s=-11, g
-00012770: 6170 5f65 3d2d 3129 0a20 2020 2020 2020  ap_e=-1).       
-00012780: 2020 2020 2020 2020 2020 2020 2070 6572               per
-00012790: 6365 6e74 5f69 645f 692e 6170 7065 6e64  cent_id_i.append
-000127a0: 2870 6964 290a 2020 2020 2020 2020 2020  (pid).          
-000127b0: 2020 2020 2020 2020 2020 626c 6f73 756d            blosum
-000127c0: 5f73 636f 7265 5f69 2e61 7070 656e 6428  _score_i.append(
-000127d0: 626c 6f73 290a 2020 2020 2020 2020 2020  blos).          
-000127e0: 2020 2020 2020 2020 2020 2369 6e63 6f72            #incor
-000127f0: 7265 6374 5f61 615f 692e 6170 7065 6e64  rect_aa_i.append
-00012800: 2877 726f 6e67 5f70 7265 6429 0a20 2020  (wrong_pred).   
-00012810: 2020 2020 2020 2020 2020 2020 2073 746f               sto
-00012820: 7261 6765 2829 0a20 2020 2020 2020 2072  rage().        r
-00012830: 6574 7572 6e20 7065 7263 656e 745f 6964  eturn percent_id
-00012840: 5f53 414d 504c 4544 5f4f 4253 4552 5645  _SAMPLED_OBSERVE
-00012850: 442c 626c 6f73 756d 5f53 414d 504c 4544  D,blosum_SAMPLED
-00012860: 5f4f 4253 4552 5645 442c 696e 636f 7272  _OBSERVED,incorr
-00012870: 6563 745f 5341 4d50 4c45 445f 4f42 5345  ect_SAMPLED_OBSE
-00012880: 5256 4544 0a0a 2020 2020 6465 6620 5065  RVED..    def Pe
-00012890: 7263 656e 745f 4944 5f54 6573 745f 4f42  rcent_ID_Test_OB
-000128a0: 535f 4f42 5345 5256 4544 2829 3a0a 2020  S_OBSERVED():.  
-000128b0: 2020 2020 2020 2247 656e 6572 6174 6520        "Generate 
-000128c0: 7468 6520 4176 6572 6167 6520 616e 6420  the Average and 
-000128d0: 5354 4420 2549 4420 616d 6f6e 6720 6f62  STD %ID among ob
-000128e0: 7320 7465 7374 220a 2020 2020 2020 2020  s test".        
-000128f0: 7065 7263 656e 745f 6964 5f4f 4253 203d  percent_id_OBS =
-00012900: 2064 6566 6175 6c74 6469 6374 286c 616d   defaultdict(lam
-00012910: 6264 613a 2064 6566 6175 6c74 6469 6374  bda: defaultdict
-00012920: 286c 616d 6264 613a 2064 6566 6175 6c74  (lambda: default
-00012930: 6469 6374 2864 6963 7429 2929 0a20 2020  dict(dict))).   
-00012940: 2020 2020 2062 6c6f 7375 6d5f 4f42 5320       blosum_OBS 
-00012950: 3d20 6465 6661 756c 7464 6963 7428 6c61  = defaultdict(la
-00012960: 6d62 6461 3a20 6465 6661 756c 7464 6963  mbda: defaultdic
-00012970: 7428 6c61 6d62 6461 3a20 6465 6661 756c  t(lambda: defaul
-00012980: 7464 6963 7428 6469 6374 2929 290a 0a20  tdict(dict))).. 
-00012990: 2020 2020 2020 2044 6174 6173 6574 5f74         Dataset_t
-000129a0: 6573 745f 736c 6963 6564 203d 2064 6174  est_sliced = dat
-000129b0: 6173 6574 5f74 6573 745b 3a2c 323a 2c30  aset_test[:,2:,0
-000129c0: 5d0a 2020 2020 2020 2020 666f 7220 692c  ].        for i,
-000129d0: 2074 3120 696e 2065 6e75 6d65 7261 7465   t1 in enumerate
-000129e0: 286e 6f64 6573 5f69 6e64 6578 6573 293a  (nodes_indexes):
-000129f0: 0a20 2020 2020 2020 2020 2020 2074 6573  .            tes
-00012a00: 745f 6f62 735f 6920 3d20 636f 6e76 6572  t_obs_i = conver
-00012a10: 745f 746f 5f6c 6574 7465 7273 2844 6174  t_to_letters(Dat
-00012a20: 6173 6574 5f74 6573 745f 736c 6963 6564  aset_test_sliced
-00012a30: 5b69 5d2c 6161 5f70 726f 6273 290a 2020  [i],aa_probs).  
-00012a40: 2020 2020 2020 2020 2020 2370 6572 6365            #perce
-00012a50: 6e74 5f69 645f 4f42 535b 4461 7461 7365  nt_id_OBS[Datase
-00012a60: 745f 7465 7374 5b69 2c20 302c 2031 5d2e  t_test[i, 0, 1].
-00012a70: 6974 656d 2829 5d5b 4461 7461 7365 745f  item()][Dataset_
-00012a80: 7465 7374 5b69 2c20 302c 2031 5d2e 6974  test[i, 0, 1].it
-00012a90: 656d 2829 5d20 3d20 7065 7263 5f69 6465  em()] = perc_ide
-00012aa0: 6e74 6974 795f 7061 6972 5f73 6571 2874  ntity_pair_seq(t
-00012ab0: 6573 745f 6f62 735f 692c 2074 6573 745f  est_obs_i, test_
-00012ac0: 6f62 735f 6929 0a20 2020 2020 2020 2020  obs_i).         
-00012ad0: 2020 2066 6f72 206a 2c20 7432 2069 6e20     for j, t2 in 
-00012ae0: 656e 756d 6572 6174 6528 6e6f 6465 735f  enumerate(nodes_
-00012af0: 696e 6465 7865 735b 693a 5d29 3a0a 2020  indexes[i:]):.  
-00012b00: 2020 2020 2020 2020 2020 2020 2020 7465                te
-00012b10: 7374 5f6f 6273 5f6a 203d 2063 6f6e 7665  st_obs_j = conve
-00012b20: 7274 5f74 6f5f 6c65 7474 6572 7328 4461  rt_to_letters(Da
-00012b30: 7461 7365 745f 7465 7374 5f73 6c69 6365  taset_test_slice
-00012b40: 645b 692b 6a5d 2c61 615f 7072 6f62 7329  d[i+j],aa_probs)
-00012b50: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00012b60: 2070 6964 5f73 636f 7265 203d 2070 6572   pid_score = per
-00012b70: 635f 6964 656e 7469 7479 5f70 6169 725f  c_identity_pair_
-00012b80: 7365 7128 7465 7374 5f6f 6273 5f69 2c74  seq(test_obs_i,t
-00012b90: 6573 745f 6f62 735f 6a29 0a20 2020 2020  est_obs_j).     
-00012ba0: 2020 2020 2020 2020 2020 2070 6572 6365             perce
-00012bb0: 6e74 5f69 645f 4f42 535b 7431 5d5b 7432  nt_id_OBS[t1][t2
-00012bc0: 5d20 3d20 7069 645f 7363 6f72 650a 2020  ] = pid_score.  
-00012bd0: 2020 2020 2020 2020 2020 2020 2020 626c                bl
-00012be0: 6f73 756d 5f73 636f 7265 203d 2073 636f  osum_score = sco
-00012bf0: 7265 5f70 6169 7277 6973 6528 7465 7374  re_pairwise(test
-00012c00: 5f6f 6273 5f69 2c74 6573 745f 6f62 735f  _obs_i,test_obs_
-00012c10: 6a2c 2062 6c6f 7375 6d2c 2067 6170 5f73  j, blosum, gap_s
-00012c20: 3d2d 3131 2c20 6761 705f 653d 2d31 290a  =-11, gap_e=-1).
-00012c30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012c40: 626c 6f73 756d 5f4f 4253 5b74 315d 5b74  blosum_OBS[t1][t
-00012c50: 325d 203d 2062 6c6f 7375 6d5f 7363 6f72  2] = blosum_scor
-00012c60: 650a 2020 2020 2020 2020 7265 7475 726e  e.        return
-00012c70: 2070 6572 6365 6e74 5f69 645f 4f42 532c   percent_id_OBS,
-00012c80: 626c 6f73 756d 5f4f 4253 0a0a 2020 2020  blosum_OBS..    
-00012c90: 6465 6620 506c 6f74 5f48 6561 746d 6170  def Plot_Heatmap
-00012ca0: 2864 662c 7469 746c 652c 7469 746c 6532  (df,title,title2
-00012cb0: 2c61 6e6e 6f74 2c6d 6173 6b2c 766d 6178  ,annot,mask,vmax
-00012cc0: 293a 0a20 2020 2020 2020 2066 6967 2c61  ):.        fig,a
-00012cd0: 7820 3d20 2070 6c74 2e73 7562 706c 6f74  x =  plt.subplot
-00012ce0: 7328 312c 2031 2c66 6967 7369 7a65 3d28  s(1, 1,figsize=(
-00012cf0: 3130 2c31 3029 290a 2020 2020 2020 2020  10,10)).        
-00012d00: 686d 6170 203d 2073 6e73 2e68 6561 746d  hmap = sns.heatm
-00012d10: 6170 2864 662c 2063 6d61 703d 2253 7065  ap(df, cmap="Spe
-00012d20: 6374 7261 6c22 2c61 6e6e 6f74 3d61 6e6e  ctral",annot=ann
-00012d30: 6f74 2c61 6e6e 6f74 5f6b 7773 3d7b 2266  ot,annot_kws={"f
-00012d40: 6f6e 7473 697a 6522 3a36 7d2c 6d61 736b  ontsize":6},mask
-00012d50: 3d6d 6173 6b2c 766d 6178 3d76 6d61 7829  =mask,vmax=vmax)
-00012d60: 0a20 2020 2020 2020 2061 782e 7365 745f  .        ax.set_
-00012d70: 7469 746c 6528 227b 7d20 3b5c 6e22 2e66  title("{} ;\n".f
-00012d80: 6f72 6d61 7428 7469 746c 6529 202b 2072  ormat(title) + r
-00012d90: 227b 7d22 2e66 6f72 6d61 7428 6164 6469  "{}".format(addi
-00012da0: 7469 6f6e 616c 5f6c 6f61 642e 6675 6c6c  tional_load.full
-00012db0: 5f6e 616d 6529 290a 2020 2020 2020 2020  _name)).        
-00012dc0: 6178 2e73 6574 5f78 6c61 6265 6c28 7469  ax.set_xlabel(ti
-00012dd0: 746c 6532 2e73 706c 6974 2822 5f22 295b  tle2.split("_")[
-00012de0: 2d31 5d29 0a20 2020 2020 2020 2061 782e  -1]).        ax.
-00012df0: 7365 745f 796c 6162 656c 2874 6974 6c65  set_ylabel(title
-00012e00: 322e 7370 6c69 7428 225f 2229 5b2d 325d  2.split("_")[-2]
-00012e10: 290a 2020 2020 2020 2020 686d 6170 2e66  ).        hmap.f
-00012e20: 6967 7572 652e 7361 7665 6669 6728 227b  igure.savefig("{
-00012e30: 7d2f 4865 6174 6d61 705f 7b7d 2e70 6e67  }/Heatmap_{}.png
-00012e40: 222e 666f 726d 6174 2872 6573 756c 7473  ".format(results
-00012e50: 5f64 6972 6563 746f 7279 2c74 6974 6c65  _directory,title
-00012e60: 3229 2c0a 2020 2020 2020 2020 2020 2020  2),.            
+000126d0: 2023 7772 6f6e 675f 7072 6564 203d 2069   #wrong_pred = i
+000126e0: 6e63 6f72 7265 6374 6c79 5f70 7265 6469  ncorrectly_predi
+000126f0: 6374 6564 5f61 6128 7465 7374 5f6f 6273  cted_aa(test_obs
+00012700: 5f69 2c73 6571 5f73 616d 706c 6564 5f74  _i,seq_sampled_t
+00012710: 6573 7429 0a20 2020 2020 2020 2020 2020  est).           
+00012720: 2020 2020 2020 2020 2062 6c6f 7320 3d20           blos = 
+00012730: 7363 6f72 655f 7061 6972 7769 7365 2874  score_pairwise(t
+00012740: 6573 745f 6f62 735f 692c 2073 6571 5f73  est_obs_i, seq_s
+00012750: 616d 706c 6564 5f74 6573 742c 2062 6c6f  ampled_test, blo
+00012760: 7375 6d2c 2067 6170 5f73 3d2d 3131 2c20  sum, gap_s=-11, 
+00012770: 6761 705f 653d 2d31 290a 2020 2020 2020  gap_e=-1).      
+00012780: 2020 2020 2020 2020 2020 2020 2020 7065                pe
+00012790: 7263 656e 745f 6964 5f69 2e61 7070 656e  rcent_id_i.appen
+000127a0: 6428 7069 6429 0a20 2020 2020 2020 2020  d(pid).         
+000127b0: 2020 2020 2020 2020 2020 2062 6c6f 7375             blosu
+000127c0: 6d5f 7363 6f72 655f 692e 6170 7065 6e64  m_score_i.append
+000127d0: 2862 6c6f 7329 0a20 2020 2020 2020 2020  (blos).         
+000127e0: 2020 2020 2020 2020 2020 2023 696e 636f             #inco
+000127f0: 7272 6563 745f 6161 5f69 2e61 7070 656e  rrect_aa_i.appen
+00012800: 6428 7772 6f6e 675f 7072 6564 290a 2020  d(wrong_pred).  
+00012810: 2020 2020 2020 2020 2020 2020 2020 7374                st
+00012820: 6f72 6167 6528 290a 2020 2020 2020 2020  orage().        
+00012830: 7265 7475 726e 2070 6572 6365 6e74 5f69  return percent_i
+00012840: 645f 5341 4d50 4c45 445f 4f42 5345 5256  d_SAMPLED_OBSERV
+00012850: 4544 2c62 6c6f 7375 6d5f 5341 4d50 4c45  ED,blosum_SAMPLE
+00012860: 445f 4f42 5345 5256 4544 2c69 6e63 6f72  D_OBSERVED,incor
+00012870: 7265 6374 5f53 414d 504c 4544 5f4f 4253  rect_SAMPLED_OBS
+00012880: 4552 5645 440a 0a20 2020 2064 6566 2050  ERVED..    def P
+00012890: 6572 6365 6e74 5f49 445f 5465 7374 5f4f  ercent_ID_Test_O
+000128a0: 4253 5f4f 4253 4552 5645 4428 293a 0a20  BS_OBSERVED():. 
+000128b0: 2020 2020 2020 2022 4765 6e65 7261 7465         "Generate
+000128c0: 2074 6865 2041 7665 7261 6765 2061 6e64   the Average and
+000128d0: 2053 5444 2025 4944 2061 6d6f 6e67 206f   STD %ID among o
+000128e0: 6273 2074 6573 7422 0a20 2020 2020 2020  bs test".       
+000128f0: 2070 6572 6365 6e74 5f69 645f 4f42 5320   percent_id_OBS 
+00012900: 3d20 6465 6661 756c 7464 6963 7428 6c61  = defaultdict(la
+00012910: 6d62 6461 3a20 6465 6661 756c 7464 6963  mbda: defaultdic
+00012920: 7428 6c61 6d62 6461 3a20 6465 6661 756c  t(lambda: defaul
+00012930: 7464 6963 7428 6469 6374 2929 290a 2020  tdict(dict))).  
+00012940: 2020 2020 2020 626c 6f73 756d 5f4f 4253        blosum_OBS
+00012950: 203d 2064 6566 6175 6c74 6469 6374 286c   = defaultdict(l
+00012960: 616d 6264 613a 2064 6566 6175 6c74 6469  ambda: defaultdi
+00012970: 6374 286c 616d 6264 613a 2064 6566 6175  ct(lambda: defau
+00012980: 6c74 6469 6374 2864 6963 7429 2929 0a0a  ltdict(dict)))..
+00012990: 2020 2020 2020 2020 4461 7461 7365 745f          Dataset_
+000129a0: 7465 7374 5f73 6c69 6365 6420 3d20 6461  test_sliced = da
+000129b0: 7461 7365 745f 7465 7374 5b3a 2c32 3a2c  taset_test[:,2:,
+000129c0: 305d 0a20 2020 2020 2020 2066 6f72 2069  0].        for i
+000129d0: 2c20 7431 2069 6e20 656e 756d 6572 6174  , t1 in enumerat
+000129e0: 6528 6e6f 6465 735f 696e 6465 7865 7329  e(nodes_indexes)
+000129f0: 3a0a 2020 2020 2020 2020 2020 2020 7465  :.            te
+00012a00: 7374 5f6f 6273 5f69 203d 2063 6f6e 7665  st_obs_i = conve
+00012a10: 7274 5f74 6f5f 6c65 7474 6572 7328 4461  rt_to_letters(Da
+00012a20: 7461 7365 745f 7465 7374 5f73 6c69 6365  taset_test_slice
+00012a30: 645b 695d 2c61 615f 7072 6f62 7329 0a20  d[i],aa_probs). 
+00012a40: 2020 2020 2020 2020 2020 2023 7065 7263             #perc
+00012a50: 656e 745f 6964 5f4f 4253 5b44 6174 6173  ent_id_OBS[Datas
+00012a60: 6574 5f74 6573 745b 692c 2030 2c20 315d  et_test[i, 0, 1]
+00012a70: 2e69 7465 6d28 295d 5b44 6174 6173 6574  .item()][Dataset
+00012a80: 5f74 6573 745b 692c 2030 2c20 315d 2e69  _test[i, 0, 1].i
+00012a90: 7465 6d28 295d 203d 2070 6572 635f 6964  tem()] = perc_id
+00012aa0: 656e 7469 7479 5f70 6169 725f 7365 7128  entity_pair_seq(
+00012ab0: 7465 7374 5f6f 6273 5f69 2c20 7465 7374  test_obs_i, test
+00012ac0: 5f6f 6273 5f69 290a 2020 2020 2020 2020  _obs_i).        
+00012ad0: 2020 2020 666f 7220 6a2c 2074 3220 696e      for j, t2 in
+00012ae0: 2065 6e75 6d65 7261 7465 286e 6f64 6573   enumerate(nodes
+00012af0: 5f69 6e64 6578 6573 5b69 3a5d 293a 0a20  _indexes[i:]):. 
+00012b00: 2020 2020 2020 2020 2020 2020 2020 2074                 t
+00012b10: 6573 745f 6f62 735f 6a20 3d20 636f 6e76  est_obs_j = conv
+00012b20: 6572 745f 746f 5f6c 6574 7465 7273 2844  ert_to_letters(D
+00012b30: 6174 6173 6574 5f74 6573 745f 736c 6963  ataset_test_slic
+00012b40: 6564 5b69 2b6a 5d2c 6161 5f70 726f 6273  ed[i+j],aa_probs
+00012b50: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+00012b60: 2020 7069 645f 7363 6f72 6520 3d20 7065    pid_score = pe
+00012b70: 7263 5f69 6465 6e74 6974 795f 7061 6972  rc_identity_pair
+00012b80: 5f73 6571 2874 6573 745f 6f62 735f 692c  _seq(test_obs_i,
+00012b90: 7465 7374 5f6f 6273 5f6a 290a 2020 2020  test_obs_j).    
+00012ba0: 2020 2020 2020 2020 2020 2020 7065 7263              perc
+00012bb0: 656e 745f 6964 5f4f 4253 5b74 315d 5b74  ent_id_OBS[t1][t
+00012bc0: 325d 203d 2070 6964 5f73 636f 7265 0a20  2] = pid_score. 
+00012bd0: 2020 2020 2020 2020 2020 2020 2020 2062                 b
+00012be0: 6c6f 7375 6d5f 7363 6f72 6520 3d20 7363  losum_score = sc
+00012bf0: 6f72 655f 7061 6972 7769 7365 2874 6573  ore_pairwise(tes
+00012c00: 745f 6f62 735f 692c 7465 7374 5f6f 6273  t_obs_i,test_obs
+00012c10: 5f6a 2c20 626c 6f73 756d 2c20 6761 705f  _j, blosum, gap_
+00012c20: 733d 2d31 312c 2067 6170 5f65 3d2d 3129  s=-11, gap_e=-1)
+00012c30: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00012c40: 2062 6c6f 7375 6d5f 4f42 535b 7431 5d5b   blosum_OBS[t1][
+00012c50: 7432 5d20 3d20 626c 6f73 756d 5f73 636f  t2] = blosum_sco
+00012c60: 7265 0a20 2020 2020 2020 2072 6574 7572  re.        retur
+00012c70: 6e20 7065 7263 656e 745f 6964 5f4f 4253  n percent_id_OBS
+00012c80: 2c62 6c6f 7375 6d5f 4f42 530a 0a20 2020  ,blosum_OBS..   
+00012c90: 2064 6566 2050 6c6f 745f 4865 6174 6d61   def Plot_Heatma
+00012ca0: 7028 6466 2c74 6974 6c65 2c74 6974 6c65  p(df,title,title
+00012cb0: 322c 616e 6e6f 742c 6d61 736b 2c76 6d61  2,annot,mask,vma
+00012cc0: 7829 3a0a 2020 2020 2020 2020 6669 672c  x):.        fig,
+00012cd0: 6178 203d 2020 706c 742e 7375 6270 6c6f  ax =  plt.subplo
+00012ce0: 7473 2831 2c20 312c 6669 6773 697a 653d  ts(1, 1,figsize=
+00012cf0: 2831 302c 3130 2929 0a20 2020 2020 2020  (10,10)).       
+00012d00: 2068 6d61 7020 3d20 736e 732e 6865 6174   hmap = sns.heat
+00012d10: 6d61 7028 6466 2c20 636d 6170 3d22 5370  map(df, cmap="Sp
+00012d20: 6563 7472 616c 222c 616e 6e6f 743d 616e  ectral",annot=an
+00012d30: 6e6f 742c 616e 6e6f 745f 6b77 733d 7b22  not,annot_kws={"
+00012d40: 666f 6e74 7369 7a65 223a 367d 2c6d 6173  fontsize":6},mas
+00012d50: 6b3d 6d61 736b 2c76 6d61 783d 766d 6178  k=mask,vmax=vmax
+00012d60: 290a 2020 2020 2020 2020 6178 2e73 6574  ).        ax.set
+00012d70: 5f74 6974 6c65 2822 7b7d 203b 5c6e 222e  _title("{} ;\n".
+00012d80: 666f 726d 6174 2874 6974 6c65 2920 2b20  format(title) + 
+00012d90: 7222 7b7d 222e 666f 726d 6174 2861 6464  r"{}".format(add
+00012da0: 6974 696f 6e61 6c5f 6c6f 6164 2e66 756c  itional_load.ful
+00012db0: 6c5f 6e61 6d65 2929 0a20 2020 2020 2020  l_name)).       
+00012dc0: 2061 782e 7365 745f 786c 6162 656c 2874   ax.set_xlabel(t
+00012dd0: 6974 6c65 322e 7370 6c69 7428 225f 2229  itle2.split("_")
+00012de0: 5b2d 315d 290a 2020 2020 2020 2020 6178  [-1]).        ax
+00012df0: 2e73 6574 5f79 6c61 6265 6c28 7469 746c  .set_ylabel(titl
+00012e00: 6532 2e73 706c 6974 2822 5f22 295b 2d32  e2.split("_")[-2
+00012e10: 5d29 0a20 2020 2020 2020 2068 6d61 702e  ]).        hmap.
+00012e20: 6669 6775 7265 2e73 6176 6566 6967 2822  figure.savefig("
+00012e30: 7b7d 2f48 6561 746d 6170 5f7b 7d2e 706e  {}/Heatmap_{}.pn
+00012e40: 6722 2e66 6f72 6d61 7428 7265 7375 6c74  g".format(result
+00012e50: 735f 6469 7265 6374 6f72 792c 7469 746c  s_directory,titl
+00012e60: 6532 292c 0a20 2020 2020 2020 2020 2020  e2),.           
 00012e70: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012e80: 666f 726d 6174 3d27 706e 6727 2c0a 2020  format='png',.  
+00012e80: 2066 6f72 6d61 743d 2770 6e67 272c 0a20   format='png',. 
 00012e90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00012ea0: 2020 2020 2020 2020 2020 6470 693d 3135            dpi=15
-00012eb0: 3029 0a20 2020 2020 2020 2070 6c74 2e63  0).        plt.c
-00012ec0: 6c66 2829 0a20 2020 2020 2020 2070 6c74  lf().        plt
-00012ed0: 2e63 6c6f 7365 2829 0a0a 2020 2020 6d75  .close()..    mu
-00012ee0: 6c74 6969 6e64 6578 3d46 616c 7365 0a20  ltiindex=False. 
-00012ef0: 2020 2069 6620 6d75 6c74 6969 6e64 6578     if multiindex
-00012f00: 3a0a 2020 2020 2020 2020 2348 6967 686c  :.        #Highl
-00012f10: 6967 6874 3a20 7361 6d70 6c65 6420 7673  ight: sampled vs
-00012f20: 206f 6273 6572 7665 640a 2020 2020 2020   observed.      
-00012f30: 2020 6469 6374 5f70 6964 5f53 414d 504c    dict_pid_SAMPL
-00012f40: 4544 5f4f 4253 2c20 6469 6374 5f62 6c6f  ED_OBS, dict_blo
-00012f50: 7375 6d5f 5341 4d50 4c45 445f 4f42 532c  sum_SAMPLED_OBS,
-00012f60: 6469 6374 5f69 6e63 6f72 7265 6374 5f53  dict_incorrect_S
-00012f70: 414d 504c 4544 5f4f 4253 203d 2050 6572  AMPLED_OBS = Per
-00012f80: 6365 6e74 5f49 445f 5465 7374 5f53 414d  cent_ID_Test_SAM
-00012f90: 504c 4544 5f4f 4253 4552 5645 4428 6d75  PLED_OBSERVED(mu
-00012fa0: 6c74 6969 6e64 6578 290a 2020 2020 2020  ltiindex).      
-00012fb0: 2020 6466 5f70 6964 5f53 414d 504c 4544    df_pid_SAMPLED
-00012fc0: 5f4f 4253 203d 2070 642e 636f 6e63 6174  _OBS = pd.concat
-00012fd0: 287b 6b3a 2070 642e 4461 7461 4672 616d  ({k: pd.DataFram
-00012fe0: 652e 6672 6f6d 5f64 6963 7428 762c 2027  e.from_dict(v, '
-00012ff0: 696e 6465 7827 2920 666f 7220 6b2c 2076  index') for k, v
-00013000: 2069 6e20 6469 6374 5f70 6964 5f53 414d   in dict_pid_SAM
-00013010: 504c 4544 5f4f 4253 2e69 7465 6d73 2829  PLED_OBS.items()
-00013020: 7d2c 2061 7869 733d 3029 0a20 2020 2020  }, axis=0).     
-00013030: 2020 2064 665f 626c 6f73 756d 5f53 414d     df_blosum_SAM
-00013040: 504c 4544 5f4f 4253 203d 2070 642e 636f  PLED_OBS = pd.co
-00013050: 6e63 6174 287b 6b3a 2070 642e 4461 7461  ncat({k: pd.Data
-00013060: 4672 616d 652e 6672 6f6d 5f64 6963 7428  Frame.from_dict(
-00013070: 762c 2027 696e 6465 7827 2920 666f 7220  v, 'index') for 
-00013080: 6b2c 2076 2069 6e20 6469 6374 5f62 6c6f  k, v in dict_blo
-00013090: 7375 6d5f 5341 4d50 4c45 445f 4f42 532e  sum_SAMPLED_OBS.
-000130a0: 6974 656d 7328 297d 2c61 7869 733d 3029  items()},axis=0)
-000130b0: 0a20 2020 2020 2020 2023 6466 5f69 6e63  .        #df_inc
-000130c0: 6f72 7265 6374 5f53 414d 504c 4544 5f4f  orrect_SAMPLED_O
-000130d0: 4253 203d 2070 642e 636f 6e63 6174 287b  BS = pd.concat({
-000130e0: 6b3a 2070 642e 4461 7461 4672 616d 652e  k: pd.DataFrame.
-000130f0: 6672 6f6d 5f64 6963 7428 762c 2027 696e  from_dict(v, 'in
-00013100: 6465 7827 2920 666f 7220 6b2c 2076 2069  dex') for k, v i
-00013110: 6e20 6469 6374 5f69 6e63 6f72 7265 6374  n dict_incorrect
-00013120: 5f53 414d 504c 4544 5f4f 4253 2e69 7465  _SAMPLED_OBS.ite
-00013130: 6d73 2829 7d2c 2061 7869 733d 3029 0a20  ms()}, axis=0). 
-00013140: 2020 2020 2020 2023 4869 6768 6c69 6768         #Highligh
-00013150: 743a 2053 616d 706c 6564 2076 7320 5361  t: Sampled vs Sa
-00013160: 6d70 6c65 640a 2020 2020 2020 2020 6469  mpled.        di
-00013170: 6374 5f70 6964 5f53 414d 504c 4544 5f53  ct_pid_SAMPLED_S
-00013180: 414d 504c 4544 2c64 6963 745f 626c 6f73  AMPLED,dict_blos
-00013190: 756d 5f53 414d 504c 4544 5f53 414d 504c  um_SAMPLED_SAMPL
-000131a0: 4544 3d50 6572 6365 6e74 5f49 445f 5465  ED=Percent_ID_Te
-000131b0: 7374 5f53 414d 504c 4544 5f53 414d 504c  st_SAMPLED_SAMPL
-000131c0: 4544 286d 756c 7469 696e 6465 7829 0a20  ED(multiindex). 
-000131d0: 2020 2020 2020 2064 665f 7069 645f 5341         df_pid_SA
-000131e0: 4d50 4c45 445f 5341 4d50 4c45 4420 3d20  MPLED_SAMPLED = 
-000131f0: 7064 2e63 6f6e 6361 7428 7b6b 3a20 7064  pd.concat({k: pd
-00013200: 2e44 6174 6146 7261 6d65 2e66 726f 6d5f  .DataFrame.from_
-00013210: 6469 6374 2876 2c20 2769 6e64 6578 2729  dict(v, 'index')
-00013220: 2066 6f72 206b 2c20 7620 696e 2064 6963   for k, v in dic
-00013230: 745f 7069 645f 5341 4d50 4c45 445f 5341  t_pid_SAMPLED_SA
-00013240: 4d50 4c45 442e 6974 656d 7328 297d 2c20  MPLED.items()}, 
-00013250: 6178 6973 3d30 290a 2020 2020 2020 2020  axis=0).        
-00013260: 6466 5f62 6c6f 7375 6d5f 5341 4d50 4c45  df_blosum_SAMPLE
-00013270: 445f 5341 4d50 4c45 4420 3d20 7064 2e63  D_SAMPLED = pd.c
-00013280: 6f6e 6361 7428 7b6b 3a20 7064 2e44 6174  oncat({k: pd.Dat
-00013290: 6146 7261 6d65 2e66 726f 6d5f 6469 6374  aFrame.from_dict
-000132a0: 2876 2c20 2769 6e64 6578 2729 2066 6f72  (v, 'index') for
-000132b0: 206b 2c20 7620 696e 2064 6963 745f 626c   k, v in dict_bl
-000132c0: 6f73 756d 5f53 414d 504c 4544 5f53 414d  osum_SAMPLED_SAM
-000132d0: 504c 4544 2e69 7465 6d73 2829 7d2c 2061  PLED.items()}, a
-000132e0: 7869 733d 3029 0a0a 2020 2020 656c 7365  xis=0)..    else
-000132f0: 3a0a 2020 2020 2020 2020 2320 4869 6768  :.        # High
-00013300: 6c69 6768 743a 2073 616d 706c 6564 2076  light: sampled v
-00013310: 7320 6f62 7365 7276 6564 0a20 2020 2020  s observed.     
-00013320: 2020 2064 6963 745f 7069 645f 5341 4d50     dict_pid_SAMP
-00013330: 4c45 445f 4f42 532c 2064 6963 745f 626c  LED_OBS, dict_bl
-00013340: 6f73 756d 5f53 414d 504c 4544 5f4f 4253  osum_SAMPLED_OBS
-00013350: 2c64 6963 745f 696e 636f 7272 6563 745f  ,dict_incorrect_
-00013360: 5341 4d50 4c45 445f 4f42 5320 3d20 5065  SAMPLED_OBS = Pe
-00013370: 7263 656e 745f 4944 5f54 6573 745f 5341  rcent_ID_Test_SA
-00013380: 4d50 4c45 445f 4f42 5345 5256 4544 286d  MPLED_OBSERVED(m
-00013390: 756c 7469 696e 6465 7829 0a20 2020 2020  ultiindex).     
-000133a0: 2020 2064 665f 7069 645f 5341 4d50 4c45     df_pid_SAMPLE
-000133b0: 445f 4f42 5320 3d20 7064 2e44 6174 6146  D_OBS = pd.DataF
-000133c0: 7261 6d65 2864 6963 745f 7069 645f 5341  rame(dict_pid_SA
-000133d0: 4d50 4c45 445f 4f42 5329 0a20 2020 2020  MPLED_OBS).     
-000133e0: 2020 2064 665f 626c 6f73 756d 5f53 414d     df_blosum_SAM
-000133f0: 504c 4544 5f4f 4253 203d 2070 642e 4461  PLED_OBS = pd.Da
-00013400: 7461 4672 616d 6528 6469 6374 5f62 6c6f  taFrame(dict_blo
-00013410: 7375 6d5f 5341 4d50 4c45 445f 4f42 5329  sum_SAMPLED_OBS)
-00013420: 0a20 2020 2020 2020 2023 6466 5f69 6e63  .        #df_inc
-00013430: 6f72 7265 6374 5f53 414d 504c 4544 5f4f  orrect_SAMPLED_O
-00013440: 4253 203d 2070 642e 4461 7461 4672 616d  BS = pd.DataFram
-00013450: 6528 6469 6374 5f69 6e63 6f72 7265 6374  e(dict_incorrect
-00013460: 5f53 414d 504c 4544 5f4f 4253 292e 746f  _SAMPLED_OBS).to
-00013470: 5f6e 756d 7079 2829 0a20 2020 2020 2020  _numpy().       
-00013480: 2023 4869 6768 6c69 6768 743a 2053 616d   #Highlight: Sam
-00013490: 706c 6564 2076 7320 5361 6d70 6c65 640a  pled vs Sampled.
-000134a0: 2020 2020 2020 2020 6469 6374 5f70 6964          dict_pid
-000134b0: 5f53 414d 504c 4544 5f53 414d 504c 4544  _SAMPLED_SAMPLED
-000134c0: 2c20 6469 6374 5f62 6c6f 7375 6d5f 5341  , dict_blosum_SA
-000134d0: 4d50 4c45 445f 5341 4d50 4c45 4420 3d20  MPLED_SAMPLED = 
-000134e0: 5065 7263 656e 745f 4944 5f54 6573 745f  Percent_ID_Test_
-000134f0: 5341 4d50 4c45 445f 5341 4d50 4c45 4428  SAMPLED_SAMPLED(
-00013500: 6d75 6c74 6969 6e64 6578 290a 2020 2020  multiindex).    
-00013510: 2020 2020 6466 5f70 6964 5f53 414d 504c      df_pid_SAMPL
-00013520: 4544 5f53 414d 504c 4544 203d 2070 642e  ED_SAMPLED = pd.
-00013530: 4461 7461 4672 616d 6528 6469 6374 5f70  DataFrame(dict_p
-00013540: 6964 5f53 414d 504c 4544 5f53 414d 504c  id_SAMPLED_SAMPL
-00013550: 4544 290a 2020 2020 2020 2020 6466 5f62  ED).        df_b
-00013560: 6c6f 7375 6d5f 5341 4d50 4c45 445f 5341  losum_SAMPLED_SA
-00013570: 4d50 4c45 4420 3d20 7064 2e44 6174 6146  MPLED = pd.DataF
-00013580: 7261 6d65 2864 6963 745f 626c 6f73 756d  rame(dict_blosum
-00013590: 5f53 414d 504c 4544 5f53 414d 504c 4544  _SAMPLED_SAMPLED
-000135a0: 290a 0a0a 2020 2020 6469 6374 5f70 6964  )...    dict_pid
-000135b0: 5f4f 4253 5f4f 4253 2c64 6963 745f 626c  _OBS_OBS,dict_bl
-000135c0: 6f73 756d 5f4f 4253 5f4f 4253 203d 2050  osum_OBS_OBS = P
-000135d0: 6572 6365 6e74 5f49 445f 5465 7374 5f4f  ercent_ID_Test_O
-000135e0: 4253 5f4f 4253 4552 5645 4428 290a 2020  BS_OBSERVED().  
-000135f0: 2020 6466 5f70 6964 5f4f 4253 5f4f 4253    df_pid_OBS_OBS
-00013600: 203d 2070 642e 4461 7461 4672 616d 6528   = pd.DataFrame(
-00013610: 6469 6374 5f70 6964 5f4f 4253 5f4f 4253  dict_pid_OBS_OBS
-00013620: 290a 2020 2020 6466 5f62 6c6f 7375 6d5f  ).    df_blosum_
-00013630: 4f42 535f 4f42 5320 3d20 7064 2e44 6174  OBS_OBS = pd.Dat
-00013640: 6146 7261 6d65 2864 6963 745f 626c 6f73  aFrame(dict_blos
-00013650: 756d 5f4f 4253 5f4f 4253 290a 0a20 2020  um_OBS_OBS)..   
-00013660: 2023 6466 5f6c 7420 3d20 6466 5f4f 4253   #df_lt = df_OBS
-00013670: 5f4f 4253 2e77 6865 7265 286e 702e 7472  _OBS.where(np.tr
-00013680: 696c 286e 702e 6f6e 6573 2864 665f 4f42  il(np.ones(df_OB
-00013690: 535f 4f42 532e 7368 6170 6529 292e 6173  S_OBS.shape)).as
-000136a0: 7479 7065 286e 702e 626f 6f6c 2929 0a20  type(np.bool)). 
-000136b0: 2020 2061 6e6e 6f74 203d 2046 616c 7365     annot = False
-000136c0: 0a20 2020 2050 6c6f 745f 4865 6174 6d61  .    Plot_Heatma
-000136d0: 7028 6466 5f70 6964 5f4f 4253 5f4f 4253  p(df_pid_OBS_OBS
-000136e0: 2c22 4f42 5320 7365 7173 2076 7320 4f42  ,"OBS seqs vs OB
-000136f0: 5320 7365 7173 2025 4944 222c 2250 4944  S seqs %ID","PID
-00013700: 5f7b 7d5f 4f42 535f 4f42 5322 2e66 6f72  _{}_OBS_OBS".for
-00013710: 6d61 7428 666f 6c64 6572 292c 616e 6e6f  mat(folder),anno
-00013720: 743d 616e 6e6f 742c 6d61 736b 3d4e 6f6e  t=annot,mask=Non
-00013730: 652c 766d 6178 3d31 3030 290a 2020 2020  e,vmax=100).    
-00013740: 506c 6f74 5f48 6561 746d 6170 2864 665f  Plot_Heatmap(df_
-00013750: 7069 645f 5341 4d50 4c45 445f 4f42 532c  pid_SAMPLED_OBS,
-00013760: 2022 4f42 5320 7673 2053 414d 504c 4544   "OBS vs SAMPLED
-00013770: 2073 6571 7320 4156 4552 4147 4520 2549   seqs AVERAGE %I
-00013780: 4422 2c20 2250 4944 5f7b 7d5f 4f42 535f  D", "PID_{}_OBS_
-00013790: 5341 4d50 4c45 4422 2e66 6f72 6d61 7428  SAMPLED".format(
-000137a0: 666f 6c64 6572 292c 616e 6e6f 743d 616e  folder),annot=an
-000137b0: 6e6f 742c 6d61 736b 3d4e 6f6e 652c 766d  not,mask=None,vm
-000137c0: 6178 3d31 3030 290a 2020 2020 506c 6f74  ax=100).    Plot
-000137d0: 5f48 6561 746d 6170 2864 665f 7069 645f  _Heatmap(df_pid_
-000137e0: 5341 4d50 4c45 445f 5341 4d50 4c45 442c  SAMPLED_SAMPLED,
-000137f0: 2022 5341 4d50 4c45 4420 7673 2053 414d   "SAMPLED vs SAM
-00013800: 504c 4544 2073 6571 7320 4156 4552 4147  PLED seqs AVERAG
-00013810: 4520 2549 4422 2c20 2250 4944 5f7b 7d5f  E %ID", "PID_{}_
-00013820: 5341 4d50 4c45 445f 5341 4d50 4c45 4422  SAMPLED_SAMPLED"
-00013830: 2e66 6f72 6d61 7428 666f 6c64 6572 292c  .format(folder),
-00013840: 616e 6e6f 743d 616e 6e6f 742c 6d61 736b  annot=annot,mask
-00013850: 3d4e 6f6e 652c 766d 6178 3d31 3030 290a  =None,vmax=100).
-00013860: 0a20 2020 2076 6d61 7820 3d20 6d61 7828  .    vmax = max(
-00013870: 5b64 665f 626c 6f73 756d 5f4f 4253 5f4f  [df_blosum_OBS_O
-00013880: 4253 2e6d 6178 2829 2e6d 6178 2829 2c64  BS.max().max(),d
-00013890: 665f 626c 6f73 756d 5f53 414d 504c 4544  f_blosum_SAMPLED
-000138a0: 5f53 414d 504c 4544 2e6d 6178 2829 2e6d  _SAMPLED.max().m
-000138b0: 6178 2829 2c64 665f 626c 6f73 756d 5f53  ax(),df_blosum_S
-000138c0: 414d 504c 4544 5f4f 4253 2e6d 6178 2829  AMPLED_OBS.max()
-000138d0: 2e6d 6178 2829 5d29 0a20 2020 2050 6c6f  .max()]).    Plo
-000138e0: 745f 4865 6174 6d61 7028 6466 5f62 6c6f  t_Heatmap(df_blo
-000138f0: 7375 6d5f 4f42 535f 4f42 532c 2022 4f42  sum_OBS_OBS, "OB
-00013900: 5320 7365 7173 2076 7320 4f42 5320 7365  S seqs vs OBS se
-00013910: 7173 2042 6c6f 7375 6d20 5363 6f72 6522  qs Blosum Score"
-00013920: 2c20 2242 6c6f 7375 6d5f 7b7d 5f4f 4253  , "Blosum_{}_OBS
-00013930: 5f4f 4253 222e 666f 726d 6174 2866 6f6c  _OBS".format(fol
-00013940: 6465 7229 2c61 6e6e 6f74 3d61 6e6e 6f74  der),annot=annot
-00013950: 2c6d 6173 6b3d 4e6f 6e65 2c76 6d61 783d  ,mask=None,vmax=
-00013960: 766d 6178 290a 2020 2020 506c 6f74 5f48  vmax).    Plot_H
-00013970: 6561 746d 6170 2864 665f 626c 6f73 756d  eatmap(df_blosum
-00013980: 5f53 414d 504c 4544 5f4f 4253 2c20 224f  _SAMPLED_OBS, "O
-00013990: 4253 2076 7320 5341 4d50 4c45 4420 7365  BS vs SAMPLED se
-000139a0: 7173 2041 5645 5241 4745 2042 6c6f 7375  qs AVERAGE Blosu
-000139b0: 6d20 5363 6f72 6522 2c20 2242 6c6f 7375  m Score", "Blosu
-000139c0: 6d5f 7b7d 5f4f 4253 5f53 414d 504c 4544  m_{}_OBS_SAMPLED
-000139d0: 222e 666f 726d 6174 2866 6f6c 6465 7229  ".format(folder)
-000139e0: 2c61 6e6e 6f74 3d61 6e6e 6f74 2c6d 6173  ,annot=annot,mas
-000139f0: 6b3d 4e6f 6e65 2c76 6d61 783d 766d 6178  k=None,vmax=vmax
-00013a00: 290a 2020 2020 506c 6f74 5f48 6561 746d  ).    Plot_Heatm
-00013a10: 6170 2864 665f 626c 6f73 756d 5f53 414d  ap(df_blosum_SAM
-00013a20: 504c 4544 5f53 414d 504c 4544 2c20 2253  PLED_SAMPLED, "S
-00013a30: 414d 504c 4544 2076 7320 5341 4d50 4c45  AMPLED vs SAMPLE
-00013a40: 4420 7365 7173 2041 5645 5241 4745 2042  D seqs AVERAGE B
-00013a50: 6c6f 7375 6d20 5363 6f72 6522 2c20 2242  losum Score", "B
-00013a60: 6c6f 7375 6d5f 7b7d 5f53 414d 504c 4544  losum_{}_SAMPLED
-00013a70: 5f53 414d 504c 4544 222e 666f 726d 6174  _SAMPLED".format
-00013a80: 2866 6f6c 6465 7229 2c61 6e6e 6f74 3d61  (folder),annot=a
-00013a90: 6e6e 6f74 2c6d 6173 6b3d 4e6f 6e65 2c76  nnot,mask=None,v
-00013aa0: 6d61 7820 3d20 766d 6178 290a 6465 6620  max = vmax).def 
-00013ab0: 6275 696c 645f 6461 7461 6672 616d 6573  build_dataframes
-00013ac0: 5f6f 7665 726c 6170 7069 6e67 5f68 6973  _overlapping_his
-00013ad0: 746f 6772 616d 7328 6161 5f73 6571 7565  tograms(aa_seque
-00013ae0: 6e63 6573 5f70 7265 6469 6374 696f 6e73  nces_predictions
-00013af0: 2c20 6461 7461 7365 745f 7472 6169 6e2c  , dataset_train,
-00013b00: 6461 7461 7365 745f 7465 7374 2c20 6e61  dataset_test, na
-00013b10: 6d65 2c20 6e5f 7361 6d70 6c65 732c 6e6f  me, n_samples,no
-00013b20: 6465 735f 696e 6465 7865 732c 7265 7375  des_indexes,resu
-00013b30: 6c74 735f 6469 7265 6374 6f72 792c 6161  lts_directory,aa
-00013b40: 5f70 726f 6273 2c63 6f72 7265 7370 6f6e  _probs,correspon
-00013b50: 6465 6e63 655f 6469 6374 3d4e 6f6e 6529  dence_dict=None)
-00013b60: 3a0a 2020 2020 2222 2250 6c6f 7420 6869  :.    """Plot hi
-00013b70: 7374 6f67 7261 6d20 7768 6f73 6520 6261  stogram whose ba
-00013b80: 7273 2072 6570 7265 7365 6e74 2074 6865  rs represent the
-00013b90: 2070 6572 6365 6e74 6167 6520 6964 656e   percentage iden
-00013ba0: 7469 7479 2062 6574 7765 656e 2074 6865  tity between the
-00013bb0: 2074 7275 6520 616e 6365 7374 6f72 7320   true ancestors 
-00013bc0: 7673 2074 7275 6520 6c65 6176 6573 2061  vs true leaves a
-00013bd0: 6e64 2074 6865 2070 7265 6469 6374 6564  nd the predicted
-00013be0: 2061 6e63 6573 746f 7273 2076 7320 7472   ancestors vs tr
-00013bf0: 7565 206c 6561 7665 730a 2020 2020 3a70  ue leaves.    :p
-00013c00: 6172 616d 2074 656e 736f 7220 6161 5f73  aram tensor aa_s
-00013c10: 6571 7565 6e63 6573 5f70 7265 6469 6374  equences_predict
-00013c20: 696f 6e73 3a0a 2020 2020 3a70 6172 616d  ions:.    :param
-00013c30: 2074 656e 736f 7220 6461 7461 7365 745f   tensor dataset_
-00013c40: 7472 6169 6e0a 2020 2020 3a70 6172 616d  train.    :param
-00013c50: 2074 656e 736f 7220 6461 7461 7365 745f   tensor dataset_
-00013c60: 7465 7374 0a20 2020 203a 7061 7261 6d20  test.    :param 
-00013c70: 7374 7220 6e61 6d65 3a20 6461 7461 7365  str name: datase
-00013c80: 7420 6e61 6d65 2070 726f 6a65 6374 0a20  t name project. 
-00013c90: 2020 203a 7061 7261 6d20 7465 6e73 6f72     :param tensor
-00013ca0: 2061 615f 7365 7175 656e 6365 735f 7072   aa_sequences_pr
-00013cb0: 6564 6963 7469 6f6e 730a 2020 2020 3a70  edictions.    :p
-00013cc0: 6172 616d 2069 6e74 206e 5f73 616d 706c  aram int n_sampl
-00013cd0: 6573 3a20 6e75 6d62 6572 206f 6620 7361  es: number of sa
-00013ce0: 6d70 6c65 6420 7072 6564 6963 7469 6f6e  mpled prediction
-00013cf0: 730a 2020 2020 3a70 6172 616d 206e 6f64  s.    :param nod
-00013d00: 6573 5f69 6e64 6578 6573 3a0a 2020 2020  es_indexes:.    
-00013d10: 3a70 6172 616d 2073 7472 2072 6573 756c  :param str resul
-00013d20: 7473 5f64 6972 6563 746f 7279 0a20 2020  ts_directory.   
-00013d30: 203a 7061 7261 6d20 696e 7420 6161 5f70   :param int aa_p
-00013d40: 726f 623a 2061 6d69 6e6f 2061 6369 6420  rob: amino acid 
-00013d50: 7072 6f62 6162 696c 6974 6965 7320 2869  probabilities (i
-00013d60: 2e65 2032 3129 0a20 2020 203a 7061 7261  .e 21).    :para
-00013d70: 6d20 6469 6374 2063 6f72 7265 7370 6f6e  m dict correspon
-00013d80: 6465 6e63 655f 6469 6374 3a20 636f 7272  dence_dict: corr
-00013d90: 6573 706f 6e64 656e 6365 2062 6574 7765  espondence betwe
-00013da0: 656e 2074 6865 2074 7265 6520 6c65 7665  en the tree leve
-00013db0: 6c20 6f72 6465 7220 696e 6465 7865 7320  l order indexes 
-00013dc0: 616e 6420 7468 6520 2d2d 3e20 6f6e 6c79  and the --> only
-00013dd0: 2075 7365 6675 6c20 666f 7220 6265 6e63   useful for benc
-00013de0: 686d 6172 6b5f 7261 6e64 616c 6c5f 6f72  hmark_randall_or
-00013df0: 6967 696e 616c 5f6e 616d 696e 670a 2020  iginal_naming.  
-00013e00: 2020 2222 220a 2020 2020 2268 7474 7073    """.    "https
-00013e10: 3a2f 2f6d 6174 706c 6f74 6c69 622e 6f72  ://matplotlib.or
-00013e20: 672f 332e 312e 312f 6761 6c6c 6572 792f  g/3.1.1/gallery/
-00013e30: 756e 6974 732f 6261 725f 756e 6974 5f64  units/bar_unit_d
-00013e40: 656d 6f2e 6874 6d6c 2373 7068 782d 676c  emo.html#sphx-gl
-00013e50: 722d 6761 6c6c 6572 792d 756e 6974 732d  r-gallery-units-
-00013e60: 6261 722d 756e 6974 2d64 656d 6f2d 7079  bar-unit-demo-py
-00013e70: 220a 2020 2020 7072 696e 7428 2242 7569  ".    print("Bui
-00013e80: 6c64 696e 6720 4f76 6572 6c61 7070 696e  lding Overlappin
-00013e90: 6720 4869 7374 6f67 7261 6d2e 2e2e 2229  g Histogram...")
-00013ea0: 0a20 2020 206e 5f63 6869 6c64 7265 6e20  .    n_children 
-00013eb0: 3d20 6c65 6e28 6e6f 6465 735f 696e 6465  = len(nodes_inde
-00013ec0: 7865 7329 0a20 2020 2064 6566 2050 6572  xes).    def Per
-00013ed0: 6365 6e74 5f49 445f 4f42 535f 4f42 5328  cent_ID_OBS_OBS(
-00013ee0: 293a 0a20 2020 2020 2020 2022 4765 6e65  ):.        "Gene
-00013ef0: 7261 7465 2074 6865 2020 2549 4420 6f66  rate the  %ID of
-00013f00: 2074 6865 204f 4253 2054 4553 5420 7365   the OBS TEST se
-00013f10: 7175 656e 6365 7320 6167 6169 6e73 7420  quences against 
-00013f20: 7468 6520 4f42 5320 7472 6169 6e22 0a20  the OBS train". 
-00013f30: 2020 2020 2020 2070 6572 6365 6e74 5f69         percent_i
-00013f40: 645f 4f42 5320 3d20 6469 6374 2e66 726f  d_OBS = dict.fro
-00013f50: 6d6b 6579 7328 6461 7461 7365 745f 7472  mkeys(dataset_tr
-00013f60: 6169 6e5b 3a2c 302c 315d 2e74 6f6c 6973  ain[:,0,1].tolis
-00013f70: 7428 292c 2064 6963 742e 6672 6f6d 6b65  t(), dict.fromke
-00013f80: 7973 2864 6174 6173 6574 5f74 6573 745b  ys(dataset_test[
-00013f90: 3a2c 302c 315d 2e74 6f6c 6973 7428 292c  :,0,1].tolist(),
-00013fa0: 5b5d 2929 0a20 2020 2020 2020 2066 6f72  [])).        for
-00013fb0: 2069 2c74 7261 696e 7365 7120 696e 2065   i,trainseq in e
-00013fc0: 6e75 6d65 7261 7465 2864 6174 6173 6574  numerate(dataset
-00013fd0: 5f74 7261 696e 293a 0a20 2020 2020 2020  _train):.       
-00013fe0: 2020 2020 2073 6571 5f54 5241 494e 5f6f       seq_TRAIN_o
-00013ff0: 6273 5f6c 6574 7465 7273 5f69 203d 2063  bs_letters_i = c
-00014000: 6f6e 7665 7274 5f74 6f5f 6c65 7474 6572  onvert_to_letter
-00014010: 7328 6461 7461 7365 745f 7472 6169 6e5b  s(dataset_train[
-00014020: 692c 2032 3a2c 305d 2c61 615f 7072 6f62  i, 2:,0],aa_prob
-00014030: 7329 0a20 2020 2020 2020 2020 2020 2066  s).            f
-00014040: 6f72 206a 2c74 6573 7473 6571 2069 6e20  or j,testseq in 
-00014050: 656e 756d 6572 6174 6528 6461 7461 7365  enumerate(datase
-00014060: 745f 7465 7374 293a 0a20 2020 2020 2020  t_test):.       
-00014070: 2020 2020 2020 2020 2073 6571 5f54 4553           seq_TES
-00014080: 545f 6f62 735f 6c65 7474 6572 735f 6920  T_obs_letters_i 
-00014090: 3d20 636f 6e76 6572 745f 746f 5f6c 6574  = convert_to_let
-000140a0: 7465 7273 2864 6174 6173 6574 5f74 6573  ters(dataset_tes
-000140b0: 745b 6a2c 2032 3a2c 305d 2c61 615f 7072  t[j, 2:,0],aa_pr
-000140c0: 6f62 7329 0a20 2020 2020 2020 2020 2020  obs).           
-000140d0: 2020 2020 2070 6964 203d 7065 7263 5f69       pid =perc_i
-000140e0: 6465 6e74 6974 795f 7061 6972 5f73 6571  dentity_pair_seq
-000140f0: 2873 6571 5f54 5241 494e 5f6f 6273 5f6c  (seq_TRAIN_obs_l
-00014100: 6574 7465 7273 5f69 2c73 6571 5f54 4553  etters_i,seq_TES
-00014110: 545f 6f62 735f 6c65 7474 6572 735f 6929  T_obs_letters_i)
-00014120: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00014130: 2070 6572 6365 6e74 5f69 645f 4f42 535b   percent_id_OBS[
-00014140: 6461 7461 7365 745f 7472 6169 6e5b 692c  dataset_train[i,
-00014150: 302c 315d 2e69 7465 6d28 295d 5b64 6174  0,1].item()][dat
-00014160: 6173 6574 5f74 6573 745b 6a2c 302c 315d  aset_test[j,0,1]
-00014170: 2e69 7465 6d28 295d 203d 2070 6964 0a20  .item()] = pid. 
-00014180: 2020 2020 2020 2072 6574 7572 6e20 7065         return pe
-00014190: 7263 656e 745f 6964 5f4f 4253 0a20 2020  rcent_id_OBS.   
-000141a0: 2064 6566 2050 6572 6365 6e74 5f49 445f   def Percent_ID_
-000141b0: 5052 4544 5f4f 4253 2829 3a0a 2020 2020  PRED_OBS():.    
-000141c0: 2020 2020 2247 656e 6572 6174 6520 7468      "Generate th
-000141d0: 6520 4176 6572 6167 6520 616e 6420 5354  e Average and ST
-000141e0: 4420 2549 4420 6f66 2074 6865 2073 616d  D %ID of the sam
-000141f0: 706c 6564 2073 6571 7565 6e63 6573 2061  pled sequences a
-00014200: 6761 696e 7374 2074 6865 204f 4253 2074  gainst the OBS t
-00014210: 7261 696e 220a 2020 2020 2020 2020 2370  rain".        #p
-00014220: 6572 6365 6e74 5f69 645f 5052 4544 203d  ercent_id_PRED =
-00014230: 2064 6963 742e 6672 6f6d 6b65 7973 2844   dict.fromkeys(D
-00014240: 6174 6173 6574 5f74 7261 696e 5b3a 2c30  ataset_train[:,0
-00014250: 2c31 5d2e 746f 6c69 7374 2829 2c20 6469  ,1].tolist(), di
-00014260: 6374 2e66 726f 6d6b 6579 7328 7072 6564  ct.fromkeys(pred
-00014270: 6963 7469 6f6e 735f 7361 6d70 6c65 735b  ictions_samples[
-00014280: 302c 3a2c 315d 2e74 6f6c 6973 7428 292c  0,:,1].tolist(),
-00014290: 7b22 4176 6572 6167 6522 3a5b 5d2c 2253  {"Average":[],"S
-000142a0: 7464 223a 5b5d 7d29 2920 237b 2241 7665  td":[]})) #{"Ave
-000142b0: 7261 6765 223a 5b5d 2c22 5374 6422 3a5b  rage":[],"Std":[
-000142c0: 5d7d 0a20 2020 2020 2020 2070 6572 6365  ]}.        perce
-000142d0: 6e74 5f69 645f 5052 4544 203d 2064 6566  nt_id_PRED = def
-000142e0: 6175 6c74 6469 6374 286c 616d 6264 613a  aultdict(lambda:
-000142f0: 2064 6566 6175 6c74 6469 6374 286c 616d   defaultdict(lam
-00014300: 6264 613a 2064 6566 6175 6c74 6469 6374  bda: defaultdict
-00014310: 2864 6963 7429 2929 0a20 2020 2020 2020  (dict))).       
-00014320: 2066 6f72 2069 2c20 7365 7120 696e 2065   for i, seq in e
-00014330: 6e75 6d65 7261 7465 2864 6174 6173 6574  numerate(dataset
-00014340: 5f74 7261 696e 293a 0a20 2020 2020 2020  _train):.       
-00014350: 2020 2020 2073 6571 5f54 5241 494e 5f6f       seq_TRAIN_o
-00014360: 6273 5f6c 6574 7465 7273 5f69 203d 2063  bs_letters_i = c
-00014370: 6f6e 7665 7274 5f74 6f5f 6c65 7474 6572  onvert_to_letter
-00014380: 7328 6461 7461 7365 745f 7472 6169 6e5b  s(dataset_train[
-00014390: 692c 2032 3a2c 305d 2c61 615f 7072 6f62  i, 2:,0],aa_prob
-000143a0: 7329 0a20 2020 2020 2020 2020 2020 2066  s).            f
-000143b0: 6f72 206a 2069 6e20 7261 6e67 6528 6e5f  or j in range(n_
-000143c0: 6368 696c 6472 656e 293a 2366 6f72 2074  children):#for t
-000143d0: 6573 7420 6e6f 6465 2023 6e5f 6368 696c  est node #n_chil
-000143e0: 6472 656e 0a20 2020 2020 2020 2020 2020  dren.           
-000143f0: 2020 2020 2061 6c6c 5f73 616d 706c 6564       all_sampled
-00014400: 5f69 6e64 6578 203d 2061 615f 7365 7175  _index = aa_sequ
-00014410: 656e 6365 735f 7072 6564 6963 7469 6f6e  ences_prediction
-00014420: 735b 3a2c 6a2c 333a 5d20 2023 2041 6c6c  s[:,j,3:]  # All
-00014430: 2073 616d 706c 6573 2066 6f72 2073 616d   samples for sam
-00014440: 6520 7465 7374 2073 6571 0a20 2020 2020  e test seq.     
-00014450: 2020 2020 2020 2020 2020 2070 6572 6365             perce
-00014460: 6e74 5f69 645f 693d 5b5d 2023 616c 6c20  nt_id_i=[] #all 
-00014470: 7361 6d70 6c65 7320 666f 7220 7468 6520  samples for the 
-00014480: 7361 6d65 2074 6573 7420 7365 7175 656e  same test sequen
-00014490: 6365 0a20 2020 2020 2020 2020 2020 2020  ce.             
-000144a0: 2020 2066 6f72 2077 2069 6e20 7261 6e67     for w in rang
-000144b0: 6528 6e5f 7361 6d70 6c65 7329 3a0a 2020  e(n_samples):.  
+00012ea0: 2020 2020 2020 2020 2020 2064 7069 3d31             dpi=1
+00012eb0: 3530 290a 2020 2020 2020 2020 706c 742e  50).        plt.
+00012ec0: 636c 6628 290a 2020 2020 2020 2020 706c  clf().        pl
+00012ed0: 742e 636c 6f73 6528 290a 0a20 2020 206d  t.close()..    m
+00012ee0: 756c 7469 696e 6465 783d 4661 6c73 650a  ultiindex=False.
+00012ef0: 2020 2020 6966 206d 756c 7469 696e 6465      if multiinde
+00012f00: 783a 0a20 2020 2020 2020 2023 4869 6768  x:.        #High
+00012f10: 6c69 6768 743a 2073 616d 706c 6564 2076  light: sampled v
+00012f20: 7320 6f62 7365 7276 6564 0a20 2020 2020  s observed.     
+00012f30: 2020 2064 6963 745f 7069 645f 5341 4d50     dict_pid_SAMP
+00012f40: 4c45 445f 4f42 532c 2064 6963 745f 626c  LED_OBS, dict_bl
+00012f50: 6f73 756d 5f53 414d 504c 4544 5f4f 4253  osum_SAMPLED_OBS
+00012f60: 2c64 6963 745f 696e 636f 7272 6563 745f  ,dict_incorrect_
+00012f70: 5341 4d50 4c45 445f 4f42 5320 3d20 5065  SAMPLED_OBS = Pe
+00012f80: 7263 656e 745f 4944 5f54 6573 745f 5341  rcent_ID_Test_SA
+00012f90: 4d50 4c45 445f 4f42 5345 5256 4544 286d  MPLED_OBSERVED(m
+00012fa0: 756c 7469 696e 6465 7829 0a20 2020 2020  ultiindex).     
+00012fb0: 2020 2064 665f 7069 645f 5341 4d50 4c45     df_pid_SAMPLE
+00012fc0: 445f 4f42 5320 3d20 7064 2e63 6f6e 6361  D_OBS = pd.conca
+00012fd0: 7428 7b6b 3a20 7064 2e44 6174 6146 7261  t({k: pd.DataFra
+00012fe0: 6d65 2e66 726f 6d5f 6469 6374 2876 2c20  me.from_dict(v, 
+00012ff0: 2769 6e64 6578 2729 2066 6f72 206b 2c20  'index') for k, 
+00013000: 7620 696e 2064 6963 745f 7069 645f 5341  v in dict_pid_SA
+00013010: 4d50 4c45 445f 4f42 532e 6974 656d 7328  MPLED_OBS.items(
+00013020: 297d 2c20 6178 6973 3d30 290a 2020 2020  )}, axis=0).    
+00013030: 2020 2020 6466 5f62 6c6f 7375 6d5f 5341      df_blosum_SA
+00013040: 4d50 4c45 445f 4f42 5320 3d20 7064 2e63  MPLED_OBS = pd.c
+00013050: 6f6e 6361 7428 7b6b 3a20 7064 2e44 6174  oncat({k: pd.Dat
+00013060: 6146 7261 6d65 2e66 726f 6d5f 6469 6374  aFrame.from_dict
+00013070: 2876 2c20 2769 6e64 6578 2729 2066 6f72  (v, 'index') for
+00013080: 206b 2c20 7620 696e 2064 6963 745f 626c   k, v in dict_bl
+00013090: 6f73 756d 5f53 414d 504c 4544 5f4f 4253  osum_SAMPLED_OBS
+000130a0: 2e69 7465 6d73 2829 7d2c 6178 6973 3d30  .items()},axis=0
+000130b0: 290a 2020 2020 2020 2020 2364 665f 696e  ).        #df_in
+000130c0: 636f 7272 6563 745f 5341 4d50 4c45 445f  correct_SAMPLED_
+000130d0: 4f42 5320 3d20 7064 2e63 6f6e 6361 7428  OBS = pd.concat(
+000130e0: 7b6b 3a20 7064 2e44 6174 6146 7261 6d65  {k: pd.DataFrame
+000130f0: 2e66 726f 6d5f 6469 6374 2876 2c20 2769  .from_dict(v, 'i
+00013100: 6e64 6578 2729 2066 6f72 206b 2c20 7620  ndex') for k, v 
+00013110: 696e 2064 6963 745f 696e 636f 7272 6563  in dict_incorrec
+00013120: 745f 5341 4d50 4c45 445f 4f42 532e 6974  t_SAMPLED_OBS.it
+00013130: 656d 7328 297d 2c20 6178 6973 3d30 290a  ems()}, axis=0).
+00013140: 2020 2020 2020 2020 2348 6967 686c 6967          #Highlig
+00013150: 6874 3a20 5361 6d70 6c65 6420 7673 2053  ht: Sampled vs S
+00013160: 616d 706c 6564 0a20 2020 2020 2020 2064  ampled.        d
+00013170: 6963 745f 7069 645f 5341 4d50 4c45 445f  ict_pid_SAMPLED_
+00013180: 5341 4d50 4c45 442c 6469 6374 5f62 6c6f  SAMPLED,dict_blo
+00013190: 7375 6d5f 5341 4d50 4c45 445f 5341 4d50  sum_SAMPLED_SAMP
+000131a0: 4c45 443d 5065 7263 656e 745f 4944 5f54  LED=Percent_ID_T
+000131b0: 6573 745f 5341 4d50 4c45 445f 5341 4d50  est_SAMPLED_SAMP
+000131c0: 4c45 4428 6d75 6c74 6969 6e64 6578 290a  LED(multiindex).
+000131d0: 2020 2020 2020 2020 6466 5f70 6964 5f53          df_pid_S
+000131e0: 414d 504c 4544 5f53 414d 504c 4544 203d  AMPLED_SAMPLED =
+000131f0: 2070 642e 636f 6e63 6174 287b 6b3a 2070   pd.concat({k: p
+00013200: 642e 4461 7461 4672 616d 652e 6672 6f6d  d.DataFrame.from
+00013210: 5f64 6963 7428 762c 2027 696e 6465 7827  _dict(v, 'index'
+00013220: 2920 666f 7220 6b2c 2076 2069 6e20 6469  ) for k, v in di
+00013230: 6374 5f70 6964 5f53 414d 504c 4544 5f53  ct_pid_SAMPLED_S
+00013240: 414d 504c 4544 2e69 7465 6d73 2829 7d2c  AMPLED.items()},
+00013250: 2061 7869 733d 3029 0a20 2020 2020 2020   axis=0).       
+00013260: 2064 665f 626c 6f73 756d 5f53 414d 504c   df_blosum_SAMPL
+00013270: 4544 5f53 414d 504c 4544 203d 2070 642e  ED_SAMPLED = pd.
+00013280: 636f 6e63 6174 287b 6b3a 2070 642e 4461  concat({k: pd.Da
+00013290: 7461 4672 616d 652e 6672 6f6d 5f64 6963  taFrame.from_dic
+000132a0: 7428 762c 2027 696e 6465 7827 2920 666f  t(v, 'index') fo
+000132b0: 7220 6b2c 2076 2069 6e20 6469 6374 5f62  r k, v in dict_b
+000132c0: 6c6f 7375 6d5f 5341 4d50 4c45 445f 5341  losum_SAMPLED_SA
+000132d0: 4d50 4c45 442e 6974 656d 7328 297d 2c20  MPLED.items()}, 
+000132e0: 6178 6973 3d30 290a 0a20 2020 2065 6c73  axis=0)..    els
+000132f0: 653a 0a20 2020 2020 2020 2023 2048 6967  e:.        # Hig
+00013300: 686c 6967 6874 3a20 7361 6d70 6c65 6420  hlight: sampled 
+00013310: 7673 206f 6273 6572 7665 640a 2020 2020  vs observed.    
+00013320: 2020 2020 6469 6374 5f70 6964 5f53 414d      dict_pid_SAM
+00013330: 504c 4544 5f4f 4253 2c20 6469 6374 5f62  PLED_OBS, dict_b
+00013340: 6c6f 7375 6d5f 5341 4d50 4c45 445f 4f42  losum_SAMPLED_OB
+00013350: 532c 6469 6374 5f69 6e63 6f72 7265 6374  S,dict_incorrect
+00013360: 5f53 414d 504c 4544 5f4f 4253 203d 2050  _SAMPLED_OBS = P
+00013370: 6572 6365 6e74 5f49 445f 5465 7374 5f53  ercent_ID_Test_S
+00013380: 414d 504c 4544 5f4f 4253 4552 5645 4428  AMPLED_OBSERVED(
+00013390: 6d75 6c74 6969 6e64 6578 290a 2020 2020  multiindex).    
+000133a0: 2020 2020 6466 5f70 6964 5f53 414d 504c      df_pid_SAMPL
+000133b0: 4544 5f4f 4253 203d 2070 642e 4461 7461  ED_OBS = pd.Data
+000133c0: 4672 616d 6528 6469 6374 5f70 6964 5f53  Frame(dict_pid_S
+000133d0: 414d 504c 4544 5f4f 4253 290a 2020 2020  AMPLED_OBS).    
+000133e0: 2020 2020 6466 5f62 6c6f 7375 6d5f 5341      df_blosum_SA
+000133f0: 4d50 4c45 445f 4f42 5320 3d20 7064 2e44  MPLED_OBS = pd.D
+00013400: 6174 6146 7261 6d65 2864 6963 745f 626c  ataFrame(dict_bl
+00013410: 6f73 756d 5f53 414d 504c 4544 5f4f 4253  osum_SAMPLED_OBS
+00013420: 290a 2020 2020 2020 2020 2364 665f 696e  ).        #df_in
+00013430: 636f 7272 6563 745f 5341 4d50 4c45 445f  correct_SAMPLED_
+00013440: 4f42 5320 3d20 7064 2e44 6174 6146 7261  OBS = pd.DataFra
+00013450: 6d65 2864 6963 745f 696e 636f 7272 6563  me(dict_incorrec
+00013460: 745f 5341 4d50 4c45 445f 4f42 5329 2e74  t_SAMPLED_OBS).t
+00013470: 6f5f 6e75 6d70 7928 290a 2020 2020 2020  o_numpy().      
+00013480: 2020 2348 6967 686c 6967 6874 3a20 5361    #Highlight: Sa
+00013490: 6d70 6c65 6420 7673 2053 616d 706c 6564  mpled vs Sampled
+000134a0: 0a20 2020 2020 2020 2064 6963 745f 7069  .        dict_pi
+000134b0: 645f 5341 4d50 4c45 445f 5341 4d50 4c45  d_SAMPLED_SAMPLE
+000134c0: 442c 2064 6963 745f 626c 6f73 756d 5f53  D, dict_blosum_S
+000134d0: 414d 504c 4544 5f53 414d 504c 4544 203d  AMPLED_SAMPLED =
+000134e0: 2050 6572 6365 6e74 5f49 445f 5465 7374   Percent_ID_Test
+000134f0: 5f53 414d 504c 4544 5f53 414d 504c 4544  _SAMPLED_SAMPLED
+00013500: 286d 756c 7469 696e 6465 7829 0a20 2020  (multiindex).   
+00013510: 2020 2020 2064 665f 7069 645f 5341 4d50       df_pid_SAMP
+00013520: 4c45 445f 5341 4d50 4c45 4420 3d20 7064  LED_SAMPLED = pd
+00013530: 2e44 6174 6146 7261 6d65 2864 6963 745f  .DataFrame(dict_
+00013540: 7069 645f 5341 4d50 4c45 445f 5341 4d50  pid_SAMPLED_SAMP
+00013550: 4c45 4429 0a20 2020 2020 2020 2064 665f  LED).        df_
+00013560: 626c 6f73 756d 5f53 414d 504c 4544 5f53  blosum_SAMPLED_S
+00013570: 414d 504c 4544 203d 2070 642e 4461 7461  AMPLED = pd.Data
+00013580: 4672 616d 6528 6469 6374 5f62 6c6f 7375  Frame(dict_blosu
+00013590: 6d5f 5341 4d50 4c45 445f 5341 4d50 4c45  m_SAMPLED_SAMPLE
+000135a0: 4429 0a0a 0a20 2020 2064 6963 745f 7069  D)...    dict_pi
+000135b0: 645f 4f42 535f 4f42 532c 6469 6374 5f62  d_OBS_OBS,dict_b
+000135c0: 6c6f 7375 6d5f 4f42 535f 4f42 5320 3d20  losum_OBS_OBS = 
+000135d0: 5065 7263 656e 745f 4944 5f54 6573 745f  Percent_ID_Test_
+000135e0: 4f42 535f 4f42 5345 5256 4544 2829 0a20  OBS_OBSERVED(). 
+000135f0: 2020 2064 665f 7069 645f 4f42 535f 4f42     df_pid_OBS_OB
+00013600: 5320 3d20 7064 2e44 6174 6146 7261 6d65  S = pd.DataFrame
+00013610: 2864 6963 745f 7069 645f 4f42 535f 4f42  (dict_pid_OBS_OB
+00013620: 5329 0a20 2020 2064 665f 626c 6f73 756d  S).    df_blosum
+00013630: 5f4f 4253 5f4f 4253 203d 2070 642e 4461  _OBS_OBS = pd.Da
+00013640: 7461 4672 616d 6528 6469 6374 5f62 6c6f  taFrame(dict_blo
+00013650: 7375 6d5f 4f42 535f 4f42 5329 0a0a 2020  sum_OBS_OBS)..  
+00013660: 2020 2364 665f 6c74 203d 2064 665f 4f42    #df_lt = df_OB
+00013670: 535f 4f42 532e 7768 6572 6528 6e70 2e74  S_OBS.where(np.t
+00013680: 7269 6c28 6e70 2e6f 6e65 7328 6466 5f4f  ril(np.ones(df_O
+00013690: 4253 5f4f 4253 2e73 6861 7065 2929 2e61  BS_OBS.shape)).a
+000136a0: 7374 7970 6528 6e70 2e62 6f6f 6c29 290a  stype(np.bool)).
+000136b0: 2020 2020 616e 6e6f 7420 3d20 4661 6c73      annot = Fals
+000136c0: 650a 2020 2020 506c 6f74 5f48 6561 746d  e.    Plot_Heatm
+000136d0: 6170 2864 665f 7069 645f 4f42 535f 4f42  ap(df_pid_OBS_OB
+000136e0: 532c 224f 4253 2073 6571 7320 7673 204f  S,"OBS seqs vs O
+000136f0: 4253 2073 6571 7320 2549 4422 2c22 5049  BS seqs %ID","PI
+00013700: 445f 7b7d 5f4f 4253 5f4f 4253 222e 666f  D_{}_OBS_OBS".fo
+00013710: 726d 6174 2866 6f6c 6465 7229 2c61 6e6e  rmat(folder),ann
+00013720: 6f74 3d61 6e6e 6f74 2c6d 6173 6b3d 4e6f  ot=annot,mask=No
+00013730: 6e65 2c76 6d61 783d 3130 3029 0a20 2020  ne,vmax=100).   
+00013740: 2050 6c6f 745f 4865 6174 6d61 7028 6466   Plot_Heatmap(df
+00013750: 5f70 6964 5f53 414d 504c 4544 5f4f 4253  _pid_SAMPLED_OBS
+00013760: 2c20 224f 4253 2076 7320 5341 4d50 4c45  , "OBS vs SAMPLE
+00013770: 4420 7365 7173 2041 5645 5241 4745 2025  D seqs AVERAGE %
+00013780: 4944 222c 2022 5049 445f 7b7d 5f4f 4253  ID", "PID_{}_OBS
+00013790: 5f53 414d 504c 4544 222e 666f 726d 6174  _SAMPLED".format
+000137a0: 2866 6f6c 6465 7229 2c61 6e6e 6f74 3d61  (folder),annot=a
+000137b0: 6e6e 6f74 2c6d 6173 6b3d 4e6f 6e65 2c76  nnot,mask=None,v
+000137c0: 6d61 783d 3130 3029 0a20 2020 2050 6c6f  max=100).    Plo
+000137d0: 745f 4865 6174 6d61 7028 6466 5f70 6964  t_Heatmap(df_pid
+000137e0: 5f53 414d 504c 4544 5f53 414d 504c 4544  _SAMPLED_SAMPLED
+000137f0: 2c20 2253 414d 504c 4544 2076 7320 5341  , "SAMPLED vs SA
+00013800: 4d50 4c45 4420 7365 7173 2041 5645 5241  MPLED seqs AVERA
+00013810: 4745 2025 4944 222c 2022 5049 445f 7b7d  GE %ID", "PID_{}
+00013820: 5f53 414d 504c 4544 5f53 414d 504c 4544  _SAMPLED_SAMPLED
+00013830: 222e 666f 726d 6174 2866 6f6c 6465 7229  ".format(folder)
+00013840: 2c61 6e6e 6f74 3d61 6e6e 6f74 2c6d 6173  ,annot=annot,mas
+00013850: 6b3d 4e6f 6e65 2c76 6d61 783d 3130 3029  k=None,vmax=100)
+00013860: 0a0a 2020 2020 766d 6178 203d 206d 6178  ..    vmax = max
+00013870: 285b 6466 5f62 6c6f 7375 6d5f 4f42 535f  ([df_blosum_OBS_
+00013880: 4f42 532e 6d61 7828 292e 6d61 7828 292c  OBS.max().max(),
+00013890: 6466 5f62 6c6f 7375 6d5f 5341 4d50 4c45  df_blosum_SAMPLE
+000138a0: 445f 5341 4d50 4c45 442e 6d61 7828 292e  D_SAMPLED.max().
+000138b0: 6d61 7828 292c 6466 5f62 6c6f 7375 6d5f  max(),df_blosum_
+000138c0: 5341 4d50 4c45 445f 4f42 532e 6d61 7828  SAMPLED_OBS.max(
+000138d0: 292e 6d61 7828 295d 290a 2020 2020 506c  ).max()]).    Pl
+000138e0: 6f74 5f48 6561 746d 6170 2864 665f 626c  ot_Heatmap(df_bl
+000138f0: 6f73 756d 5f4f 4253 5f4f 4253 2c20 224f  osum_OBS_OBS, "O
+00013900: 4253 2073 6571 7320 7673 204f 4253 2073  BS seqs vs OBS s
+00013910: 6571 7320 426c 6f73 756d 2053 636f 7265  eqs Blosum Score
+00013920: 222c 2022 426c 6f73 756d 5f7b 7d5f 4f42  ", "Blosum_{}_OB
+00013930: 535f 4f42 5322 2e66 6f72 6d61 7428 666f  S_OBS".format(fo
+00013940: 6c64 6572 292c 616e 6e6f 743d 616e 6e6f  lder),annot=anno
+00013950: 742c 6d61 736b 3d4e 6f6e 652c 766d 6178  t,mask=None,vmax
+00013960: 3d76 6d61 7829 0a20 2020 2050 6c6f 745f  =vmax).    Plot_
+00013970: 4865 6174 6d61 7028 6466 5f62 6c6f 7375  Heatmap(df_blosu
+00013980: 6d5f 5341 4d50 4c45 445f 4f42 532c 2022  m_SAMPLED_OBS, "
+00013990: 4f42 5320 7673 2053 414d 504c 4544 2073  OBS vs SAMPLED s
+000139a0: 6571 7320 4156 4552 4147 4520 426c 6f73  eqs AVERAGE Blos
+000139b0: 756d 2053 636f 7265 222c 2022 426c 6f73  um Score", "Blos
+000139c0: 756d 5f7b 7d5f 4f42 535f 5341 4d50 4c45  um_{}_OBS_SAMPLE
+000139d0: 4422 2e66 6f72 6d61 7428 666f 6c64 6572  D".format(folder
+000139e0: 292c 616e 6e6f 743d 616e 6e6f 742c 6d61  ),annot=annot,ma
+000139f0: 736b 3d4e 6f6e 652c 766d 6178 3d76 6d61  sk=None,vmax=vma
+00013a00: 7829 0a20 2020 2050 6c6f 745f 4865 6174  x).    Plot_Heat
+00013a10: 6d61 7028 6466 5f62 6c6f 7375 6d5f 5341  map(df_blosum_SA
+00013a20: 4d50 4c45 445f 5341 4d50 4c45 442c 2022  MPLED_SAMPLED, "
+00013a30: 5341 4d50 4c45 4420 7673 2053 414d 504c  SAMPLED vs SAMPL
+00013a40: 4544 2073 6571 7320 4156 4552 4147 4520  ED seqs AVERAGE 
+00013a50: 426c 6f73 756d 2053 636f 7265 222c 2022  Blosum Score", "
+00013a60: 426c 6f73 756d 5f7b 7d5f 5341 4d50 4c45  Blosum_{}_SAMPLE
+00013a70: 445f 5341 4d50 4c45 4422 2e66 6f72 6d61  D_SAMPLED".forma
+00013a80: 7428 666f 6c64 6572 292c 616e 6e6f 743d  t(folder),annot=
+00013a90: 616e 6e6f 742c 6d61 736b 3d4e 6f6e 652c  annot,mask=None,
+00013aa0: 766d 6178 203d 2076 6d61 7829 0a64 6566  vmax = vmax).def
+00013ab0: 2062 7569 6c64 5f64 6174 6166 7261 6d65   build_dataframe
+00013ac0: 735f 6f76 6572 6c61 7070 696e 675f 6869  s_overlapping_hi
+00013ad0: 7374 6f67 7261 6d73 2861 615f 7365 7175  stograms(aa_sequ
+00013ae0: 656e 6365 735f 7072 6564 6963 7469 6f6e  ences_prediction
+00013af0: 732c 2064 6174 6173 6574 5f74 7261 696e  s, dataset_train
+00013b00: 2c64 6174 6173 6574 5f74 6573 742c 206e  ,dataset_test, n
+00013b10: 616d 652c 206e 5f73 616d 706c 6573 2c6e  ame, n_samples,n
+00013b20: 6f64 6573 5f69 6e64 6578 6573 2c72 6573  odes_indexes,res
+00013b30: 756c 7473 5f64 6972 6563 746f 7279 2c61  ults_directory,a
+00013b40: 615f 7072 6f62 732c 636f 7272 6573 706f  a_probs,correspo
+00013b50: 6e64 656e 6365 5f64 6963 743d 4e6f 6e65  ndence_dict=None
+00013b60: 293a 0a20 2020 2022 2222 506c 6f74 2068  ):.    """Plot h
+00013b70: 6973 746f 6772 616d 2077 686f 7365 2062  istogram whose b
+00013b80: 6172 7320 7265 7072 6573 656e 7420 7468  ars represent th
+00013b90: 6520 7065 7263 656e 7461 6765 2069 6465  e percentage ide
+00013ba0: 6e74 6974 7920 6265 7477 6565 6e20 7468  ntity between th
+00013bb0: 6520 7472 7565 2061 6e63 6573 746f 7273  e true ancestors
+00013bc0: 2076 7320 7472 7565 206c 6561 7665 7320   vs true leaves 
+00013bd0: 616e 6420 7468 6520 7072 6564 6963 7465  and the predicte
+00013be0: 6420 616e 6365 7374 6f72 7320 7673 2074  d ancestors vs t
+00013bf0: 7275 6520 6c65 6176 6573 0a20 2020 203a  rue leaves.    :
+00013c00: 7061 7261 6d20 7465 6e73 6f72 2061 615f  param tensor aa_
+00013c10: 7365 7175 656e 6365 735f 7072 6564 6963  sequences_predic
+00013c20: 7469 6f6e 733a 0a20 2020 203a 7061 7261  tions:.    :para
+00013c30: 6d20 7465 6e73 6f72 2064 6174 6173 6574  m tensor dataset
+00013c40: 5f74 7261 696e 0a20 2020 203a 7061 7261  _train.    :para
+00013c50: 6d20 7465 6e73 6f72 2064 6174 6173 6574  m tensor dataset
+00013c60: 5f74 6573 740a 2020 2020 3a70 6172 616d  _test.    :param
+00013c70: 2073 7472 206e 616d 653a 2064 6174 6173   str name: datas
+00013c80: 6574 206e 616d 6520 7072 6f6a 6563 740a  et name project.
+00013c90: 2020 2020 3a70 6172 616d 2074 656e 736f      :param tenso
+00013ca0: 7220 6161 5f73 6571 7565 6e63 6573 5f70  r aa_sequences_p
+00013cb0: 7265 6469 6374 696f 6e73 0a20 2020 203a  redictions.    :
+00013cc0: 7061 7261 6d20 696e 7420 6e5f 7361 6d70  param int n_samp
+00013cd0: 6c65 733a 206e 756d 6265 7220 6f66 2073  les: number of s
+00013ce0: 616d 706c 6564 2070 7265 6469 6374 696f  ampled predictio
+00013cf0: 6e73 0a20 2020 203a 7061 7261 6d20 6e6f  ns.    :param no
+00013d00: 6465 735f 696e 6465 7865 733a 0a20 2020  des_indexes:.   
+00013d10: 203a 7061 7261 6d20 7374 7220 7265 7375   :param str resu
+00013d20: 6c74 735f 6469 7265 6374 6f72 790a 2020  lts_directory.  
+00013d30: 2020 3a70 6172 616d 2069 6e74 2061 615f    :param int aa_
+00013d40: 7072 6f62 3a20 616d 696e 6f20 6163 6964  prob: amino acid
+00013d50: 2070 726f 6261 6269 6c69 7469 6573 2028   probabilities (
+00013d60: 692e 6520 3231 290a 2020 2020 3a70 6172  i.e 21).    :par
+00013d70: 616d 2064 6963 7420 636f 7272 6573 706f  am dict correspo
+00013d80: 6e64 656e 6365 5f64 6963 743a 2063 6f72  ndence_dict: cor
+00013d90: 7265 7370 6f6e 6465 6e63 6520 6265 7477  respondence betw
+00013da0: 6565 6e20 7468 6520 7472 6565 206c 6576  een the tree lev
+00013db0: 656c 206f 7264 6572 2069 6e64 6578 6573  el order indexes
+00013dc0: 2061 6e64 2074 6865 202d 2d3e 206f 6e6c   and the --> onl
+00013dd0: 7920 7573 6566 756c 2066 6f72 2062 656e  y useful for ben
+00013de0: 6368 6d61 726b 5f72 616e 6461 6c6c 5f6f  chmark_randall_o
+00013df0: 7269 6769 6e61 6c5f 6e61 6d69 6e67 0a20  riginal_naming. 
+00013e00: 2020 2022 2222 0a20 2020 2022 6874 7470     """.    "http
+00013e10: 733a 2f2f 6d61 7470 6c6f 746c 6962 2e6f  s://matplotlib.o
+00013e20: 7267 2f33 2e31 2e31 2f67 616c 6c65 7279  rg/3.1.1/gallery
+00013e30: 2f75 6e69 7473 2f62 6172 5f75 6e69 745f  /units/bar_unit_
+00013e40: 6465 6d6f 2e68 746d 6c23 7370 6878 2d67  demo.html#sphx-g
+00013e50: 6c72 2d67 616c 6c65 7279 2d75 6e69 7473  lr-gallery-units
+00013e60: 2d62 6172 2d75 6e69 742d 6465 6d6f 2d70  -bar-unit-demo-p
+00013e70: 7922 0a20 2020 2070 7269 6e74 2822 4275  y".    print("Bu
+00013e80: 696c 6469 6e67 204f 7665 726c 6170 7069  ilding Overlappi
+00013e90: 6e67 2048 6973 746f 6772 616d 2e2e 2e22  ng Histogram..."
+00013ea0: 290a 2020 2020 6e5f 6368 696c 6472 656e  ).    n_children
+00013eb0: 203d 206c 656e 286e 6f64 6573 5f69 6e64   = len(nodes_ind
+00013ec0: 6578 6573 290a 2020 2020 6465 6620 5065  exes).    def Pe
+00013ed0: 7263 656e 745f 4944 5f4f 4253 5f4f 4253  rcent_ID_OBS_OBS
+00013ee0: 2829 3a0a 2020 2020 2020 2020 2247 656e  ():.        "Gen
+00013ef0: 6572 6174 6520 7468 6520 2025 4944 206f  erate the  %ID o
+00013f00: 6620 7468 6520 4f42 5320 5445 5354 2073  f the OBS TEST s
+00013f10: 6571 7565 6e63 6573 2061 6761 696e 7374  equences against
+00013f20: 2074 6865 204f 4253 2074 7261 696e 220a   the OBS train".
+00013f30: 2020 2020 2020 2020 7065 7263 656e 745f          percent_
+00013f40: 6964 5f4f 4253 203d 2064 6963 742e 6672  id_OBS = dict.fr
+00013f50: 6f6d 6b65 7973 2864 6174 6173 6574 5f74  omkeys(dataset_t
+00013f60: 7261 696e 5b3a 2c30 2c31 5d2e 746f 6c69  rain[:,0,1].toli
+00013f70: 7374 2829 2c20 6469 6374 2e66 726f 6d6b  st(), dict.fromk
+00013f80: 6579 7328 6461 7461 7365 745f 7465 7374  eys(dataset_test
+00013f90: 5b3a 2c30 2c31 5d2e 746f 6c69 7374 2829  [:,0,1].tolist()
+00013fa0: 2c5b 5d29 290a 2020 2020 2020 2020 666f  ,[])).        fo
+00013fb0: 7220 692c 7472 6169 6e73 6571 2069 6e20  r i,trainseq in 
+00013fc0: 656e 756d 6572 6174 6528 6461 7461 7365  enumerate(datase
+00013fd0: 745f 7472 6169 6e29 3a0a 2020 2020 2020  t_train):.      
+00013fe0: 2020 2020 2020 7365 715f 5452 4149 4e5f        seq_TRAIN_
+00013ff0: 6f62 735f 6c65 7474 6572 735f 6920 3d20  obs_letters_i = 
+00014000: 636f 6e76 6572 745f 746f 5f6c 6574 7465  convert_to_lette
+00014010: 7273 2864 6174 6173 6574 5f74 7261 696e  rs(dataset_train
+00014020: 5b69 2c20 323a 2c30 5d2c 6161 5f70 726f  [i, 2:,0],aa_pro
+00014030: 6273 290a 2020 2020 2020 2020 2020 2020  bs).            
+00014040: 666f 7220 6a2c 7465 7374 7365 7120 696e  for j,testseq in
+00014050: 2065 6e75 6d65 7261 7465 2864 6174 6173   enumerate(datas
+00014060: 6574 5f74 6573 7429 3a0a 2020 2020 2020  et_test):.      
+00014070: 2020 2020 2020 2020 2020 7365 715f 5445            seq_TE
+00014080: 5354 5f6f 6273 5f6c 6574 7465 7273 5f69  ST_obs_letters_i
+00014090: 203d 2063 6f6e 7665 7274 5f74 6f5f 6c65   = convert_to_le
+000140a0: 7474 6572 7328 6461 7461 7365 745f 7465  tters(dataset_te
+000140b0: 7374 5b6a 2c20 323a 2c30 5d2c 6161 5f70  st[j, 2:,0],aa_p
+000140c0: 726f 6273 290a 2020 2020 2020 2020 2020  robs).          
+000140d0: 2020 2020 2020 7069 6420 3d70 6572 635f        pid =perc_
+000140e0: 6964 656e 7469 7479 5f70 6169 725f 7365  identity_pair_se
+000140f0: 7128 7365 715f 5452 4149 4e5f 6f62 735f  q(seq_TRAIN_obs_
+00014100: 6c65 7474 6572 735f 692c 7365 715f 5445  letters_i,seq_TE
+00014110: 5354 5f6f 6273 5f6c 6574 7465 7273 5f69  ST_obs_letters_i
+00014120: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
+00014130: 2020 7065 7263 656e 745f 6964 5f4f 4253    percent_id_OBS
+00014140: 5b64 6174 6173 6574 5f74 7261 696e 5b69  [dataset_train[i
+00014150: 2c30 2c31 5d2e 6974 656d 2829 5d5b 6461  ,0,1].item()][da
+00014160: 7461 7365 745f 7465 7374 5b6a 2c30 2c31  taset_test[j,0,1
+00014170: 5d2e 6974 656d 2829 5d20 3d20 7069 640a  ].item()] = pid.
+00014180: 2020 2020 2020 2020 7265 7475 726e 2070          return p
+00014190: 6572 6365 6e74 5f69 645f 4f42 530a 2020  ercent_id_OBS.  
+000141a0: 2020 6465 6620 5065 7263 656e 745f 4944    def Percent_ID
+000141b0: 5f50 5245 445f 4f42 5328 293a 0a20 2020  _PRED_OBS():.   
+000141c0: 2020 2020 2022 4765 6e65 7261 7465 2074       "Generate t
+000141d0: 6865 2041 7665 7261 6765 2061 6e64 2053  he Average and S
+000141e0: 5444 2025 4944 206f 6620 7468 6520 7361  TD %ID of the sa
+000141f0: 6d70 6c65 6420 7365 7175 656e 6365 7320  mpled sequences 
+00014200: 6167 6169 6e73 7420 7468 6520 4f42 5320  against the OBS 
+00014210: 7472 6169 6e22 0a20 2020 2020 2020 2023  train".        #
+00014220: 7065 7263 656e 745f 6964 5f50 5245 4420  percent_id_PRED 
+00014230: 3d20 6469 6374 2e66 726f 6d6b 6579 7328  = dict.fromkeys(
+00014240: 4461 7461 7365 745f 7472 6169 6e5b 3a2c  Dataset_train[:,
+00014250: 302c 315d 2e74 6f6c 6973 7428 292c 2064  0,1].tolist(), d
+00014260: 6963 742e 6672 6f6d 6b65 7973 2870 7265  ict.fromkeys(pre
+00014270: 6469 6374 696f 6e73 5f73 616d 706c 6573  dictions_samples
+00014280: 5b30 2c3a 2c31 5d2e 746f 6c69 7374 2829  [0,:,1].tolist()
+00014290: 2c7b 2241 7665 7261 6765 223a 5b5d 2c22  ,{"Average":[],"
+000142a0: 5374 6422 3a5b 5d7d 2929 2023 7b22 4176  Std":[]})) #{"Av
+000142b0: 6572 6167 6522 3a5b 5d2c 2253 7464 223a  erage":[],"Std":
+000142c0: 5b5d 7d0a 2020 2020 2020 2020 7065 7263  []}.        perc
+000142d0: 656e 745f 6964 5f50 5245 4420 3d20 6465  ent_id_PRED = de
+000142e0: 6661 756c 7464 6963 7428 6c61 6d62 6461  faultdict(lambda
+000142f0: 3a20 6465 6661 756c 7464 6963 7428 6c61  : defaultdict(la
+00014300: 6d62 6461 3a20 6465 6661 756c 7464 6963  mbda: defaultdic
+00014310: 7428 6469 6374 2929 290a 2020 2020 2020  t(dict))).      
+00014320: 2020 666f 7220 692c 2073 6571 2069 6e20    for i, seq in 
+00014330: 656e 756d 6572 6174 6528 6461 7461 7365  enumerate(datase
+00014340: 745f 7472 6169 6e29 3a0a 2020 2020 2020  t_train):.      
+00014350: 2020 2020 2020 7365 715f 5452 4149 4e5f        seq_TRAIN_
+00014360: 6f62 735f 6c65 7474 6572 735f 6920 3d20  obs_letters_i = 
+00014370: 636f 6e76 6572 745f 746f 5f6c 6574 7465  convert_to_lette
+00014380: 7273 2864 6174 6173 6574 5f74 7261 696e  rs(dataset_train
+00014390: 5b69 2c20 323a 2c30 5d2c 6161 5f70 726f  [i, 2:,0],aa_pro
+000143a0: 6273 290a 2020 2020 2020 2020 2020 2020  bs).            
+000143b0: 666f 7220 6a20 696e 2072 616e 6765 286e  for j in range(n
+000143c0: 5f63 6869 6c64 7265 6e29 3a23 666f 7220  _children):#for 
+000143d0: 7465 7374 206e 6f64 6520 236e 5f63 6869  test node #n_chi
+000143e0: 6c64 7265 6e0a 2020 2020 2020 2020 2020  ldren.          
+000143f0: 2020 2020 2020 616c 6c5f 7361 6d70 6c65        all_sample
+00014400: 645f 696e 6465 7820 3d20 6161 5f73 6571  d_index = aa_seq
+00014410: 7565 6e63 6573 5f70 7265 6469 6374 696f  uences_predictio
+00014420: 6e73 5b3a 2c6a 2c33 3a5d 2020 2320 416c  ns[:,j,3:]  # Al
+00014430: 6c20 7361 6d70 6c65 7320 666f 7220 7361  l samples for sa
+00014440: 6d65 2074 6573 7420 7365 710a 2020 2020  me test seq.    
+00014450: 2020 2020 2020 2020 2020 2020 7065 7263              perc
+00014460: 656e 745f 6964 5f69 3d5b 5d20 2361 6c6c  ent_id_i=[] #all
+00014470: 2073 616d 706c 6573 2066 6f72 2074 6865   samples for the
+00014480: 2073 616d 6520 7465 7374 2073 6571 7565   same test seque
+00014490: 6e63 650a 2020 2020 2020 2020 2020 2020  nce.            
+000144a0: 2020 2020 666f 7220 7720 696e 2072 616e      for w in ran
+000144b0: 6765 286e 5f73 616d 706c 6573 293a 0a20  ge(n_samples):. 
 000144c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000144d0: 2020 7365 715f 7361 6d70 6c65 645f 7465    seq_sampled_te
-000144e0: 7374 203d 2063 6f6e 7665 7274 5f74 6f5f  st = convert_to_
-000144f0: 6c65 7474 6572 7328 616c 6c5f 7361 6d70  letters(all_samp
-00014500: 6c65 645f 696e 6465 785b 775d 2c61 615f  led_index[w],aa_
-00014510: 7072 6f62 7329 0a20 2020 2020 2020 2020  probs).         
-00014520: 2020 2020 2020 2020 2020 2070 6964 203d             pid =
-00014530: 2070 6572 635f 6964 656e 7469 7479 5f70   perc_identity_p
-00014540: 6169 725f 7365 7128 7365 715f 5452 4149  air_seq(seq_TRAI
-00014550: 4e5f 6f62 735f 6c65 7474 6572 735f 692c  N_obs_letters_i,
-00014560: 2073 6571 5f73 616d 706c 6564 5f74 6573   seq_sampled_tes
-00014570: 7429 0a20 2020 2020 2020 2020 2020 2020  t).             
-00014580: 2020 2020 2020 2070 6572 6365 6e74 5f69         percent_i
-00014590: 645f 692e 6170 7065 6e64 2870 6964 290a  d_i.append(pid).
-000145a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000145b0: 7065 7263 656e 745f 6964 5f50 5245 445b  percent_id_PRED[
-000145c0: 6461 7461 7365 745f 7472 6169 6e5b 692c  dataset_train[i,
-000145d0: 302c 315d 2e69 7465 6d28 295d 5b61 615f  0,1].item()][aa_
-000145e0: 7365 7175 656e 6365 735f 7072 6564 6963  sequences_predic
-000145f0: 7469 6f6e 735b 302c 6a2c 315d 5d5b 2241  tions[0,j,1]]["A
-00014600: 7665 7261 6765 225d 203d 206e 702e 6d65  verage"] = np.me
-00014610: 616e 286e 702e 6172 7261 7928 7065 7263  an(np.array(perc
-00014620: 656e 745f 6964 5f69 292c 6178 6973 3d30  ent_id_i),axis=0
-00014630: 290a 2020 2020 2020 2020 2020 2020 2020  ).              
-00014640: 2020 7065 7263 656e 745f 6964 5f50 5245    percent_id_PRE
-00014650: 445b 6461 7461 7365 745f 7472 6169 6e5b  D[dataset_train[
-00014660: 692c 302c 315d 2e69 7465 6d28 295d 5b61  i,0,1].item()][a
-00014670: 615f 7365 7175 656e 6365 735f 7072 6564  a_sequences_pred
-00014680: 6963 7469 6f6e 735b 302c 6a2c 315d 5d5b  ictions[0,j,1]][
-00014690: 2253 7464 225d 203d 206e 702e 7374 6428  "Std"] = np.std(
-000146a0: 6e70 2e61 7272 6179 2870 6572 6365 6e74  np.array(percent
-000146b0: 5f69 645f 6929 2c20 6178 6973 3d30 290a  _id_i), axis=0).
-000146c0: 2020 2020 2020 2020 7265 7475 726e 2070          return p
-000146d0: 6572 6365 6e74 5f69 645f 5052 4544 0a0a  ercent_id_PRED..
-000146e0: 0a20 2020 2070 6572 6365 6e74 5f69 645f  .    percent_id_
-000146f0: 5052 4544 203d 2050 6572 6365 6e74 5f49  PRED = Percent_I
-00014700: 445f 5052 4544 5f4f 4253 2829 0a20 2020  D_PRED_OBS().   
-00014710: 2070 6572 6365 6e74 5f69 645f 4f42 5320   percent_id_OBS 
-00014720: 3d20 5065 7263 656e 745f 4944 5f4f 4253  = Percent_ID_OBS
-00014730: 5f4f 4253 2829 0a0a 2020 2020 6669 6720  _OBS()..    fig 
-00014740: 3d20 706c 742e 6669 6775 7265 2866 6967  = plt.figure(fig
-00014750: 7369 7a65 3d28 3138 2c20 3529 290a 2020  size=(18, 5)).  
-00014760: 2020 6178 203d 2066 6967 2e61 6464 5f73    ax = fig.add_s
-00014770: 7562 706c 6f74 2831 3131 290a 0a20 2020  ubplot(111)..   
-00014780: 2023 2320 496e 6465 7865 7320 7768 6572   ## Indexes wher
-00014790: 6520 7468 6520 6261 7273 2077 696c 6c20  e the bars will 
-000147a0: 676f 0a20 2020 2069 6e64 5f6c 696e 6573  go.    ind_lines
-000147b0: 203d 206e 702e 6172 616e 6765 2831 302c   = np.arange(10,
-000147c0: 6c65 6e28 6461 7461 7365 745f 7472 6169  len(dataset_trai
-000147d0: 6e29 2a31 302c 3130 2029 0a20 2020 2069  n)*10,10 ).    i
-000147e0: 6e64 5f74 7261 696e 203d 206e 702e 6172  nd_train = np.ar
-000147f0: 616e 6765 2835 2c6c 656e 2864 6174 6173  ange(5,len(datas
-00014800: 6574 5f74 7261 696e 292a 3130 2c31 3020  et_train)*10,10 
-00014810: 2920 2023 2074 6865 2078 206c 6f63 6174  )  # the x locat
-00014820: 696f 6e73 2066 6f72 2074 6865 2074 7261  ions for the tra
-00014830: 696e 206e 6f64 6573 0a20 2020 2077 6964  in nodes.    wid
-00014840: 7468 203d 2031 302f 6c65 6e28 6461 7461  th = 10/len(data
-00014850: 7365 745f 7465 7374 2920 2d30 2e33 2023  set_test) -0.3 #
-00014860: 2074 6865 2077 6964 7468 206f 6620 7468   the width of th
-00014870: 6520 6261 7273 0a20 2020 2069 6e64 5f74  e bars.    ind_t
-00014880: 6573 7420 3d20 6e70 2e61 7261 6e67 6528  est = np.arange(
-00014890: 302c 3130 2c77 6964 7468 202b 2030 2e33  0,10,width + 0.3
-000148a0: 2920 2023 2074 6865 2078 206c 6f63 6174  )  # the x locat
-000148b0: 696f 6e73 2066 6f72 2074 6865 2074 6573  ions for the tes
-000148c0: 7420 6e6f 6465 732c 2074 616b 696e 6720  t nodes, taking 
-000148d0: 696e 746f 2061 6363 6f75 6e74 2074 6865  into account the
-000148e0: 2073 7061 6365 2062 6574 7765 656e 2074   space between t
-000148f0: 6865 6d3f 0a20 2020 2073 7461 7274 203d  hem?.    start =
-00014900: 2030 0a20 2020 2062 6c75 652c 203d 2073   0.    blue, = s
-00014910: 6e73 2e63 6f6c 6f72 5f70 616c 6574 7465  ns.color_palette
-00014920: 2822 6d75 7465 6422 2c20 3129 0a20 2020  ("muted", 1).   
-00014930: 2066 6f72 2074 7261 696e 5f6e 6f64 6520   for train_node 
-00014940: 696e 2064 6174 6173 6574 5f74 7261 696e  in dataset_train
-00014950: 5b3a 2c30 2c31 5d3a 0a20 2020 2020 2020  [:,0,1]:.       
-00014960: 2066 6f72 2069 6478 2c74 6573 745f 6e6f   for idx,test_no
-00014970: 6465 2069 6e20 656e 756d 6572 6174 6528  de in enumerate(
-00014980: 6461 7461 7365 745f 7465 7374 5b3a 2c30  dataset_test[:,0
-00014990: 2c31 5d29 3a20 2354 4f44 4f3a 206f 7574  ,1]): #TODO: out
-000149a0: 206f 6620 7261 6e67 650a 2020 2020 2020   of range.      
-000149b0: 2020 2020 2020 7265 6374 7332 203d 2061        rects2 = a
-000149c0: 782e 6261 7228 696e 645f 7465 7374 5b69  x.bar(ind_test[i
-000149d0: 6478 5d20 2b20 7374 6172 7420 2c0a 2020  dx] + start ,.  
+000144d0: 2020 2073 6571 5f73 616d 706c 6564 5f74     seq_sampled_t
+000144e0: 6573 7420 3d20 636f 6e76 6572 745f 746f  est = convert_to
+000144f0: 5f6c 6574 7465 7273 2861 6c6c 5f73 616d  _letters(all_sam
+00014500: 706c 6564 5f69 6e64 6578 5b77 5d2c 6161  pled_index[w],aa
+00014510: 5f70 726f 6273 290a 2020 2020 2020 2020  _probs).        
+00014520: 2020 2020 2020 2020 2020 2020 7069 6420              pid 
+00014530: 3d20 7065 7263 5f69 6465 6e74 6974 795f  = perc_identity_
+00014540: 7061 6972 5f73 6571 2873 6571 5f54 5241  pair_seq(seq_TRA
+00014550: 494e 5f6f 6273 5f6c 6574 7465 7273 5f69  IN_obs_letters_i
+00014560: 2c20 7365 715f 7361 6d70 6c65 645f 7465  , seq_sampled_te
+00014570: 7374 290a 2020 2020 2020 2020 2020 2020  st).            
+00014580: 2020 2020 2020 2020 7065 7263 656e 745f          percent_
+00014590: 6964 5f69 2e61 7070 656e 6428 7069 6429  id_i.append(pid)
+000145a0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+000145b0: 2070 6572 6365 6e74 5f69 645f 5052 4544   percent_id_PRED
+000145c0: 5b64 6174 6173 6574 5f74 7261 696e 5b69  [dataset_train[i
+000145d0: 2c30 2c31 5d2e 6974 656d 2829 5d5b 6161  ,0,1].item()][aa
+000145e0: 5f73 6571 7565 6e63 6573 5f70 7265 6469  _sequences_predi
+000145f0: 6374 696f 6e73 5b30 2c6a 2c31 5d5d 5b22  ctions[0,j,1]]["
+00014600: 4176 6572 6167 6522 5d20 3d20 6e70 2e6d  Average"] = np.m
+00014610: 6561 6e28 6e70 2e61 7272 6179 2870 6572  ean(np.array(per
+00014620: 6365 6e74 5f69 645f 6929 2c61 7869 733d  cent_id_i),axis=
+00014630: 3029 0a20 2020 2020 2020 2020 2020 2020  0).             
+00014640: 2020 2070 6572 6365 6e74 5f69 645f 5052     percent_id_PR
+00014650: 4544 5b64 6174 6173 6574 5f74 7261 696e  ED[dataset_train
+00014660: 5b69 2c30 2c31 5d2e 6974 656d 2829 5d5b  [i,0,1].item()][
+00014670: 6161 5f73 6571 7565 6e63 6573 5f70 7265  aa_sequences_pre
+00014680: 6469 6374 696f 6e73 5b30 2c6a 2c31 5d5d  dictions[0,j,1]]
+00014690: 5b22 5374 6422 5d20 3d20 6e70 2e73 7464  ["Std"] = np.std
+000146a0: 286e 702e 6172 7261 7928 7065 7263 656e  (np.array(percen
+000146b0: 745f 6964 5f69 292c 2061 7869 733d 3029  t_id_i), axis=0)
+000146c0: 0a20 2020 2020 2020 2072 6574 7572 6e20  .        return 
+000146d0: 7065 7263 656e 745f 6964 5f50 5245 440a  percent_id_PRED.
+000146e0: 0a0a 2020 2020 7065 7263 656e 745f 6964  ..    percent_id
+000146f0: 5f50 5245 4420 3d20 5065 7263 656e 745f  _PRED = Percent_
+00014700: 4944 5f50 5245 445f 4f42 5328 290a 2020  ID_PRED_OBS().  
+00014710: 2020 7065 7263 656e 745f 6964 5f4f 4253    percent_id_OBS
+00014720: 203d 2050 6572 6365 6e74 5f49 445f 4f42   = Percent_ID_OB
+00014730: 535f 4f42 5328 290a 0a20 2020 2066 6967  S_OBS()..    fig
+00014740: 203d 2070 6c74 2e66 6967 7572 6528 6669   = plt.figure(fi
+00014750: 6773 697a 653d 2831 382c 2035 2929 0a20  gsize=(18, 5)). 
+00014760: 2020 2061 7820 3d20 6669 672e 6164 645f     ax = fig.add_
+00014770: 7375 6270 6c6f 7428 3131 3129 0a0a 2020  subplot(111)..  
+00014780: 2020 2323 2049 6e64 6578 6573 2077 6865    ## Indexes whe
+00014790: 7265 2074 6865 2062 6172 7320 7769 6c6c  re the bars will
+000147a0: 2067 6f0a 2020 2020 696e 645f 6c69 6e65   go.    ind_line
+000147b0: 7320 3d20 6e70 2e61 7261 6e67 6528 3130  s = np.arange(10
+000147c0: 2c6c 656e 2864 6174 6173 6574 5f74 7261  ,len(dataset_tra
+000147d0: 696e 292a 3130 2c31 3020 290a 2020 2020  in)*10,10 ).    
+000147e0: 696e 645f 7472 6169 6e20 3d20 6e70 2e61  ind_train = np.a
+000147f0: 7261 6e67 6528 352c 6c65 6e28 6461 7461  range(5,len(data
+00014800: 7365 745f 7472 6169 6e29 2a31 302c 3130  set_train)*10,10
+00014810: 2029 2020 2320 7468 6520 7820 6c6f 6361   )  # the x loca
+00014820: 7469 6f6e 7320 666f 7220 7468 6520 7472  tions for the tr
+00014830: 6169 6e20 6e6f 6465 730a 2020 2020 7769  ain nodes.    wi
+00014840: 6474 6820 3d20 3130 2f6c 656e 2864 6174  dth = 10/len(dat
+00014850: 6173 6574 5f74 6573 7429 202d 302e 3320  aset_test) -0.3 
+00014860: 2320 7468 6520 7769 6474 6820 6f66 2074  # the width of t
+00014870: 6865 2062 6172 730a 2020 2020 696e 645f  he bars.    ind_
+00014880: 7465 7374 203d 206e 702e 6172 616e 6765  test = np.arange
+00014890: 2830 2c31 302c 7769 6474 6820 2b20 302e  (0,10,width + 0.
+000148a0: 3329 2020 2320 7468 6520 7820 6c6f 6361  3)  # the x loca
+000148b0: 7469 6f6e 7320 666f 7220 7468 6520 7465  tions for the te
+000148c0: 7374 206e 6f64 6573 2c20 7461 6b69 6e67  st nodes, taking
+000148d0: 2069 6e74 6f20 6163 636f 756e 7420 7468   into account th
+000148e0: 6520 7370 6163 6520 6265 7477 6565 6e20  e space between 
+000148f0: 7468 656d 3f0a 2020 2020 7374 6172 7420  them?.    start 
+00014900: 3d20 300a 2020 2020 626c 7565 2c20 3d20  = 0.    blue, = 
+00014910: 736e 732e 636f 6c6f 725f 7061 6c65 7474  sns.color_palett
+00014920: 6528 226d 7574 6564 222c 2031 290a 2020  e("muted", 1).  
+00014930: 2020 666f 7220 7472 6169 6e5f 6e6f 6465    for train_node
+00014940: 2069 6e20 6461 7461 7365 745f 7472 6169   in dataset_trai
+00014950: 6e5b 3a2c 302c 315d 3a0a 2020 2020 2020  n[:,0,1]:.      
+00014960: 2020 666f 7220 6964 782c 7465 7374 5f6e    for idx,test_n
+00014970: 6f64 6520 696e 2065 6e75 6d65 7261 7465  ode in enumerate
+00014980: 2864 6174 6173 6574 5f74 6573 745b 3a2c  (dataset_test[:,
+00014990: 302c 315d 293a 2023 544f 444f 3a20 6f75  0,1]): #TODO: ou
+000149a0: 7420 6f66 2072 616e 6765 0a20 2020 2020  t of range.     
+000149b0: 2020 2020 2020 2072 6563 7473 3220 3d20         rects2 = 
+000149c0: 6178 2e62 6172 2869 6e64 5f74 6573 745b  ax.bar(ind_test[
+000149d0: 6964 785d 202b 2073 7461 7274 202c 0a20  idx] + start ,. 
 000149e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000149f0: 2020 2020 2020 2020 2020 7065 7263 656e            percen
-00014a00: 745f 6964 5f4f 4253 5b74 7261 696e 5f6e  t_id_OBS[train_n
-00014a10: 6f64 652e 6974 656d 2829 5d5b 7465 7374  ode.item()][test
-00014a20: 5f6e 6f64 652e 6974 656d 2829 5d2c 0a20  _node.item()],. 
+000149f0: 2020 2020 2020 2020 2020 2070 6572 6365             perce
+00014a00: 6e74 5f69 645f 4f42 535b 7472 6169 6e5f  nt_id_OBS[train_
+00014a10: 6e6f 6465 2e69 7465 6d28 295d 5b74 6573  node.item()][tes
+00014a20: 745f 6e6f 6465 2e69 7465 6d28 295d 2c0a  t_node.item()],.
 00014a30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014a40: 2020 2020 2020 2020 2020 2077 6964 7468             width
-00014a50: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
-00014a60: 2020 2020 2020 2020 2020 2020 2020 636f                co
-00014a70: 6c6f 723d 2767 7265 656e 272c 0a20 2020  lor='green',.   
+00014a40: 2020 2020 2020 2020 2020 2020 7769 6474              widt
+00014a50: 682c 0a20 2020 2020 2020 2020 2020 2020  h,.             
+00014a60: 2020 2020 2020 2020 2020 2020 2020 2063                 c
+00014a70: 6f6c 6f72 3d27 6772 6565 6e27 2c0a 2020  olor='green',.  
 00014a80: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014a90: 2020 2020 2020 2020 2079 6572 723d 5b30           yerr=[0
-00014aa0: 5d2c 0a20 2020 2020 2020 2020 2020 2020  ],.             
-00014ab0: 2020 2020 2020 2020 2020 2020 2020 2065                 e
-00014ac0: 7272 6f72 5f6b 773d 6469 6374 2865 6c69  rror_kw=dict(eli
-00014ad0: 6e65 7769 6474 683d 302e 3031 2c20 6563  newidth=0.01, ec
-00014ae0: 6f6c 6f72 3d27 7265 6427 292c 0a20 2020  olor='red'),.   
+00014a90: 2020 2020 2020 2020 2020 7965 7272 3d5b            yerr=[
+00014aa0: 305d 2c0a 2020 2020 2020 2020 2020 2020  0],.            
+00014ab0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00014ac0: 6572 726f 725f 6b77 3d64 6963 7428 656c  error_kw=dict(el
+00014ad0: 696e 6577 6964 7468 3d30 2e30 312c 2065  inewidth=0.01, e
+00014ae0: 636f 6c6f 723d 2772 6564 2729 2c0a 2020  color='red'),.  
 00014af0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014b00: 2020 2020 2020 2020 206c 6162 656c 3d22           label="
-00014b10: 4f42 5345 5256 4544 2229 0a20 2020 2020  OBSERVED").     
-00014b20: 2020 2020 2020 2072 6563 7473 3120 3d20         rects1 = 
-00014b30: 6178 2e62 6172 2869 6e64 5f74 6573 745b  ax.bar(ind_test[
-00014b40: 6964 785d 202b 2073 7461 7274 2c0a 2020  idx] + start,.  
+00014b00: 2020 2020 2020 2020 2020 6c61 6265 6c3d            label=
+00014b10: 224f 4253 4552 5645 4422 290a 2020 2020  "OBSERVED").    
+00014b20: 2020 2020 2020 2020 7265 6374 7331 203d          rects1 =
+00014b30: 2061 782e 6261 7228 696e 645f 7465 7374   ax.bar(ind_test
+00014b40: 5b69 6478 5d20 2b20 7374 6172 742c 0a20  [idx] + start,. 
 00014b50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014b60: 2020 2020 2020 2020 2020 7065 7263 656e            percen
-00014b70: 745f 6964 5f50 5245 445b 7472 6169 6e5f  t_id_PRED[train_
-00014b80: 6e6f 6465 2e69 7465 6d28 295d 5b74 6573  node.item()][tes
-00014b90: 745f 6e6f 6465 2e69 7465 6d28 295d 5b22  t_node.item()]["
-00014ba0: 4176 6572 6167 6522 5d2c 0a20 2020 2020  Average"],.     
+00014b60: 2020 2020 2020 2020 2020 2070 6572 6365             perce
+00014b70: 6e74 5f69 645f 5052 4544 5b74 7261 696e  nt_id_PRED[train
+00014b80: 5f6e 6f64 652e 6974 656d 2829 5d5b 7465  _node.item()][te
+00014b90: 7374 5f6e 6f64 652e 6974 656d 2829 5d5b  st_node.item()][
+00014ba0: 2241 7665 7261 6765 225d 2c0a 2020 2020  "Average"],.    
 00014bb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014bc0: 2020 2020 2020 2077 6964 7468 2c0a 2020         width,.  
+00014bc0: 2020 2020 2020 2020 7769 6474 682c 0a20          width,. 
 00014bd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014be0: 2020 2020 2020 2020 2020 636f 6c6f 723d            color=
-00014bf0: 226f 7261 6e67 6522 2c0a 2020 2020 2020  "orange",.      
+00014be0: 2020 2020 2020 2020 2020 2063 6f6c 6f72             color
+00014bf0: 3d22 6f72 616e 6765 222c 0a20 2020 2020  ="orange",.     
 00014c00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014c10: 2020 2020 2020 616c 7068 613d 312c 0a20        alpha=1,. 
+00014c10: 2020 2020 2020 2061 6c70 6861 3d31 2c0a         alpha=1,.
 00014c20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014c30: 2020 2020 2020 2020 2020 2079 6572 723d             yerr=
-00014c40: 7065 7263 656e 745f 6964 5f50 5245 445b  percent_id_PRED[
-00014c50: 7472 6169 6e5f 6e6f 6465 2e69 7465 6d28  train_node.item(
-00014c60: 295d 5b74 6573 745f 6e6f 6465 2e69 7465  )][test_node.ite
-00014c70: 6d28 295d 5b22 5374 6422 5d2c 0a20 2020  m()]["Std"],.   
+00014c30: 2020 2020 2020 2020 2020 2020 7965 7272              yerr
+00014c40: 3d70 6572 6365 6e74 5f69 645f 5052 4544  =percent_id_PRED
+00014c50: 5b74 7261 696e 5f6e 6f64 652e 6974 656d  [train_node.item
+00014c60: 2829 5d5b 7465 7374 5f6e 6f64 652e 6974  ()][test_node.it
+00014c70: 656d 2829 5d5b 2253 7464 225d 2c0a 2020  em()]["Std"],.  
 00014c80: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014c90: 2020 2020 2020 2020 2065 7272 6f72 5f6b           error_k
-00014ca0: 773d 6469 6374 2865 6c69 6e65 7769 6474  w=dict(elinewidt
-00014cb0: 683d 302e 3031 2c20 6563 6f6c 6f72 3d27  h=0.01, ecolor='
-00014cc0: 7265 6427 292c 0a20 2020 2020 2020 2020  red'),.         
+00014c90: 2020 2020 2020 2020 2020 6572 726f 725f            error_
+00014ca0: 6b77 3d64 6963 7428 656c 696e 6577 6964  kw=dict(elinewid
+00014cb0: 7468 3d30 2e30 312c 2065 636f 6c6f 723d  th=0.01, ecolor=
+00014cc0: 2772 6564 2729 2c0a 2020 2020 2020 2020  'red'),.        
 00014cd0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00014ce0: 2020 206c 6162 656c 3d22 5341 4d50 4c45     label="SAMPLE
-00014cf0: 4422 290a 0a20 2020 2020 2020 2073 7461  D")..        sta
-00014d00: 7274 202b 3d20 3130 0a20 2020 2023 2061  rt += 10.    # a
-00014d10: 7865 7320 616e 6420 6c61 6265 6c73 0a20  xes and labels. 
-00014d20: 2020 2061 782e 7365 745f 786c 696d 282d     ax.set_xlim(-
-00014d30: 7769 6474 682c 206e 702e 6d61 7828 696e  width, np.max(in
-00014d40: 645f 7472 6169 6e29 202b 3529 0a20 2020  d_train) +5).   
-00014d50: 2061 782e 7365 745f 796c 696d 2830 2c20   ax.set_ylim(0, 
-00014d60: 3130 3029 0a20 2020 2061 782e 7365 745f  100).    ax.set_
-00014d70: 796c 6162 656c 2827 2549 4427 290a 2020  ylabel('%ID').  
-00014d80: 2020 6178 2e73 6574 5f74 6974 6c65 2827    ax.set_title('
-00014d90: 4f76 6572 6c61 7070 696e 6720 4869 7374  Overlapping Hist
-00014da0: 6f67 7261 6d27 290a 2020 2020 7854 6963  ogram').    xTic
-00014db0: 6b4d 6172 6b73 203d 205b 2754 7261 696e  kMarks = ['Train
-00014dc0: 4e6f 6465 5f7b 7d27 2e66 6f72 6d61 7428  Node_{}'.format(
-00014dd0: 696e 7428 692e 6974 656d 2829 2929 2066  int(i.item())) f
-00014de0: 6f72 2069 2069 6e20 6461 7461 7365 745f  or i in dataset_
-00014df0: 7472 6169 6e5b 3a2c 302c 315d 5d0a 2020  train[:,0,1]].  
-00014e00: 2020 6178 2e73 6574 5f78 7469 636b 7328    ax.set_xticks(
-00014e10: 696e 645f 7472 6169 6e29 0a20 2020 2078  ind_train).    x
-00014e20: 7469 636b 4e61 6d65 7320 3d20 6178 2e73  tickNames = ax.s
-00014e30: 6574 5f78 7469 636b 6c61 6265 6c73 2878  et_xticklabels(x
-00014e40: 5469 636b 4d61 726b 7329 0a20 2020 2070  TickMarks).    p
-00014e50: 6c74 2e73 6574 7028 7874 6963 6b4e 616d  lt.setp(xtickNam
-00014e60: 6573 2c20 726f 7461 7469 6f6e 3d34 352c  es, rotation=45,
-00014e70: 2066 6f6e 7473 697a 653d 3629 0a20 2020   fontsize=6).   
-00014e80: 2070 6c74 2e76 6c69 6e65 7328 696e 645f   plt.vlines(ind_
-00014e90: 6c69 6e65 732c 796d 696e 3d30 2c79 6d61  lines,ymin=0,yma
-00014ea0: 783d 3130 3029 0a20 2020 2023 2320 6164  x=100).    ## ad
-00014eb0: 6420 6120 6c65 6765 6e64 2d2d 3e20 4e6f  d a legend--> No
-00014ec0: 7420 776f 726b 696e 6720 2869 6e73 6964  t working (insid
-00014ed0: 6520 6f72 206f 7574 7369 6465 206c 6f6f  e or outside loo
-00014ee0: 7029 0a20 2020 2061 782e 6c65 6765 6e64  p).    ax.legend
-00014ef0: 2828 7265 6374 7331 2c72 6563 7473 3229  ((rects1,rects2)
-00014f00: 2c20 2827 5361 6d70 6c65 6427 2c22 4f62  , ('Sampled',"Ob
-00014f10: 7365 7276 6564 2229 2c6c 6f63 3d31 2c62  served"),loc=1,b
-00014f20: 626f 785f 746f 5f61 6e63 686f 723d 2831  box_to_anchor=(1
-00014f30: 2e31 352c 2031 2e31 2929 0a20 2020 2070  .15, 1.1)).    p
-00014f40: 6c74 2e73 6176 6566 6967 2822 7b7d 2f4f  lt.savefig("{}/O
-00014f50: 7665 726c 6170 7069 6e67 4869 7374 6f67  verlappingHistog
-00014f60: 7261 6d22 2e66 6f72 6d61 7428 7265 7375  ram".format(resu
-00014f70: 6c74 735f 6469 7265 6374 6f72 7929 290a  lts_directory)).
-00014f80: 6465 6620 7472 6565 5f70 6f73 6974 696f  def tree_positio
-00014f90: 6e61 6c5f 656d 6265 6464 696e 6773 2861  nal_embeddings(a
-00014fa0: 6e63 6573 746f 725f 696e 666f 5f64 6963  ncestor_info_dic
-00014fb0: 742c 7472 6565 5f62 795f 6c65 7665 6c73  t,tree_by_levels
-00014fc0: 5f64 6963 7429 3a0a 2020 2020 2222 2254  _dict):.    """T
-00014fd0: 6865 206e 6f64 6573 206f 7264 6572 6564  he nodes ordered
-00014fe0: 2062 7920 7472 6565 206c 6576 656c 206f   by tree level o
-00014ff0: 7264 6572 2061 7265 2063 6f6e 7665 7274  rder are convert
-00015000: 6564 2069 6e74 6f20 736f 6d65 7768 6174  ed into somewhat
-00015010: 206f 6e65 2d68 6f74 2065 6e63 6f64 6564   one-hot encoded
-00015020: 206c 696b 6520 656d 6265 6464 696e 6773   like embeddings
-00015030: 0a20 2020 2049 6d70 6c65 6d65 6e74 6174  .    Implementat
-00015040: 6564 2066 6f6c 6c6f 7769 6e67 0a20 2020  ed following.   
-00015050: 2068 7474 7073 3a2f 2f70 6170 6572 732e   https://papers.
-00015060: 6e69 7073 2e63 632f 7061 7065 722f 3230  nips.cc/paper/20
-00015070: 3139 2f66 696c 652f 3665 3039 3137 3436  19/file/6e091746
-00015080: 3932 3134 6438 6662 6438 6335 3137 6463  9214d8fbd8c517dc
-00015090: 6463 3662 3864 6366 2d50 6170 6572 2e70  dc6b8dcf-Paper.p
-000150a0: 6466 0a20 2020 2049 6e70 7574 3a20 5472  df.    Input: Tr
-000150b0: 6565 2062 7920 6c65 7665 6c73 2069 6e20  ee by levels in 
-000150c0: 7472 6176 6572 7361 6c20 6f72 6465 720a  traversal order.
-000150d0: 2020 2020 4f75 7470 7574 3a20 5265 7072      Output: Repr
-000150e0: 6573 656e 7461 7469 6f6e 7320 6f66 2074  esentations of t
-000150f0: 6865 2070 6f73 6974 696f 6e20 6f66 2074  he position of t
-00015100: 6865 206e 6f64 6520 696e 2074 6865 2074  he node in the t
-00015110: 7265 6520 696e 206f 6e65 2068 6f74 2062  ree in one hot b
-00015120: 6974 2065 6e63 6f64 696e 672e 2054 6865  it encoding. The
-00015130: 7920 7265 7072 6573 656e 7420 7468 6520  y represent the 
-00015140: 6f6e 6520 686f 7420 656e 636f 6465 6420  one hot encoded 
-00015150: 7061 7468 2066 726f 6d20 726f 6f74 2074  path from root t
-00015160: 6f20 7468 6520 6e6f 6465 2022 2222 0a20  o the node """. 
-00015170: 2020 2064 6567 7265 655f 7472 6565 203d     degree_tree =
-00015180: 320a 2020 2020 6e5f 6c65 7665 6c73 203d  2.    n_levels =
-00015190: 206c 656e 2874 7265 655f 6279 5f6c 6576   len(tree_by_lev
-000151a0: 656c 735f 6469 6374 290a 2020 2020 6e6f  els_dict).    no
-000151b0: 6465 735f 7265 7072 6573 656e 7461 7469  des_representati
-000151c0: 6f6e 733d 6465 6661 756c 7464 6963 7428  ons=defaultdict(
-000151d0: 290a 2020 2020 6e6f 6465 5f72 6570 7265  ).    node_repre
-000151e0: 7365 6e74 6174 696f 6e20 3d20 5b30 5d2a  sentation = [0]*
-000151f0: 2864 6567 7265 655f 7472 6565 2a6e 5f6c  (degree_tree*n_l
-00015200: 6576 656c 7329 2023 696e 6974 6961 7465  evels) #initiate
-00015210: 2074 6865 2072 6f6f 7420 746f 2061 6c6c   the root to all
-00015220: 2030 0a20 2020 2066 6f72 206c 6576 656c   0.    for level
-00015230: 5f6e 616d 652c 206c 6576 656c 5f6e 6f64  _name, level_nod
-00015240: 6573 2069 6e20 736f 7274 6564 2874 7265  es in sorted(tre
-00015250: 655f 6279 5f6c 6576 656c 735f 6469 6374  e_by_levels_dict
-00015260: 2e69 7465 6d73 2829 293a 0a20 2020 2020  .items()):.     
-00015270: 2020 2061 6c6c 5f61 6e63 6573 746f 7273     all_ancestors
-00015280: 5f69 6e5f 6c65 7665 6c20 3d20 7b69 6e64  _in_level = {ind
-00015290: 6578 3a61 6e63 6573 746f 725f 696e 666f  ex:ancestor_info
-000152a0: 5f64 6963 745b 6e6f 6465 5d20 666f 7220  _dict[node] for 
-000152b0: 696e 6465 782c 6e6f 6465 2069 6e20 656e  index,node in en
-000152c0: 756d 6572 6174 6528 6c65 7665 6c5f 6e6f  umerate(level_no
-000152d0: 6465 7329 7d20 2366 696e 6420 7468 6520  des)} #find the 
-000152e0: 696e 6465 7865 7320 6f66 2074 6865 206e  indexes of the n
-000152f0: 6f64 6573 2073 6861 7269 6e67 2074 6865  odes sharing the
-00015300: 2073 616d 6520 616e 6365 7374 6f72 730a   same ancestors.
-00015310: 2020 2020 2020 2020 6966 206e 702e 6e61          if np.na
-00015320: 6e20 696e 2061 6c6c 5f61 6e63 6573 746f  n in all_ancesto
-00015330: 7273 5f69 6e5f 6c65 7665 6c2e 7661 6c75  rs_in_level.valu
-00015340: 6573 2829 3a23 6964 656e 7469 6679 2072  es():#identify r
-00015350: 6f6f 742c 2062 6563 6175 7365 2069 7420  oot, because it 
-00015360: 6861 7320 6e61 6e20 6173 2061 6e20 616e  has nan as an an
-00015370: 6365 7374 6f72 0a20 2020 2020 2020 2020  cestor.         
-00015380: 2020 206e 6f64 6573 5f72 6570 7265 7365     nodes_represe
-00015390: 6e74 6174 696f 6e73 5b6c 6576 656c 5f6e  ntations[level_n
-000153a0: 6f64 6573 5b30 5d5d 203d 206e 6f64 655f  odes[0]] = node_
-000153b0: 7265 7072 6573 656e 7461 7469 6f6e 0a20  representation. 
-000153c0: 2020 2020 2020 2065 6c73 653a 0a20 2020         else:.   
-000153d0: 2020 2020 2020 2020 206e 6f64 655f 7265           node_re
-000153e0: 7072 6573 656e 7461 7469 6f6e 203d 206e  presentation = n
-000153f0: 6f64 655f 7265 7072 6573 656e 7461 7469  ode_representati
-00015400: 6f6e 5b3a 2d64 6567 7265 655f 7472 6565  on[:-degree_tree
-00015410: 5d0a 2020 2020 2020 2020 2020 2020 6c65  ].            le
-00015420: 7665 6c5f 7061 6972 7320 3d20 5b6c 6576  vel_pairs = [lev
-00015430: 656c 5f6e 6f64 6573 5b69 3a69 202b 2064  el_nodes[i:i + d
-00015440: 6567 7265 655f 7472 6565 5d20 666f 7220  egree_tree] for 
-00015450: 6920 696e 2072 616e 6765 2830 2c20 6c65  i in range(0, le
-00015460: 6e28 6c65 7665 6c5f 6e6f 6465 7329 2c20  n(level_nodes), 
-00015470: 6465 6772 6565 5f74 7265 6529 5d20 2344  degree_tree)] #D
-00015480: 6976 6964 6520 6c65 7665 6c20 696e 2067  ivide level in g
-00015490: 726f 7570 7320 6f66 2032 2c20 7468 6520  roups of 2, the 
-000154a0: 7472 6565 2069 7320 696e 2074 7265 6520  tree is in tree 
-000154b0: 7472 6176 6572 7361 6c2c 2073 6f20 6576  traversal, so ev
-000154c0: 6572 7974 6869 6e67 2073 686f 7564 6c20  erything shoudl 
-000154d0: 6265 206f 7264 6572 6564 0a20 2020 2020  be ordered.     
-000154e0: 2020 2020 2020 2066 6f72 206c 6576 656c         for level
-000154f0: 5f70 6169 7220 696e 206c 6576 656c 5f70  _pair in level_p
-00015500: 6169 7273 3a20 2374 6865 7365 2070 6169  airs: #these pai
-00015510: 7273 2068 6176 6520 7361 6d65 2061 6e63  rs have same anc
-00015520: 6573 746f 7220 6265 6361 7573 6520 6f66  estor because of
-00015530: 2074 7265 6520 6c65 7665 6c73 2061 7265   tree levels are
-00015540: 206f 7267 616e 697a 6564 2069 6e20 7472   organized in tr
-00015550: 6176 6572 7361 6c20 6f72 6465 720a 2020  aversal order.  
-00015560: 2020 2020 2020 2020 2020 2020 2020 2367                #g
-00015570: 6574 2074 6865 2061 6e63 6573 746f 7220  et the ancestor 
-00015580: 6e6f 6465 2072 6570 7265 7365 6e74 6174  node representat
-00015590: 696f 6e0a 2020 2020 2020 2020 2020 2020  ion.            
-000155a0: 2020 2020 616e 6365 7374 6f72 203d 2061      ancestor = a
-000155b0: 6e63 6573 746f 725f 696e 666f 5f64 6963  ncestor_info_dic
-000155c0: 745b 6c65 7665 6c5f 7061 6972 5b30 5d5d  t[level_pair[0]]
-000155d0: 2023 6a75 7374 2075 7365 206f 6e65 206f   #just use one o
-000155e0: 6620 7468 656d 2074 6f20 6765 7420 7468  f them to get th
-000155f0: 6520 616e 6365 7374 6f72 0a20 2020 2020  e ancestor.     
-00015600: 2020 2020 2020 2020 2020 206e 6f64 655f             node_
-00015610: 7265 7072 6573 656e 7461 7469 6f6e 203d  representation =
-00015620: 206e 6f64 6573 5f72 6570 7265 7365 6e74   nodes_represent
-00015630: 6174 696f 6e73 5b61 6e63 6573 746f 725d  ations[ancestor]
-00015640: 5b3a 2d64 6567 7265 655f 7472 6565 5d0a  [:-degree_tree].
-00015650: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015660: 6164 6465 645f 7265 7072 6573 656e 7461  added_representa
-00015670: 7469 6f6e 203d 5b30 5d2a 320a 2020 2020  tion =[0]*2.    
-00015680: 2020 2020 2020 2020 2020 2020 666f 7220              for 
-00015690: 696e 6465 782c 6e6f 6465 2069 6e20 656e  index,node in en
-000156a0: 756d 6572 6174 6528 6c65 7665 6c5f 7061  umerate(level_pa
-000156b0: 6972 293a 0a20 2020 2020 2020 2020 2020  ir):.           
-000156c0: 2020 2020 2020 2020 2061 6464 6564 5f72           added_r
-000156d0: 6570 7265 7365 6e74 6174 696f 6e5b 696e  epresentation[in
-000156e0: 6465 785d 203d 310a 2020 2020 2020 2020  dex] =1.        
-000156f0: 2020 2020 2020 2020 2020 2020 6e6f 6465              node
-00015700: 5f72 6570 7265 7365 6e74 6174 696f 6e20  _representation 
-00015710: 3d20 6164 6465 645f 7265 7072 6573 656e  = added_represen
-00015720: 7461 7469 6f6e 202b 206e 6f64 655f 7265  tation + node_re
-00015730: 7072 6573 656e 7461 7469 6f6e 0a20 2020  presentation.   
+00014ce0: 2020 2020 6c61 6265 6c3d 2253 414d 504c      label="SAMPL
+00014cf0: 4544 2229 0a0a 2020 2020 2020 2020 7374  ED")..        st
+00014d00: 6172 7420 2b3d 2031 300a 2020 2020 2320  art += 10.    # 
+00014d10: 6178 6573 2061 6e64 206c 6162 656c 730a  axes and labels.
+00014d20: 2020 2020 6178 2e73 6574 5f78 6c69 6d28      ax.set_xlim(
+00014d30: 2d77 6964 7468 2c20 6e70 2e6d 6178 2869  -width, np.max(i
+00014d40: 6e64 5f74 7261 696e 2920 2b35 290a 2020  nd_train) +5).  
+00014d50: 2020 6178 2e73 6574 5f79 6c69 6d28 302c    ax.set_ylim(0,
+00014d60: 2031 3030 290a 2020 2020 6178 2e73 6574   100).    ax.set
+00014d70: 5f79 6c61 6265 6c28 2725 4944 2729 0a20  _ylabel('%ID'). 
+00014d80: 2020 2061 782e 7365 745f 7469 746c 6528     ax.set_title(
+00014d90: 274f 7665 726c 6170 7069 6e67 2048 6973  'Overlapping His
+00014da0: 746f 6772 616d 2729 0a20 2020 2078 5469  togram').    xTi
+00014db0: 636b 4d61 726b 7320 3d20 5b27 5472 6169  ckMarks = ['Trai
+00014dc0: 6e4e 6f64 655f 7b7d 272e 666f 726d 6174  nNode_{}'.format
+00014dd0: 2869 6e74 2869 2e69 7465 6d28 2929 2920  (int(i.item())) 
+00014de0: 666f 7220 6920 696e 2064 6174 6173 6574  for i in dataset
+00014df0: 5f74 7261 696e 5b3a 2c30 2c31 5d5d 0a20  _train[:,0,1]]. 
+00014e00: 2020 2061 782e 7365 745f 7874 6963 6b73     ax.set_xticks
+00014e10: 2869 6e64 5f74 7261 696e 290a 2020 2020  (ind_train).    
+00014e20: 7874 6963 6b4e 616d 6573 203d 2061 782e  xtickNames = ax.
+00014e30: 7365 745f 7874 6963 6b6c 6162 656c 7328  set_xticklabels(
+00014e40: 7854 6963 6b4d 6172 6b73 290a 2020 2020  xTickMarks).    
+00014e50: 706c 742e 7365 7470 2878 7469 636b 4e61  plt.setp(xtickNa
+00014e60: 6d65 732c 2072 6f74 6174 696f 6e3d 3435  mes, rotation=45
+00014e70: 2c20 666f 6e74 7369 7a65 3d36 290a 2020  , fontsize=6).  
+00014e80: 2020 706c 742e 766c 696e 6573 2869 6e64    plt.vlines(ind
+00014e90: 5f6c 696e 6573 2c79 6d69 6e3d 302c 796d  _lines,ymin=0,ym
+00014ea0: 6178 3d31 3030 290a 2020 2020 2323 2061  ax=100).    ## a
+00014eb0: 6464 2061 206c 6567 656e 642d 2d3e 204e  dd a legend--> N
+00014ec0: 6f74 2077 6f72 6b69 6e67 2028 696e 7369  ot working (insi
+00014ed0: 6465 206f 7220 6f75 7473 6964 6520 6c6f  de or outside lo
+00014ee0: 6f70 290a 2020 2020 6178 2e6c 6567 656e  op).    ax.legen
+00014ef0: 6428 2872 6563 7473 312c 7265 6374 7332  d((rects1,rects2
+00014f00: 292c 2028 2753 616d 706c 6564 272c 224f  ), ('Sampled',"O
+00014f10: 6273 6572 7665 6422 292c 6c6f 633d 312c  bserved"),loc=1,
+00014f20: 6262 6f78 5f74 6f5f 616e 6368 6f72 3d28  bbox_to_anchor=(
+00014f30: 312e 3135 2c20 312e 3129 290a 2020 2020  1.15, 1.1)).    
+00014f40: 706c 742e 7361 7665 6669 6728 227b 7d2f  plt.savefig("{}/
+00014f50: 4f76 6572 6c61 7070 696e 6748 6973 746f  OverlappingHisto
+00014f60: 6772 616d 222e 666f 726d 6174 2872 6573  gram".format(res
+00014f70: 756c 7473 5f64 6972 6563 746f 7279 2929  ults_directory))
+00014f80: 0a64 6566 2074 7265 655f 706f 7369 7469  .def tree_positi
+00014f90: 6f6e 616c 5f65 6d62 6564 6469 6e67 7328  onal_embeddings(
+00014fa0: 616e 6365 7374 6f72 5f69 6e66 6f5f 6469  ancestor_info_di
+00014fb0: 6374 2c74 7265 655f 6279 5f6c 6576 656c  ct,tree_by_level
+00014fc0: 735f 6469 6374 293a 0a20 2020 2022 2222  s_dict):.    """
+00014fd0: 5468 6520 6e6f 6465 7320 6f72 6465 7265  The nodes ordere
+00014fe0: 6420 6279 2074 7265 6520 6c65 7665 6c20  d by tree level 
+00014ff0: 6f72 6465 7220 6172 6520 636f 6e76 6572  order are conver
+00015000: 7465 6420 696e 746f 2073 6f6d 6577 6861  ted into somewha
+00015010: 7420 6f6e 652d 686f 7420 656e 636f 6465  t one-hot encode
+00015020: 6420 6c69 6b65 2065 6d62 6564 6469 6e67  d like embedding
+00015030: 730a 2020 2020 496d 706c 656d 656e 7461  s.    Implementa
+00015040: 7465 6420 666f 6c6c 6f77 696e 670a 2020  ted following.  
+00015050: 2020 6874 7470 733a 2f2f 7061 7065 7273    https://papers
+00015060: 2e6e 6970 732e 6363 2f70 6170 6572 2f32  .nips.cc/paper/2
+00015070: 3031 392f 6669 6c65 2f36 6530 3931 3734  019/file/6e09174
+00015080: 3639 3231 3464 3866 6264 3863 3531 3764  69214d8fbd8c517d
+00015090: 6364 6336 6238 6463 662d 5061 7065 722e  cdc6b8dcf-Paper.
+000150a0: 7064 660a 2020 2020 496e 7075 743a 2054  pdf.    Input: T
+000150b0: 7265 6520 6279 206c 6576 656c 7320 696e  ree by levels in
+000150c0: 2074 7261 7665 7273 616c 206f 7264 6572   traversal order
+000150d0: 0a20 2020 204f 7574 7075 743a 2052 6570  .    Output: Rep
+000150e0: 7265 7365 6e74 6174 696f 6e73 206f 6620  resentations of 
+000150f0: 7468 6520 706f 7369 7469 6f6e 206f 6620  the position of 
+00015100: 7468 6520 6e6f 6465 2069 6e20 7468 6520  the node in the 
+00015110: 7472 6565 2069 6e20 6f6e 6520 686f 7420  tree in one hot 
+00015120: 6269 7420 656e 636f 6469 6e67 2e20 5468  bit encoding. Th
+00015130: 6579 2072 6570 7265 7365 6e74 2074 6865  ey represent the
+00015140: 206f 6e65 2068 6f74 2065 6e63 6f64 6564   one hot encoded
+00015150: 2070 6174 6820 6672 6f6d 2072 6f6f 7420   path from root 
+00015160: 746f 2074 6865 206e 6f64 6520 2222 220a  to the node """.
+00015170: 2020 2020 6465 6772 6565 5f74 7265 6520      degree_tree 
+00015180: 3d32 0a20 2020 206e 5f6c 6576 656c 7320  =2.    n_levels 
+00015190: 3d20 6c65 6e28 7472 6565 5f62 795f 6c65  = len(tree_by_le
+000151a0: 7665 6c73 5f64 6963 7429 0a20 2020 206e  vels_dict).    n
+000151b0: 6f64 6573 5f72 6570 7265 7365 6e74 6174  odes_representat
+000151c0: 696f 6e73 3d64 6566 6175 6c74 6469 6374  ions=defaultdict
+000151d0: 2829 0a20 2020 206e 6f64 655f 7265 7072  ().    node_repr
+000151e0: 6573 656e 7461 7469 6f6e 203d 205b 305d  esentation = [0]
+000151f0: 2a28 6465 6772 6565 5f74 7265 652a 6e5f  *(degree_tree*n_
+00015200: 6c65 7665 6c73 2920 2369 6e69 7469 6174  levels) #initiat
+00015210: 6520 7468 6520 726f 6f74 2074 6f20 616c  e the root to al
+00015220: 6c20 300a 2020 2020 666f 7220 6c65 7665  l 0.    for leve
+00015230: 6c5f 6e61 6d65 2c20 6c65 7665 6c5f 6e6f  l_name, level_no
+00015240: 6465 7320 696e 2073 6f72 7465 6428 7472  des in sorted(tr
+00015250: 6565 5f62 795f 6c65 7665 6c73 5f64 6963  ee_by_levels_dic
+00015260: 742e 6974 656d 7328 2929 3a0a 2020 2020  t.items()):.    
+00015270: 2020 2020 616c 6c5f 616e 6365 7374 6f72      all_ancestor
+00015280: 735f 696e 5f6c 6576 656c 203d 207b 696e  s_in_level = {in
+00015290: 6465 783a 616e 6365 7374 6f72 5f69 6e66  dex:ancestor_inf
+000152a0: 6f5f 6469 6374 5b6e 6f64 655d 2066 6f72  o_dict[node] for
+000152b0: 2069 6e64 6578 2c6e 6f64 6520 696e 2065   index,node in e
+000152c0: 6e75 6d65 7261 7465 286c 6576 656c 5f6e  numerate(level_n
+000152d0: 6f64 6573 297d 2023 6669 6e64 2074 6865  odes)} #find the
+000152e0: 2069 6e64 6578 6573 206f 6620 7468 6520   indexes of the 
+000152f0: 6e6f 6465 7320 7368 6172 696e 6720 7468  nodes sharing th
+00015300: 6520 7361 6d65 2061 6e63 6573 746f 7273  e same ancestors
+00015310: 0a20 2020 2020 2020 2069 6620 6e70 2e6e  .        if np.n
+00015320: 616e 2069 6e20 616c 6c5f 616e 6365 7374  an in all_ancest
+00015330: 6f72 735f 696e 5f6c 6576 656c 2e76 616c  ors_in_level.val
+00015340: 7565 7328 293a 2369 6465 6e74 6966 7920  ues():#identify 
+00015350: 726f 6f74 2c20 6265 6361 7573 6520 6974  root, because it
+00015360: 2068 6173 206e 616e 2061 7320 616e 2061   has nan as an a
+00015370: 6e63 6573 746f 720a 2020 2020 2020 2020  ncestor.        
+00015380: 2020 2020 6e6f 6465 735f 7265 7072 6573      nodes_repres
+00015390: 656e 7461 7469 6f6e 735b 6c65 7665 6c5f  entations[level_
+000153a0: 6e6f 6465 735b 305d 5d20 3d20 6e6f 6465  nodes[0]] = node
+000153b0: 5f72 6570 7265 7365 6e74 6174 696f 6e0a  _representation.
+000153c0: 2020 2020 2020 2020 656c 7365 3a0a 2020          else:.  
+000153d0: 2020 2020 2020 2020 2020 6e6f 6465 5f72            node_r
+000153e0: 6570 7265 7365 6e74 6174 696f 6e20 3d20  epresentation = 
+000153f0: 6e6f 6465 5f72 6570 7265 7365 6e74 6174  node_representat
+00015400: 696f 6e5b 3a2d 6465 6772 6565 5f74 7265  ion[:-degree_tre
+00015410: 655d 0a20 2020 2020 2020 2020 2020 206c  e].            l
+00015420: 6576 656c 5f70 6169 7273 203d 205b 6c65  evel_pairs = [le
+00015430: 7665 6c5f 6e6f 6465 735b 693a 6920 2b20  vel_nodes[i:i + 
+00015440: 6465 6772 6565 5f74 7265 655d 2066 6f72  degree_tree] for
+00015450: 2069 2069 6e20 7261 6e67 6528 302c 206c   i in range(0, l
+00015460: 656e 286c 6576 656c 5f6e 6f64 6573 292c  en(level_nodes),
+00015470: 2064 6567 7265 655f 7472 6565 295d 2023   degree_tree)] #
+00015480: 4469 7669 6465 206c 6576 656c 2069 6e20  Divide level in 
+00015490: 6772 6f75 7073 206f 6620 322c 2074 6865  groups of 2, the
+000154a0: 2074 7265 6520 6973 2069 6e20 7472 6565   tree is in tree
+000154b0: 2074 7261 7665 7273 616c 2c20 736f 2065   traversal, so e
+000154c0: 7665 7279 7468 696e 6720 7368 6f75 646c  verything shoudl
+000154d0: 2062 6520 6f72 6465 7265 640a 2020 2020   be ordered.    
+000154e0: 2020 2020 2020 2020 666f 7220 6c65 7665          for leve
+000154f0: 6c5f 7061 6972 2069 6e20 6c65 7665 6c5f  l_pair in level_
+00015500: 7061 6972 733a 2023 7468 6573 6520 7061  pairs: #these pa
+00015510: 6972 7320 6861 7665 2073 616d 6520 616e  irs have same an
+00015520: 6365 7374 6f72 2062 6563 6175 7365 206f  cestor because o
+00015530: 6620 7472 6565 206c 6576 656c 7320 6172  f tree levels ar
+00015540: 6520 6f72 6761 6e69 7a65 6420 696e 2074  e organized in t
+00015550: 7261 7665 7273 616c 206f 7264 6572 0a20  raversal order. 
+00015560: 2020 2020 2020 2020 2020 2020 2020 2023                 #
+00015570: 6765 7420 7468 6520 616e 6365 7374 6f72  get the ancestor
+00015580: 206e 6f64 6520 7265 7072 6573 656e 7461   node representa
+00015590: 7469 6f6e 0a20 2020 2020 2020 2020 2020  tion.           
+000155a0: 2020 2020 2061 6e63 6573 746f 7220 3d20       ancestor = 
+000155b0: 616e 6365 7374 6f72 5f69 6e66 6f5f 6469  ancestor_info_di
+000155c0: 6374 5b6c 6576 656c 5f70 6169 725b 305d  ct[level_pair[0]
+000155d0: 5d20 236a 7573 7420 7573 6520 6f6e 6520  ] #just use one 
+000155e0: 6f66 2074 6865 6d20 746f 2067 6574 2074  of them to get t
+000155f0: 6865 2061 6e63 6573 746f 720a 2020 2020  he ancestor.    
+00015600: 2020 2020 2020 2020 2020 2020 6e6f 6465              node
+00015610: 5f72 6570 7265 7365 6e74 6174 696f 6e20  _representation 
+00015620: 3d20 6e6f 6465 735f 7265 7072 6573 656e  = nodes_represen
+00015630: 7461 7469 6f6e 735b 616e 6365 7374 6f72  tations[ancestor
+00015640: 5d5b 3a2d 6465 6772 6565 5f74 7265 655d  ][:-degree_tree]
+00015650: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00015660: 2061 6464 6564 5f72 6570 7265 7365 6e74   added_represent
+00015670: 6174 696f 6e20 3d5b 305d 2a32 0a20 2020  ation =[0]*2.   
+00015680: 2020 2020 2020 2020 2020 2020 2066 6f72               for
+00015690: 2069 6e64 6578 2c6e 6f64 6520 696e 2065   index,node in e
+000156a0: 6e75 6d65 7261 7465 286c 6576 656c 5f70  numerate(level_p
+000156b0: 6169 7229 3a0a 2020 2020 2020 2020 2020  air):.          
+000156c0: 2020 2020 2020 2020 2020 6164 6465 645f            added_
+000156d0: 7265 7072 6573 656e 7461 7469 6f6e 5b69  representation[i
+000156e0: 6e64 6578 5d20 3d31 0a20 2020 2020 2020  ndex] =1.       
+000156f0: 2020 2020 2020 2020 2020 2020 206e 6f64               nod
+00015700: 655f 7265 7072 6573 656e 7461 7469 6f6e  e_representation
+00015710: 203d 2061 6464 6564 5f72 6570 7265 7365   = added_represe
+00015720: 6e74 6174 696f 6e20 2b20 6e6f 6465 5f72  ntation + node_r
+00015730: 6570 7265 7365 6e74 6174 696f 6e0a 2020  epresentation.  
 00015740: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015750: 206e 6f64 6573 5f72 6570 7265 7365 6e74   nodes_represent
-00015760: 6174 696f 6e73 5b6e 6f64 655d 203d 206e  ations[node] = n
-00015770: 6f64 655f 7265 7072 6573 656e 7461 7469  ode_representati
-00015780: 6f6e 0a20 2020 2020 2020 2020 2020 2020  on.             
-00015790: 2020 2020 2020 2061 6464 6564 5f72 6570         added_rep
-000157a0: 7265 7365 6e74 6174 696f 6e20 3d20 5b30  resentation = [0
-000157b0: 5d20 2a20 3220 2372 6573 7461 7274 0a20  ] * 2 #restart. 
+00015750: 2020 6e6f 6465 735f 7265 7072 6573 656e    nodes_represen
+00015760: 7461 7469 6f6e 735b 6e6f 6465 5d20 3d20  tations[node] = 
+00015770: 6e6f 6465 5f72 6570 7265 7365 6e74 6174  node_representat
+00015780: 696f 6e0a 2020 2020 2020 2020 2020 2020  ion.            
+00015790: 2020 2020 2020 2020 6164 6465 645f 7265          added_re
+000157a0: 7072 6573 656e 7461 7469 6f6e 203d 205b  presentation = [
+000157b0: 305d 202a 2032 2023 7265 7374 6172 740a  0] * 2 #restart.
 000157c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000157d0: 2020 206e 6f64 655f 7265 7072 6573 656e     node_represen
-000157e0: 7461 7469 6f6e 203d 206e 6f64 6573 5f72  tation = nodes_r
-000157f0: 6570 7265 7365 6e74 6174 696f 6e73 5b61  epresentations[a
-00015800: 6e63 6573 746f 725d 5b3a 2d64 6567 7265  ncestor][:-degre
-00015810: 655f 7472 6565 5d20 2372 6573 7461 7274  e_tree] #restart
-00015820: 0a20 2020 206e 6f64 6573 5f72 6570 7265  .    nodes_repre
-00015830: 7365 6e74 6174 696f 6e73 5f61 7272 6179  sentations_array
-00015840: 203d 206e 702e 7673 7461 636b 286c 6973   = np.vstack(lis
-00015850: 7428 6e6f 6465 735f 7265 7072 6573 656e  t(nodes_represen
-00015860: 7461 7469 6f6e 732e 7661 6c75 6573 2829  tations.values()
-00015870: 2929 0a20 2020 206e 6f64 6573 5f72 6570  )).    nodes_rep
-00015880: 7265 7365 6e74 6174 696f 6e73 5f61 7272  resentations_arr
-00015890: 6179 203d 206e 702e 635f 5b20 6e70 2e61  ay = np.c_[ np.a
-000158a0: 7272 6179 286c 6973 7428 6e6f 6465 735f  rray(list(nodes_
-000158b0: 7265 7072 6573 656e 7461 7469 6f6e 732e  representations.
-000158c0: 6b65 7973 2829 2929 5b3a 2c4e 6f6e 655d  keys()))[:,None]
-000158d0: 2c20 6e6f 6465 735f 7265 7072 6573 656e  , nodes_represen
-000158e0: 7461 7469 6f6e 735f 6172 7261 795d 0a20  tations_array]. 
-000158f0: 2020 2072 6574 7572 6e20 6e6f 6465 735f     return nodes_
-00015900: 7265 7072 6573 656e 7461 7469 6f6e 735f  representations_
-00015910: 6172 7261 790a 6465 6620 6578 7472 615f  array.def extra_
-00015920: 7072 6f63 6573 7369 6e67 2861 6e63 6573  processing(ances
-00015930: 746f 725f 696e 666f 2c70 6174 7269 7374  tor_info,patrist
-00015940: 6963 5f6d 6174 7269 782c 7265 7375 6c74  ic_matrix,result
-00015950: 735f 6469 722c 6172 6773 2c62 7569 6c64  s_dir,args,build
-00015960: 5f63 6f6e 6669 6729 3a0a 2020 2020 2222  _config):.    ""
-00015970: 2243 6f6d 7075 7465 7320 736f 6d65 2064  "Computes some d
-00015980: 6963 7469 6f6e 6172 6965 7320 7769 7468  ictionaries with
-00015990: 2074 6865 206e 6f64 6573 206f 7264 6572   the nodes order
-000159a0: 6564 2069 6e20 7472 6565 206c 6576 656c  ed in tree level
-000159b0: 206f 7264 6572 2c20 7072 6f63 6573 7365   order, processe
-000159c0: 7320 7468 6520 626c 6f73 756d 206d 6174  s the blosum mat
-000159d0: 7269 7820 616e 6420 6275 696c 6420 6120  rix and build a 
-000159e0: 6772 6170 6820 7769 7468 2074 6865 2074  graph with the t
-000159f0: 7265 6520 6e6f 6465 7320 616e 6420 7468  ree nodes and th
-00015a00: 6520 6272 616e 6368 206c 656e 6774 6873  e branch lengths
-00015a10: 2061 7320 6564 6765 730a 2020 2020 203a   as edges.     :
-00015a20: 7061 7261 6d20 6e61 6d65 6474 7570 6c65  param namedtuple
-00015a30: 2061 6e63 6573 746f 725f 696e 666f 0a20   ancestor_info. 
-00015a40: 2020 2020 3a70 6172 616d 2074 656e 736f      :param tenso
-00015a50: 7220 7061 7472 6973 7469 635f 6d61 7472  r patristic_matr
-00015a60: 6978 3a20 5b6e 5f6c 6561 7665 7320 2b20  ix: [n_leaves + 
-00015a70: 6e5f 696e 7465 726e 616c 202b 2031 2c20  n_internal + 1, 
-00015a80: 6e5f 6c65 6176 6573 202b 206e 5f69 6e74  n_leaves + n_int
-00015a90: 6572 6e61 6c20 2b20 315d 0a20 2020 2020  ernal + 1].     
-00015aa0: 3a70 6172 616d 2073 7472 2072 6573 756c  :param str resul
-00015ab0: 7473 5f64 6972 0a20 2020 2020 3a70 6172  ts_dir.     :par
-00015ac0: 616d 206e 616d 6564 7475 706c 6520 6172  am namedtuple ar
-00015ad0: 6773 0a20 2020 2020 3a70 6172 616d 206e  gs.     :param n
-00015ae0: 616d 6564 7475 706c 6520 6275 696c 645f  amedtuple build_
-00015af0: 636f 6e66 6967 2222 220a 2020 2020 4164  config""".    Ad
-00015b00: 6469 7469 6f6e 616c 496e 666f 203d 206e  ditionalInfo = n
-00015b10: 616d 6564 7475 706c 6528 2241 6464 6974  amedtuple("Addit
-00015b20: 696f 6e61 6c49 6e66 6f22 2c0a 2020 2020  ionalInfo",.    
+000157d0: 2020 2020 6e6f 6465 5f72 6570 7265 7365      node_represe
+000157e0: 6e74 6174 696f 6e20 3d20 6e6f 6465 735f  ntation = nodes_
+000157f0: 7265 7072 6573 656e 7461 7469 6f6e 735b  representations[
+00015800: 616e 6365 7374 6f72 5d5b 3a2d 6465 6772  ancestor][:-degr
+00015810: 6565 5f74 7265 655d 2023 7265 7374 6172  ee_tree] #restar
+00015820: 740a 2020 2020 6e6f 6465 735f 7265 7072  t.    nodes_repr
+00015830: 6573 656e 7461 7469 6f6e 735f 6172 7261  esentations_arra
+00015840: 7920 3d20 6e70 2e76 7374 6163 6b28 6c69  y = np.vstack(li
+00015850: 7374 286e 6f64 6573 5f72 6570 7265 7365  st(nodes_represe
+00015860: 6e74 6174 696f 6e73 2e76 616c 7565 7328  ntations.values(
+00015870: 2929 290a 2020 2020 6e6f 6465 735f 7265  ))).    nodes_re
+00015880: 7072 6573 656e 7461 7469 6f6e 735f 6172  presentations_ar
+00015890: 7261 7920 3d20 6e70 2e63 5f5b 206e 702e  ray = np.c_[ np.
+000158a0: 6172 7261 7928 6c69 7374 286e 6f64 6573  array(list(nodes
+000158b0: 5f72 6570 7265 7365 6e74 6174 696f 6e73  _representations
+000158c0: 2e6b 6579 7328 2929 295b 3a2c 4e6f 6e65  .keys()))[:,None
+000158d0: 5d2c 206e 6f64 6573 5f72 6570 7265 7365  ], nodes_represe
+000158e0: 6e74 6174 696f 6e73 5f61 7272 6179 5d0a  ntations_array].
+000158f0: 2020 2020 7265 7475 726e 206e 6f64 6573      return nodes
+00015900: 5f72 6570 7265 7365 6e74 6174 696f 6e73  _representations
+00015910: 5f61 7272 6179 0a64 6566 2065 7874 7261  _array.def extra
+00015920: 5f70 726f 6365 7373 696e 6728 616e 6365  _processing(ance
+00015930: 7374 6f72 5f69 6e66 6f2c 7061 7472 6973  stor_info,patris
+00015940: 7469 635f 6d61 7472 6978 2c72 6573 756c  tic_matrix,resul
+00015950: 7473 5f64 6972 2c61 7267 732c 6275 696c  ts_dir,args,buil
+00015960: 645f 636f 6e66 6967 293a 0a20 2020 2022  d_config):.    "
+00015970: 2222 436f 6d70 7574 6573 2073 6f6d 6520  ""Computes some 
+00015980: 6469 6374 696f 6e61 7269 6573 2077 6974  dictionaries wit
+00015990: 6820 7468 6520 6e6f 6465 7320 6f72 6465  h the nodes orde
+000159a0: 7265 6420 696e 2074 7265 6520 6c65 7665  red in tree leve
+000159b0: 6c20 6f72 6465 722c 2070 726f 6365 7373  l order, process
+000159c0: 6573 2074 6865 2062 6c6f 7375 6d20 6d61  es the blosum ma
+000159d0: 7472 6978 2061 6e64 2062 7569 6c64 2061  trix and build a
+000159e0: 2067 7261 7068 2077 6974 6820 7468 6520   graph with the 
+000159f0: 7472 6565 206e 6f64 6573 2061 6e64 2074  tree nodes and t
+00015a00: 6865 2062 7261 6e63 6820 6c65 6e67 7468  he branch length
+00015a10: 7320 6173 2065 6467 6573 0a20 2020 2020  s as edges.     
+00015a20: 3a70 6172 616d 206e 616d 6564 7475 706c  :param namedtupl
+00015a30: 6520 616e 6365 7374 6f72 5f69 6e66 6f0a  e ancestor_info.
+00015a40: 2020 2020 203a 7061 7261 6d20 7465 6e73       :param tens
+00015a50: 6f72 2070 6174 7269 7374 6963 5f6d 6174  or patristic_mat
+00015a60: 7269 783a 205b 6e5f 6c65 6176 6573 202b  rix: [n_leaves +
+00015a70: 206e 5f69 6e74 6572 6e61 6c20 2b20 312c   n_internal + 1,
+00015a80: 206e 5f6c 6561 7665 7320 2b20 6e5f 696e   n_leaves + n_in
+00015a90: 7465 726e 616c 202b 2031 5d0a 2020 2020  ternal + 1].    
+00015aa0: 203a 7061 7261 6d20 7374 7220 7265 7375   :param str resu
+00015ab0: 6c74 735f 6469 720a 2020 2020 203a 7061  lts_dir.     :pa
+00015ac0: 7261 6d20 6e61 6d65 6474 7570 6c65 2061  ram namedtuple a
+00015ad0: 7267 730a 2020 2020 203a 7061 7261 6d20  rgs.     :param 
+00015ae0: 6e61 6d65 6474 7570 6c65 2062 7569 6c64  namedtuple build
+00015af0: 5f63 6f6e 6669 6722 2222 0a20 2020 2041  _config""".    A
+00015b00: 6464 6974 696f 6e61 6c49 6e66 6f20 3d20  dditionalInfo = 
+00015b10: 6e61 6d65 6474 7570 6c65 2822 4164 6469  namedtuple("Addi
+00015b20: 7469 6f6e 616c 496e 666f 222c 0a20 2020  tionalInfo",.   
 00015b30: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015b40: 2020 2020 2020 2020 5b22 626c 6f73 756d          ["blosum
-00015b50: 2220 2c22 626c 6f73 756d 5f64 6963 7422  " ,"blosum_dict"
-00015b60: 2c22 6368 696c 6472 656e 5f64 6963 7422  ,"children_dict"
-00015b70: 2c22 616e 6365 7374 6f72 5f69 6e66 6f5f  ,"ancestor_info_
-00015b80: 6469 6374 222c 2022 7472 6565 5f62 795f  dict", "tree_by_
-00015b90: 6c65 7665 6c73 5f64 6963 7422 2c20 2274  levels_dict", "t
-00015ba0: 7265 655f 6279 5f6c 6576 656c 735f 6172  ree_by_levels_ar
-00015bb0: 7261 7922 2c0a 2020 2020 2020 2020 2020  ray",.          
+00015b40: 2020 2020 2020 2020 205b 2262 6c6f 7375           ["blosu
+00015b50: 6d22 202c 2262 6c6f 7375 6d5f 6469 6374  m" ,"blosum_dict
+00015b60: 222c 2263 6869 6c64 7265 6e5f 6469 6374  ","children_dict
+00015b70: 222c 2261 6e63 6573 746f 725f 696e 666f  ","ancestor_info
+00015b80: 5f64 6963 7422 2c20 2274 7265 655f 6279  _dict", "tree_by
+00015b90: 5f6c 6576 656c 735f 6469 6374 222c 2022  _levels_dict", "
+00015ba0: 7472 6565 5f62 795f 6c65 7665 6c73 5f61  tree_by_levels_a
+00015bb0: 7272 6179 222c 0a20 2020 2020 2020 2020  rray",.         
 00015bc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00015bd0: 2020 2022 7061 7472 6973 7469 635f 696e     "patristic_in
-00015be0: 666f 222c 2022 6772 6170 685f 636f 6f22  fo", "graph_coo"
-00015bf0: 2c22 7061 7472 6973 7469 635f 6675 6c6c  ,"patristic_full
-00015c00: 5f73 7061 7273 6522 2c22 6e6f 6465 735f  _sparse","nodes_
-00015c10: 7265 7072 6573 656e 7461 7469 6f6e 735f  representations_
-00015c20: 6172 7261 7922 2c22 6467 6c5f 6772 6170  array","dgl_grap
-00015c30: 6822 5d29 0a0a 2020 2020 616e 6365 7374  h"])..    ancest
-00015c40: 6f72 5f69 6e66 6f5f 6469 6374 203d 2064  or_info_dict = d
-00015c50: 6963 7428 7a69 7028 616e 6365 7374 6f72  ict(zip(ancestor
-00015c60: 5f69 6e66 6f5b 3a2c 305d 2e74 6f6c 6973  _info[:,0].tolis
-00015c70: 7428 292c 616e 6365 7374 6f72 5f69 6e66  t(),ancestor_inf
-00015c80: 6f5b 3a2c 325d 2e74 6f6c 6973 7428 2929  o[:,2].tolist())
-00015c90: 290a 2020 2020 7069 636b 6c65 2e64 756d  ).    pickle.dum
-00015ca0: 7028 616e 6365 7374 6f72 5f69 6e66 6f5f  p(ancestor_info_
-00015cb0: 6469 6374 2c20 6f70 656e 2827 7b7d 2f41  dict, open('{}/A
-00015cc0: 6e63 6573 746f 725f 696e 666f 5f64 6963  ncestor_info_dic
-00015cd0: 742e 7027 2e66 6f72 6d61 7428 7265 7375  t.p'.format(resu
-00015ce0: 6c74 735f 6469 7229 2c20 2777 6227 292c  lts_dir), 'wb'),
-00015cf0: 2070 726f 746f 636f 6c3d 7069 636b 6c65   protocol=pickle
-00015d00: 2e48 4947 4845 5354 5f50 524f 544f 434f  .HIGHEST_PROTOCO
-00015d10: 4c29 0a20 2020 2069 6620 6973 696e 7374  L).    if isinst
-00015d20: 616e 6365 2870 6174 7269 7374 6963 5f6d  ance(patristic_m
-00015d30: 6174 7269 782c 6e70 2e6e 6461 7272 6179  atrix,np.ndarray
-00015d40: 293a 0a20 2020 2020 2020 2070 6174 7269  ):.        patri
-00015d50: 7374 6963 5f69 6e66 6f20 3d20 6469 6374  stic_info = dict
-00015d60: 287a 6970 286c 6973 7428 7261 6e67 6528  (zip(list(range(
-00015d70: 6c65 6e28 7061 7472 6973 7469 635f 6d61  len(patristic_ma
-00015d80: 7472 6978 292d 3129 292c 7061 7472 6973  trix)-1)),patris
-00015d90: 7469 635f 6d61 7472 6978 5b31 3a2c 305d  tic_matrix[1:,0]
-00015da0: 2e61 7374 7970 6528 696e 7429 2e74 6f6c  .astype(int).tol
-00015db0: 6973 7428 2929 2920 2353 6b69 7020 7468  ist())) #Skip th
-00015dc0: 6520 6661 6b65 2068 6561 6465 720a 2020  e fake header.  
-00015dd0: 2020 656c 7365 3a0a 0a20 2020 2020 2020    else:..       
-00015de0: 2070 6174 7269 7374 6963 5f69 6e66 6f20   patristic_info 
-00015df0: 3d64 6963 7428 7a69 7028 6c69 7374 2872  =dict(zip(list(r
-00015e00: 616e 6765 286c 656e 2870 6174 7269 7374  ange(len(patrist
-00015e10: 6963 5f6d 6174 7269 7829 2d31 2929 2c70  ic_matrix)-1)),p
-00015e20: 6174 7269 7374 6963 5f6d 6174 7269 785b  atristic_matrix[
-00015e30: 313a 2c30 5d2e 7479 7065 2874 6f72 6368  1:,0].type(torch
-00015e40: 2e69 6e74 292e 746f 6c69 7374 2829 2929  .int).tolist()))
-00015e50: 0a0a 2020 2020 616e 6365 7374 6f72 735f  ..    ancestors_
-00015e60: 696e 666f 5f66 6c69 7070 6564 203d 206e  info_flipped = n
-00015e70: 702e 666c 6970 2861 6e63 6573 746f 725f  p.flip(ancestor_
-00015e80: 696e 666f 2c20 6178 6973 3d31 2920 2353  info, axis=1) #S
-00015e90: 6f20 7765 2063 616e 2065 7661 6c75 6174  o we can evaluat
-00015ea0: 6520 696e 2072 6576 6572 7365 2c20 7765  e in reverse, we
-00015eb0: 2073 7461 7274 2077 6974 6820 7468 6520   start with the 
-00015ec0: 6c61 7374 2063 6f6c 756d 6e0a 0a20 2020  last column..   
-00015ed0: 2074 7265 655f 6279 5f6c 6576 656c 7320   tree_by_levels 
-00015ee0: 3d20 5b5d 0a20 2020 2066 6f72 2063 6f6c  = [].    for col
-00015ef0: 756d 6e20 696e 2061 6e63 6573 746f 7273  umn in ancestors
-00015f00: 5f69 6e66 6f5f 666c 6970 7065 642e 543a  _info_flipped.T:
-00015f10: 0a20 2020 2020 2020 2069 6e64 6578 6573  .        indexes
-00015f20: 203d 206e 702e 7768 6572 6528 636f 6c75   = np.where(colu
-00015f30: 6d6e 203d 3d20 3029 0a20 2020 2020 2020  mn == 0).       
-00015f40: 206c 6576 656c 203d 2061 6e63 6573 746f   level = ancesto
-00015f50: 7273 5f69 6e66 6f5f 666c 6970 7065 642e  rs_info_flipped.
-00015f60: 545b 2d31 5d5b 696e 6465 7865 735d 0a20  T[-1][indexes]. 
-00015f70: 2020 2020 2020 2074 7265 655f 6279 5f6c         tree_by_l
-00015f80: 6576 656c 732e 6170 7065 6e64 286c 6576  evels.append(lev
-00015f90: 656c 290a 0a20 2020 2074 7265 655f 6279  el)..    tree_by
-00015fa0: 5f6c 6576 656c 7320 3d20 7472 6565 5f62  _levels = tree_b
-00015fb0: 795f 6c65 7665 6c73 5b3a 2d31 5d20 2374  y_levels[:-1] #t
-00015fc0: 6865 2072 6f6f 7420 6765 7473 2061 6464  he root gets add
-00015fd0: 6564 2074 7769 6365 2062 6563 6175 7365  ed twice because
-00015fe0: 2074 6865 2066 6972 7374 2063 6f6c 756d   the first colum
-00015ff0: 6e20 6172 6520 7468 6520 6e6f 6465 2069  n are the node i
-00016000: 6e64 6578 6573 2c20 6361 6e6e 6f74 2062  ndexes, cannot b
-00016010: 6520 6669 7865 642c 206f 7468 6572 7769  e fixed, otherwi
-00016020: 7365 2077 6520 6361 6e6e 6f74 2064 6f20  se we cannot do 
-00016030: 7468 6520 7472 6963 6b0a 0a20 2020 2074  the trick..    t
-00016040: 7265 655f 6279 5f6c 6576 656c 735f 6469  ree_by_levels_di
-00016050: 6374 203d 2064 6963 7428 7a69 7028 6c69  ct = dict(zip(li
-00016060: 7374 2872 6576 6572 7365 6428 7261 6e67  st(reversed(rang
-00016070: 6528 6c65 6e28 7472 6565 5f62 795f 6c65  e(len(tree_by_le
-00016080: 7665 6c73 2929 2929 2c74 7265 655f 6279  vels)))),tree_by
-00016090: 5f6c 6576 656c 7329 290a 2020 2020 7069  _levels)).    pi
-000160a0: 636b 6c65 2e64 756d 7028 7472 6565 5f62  ckle.dump(tree_b
-000160b0: 795f 6c65 7665 6c73 5f64 6963 742c 206f  y_levels_dict, o
-000160c0: 7065 6e28 277b 7d2f 5472 6565 5f62 795f  pen('{}/Tree_by_
-000160d0: 6c65 7665 6c73 5f64 6963 742e 7027 2e66  levels_dict.p'.f
-000160e0: 6f72 6d61 7428 7265 7375 6c74 735f 6469  ormat(results_di
-000160f0: 7229 2c20 2777 6227 292c 2070 726f 746f  r), 'wb'), proto
-00016100: 636f 6c3d 7069 636b 6c65 2e48 4947 4845  col=pickle.HIGHE
-00016110: 5354 5f50 524f 544f 434f 4c29 0a0a 2020  ST_PROTOCOL)..  
-00016120: 2020 6c65 6e67 7468 203d 206d 6178 286d    length = max(m
-00016130: 6170 286c 656e 2c20 7472 6565 5f62 795f  ap(len, tree_by_
-00016140: 6c65 7665 6c73 5f64 6963 742e 7661 6c75  levels_dict.valu
-00016150: 6573 2829 2929 0a20 2020 2074 7265 655f  es())).    tree_
-00016160: 6279 5f6c 6576 656c 735f 6172 7261 7920  by_levels_array 
-00016170: 3d20 6e70 2e61 7272 6179 285b 7869 2e74  = np.array([xi.t
-00016180: 6f6c 6973 7428 2920 2b20 5b4e 6f6e 655d  olist() + [None]
-00016190: 202a 2028 6c65 6e67 7468 202d 206c 656e   * (length - len
-000161a0: 2878 6929 2920 666f 7220 7869 2069 6e20  (xi)) for xi in 
-000161b0: 7472 6565 5f62 795f 6c65 7665 6c73 5d29  tree_by_levels])
-000161c0: 0a20 2020 2023 2074 7265 655f 6c65 7665  .    # tree_leve
-000161d0: 6c73 5f32 203d 2043 7265 6174 655f 7472  ls_2 = Create_tr
-000161e0: 6565 5f62 795f 6c65 7665 6c73 2863 6869  ee_by_levels(chi
-000161f0: 6c64 7265 6e5f 6469 6374 290a 0a20 2020  ldren_dict)..   
-00016200: 2063 6869 6c64 7265 6e5f 6469 6374 203d   children_dict =
-00016210: 207b 7d0a 2020 2020 666f 7220 6b2c 2076   {}.    for k, v
-00016220: 2069 6e20 616e 6365 7374 6f72 5f69 6e66   in ancestor_inf
-00016230: 6f5f 6469 6374 2e69 7465 6d73 2829 3a0a  o_dict.items():.
-00016240: 2020 2020 2020 2020 6368 696c 6472 656e          children
-00016250: 5f64 6963 745b 765d 203d 2063 6869 6c64  _dict[v] = child
-00016260: 7265 6e5f 6469 6374 2e67 6574 2876 2c20  ren_dict.get(v, 
-00016270: 5b5d 290a 2020 2020 2020 2020 6368 696c  []).        chil
-00016280: 6472 656e 5f64 6963 745b 765d 2e61 7070  dren_dict[v].app
-00016290: 656e 6428 6b29 0a20 2020 2070 6963 6b6c  end(k).    pickl
-000162a0: 652e 6475 6d70 2863 6869 6c64 7265 6e5f  e.dump(children_
-000162b0: 6469 6374 2c20 6f70 656e 2827 7b7d 2f43  dict, open('{}/C
-000162c0: 6869 6c64 7265 6e5f 6469 6374 2e70 272e  hildren_dict.p'.
-000162d0: 666f 726d 6174 2872 6573 756c 7473 5f64  format(results_d
-000162e0: 6972 292c 2027 7762 2729 2c0a 2020 2020  ir), 'wb'),.    
-000162f0: 2020 2020 2020 2020 2020 2020 7072 6f74              prot
-00016300: 6f63 6f6c 3d70 6963 6b6c 652e 4849 4748  ocol=pickle.HIGH
-00016310: 4553 545f 5052 4f54 4f43 4f4c 290a 0a20  EST_PROTOCOL).. 
-00016320: 2020 2023 2268 7474 7073 3a2f 2f70 7974     #"https://pyt
-00016330: 6f72 6368 2d67 656f 6d65 7472 6963 2e72  orch-geometric.r
-00016340: 6561 6474 6865 646f 6373 2e69 6f2f 656e  eadthedocs.io/en
-00016350: 2f6c 6174 6573 742f 6e6f 7465 732f 696e  /latest/notes/in
-00016360: 7472 6f64 7563 7469 6f6e 2e68 746d 6c22  troduction.html"
-00016370: 0a20 2020 2023 4869 6768 6c69 6768 743a  .    #Highlight:
-00016380: 4275 696c 6420 7468 6520 6d61 7472 6978  Build the matrix
-00016390: 2072 6570 7265 7365 6e74 6174 696f 6e20   representation 
-000163a0: 6f66 2074 6865 2074 7265 6520 6772 6170  of the tree grap
-000163b0: 683a 2044 6972 6563 7465 6420 2872 6f6f  h: Directed (roo
-000163c0: 742d 3e20 6c65 6176 6573 2920 7765 6967  t-> leaves) weig
-000163d0: 6874 6564 2028 7061 7472 6973 7469 6320  hted (patristic 
-000163e0: 6469 7374 2920 6772 6170 680a 2020 2020  dist) graph.    
-000163f0: 6966 2062 7569 6c64 5f63 6f6e 6669 672e  if build_config.
-00016400: 6275 696c 645f 6772 6170 683a 2023 544f  build_graph: #TO
-00016410: 444f 3a20 4669 7820 696e 2073 6572 7665  DO: Fix in serve
-00016420: 720a 2020 2020 2020 2020 7072 696e 7428  r.        print(
-00016430: 2242 7569 6c64 696e 6720 6120 7472 6565  "Building a tree
-00016440: 2d67 7261 7068 2229 0a20 2020 2020 2020  -graph").       
-00016450: 2023 6d61 6b65 2067 7261 7068 2061 646a   #make graph adj
-00016460: 6163 656e 7420 6d61 7472 6978 0a20 2020  acent matrix.   
-00016470: 2020 2020 2070 6174 7269 7374 6963 5f6d       patristic_m
-00016480: 6174 7269 7820 3d20 7061 7472 6973 7469  atrix = patristi
-00016490: 635f 6d61 7472 6978 5b70 6174 7269 7374  c_matrix[patrist
-000164a0: 6963 5f6d 6174 7269 785b 3a2c 2030 5d2e  ic_matrix[:, 0].
-000164b0: 6172 6773 6f72 7428 295d 0a20 2020 2020  argsort()].     
-000164c0: 2020 2070 6174 7269 7374 6963 5f6d 6174     patristic_mat
-000164d0: 7269 7820 3d20 7061 7472 6973 7469 635f  rix = patristic_
-000164e0: 6d61 7472 6978 5b3a 2c20 7061 7472 6973  matrix[:, patris
-000164f0: 7469 635f 6d61 7472 6978 5b30 2c20 3a5d  tic_matrix[0, :]
-00016500: 2e61 7267 736f 7274 2829 5d0a 0a20 2020  .argsort()]..   
-00016510: 2020 2020 2067 7261 7068 5f6e 6f64 655f       graph_node_
-00016520: 6d61 7472 6978 203d 206e 702e 7a65 726f  matrix = np.zero
-00016530: 735f 6c69 6b65 2870 6174 7269 7374 6963  s_like(patristic
-00016540: 5f6d 6174 7269 7829 2023 4469 7265 6374  _matrix) #Direct
-00016550: 6564 2028 726f 6f74 2d3e 206c 6561 7665  ed (root-> leave
-00016560: 7329 204e 4f54 2077 6569 6768 7465 6420  s) NOT weighted 
-00016570: 6772 6170 680a 2020 2020 2020 2020 6564  graph.        ed
-00016580: 6765 5f77 6569 6768 745f 6d61 7472 6978  ge_weight_matrix
-00016590: 203d 206e 702e 7a65 726f 735f 6c69 6b65   = np.zeros_like
-000165a0: 2870 6174 7269 7374 6963 5f6d 6174 7269  (patristic_matri
-000165b0: 7829 2023 4469 7265 6374 6564 2028 726f  x) #Directed (ro
-000165c0: 6f74 2d3e 206c 6561 7665 7329 2077 6569  ot-> leaves) wei
-000165d0: 6768 7465 6420 2870 6174 7269 7374 6963  ghted (patristic
-000165e0: 2064 6973 7429 2067 7261 7068 0a20 2020   dist) graph.   
-000165f0: 2020 2020 2067 7261 7068 5f6e 6f64 655f       graph_node_
-00016600: 6d61 7472 6978 5b30 2c3a 5d20 3d20 7061  matrix[0,:] = pa
-00016610: 7472 6973 7469 635f 6d61 7472 6978 5b30  tristic_matrix[0
-00016620: 2c3a 5d0a 2020 2020 2020 2020 6772 6170  ,:].        grap
-00016630: 685f 6e6f 6465 5f6d 6174 7269 785b 3a2c  h_node_matrix[:,
-00016640: 305d 203d 2070 6174 7269 7374 6963 5f6d  0] = patristic_m
-00016650: 6174 7269 785b 3a2c 305d 0a20 2020 2020  atrix[:,0].     
-00016660: 2020 2065 6467 655f 7765 6967 6874 5f6d     edge_weight_m
-00016670: 6174 7269 785b 302c 203a 5d20 3d20 7061  atrix[0, :] = pa
-00016680: 7472 6973 7469 635f 6d61 7472 6978 5b30  tristic_matrix[0
-00016690: 2c20 3a5d 0a20 2020 2020 2020 2065 6467  , :].        edg
-000166a0: 655f 7765 6967 6874 5f6d 6174 7269 785b  e_weight_matrix[
-000166b0: 3a2c 2030 5d20 3d20 7061 7472 6973 7469  :, 0] = patristi
-000166c0: 635f 6d61 7472 6978 5b3a 2c20 305d 0a20  c_matrix[:, 0]. 
-000166d0: 2020 2020 2020 2066 6f72 2061 6e63 6573         for ances
-000166e0: 746f 722c 6368 696c 6472 656e 2069 6e20  tor,children in 
-000166f0: 6368 696c 6472 656e 5f64 6963 742e 6974  children_dict.it
-00016700: 656d 7328 293a 0a20 2020 2020 2020 2020  ems():.         
-00016710: 2020 2061 6e63 6573 746f 725f 6964 7820     ancestor_idx 
-00016720: 3d20 6e70 2e69 6e31 6428 6772 6170 685f  = np.in1d(graph_
-00016730: 6e6f 6465 5f6d 6174 7269 785b 302c 3a5d  node_matrix[0,:]
-00016740: 2c20 616e 6365 7374 6f72 292e 6e6f 6e7a  , ancestor).nonz
-00016750: 6572 6f28 295b 305d 0a20 2020 2020 2020  ero()[0].       
-00016760: 2020 2020 2063 6869 6c64 7265 6e5f 6964       children_id
-00016770: 783d 206e 702e 696e 3164 2867 7261 7068  x= np.in1d(graph
-00016780: 5f6e 6f64 655f 6d61 7472 6978 5b30 2c3a  _node_matrix[0,:
-00016790: 5d2c 2063 6869 6c64 7265 6e29 2e6e 6f6e  ], children).non
-000167a0: 7a65 726f 2829 5b30 5d0a 2020 2020 2020  zero()[0].      
-000167b0: 2020 2020 2020 6772 6170 685f 6e6f 6465        graph_node
-000167c0: 5f6d 6174 7269 785b 616e 6365 7374 6f72  _matrix[ancestor
-000167d0: 5f69 6478 2c63 6869 6c64 7265 6e5f 6964  _idx,children_id
-000167e0: 785d 203d 2031 2023 636f 6e74 6169 6e73  x] = 1 #contains
-000167f0: 206e 6f20 6564 6765 7320 7765 6967 6874   no edges weight
-00016800: 730a 2020 2020 2020 2020 2020 2020 6772  s.            gr
-00016810: 6170 685f 6e6f 6465 5f6d 6174 7269 785b  aph_node_matrix[
-00016820: 6368 696c 6472 656e 5f69 6478 2c61 6e63  children_idx,anc
-00016830: 6573 746f 725f 6964 785d 203d 2031 0a20  estor_idx] = 1. 
-00016840: 2020 2020 2020 2020 2020 2065 6467 655f             edge_
-00016850: 7765 6967 6874 5f6d 6174 7269 785b 616e  weight_matrix[an
-00016860: 6365 7374 6f72 5f69 6478 2c20 6368 696c  cestor_idx, chil
-00016870: 6472 656e 5f69 6478 5d20 3d20 7061 7472  dren_idx] = patr
-00016880: 6973 7469 635f 6d61 7472 6978 5b61 6e63  istic_matrix[anc
-00016890: 6573 746f 725f 6964 782c 2063 6869 6c64  estor_idx, child
-000168a0: 7265 6e5f 6964 785d 0a20 2020 2020 2020  ren_idx].       
-000168b0: 2020 2020 2023 6564 6765 5f77 6569 6768       #edge_weigh
-000168c0: 745f 6d61 7472 6978 5b63 6869 6c64 7265  t_matrix[childre
-000168d0: 6e5f 6964 782c 2061 6e63 6573 746f 725f  n_idx, ancestor_
-000168e0: 6964 785d 203d 2070 6174 7269 7374 6963  idx] = patristic
-000168f0: 5f6d 6174 7269 785b 6368 696c 6472 656e  _matrix[children
-00016900: 5f69 6478 2c20 616e 6365 7374 6f72 5f69  _idx, ancestor_i
-00016910: 6478 5d20 2348 6967 686c 6967 6874 3a20  dx] #Highlight: 
-00016920: 4e6f 7420 7573 6520 746f 2061 766f 6964  Not use to avoid
-00016930: 2063 7265 6174 696e 6720 6120 7365 6c66   creating a self
-00016940: 206c 6f6f 7065 6420 6772 6170 6820 6f72   looped graph or
-00016950: 2062 6964 6972 6563 7469 6f6e 616c 2067   bidirectional g
-00016960: 7261 7068 730a 2020 2020 2020 2020 2343  raphs.        #C
-00016970: 7265 6174 6520 6120 636f 6f72 6469 6e61  reate a coordina
-00016980: 7465 7320 6d61 7472 6978 2c20 6261 7369  tes matrix, basi
-00016990: 6361 6c6c 7920 706f 696e 7473 206f 7574  cally points out
-000169a0: 2077 6869 6368 2072 6567 696f 6e73 2066   which regions f
-000169b0: 726f 6d20 7468 6520 6d61 7472 6978 2064  rom the matrix d
-000169c0: 6f20 6e6f 7420 6861 7665 2030 2076 616c  o not have 0 val
-000169d0: 7565 7320 616e 6420 7468 6572 6566 6f72  ues and therefor
-000169e0: 6520 6172 6520 636f 6e6e 6563 7465 640a  e are connected.
-000169f0: 2020 2020 2020 2020 7765 6967 6874 735f          weights_
-00016a00: 636f 6f20 3d20 636f 6f5f 6d61 7472 6978  coo = coo_matrix
-00016a10: 2865 6467 655f 7765 6967 6874 5f6d 6174  (edge_weight_mat
-00016a20: 7269 785b 313a 2c31 3a5d 290a 2020 2020  rix[1:,1:]).    
-00016a30: 2020 2020 7472 793a 0a20 2020 2020 2020      try:.       
-00016a40: 2020 2020 2023 5468 6520 6772 6170 6820       #The graph 
-00016a50: 636f 6e6e 6563 7469 7669 7479 2028 6564  connectivity (ed
-00016a60: 6765 2069 6e64 6578 2920 7368 6f75 6c64  ge index) should
-00016a70: 2062 6520 636f 6e66 696e 6564 2077 6974   be confined wit
-00016a80: 6820 7468 6520 434f 4f20 666f 726d 6174  h the COO format
-00016a90: 2c0a 2020 2020 2020 2020 2020 2020 2320  ,.            # 
-00016aa0: 692e 652e 2074 6865 2066 6972 7374 206c  i.e. the first l
-00016ab0: 6973 7420 636f 6e74 6169 6e73 2074 6865  ist contains the
-00016ac0: 2069 6e64 6578 206f 6620 7468 6520 736f   index of the so
-00016ad0: 7572 6365 206e 6f64 6573 2c20 7768 696c  urce nodes, whil
-00016ae0: 6520 7468 6520 696e 6465 7820 6f66 2074  e the index of t
-00016af0: 6172 6765 7420 6e6f 6465 7320 6973 2073  arget nodes is s
-00016b00: 7065 6369 6669 6564 2069 6e20 7468 6520  pecified in the 
-00016b10: 7365 636f 6e64 206c 6973 742e 0a20 2020  second list..   
-00016b20: 2020 2020 2020 2020 2023 4e6f 7465 2074           #Note t
-00016b30: 6861 7420 7468 6520 6f72 6465 7220 6f66  hat the order of
-00016b40: 2074 6865 2065 6467 6520 696e 6465 7820   the edge index 
-00016b50: 6973 2069 7272 656c 6576 616e 7420 746f  is irrelevant to
-00016b60: 2074 6865 2044 6174 6120 6f62 6a65 6374   the Data object
-00016b70: 2079 6f75 2063 7265 6174 6520 7369 6e63   you create sinc
-00016b80: 6520 7375 6368 2069 6e66 6f72 6d61 7469  e such informati
-00016b90: 6f6e 2069 7320 6f6e 6c79 2066 6f72 2063  on is only for c
-00016ba0: 6f6d 7075 7469 6e67 2074 6865 2061 646a  omputing the adj
-00016bb0: 6163 656e 6379 206d 6174 7269 782e 0a20  acency matrix.. 
-00016bc0: 2020 2020 2020 2020 2020 2066 726f 6d20             from 
-00016bd0: 746f 7263 685f 6765 6f6d 6574 7269 632e  torch_geometric.
-00016be0: 7574 696c 732e 636f 6e76 6572 7420 696d  utils.convert im
-00016bf0: 706f 7274 2066 726f 6d5f 7363 6970 795f  port from_scipy_
-00016c00: 7370 6172 7365 5f6d 6174 7269 780a 2020  sparse_matrix.  
-00016c10: 2020 2020 2020 2020 2020 6564 6765 5f69            edge_i
-00016c20: 6e64 6578 2c65 6467 655f 7765 6967 6874  ndex,edge_weight
-00016c30: 203d 2066 726f 6d5f 7363 6970 795f 7370   = from_scipy_sp
-00016c40: 6172 7365 5f6d 6174 7269 7828 7765 6967  arse_matrix(weig
-00016c50: 6874 735f 636f 6f29 0a20 2020 2020 2020  hts_coo).       
-00016c60: 2020 2020 2023 2064 676c 5f67 7261 7068       # dgl_graph
-00016c70: 203d 2064 676c 2e44 474c 4772 6170 6828   = dgl.DGLGraph(
-00016c80: 2920 2367 7261 7068 2066 6f72 2054 7265  ) #graph for Tre
-00016c90: 654c 5354 4d0a 2020 2020 2020 2020 2020  eLSTM.          
-00016ca0: 2020 2320 6467 6c5f 6772 6170 6820 3d20    # dgl_graph = 
-00016cb0: 6467 6c2e 4447 4c48 6574 6572 6f47 7261  dgl.DGLHeteroGra
-00016cc0: 7068 2829 0a20 2020 2020 2020 2020 2020  ph().           
-00016cd0: 2023 2064 676c 5f67 7261 7068 2e61 6464   # dgl_graph.add
-00016ce0: 5f6e 6f64 6573 2870 6174 7269 7374 6963  _nodes(patristic
-00016cf0: 5f6d 6174 7269 785b 313a 2c20 313a 5d2e  _matrix[1:, 1:].
-00016d00: 7368 6170 655b 305d 290a 2020 2020 2020  shape[0]).      
-00016d10: 2020 2020 2020 2320 6467 6c5f 6772 6170        # dgl_grap
-00016d20: 682e 6164 645f 6564 6765 7328 6564 6765  h.add_edges(edge
-00016d30: 5f69 6e64 6578 5b30 5d2e 6375 6461 2829  _index[0].cuda()
-00016d40: 2c20 6564 6765 5f69 6e64 6578 5b31 5d2e  , edge_index[1].
-00016d50: 6375 6461 2829 290a 2020 2020 2020 2020  cuda()).        
-00016d60: 2020 2020 2320 6467 6c5f 6772 6170 682e      # dgl_graph.
-00016d70: 6564 6174 615b 2779 275d 203d 2065 6467  edata['y'] = edg
-00016d80: 655f 7765 6967 6874 2e63 7564 6128 290a  e_weight.cuda().
-00016d90: 2020 2020 2020 2020 2020 2020 6966 2061              if a
-00016da0: 7267 732e 7573 655f 6375 6461 3a0a 2020  rgs.use_cuda:.  
-00016db0: 2020 2020 2020 2020 2020 2020 2020 6772                gr
-00016dc0: 6170 685f 636f 6f20 3d20 2865 6467 655f  aph_coo = (edge_
-00016dd0: 696e 6465 782e 6375 6461 2829 2c20 6564  index.cuda(), ed
-00016de0: 6765 5f77 6569 6768 742e 6375 6461 2829  ge_weight.cuda()
-00016df0: 2920 2023 2067 7261 7068 2066 6f72 2070  )  # graph for p
-00016e00: 7974 6f72 6368 2067 656f 6d65 7472 6963  ytorch geometric
-00016e10: 2066 6f72 2047 4e4e 0a20 2020 2020 2020   for GNN.       
-00016e20: 2020 2020 2020 2020 2064 676c 5f67 7261           dgl_gra
-00016e30: 7068 203d 2064 676c 2e63 6f6e 7665 7274  ph = dgl.convert
-00016e40: 2e66 726f 6d5f 7363 6970 7928 7765 6967  .from_scipy(weig
-00016e50: 6874 735f 636f 6f29 2e74 6f28 2263 7564  hts_coo).to("cud
-00016e60: 6122 290a 2020 2020 2020 2020 2020 2020  a").            
-00016e70: 2020 2020 6467 6c5f 6772 6170 682e 6564      dgl_graph.ed
-00016e80: 6174 615b 2779 275d 203d 2065 6467 655f  ata['y'] = edge_
-00016e90: 7765 6967 6874 2e74 6f28 2263 7564 6122  weight.to("cuda"
-00016ea0: 290a 2020 2020 2020 2020 2020 2020 656c  ).            el
-00016eb0: 7365 3a0a 2020 2020 2020 2020 2020 2020  se:.            
-00016ec0: 2020 2020 6772 6170 685f 636f 6f20 3d20      graph_coo = 
-00016ed0: 2865 6467 655f 696e 6465 782e 6370 7528  (edge_index.cpu(
-00016ee0: 292c 2065 6467 655f 7765 6967 6874 2e63  ), edge_weight.c
-00016ef0: 7075 2829 2920 2023 2067 7261 7068 2066  pu())  # graph f
-00016f00: 6f72 2070 7974 6f72 6368 2067 656f 6d65  or pytorch geome
-00016f10: 7472 6963 2066 6f72 2047 4e4e 0a20 2020  tric for GNN.   
-00016f20: 2020 2020 2020 2020 2020 2020 2064 676c               dgl
-00016f30: 5f67 7261 7068 203d 2064 676c 2e63 6f6e  _graph = dgl.con
-00016f40: 7665 7274 2e66 726f 6d5f 7363 6970 7928  vert.from_scipy(
-00016f50: 7765 6967 6874 735f 636f 6f29 2e74 6f28  weights_coo).to(
-00016f60: 2263 7075 2229 0a20 2020 2020 2020 2020  "cpu").         
-00016f70: 2020 2020 2020 2064 676c 5f67 7261 7068         dgl_graph
-00016f80: 2e65 6461 7461 5b27 7927 5d20 3d20 6564  .edata['y'] = ed
-00016f90: 6765 5f77 6569 6768 742e 6370 7528 290a  ge_weight.cpu().
-00016fa0: 2020 2020 2020 2020 2020 2020 7072 696e              prin
-00016fb0: 7428 2274 7265 652d 6772 6170 6820 6669  t("tree-graph fi
-00016fc0: 6e69 7368 6564 2229 0a0a 2020 2020 2020  nished")..      
-00016fd0: 2020 6578 6365 7074 3a0a 2020 2020 2020    except:.      
-00016fe0: 2020 2020 2020 6772 6170 685f 636f 6f20        graph_coo 
-00016ff0: 3d20 4e6f 6e65 0a20 2020 2020 2020 2020  = None.         
-00017000: 2020 2064 676c 5f67 7261 7068 203d 204e     dgl_graph = N
-00017010: 6f6e 650a 0a20 2020 2065 6c73 653a 0a20  one..    else:. 
-00017020: 2020 2020 2020 2067 7261 7068 5f63 6f6f         graph_coo
-00017030: 3d4e 6f6e 650a 2020 2020 2020 2020 6564  =None.        ed
-00017040: 6765 5f77 6569 6768 745f 6d61 7472 6978  ge_weight_matrix
-00017050: 3d4e 6f6e 650a 2020 2020 2020 2020 6467  =None.        dg
-00017060: 6c5f 6772 6170 6820 3d20 4e6f 6e65 0a20  l_graph = None. 
-00017070: 2020 2062 6c6f 7375 6d5f 6172 7261 792c     blosum_array,
-00017080: 626c 6f73 756d 5f64 6963 7420 3d20 6372  blosum_dict = cr
-00017090: 6561 7465 5f62 6c6f 7375 6d28 6275 696c  eate_blosum(buil
-000170a0: 645f 636f 6e66 6967 2e61 615f 7072 6f62  d_config.aa_prob
-000170b0: 732c 6172 6773 2e73 7562 735f 6d61 7472  s,args.subs_matr
-000170c0: 6978 290a 0a20 2020 2023 6467 6c5f 6772  ix)..    #dgl_gr
-000170d0: 6170 682e 6e64 6174 615b 2778 275d 203d  aph.ndata['x'] =
-000170e0: 2074 6f72 6368 2e7a 6572 6f73 2828 332c   torch.zeros((3,
-000170f0: 2035 2929 2023 746f 2061 6464 206c 6174   5)) #to add lat
-00017100: 6572 2069 6e20 7468 6520 6d6f 6465 6c20  er in the model 
-00017110: 6974 2077 696c 6c20 6265 2074 6865 206c  it will be the l
-00017120: 6174 656e 7420 7370 6163 6520 7468 6174  atent space that
-00017130: 2067 6574 7320 7472 616e 7366 6f72 6d65   gets transforme
-00017140: 6420 746f 206c 6f67 6974 730a 2020 2020  d to logits.    
-00017150: 2347 2e6e 6f64 6573 5b5b 302c 2032 5d5d  #G.nodes[[0, 2]]
-00017160: 2e64 6174 615b 2778 275d 203d 2074 682e  .data['x'] = th.
-00017170: 6f6e 6573 2828 322c 2035 2929 0a0a 2020  ones((2, 5))..  
-00017180: 2020 6e6f 6465 735f 7265 7072 6573 656e    nodes_represen
-00017190: 7461 7469 6f6e 735f 6172 7261 7920 3d20  tations_array = 
-000171a0: 7472 6565 5f70 6f73 6974 696f 6e61 6c5f  tree_positional_
-000171b0: 656d 6265 6464 696e 6773 2861 6e63 6573  embeddings(ances
-000171c0: 746f 725f 696e 666f 5f64 6963 742c 7472  tor_info_dict,tr
-000171d0: 6565 5f62 795f 6c65 7665 6c73 5f64 6963  ee_by_levels_dic
-000171e0: 7429 0a20 2020 2061 6464 6974 696f 6e61  t).    additiona
-000171f0: 6c5f 696e 666f 203d 2041 6464 6974 696f  l_info = Additio
-00017200: 6e61 6c49 6e66 6f28 626c 6f73 756d 3d74  nalInfo(blosum=t
-00017210: 6f72 6368 2e66 726f 6d5f 6e75 6d70 7928  orch.from_numpy(
-00017220: 626c 6f73 756d 5f61 7272 6179 292c 0a20  blosum_array),. 
+00015bd0: 2020 2020 2270 6174 7269 7374 6963 5f69      "patristic_i
+00015be0: 6e66 6f22 2c20 2267 7261 7068 5f63 6f6f  nfo", "graph_coo
+00015bf0: 222c 2270 6174 7269 7374 6963 5f66 756c  ","patristic_ful
+00015c00: 6c5f 7370 6172 7365 222c 226e 6f64 6573  l_sparse","nodes
+00015c10: 5f72 6570 7265 7365 6e74 6174 696f 6e73  _representations
+00015c20: 5f61 7272 6179 222c 2264 676c 5f67 7261  _array","dgl_gra
+00015c30: 7068 225d 290a 0a20 2020 2061 6e63 6573  ph"])..    ances
+00015c40: 746f 725f 696e 666f 5f64 6963 7420 3d20  tor_info_dict = 
+00015c50: 6469 6374 287a 6970 2861 6e63 6573 746f  dict(zip(ancesto
+00015c60: 725f 696e 666f 5b3a 2c30 5d2e 746f 6c69  r_info[:,0].toli
+00015c70: 7374 2829 2c61 6e63 6573 746f 725f 696e  st(),ancestor_in
+00015c80: 666f 5b3a 2c32 5d2e 746f 6c69 7374 2829  fo[:,2].tolist()
+00015c90: 2929 0a20 2020 2070 6963 6b6c 652e 6475  )).    pickle.du
+00015ca0: 6d70 2861 6e63 6573 746f 725f 696e 666f  mp(ancestor_info
+00015cb0: 5f64 6963 742c 206f 7065 6e28 277b 7d2f  _dict, open('{}/
+00015cc0: 416e 6365 7374 6f72 5f69 6e66 6f5f 6469  Ancestor_info_di
+00015cd0: 6374 2e70 272e 666f 726d 6174 2872 6573  ct.p'.format(res
+00015ce0: 756c 7473 5f64 6972 292c 2027 7762 2729  ults_dir), 'wb')
+00015cf0: 2c20 7072 6f74 6f63 6f6c 3d70 6963 6b6c  , protocol=pickl
+00015d00: 652e 4849 4748 4553 545f 5052 4f54 4f43  e.HIGHEST_PROTOC
+00015d10: 4f4c 290a 2020 2020 6966 2069 7369 6e73  OL).    if isins
+00015d20: 7461 6e63 6528 7061 7472 6973 7469 635f  tance(patristic_
+00015d30: 6d61 7472 6978 2c6e 702e 6e64 6172 7261  matrix,np.ndarra
+00015d40: 7929 3a0a 2020 2020 2020 2020 7061 7472  y):.        patr
+00015d50: 6973 7469 635f 696e 666f 203d 2064 6963  istic_info = dic
+00015d60: 7428 7a69 7028 6c69 7374 2872 616e 6765  t(zip(list(range
+00015d70: 286c 656e 2870 6174 7269 7374 6963 5f6d  (len(patristic_m
+00015d80: 6174 7269 7829 2d31 2929 2c70 6174 7269  atrix)-1)),patri
+00015d90: 7374 6963 5f6d 6174 7269 785b 313a 2c30  stic_matrix[1:,0
+00015da0: 5d2e 6173 7479 7065 2869 6e74 292e 746f  ].astype(int).to
+00015db0: 6c69 7374 2829 2929 2023 536b 6970 2074  list())) #Skip t
+00015dc0: 6865 2066 616b 6520 6865 6164 6572 0a20  he fake header. 
+00015dd0: 2020 2065 6c73 653a 0a0a 2020 2020 2020     else:..      
+00015de0: 2020 7061 7472 6973 7469 635f 696e 666f    patristic_info
+00015df0: 203d 6469 6374 287a 6970 286c 6973 7428   =dict(zip(list(
+00015e00: 7261 6e67 6528 6c65 6e28 7061 7472 6973  range(len(patris
+00015e10: 7469 635f 6d61 7472 6978 292d 3129 292c  tic_matrix)-1)),
+00015e20: 7061 7472 6973 7469 635f 6d61 7472 6978  patristic_matrix
+00015e30: 5b31 3a2c 305d 2e74 7970 6528 746f 7263  [1:,0].type(torc
+00015e40: 682e 696e 7429 2e74 6f6c 6973 7428 2929  h.int).tolist())
+00015e50: 290a 0a20 2020 2061 6e63 6573 746f 7273  )..    ancestors
+00015e60: 5f69 6e66 6f5f 666c 6970 7065 6420 3d20  _info_flipped = 
+00015e70: 6e70 2e66 6c69 7028 616e 6365 7374 6f72  np.flip(ancestor
+00015e80: 5f69 6e66 6f2c 2061 7869 733d 3129 2023  _info, axis=1) #
+00015e90: 536f 2077 6520 6361 6e20 6576 616c 7561  So we can evalua
+00015ea0: 7465 2069 6e20 7265 7665 7273 652c 2077  te in reverse, w
+00015eb0: 6520 7374 6172 7420 7769 7468 2074 6865  e start with the
+00015ec0: 206c 6173 7420 636f 6c75 6d6e 0a0a 2020   last column..  
+00015ed0: 2020 7472 6565 5f62 795f 6c65 7665 6c73    tree_by_levels
+00015ee0: 203d 205b 5d0a 2020 2020 666f 7220 636f   = [].    for co
+00015ef0: 6c75 6d6e 2069 6e20 616e 6365 7374 6f72  lumn in ancestor
+00015f00: 735f 696e 666f 5f66 6c69 7070 6564 2e54  s_info_flipped.T
+00015f10: 3a0a 2020 2020 2020 2020 696e 6465 7865  :.        indexe
+00015f20: 7320 3d20 6e70 2e77 6865 7265 2863 6f6c  s = np.where(col
+00015f30: 756d 6e20 3d3d 2030 290a 2020 2020 2020  umn == 0).      
+00015f40: 2020 6c65 7665 6c20 3d20 616e 6365 7374    level = ancest
+00015f50: 6f72 735f 696e 666f 5f66 6c69 7070 6564  ors_info_flipped
+00015f60: 2e54 5b2d 315d 5b69 6e64 6578 6573 5d0a  .T[-1][indexes].
+00015f70: 2020 2020 2020 2020 7472 6565 5f62 795f          tree_by_
+00015f80: 6c65 7665 6c73 2e61 7070 656e 6428 6c65  levels.append(le
+00015f90: 7665 6c29 0a0a 2020 2020 7472 6565 5f62  vel)..    tree_b
+00015fa0: 795f 6c65 7665 6c73 203d 2074 7265 655f  y_levels = tree_
+00015fb0: 6279 5f6c 6576 656c 735b 3a2d 315d 2023  by_levels[:-1] #
+00015fc0: 7468 6520 726f 6f74 2067 6574 7320 6164  the root gets ad
+00015fd0: 6465 6420 7477 6963 6520 6265 6361 7573  ded twice becaus
+00015fe0: 6520 7468 6520 6669 7273 7420 636f 6c75  e the first colu
+00015ff0: 6d6e 2061 7265 2074 6865 206e 6f64 6520  mn are the node 
+00016000: 696e 6465 7865 732c 2063 616e 6e6f 7420  indexes, cannot 
+00016010: 6265 2066 6978 6564 2c20 6f74 6865 7277  be fixed, otherw
+00016020: 6973 6520 7765 2063 616e 6e6f 7420 646f  ise we cannot do
+00016030: 2074 6865 2074 7269 636b 0a0a 2020 2020   the trick..    
+00016040: 7472 6565 5f62 795f 6c65 7665 6c73 5f64  tree_by_levels_d
+00016050: 6963 7420 3d20 6469 6374 287a 6970 286c  ict = dict(zip(l
+00016060: 6973 7428 7265 7665 7273 6564 2872 616e  ist(reversed(ran
+00016070: 6765 286c 656e 2874 7265 655f 6279 5f6c  ge(len(tree_by_l
+00016080: 6576 656c 7329 2929 292c 7472 6565 5f62  evels)))),tree_b
+00016090: 795f 6c65 7665 6c73 2929 0a20 2020 2070  y_levels)).    p
+000160a0: 6963 6b6c 652e 6475 6d70 2874 7265 655f  ickle.dump(tree_
+000160b0: 6279 5f6c 6576 656c 735f 6469 6374 2c20  by_levels_dict, 
+000160c0: 6f70 656e 2827 7b7d 2f54 7265 655f 6279  open('{}/Tree_by
+000160d0: 5f6c 6576 656c 735f 6469 6374 2e70 272e  _levels_dict.p'.
+000160e0: 666f 726d 6174 2872 6573 756c 7473 5f64  format(results_d
+000160f0: 6972 292c 2027 7762 2729 2c20 7072 6f74  ir), 'wb'), prot
+00016100: 6f63 6f6c 3d70 6963 6b6c 652e 4849 4748  ocol=pickle.HIGH
+00016110: 4553 545f 5052 4f54 4f43 4f4c 290a 0a20  EST_PROTOCOL).. 
+00016120: 2020 206c 656e 6774 6820 3d20 6d61 7828     length = max(
+00016130: 6d61 7028 6c65 6e2c 2074 7265 655f 6279  map(len, tree_by
+00016140: 5f6c 6576 656c 735f 6469 6374 2e76 616c  _levels_dict.val
+00016150: 7565 7328 2929 290a 2020 2020 7472 6565  ues())).    tree
+00016160: 5f62 795f 6c65 7665 6c73 5f61 7272 6179  _by_levels_array
+00016170: 203d 206e 702e 6172 7261 7928 5b78 692e   = np.array([xi.
+00016180: 746f 6c69 7374 2829 202b 205b 4e6f 6e65  tolist() + [None
+00016190: 5d20 2a20 286c 656e 6774 6820 2d20 6c65  ] * (length - le
+000161a0: 6e28 7869 2929 2066 6f72 2078 6920 696e  n(xi)) for xi in
+000161b0: 2074 7265 655f 6279 5f6c 6576 656c 735d   tree_by_levels]
+000161c0: 290a 2020 2020 2320 7472 6565 5f6c 6576  ).    # tree_lev
+000161d0: 656c 735f 3220 3d20 4372 6561 7465 5f74  els_2 = Create_t
+000161e0: 7265 655f 6279 5f6c 6576 656c 7328 6368  ree_by_levels(ch
+000161f0: 696c 6472 656e 5f64 6963 7429 0a0a 2020  ildren_dict)..  
+00016200: 2020 6368 696c 6472 656e 5f64 6963 7420    children_dict 
+00016210: 3d20 7b7d 0a20 2020 2066 6f72 206b 2c20  = {}.    for k, 
+00016220: 7620 696e 2061 6e63 6573 746f 725f 696e  v in ancestor_in
+00016230: 666f 5f64 6963 742e 6974 656d 7328 293a  fo_dict.items():
+00016240: 0a20 2020 2020 2020 2063 6869 6c64 7265  .        childre
+00016250: 6e5f 6469 6374 5b76 5d20 3d20 6368 696c  n_dict[v] = chil
+00016260: 6472 656e 5f64 6963 742e 6765 7428 762c  dren_dict.get(v,
+00016270: 205b 5d29 0a20 2020 2020 2020 2063 6869   []).        chi
+00016280: 6c64 7265 6e5f 6469 6374 5b76 5d2e 6170  ldren_dict[v].ap
+00016290: 7065 6e64 286b 290a 2020 2020 7069 636b  pend(k).    pick
+000162a0: 6c65 2e64 756d 7028 6368 696c 6472 656e  le.dump(children
+000162b0: 5f64 6963 742c 206f 7065 6e28 277b 7d2f  _dict, open('{}/
+000162c0: 4368 696c 6472 656e 5f64 6963 742e 7027  Children_dict.p'
+000162d0: 2e66 6f72 6d61 7428 7265 7375 6c74 735f  .format(results_
+000162e0: 6469 7229 2c20 2777 6227 292c 0a20 2020  dir), 'wb'),.   
+000162f0: 2020 2020 2020 2020 2020 2020 2070 726f               pro
+00016300: 746f 636f 6c3d 7069 636b 6c65 2e48 4947  tocol=pickle.HIG
+00016310: 4845 5354 5f50 524f 544f 434f 4c29 0a0a  HEST_PROTOCOL)..
+00016320: 2020 2020 2322 6874 7470 733a 2f2f 7079      #"https://py
+00016330: 746f 7263 682d 6765 6f6d 6574 7269 632e  torch-geometric.
+00016340: 7265 6164 7468 6564 6f63 732e 696f 2f65  readthedocs.io/e
+00016350: 6e2f 6c61 7465 7374 2f6e 6f74 6573 2f69  n/latest/notes/i
+00016360: 6e74 726f 6475 6374 696f 6e2e 6874 6d6c  ntroduction.html
+00016370: 220a 2020 2020 2348 6967 686c 6967 6874  ".    #Highlight
+00016380: 3a42 7569 6c64 2074 6865 206d 6174 7269  :Build the matri
+00016390: 7820 7265 7072 6573 656e 7461 7469 6f6e  x representation
+000163a0: 206f 6620 7468 6520 7472 6565 2067 7261   of the tree gra
+000163b0: 7068 3a20 4469 7265 6374 6564 2028 726f  ph: Directed (ro
+000163c0: 6f74 2d3e 206c 6561 7665 7329 2077 6569  ot-> leaves) wei
+000163d0: 6768 7465 6420 2870 6174 7269 7374 6963  ghted (patristic
+000163e0: 2064 6973 7429 2067 7261 7068 0a20 2020   dist) graph.   
+000163f0: 2069 6620 6275 696c 645f 636f 6e66 6967   if build_config
+00016400: 2e62 7569 6c64 5f67 7261 7068 3a20 2354  .build_graph: #T
+00016410: 4f44 4f3a 2046 6978 2069 6e20 7365 7276  ODO: Fix in serv
+00016420: 6572 0a20 2020 2020 2020 2070 7269 6e74  er.        print
+00016430: 2822 4275 696c 6469 6e67 2061 2074 7265  ("Building a tre
+00016440: 652d 6772 6170 6822 290a 2020 2020 2020  e-graph").      
+00016450: 2020 236d 616b 6520 6772 6170 6820 6164    #make graph ad
+00016460: 6a61 6365 6e74 206d 6174 7269 780a 2020  jacent matrix.  
+00016470: 2020 2020 2020 7061 7472 6973 7469 635f        patristic_
+00016480: 6d61 7472 6978 203d 2070 6174 7269 7374  matrix = patrist
+00016490: 6963 5f6d 6174 7269 785b 7061 7472 6973  ic_matrix[patris
+000164a0: 7469 635f 6d61 7472 6978 5b3a 2c20 305d  tic_matrix[:, 0]
+000164b0: 2e61 7267 736f 7274 2829 5d0a 2020 2020  .argsort()].    
+000164c0: 2020 2020 7061 7472 6973 7469 635f 6d61      patristic_ma
+000164d0: 7472 6978 203d 2070 6174 7269 7374 6963  trix = patristic
+000164e0: 5f6d 6174 7269 785b 3a2c 2070 6174 7269  _matrix[:, patri
+000164f0: 7374 6963 5f6d 6174 7269 785b 302c 203a  stic_matrix[0, :
+00016500: 5d2e 6172 6773 6f72 7428 295d 0a0a 2020  ].argsort()]..  
+00016510: 2020 2020 2020 6772 6170 685f 6e6f 6465        graph_node
+00016520: 5f6d 6174 7269 7820 3d20 6e70 2e7a 6572  _matrix = np.zer
+00016530: 6f73 5f6c 696b 6528 7061 7472 6973 7469  os_like(patristi
+00016540: 635f 6d61 7472 6978 2920 2344 6972 6563  c_matrix) #Direc
+00016550: 7465 6420 2872 6f6f 742d 3e20 6c65 6176  ted (root-> leav
+00016560: 6573 2920 4e4f 5420 7765 6967 6874 6564  es) NOT weighted
+00016570: 2067 7261 7068 0a20 2020 2020 2020 2065   graph.        e
+00016580: 6467 655f 7765 6967 6874 5f6d 6174 7269  dge_weight_matri
+00016590: 7820 3d20 6e70 2e7a 6572 6f73 5f6c 696b  x = np.zeros_lik
+000165a0: 6528 7061 7472 6973 7469 635f 6d61 7472  e(patristic_matr
+000165b0: 6978 2920 2344 6972 6563 7465 6420 2872  ix) #Directed (r
+000165c0: 6f6f 742d 3e20 6c65 6176 6573 2920 7765  oot-> leaves) we
+000165d0: 6967 6874 6564 2028 7061 7472 6973 7469  ighted (patristi
+000165e0: 6320 6469 7374 2920 6772 6170 680a 2020  c dist) graph.  
+000165f0: 2020 2020 2020 6772 6170 685f 6e6f 6465        graph_node
+00016600: 5f6d 6174 7269 785b 302c 3a5d 203d 2070  _matrix[0,:] = p
+00016610: 6174 7269 7374 6963 5f6d 6174 7269 785b  atristic_matrix[
+00016620: 302c 3a5d 0a20 2020 2020 2020 2067 7261  0,:].        gra
+00016630: 7068 5f6e 6f64 655f 6d61 7472 6978 5b3a  ph_node_matrix[:
+00016640: 2c30 5d20 3d20 7061 7472 6973 7469 635f  ,0] = patristic_
+00016650: 6d61 7472 6978 5b3a 2c30 5d0a 2020 2020  matrix[:,0].    
+00016660: 2020 2020 6564 6765 5f77 6569 6768 745f      edge_weight_
+00016670: 6d61 7472 6978 5b30 2c20 3a5d 203d 2070  matrix[0, :] = p
+00016680: 6174 7269 7374 6963 5f6d 6174 7269 785b  atristic_matrix[
+00016690: 302c 203a 5d0a 2020 2020 2020 2020 6564  0, :].        ed
+000166a0: 6765 5f77 6569 6768 745f 6d61 7472 6978  ge_weight_matrix
+000166b0: 5b3a 2c20 305d 203d 2070 6174 7269 7374  [:, 0] = patrist
+000166c0: 6963 5f6d 6174 7269 785b 3a2c 2030 5d0a  ic_matrix[:, 0].
+000166d0: 2020 2020 2020 2020 666f 7220 616e 6365          for ance
+000166e0: 7374 6f72 2c63 6869 6c64 7265 6e20 696e  stor,children in
+000166f0: 2063 6869 6c64 7265 6e5f 6469 6374 2e69   children_dict.i
+00016700: 7465 6d73 2829 3a0a 2020 2020 2020 2020  tems():.        
+00016710: 2020 2020 616e 6365 7374 6f72 5f69 6478      ancestor_idx
+00016720: 203d 206e 702e 696e 3164 2867 7261 7068   = np.in1d(graph
+00016730: 5f6e 6f64 655f 6d61 7472 6978 5b30 2c3a  _node_matrix[0,:
+00016740: 5d2c 2061 6e63 6573 746f 7229 2e6e 6f6e  ], ancestor).non
+00016750: 7a65 726f 2829 5b30 5d0a 2020 2020 2020  zero()[0].      
+00016760: 2020 2020 2020 6368 696c 6472 656e 5f69        children_i
+00016770: 6478 3d20 6e70 2e69 6e31 6428 6772 6170  dx= np.in1d(grap
+00016780: 685f 6e6f 6465 5f6d 6174 7269 785b 302c  h_node_matrix[0,
+00016790: 3a5d 2c20 6368 696c 6472 656e 292e 6e6f  :], children).no
+000167a0: 6e7a 6572 6f28 295b 305d 0a20 2020 2020  nzero()[0].     
+000167b0: 2020 2020 2020 2067 7261 7068 5f6e 6f64         graph_nod
+000167c0: 655f 6d61 7472 6978 5b61 6e63 6573 746f  e_matrix[ancesto
+000167d0: 725f 6964 782c 6368 696c 6472 656e 5f69  r_idx,children_i
+000167e0: 6478 5d20 3d20 3120 2363 6f6e 7461 696e  dx] = 1 #contain
+000167f0: 7320 6e6f 2065 6467 6573 2077 6569 6768  s no edges weigh
+00016800: 7473 0a20 2020 2020 2020 2020 2020 2067  ts.            g
+00016810: 7261 7068 5f6e 6f64 655f 6d61 7472 6978  raph_node_matrix
+00016820: 5b63 6869 6c64 7265 6e5f 6964 782c 616e  [children_idx,an
+00016830: 6365 7374 6f72 5f69 6478 5d20 3d20 310a  cestor_idx] = 1.
+00016840: 2020 2020 2020 2020 2020 2020 6564 6765              edge
+00016850: 5f77 6569 6768 745f 6d61 7472 6978 5b61  _weight_matrix[a
+00016860: 6e63 6573 746f 725f 6964 782c 2063 6869  ncestor_idx, chi
+00016870: 6c64 7265 6e5f 6964 785d 203d 2070 6174  ldren_idx] = pat
+00016880: 7269 7374 6963 5f6d 6174 7269 785b 616e  ristic_matrix[an
+00016890: 6365 7374 6f72 5f69 6478 2c20 6368 696c  cestor_idx, chil
+000168a0: 6472 656e 5f69 6478 5d0a 2020 2020 2020  dren_idx].      
+000168b0: 2020 2020 2020 2365 6467 655f 7765 6967        #edge_weig
+000168c0: 6874 5f6d 6174 7269 785b 6368 696c 6472  ht_matrix[childr
+000168d0: 656e 5f69 6478 2c20 616e 6365 7374 6f72  en_idx, ancestor
+000168e0: 5f69 6478 5d20 3d20 7061 7472 6973 7469  _idx] = patristi
+000168f0: 635f 6d61 7472 6978 5b63 6869 6c64 7265  c_matrix[childre
+00016900: 6e5f 6964 782c 2061 6e63 6573 746f 725f  n_idx, ancestor_
+00016910: 6964 785d 2023 4869 6768 6c69 6768 743a  idx] #Highlight:
+00016920: 204e 6f74 2075 7365 2074 6f20 6176 6f69   Not use to avoi
+00016930: 6420 6372 6561 7469 6e67 2061 2073 656c  d creating a sel
+00016940: 6620 6c6f 6f70 6564 2067 7261 7068 206f  f looped graph o
+00016950: 7220 6269 6469 7265 6374 696f 6e61 6c20  r bidirectional 
+00016960: 6772 6170 6873 0a20 2020 2020 2020 2023  graphs.        #
+00016970: 4372 6561 7465 2061 2063 6f6f 7264 696e  Create a coordin
+00016980: 6174 6573 206d 6174 7269 782c 2062 6173  ates matrix, bas
+00016990: 6963 616c 6c79 2070 6f69 6e74 7320 6f75  ically points ou
+000169a0: 7420 7768 6963 6820 7265 6769 6f6e 7320  t which regions 
+000169b0: 6672 6f6d 2074 6865 206d 6174 7269 7820  from the matrix 
+000169c0: 646f 206e 6f74 2068 6176 6520 3020 7661  do not have 0 va
+000169d0: 6c75 6573 2061 6e64 2074 6865 7265 666f  lues and therefo
+000169e0: 7265 2061 7265 2063 6f6e 6e65 6374 6564  re are connected
+000169f0: 0a20 2020 2020 2020 2077 6569 6768 7473  .        weights
+00016a00: 5f63 6f6f 203d 2063 6f6f 5f6d 6174 7269  _coo = coo_matri
+00016a10: 7828 6564 6765 5f77 6569 6768 745f 6d61  x(edge_weight_ma
+00016a20: 7472 6978 5b31 3a2c 313a 5d29 0a20 2020  trix[1:,1:]).   
+00016a30: 2020 2020 2074 7279 3a0a 2020 2020 2020       try:.      
+00016a40: 2020 2020 2020 2354 6865 2067 7261 7068        #The graph
+00016a50: 2063 6f6e 6e65 6374 6976 6974 7920 2865   connectivity (e
+00016a60: 6467 6520 696e 6465 7829 2073 686f 756c  dge index) shoul
+00016a70: 6420 6265 2063 6f6e 6669 6e65 6420 7769  d be confined wi
+00016a80: 7468 2074 6865 2043 4f4f 2066 6f72 6d61  th the COO forma
+00016a90: 742c 0a20 2020 2020 2020 2020 2020 2023  t,.            #
+00016aa0: 2069 2e65 2e20 7468 6520 6669 7273 7420   i.e. the first 
+00016ab0: 6c69 7374 2063 6f6e 7461 696e 7320 7468  list contains th
+00016ac0: 6520 696e 6465 7820 6f66 2074 6865 2073  e index of the s
+00016ad0: 6f75 7263 6520 6e6f 6465 732c 2077 6869  ource nodes, whi
+00016ae0: 6c65 2074 6865 2069 6e64 6578 206f 6620  le the index of 
+00016af0: 7461 7267 6574 206e 6f64 6573 2069 7320  target nodes is 
+00016b00: 7370 6563 6966 6965 6420 696e 2074 6865  specified in the
+00016b10: 2073 6563 6f6e 6420 6c69 7374 2e0a 2020   second list..  
+00016b20: 2020 2020 2020 2020 2020 234e 6f74 6520            #Note 
+00016b30: 7468 6174 2074 6865 206f 7264 6572 206f  that the order o
+00016b40: 6620 7468 6520 6564 6765 2069 6e64 6578  f the edge index
+00016b50: 2069 7320 6972 7265 6c65 7661 6e74 2074   is irrelevant t
+00016b60: 6f20 7468 6520 4461 7461 206f 626a 6563  o the Data objec
+00016b70: 7420 796f 7520 6372 6561 7465 2073 696e  t you create sin
+00016b80: 6365 2073 7563 6820 696e 666f 726d 6174  ce such informat
+00016b90: 696f 6e20 6973 206f 6e6c 7920 666f 7220  ion is only for 
+00016ba0: 636f 6d70 7574 696e 6720 7468 6520 6164  computing the ad
+00016bb0: 6a61 6365 6e63 7920 6d61 7472 6978 2e0a  jacency matrix..
+00016bc0: 2020 2020 2020 2020 2020 2020 6672 6f6d              from
+00016bd0: 2074 6f72 6368 5f67 656f 6d65 7472 6963   torch_geometric
+00016be0: 2e75 7469 6c73 2e63 6f6e 7665 7274 2069  .utils.convert i
+00016bf0: 6d70 6f72 7420 6672 6f6d 5f73 6369 7079  mport from_scipy
+00016c00: 5f73 7061 7273 655f 6d61 7472 6978 0a20  _sparse_matrix. 
+00016c10: 2020 2020 2020 2020 2020 2065 6467 655f             edge_
+00016c20: 696e 6465 782c 6564 6765 5f77 6569 6768  index,edge_weigh
+00016c30: 7420 3d20 6672 6f6d 5f73 6369 7079 5f73  t = from_scipy_s
+00016c40: 7061 7273 655f 6d61 7472 6978 2877 6569  parse_matrix(wei
+00016c50: 6768 7473 5f63 6f6f 290a 2020 2020 2020  ghts_coo).      
+00016c60: 2020 2020 2020 2320 6467 6c5f 6772 6170        # dgl_grap
+00016c70: 6820 3d20 6467 6c2e 4447 4c47 7261 7068  h = dgl.DGLGraph
+00016c80: 2829 2023 6772 6170 6820 666f 7220 5472  () #graph for Tr
+00016c90: 6565 4c53 544d 0a20 2020 2020 2020 2020  eeLSTM.         
+00016ca0: 2020 2023 2064 676c 5f67 7261 7068 203d     # dgl_graph =
+00016cb0: 2064 676c 2e44 474c 4865 7465 726f 4772   dgl.DGLHeteroGr
+00016cc0: 6170 6828 290a 2020 2020 2020 2020 2020  aph().          
+00016cd0: 2020 2320 6467 6c5f 6772 6170 682e 6164    # dgl_graph.ad
+00016ce0: 645f 6e6f 6465 7328 7061 7472 6973 7469  d_nodes(patristi
+00016cf0: 635f 6d61 7472 6978 5b31 3a2c 2031 3a5d  c_matrix[1:, 1:]
+00016d00: 2e73 6861 7065 5b30 5d29 0a20 2020 2020  .shape[0]).     
+00016d10: 2020 2020 2020 2023 2064 676c 5f67 7261         # dgl_gra
+00016d20: 7068 2e61 6464 5f65 6467 6573 2865 6467  ph.add_edges(edg
+00016d30: 655f 696e 6465 785b 305d 2e63 7564 6128  e_index[0].cuda(
+00016d40: 292c 2065 6467 655f 696e 6465 785b 315d  ), edge_index[1]
+00016d50: 2e63 7564 6128 2929 0a20 2020 2020 2020  .cuda()).       
+00016d60: 2020 2020 2023 2064 676c 5f67 7261 7068       # dgl_graph
+00016d70: 2e65 6461 7461 5b27 7927 5d20 3d20 6564  .edata['y'] = ed
+00016d80: 6765 5f77 6569 6768 742e 6375 6461 2829  ge_weight.cuda()
+00016d90: 0a20 2020 2020 2020 2020 2020 2069 6620  .            if 
+00016da0: 6172 6773 2e75 7365 5f63 7564 613a 0a20  args.use_cuda:. 
+00016db0: 2020 2020 2020 2020 2020 2020 2020 2067                 g
+00016dc0: 7261 7068 5f63 6f6f 203d 2028 6564 6765  raph_coo = (edge
+00016dd0: 5f69 6e64 6578 2e63 7564 6128 292c 2065  _index.cuda(), e
+00016de0: 6467 655f 7765 6967 6874 2e63 7564 6128  dge_weight.cuda(
+00016df0: 2929 2020 2320 6772 6170 6820 666f 7220  ))  # graph for 
+00016e00: 7079 746f 7263 6820 6765 6f6d 6574 7269  pytorch geometri
+00016e10: 6320 666f 7220 474e 4e0a 2020 2020 2020  c for GNN.      
+00016e20: 2020 2020 2020 2020 2020 6467 6c5f 6772            dgl_gr
+00016e30: 6170 6820 3d20 6467 6c2e 636f 6e76 6572  aph = dgl.conver
+00016e40: 742e 6672 6f6d 5f73 6369 7079 2877 6569  t.from_scipy(wei
+00016e50: 6768 7473 5f63 6f6f 292e 746f 2822 6375  ghts_coo).to("cu
+00016e60: 6461 2229 0a20 2020 2020 2020 2020 2020  da").           
+00016e70: 2020 2020 2064 676c 5f67 7261 7068 2e65       dgl_graph.e
+00016e80: 6461 7461 5b27 7927 5d20 3d20 6564 6765  data['y'] = edge
+00016e90: 5f77 6569 6768 742e 746f 2822 6375 6461  _weight.to("cuda
+00016ea0: 2229 0a20 2020 2020 2020 2020 2020 2065  ").            e
+00016eb0: 6c73 653a 0a20 2020 2020 2020 2020 2020  lse:.           
+00016ec0: 2020 2020 2067 7261 7068 5f63 6f6f 203d       graph_coo =
+00016ed0: 2028 6564 6765 5f69 6e64 6578 2e63 7075   (edge_index.cpu
+00016ee0: 2829 2c20 6564 6765 5f77 6569 6768 742e  (), edge_weight.
+00016ef0: 6370 7528 2929 2020 2320 6772 6170 6820  cpu())  # graph 
+00016f00: 666f 7220 7079 746f 7263 6820 6765 6f6d  for pytorch geom
+00016f10: 6574 7269 6320 666f 7220 474e 4e0a 2020  etric for GNN.  
+00016f20: 2020 2020 2020 2020 2020 2020 2020 6467                dg
+00016f30: 6c5f 6772 6170 6820 3d20 6467 6c2e 636f  l_graph = dgl.co
+00016f40: 6e76 6572 742e 6672 6f6d 5f73 6369 7079  nvert.from_scipy
+00016f50: 2877 6569 6768 7473 5f63 6f6f 292e 746f  (weights_coo).to
+00016f60: 2822 6370 7522 290a 2020 2020 2020 2020  ("cpu").        
+00016f70: 2020 2020 2020 2020 6467 6c5f 6772 6170          dgl_grap
+00016f80: 682e 6564 6174 615b 2779 275d 203d 2065  h.edata['y'] = e
+00016f90: 6467 655f 7765 6967 6874 2e63 7075 2829  dge_weight.cpu()
+00016fa0: 0a20 2020 2020 2020 2020 2020 2070 7269  .            pri
+00016fb0: 6e74 2822 7472 6565 2d67 7261 7068 2066  nt("tree-graph f
+00016fc0: 696e 6973 6865 6422 290a 0a20 2020 2020  inished")..     
+00016fd0: 2020 2065 7863 6570 743a 0a20 2020 2020     except:.     
+00016fe0: 2020 2020 2020 2067 7261 7068 5f63 6f6f         graph_coo
+00016ff0: 203d 204e 6f6e 650a 2020 2020 2020 2020   = None.        
+00017000: 2020 2020 6467 6c5f 6772 6170 6820 3d20      dgl_graph = 
+00017010: 4e6f 6e65 0a0a 2020 2020 656c 7365 3a0a  None..    else:.
+00017020: 2020 2020 2020 2020 6772 6170 685f 636f          graph_co
+00017030: 6f3d 4e6f 6e65 0a20 2020 2020 2020 2065  o=None.        e
+00017040: 6467 655f 7765 6967 6874 5f6d 6174 7269  dge_weight_matri
+00017050: 783d 4e6f 6e65 0a20 2020 2020 2020 2064  x=None.        d
+00017060: 676c 5f67 7261 7068 203d 204e 6f6e 650a  gl_graph = None.
+00017070: 2020 2020 626c 6f73 756d 5f61 7272 6179      blosum_array
+00017080: 2c62 6c6f 7375 6d5f 6469 6374 203d 2063  ,blosum_dict = c
+00017090: 7265 6174 655f 626c 6f73 756d 2862 7569  reate_blosum(bui
+000170a0: 6c64 5f63 6f6e 6669 672e 6161 5f70 726f  ld_config.aa_pro
+000170b0: 6273 2c61 7267 732e 7375 6273 5f6d 6174  bs,args.subs_mat
+000170c0: 7269 7829 0a0a 2020 2020 2364 676c 5f67  rix)..    #dgl_g
+000170d0: 7261 7068 2e6e 6461 7461 5b27 7827 5d20  raph.ndata['x'] 
+000170e0: 3d20 746f 7263 682e 7a65 726f 7328 2833  = torch.zeros((3
+000170f0: 2c20 3529 2920 2374 6f20 6164 6420 6c61  , 5)) #to add la
+00017100: 7465 7220 696e 2074 6865 206d 6f64 656c  ter in the model
+00017110: 2069 7420 7769 6c6c 2062 6520 7468 6520   it will be the 
+00017120: 6c61 7465 6e74 2073 7061 6365 2074 6861  latent space tha
+00017130: 7420 6765 7473 2074 7261 6e73 666f 726d  t gets transform
+00017140: 6564 2074 6f20 6c6f 6769 7473 0a20 2020  ed to logits.   
+00017150: 2023 472e 6e6f 6465 735b 5b30 2c20 325d   #G.nodes[[0, 2]
+00017160: 5d2e 6461 7461 5b27 7827 5d20 3d20 7468  ].data['x'] = th
+00017170: 2e6f 6e65 7328 2832 2c20 3529 290a 0a20  .ones((2, 5)).. 
+00017180: 2020 206e 6f64 6573 5f72 6570 7265 7365     nodes_represe
+00017190: 6e74 6174 696f 6e73 5f61 7272 6179 203d  ntations_array =
+000171a0: 2074 7265 655f 706f 7369 7469 6f6e 616c   tree_positional
+000171b0: 5f65 6d62 6564 6469 6e67 7328 616e 6365  _embeddings(ance
+000171c0: 7374 6f72 5f69 6e66 6f5f 6469 6374 2c74  stor_info_dict,t
+000171d0: 7265 655f 6279 5f6c 6576 656c 735f 6469  ree_by_levels_di
+000171e0: 6374 290a 2020 2020 6164 6469 7469 6f6e  ct).    addition
+000171f0: 616c 5f69 6e66 6f20 3d20 4164 6469 7469  al_info = Additi
+00017200: 6f6e 616c 496e 666f 2862 6c6f 7375 6d3d  onalInfo(blosum=
+00017210: 746f 7263 682e 6672 6f6d 5f6e 756d 7079  torch.from_numpy
+00017220: 2862 6c6f 7375 6d5f 6172 7261 7929 2c0a  (blosum_array),.
 00017230: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00017240: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017250: 2020 2020 626c 6f73 756d 5f64 6963 743d      blosum_dict=
-00017260: 626c 6f73 756d 5f64 6963 742c 0a20 2020  blosum_dict,.   
+00017250: 2020 2020 2062 6c6f 7375 6d5f 6469 6374       blosum_dict
+00017260: 3d62 6c6f 7375 6d5f 6469 6374 2c0a 2020  =blosum_dict,.  
 00017270: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00017280: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017290: 2020 6368 696c 6472 656e 5f64 6963 7420    children_dict 
-000172a0: 3d20 6368 696c 6472 656e 5f64 6963 742c  = children_dict,
-000172b0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00017290: 2020 2063 6869 6c64 7265 6e5f 6469 6374     children_dict
+000172a0: 203d 2063 6869 6c64 7265 6e5f 6469 6374   = children_dict
+000172b0: 2c0a 2020 2020 2020 2020 2020 2020 2020  ,.              
 000172c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000172d0: 2020 2020 2020 616e 6365 7374 6f72 5f69        ancestor_i
-000172e0: 6e66 6f5f 6469 6374 3d61 6e63 6573 746f  nfo_dict=ancesto
-000172f0: 725f 696e 666f 5f64 6963 742c 0a20 2020  r_info_dict,.   
+000172d0: 2020 2020 2020 2061 6e63 6573 746f 725f         ancestor_
+000172e0: 696e 666f 5f64 6963 743d 616e 6365 7374  info_dict=ancest
+000172f0: 6f72 5f69 6e66 6f5f 6469 6374 2c0a 2020  or_info_dict,.  
 00017300: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00017310: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017320: 2020 7472 6565 5f62 795f 6c65 7665 6c73    tree_by_levels
-00017330: 5f64 6963 743d 7472 6565 5f62 795f 6c65  _dict=tree_by_le
-00017340: 7665 6c73 5f64 6963 742c 0a20 2020 2020  vels_dict,.     
+00017320: 2020 2074 7265 655f 6279 5f6c 6576 656c     tree_by_level
+00017330: 735f 6469 6374 3d74 7265 655f 6279 5f6c  s_dict=tree_by_l
+00017340: 6576 656c 735f 6469 6374 2c0a 2020 2020  evels_dict,.    
 00017350: 2020 2020 2020 2020 2020 2020 2020 2020                  
 00017360: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017370: 7472 6565 5f62 795f 6c65 7665 6c73 5f61  tree_by_levels_a
-00017380: 7272 6179 3d74 7265 655f 6279 5f6c 6576  rray=tree_by_lev
-00017390: 656c 735f 6172 7261 792c 0a20 2020 2020  els_array,.     
+00017370: 2074 7265 655f 6279 5f6c 6576 656c 735f   tree_by_levels_
+00017380: 6172 7261 793d 7472 6565 5f62 795f 6c65  array=tree_by_le
+00017390: 7665 6c73 5f61 7272 6179 2c0a 2020 2020  vels_array,.    
 000173a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
 000173b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000173c0: 7061 7472 6973 7469 635f 696e 666f 3d70  patristic_info=p
-000173d0: 6174 7269 7374 6963 5f69 6e66 6f2c 0a20  atristic_info,. 
+000173c0: 2070 6174 7269 7374 6963 5f69 6e66 6f3d   patristic_info=
+000173d0: 7061 7472 6973 7469 635f 696e 666f 2c0a  patristic_info,.
 000173e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
 000173f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017400: 2020 2020 6772 6170 685f 636f 6f3d 6772      graph_coo=gr
-00017410: 6170 685f 636f 6f2c 0a20 2020 2020 2020  aph_coo,.       
+00017400: 2020 2020 2067 7261 7068 5f63 6f6f 3d67       graph_coo=g
+00017410: 7261 7068 5f63 6f6f 2c0a 2020 2020 2020  raph_coo,.      
 00017420: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017430: 2020 2020 2020 2020 2020 2020 2020 7061                pa
-00017440: 7472 6973 7469 635f 6675 6c6c 5f73 7061  tristic_full_spa
-00017450: 7273 6520 3d20 6564 6765 5f77 6569 6768  rse = edge_weigh
-00017460: 745f 6d61 7472 6978 2c0a 2020 2020 2020  t_matrix,.      
+00017430: 2020 2020 2020 2020 2020 2020 2020 2070                 p
+00017440: 6174 7269 7374 6963 5f66 756c 6c5f 7370  atristic_full_sp
+00017450: 6172 7365 203d 2065 6467 655f 7765 6967  arse = edge_weig
+00017460: 6874 5f6d 6174 7269 782c 0a20 2020 2020  ht_matrix,.     
 00017470: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00017480: 2020 2020 2020 2020 2020 2020 2020 206e                 n
-00017490: 6f64 6573 5f72 6570 7265 7365 6e74 6174  odes_representat
-000174a0: 696f 6e73 5f61 7272 6179 3d74 6f72 6368  ions_array=torch
-000174b0: 2e66 726f 6d5f 6e75 6d70 7928 6e6f 6465  .from_numpy(node
-000174c0: 735f 7265 7072 6573 656e 7461 7469 6f6e  s_representation
-000174d0: 735f 6172 7261 7929 2c0a 2020 2020 2020  s_array),.      
+00017480: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017490: 6e6f 6465 735f 7265 7072 6573 656e 7461  nodes_representa
+000174a0: 7469 6f6e 735f 6172 7261 793d 746f 7263  tions_array=torc
+000174b0: 682e 6672 6f6d 5f6e 756d 7079 286e 6f64  h.from_numpy(nod
+000174c0: 6573 5f72 6570 7265 7365 6e74 6174 696f  es_representatio
+000174d0: 6e73 5f61 7272 6179 292c 0a20 2020 2020  ns_array),.     
 000174e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000174f0: 2020 2020 2020 2020 2020 2020 2020 2064                 d
-00017500: 676c 5f67 7261 7068 3d64 676c 5f67 7261  gl_graph=dgl_gra
-00017510: 7068 290a 0a0a 2020 2020 7265 7475 726e  ph)...    return
-00017520: 2061 6464 6974 696f 6e61 6c5f 696e 666f   additional_info
-00017530: 0a64 6566 2063 7265 6174 655f 7472 6565  .def create_tree
-00017540: 5f62 795f 6c65 7665 6c73 2863 6869 6c64  _by_levels(child
-00017550: 7265 6e5f 6469 6374 293a 0a20 2020 2022  ren_dict):.    "
-00017560: 2222 416c 7465 726e 6174 6976 6520 6d65  ""Alternative me
-00017570: 7468 6f64 2074 6f20 6275 696c 6420 7468  thod to build th
-00017580: 6520 7472 6565 2062 7920 6c65 7665 6c73  e tree by levels
-00017590: 2061 7320 7375 6767 6573 7465 6420 6279   as suggested by
-000175a0: 2052 6f62 6572 7422 2222 0a20 2020 2074   Robert""".    t
-000175b0: 7265 655f 6c65 7665 6c73 203d 205b 5b30  ree_levels = [[0
-000175c0: 5d5d 0a20 2020 2063 7572 7265 6e74 5f6e  ]].    current_n
-000175d0: 6f64 6573 203d 205b 305d 0a20 2020 2077  odes = [0].    w
-000175e0: 6869 6c65 206c 656e 2863 7572 7265 6e74  hile len(current
-000175f0: 5f6e 6f64 6573 2920 3e20 303a 0a20 2020  _nodes) > 0:.   
-00017600: 2020 2020 2063 6869 6c64 7265 6e5f 6e6f       children_no
-00017610: 6465 7320 3d20 5b5d 0a20 2020 2020 2020  des = [].       
-00017620: 2066 6f72 206e 6f64 6520 696e 2063 7572   for node in cur
-00017630: 7265 6e74 5f6e 6f64 6573 3a0a 2020 2020  rent_nodes:.    
-00017640: 2020 2020 2020 2020 6966 206e 6f64 6520          if node 
-00017650: 696e 2063 6869 6c64 7265 6e5f 6469 6374  in children_dict
-00017660: 2e6b 6579 7328 293a 0a20 2020 2020 2020  .keys():.       
-00017670: 2020 2020 2020 2020 2063 6869 6c64 7265           childre
-00017680: 6e5f 6e6f 6465 7320 2b3d 2063 6869 6c64  n_nodes += child
-00017690: 7265 6e5f 6469 6374 5b6e 6f64 655d 0a0a  ren_dict[node]..
-000176a0: 2020 2020 2020 2020 7472 6565 5f6c 6576          tree_lev
-000176b0: 656c 732e 6170 7065 6e64 2863 6869 6c64  els.append(child
-000176c0: 7265 6e5f 6e6f 6465 7329 0a20 2020 2020  ren_nodes).     
-000176d0: 2020 2063 7572 7265 6e74 5f6e 6f64 6573     current_nodes
-000176e0: 203d 2063 6869 6c64 7265 6e5f 6e6f 6465   = children_node
-000176f0: 730a 0a20 2020 2072 6574 7572 6e20 7472  s..    return tr
-00017700: 6565 5f6c 6576 656c 730a 636c 6173 7320  ee_levels.class 
-00017710: 4d79 4461 7461 7365 7428 4461 7461 7365  MyDataset(Datase
-00017720: 7429 3a23 544f 444f 3a20 7265 6d6f 7665  t):#TODO: remove
-00017730: 0a20 2020 2064 6566 205f 5f69 6e69 745f  .    def __init_
-00017740: 5f28 7365 6c66 2c6c 6162 656c 732c 6461  _(self,labels,da
-00017750: 7461 5f61 7272 6179 7329 3a0a 2020 2020  ta_arrays):.    
-00017760: 2020 2020 7365 6c66 2e6c 6162 656c 7320      self.labels 
-00017770: 3d20 6c61 6265 6c73 0a20 2020 2020 2020  = labels.       
-00017780: 2073 656c 662e 6461 7461 5f61 7272 6179   self.data_array
-00017790: 7320 3d20 6461 7461 5f61 7272 6179 730a  s = data_arrays.
-000177a0: 0a20 2020 2064 6566 205f 5f67 6574 6974  .    def __getit
-000177b0: 656d 5f5f 2873 656c 662c 2069 6e64 6578  em__(self, index
-000177c0: 293a 2023 7365 7473 2061 5b69 5d0a 2020  ): #sets a[i].  
-000177d0: 2020 2020 2020 6c61 6265 6c20 3d20 7365        label = se
-000177e0: 6c66 2e6c 6162 656c 735b 696e 6465 785d  lf.labels[index]
-000177f0: 0a20 2020 2020 2020 2064 6174 6120 3d20  .        data = 
-00017800: 7365 6c66 2e64 6174 615f 6172 7261 7973  self.data_arrays
-00017810: 5b69 6e64 6578 5d0a 2020 2020 2020 2020  [index].        
-00017820: 7265 7475 726e 207b 2766 616d 696c 795f  return {'family_
-00017830: 6e61 6d65 273a 206c 6162 656c 2c20 2766  name': label, 'f
-00017840: 616d 696c 795f 6461 7461 273a 2064 6174  amily_data': dat
-00017850: 617d 0a20 2020 2064 6566 205f 5f6c 656e  a}.    def __len
-00017860: 5f5f 2873 656c 6629 3a0a 2020 2020 2020  __(self):.      
-00017870: 2020 7265 7475 726e 206c 656e 2873 656c    return len(sel
-00017880: 662e 6c61 6265 6c73 290a 6465 6620 6465  f.labels).def de
-00017890: 6669 6e65 5f62 6174 6368 5f73 697a 6528  fine_batch_size(
-000178a0: 6e2c 6261 7463 685f 7369 7a65 3d54 7275  n,batch_size=Tru
-000178b0: 652c 2062 656e 6368 6d61 726b 696e 673d  e, benchmarking=
-000178c0: 4661 6c73 6529 3a0a 2020 2020 2222 2241  False):.    """A
-000178d0: 7574 6f6d 6174 6963 2063 616c 6375 6c61  utomatic calcula
-000178e0: 7469 6f6e 2074 6865 2061 7661 696c 6162  tion the availab
-000178f0: 6c65 2064 6976 6973 6f72 7320 6f66 2074  le divisors of t
-00017900: 6865 206e 756d 6265 7220 6f66 2074 7261  he number of tra
-00017910: 696e 696e 6720 6461 7461 2070 6f69 6e74  ining data point
-00017920: 7320 286e 292e 0a20 2020 2054 6869 7320  s (n)..    This 
-00017930: 6865 6c70 7320 746f 2073 7567 6765 7374  helps to suggest
-00017940: 2061 6e20 6170 7072 6f70 6961 7465 2069   an appropiate i
-00017950: 6e74 6567 6572 2062 6174 6368 2073 697a  nteger batch siz
-00017960: 6520 6e75 6d62 6572 2c20 7468 6174 2073  e number, that s
-00017970: 706c 6974 7320 6576 656e 6c79 2074 6865  plits evenly the
-00017980: 2064 6174 610a 2020 2020 3a70 6172 616d   data.    :param
-00017990: 2069 6e74 206e 3a20 6e75 6d62 6572 206f   int n: number o
-000179a0: 6620 7365 7175 656e 6365 730a 2020 2020  f sequences.    
-000179b0: 3a70 6172 616d 2062 6f6f 6c20 6261 7463  :param bool batc
-000179c0: 685f 7369 7a65 0a20 2020 203a 7061 7261  h_size.    :para
-000179d0: 6d20 6265 6e63 686d 6172 6b69 6e67 2222  m benchmarking""
-000179e0: 220a 2020 2020 6966 206e 6f74 2062 656e  ".    if not ben
-000179f0: 6368 6d61 726b 696e 673a 0a20 2020 2020  chmarking:.     
-00017a00: 2020 2061 7373 6572 7420 6e20 3e3d 2031     assert n >= 1
-00017a10: 3030 202c 224e 6f74 2077 6f72 7468 2062  00 ,"Not worth b
-00017a20: 6174 6368 696e 672c 206e 756d 6265 7220  atching, number 
-00017a30: 6f66 2073 6571 7565 6e63 6573 2069 7320  of sequences is 
-00017a40: 3c20 3130 3020 220a 2020 2020 6469 7669  < 100 ".    divi
-00017a50: 736f 7273 203d 2044 7261 7570 6e69 724d  sors = DraupnirM
-00017a60: 6f64 656c 5574 696c 732e 7072 696e 745f  odelUtils.print_
-00017a70: 6469 7669 736f 7273 286e 290a 2020 2020  divisors(n).    
-00017a80: 6e5f 6469 6769 7473 203d 206c 656e 286c  n_digits = len(l
-00017a90: 6973 7428 7374 7228 6e29 2929 2023 2031  ist(str(n))) # 1
-00017aa0: 3030 2068 6173 2033 2064 6967 6974 732c  00 has 3 digits,
-00017ab0: 2031 3030 3020 6861 7320 3420 6469 6769   1000 has 4 digi
-00017ac0: 7473 2c20 3130 3030 3020 6861 7320 3520  ts, 10000 has 5 
-00017ad0: 6469 6769 7473 2061 6e64 2073 6f20 6f6e  digits and so on
-00017ae0: 0a20 2020 206e 5f63 6875 6e6b 7320 3d20  .    n_chunks = 
-00017af0: 5b6d 696e 2864 6976 6973 6f72 735b 6e5f  [min(divisors[n_
-00017b00: 6469 6769 7473 2d31 3a5d 2920 6966 206c  digits-1:]) if l
-00017b10: 656e 2864 6976 6973 6f72 7329 203e 206e  en(divisors) > n
-00017b20: 5f64 6967 6974 732d 3120 656c 7365 2031  _digits-1 else 1
-00017b30: 5d5b 305d 2020 2320 736d 616c 6c65 7374  ][0]  # smallest
-00017b40: 2064 6976 6973 6f72 2074 6861 7420 6973   divisor that is
-00017b50: 206e 6f74 2031 2c20 6f72 206a 7573 7420   not 1, or just 
-00017b60: 7573 6520 3120 2c20 7768 656e 206e 6f20  use 1 , when no 
-00017b70: 6f74 6865 7220 6469 7669 736f 7273 2061  other divisors a
-00017b80: 7265 2061 7661 696c 6162 6c65 2028 7072  re available (pr
-00017b90: 696d 6520 6e75 6d62 6572 7329 0a20 2020  ime numbers).   
-00017ba0: 2062 6174 6368 7369 7a65 203d 2069 6e74   batchsize = int
-00017bb0: 2844 7261 7570 6e69 724d 6f64 656c 5574  (DraupnirModelUt
-00017bc0: 696c 732e 696e 7465 7276 616c 7328 6e5f  ils.intervals(n_
-00017bd0: 6368 756e 6b73 2c20 6e29 5b30 5d5b 315d  chunks, n)[0][1]
-00017be0: 290a 2020 2020 6966 2062 6174 6368 5f73  ).    if batch_s
-00017bf0: 697a 653a 7265 7475 726e 2062 6174 6368  ize:return batch
-00017c00: 7369 7a65 0a20 2020 2065 6c73 653a 7265  size.    else:re
-00017c10: 7475 726e 206e 5f63 6875 6e6b 730a 6465  turn n_chunks.de
-00017c20: 6620 636f 7661 7269 616e 6365 2878 2c79  f covariance(x,y
-00017c30: 293a 0a20 2020 2022 2222 436f 6d70 7574  ):.    """Comput
-00017c40: 6573 2063 726f 7373 2d63 6f76 6172 6961  es cross-covaria
-00017c50: 6e63 6520 6265 7477 6565 6e20 7665 6374  nce between vect
-00017c60: 6f72 732c 2061 6e64 2069 7320 6465 6669  ors, and is defi
-00017c70: 6e65 6420 6279 2063 6f76 5b58 2c59 5d3d  ned by cov[X,Y]=
-00017c80: 455b 2858 e288 92ce bc58 2928 59e2 8892  E[(X.....X)(Y...
-00017c90: cebc 5929 545d 2222 220a 2020 2020 4120  ..Y)T]""".    A 
-00017ca0: 3d20 746f 7263 682e 7371 7274 2874 6f72  = torch.sqrt(tor
-00017cb0: 6368 2e61 7261 6e67 6528 3132 292e 7265  ch.arange(12).re
-00017cc0: 7368 6170 6528 332c 2034 2929 2020 2320  shape(3, 4))  # 
-00017cd0: 736f 6d65 2033 2062 7920 3420 6172 7261  some 3 by 4 arra
-00017ce0: 790a 2020 2020 6220 3d20 746f 7263 682e  y.    b = torch.
-00017cf0: 7465 6e73 6f72 285b 5b32 5d2c 205b 345d  tensor([[2], [4]
-00017d00: 2c20 5b35 5d5d 2920 2023 2073 6f6d 6520  , [5]])  # some 
-00017d10: 3320 6279 2031 2076 6563 746f 720a 2020  3 by 1 vector.  
-00017d20: 2020 636f 7620 3d20 746f 7263 682e 646f    cov = torch.do
-00017d30: 7428 622e 5420 2d20 622e 6d65 616e 2829  t(b.T - b.mean()
-00017d40: 2c20 4120 2d20 412e 6d65 616e 2864 696d  , A - A.mean(dim
-00017d50: 3d30 2929 202f 2028 622e 7368 6170 655b  =0)) / (b.shape[
-00017d60: 305d 202d 2031 290a 2020 2020 7265 7475  0] - 1).    retu
-00017d70: 726e 2063 6f76 0a64 6566 2066 696e 645f  rn cov.def find_
-00017d80: 6e61 6e28 6172 7261 792c 6e61 6e3d 5472  nan(array,nan=Tr
-00017d90: 7565 293a 0a20 2020 2022 2222 4472 6f70  ue):.    """Drop
-00017da0: 7320 726f 7773 2069 6e20 6120 7465 6e73  s rows in a tens
-00017db0: 6f72 2063 6f6e 7461 696e 696e 6720 616e  or containing an
-00017dc0: 7920 6e61 6e20 7661 6c75 6573 0a20 2020  y nan values.   
-00017dd0: 203a 7061 7261 6d20 7465 6e73 6f72 2061   :param tensor a
-00017de0: 7272 6179 2222 220a 2020 2020 7368 6170  rray""".    shap
-00017df0: 6520 3d20 6172 7261 792e 7368 6170 650a  e = array.shape.
-00017e00: 2020 2020 7465 6e73 6f72 5f72 6573 6861      tensor_resha
-00017e10: 7065 6420 3d20 6172 7261 792e 7265 7368  ped = array.resh
-00017e20: 6170 6528 7368 6170 655b 305d 2c20 2d31  ape(shape[0], -1
-00017e30: 290a 2020 2020 2320 4472 6f70 2061 6c6c  ).    # Drop all
-00017e40: 2072 6f77 7320 636f 6e74 6169 6e69 6e67   rows containing
-00017e50: 2061 6e79 206e 616e 3a0a 2020 2020 6966   any nan:.    if
-00017e60: 206e 6f74 206e 616e 3a0a 2020 2020 2020   not nan:.      
-00017e70: 2020 7465 6e73 6f72 5f72 6573 6861 7065    tensor_reshape
-00017e80: 6420 3d20 7465 6e73 6f72 5f72 6573 6861  d = tensor_resha
-00017e90: 7065 645b 7e74 6f72 6368 2e61 6e79 2874  ped[~torch.any(t
-00017ea0: 656e 736f 725f 7265 7368 6170 6564 2e69  ensor_reshaped.i
-00017eb0: 736e 616e 2829 2c20 6469 6d3d 3129 5d0a  snan(), dim=1)].
-00017ec0: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
-00017ed0: 2020 7465 6e73 6f72 5f72 6573 6861 7065    tensor_reshape
-00017ee0: 6420 3d20 7465 6e73 6f72 5f72 6573 6861  d = tensor_resha
-00017ef0: 7065 645b 746f 7263 682e 616e 7928 7465  ped[torch.any(te
-00017f00: 6e73 6f72 5f72 6573 6861 7065 642e 6973  nsor_reshaped.is
-00017f10: 6e61 6e28 292c 2064 696d 3d31 295d 0a20  nan(), dim=1)]. 
-00017f20: 2020 2023 2052 6573 6861 7065 2062 6163     # Reshape bac
-00017f30: 6b3a 0a20 2020 2061 7272 6179 203d 2074  k:.    array = t
-00017f40: 656e 736f 725f 7265 7368 6170 6564 2e72  ensor_reshaped.r
-00017f50: 6573 6861 7065 2874 656e 736f 725f 7265  eshape(tensor_re
-00017f60: 7368 6170 6564 2e73 6861 7065 5b30 5d2c  shaped.shape[0],
-00017f70: 202a 7368 6170 655b 313a 5d29 0a20 2020   *shape[1:]).   
-00017f80: 2072 6574 7572 6e20 6172 7261 790a 6465   return array.de
-00017f90: 6620 7261 6d61 6368 616e 6472 616e 5f70  f ramachandran_p
-00017fa0: 6c6f 7428 2064 6174 615f 616e 676c 6573  lot( data_angles
-00017fb0: 2c73 6176 655f 6469 7265 6374 6f72 792c  ,save_directory,
-00017fc0: 706c 6f74 5f74 6974 6c65 2c20 6f6e 655f  plot_title, one_
-00017fd0: 686f 745f 656e 636f 6465 6420 3d20 4661  hot_encoded = Fa
-00017fe0: 6c73 6520 293a 0a20 2020 2022 2222 506c  lse ):.    """Pl
-00017ff0: 6f74 7320 5261 6d61 6368 616e 6472 616e  ots Ramachandran
-00018000: 2070 6c6f 7473 2062 6574 7765 656e 2074   plots between t
-00018010: 6865 2070 6869 2061 6e64 2070 7369 2061  he phi and psi a
-00018020: 6e67 6c65 7320 696e 2074 6865 2064 6174  ngles in the dat
-00018030: 6173 6574 0a20 2020 203a 7061 7261 6d20  aset.    :param 
-00018040: 6e75 6d70 792d 6172 7261 7920 6461 7461  numpy-array data
-00018050: 5f61 6e67 6c65 7320 3a20 3364 2061 7272  _angles : 3d arr
-00018060: 6179 205b 6e5f 7365 712c 2061 6c69 676e  ay [n_seq, align
-00018070: 5f6c 656e 2c20 3330 5d2c 2077 6865 7265  _len, 30], where
-00018080: 2063 6f6c 756d 6e73 2031 2061 6e64 2032   columns 1 and 2
-00018090: 206f 7220 3231 2061 6e64 2032 3220 636f   or 21 and 22 co
-000180a0: 6e74 6169 6e20 7468 6520 616e 676c 6573  ntain the angles
-000180b0: 2069 6e20 7468 6520 6361 7365 206f 6620   in the case of 
-000180c0: 4e4f 5420 6f6e 652d 686f 7420 656e 636f  NOT one-hot enco
-000180d0: 6465 6420 616e 6420 6f6e 652d 686f 7420  ded and one-hot 
-000180e0: 656e 636f 6465 642c 2072 6573 7065 6374  encoded, respect
-000180f0: 6976 656c 790a 2020 2020 3a70 6172 616d  ively.    :param
-00018100: 2073 7472 2073 6176 655f 6469 7265 6374   str save_direct
-00018110: 6f72 793a 2070 6174 6820 746f 2073 6176  ory: path to sav
-00018120: 696e 6720 6469 7265 6374 6f72 790a 2020  ing directory.  
-00018130: 2020 3a70 6172 616d 2073 7472 2070 6c6f    :param str plo
-00018140: 745f 7469 746c 650a 2020 2020 3a70 6172  t_title.    :par
-00018150: 616d 2062 6f6f 6c20 6f6e 655f 686f 745f  am bool one_hot_
-00018160: 656e 636f 6465 640a 2020 2020 2222 220a  encoded.    """.
-00018170: 2020 2020 706c 742e 636c 6628 290a 2020      plt.clf().  
-00018180: 2020 6966 206f 6e65 5f68 6f74 5f65 6e63    if one_hot_enc
-00018190: 6f64 6564 3a0a 2020 2020 2020 2020 7068  oded:.        ph
-000181a0: 6920 3d20 6461 7461 5f61 6e67 6c65 735b  i = data_angles[
-000181b0: 3a2c 3a2c 3231 5d2e 7265 7368 6170 6528  :,:,21].reshape(
-000181c0: 2864 6174 615f 616e 676c 6573 2e73 6861  (data_angles.sha
-000181d0: 7065 5b30 5d2a 6461 7461 5f61 6e67 6c65  pe[0]*data_angle
-000181e0: 732e 7368 6170 655b 315d 2929 2e61 7374  s.shape[1])).ast
-000181f0: 7970 6528 666c 6f61 7429 2023 2d20 6e70  ype(float) #- np
-00018200: 2e70 690a 0a20 2020 2020 2020 2070 7369  .pi..        psi
-00018210: 203d 2064 6174 615f 616e 676c 6573 5b3a   = data_angles[:
-00018220: 2c3a 2c32 325d 2e72 6573 6861 7065 2828  ,:,22].reshape((
-00018230: 6461 7461 5f61 6e67 6c65 732e 7368 6170  data_angles.shap
-00018240: 655b 305d 2a64 6174 615f 616e 676c 6573  e[0]*data_angles
-00018250: 2e73 6861 7065 5b31 5d29 292e 6173 7479  .shape[1])).asty
-00018260: 7065 2866 6c6f 6174 2920 232d 206e 702e  pe(float) #- np.
-00018270: 7069 0a20 2020 2065 6c73 653a 0a20 2020  pi.    else:.   
-00018280: 2020 2020 2070 6869 203d 2064 6174 615f       phi = data_
-00018290: 616e 676c 6573 5b3a 2c3a 2c31 5d2e 7265  angles[:,:,1].re
-000182a0: 7368 6170 6528 2864 6174 615f 616e 676c  shape((data_angl
-000182b0: 6573 2e73 6861 7065 5b30 5d2a 6461 7461  es.shape[0]*data
-000182c0: 5f61 6e67 6c65 732e 7368 6170 655b 315d  _angles.shape[1]
-000182d0: 2929 2e61 7374 7970 6528 666c 6f61 7429  )).astype(float)
-000182e0: 0a20 2020 2020 2020 2070 7369 203d 2064  .        psi = d
-000182f0: 6174 615f 616e 676c 6573 5b3a 2c3a 2c32  ata_angles[:,:,2
-00018300: 5d2e 7265 7368 6170 6528 2864 6174 615f  ].reshape((data_
-00018310: 616e 676c 6573 2e73 6861 7065 5b30 5d2a  angles.shape[0]*
-00018320: 6461 7461 5f61 6e67 6c65 732e 7368 6170  data_angles.shap
-00018330: 655b 315d 2929 2e61 7374 7970 6528 666c  e[1])).astype(fl
-00018340: 6f61 7429 0a0a 2020 2020 6669 6720 3d20  oat)..    fig = 
-00018350: 706c 742e 6669 6775 7265 2866 6967 7369  plt.figure(figsi
-00018360: 7a65 3d28 352c 3529 290a 2020 2020 706c  ze=(5,5)).    pl
-00018370: 742e 6869 7374 3264 2820 7068 692c 2070  t.hist2d( phi, p
-00018380: 7369 2c20 6269 6e73 203d 2036 3238 2c20  si, bins = 628, 
-00018390: 6e6f 726d 203d 204c 6f67 4e6f 726d 2829  norm = LogNorm()
-000183a0: 2923 2c20 636d 6170 203d 2070 6c74 2e63  )#, cmap = plt.c
-000183b0: 6d2e 6a65 7420 290a 2020 2020 706c 742e  m.jet ).    plt.
-000183c0: 796c 696d 282d 6e70 2e70 692c 206e 702e  ylim(-np.pi, np.
-000183d0: 7069 290a 2020 2020 706c 742e 786c 696d  pi).    plt.xlim
-000183e0: 282d 6e70 2e70 692c 206e 702e 7069 290a  (-np.pi, np.pi).
-000183f0: 2020 2020 706c 742e 7469 746c 6528 706c      plt.title(pl
-00018400: 6f74 5f74 6974 6c65 2c66 6f6e 7473 697a  ot_title,fontsiz
-00018410: 653d 3132 290a 2020 2020 706c 742e 786c  e=12).    plt.xl
-00018420: 6162 656c 2827 cf86 2729 0a20 2020 2070  abel('..').    p
-00018430: 6c74 2e79 6c61 6265 6c28 27cf 8827 290a  lt.ylabel('..').
-00018440: 2020 2020 706c 742e 7361 7665 6669 6728      plt.savefig(
-00018450: 7361 7665 5f64 6972 6563 746f 7279 290a  save_directory).
-00018460: 2020 2020 706c 742e 636c 6628 290a 2020      plt.clf().  
-00018470: 2020 706c 742e 636c 6f73 6528 6669 6729    plt.close(fig)
-00018480: 0a64 6566 2072 616d 6163 6861 6e64 7261  .def ramachandra
-00018490: 6e5f 706c 6f74 5f73 616d 706c 6564 2870  n_plot_sampled(p
-000184a0: 6869 2c70 7369 2c73 6176 655f 6469 7265  hi,psi,save_dire
-000184b0: 6374 6f72 792c 706c 6f74 5f74 6974 6c65  ctory,plot_title
-000184c0: 2c70 6c6f 745f 6b61 7070 6173 3d46 616c  ,plot_kappas=Fal
-000184d0: 7365 293a 0a20 2020 2022 2222 506c 6f74  se):.    """Plot
-000184e0: 7320 5261 6d61 6368 616e 6472 616e 2070  s Ramachandran p
-000184f0: 6c6f 7473 2062 6574 7765 656e 2074 6865  lots between the
-00018500: 2070 6869 2061 6e64 2070 7369 2061 6e67   phi and psi ang
-00018510: 6c65 7320 696e 2074 6865 2064 6174 6173  les in the datas
-00018520: 6574 0a20 2020 203a 7061 7261 6d20 7465  et.    :param te
-00018530: 6e73 6f72 2070 6869 0a20 2020 203a 7061  nsor phi.    :pa
-00018540: 7261 6d20 7465 6e73 6f72 2070 7369 0a20  ram tensor psi. 
-00018550: 2020 203a 7061 7261 6d20 7374 7220 7361     :param str sa
-00018560: 7665 5f64 6972 6563 746f 7279 3a20 7061  ve_directory: pa
-00018570: 7468 2074 6f20 7361 7669 6e67 2064 6972  th to saving dir
-00018580: 6563 746f 7279 0a20 2020 203a 7061 7261  ectory.    :para
-00018590: 6d20 7374 7220 706c 6f74 5f74 6974 6c65  m str plot_title
-000185a0: 0a20 2020 203a 7061 7261 6d20 626f 6f6c  .    :param bool
-000185b0: 2070 6c6f 745f 6b61 7070 6173 0a20 2020   plot_kappas.   
-000185c0: 2022 2222 0a20 2020 2069 6620 6973 696e   """.    if isin
-000185d0: 7374 616e 6365 2870 6869 2c74 6f72 6368  stance(phi,torch
-000185e0: 2e54 656e 736f 7229 3a0a 2020 2020 2020  .Tensor):.      
-000185f0: 2020 7068 6920 3d20 7068 692e 7669 6577    phi = phi.view
-00018600: 282d 3129 2e63 7075 2829 2e64 6574 6163  (-1).cpu().detac
-00018610: 6828 292e 6e75 6d70 7928 290a 2020 2020  h().numpy().    
-00018620: 2020 2020 7073 6920 3d20 7073 692e 7669      psi = psi.vi
-00018630: 6577 282d 3129 2e63 7075 2829 2e64 6574  ew(-1).cpu().det
-00018640: 6163 6828 292e 6e75 6d70 7928 290a 2020  ach().numpy().  
-00018650: 2020 2361 7865 7320 3d20 5b5b 2d6e 702e    #axes = [[-np.
-00018660: 7069 2c20 6e70 2e70 695d 2c20 5b2d 6e70  pi, np.pi], [-np
-00018670: 2e70 692c 206e 702e 7069 5d5d 0a20 2020  .pi, np.pi]].   
-00018680: 2070 6c74 2e66 6967 7572 6528 6669 6773   plt.figure(figs
-00018690: 697a 653d 2835 2c35 2929 0a20 2020 2070  ize=(5,5)).    p
-000186a0: 6c74 2e68 6973 7432 6428 2070 6869 2c20  lt.hist2d( phi, 
-000186b0: 7073 692c 2062 696e 7320 3d20 3632 382c  psi, bins = 628,
-000186c0: 206e 6f72 6d20 3d20 4c6f 674e 6f72 6d28   norm = LogNorm(
-000186d0: 2929 232c 2063 6d61 7020 3d20 706c 742e  ))#, cmap = plt.
-000186e0: 636d 2e6a 6574 2029 0a20 2020 2069 6620  cm.jet ).    if 
-000186f0: 706c 6f74 5f6b 6170 7061 733a 0a20 2020  plot_kappas:.   
-00018700: 2020 2020 2070 6c74 2e78 6c61 6265 6c28       plt.xlabel(
-00018710: 274b 6170 7061 20cf 8627 290a 2020 2020  'Kappa ..').    
-00018720: 2020 2020 706c 742e 796c 6162 656c 2827      plt.ylabel('
-00018730: 4b61 7070 6120 cf88 2729 0a20 2020 2065  Kappa ..').    e
-00018740: 6c73 653a 0a20 2020 2020 2020 2070 6c74  lse:.        plt
-00018750: 2e79 6c69 6d28 2d6e 702e 7069 2c20 6e70  .ylim(-np.pi, np
-00018760: 2e70 6929 0a20 2020 2020 2020 2070 6c74  .pi).        plt
-00018770: 2e78 6c69 6d28 2d6e 702e 7069 2c20 6e70  .xlim(-np.pi, np
-00018780: 2e70 6929 0a20 2020 2020 2020 2070 6c74  .pi).        plt
-00018790: 2e78 6c61 6265 6c28 27cf 8627 290a 2020  .xlabel('..').  
-000187a0: 2020 2020 2020 706c 742e 796c 6162 656c        plt.ylabel
-000187b0: 2827 cf88 2729 0a20 2020 2070 6c74 2e74  ('..').    plt.t
-000187c0: 6974 6c65 2870 6c6f 745f 7469 746c 652c  itle(plot_title,
-000187d0: 666f 6e74 7369 7a65 3d31 3229 0a20 2020  fontsize=12).   
-000187e0: 2070 6c74 2e73 6176 6566 6967 2873 6176   plt.savefig(sav
-000187f0: 655f 6469 7265 6374 6f72 7929 0a20 2020  e_directory).   
-00018800: 2070 6c74 2e63 6c66 2829 0a20 2020 2070   plt.clf().    p
-00018810: 6c74 2e63 6c6f 7365 2829 0a64 6566 2067  lt.close().def g
-00018820: 7261 6469 656e 7473 5f70 6c6f 7428 6772  radients_plot(gr
-00018830: 6164 6965 6e74 5f6e 6f72 6d73 2c65 706f  adient_norms,epo
-00018840: 6368 732c 6469 7265 6374 6f72 7929 3a0a  chs,directory):.
-00018850: 2020 2020 2222 2256 6973 7561 6c69 7a61      """Visualiza
-00018860: 7469 6f6e 206f 6620 7468 6520 6772 6164  tion of the grad
-00018870: 6965 6e74 2064 6573 6365 6e74 2070 6572  ient descent per
-00018880: 206d 6f64 656c 2070 6172 616d 6574 6572   model parameter
-00018890: 0a20 2020 203a 7061 7261 6d20 6469 6374  .    :param dict
-000188a0: 2067 7261 6469 656e 745f 6e6f 726d 733a   gradient_norms:
-000188b0: 2064 6963 7469 6f6e 6172 7920 7769 7468   dictionary with
-000188c0: 2074 6865 206d 6f64 656c 2070 6172 616d   the model param
-000188d0: 6574 6572 7320 616e 6420 7468 6520 7061  eters and the pa
-000188e0: 7274 6961 6c20 6465 7269 7661 7469 7665  rtial derivative
-000188f0: 730a 2020 2020 3a70 6172 616d 2069 6e74  s.    :param int
-00018900: 2065 706f 6368 7322 2222 0a0a 2020 2020   epochs"""..    
-00018910: 6669 6720 3d20 706c 742e 6669 6775 7265  fig = plt.figure
-00018920: 2866 6967 7369 7a65 3d28 3136 2c20 3929  (figsize=(16, 9)
-00018930: 2c20 6470 693d 3130 3029 2e73 6574 5f66  , dpi=100).set_f
-00018940: 6163 6563 6f6c 6f72 2827 7768 6974 6527  acecolor('white'
-00018950: 290a 2020 2020 6178 203d 2070 6c74 2e73  ).    ax = plt.s
-00018960: 7562 706c 6f74 2831 3231 290a 2020 2020  ubplot(121).    
-00018970: 636f 6c6f 725f 6d61 7020 3d20 636d 2e72  color_map = cm.r
-00018980: 6169 6e62 6f77 286e 702e 6c69 6e73 7061  ainbow(np.linspa
-00018990: 6365 2830 2c20 312c 2034 3229 290a 2020  ce(0, 1, 42)).  
-000189a0: 2020 666f 7220 286e 616d 655f 692c 2067    for (name_i, g
-000189b0: 7261 645f 6e6f 726d 7329 2c20 636f 6c6f  rad_norms), colo
-000189c0: 7220 696e 207a 6970 2867 7261 6469 656e  r in zip(gradien
-000189d0: 745f 6e6f 726d 732e 6974 656d 7328 292c  t_norms.items(),
-000189e0: 2063 6f6c 6f72 5f6d 6170 293a 0a20 2020   color_map):.   
-000189f0: 2020 2020 2061 782e 706c 6f74 2867 7261       ax.plot(gra
-00018a00: 645f 6e6f 726d 732c 206c 6162 656c 3d6e  d_norms, label=n
-00018a10: 616d 655f 692c 2063 3d63 6f6c 6f72 290a  ame_i, c=color).
-00018a20: 2020 2020 706c 742e 786c 6162 656c 2827      plt.xlabel('
-00018a30: 6974 6572 7327 290a 2020 2020 706c 742e  iters').    plt.
-00018a40: 796c 6162 656c 2827 6772 6164 6965 6e74  ylabel('gradient
-00018a50: 206e 6f72 6d27 290a 2020 2020 706c 742e   norm').    plt.
-00018a60: 7973 6361 6c65 2827 6c6f 6727 290a 2020  yscale('log').  
-00018a70: 2020 6178 2e6c 6567 656e 6428 6c6f 633d    ax.legend(loc=
-00018a80: 2775 7070 6572 2063 656e 7465 7227 2c20  'upper center', 
-00018a90: 6262 6f78 5f74 6f5f 616e 6368 6f72 3d28  bbox_to_anchor=(
-00018aa0: 312e 352c 2031 292c 2073 6861 646f 773d  1.5, 1), shadow=
-00018ab0: 5472 7565 2c20 6e63 6f6c 3d31 2c20 7072  True, ncol=1, pr
-00018ac0: 6f70 3d7b 2773 697a 6527 3a20 387d 290a  op={'size': 8}).
-00018ad0: 2020 2020 706c 742e 7469 746c 6528 2747      plt.title('G
-00018ae0: 7261 6469 656e 7420 6e6f 726d 7320 6475  radient norms du
-00018af0: 7269 6e67 2053 5649 2729 0a20 2020 2070  ring SVI').    p
-00018b00: 6c74 2e73 6176 6566 6967 2822 7b7d 2f47  lt.savefig("{}/G
-00018b10: 7261 6469 656e 7473 5f7b 7d5f 6570 6f63  radients_{}_epoc
-00018b20: 6873 2e70 6e67 222e 666f 726d 6174 2864  hs.png".format(d
-00018b30: 6972 6563 746f 7279 2c65 706f 6368 7329  irectory,epochs)
-00018b40: 290a 2020 2020 706c 742e 636c 6628 290a  ).    plt.clf().
-00018b50: 2020 2020 706c 742e 636c 6f73 6528 290a      plt.close().
+000174f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00017500: 6467 6c5f 6772 6170 683d 6467 6c5f 6772  dgl_graph=dgl_gr
+00017510: 6170 6829 0a0a 0a20 2020 2072 6574 7572  aph)...    retur
+00017520: 6e20 6164 6469 7469 6f6e 616c 5f69 6e66  n additional_inf
+00017530: 6f0a 6465 6620 6372 6561 7465 5f74 7265  o.def create_tre
+00017540: 655f 6279 5f6c 6576 656c 7328 6368 696c  e_by_levels(chil
+00017550: 6472 656e 5f64 6963 7429 3a0a 2020 2020  dren_dict):.    
+00017560: 2222 2241 6c74 6572 6e61 7469 7665 206d  """Alternative m
+00017570: 6574 686f 6420 746f 2062 7569 6c64 2074  ethod to build t
+00017580: 6865 2074 7265 6520 6279 206c 6576 656c  he tree by level
+00017590: 7320 6173 2073 7567 6765 7374 6564 2062  s as suggested b
+000175a0: 7920 526f 6265 7274 2222 220a 2020 2020  y Robert""".    
+000175b0: 7472 6565 5f6c 6576 656c 7320 3d20 5b5b  tree_levels = [[
+000175c0: 305d 5d0a 2020 2020 6375 7272 656e 745f  0]].    current_
+000175d0: 6e6f 6465 7320 3d20 5b30 5d0a 2020 2020  nodes = [0].    
+000175e0: 7768 696c 6520 6c65 6e28 6375 7272 656e  while len(curren
+000175f0: 745f 6e6f 6465 7329 203e 2030 3a0a 2020  t_nodes) > 0:.  
+00017600: 2020 2020 2020 6368 696c 6472 656e 5f6e        children_n
+00017610: 6f64 6573 203d 205b 5d0a 2020 2020 2020  odes = [].      
+00017620: 2020 666f 7220 6e6f 6465 2069 6e20 6375    for node in cu
+00017630: 7272 656e 745f 6e6f 6465 733a 0a20 2020  rrent_nodes:.   
+00017640: 2020 2020 2020 2020 2069 6620 6e6f 6465           if node
+00017650: 2069 6e20 6368 696c 6472 656e 5f64 6963   in children_dic
+00017660: 742e 6b65 7973 2829 3a0a 2020 2020 2020  t.keys():.      
+00017670: 2020 2020 2020 2020 2020 6368 696c 6472            childr
+00017680: 656e 5f6e 6f64 6573 202b 3d20 6368 696c  en_nodes += chil
+00017690: 6472 656e 5f64 6963 745b 6e6f 6465 5d0a  dren_dict[node].
+000176a0: 0a20 2020 2020 2020 2074 7265 655f 6c65  .        tree_le
+000176b0: 7665 6c73 2e61 7070 656e 6428 6368 696c  vels.append(chil
+000176c0: 6472 656e 5f6e 6f64 6573 290a 2020 2020  dren_nodes).    
+000176d0: 2020 2020 6375 7272 656e 745f 6e6f 6465      current_node
+000176e0: 7320 3d20 6368 696c 6472 656e 5f6e 6f64  s = children_nod
+000176f0: 6573 0a0a 2020 2020 7265 7475 726e 2074  es..    return t
+00017700: 7265 655f 6c65 7665 6c73 0a63 6c61 7373  ree_levels.class
+00017710: 204d 7944 6174 6173 6574 2844 6174 6173   MyDataset(Datas
+00017720: 6574 293a 2354 4f44 4f3a 2072 656d 6f76  et):#TODO: remov
+00017730: 650a 2020 2020 6465 6620 5f5f 696e 6974  e.    def __init
+00017740: 5f5f 2873 656c 662c 6c61 6265 6c73 2c64  __(self,labels,d
+00017750: 6174 615f 6172 7261 7973 293a 0a20 2020  ata_arrays):.   
+00017760: 2020 2020 2073 656c 662e 6c61 6265 6c73       self.labels
+00017770: 203d 206c 6162 656c 730a 2020 2020 2020   = labels.      
+00017780: 2020 7365 6c66 2e64 6174 615f 6172 7261    self.data_arra
+00017790: 7973 203d 2064 6174 615f 6172 7261 7973  ys = data_arrays
+000177a0: 0a0a 2020 2020 6465 6620 5f5f 6765 7469  ..    def __geti
+000177b0: 7465 6d5f 5f28 7365 6c66 2c20 696e 6465  tem__(self, inde
+000177c0: 7829 3a20 2373 6574 7320 615b 695d 0a20  x): #sets a[i]. 
+000177d0: 2020 2020 2020 206c 6162 656c 203d 2073         label = s
+000177e0: 656c 662e 6c61 6265 6c73 5b69 6e64 6578  elf.labels[index
+000177f0: 5d0a 2020 2020 2020 2020 6461 7461 203d  ].        data =
+00017800: 2073 656c 662e 6461 7461 5f61 7272 6179   self.data_array
+00017810: 735b 696e 6465 785d 0a20 2020 2020 2020  s[index].       
+00017820: 2072 6574 7572 6e20 7b27 6661 6d69 6c79   return {'family
+00017830: 5f6e 616d 6527 3a20 6c61 6265 6c2c 2027  _name': label, '
+00017840: 6661 6d69 6c79 5f64 6174 6127 3a20 6461  family_data': da
+00017850: 7461 7d0a 2020 2020 6465 6620 5f5f 6c65  ta}.    def __le
+00017860: 6e5f 5f28 7365 6c66 293a 0a20 2020 2020  n__(self):.     
+00017870: 2020 2072 6574 7572 6e20 6c65 6e28 7365     return len(se
+00017880: 6c66 2e6c 6162 656c 7329 0a64 6566 2064  lf.labels).def d
+00017890: 6566 696e 655f 6261 7463 685f 7369 7a65  efine_batch_size
+000178a0: 286e 2c62 6174 6368 5f73 697a 653d 5472  (n,batch_size=Tr
+000178b0: 7565 2c20 6265 6e63 686d 6172 6b69 6e67  ue, benchmarking
+000178c0: 3d46 616c 7365 293a 0a20 2020 2022 2222  =False):.    """
+000178d0: 4175 746f 6d61 7469 6320 6361 6c63 756c  Automatic calcul
+000178e0: 6174 696f 6e20 7468 6520 6176 6169 6c61  ation the availa
+000178f0: 626c 6520 6469 7669 736f 7273 206f 6620  ble divisors of 
+00017900: 7468 6520 6e75 6d62 6572 206f 6620 7472  the number of tr
+00017910: 6169 6e69 6e67 2064 6174 6120 706f 696e  aining data poin
+00017920: 7473 2028 6e29 2e0a 2020 2020 5468 6973  ts (n)..    This
+00017930: 2068 656c 7073 2074 6f20 7375 6767 6573   helps to sugges
+00017940: 7420 616e 2061 7070 726f 7069 6174 6520  t an appropiate 
+00017950: 696e 7465 6765 7220 6261 7463 6820 7369  integer batch si
+00017960: 7a65 206e 756d 6265 722c 2074 6861 7420  ze number, that 
+00017970: 7370 6c69 7473 2065 7665 6e6c 7920 7468  splits evenly th
+00017980: 6520 6461 7461 0a20 2020 203a 7061 7261  e data.    :para
+00017990: 6d20 696e 7420 6e3a 206e 756d 6265 7220  m int n: number 
+000179a0: 6f66 2073 6571 7565 6e63 6573 0a20 2020  of sequences.   
+000179b0: 203a 7061 7261 6d20 626f 6f6c 2062 6174   :param bool bat
+000179c0: 6368 5f73 697a 650a 2020 2020 3a70 6172  ch_size.    :par
+000179d0: 616d 2062 656e 6368 6d61 726b 696e 6722  am benchmarking"
+000179e0: 2222 0a20 2020 2069 6620 6e6f 7420 6265  "".    if not be
+000179f0: 6e63 686d 6172 6b69 6e67 3a0a 2020 2020  nchmarking:.    
+00017a00: 2020 2020 6173 7365 7274 206e 203e 3d20      assert n >= 
+00017a10: 3130 3020 2c22 4e6f 7420 776f 7274 6820  100 ,"Not worth 
+00017a20: 6261 7463 6869 6e67 2c20 6e75 6d62 6572  batching, number
+00017a30: 206f 6620 7365 7175 656e 6365 7320 6973   of sequences is
+00017a40: 203c 2031 3030 2022 0a20 2020 2064 6976   < 100 ".    div
+00017a50: 6973 6f72 7320 3d20 4472 6175 706e 6972  isors = Draupnir
+00017a60: 4d6f 6465 6c55 7469 6c73 2e70 7269 6e74  ModelUtils.print
+00017a70: 5f64 6976 6973 6f72 7328 6e29 0a20 2020  _divisors(n).   
+00017a80: 206e 5f64 6967 6974 7320 3d20 6c65 6e28   n_digits = len(
+00017a90: 6c69 7374 2873 7472 286e 2929 2920 2320  list(str(n))) # 
+00017aa0: 3130 3020 6861 7320 3320 6469 6769 7473  100 has 3 digits
+00017ab0: 2c20 3130 3030 2068 6173 2034 2064 6967  , 1000 has 4 dig
+00017ac0: 6974 732c 2031 3030 3030 2068 6173 2035  its, 10000 has 5
+00017ad0: 2064 6967 6974 7320 616e 6420 736f 206f   digits and so o
+00017ae0: 6e0a 2020 2020 6e5f 6368 756e 6b73 203d  n.    n_chunks =
+00017af0: 205b 6d69 6e28 6469 7669 736f 7273 5b6e   [min(divisors[n
+00017b00: 5f64 6967 6974 732d 313a 5d29 2069 6620  _digits-1:]) if 
+00017b10: 6c65 6e28 6469 7669 736f 7273 2920 3e20  len(divisors) > 
+00017b20: 6e5f 6469 6769 7473 2d31 2065 6c73 6520  n_digits-1 else 
+00017b30: 315d 5b30 5d20 2023 2073 6d61 6c6c 6573  1][0]  # smalles
+00017b40: 7420 6469 7669 736f 7220 7468 6174 2069  t divisor that i
+00017b50: 7320 6e6f 7420 312c 206f 7220 6a75 7374  s not 1, or just
+00017b60: 2075 7365 2031 202c 2077 6865 6e20 6e6f   use 1 , when no
+00017b70: 206f 7468 6572 2064 6976 6973 6f72 7320   other divisors 
+00017b80: 6172 6520 6176 6169 6c61 626c 6520 2870  are available (p
+00017b90: 7269 6d65 206e 756d 6265 7273 290a 2020  rime numbers).  
+00017ba0: 2020 6261 7463 6873 697a 6520 3d20 696e    batchsize = in
+00017bb0: 7428 4472 6175 706e 6972 4d6f 6465 6c55  t(DraupnirModelU
+00017bc0: 7469 6c73 2e69 6e74 6572 7661 6c73 286e  tils.intervals(n
+00017bd0: 5f63 6875 6e6b 732c 206e 295b 305d 5b31  _chunks, n)[0][1
+00017be0: 5d29 0a20 2020 2069 6620 6261 7463 685f  ]).    if batch_
+00017bf0: 7369 7a65 3a72 6574 7572 6e20 6261 7463  size:return batc
+00017c00: 6873 697a 650a 2020 2020 656c 7365 3a72  hsize.    else:r
+00017c10: 6574 7572 6e20 6e5f 6368 756e 6b73 0a64  eturn n_chunks.d
+00017c20: 6566 2063 6f76 6172 6961 6e63 6528 782c  ef covariance(x,
+00017c30: 7929 3a0a 2020 2020 2222 2243 6f6d 7075  y):.    """Compu
+00017c40: 7465 7320 6372 6f73 732d 636f 7661 7269  tes cross-covari
+00017c50: 616e 6365 2062 6574 7765 656e 2076 6563  ance between vec
+00017c60: 746f 7273 2c20 616e 6420 6973 2064 6566  tors, and is def
+00017c70: 696e 6564 2062 7920 636f 765b 582c 595d  ined by cov[X,Y]
+00017c80: 3d45 5b28 58e2 8892 cebc 5829 2859 e288  =E[(X.....X)(Y..
+00017c90: 92ce bc59 2954 5d22 2222 0a20 2020 2041  ...Y)T]""".    A
+00017ca0: 203d 2074 6f72 6368 2e73 7172 7428 746f   = torch.sqrt(to
+00017cb0: 7263 682e 6172 616e 6765 2831 3229 2e72  rch.arange(12).r
+00017cc0: 6573 6861 7065 2833 2c20 3429 2920 2023  eshape(3, 4))  #
+00017cd0: 2073 6f6d 6520 3320 6279 2034 2061 7272   some 3 by 4 arr
+00017ce0: 6179 0a20 2020 2062 203d 2074 6f72 6368  ay.    b = torch
+00017cf0: 2e74 656e 736f 7228 5b5b 325d 2c20 5b34  .tensor([[2], [4
+00017d00: 5d2c 205b 355d 5d29 2020 2320 736f 6d65  ], [5]])  # some
+00017d10: 2033 2062 7920 3120 7665 6374 6f72 0a20   3 by 1 vector. 
+00017d20: 2020 2063 6f76 203d 2074 6f72 6368 2e64     cov = torch.d
+00017d30: 6f74 2862 2e54 202d 2062 2e6d 6561 6e28  ot(b.T - b.mean(
+00017d40: 292c 2041 202d 2041 2e6d 6561 6e28 6469  ), A - A.mean(di
+00017d50: 6d3d 3029 2920 2f20 2862 2e73 6861 7065  m=0)) / (b.shape
+00017d60: 5b30 5d20 2d20 3129 0a20 2020 2072 6574  [0] - 1).    ret
+00017d70: 7572 6e20 636f 760a 6465 6620 6669 6e64  urn cov.def find
+00017d80: 5f6e 616e 2861 7272 6179 2c6e 616e 3d54  _nan(array,nan=T
+00017d90: 7275 6529 3a0a 2020 2020 2222 2244 726f  rue):.    """Dro
+00017da0: 7073 2072 6f77 7320 696e 2061 2074 656e  ps rows in a ten
+00017db0: 736f 7220 636f 6e74 6169 6e69 6e67 2061  sor containing a
+00017dc0: 6e79 206e 616e 2076 616c 7565 730a 2020  ny nan values.  
+00017dd0: 2020 3a70 6172 616d 2074 656e 736f 7220    :param tensor 
+00017de0: 6172 7261 7922 2222 0a20 2020 2073 6861  array""".    sha
+00017df0: 7065 203d 2061 7272 6179 2e73 6861 7065  pe = array.shape
+00017e00: 0a20 2020 2074 656e 736f 725f 7265 7368  .    tensor_resh
+00017e10: 6170 6564 203d 2061 7272 6179 2e72 6573  aped = array.res
+00017e20: 6861 7065 2873 6861 7065 5b30 5d2c 202d  hape(shape[0], -
+00017e30: 3129 0a20 2020 2023 2044 726f 7020 616c  1).    # Drop al
+00017e40: 6c20 726f 7773 2063 6f6e 7461 696e 696e  l rows containin
+00017e50: 6720 616e 7920 6e61 6e3a 0a20 2020 2069  g any nan:.    i
+00017e60: 6620 6e6f 7420 6e61 6e3a 0a20 2020 2020  f not nan:.     
+00017e70: 2020 2074 656e 736f 725f 7265 7368 6170     tensor_reshap
+00017e80: 6564 203d 2074 656e 736f 725f 7265 7368  ed = tensor_resh
+00017e90: 6170 6564 5b7e 746f 7263 682e 616e 7928  aped[~torch.any(
+00017ea0: 7465 6e73 6f72 5f72 6573 6861 7065 642e  tensor_reshaped.
+00017eb0: 6973 6e61 6e28 292c 2064 696d 3d31 295d  isnan(), dim=1)]
+00017ec0: 0a20 2020 2065 6c73 653a 0a20 2020 2020  .    else:.     
+00017ed0: 2020 2074 656e 736f 725f 7265 7368 6170     tensor_reshap
+00017ee0: 6564 203d 2074 656e 736f 725f 7265 7368  ed = tensor_resh
+00017ef0: 6170 6564 5b74 6f72 6368 2e61 6e79 2874  aped[torch.any(t
+00017f00: 656e 736f 725f 7265 7368 6170 6564 2e69  ensor_reshaped.i
+00017f10: 736e 616e 2829 2c20 6469 6d3d 3129 5d0a  snan(), dim=1)].
+00017f20: 2020 2020 2320 5265 7368 6170 6520 6261      # Reshape ba
+00017f30: 636b 3a0a 2020 2020 6172 7261 7920 3d20  ck:.    array = 
+00017f40: 7465 6e73 6f72 5f72 6573 6861 7065 642e  tensor_reshaped.
+00017f50: 7265 7368 6170 6528 7465 6e73 6f72 5f72  reshape(tensor_r
+00017f60: 6573 6861 7065 642e 7368 6170 655b 305d  eshaped.shape[0]
+00017f70: 2c20 2a73 6861 7065 5b31 3a5d 290a 2020  , *shape[1:]).  
+00017f80: 2020 7265 7475 726e 2061 7272 6179 0a64    return array.d
+00017f90: 6566 2072 616d 6163 6861 6e64 7261 6e5f  ef ramachandran_
+00017fa0: 706c 6f74 2820 6461 7461 5f61 6e67 6c65  plot( data_angle
+00017fb0: 732c 7361 7665 5f64 6972 6563 746f 7279  s,save_directory
+00017fc0: 2c70 6c6f 745f 7469 746c 652c 206f 6e65  ,plot_title, one
+00017fd0: 5f68 6f74 5f65 6e63 6f64 6564 203d 2046  _hot_encoded = F
+00017fe0: 616c 7365 2029 3a0a 2020 2020 2222 2250  alse ):.    """P
+00017ff0: 6c6f 7473 2052 616d 6163 6861 6e64 7261  lots Ramachandra
+00018000: 6e20 706c 6f74 7320 6265 7477 6565 6e20  n plots between 
+00018010: 7468 6520 7068 6920 616e 6420 7073 6920  the phi and psi 
+00018020: 616e 676c 6573 2069 6e20 7468 6520 6461  angles in the da
+00018030: 7461 7365 740a 2020 2020 3a70 6172 616d  taset.    :param
+00018040: 206e 756d 7079 2d61 7272 6179 2064 6174   numpy-array dat
+00018050: 615f 616e 676c 6573 203a 2033 6420 6172  a_angles : 3d ar
+00018060: 7261 7920 5b6e 5f73 6571 2c20 616c 6967  ray [n_seq, alig
+00018070: 6e5f 6c65 6e2c 2033 305d 2c20 7768 6572  n_len, 30], wher
+00018080: 6520 636f 6c75 6d6e 7320 3120 616e 6420  e columns 1 and 
+00018090: 3220 6f72 2032 3120 616e 6420 3232 2063  2 or 21 and 22 c
+000180a0: 6f6e 7461 696e 2074 6865 2061 6e67 6c65  ontain the angle
+000180b0: 7320 696e 2074 6865 2063 6173 6520 6f66  s in the case of
+000180c0: 204e 4f54 206f 6e65 2d68 6f74 2065 6e63   NOT one-hot enc
+000180d0: 6f64 6564 2061 6e64 206f 6e65 2d68 6f74  oded and one-hot
+000180e0: 2065 6e63 6f64 6564 2c20 7265 7370 6563   encoded, respec
+000180f0: 7469 7665 6c79 0a20 2020 203a 7061 7261  tively.    :para
+00018100: 6d20 7374 7220 7361 7665 5f64 6972 6563  m str save_direc
+00018110: 746f 7279 3a20 7061 7468 2074 6f20 7361  tory: path to sa
+00018120: 7669 6e67 2064 6972 6563 746f 7279 0a20  ving directory. 
+00018130: 2020 203a 7061 7261 6d20 7374 7220 706c     :param str pl
+00018140: 6f74 5f74 6974 6c65 0a20 2020 203a 7061  ot_title.    :pa
+00018150: 7261 6d20 626f 6f6c 206f 6e65 5f68 6f74  ram bool one_hot
+00018160: 5f65 6e63 6f64 6564 0a20 2020 2022 2222  _encoded.    """
+00018170: 0a20 2020 2070 6c74 2e63 6c66 2829 0a20  .    plt.clf(). 
+00018180: 2020 2069 6620 6f6e 655f 686f 745f 656e     if one_hot_en
+00018190: 636f 6465 643a 0a20 2020 2020 2020 2070  coded:.        p
+000181a0: 6869 203d 2064 6174 615f 616e 676c 6573  hi = data_angles
+000181b0: 5b3a 2c3a 2c32 315d 2e72 6573 6861 7065  [:,:,21].reshape
+000181c0: 2828 6461 7461 5f61 6e67 6c65 732e 7368  ((data_angles.sh
+000181d0: 6170 655b 305d 2a64 6174 615f 616e 676c  ape[0]*data_angl
+000181e0: 6573 2e73 6861 7065 5b31 5d29 292e 6173  es.shape[1])).as
+000181f0: 7479 7065 2866 6c6f 6174 2920 232d 206e  type(float) #- n
+00018200: 702e 7069 0a0a 2020 2020 2020 2020 7073  p.pi..        ps
+00018210: 6920 3d20 6461 7461 5f61 6e67 6c65 735b  i = data_angles[
+00018220: 3a2c 3a2c 3232 5d2e 7265 7368 6170 6528  :,:,22].reshape(
+00018230: 2864 6174 615f 616e 676c 6573 2e73 6861  (data_angles.sha
+00018240: 7065 5b30 5d2a 6461 7461 5f61 6e67 6c65  pe[0]*data_angle
+00018250: 732e 7368 6170 655b 315d 2929 2e61 7374  s.shape[1])).ast
+00018260: 7970 6528 666c 6f61 7429 2023 2d20 6e70  ype(float) #- np
+00018270: 2e70 690a 2020 2020 656c 7365 3a0a 2020  .pi.    else:.  
+00018280: 2020 2020 2020 7068 6920 3d20 6461 7461        phi = data
+00018290: 5f61 6e67 6c65 735b 3a2c 3a2c 315d 2e72  _angles[:,:,1].r
+000182a0: 6573 6861 7065 2828 6461 7461 5f61 6e67  eshape((data_ang
+000182b0: 6c65 732e 7368 6170 655b 305d 2a64 6174  les.shape[0]*dat
+000182c0: 615f 616e 676c 6573 2e73 6861 7065 5b31  a_angles.shape[1
+000182d0: 5d29 292e 6173 7479 7065 2866 6c6f 6174  ])).astype(float
+000182e0: 290a 2020 2020 2020 2020 7073 6920 3d20  ).        psi = 
+000182f0: 6461 7461 5f61 6e67 6c65 735b 3a2c 3a2c  data_angles[:,:,
+00018300: 325d 2e72 6573 6861 7065 2828 6461 7461  2].reshape((data
+00018310: 5f61 6e67 6c65 732e 7368 6170 655b 305d  _angles.shape[0]
+00018320: 2a64 6174 615f 616e 676c 6573 2e73 6861  *data_angles.sha
+00018330: 7065 5b31 5d29 292e 6173 7479 7065 2866  pe[1])).astype(f
+00018340: 6c6f 6174 290a 0a20 2020 2066 6967 203d  loat)..    fig =
+00018350: 2070 6c74 2e66 6967 7572 6528 6669 6773   plt.figure(figs
+00018360: 697a 653d 2835 2c35 2929 0a20 2020 2070  ize=(5,5)).    p
+00018370: 6c74 2e68 6973 7432 6428 2070 6869 2c20  lt.hist2d( phi, 
+00018380: 7073 692c 2062 696e 7320 3d20 3632 382c  psi, bins = 628,
+00018390: 206e 6f72 6d20 3d20 4c6f 674e 6f72 6d28   norm = LogNorm(
+000183a0: 2929 232c 2063 6d61 7020 3d20 706c 742e  ))#, cmap = plt.
+000183b0: 636d 2e6a 6574 2029 0a20 2020 2070 6c74  cm.jet ).    plt
+000183c0: 2e79 6c69 6d28 2d6e 702e 7069 2c20 6e70  .ylim(-np.pi, np
+000183d0: 2e70 6929 0a20 2020 2070 6c74 2e78 6c69  .pi).    plt.xli
+000183e0: 6d28 2d6e 702e 7069 2c20 6e70 2e70 6929  m(-np.pi, np.pi)
+000183f0: 0a20 2020 2070 6c74 2e74 6974 6c65 2870  .    plt.title(p
+00018400: 6c6f 745f 7469 746c 652c 666f 6e74 7369  lot_title,fontsi
+00018410: 7a65 3d31 3229 0a20 2020 2070 6c74 2e78  ze=12).    plt.x
+00018420: 6c61 6265 6c28 27cf 8627 290a 2020 2020  label('..').    
+00018430: 706c 742e 796c 6162 656c 2827 cf88 2729  plt.ylabel('..')
+00018440: 0a20 2020 2070 6c74 2e73 6176 6566 6967  .    plt.savefig
+00018450: 2873 6176 655f 6469 7265 6374 6f72 7929  (save_directory)
+00018460: 0a20 2020 2070 6c74 2e63 6c66 2829 0a20  .    plt.clf(). 
+00018470: 2020 2070 6c74 2e63 6c6f 7365 2866 6967     plt.close(fig
+00018480: 290a 6465 6620 7261 6d61 6368 616e 6472  ).def ramachandr
+00018490: 616e 5f70 6c6f 745f 7361 6d70 6c65 6428  an_plot_sampled(
+000184a0: 7068 692c 7073 692c 7361 7665 5f64 6972  phi,psi,save_dir
+000184b0: 6563 746f 7279 2c70 6c6f 745f 7469 746c  ectory,plot_titl
+000184c0: 652c 706c 6f74 5f6b 6170 7061 733d 4661  e,plot_kappas=Fa
+000184d0: 6c73 6529 3a0a 2020 2020 2222 2250 6c6f  lse):.    """Plo
+000184e0: 7473 2052 616d 6163 6861 6e64 7261 6e20  ts Ramachandran 
+000184f0: 706c 6f74 7320 6265 7477 6565 6e20 7468  plots between th
+00018500: 6520 7068 6920 616e 6420 7073 6920 616e  e phi and psi an
+00018510: 676c 6573 2069 6e20 7468 6520 6461 7461  gles in the data
+00018520: 7365 740a 2020 2020 3a70 6172 616d 2074  set.    :param t
+00018530: 656e 736f 7220 7068 690a 2020 2020 3a70  ensor phi.    :p
+00018540: 6172 616d 2074 656e 736f 7220 7073 690a  aram tensor psi.
+00018550: 2020 2020 3a70 6172 616d 2073 7472 2073      :param str s
+00018560: 6176 655f 6469 7265 6374 6f72 793a 2070  ave_directory: p
+00018570: 6174 6820 746f 2073 6176 696e 6720 6469  ath to saving di
+00018580: 7265 6374 6f72 790a 2020 2020 3a70 6172  rectory.    :par
+00018590: 616d 2073 7472 2070 6c6f 745f 7469 746c  am str plot_titl
+000185a0: 650a 2020 2020 3a70 6172 616d 2062 6f6f  e.    :param boo
+000185b0: 6c20 706c 6f74 5f6b 6170 7061 730a 2020  l plot_kappas.  
+000185c0: 2020 2222 220a 2020 2020 6966 2069 7369    """.    if isi
+000185d0: 6e73 7461 6e63 6528 7068 692c 746f 7263  nstance(phi,torc
+000185e0: 682e 5465 6e73 6f72 293a 0a20 2020 2020  h.Tensor):.     
+000185f0: 2020 2070 6869 203d 2070 6869 2e76 6965     phi = phi.vie
+00018600: 7728 2d31 292e 6370 7528 292e 6465 7461  w(-1).cpu().deta
+00018610: 6368 2829 2e6e 756d 7079 2829 0a20 2020  ch().numpy().   
+00018620: 2020 2020 2070 7369 203d 2070 7369 2e76       psi = psi.v
+00018630: 6965 7728 2d31 292e 6370 7528 292e 6465  iew(-1).cpu().de
+00018640: 7461 6368 2829 2e6e 756d 7079 2829 0a20  tach().numpy(). 
+00018650: 2020 2023 6178 6573 203d 205b 5b2d 6e70     #axes = [[-np
+00018660: 2e70 692c 206e 702e 7069 5d2c 205b 2d6e  .pi, np.pi], [-n
+00018670: 702e 7069 2c20 6e70 2e70 695d 5d0a 2020  p.pi, np.pi]].  
+00018680: 2020 706c 742e 6669 6775 7265 2866 6967    plt.figure(fig
+00018690: 7369 7a65 3d28 352c 3529 290a 2020 2020  size=(5,5)).    
+000186a0: 706c 742e 6869 7374 3264 2820 7068 692c  plt.hist2d( phi,
+000186b0: 2070 7369 2c20 6269 6e73 203d 2036 3238   psi, bins = 628
+000186c0: 2c20 6e6f 726d 203d 204c 6f67 4e6f 726d  , norm = LogNorm
+000186d0: 2829 2923 2c20 636d 6170 203d 2070 6c74  ())#, cmap = plt
+000186e0: 2e63 6d2e 6a65 7420 290a 2020 2020 6966  .cm.jet ).    if
+000186f0: 2070 6c6f 745f 6b61 7070 6173 3a0a 2020   plot_kappas:.  
+00018700: 2020 2020 2020 706c 742e 786c 6162 656c        plt.xlabel
+00018710: 2827 4b61 7070 6120 cf86 2729 0a20 2020  ('Kappa ..').   
+00018720: 2020 2020 2070 6c74 2e79 6c61 6265 6c28       plt.ylabel(
+00018730: 274b 6170 7061 20cf 8827 290a 2020 2020  'Kappa ..').    
+00018740: 656c 7365 3a0a 2020 2020 2020 2020 706c  else:.        pl
+00018750: 742e 796c 696d 282d 6e70 2e70 692c 206e  t.ylim(-np.pi, n
+00018760: 702e 7069 290a 2020 2020 2020 2020 706c  p.pi).        pl
+00018770: 742e 786c 696d 282d 6e70 2e70 692c 206e  t.xlim(-np.pi, n
+00018780: 702e 7069 290a 2020 2020 2020 2020 706c  p.pi).        pl
+00018790: 742e 786c 6162 656c 2827 cf86 2729 0a20  t.xlabel('..'). 
+000187a0: 2020 2020 2020 2070 6c74 2e79 6c61 6265         plt.ylabe
+000187b0: 6c28 27cf 8827 290a 2020 2020 706c 742e  l('..').    plt.
+000187c0: 7469 746c 6528 706c 6f74 5f74 6974 6c65  title(plot_title
+000187d0: 2c66 6f6e 7473 697a 653d 3132 290a 2020  ,fontsize=12).  
+000187e0: 2020 706c 742e 7361 7665 6669 6728 7361    plt.savefig(sa
+000187f0: 7665 5f64 6972 6563 746f 7279 290a 2020  ve_directory).  
+00018800: 2020 706c 742e 636c 6628 290a 2020 2020    plt.clf().    
+00018810: 706c 742e 636c 6f73 6528 290a 6465 6620  plt.close().def 
+00018820: 6772 6164 6965 6e74 735f 706c 6f74 2867  gradients_plot(g
+00018830: 7261 6469 656e 745f 6e6f 726d 732c 6570  radient_norms,ep
+00018840: 6f63 6873 2c64 6972 6563 746f 7279 293a  ochs,directory):
+00018850: 0a20 2020 2022 2222 5669 7375 616c 697a  .    """Visualiz
+00018860: 6174 696f 6e20 6f66 2074 6865 2067 7261  ation of the gra
+00018870: 6469 656e 7420 6465 7363 656e 7420 7065  dient descent pe
+00018880: 7220 6d6f 6465 6c20 7061 7261 6d65 7465  r model paramete
+00018890: 720a 2020 2020 3a70 6172 616d 2064 6963  r.    :param dic
+000188a0: 7420 6772 6164 6965 6e74 5f6e 6f72 6d73  t gradient_norms
+000188b0: 3a20 6469 6374 696f 6e61 7279 2077 6974  : dictionary wit
+000188c0: 6820 7468 6520 6d6f 6465 6c20 7061 7261  h the model para
+000188d0: 6d65 7465 7273 2061 6e64 2074 6865 2070  meters and the p
+000188e0: 6172 7469 616c 2064 6572 6976 6174 6976  artial derivativ
+000188f0: 6573 0a20 2020 203a 7061 7261 6d20 696e  es.    :param in
+00018900: 7420 6570 6f63 6873 2222 220a 0a20 2020  t epochs"""..   
+00018910: 2066 6967 203d 2070 6c74 2e66 6967 7572   fig = plt.figur
+00018920: 6528 6669 6773 697a 653d 2831 362c 2039  e(figsize=(16, 9
+00018930: 292c 2064 7069 3d31 3030 292e 7365 745f  ), dpi=100).set_
+00018940: 6661 6365 636f 6c6f 7228 2777 6869 7465  facecolor('white
+00018950: 2729 0a20 2020 2061 7820 3d20 706c 742e  ').    ax = plt.
+00018960: 7375 6270 6c6f 7428 3132 3129 0a20 2020  subplot(121).   
+00018970: 2063 6f6c 6f72 5f6d 6170 203d 2063 6d2e   color_map = cm.
+00018980: 7261 696e 626f 7728 6e70 2e6c 696e 7370  rainbow(np.linsp
+00018990: 6163 6528 302c 2031 2c20 3432 2929 0a20  ace(0, 1, 42)). 
+000189a0: 2020 2066 6f72 2028 6e61 6d65 5f69 2c20     for (name_i, 
+000189b0: 6772 6164 5f6e 6f72 6d73 292c 2063 6f6c  grad_norms), col
+000189c0: 6f72 2069 6e20 7a69 7028 6772 6164 6965  or in zip(gradie
+000189d0: 6e74 5f6e 6f72 6d73 2e69 7465 6d73 2829  nt_norms.items()
+000189e0: 2c20 636f 6c6f 725f 6d61 7029 3a0a 2020  , color_map):.  
+000189f0: 2020 2020 2020 6178 2e70 6c6f 7428 6772        ax.plot(gr
+00018a00: 6164 5f6e 6f72 6d73 2c20 6c61 6265 6c3d  ad_norms, label=
+00018a10: 6e61 6d65 5f69 2c20 633d 636f 6c6f 7229  name_i, c=color)
+00018a20: 0a20 2020 2070 6c74 2e78 6c61 6265 6c28  .    plt.xlabel(
+00018a30: 2769 7465 7273 2729 0a20 2020 2070 6c74  'iters').    plt
+00018a40: 2e79 6c61 6265 6c28 2767 7261 6469 656e  .ylabel('gradien
+00018a50: 7420 6e6f 726d 2729 0a20 2020 2070 6c74  t norm').    plt
+00018a60: 2e79 7363 616c 6528 276c 6f67 2729 0a20  .yscale('log'). 
+00018a70: 2020 2061 782e 6c65 6765 6e64 286c 6f63     ax.legend(loc
+00018a80: 3d27 7570 7065 7220 6365 6e74 6572 272c  ='upper center',
+00018a90: 2062 626f 785f 746f 5f61 6e63 686f 723d   bbox_to_anchor=
+00018aa0: 2831 2e35 2c20 3129 2c20 7368 6164 6f77  (1.5, 1), shadow
+00018ab0: 3d54 7275 652c 206e 636f 6c3d 312c 2070  =True, ncol=1, p
+00018ac0: 726f 703d 7b27 7369 7a65 273a 2038 7d29  rop={'size': 8})
+00018ad0: 0a20 2020 2070 6c74 2e74 6974 6c65 2827  .    plt.title('
+00018ae0: 4772 6164 6965 6e74 206e 6f72 6d73 2064  Gradient norms d
+00018af0: 7572 696e 6720 5356 4927 290a 2020 2020  uring SVI').    
+00018b00: 706c 742e 7361 7665 6669 6728 227b 7d2f  plt.savefig("{}/
+00018b10: 4772 6164 6965 6e74 735f 7b7d 5f65 706f  Gradients_{}_epo
+00018b20: 6368 732e 706e 6722 2e66 6f72 6d61 7428  chs.png".format(
+00018b30: 6469 7265 6374 6f72 792c 6570 6f63 6873  directory,epochs
+00018b40: 2929 0a20 2020 2070 6c74 2e63 6c66 2829  )).    plt.clf()
+00018b50: 0a20 2020 2070 6c74 2e63 6c6f 7365 2829  .    plt.close()
 00018b60: 0a64 6566 2072 656e 616d 696e 6728 7472  .def renaming(tr
 00018b70: 6565 293a 0a20 2020 2020 2020 2022 2222  ee):.        """
 00018b80: 5265 6e61 6d65 2074 6865 2069 6e74 6572  Rename the inter
 00018b90: 6e61 6c20 6e6f 6465 732c 2075 6e6c 6573  nal nodes, unles
 00018ba0: 7320 7468 6520 6769 7665 6e20 6e65 7769  s the given newi
 00018bb0: 636b 2066 696c 6520 616c 7265 6164 7920  ck file already 
 00018bc0: 6861 7320 7468 6520 6e61 6d65 7320 6f6e  has the names on
@@ -6355,428 +6355,489 @@
 00018d20: 2020 2020 2020 2020 2020 2020 206e 6f64               nod
 00018d30: 652e 6e61 6d65 203d 2022 4125 6422 2025  e.name = "A%d" %
 00018d40: 2065 6467 650a 2020 2020 2020 2020 2020   edge.          
 00018d50: 2020 2020 2020 696e 7465 726e 616c 5f6e        internal_n
 00018d60: 6f64 6573 5f6e 616d 6573 2e61 7070 656e  odes_names.appen
 00018d70: 6428 6e6f 6465 2e6e 616d 6529 0a20 2020  d(node.name).   
 00018d80: 2020 2020 2020 2020 2020 2020 2065 6467               edg
-00018d90: 6520 2b3d 2031 0a0a 6465 6620 6361 6c63  e += 1..def calc
-00018da0: 756c 6174 655f 6161 5f66 7265 7175 656e  ulate_aa_frequen
-00018db0: 6369 6573 2864 6174 6173 6574 2c66 7265  cies(dataset,fre
-00018dc0: 715f 6269 6e73 293a 0a20 2020 2022 2222  q_bins):.    """
-00018dd0: 4361 6c63 756c 6174 6573 2061 2066 7265  Calculates a fre
-00018de0: 7175 656e 6379 2066 6f72 2065 6163 6820  quency for each 
-00018df0: 6f66 2074 6865 2061 6120 2620 6761 7020  of the aa & gap 
-00018e00: 6174 2065 6163 6820 706f 7369 7469 6f6e  at each position
-00018e10: 2e54 6865 206e 756d 6265 7220 6f66 2062  .The number of b
-00018e20: 696e 7320 286f 6620 7369 7a65 2031 2920  ins (of size 1) 
-00018e30: 6973 206f 6e65 206c 6172 6765 7220 7468  is one larger th
-00018e40: 616e 2074 6865 206c 6172 6765 7374 2076  an the largest v
-00018e50: 616c 7565 2069 6e20 782e 2054 6869 7320  alue in x. This 
-00018e60: 6973 2064 6f6e 6520 666f 7220 6e75 6d70  is done for nump
-00018e70: 7920 6172 7261 7973 0a20 2020 203a 7061  y arrays.    :pa
-00018e80: 7261 6d20 7465 6e73 6f72 2064 6174 6173  ram tensor datas
-00018e90: 6574 0a20 2020 203a 7061 7261 6d20 696e  et.    :param in
-00018ea0: 7420 6672 6571 5f62 696e 730a 2020 2020  t freq_bins.    
-00018eb0: 2222 220a 2020 2020 6672 6571 7320 3d20  """.    freqs = 
-00018ec0: 6e70 2e61 7070 6c79 5f61 6c6f 6e67 5f61  np.apply_along_a
-00018ed0: 7869 7328 6c61 6d62 6461 2078 3a20 6e70  xis(lambda x: np
-00018ee0: 2e62 696e 636f 756e 7428 782c 206d 696e  .bincount(x, min
-00018ef0: 6c65 6e67 7468 3d66 7265 715f 6269 6e73  length=freq_bins
-00018f00: 292c 2061 7869 733d 302c 2061 7272 3d64  ), axis=0, arr=d
-00018f10: 6174 6173 6574 2e61 7374 7970 6528 2269  ataset.astype("i
-00018f20: 6e74 3634 2229 292e 540a 2020 2020 6672  nt64")).T.    fr
-00018f30: 6571 7320 3d20 6672 6571 732f 6461 7461  eqs = freqs/data
-00018f40: 7365 742e 7368 6170 655b 305d 0a20 2020  set.shape[0].   
-00018f50: 2072 6574 7572 6e20 6672 6571 730a 6465   return freqs.de
-00018f60: 6620 6361 6c63 756c 6174 655f 6161 5f66  f calculate_aa_f
-00018f70: 7265 7175 656e 6369 6573 5f74 6f72 6368  requencies_torch
-00018f80: 2864 6174 6173 6574 2c66 7265 715f 6269  (dataset,freq_bi
-00018f90: 6e73 293a 0a20 2020 2022 2222 4361 6c63  ns):.    """Calc
-00018fa0: 756c 6174 6573 2061 2066 7265 7175 656e  ulates a frequen
-00018fb0: 6379 2066 6f72 2065 6163 6820 6f66 2074  cy for each of t
-00018fc0: 6865 2061 6120 2620 6761 7020 6174 2065  he aa & gap at e
-00018fd0: 6163 6820 706f 7369 7469 6f6e 2e54 6865  ach position.The
-00018fe0: 206e 756d 6265 7220 6f66 2062 696e 7320   number of bins 
-00018ff0: 286f 6620 7369 7a65 2031 2920 6973 206f  (of size 1) is o
-00019000: 6e65 206c 6172 6765 7220 7468 616e 2074  ne larger than t
-00019010: 6865 206c 6172 6765 7374 2076 616c 7565  he largest value
-00019020: 2069 6e20 782e 2054 6869 7320 6973 2064   in x. This is d
-00019030: 6f6e 6520 666f 7220 746f 7263 6820 7465  one for torch te
-00019040: 6e73 6f72 730a 2020 2020 3a70 6172 616d  nsors.    :param
-00019050: 2074 656e 736f 7220 6461 7461 7365 740a   tensor dataset.
-00019060: 2020 2020 3a70 6172 616d 2069 6e74 2066      :param int f
-00019070: 7265 715f 6269 6e73 0a20 2020 2022 2222  req_bins.    """
-00019080: 0a20 2020 2066 7265 7173 203d 2074 6f72  .    freqs = tor
-00019090: 6368 2e73 7461 636b 285b 746f 7263 682e  ch.stack([torch.
-000190a0: 6269 6e63 6f75 6e74 2878 5f69 2c20 6d69  bincount(x_i, mi
-000190b0: 6e6c 656e 6774 683d 6672 6571 5f62 696e  nlength=freq_bin
-000190c0: 7329 2066 6f72 2069 2c20 785f 6920 696e  s) for i, x_i in
-000190d0: 2065 6e75 6d65 7261 7465 2874 6f72 6368   enumerate(torch
-000190e0: 2e75 6e62 696e 6428 6461 7461 7365 742e  .unbind(dataset.
-000190f0: 7479 7065 2874 6f72 6368 2e69 6e74 3634  type(torch.int64
-00019100: 292c 2064 696d 3d31 292c 2030 295d 2c20  ), dim=1), 0)], 
-00019110: 6469 6d3d 3129 0a20 2020 2066 7265 7173  dim=1).    freqs
-00019120: 203d 2066 7265 7173 2e54 0a20 2020 2066   = freqs.T.    f
-00019130: 7265 7173 203d 2066 7265 7173 202f 2064  reqs = freqs / d
-00019140: 6174 6173 6574 2e73 6861 7065 5b30 5d0a  ataset.shape[0].
-00019150: 2020 2020 7265 7475 726e 2066 7265 7173      return freqs
-00019160: 0a64 6566 2063 6f6d 7061 7265 5f74 7265  .def compare_tre
-00019170: 6573 2874 312c 7432 293a 0a20 2020 2022  es(t1,t2):.    "
-00019180: 2222 436f 6d70 7574 6573 2074 6865 2052  ""Computes the R
-00019190: 6620 6469 7374 616e 6365 2061 6d6f 6e67  f distance among
-000191a0: 2074 776f 2074 7265 652c 2069 7420 6361   two tree, it ca
-000191b0: 6c63 756c 6174 6573 2074 6865 206e 756d  lculates the num
-000191c0: 6265 7220 6f66 2063 6861 6e67 6573 2072  ber of changes r
-000191d0: 6571 7569 7265 7320 746f 2074 7261 6e73  equires to trans
-000191e0: 666f 726d 206f 6e65 2074 7265 6520 696e  form one tree in
-000191f0: 746f 2074 6865 206f 7468 6572 0a20 2020  to the other.   
-00019200: 203a 7061 7261 6d20 7374 7220 7431 3a20   :param str t1: 
-00019210: 7061 7468 2074 6f20 6669 7273 7420 6e65  path to first ne
-00019220: 7477 6963 6b20 7472 6565 0a20 2020 203a  twick tree.    :
-00019230: 7061 7261 6d20 7374 7220 7432 3a20 7061  param str t2: pa
-00019240: 7468 2074 6f20 7365 636f 6e64 206e 6574  th to second net
-00019250: 7769 636b 2074 7265 6522 2222 0a20 2020  wick tree""".   
-00019260: 2063 6f6c 756d 6e73 203d 205b 2252 4620   columns = ["RF 
-00019270: 6469 7374 616e 6365 222c 224d 6178 696d  distance","Maxim
-00019280: 756d 2052 4620 6469 7374 616e 6365 222c  um RF distance",
-00019290: 2243 6f6d 6d6f 6e20 6c65 6176 6573 222c  "Common leaves",
-000192a0: 2250 6172 7469 7469 6f6e 7349 6e5f 7431  "PartitionsIn_t1
-000192b0: 5f4e 4f54 5f74 3222 2c22 5061 7274 6974  _NOT_t2","Partit
-000192c0: 696f 6e73 496e 5f74 325f 4e4f 545f 7431  ionsIn_t2_NOT_t1
-000192d0: 222c 2244 6973 6361 7264 6564 5f70 6172  ","Discarded_par
-000192e0: 7469 7469 6f6e 735f 7431 222c 2244 6973  titions_t1","Dis
-000192f0: 6361 7264 6564 5f70 6172 7469 7469 6f6e  carded_partition
-00019300: 735f 7432 225d 0a20 2020 2023 2074 3120  s_t2"].    # t1 
-00019310: 203d 2054 7265 6545 7465 3328 7431 290a   = TreeEte3(t1).
-00019320: 2020 2020 2320 7432 203d 2054 7265 6545      # t2 = TreeE
-00019330: 7465 3328 7432 290a 2020 2020 2320 7265  te3(t2).    # re
-00019340: 7375 6c74 7320 3d20 7431 2e72 6f62 696e  sults = t1.robin
-00019350: 736f 6e5f 666f 756c 6473 2874 322c 756e  son_foulds(t2,un
-00019360: 726f 6f74 6564 5f74 7265 6573 3d54 7275  rooted_trees=Tru
-00019370: 6529 0a20 2020 2063 6f6d 6d61 6e64 203d  e).    command =
-00019380: 2027 5273 6372 6970 7427 0a20 2020 2070   'Rscript'.    p
-00019390: 6174 6832 7363 7269 7074 203d 2027 2f68  ath2script = '/h
-000193a0: 6f6d 652f 6c79 732f 4472 6f70 626f 782f  ome/lys/Dropbox/
-000193b0: 5068 442f 4452 4155 504e 4952 2f54 7265  PhD/DRAUPNIR/Tre
-000193c0: 655f 4469 7374 616e 6365 2e52 270a 2020  e_Distance.R'.  
-000193d0: 2020 2320 4275 696c 6420 7375 6270 726f    # Build subpro
-000193e0: 6365 7373 2063 6f6d 6d61 6e64 0a20 2020  cess command.   
-000193f0: 2061 7267 7320 3d5b 7431 2c74 325d 0a20   args =[t1,t2]. 
-00019400: 2020 2063 6d64 203d 205b 636f 6d6d 616e     cmd = [comman
-00019410: 642c 2070 6174 6832 7363 7269 7074 5d20  d, path2script] 
-00019420: 2b20 6172 6773 0a20 2020 2064 6973 7461  + args.    dista
-00019430: 6e63 6520 3d20 7375 6270 726f 6365 7373  nce = subprocess
-00019440: 2e63 6865 636b 5f6f 7574 7075 7428 636d  .check_output(cm
-00019450: 642c 2075 6e69 7665 7273 616c 5f6e 6577  d, universal_new
-00019460: 6c69 6e65 733d 5472 7565 290a 2020 2020  lines=True).    
-00019470: 7265 7475 726e 2064 6973 7461 6e63 650a  return distance.
-00019480: 0a64 6566 2072 656d 6f76 655f 7374 6f70  .def remove_stop
-00019490: 5f63 6f64 6f6e 7328 7365 7175 656e 6365  _codons(sequence
-000194a0: 5f66 696c 652c 6973 5f70 726f 743d 4661  _file,is_prot=Fa
-000194b0: 6c73 6529 3a0a 2020 2020 2222 2252 656d  lse):.    """Rem
-000194c0: 6f76 6520 7468 6520 7374 6f70 2063 6f64  ove the stop cod
-000194d0: 6f6e 7328 2a29 2066 726f 6d20 616e 2061  ons(*) from an a
-000194e0: 6c69 676e 6d65 6e74 2066 696c 6522 2222  lignment file"""
-000194f0: 0a20 2020 2069 6620 6973 5f70 726f 743a  .    if is_prot:
-00019500: 0a20 2020 2020 2020 2070 7269 6e74 2822  .        print("
-00019510: 5265 6d6f 7669 6e67 2073 746f 7020 636f  Removing stop co
-00019520: 646f 6e73 2066 726f 6d20 7072 6f74 6569  dons from protei
-00019530: 6e22 290a 2020 2020 2020 2020 7072 696e  n").        prin
-00019540: 7428 7365 7175 656e 6365 5f66 696c 6529  t(sequence_file)
-00019550: 0a20 2020 2020 2020 2073 6571 5f66 696c  .        seq_fil
-00019560: 6520 3d20 6f70 656e 2873 6571 7565 6e63  e = open(sequenc
-00019570: 655f 6669 6c65 2c20 2772 2b27 290a 2020  e_file, 'r+').  
-00019580: 2020 2020 2020 7365 7120 3d20 7365 715f        seq = seq_
-00019590: 6669 6c65 2e72 6561 6428 290a 2020 2020  file.read().    
-000195a0: 2020 2020 7365 715f 6669 6c65 2e73 6565      seq_file.see
-000195b0: 6b28 3029 0a20 2020 2020 2020 206e 6f5f  k(0).        no_
-000195c0: 636f 646f 6e73 203d 205b 6920 6966 2069  codons = [i if i
-000195d0: 213d 222a 2220 656c 7365 2022 2d22 2066  !="*" else "-" f
-000195e0: 6f72 2069 2069 6e20 7365 715d 2023 7265  or i in seq] #re
-000195f0: 706c 6163 6520 666f 7220 6120 6761 700a  place for a gap.
-00019600: 2020 2020 2020 2020 2354 4f44 4f3a 2054          #TODO: T
-00019610: 7275 6e63 6174 6520 7468 6520 7365 7175  runcate the sequ
-00019620: 656e 6365 2069 6620 7468 6572 6520 6973  ence if there is
-00019630: 2061 2073 746f 7020 636f 646f 6e20 696e   a stop codon in
-00019640: 2074 6865 206d 6964 646c 6520 2873 746f   the middle (sto
-00019650: 7020 6164 6469 6e67 2067 6170 7320 7768  p adding gaps wh
-00019660: 656e 205c 6e20 6973 2072 6561 6368 6564  en \n is reached
-00019670: 290a 2020 2020 2020 2020 2320 6e6f 5f63  ).        # no_c
-00019680: 6f64 6f6e 733d 205b 5d0a 2020 2020 2020  odons= [].      
-00019690: 2020 2320 666f 7220 696e 6465 782c 6920    # for index,i 
-000196a0: 696e 2065 6e75 6d65 7261 7465 2873 6571  in enumerate(seq
-000196b0: 293a 0a20 2020 2020 2020 2023 2020 2020  ):.        #    
-000196c0: 2069 6620 6920 3d3d 2022 2a22 3a0a 2020   if i == "*":.  
-000196d0: 2020 2020 2020 2320 2020 2020 2020 2020        #         
-000196e0: 693d 2022 2d22 0a20 2020 2020 2020 2023  i= "-".        #
-000196f0: 2020 2020 2020 2020 206e 6578 745f 6272           next_br
-00019700: 6561 6b5f 696e 6465 7820 3d20 7365 715b  eak_index = seq[
-00019710: 696e 6465 783a 5d2e 696e 6465 7828 225c  index:].index("\
-00019720: 6e22 290a 2020 2020 2020 2020 2320 2020  n").        #   
-00019730: 2020 2020 2020 7365 715b 696e 6465 783a        seq[index:
-00019740: 6e65 7874 5f62 7265 616b 5f69 6e64 6578  next_break_index
-00019750: 5d20 3d20 222d 222a 286e 6578 745f 6272  ] = "-"*(next_br
-00019760: 6561 6b5f 696e 6465 782d 696e 6465 7829  eak_index-index)
-00019770: 0a0a 2020 2020 2020 2020 7365 715f 6669  ..        seq_fi
-00019780: 6c65 2e77 7269 7465 2822 222e 6a6f 696e  le.write("".join
-00019790: 286e 6f5f 636f 646f 6e73 2929 0a20 2020  (no_codons)).   
-000197a0: 2020 2020 2073 6571 5f66 696c 652e 7472       seq_file.tr
-000197b0: 756e 6361 7465 2829 2020 2320 7265 6d6f  uncate()  # remo
-000197c0: 7665 2063 6f6e 7465 6e74 730a 2020 2020  ve contents.    
-000197d0: 656c 7365 3a0a 2020 2020 2020 2020 7374  else:.        st
-000197e0: 6f70 5f63 6f64 6f6e 7320 3d20 5b22 5447  op_codons = ["TG
-000197f0: 4122 2c22 5441 4722 2c22 5441 4122 5d0a  A","TAG","TAA"].
-00019800: 2020 2020 2020 2020 7365 715f 6669 6c65          seq_file
-00019810: 203d 206f 7065 6e28 7365 7175 656e 6365   = open(sequence
-00019820: 5f66 696c 652c 2027 722b 2729 0a20 2020  _file, 'r+').   
-00019830: 2020 2020 2073 6571 203d 2073 6571 5f66       seq = seq_f
-00019840: 696c 652e 7265 6164 2829 0a20 2020 2020  ile.read().     
-00019850: 2020 2073 6571 5f66 696c 652e 7365 656b     seq_file.seek
-00019860: 2830 290a 2020 2020 2020 2020 2363 6f64  (0).        #cod
-00019870: 696e 675f 646e 6120 3d20 5365 7128 7365  ing_dna = Seq(se
-00019880: 7129 0a20 2020 2020 2020 2023 7072 6f74  q).        #prot
-00019890: 6569 6e20 3d20 636f 6469 6e67 5f64 6e61  ein = coding_dna
-000198a0: 2e74 7261 6e73 6c61 7465 2829 0a20 2020  .translate().   
-000198b0: 2020 2020 2063 6f64 6f6e 7320 3d20 5b73       codons = [s
-000198c0: 6571 5b69 3a69 2b33 5d20 666f 7220 6920  eq[i:i+3] for i 
-000198d0: 696e 2072 616e 6765 2830 2c20 6c65 6e28  in range(0, len(
-000198e0: 7365 7129 2c20 3329 5d0a 2020 2020 2020  seq), 3)].      
-000198f0: 2020 636f 646f 6e73 203d 5b63 6f64 2066    codons =[cod f
-00019900: 6f72 2063 6f64 2069 6e20 636f 646f 6e73  or cod in codons
-00019910: 2069 6620 6c65 6e28 636f 6429 203d 3d33   if len(cod) ==3
-00019920: 2061 6e64 2063 6f64 206e 6f74 2069 6e20   and cod not in 
-00019930: 7374 6f70 5f63 6f64 6f6e 735d 0a20 2020  stop_codons].   
-00019940: 2020 2020 2023 6368 6563 6b20 3d20 616e       #check = an
-00019950: 7928 6974 656d 2069 6e20 636f 646f 6e73  y(item in codons
-00019960: 2066 6f72 2069 7465 6d20 696e 2073 746f   for item in sto
-00019970: 705f 636f 646f 6e73 290a 2020 2020 2020  p_codons).      
-00019980: 2020 7365 715f 6669 6c65 2e77 7269 7465    seq_file.write
-00019990: 2822 222e 6a6f 696e 2863 6f64 6f6e 7329  ("".join(codons)
-000199a0: 290a 2020 2020 2020 2020 7365 715f 6669  ).        seq_fi
-000199b0: 6c65 2e74 7275 6e63 6174 6528 2920 2023  le.truncate()  #
-000199c0: 2072 656d 6f76 6520 636f 6e74 656e 7473   remove contents
-000199d0: 0a0a 6465 6620 636f 6e76 6572 745f 746f  ..def convert_to
-000199e0: 5f69 6e74 6567 6572 7328 6461 7461 7365  _integers(datase
-000199f0: 742c 6161 5f70 726f 6273 2c61 7869 7329  t,aa_probs,axis)
-00019a00: 3a0a 2020 2020 2222 2254 7261 6e73 666f  :.    """Transfo
-00019a10: 726d 7320 6f6e 6520 686f 7420 656e 636f  rms one hot enco
-00019a20: 6465 6420 616d 696e 6f20 6163 6964 7320  ded amino acids 
-00019a30: 696e 746f 2030 2d69 6e64 6578 6564 2069  into 0-indexed i
-00019a40: 6e74 6567 6572 7320 692e 6520 5b31 2c30  ntegers i.e [1,0
-00019a50: 2c30 2c30 5d20 2d2d 3e20 3020 205b 302c  ,0,0] --> 0  [0,
-00019a60: 312c 302c 305d 202d 3e20 310a 2020 2020  1,0,0] -> 1.    
-00019a70: 3a70 6172 616d 206e 756d 7079 2064 6174  :param numpy dat
-00019a80: 6173 6574 0a20 2020 203a 7061 7261 6d20  aset.    :param 
-00019a90: 6161 5f70 726f 6273 3a20 616d 696e 6f20  aa_probs: amino 
-00019aa0: 6163 6964 2070 726f 6261 6269 6c69 7469  acid probabiliti
-00019ab0: 6573 2c20 6469 6d65 6e73 696f 6e73 206f  es, dimensions o
-00019ac0: 6620 7468 6520 6f6e 652d 686f 7420 656e  f the one-hot en
-00019ad0: 636f 6469 6e67 2222 220a 2020 2020 6966  coding""".    if
-00019ae0: 2061 7869 733d 3d33 3a23 7573 6520 666f   axis==3:#use fo
-00019af0: 7220 7072 6564 6963 7469 6f6e 730a 2020  r predictions.  
-00019b00: 2020 2020 2020 6220 3d20 746f 7263 682e        b = torch.
-00019b10: 6172 676d 6178 2864 6174 6173 6574 2c20  argmax(dataset, 
-00019b20: 6469 6d3d 6178 6973 290a 2020 2020 656c  dim=axis).    el
-00019b30: 7365 3a0a 2020 2020 2020 2020 696e 7465  se:.        inte
-00019b40: 6765 7273 203d 2074 6f72 6368 2e61 7267  gers = torch.arg
-00019b50: 6d61 7828 6461 7461 7365 745b 3a2c 323a  max(dataset[:,2:
-00019b60: 2c30 3a61 615f 7072 6f62 735d 2c64 696d  ,0:aa_probs],dim
-00019b70: 3d61 7869 7329 0a20 2020 2020 2020 2062  =axis).        b
-00019b80: 203d 2074 6f72 6368 2e7a 6572 6f73 2864   = torch.zeros(d
-00019b90: 6174 6173 6574 5b3a 2c3a 2c30 5d2e 7368  ataset[:,:,0].sh
-00019ba0: 6170 6520 2b20 2833 302c 2929 2e63 7075  ape + (30,)).cpu
-00019bb0: 2829 0a20 2020 2020 2020 2023 6220 3d20  ().        #b = 
-00019bc0: 6e70 2e7a 6572 6f73 2844 6174 6173 6574  np.zeros(Dataset
-00019bd0: 5b3a 2c3a 2c30 5d2e 7368 6170 6520 2b20  [:,:,0].shape + 
-00019be0: 2833 302c 2929 0a20 2020 2020 2020 2062  (30,)).        b
-00019bf0: 5b3a 2c20 323a 2c20 305d 203d 2069 6e74  [:, 2:, 0] = int
-00019c00: 6567 6572 730a 2020 2020 2020 2020 625b  egers.        b[
-00019c10: 3a2c 3a32 5d20 3d20 6461 7461 7365 745b  :,:2] = dataset[
-00019c20: 3a2c 3a32 5d0a 2020 2020 7265 7475 726e  :,:2].    return
-00019c30: 2062 0a64 6566 2070 726f 6365 7373 5f62   b.def process_b
-00019c40: 6c6f 7375 6d28 626c 6f73 756d 2c61 615f  losum(blosum,aa_
-00019c50: 6672 6571 732c 616c 6967 6e5f 7365 715f  freqs,align_seq_
-00019c60: 6c65 6e2c 6161 5f70 726f 6273 293a 0a20  len,aa_probs):. 
-00019c70: 2020 2022 2222 0a20 2020 2043 6f6d 7075     """.    Compu
-00019c80: 7465 7320 7468 6520 6d61 7472 6963 6573  tes the matrices
-00019c90: 2072 6571 7569 7265 6420 746f 2062 7569   required to bui
-00019ca0: 6c64 2061 2062 6c6f 7375 6d20 656d 6265  ld a blosum embe
-00019cb0: 6464 696e 670a 2020 2020 3a70 6172 616d  dding.    :param
-00019cc0: 2074 656e 736f 7220 626c 6f73 756d 3a20   tensor blosum: 
-00019cd0: 424c 4f53 554d 206c 696b 656c 6968 6f6f  BLOSUM likelihoo
-00019ce0: 6420 2073 636f 7265 730a 2020 2020 3a70  d  scores.    :p
-00019cf0: 6172 616d 2074 656e 736f 7220 6161 5f66  aram tensor aa_f
-00019d00: 7265 7173 203a 2061 6d69 6e6f 2061 6369  reqs : amino aci
-00019d10: 6420 6672 6571 7565 6e63 6965 7320 7065  d frequencies pe
-00019d20: 7220 706f 7369 7469 6f6e 0a20 2020 203a  r position.    :
-00019d30: 7061 7261 6d20 616c 6967 6e5f 7365 715f  param align_seq_
-00019d40: 6c65 6e3a 2061 6c69 676e 6d65 6e74 206c  len: alignment l
-00019d50: 656e 6774 680a 2020 2020 3a70 6172 616d  ength.    :param
-00019d60: 2061 615f 7072 6f62 733a 2061 6d69 6e6f   aa_probs: amino
-00019d70: 2061 6369 6420 7072 6f62 6162 696c 6974   acid probabilit
-00019d80: 6965 732c 2074 7970 6573 206f 6620 616d  ies, types of am
-00019d90: 696e 6f20 6163 6964 7320 696e 2074 6865  ino acids in the
-00019da0: 2061 6c69 676e 6d65 6e74 0a20 2020 203a   alignment.    :
-00019db0: 6f75 7420 7465 6e73 6f72 2062 6c6f 7375  out tensor blosu
-00019dc0: 6d5f 6d61 7820 5b61 6c69 676e 5f6c 656e  m_max [align_len
-00019dd0: 2c61 615f 7072 6f62 5d3a 2062 6c6f 7375  ,aa_prob]: blosu
-00019de0: 6d20 6c69 6b65 6c69 686f 6f64 2073 636f  m likelihood sco
-00019df0: 7265 7320 666f 7220 7468 6520 6d6f 7374  res for the most
-00019e00: 2066 7265 7175 656e 7420 6161 2069 6e20   frequent aa in 
-00019e10: 7468 6520 616c 6967 6e6d 656e 7420 706f  the alignment po
-00019e20: 7369 7469 6f6e 0a20 2020 203a 6f75 7420  sition.    :out 
-00019e30: 7465 6e73 6f72 2062 6c6f 7375 6d5f 7765  tensor blosum_we
-00019e40: 6967 6874 6564 205b 616c 6967 6e5f 6c65  ighted [align_le
-00019e50: 6e2c 6161 5f70 726f 623a 2077 6569 6768  n,aa_prob: weigh
-00019e60: 7465 6420 6176 6572 6167 6520 6f66 2062  ted average of b
-00019e70: 6c6f 7375 6d20 6c69 6b65 6c69 686f 6f64  losum likelihood
-00019e80: 7320 6163 636f 7264 696e 6720 746f 2074  s according to t
-00019e90: 6865 2061 6120 6672 6571 7565 6e63 790a  he aa frequency.
-00019ea0: 2020 2020 3a6f 7574 2076 6172 6961 626c      :out variabl
-00019eb0: 655f 636f 7265 3a20 5b61 6c69 676e 5f6c  e_core: [align_l
-00019ec0: 656e 5d20 3a20 636f 756e 7473 2074 6865  en] : counts the
-00019ed0: 206e 756d 6265 7220 6f66 2064 6966 6665   number of diffe
-00019ee0: 7265 6e74 2065 6c65 6d65 6e74 7320 2861  rent elements (a
-00019ef0: 6d69 6e6f 2061 6369 6420 6469 7665 7273  mino acid divers
-00019f00: 6974 7929 2070 6572 2061 6c69 676e 6d65  ity) per alignme
-00019f10: 6e74 2070 6f73 6974 696f 6e22 2222 0a0a  nt position"""..
-00019f20: 2020 2020 6161 5f66 7265 7173 5f6d 6178      aa_freqs_max
-00019f30: 203d 2074 6f72 6368 2e61 7267 6d61 7828   = torch.argmax(
-00019f40: 6161 5f66 7265 7173 2c20 6469 6d3d 3129  aa_freqs, dim=1)
-00019f50: 2e72 6570 6561 7428 6161 5f70 726f 6273  .repeat(aa_probs
-00019f60: 2c20 3129 2e70 6572 6d75 7465 2831 2c20  , 1).permute(1, 
-00019f70: 3029 2023 5b6d 6178 5f6c 656e 2c20 6161  0) #[max_len, aa
-00019f80: 5f70 726f 6273 5d0a 2020 2020 626c 6f73  _probs].    blos
-00019f90: 756d 5f65 7870 616e 6465 6420 3d20 626c  um_expanded = bl
-00019fa0: 6f73 756d 5b31 3a2c 2031 3a5d 2e72 6570  osum[1:, 1:].rep
-00019fb0: 6561 7428 616c 6967 6e5f 7365 715f 6c65  eat(align_seq_le
-00019fc0: 6e2c 2031 2c20 3129 2020 2320 5b6d 6178  n, 1, 1)  # [max
-00019fd0: 5f6c 656e 2c61 615f 7072 6f62 732c 6161  _len,aa_probs,aa
-00019fe0: 5f70 726f 6273 5d0a 2020 2020 626c 6f73  _probs].    blos
-00019ff0: 756d 5f6d 6178 203d 2062 6c6f 7375 6d5f  um_max = blosum_
-0001a000: 6578 7061 6e64 6564 2e67 6174 6865 7228  expanded.gather(
-0001a010: 312c 2061 615f 6672 6571 735f 6d61 782e  1, aa_freqs_max.
-0001a020: 756e 7371 7565 657a 6528 3129 292e 7371  unsqueeze(1)).sq
-0001a030: 7565 657a 6528 3129 2020 2320 5b61 6c69  ueeze(1)  # [ali
-0001a040: 676e 5f73 6571 5f6c 656e 2c32 315d 2053  gn_seq_len,21] S
-0001a050: 6565 6d73 2063 6f72 7265 6374 0a0a 2020  eems correct..  
-0001a060: 2020 626c 6f73 756d 5f77 6569 6768 7465    blosum_weighte
-0001a070: 6420 3d20 6161 5f66 7265 7173 5b3a 2c3a  d = aa_freqs[:,:
-0001a080: 2c4e 6f6e 655d 2a62 6c6f 7375 6d5f 6578  ,None]*blosum_ex
-0001a090: 7061 6e64 6564 2023 2d2d 3e20 7265 706c  panded #--> repl
-0001a0a0: 6163 6520 3020 7769 7468 206e 616e 733f  ace 0 with nans?
-0001a0b0: 206f 7468 6572 7769 7365 2074 6865 2030   otherwise the 0
-0001a0c0: 2061 7265 2069 6e20 7468 6520 6d65 616e   are in the mean
-0001a0d0: 2061 7320 7765 6c6c 2e2e 2e2e 0a20 2020   as well.....   
-0001a0e0: 2062 6c6f 7375 6d5f 7765 6967 6874 6564   blosum_weighted
-0001a0f0: 203d 2062 6c6f 7375 6d5f 7765 6967 6874   = blosum_weight
-0001a100: 6564 2e6d 6561 6e28 6469 6d3d 3129 0a0a  ed.mean(dim=1)..
-0001a110: 2020 2020 7661 7269 6162 6c65 5f73 636f      variable_sco
-0001a120: 7265 203d 2074 6f72 6368 2e63 6f75 6e74  re = torch.count
-0001a130: 5f6e 6f6e 7a65 726f 2861 615f 6672 6571  _nonzero(aa_freq
-0001a140: 732c 2064 696d 3d31 292f 6161 5f70 726f  s, dim=1)/aa_pro
-0001a150: 6273 2023 6869 6768 6572 2073 636f 7265  bs #higher score
-0001a160: 2c20 6d6f 7265 2076 6172 6961 626c 650a  , more variable.
-0001a170: 0a20 2020 2072 6574 7572 6e20 626c 6f73  .    return blos
-0001a180: 756d 5f6d 6178 2c62 6c6f 7375 6d5f 7765  um_max,blosum_we
-0001a190: 6967 6874 6564 2c20 7661 7269 6162 6c65  ighted, variable
-0001a1a0: 5f73 636f 7265 0a64 6566 2062 6c6f 7375  _score.def blosu
-0001a1b0: 6d5f 656d 6265 6464 696e 675f 656e 636f  m_embedding_enco
-0001a1c0: 6465 7228 626c 6f73 756d 2c61 615f 6672  der(blosum,aa_fr
-0001a1d0: 6571 732c 616c 6967 6e5f 7365 715f 6c65  eqs,align_seq_le
-0001a1e0: 6e2c 6161 5f70 726f 6273 2c64 6174 6173  n,aa_probs,datas
-0001a1f0: 6574 5f74 7261 696e 2c20 6f6e 655f 686f  et_train, one_ho
-0001a200: 745f 656e 636f 6469 6e67 293a 0a20 2020  t_encoding):.   
-0001a210: 2022 2222 5472 616e 7366 6f72 6d73 2074   """Transforms t
-0001a220: 6865 2074 7261 696e 2064 6174 6173 6574  he train dataset
-0001a230: 2c20 7768 6963 6820 6973 2066 6f72 6d65  , which is forme
-0001a240: 6420 6279 2061 2074 656e 736f 7220 6f66  d by a tensor of
-0001a250: 2069 6e74 6567 6572 7320 7468 6174 2072   integers that r
-0001a260: 6570 7265 7365 6e74 2074 6865 2061 6d69  epresent the ami
-0001a270: 6e6f 2061 6369 6473 2c20 696e 746f 2061  no acids, into a
-0001a280: 2064 6174 6173 6574 2077 6865 7265 2065   dataset where e
-0001a290: 6163 6820 616d 696e 6f20 6163 6964 2028  ach amino acid (
-0001a2a0: 696e 7465 6765 7229 2069 7320 6578 6368  integer) is exch
-0001a2b0: 616e 6765 6420 7769 7468 2069 7473 2062  anged with its b
-0001a2c0: 6c6f 7375 6d20 7363 6f72 650a 2020 2020  losum score.    
-0001a2d0: 3a6f 7574 2074 656e 736f 7220 6161 5f74  :out tensor aa_t
-0001a2e0: 7261 696e 5f62 6c6f 7375 6d20 3a20 5472  rain_blosum : Tr
-0001a2f0: 6169 6e69 6e67 2064 6174 6173 6574 2077  aining dataset w
-0001a300: 6974 6820 7468 6520 626c 6f73 756d 2076  ith the blosum v
-0001a310: 6563 746f 7273 2069 6e73 7465 6164 206f  ectors instead o
-0001a320: 6620 7468 6520 616d 696e 6f20 6163 6964  f the amino acid
-0001a330: 7320 286e 756d 6265 7273 206f 7220 6f6e  s (numbers or on
-0001a340: 6520 686f 7420 7265 7072 6573 656e 7461  e hot representa
-0001a350: 7469 6f6e 2922 2222 0a20 2020 2069 6620  tion)""".    if 
-0001a360: 6f6e 655f 686f 745f 656e 636f 6469 6e67  one_hot_encoding
-0001a370: 3a0a 2020 2020 2020 2020 6461 7461 7365  :.        datase
-0001a380: 745f 7472 6169 6e20 3d20 636f 6e76 6572  t_train = conver
-0001a390: 745f 746f 5f69 6e74 6567 6572 7328 6461  t_to_integers(da
-0001a3a0: 7461 7365 745f 7472 6169 6e2c 6161 5f70  taset_train,aa_p
-0001a3b0: 726f 6273 2c61 7869 733d 3229 0a20 2020  robs,axis=2).   
-0001a3c0: 2061 6d69 6e6f 6163 6964 735f 7365 7173   aminoacids_seqs
-0001a3d0: 203d 2064 6174 6173 6574 5f74 7261 696e   = dataset_train
-0001a3e0: 5b3a 2c32 3a2c 305d 2e72 6570 6561 7428  [:,2:,0].repeat(
-0001a3f0: 6161 5f70 726f 6273 2c31 2c31 292e 7065  aa_probs,1,1).pe
-0001a400: 726d 7574 6528 312c 322c 3029 2023 5b4e  rmute(1,2,0) #[N
-0001a410: 2c6d 6178 5f6c 656e 2c61 6120 7265 7065  ,max_len,aa repe
-0001a420: 6174 6564 2061 615f 7072 6f62 7320 7469  ated aa_probs ti
-0001a430: 6d65 735d 2d2d 7365 656d 7320 636f 7272  mes]--seems corr
-0001a440: 6563 740a 2020 2020 626c 6f73 756d 5f65  ect.    blosum_e
-0001a450: 7870 616e 6465 6420 3d20 626c 6f73 756d  xpanded = blosum
-0001a460: 5b31 3a2c 2031 3a5d 2e72 6570 6561 7428  [1:, 1:].repeat(
-0001a470: 6461 7461 7365 745f 7472 6169 6e2e 7368  dataset_train.sh
-0001a480: 6170 655b 305d 2c61 6c69 676e 5f73 6571  ape[0],align_seq
-0001a490: 5f6c 656e 2c20 312c 2031 2920 2023 205b  _len, 1, 1)  # [
-0001a4a0: 4e2c 6d61 785f 6c65 6e2c 6161 5f70 726f  N,max_len,aa_pro
-0001a4b0: 6273 2c61 615f 7072 6f62 735d 0a20 2020  bs,aa_probs].   
-0001a4c0: 2061 615f 7472 6169 6e5f 626c 6f73 756d   aa_train_blosum
-0001a4d0: 203d 2062 6c6f 7375 6d5f 6578 7061 6e64   = blosum_expand
-0001a4e0: 6564 2e67 6174 6865 7228 332c 2061 6d69  ed.gather(3, ami
-0001a4f0: 6e6f 6163 6964 735f 7365 7173 2e74 6f28  noacids_seqs.to(
-0001a500: 746f 7263 682e 696e 7436 3429 2e75 6e73  torch.int64).uns
-0001a510: 7175 6565 7a65 2833 2929 2e73 7175 6565  queeze(3)).squee
-0001a520: 7a65 282d 3129 2020 235b 4e2c 6d61 785f  ze(-1)  #[N,max_
-0001a530: 6c65 6e2c 6161 5f70 726f 6273 5d0a 0a0a  len,aa_probs]...
-0001a540: 2020 2020 7265 7475 726e 2061 615f 7472      return aa_tr
-0001a550: 6169 6e5f 626c 6f73 756d 0a0a 6465 6620  ain_blosum..def 
-0001a560: 7374 7232 626f 6f6c 2876 293a 0a20 2020  str2bool(v):.   
-0001a570: 2022 2222 436f 6e76 6572 7473 2061 2073   """Converts a s
-0001a580: 7472 696e 6720 696e 746f 2061 2062 6f6f  tring into a boo
-0001a590: 6c65 616e 2c20 7573 6566 756c 2066 6f72  lean, useful for
-0001a5a0: 2062 6f6f 6c65 616e 2061 7267 756d 656e   boolean argumen
-0001a5b0: 7473 0a20 2020 203a 7061 7261 6d20 7374  ts.    :param st
-0001a5c0: 7220 7622 2222 0a20 2020 2069 6620 6973  r v""".    if is
-0001a5d0: 696e 7374 616e 6365 2876 2c20 626f 6f6c  instance(v, bool
-0001a5e0: 293a 0a20 2020 2020 2020 2072 6574 7572  ):.        retur
-0001a5f0: 6e20 760a 2020 2020 6966 2076 2e6c 6f77  n v.    if v.low
-0001a600: 6572 2829 2069 6e20 2827 7965 7327 2c20  er() in ('yes', 
-0001a610: 2774 7275 6527 2c20 2774 272c 2027 7927  'true', 't', 'y'
-0001a620: 2c20 2731 272c 2754 7275 6527 293a 0a20  , '1','True'):. 
-0001a630: 2020 2020 2020 2072 6574 7572 6e20 5472         return Tr
-0001a640: 7565 0a20 2020 2065 6c69 6620 762e 6c6f  ue.    elif v.lo
-0001a650: 7765 7228 2920 696e 2028 276e 6f27 2c20  wer() in ('no', 
-0001a660: 2766 616c 7365 272c 2027 6627 2c20 276e  'false', 'f', 'n
-0001a670: 272c 2027 3027 2c27 4661 6c73 6527 293a  ', '0','False'):
-0001a680: 0a20 2020 2020 2020 2072 6574 7572 6e20  .        return 
-0001a690: 4661 6c73 650a 2020 2020 656c 7365 3a0a  False.    else:.
-0001a6a0: 2020 2020 2020 2020 7261 6973 6520 6172          raise ar
-0001a6b0: 6770 6172 7365 2e41 7267 756d 656e 7454  gparse.ArgumentT
-0001a6c0: 7970 6545 7272 6f72 2827 426f 6f6c 6561  ypeError('Boolea
-0001a6d0: 6e20 7661 6c75 6520 6578 7065 6374 6564  n value expected
-0001a6e0: 2e27 290a 6465 6620 7374 7232 4e6f 6e65  .').def str2None
-0001a6f0: 2876 293a 0a20 2020 2022 2222 436f 6e76  (v):.    """Conv
-0001a700: 6572 7473 2061 2073 7472 696e 6720 696e  erts a string in
-0001a710: 746f 204e 6f6e 650a 2020 2020 3a70 6172  to None.    :par
-0001a720: 616d 2073 7472 2076 2222 220a 0a20 2020  am str v"""..   
-0001a730: 2069 6620 762e 6c6f 7765 7228 2920 696e   if v.lower() in
-0001a740: 2028 274e 6f6e 6527 2c27 6e6f 6e65 2729   ('None','none')
-0001a750: 3a0a 2020 2020 2020 2020 7265 7475 726e  :.        return
-0001a760: 204e 6f6e 650a 2020 2020 656c 6966 2069   None.    elif i
-0001a770: 7369 6e73 7461 6e63 6528 762c 7374 7229  sinstance(v,str)
-0001a780: 3a0a 2020 2020 2020 2020 7265 7475 726e  :.        return
-0001a790: 2076 0a20 2020 2065 6c73 653a 0a20 2020   v.    else:.   
-0001a7a0: 2020 2020 2076 203d 2061 7374 2e6c 6974       v = ast.lit
-0001a7b0: 6572 616c 5f65 7661 6c28 7629 0a20 2020  eral_eval(v).   
-0001a7c0: 2020 2020 2072 6574 7572 6e20 760a 0a0a       return v...
-0001a7d0: 0a0a 0a0a                                ....
+00018d90: 6520 2b3d 2031 0a64 6566 2063 616c 6375  e += 1.def calcu
+00018da0: 6c61 7465 5f61 615f 6672 6571 7565 6e63  late_aa_frequenc
+00018db0: 6965 7328 6461 7461 7365 742c 6672 6571  ies(dataset,freq
+00018dc0: 5f62 696e 7329 3a0a 2020 2020 2222 2243  _bins):.    """C
+00018dd0: 616c 6375 6c61 7465 7320 6120 6672 6571  alculates a freq
+00018de0: 7565 6e63 7920 666f 7220 6561 6368 206f  uency for each o
+00018df0: 6620 7468 6520 6161 2026 2067 6170 2061  f the aa & gap a
+00018e00: 7420 6561 6368 2070 6f73 6974 696f 6e2e  t each position.
+00018e10: 5468 6520 6e75 6d62 6572 206f 6620 6269  The number of bi
+00018e20: 6e73 2028 6f66 2073 697a 6520 3129 2069  ns (of size 1) i
+00018e30: 7320 6f6e 6520 6c61 7267 6572 2074 6861  s one larger tha
+00018e40: 6e20 7468 6520 6c61 7267 6573 7420 7661  n the largest va
+00018e50: 6c75 6520 696e 2078 2e20 5468 6973 2069  lue in x. This i
+00018e60: 7320 646f 6e65 2066 6f72 206e 756d 7079  s done for numpy
+00018e70: 2061 7272 6179 730a 2020 2020 3a70 6172   arrays.    :par
+00018e80: 616d 2074 656e 736f 7220 6461 7461 7365  am tensor datase
+00018e90: 740a 2020 2020 3a70 6172 616d 2069 6e74  t.    :param int
+00018ea0: 2066 7265 715f 6269 6e73 0a20 2020 2022   freq_bins.    "
+00018eb0: 2222 0a20 2020 2066 7265 7173 203d 206e  "".    freqs = n
+00018ec0: 702e 6170 706c 795f 616c 6f6e 675f 6178  p.apply_along_ax
+00018ed0: 6973 286c 616d 6264 6120 783a 206e 702e  is(lambda x: np.
+00018ee0: 6269 6e63 6f75 6e74 2878 2c20 6d69 6e6c  bincount(x, minl
+00018ef0: 656e 6774 683d 6672 6571 5f62 696e 7329  ength=freq_bins)
+00018f00: 2c20 6178 6973 3d30 2c20 6172 723d 6461  , axis=0, arr=da
+00018f10: 7461 7365 742e 6173 7479 7065 2822 696e  taset.astype("in
+00018f20: 7436 3422 2929 2e54 0a20 2020 2066 7265  t64")).T.    fre
+00018f30: 7173 203d 2066 7265 7173 2f64 6174 6173  qs = freqs/datas
+00018f40: 6574 2e73 6861 7065 5b30 5d0a 2020 2020  et.shape[0].    
+00018f50: 7265 7475 726e 2066 7265 7173 0a64 6566  return freqs.def
+00018f60: 2063 616c 6375 6c61 7465 5f61 615f 6672   calculate_aa_fr
+00018f70: 6571 7565 6e63 6965 735f 746f 7263 6828  equencies_torch(
+00018f80: 6461 7461 7365 742c 6672 6571 5f62 696e  dataset,freq_bin
+00018f90: 7329 3a0a 2020 2020 2222 2243 616c 6375  s):.    """Calcu
+00018fa0: 6c61 7465 7320 6120 6672 6571 7565 6e63  lates a frequenc
+00018fb0: 7920 666f 7220 6561 6368 206f 6620 7468  y for each of th
+00018fc0: 6520 6161 2026 2067 6170 2061 7420 6561  e aa & gap at ea
+00018fd0: 6368 2070 6f73 6974 696f 6e2e 5468 6520  ch position.The 
+00018fe0: 6e75 6d62 6572 206f 6620 6269 6e73 2028  number of bins (
+00018ff0: 6f66 2073 697a 6520 3129 2069 7320 6f6e  of size 1) is on
+00019000: 6520 6c61 7267 6572 2074 6861 6e20 7468  e larger than th
+00019010: 6520 6c61 7267 6573 7420 7661 6c75 6520  e largest value 
+00019020: 696e 2078 2e20 5468 6973 2069 7320 646f  in x. This is do
+00019030: 6e65 2066 6f72 2074 6f72 6368 2074 656e  ne for torch ten
+00019040: 736f 7273 0a20 2020 203a 7061 7261 6d20  sors.    :param 
+00019050: 7465 6e73 6f72 2064 6174 6173 6574 0a20  tensor dataset. 
+00019060: 2020 203a 7061 7261 6d20 696e 7420 6672     :param int fr
+00019070: 6571 5f62 696e 730a 2020 2020 2222 220a  eq_bins.    """.
+00019080: 2020 2020 6672 6571 7320 3d20 746f 7263      freqs = torc
+00019090: 682e 7374 6163 6b28 5b74 6f72 6368 2e62  h.stack([torch.b
+000190a0: 696e 636f 756e 7428 785f 692c 206d 696e  incount(x_i, min
+000190b0: 6c65 6e67 7468 3d66 7265 715f 6269 6e73  length=freq_bins
+000190c0: 2920 666f 7220 692c 2078 5f69 2069 6e20  ) for i, x_i in 
+000190d0: 656e 756d 6572 6174 6528 746f 7263 682e  enumerate(torch.
+000190e0: 756e 6269 6e64 2864 6174 6173 6574 2e74  unbind(dataset.t
+000190f0: 7970 6528 746f 7263 682e 696e 7436 3429  ype(torch.int64)
+00019100: 2c20 6469 6d3d 3129 2c20 3029 5d2c 2064  , dim=1), 0)], d
+00019110: 696d 3d31 290a 2020 2020 6672 6571 7320  im=1).    freqs 
+00019120: 3d20 6672 6571 732e 540a 2020 2020 6672  = freqs.T.    fr
+00019130: 6571 7320 3d20 6672 6571 7320 2f20 6461  eqs = freqs / da
+00019140: 7461 7365 742e 7368 6170 655b 305d 0a20  taset.shape[0]. 
+00019150: 2020 2072 6574 7572 6e20 6672 6571 730a     return freqs.
+00019160: 6465 6620 636f 6d70 6172 655f 7472 6565  def compare_tree
+00019170: 7328 7431 2c74 3229 3a0a 2020 2020 2222  s(t1,t2):.    ""
+00019180: 2243 6f6d 7075 7465 7320 7468 6520 5266  "Computes the Rf
+00019190: 2064 6973 7461 6e63 6520 616d 6f6e 6720   distance among 
+000191a0: 7477 6f20 7472 6565 2c20 6974 2063 616c  two tree, it cal
+000191b0: 6375 6c61 7465 7320 7468 6520 6e75 6d62  culates the numb
+000191c0: 6572 206f 6620 6368 616e 6765 7320 7265  er of changes re
+000191d0: 7175 6972 6573 2074 6f20 7472 616e 7366  quires to transf
+000191e0: 6f72 6d20 6f6e 6520 7472 6565 2069 6e74  orm one tree int
+000191f0: 6f20 7468 6520 6f74 6865 720a 2020 2020  o the other.    
+00019200: 3a70 6172 616d 2073 7472 2074 313a 2070  :param str t1: p
+00019210: 6174 6820 746f 2066 6972 7374 206e 6574  ath to first net
+00019220: 7769 636b 2074 7265 650a 2020 2020 3a70  wick tree.    :p
+00019230: 6172 616d 2073 7472 2074 323a 2070 6174  aram str t2: pat
+00019240: 6820 746f 2073 6563 6f6e 6420 6e65 7477  h to second netw
+00019250: 6963 6b20 7472 6565 2222 220a 2020 2020  ick tree""".    
+00019260: 636f 6c75 6d6e 7320 3d20 5b22 5246 2064  columns = ["RF d
+00019270: 6973 7461 6e63 6522 2c22 4d61 7869 6d75  istance","Maximu
+00019280: 6d20 5246 2064 6973 7461 6e63 6522 2c22  m RF distance","
+00019290: 436f 6d6d 6f6e 206c 6561 7665 7322 2c22  Common leaves","
+000192a0: 5061 7274 6974 696f 6e73 496e 5f74 315f  PartitionsIn_t1_
+000192b0: 4e4f 545f 7432 222c 2250 6172 7469 7469  NOT_t2","Partiti
+000192c0: 6f6e 7349 6e5f 7432 5f4e 4f54 5f74 3122  onsIn_t2_NOT_t1"
+000192d0: 2c22 4469 7363 6172 6465 645f 7061 7274  ,"Discarded_part
+000192e0: 6974 696f 6e73 5f74 3122 2c22 4469 7363  itions_t1","Disc
+000192f0: 6172 6465 645f 7061 7274 6974 696f 6e73  arded_partitions
+00019300: 5f74 3222 5d0a 2020 2020 2320 7431 2020  _t2"].    # t1  
+00019310: 3d20 5472 6565 4574 6533 2874 3129 0a20  = TreeEte3(t1). 
+00019320: 2020 2023 2074 3220 3d20 5472 6565 4574     # t2 = TreeEt
+00019330: 6533 2874 3229 0a20 2020 2023 2072 6573  e3(t2).    # res
+00019340: 756c 7473 203d 2074 312e 726f 6269 6e73  ults = t1.robins
+00019350: 6f6e 5f66 6f75 6c64 7328 7432 2c75 6e72  on_foulds(t2,unr
+00019360: 6f6f 7465 645f 7472 6565 733d 5472 7565  ooted_trees=True
+00019370: 290a 2020 2020 636f 6d6d 616e 6420 3d20  ).    command = 
+00019380: 2752 7363 7269 7074 270a 2020 2020 7061  'Rscript'.    pa
+00019390: 7468 3273 6372 6970 7420 3d20 272f 686f  th2script = '/ho
+000193a0: 6d65 2f6c 7973 2f44 726f 7062 6f78 2f50  me/lys/Dropbox/P
+000193b0: 6844 2f44 5241 5550 4e49 522f 5472 6565  hD/DRAUPNIR/Tree
+000193c0: 5f44 6973 7461 6e63 652e 5227 0a20 2020  _Distance.R'.   
+000193d0: 2023 2042 7569 6c64 2073 7562 7072 6f63   # Build subproc
+000193e0: 6573 7320 636f 6d6d 616e 640a 2020 2020  ess command.    
+000193f0: 6172 6773 203d 5b74 312c 7432 5d0a 2020  args =[t1,t2].  
+00019400: 2020 636d 6420 3d20 5b63 6f6d 6d61 6e64    cmd = [command
+00019410: 2c20 7061 7468 3273 6372 6970 745d 202b  , path2script] +
+00019420: 2061 7267 730a 2020 2020 6469 7374 616e   args.    distan
+00019430: 6365 203d 2073 7562 7072 6f63 6573 732e  ce = subprocess.
+00019440: 6368 6563 6b5f 6f75 7470 7574 2863 6d64  check_output(cmd
+00019450: 2c20 756e 6976 6572 7361 6c5f 6e65 776c  , universal_newl
+00019460: 696e 6573 3d54 7275 6529 0a20 2020 2072  ines=True).    r
+00019470: 6574 7572 6e20 6469 7374 616e 6365 0a64  eturn distance.d
+00019480: 6566 2072 656d 6f76 655f 7374 6f70 5f63  ef remove_stop_c
+00019490: 6f64 6f6e 7328 7365 7175 656e 6365 5f66  odons(sequence_f
+000194a0: 696c 652c 6973 5f70 726f 743d 4661 6c73  ile,is_prot=Fals
+000194b0: 6529 3a0a 2020 2020 2222 2252 656d 6f76  e):.    """Remov
+000194c0: 6520 7468 6520 7374 6f70 2063 6f64 6f6e  e the stop codon
+000194d0: 7328 2a29 2066 726f 6d20 616e 2061 6c69  s(*) from an ali
+000194e0: 676e 6d65 6e74 2066 696c 6522 2222 0a20  gnment file""". 
+000194f0: 2020 2069 6620 6973 5f70 726f 743a 0a20     if is_prot:. 
+00019500: 2020 2020 2020 2070 7269 6e74 2822 5265         print("Re
+00019510: 6d6f 7669 6e67 2073 746f 7020 636f 646f  moving stop codo
+00019520: 6e73 2066 726f 6d20 7072 6f74 6569 6e22  ns from protein"
+00019530: 290a 2020 2020 2020 2020 7072 696e 7428  ).        print(
+00019540: 7365 7175 656e 6365 5f66 696c 6529 0a20  sequence_file). 
+00019550: 2020 2020 2020 2073 6571 5f66 696c 6520         seq_file 
+00019560: 3d20 6f70 656e 2873 6571 7565 6e63 655f  = open(sequence_
+00019570: 6669 6c65 2c20 2772 2b27 290a 2020 2020  file, 'r+').    
+00019580: 2020 2020 7365 7120 3d20 7365 715f 6669      seq = seq_fi
+00019590: 6c65 2e72 6561 6428 290a 2020 2020 2020  le.read().      
+000195a0: 2020 7365 715f 6669 6c65 2e73 6565 6b28    seq_file.seek(
+000195b0: 3029 0a20 2020 2020 2020 206e 6f5f 636f  0).        no_co
+000195c0: 646f 6e73 203d 205b 6920 6966 2069 213d  dons = [i if i!=
+000195d0: 222a 2220 656c 7365 2022 2d22 2066 6f72  "*" else "-" for
+000195e0: 2069 2069 6e20 7365 715d 2023 7265 706c   i in seq] #repl
+000195f0: 6163 6520 666f 7220 6120 6761 700a 2020  ace for a gap.  
+00019600: 2020 2020 2020 2354 4f44 4f3a 2054 7275        #TODO: Tru
+00019610: 6e63 6174 6520 7468 6520 7365 7175 656e  ncate the sequen
+00019620: 6365 2069 6620 7468 6572 6520 6973 2061  ce if there is a
+00019630: 2073 746f 7020 636f 646f 6e20 696e 2074   stop codon in t
+00019640: 6865 206d 6964 646c 6520 2873 746f 7020  he middle (stop 
+00019650: 6164 6469 6e67 2067 6170 7320 7768 656e  adding gaps when
+00019660: 205c 6e20 6973 2072 6561 6368 6564 290a   \n is reached).
+00019670: 2020 2020 2020 2020 2320 6e6f 5f63 6f64          # no_cod
+00019680: 6f6e 733d 205b 5d0a 2020 2020 2020 2020  ons= [].        
+00019690: 2320 666f 7220 696e 6465 782c 6920 696e  # for index,i in
+000196a0: 2065 6e75 6d65 7261 7465 2873 6571 293a   enumerate(seq):
+000196b0: 0a20 2020 2020 2020 2023 2020 2020 2069  .        #     i
+000196c0: 6620 6920 3d3d 2022 2a22 3a0a 2020 2020  f i == "*":.    
+000196d0: 2020 2020 2320 2020 2020 2020 2020 693d      #         i=
+000196e0: 2022 2d22 0a20 2020 2020 2020 2023 2020   "-".        #  
+000196f0: 2020 2020 2020 206e 6578 745f 6272 6561         next_brea
+00019700: 6b5f 696e 6465 7820 3d20 7365 715b 696e  k_index = seq[in
+00019710: 6465 783a 5d2e 696e 6465 7828 225c 6e22  dex:].index("\n"
+00019720: 290a 2020 2020 2020 2020 2320 2020 2020  ).        #     
+00019730: 2020 2020 7365 715b 696e 6465 783a 6e65      seq[index:ne
+00019740: 7874 5f62 7265 616b 5f69 6e64 6578 5d20  xt_break_index] 
+00019750: 3d20 222d 222a 286e 6578 745f 6272 6561  = "-"*(next_brea
+00019760: 6b5f 696e 6465 782d 696e 6465 7829 0a0a  k_index-index)..
+00019770: 2020 2020 2020 2020 7365 715f 6669 6c65          seq_file
+00019780: 2e77 7269 7465 2822 222e 6a6f 696e 286e  .write("".join(n
+00019790: 6f5f 636f 646f 6e73 2929 0a20 2020 2020  o_codons)).     
+000197a0: 2020 2073 6571 5f66 696c 652e 7472 756e     seq_file.trun
+000197b0: 6361 7465 2829 2020 2320 7265 6d6f 7665  cate()  # remove
+000197c0: 2063 6f6e 7465 6e74 730a 2020 2020 656c   contents.    el
+000197d0: 7365 3a0a 2020 2020 2020 2020 7374 6f70  se:.        stop
+000197e0: 5f63 6f64 6f6e 7320 3d20 5b22 5447 4122  _codons = ["TGA"
+000197f0: 2c22 5441 4722 2c22 5441 4122 5d0a 2020  ,"TAG","TAA"].  
+00019800: 2020 2020 2020 7365 715f 6669 6c65 203d        seq_file =
+00019810: 206f 7065 6e28 7365 7175 656e 6365 5f66   open(sequence_f
+00019820: 696c 652c 2027 722b 2729 0a20 2020 2020  ile, 'r+').     
+00019830: 2020 2073 6571 203d 2073 6571 5f66 696c     seq = seq_fil
+00019840: 652e 7265 6164 2829 0a20 2020 2020 2020  e.read().       
+00019850: 2073 6571 5f66 696c 652e 7365 656b 2830   seq_file.seek(0
+00019860: 290a 2020 2020 2020 2020 2363 6f64 696e  ).        #codin
+00019870: 675f 646e 6120 3d20 5365 7128 7365 7129  g_dna = Seq(seq)
+00019880: 0a20 2020 2020 2020 2023 7072 6f74 6569  .        #protei
+00019890: 6e20 3d20 636f 6469 6e67 5f64 6e61 2e74  n = coding_dna.t
+000198a0: 7261 6e73 6c61 7465 2829 0a20 2020 2020  ranslate().     
+000198b0: 2020 2063 6f64 6f6e 7320 3d20 5b73 6571     codons = [seq
+000198c0: 5b69 3a69 2b33 5d20 666f 7220 6920 696e  [i:i+3] for i in
+000198d0: 2072 616e 6765 2830 2c20 6c65 6e28 7365   range(0, len(se
+000198e0: 7129 2c20 3329 5d0a 2020 2020 2020 2020  q), 3)].        
+000198f0: 636f 646f 6e73 203d 5b63 6f64 2066 6f72  codons =[cod for
+00019900: 2063 6f64 2069 6e20 636f 646f 6e73 2069   cod in codons i
+00019910: 6620 6c65 6e28 636f 6429 203d 3d33 2061  f len(cod) ==3 a
+00019920: 6e64 2063 6f64 206e 6f74 2069 6e20 7374  nd cod not in st
+00019930: 6f70 5f63 6f64 6f6e 735d 0a20 2020 2020  op_codons].     
+00019940: 2020 2023 6368 6563 6b20 3d20 616e 7928     #check = any(
+00019950: 6974 656d 2069 6e20 636f 646f 6e73 2066  item in codons f
+00019960: 6f72 2069 7465 6d20 696e 2073 746f 705f  or item in stop_
+00019970: 636f 646f 6e73 290a 2020 2020 2020 2020  codons).        
+00019980: 7365 715f 6669 6c65 2e77 7269 7465 2822  seq_file.write("
+00019990: 222e 6a6f 696e 2863 6f64 6f6e 7329 290a  ".join(codons)).
+000199a0: 2020 2020 2020 2020 7365 715f 6669 6c65          seq_file
+000199b0: 2e74 7275 6e63 6174 6528 2920 2023 2072  .truncate()  # r
+000199c0: 656d 6f76 6520 636f 6e74 656e 7473 0a64  emove contents.d
+000199d0: 6566 2063 6f6e 7665 7274 5f74 6f5f 696e  ef convert_to_in
+000199e0: 7465 6765 7273 2864 6174 6173 6574 2c61  tegers(dataset,a
+000199f0: 615f 7072 6f62 732c 6178 6973 293a 0a20  a_probs,axis):. 
+00019a00: 2020 2022 2222 5472 616e 7366 6f72 6d73     """Transforms
+00019a10: 206f 6e65 2068 6f74 2065 6e63 6f64 6564   one hot encoded
+00019a20: 2061 6d69 6e6f 2061 6369 6473 2069 6e74   amino acids int
+00019a30: 6f20 302d 696e 6465 7865 6420 696e 7465  o 0-indexed inte
+00019a40: 6765 7273 2069 2e65 205b 312c 302c 302c  gers i.e [1,0,0,
+00019a50: 305d 202d 2d3e 2030 2020 5b30 2c31 2c30  0] --> 0  [0,1,0
+00019a60: 2c30 5d20 2d3e 2031 0a20 2020 203a 7061  ,0] -> 1.    :pa
+00019a70: 7261 6d20 6e75 6d70 7920 6461 7461 7365  ram numpy datase
+00019a80: 740a 2020 2020 3a70 6172 616d 2061 615f  t.    :param aa_
+00019a90: 7072 6f62 733a 2061 6d69 6e6f 2061 6369  probs: amino aci
+00019aa0: 6420 7072 6f62 6162 696c 6974 6965 732c  d probabilities,
+00019ab0: 2064 696d 656e 7369 6f6e 7320 6f66 2074   dimensions of t
+00019ac0: 6865 206f 6e65 2d68 6f74 2065 6e63 6f64  he one-hot encod
+00019ad0: 696e 6722 2222 0a20 2020 2069 6620 6178  ing""".    if ax
+00019ae0: 6973 3d3d 333a 2375 7365 2066 6f72 2070  is==3:#use for p
+00019af0: 7265 6469 6374 696f 6e73 0a20 2020 2020  redictions.     
+00019b00: 2020 2062 203d 2074 6f72 6368 2e61 7267     b = torch.arg
+00019b10: 6d61 7828 6461 7461 7365 742c 2064 696d  max(dataset, dim
+00019b20: 3d61 7869 7329 0a20 2020 2065 6c73 653a  =axis).    else:
+00019b30: 0a20 2020 2020 2020 2069 6e74 6567 6572  .        integer
+00019b40: 7320 3d20 746f 7263 682e 6172 676d 6178  s = torch.argmax
+00019b50: 2864 6174 6173 6574 5b3a 2c32 3a2c 303a  (dataset[:,2:,0:
+00019b60: 6161 5f70 726f 6273 5d2c 6469 6d3d 6178  aa_probs],dim=ax
+00019b70: 6973 290a 2020 2020 2020 2020 6220 3d20  is).        b = 
+00019b80: 746f 7263 682e 7a65 726f 7328 6461 7461  torch.zeros(data
+00019b90: 7365 745b 3a2c 3a2c 305d 2e73 6861 7065  set[:,:,0].shape
+00019ba0: 202b 2028 3330 2c29 292e 6370 7528 290a   + (30,)).cpu().
+00019bb0: 2020 2020 2020 2020 2362 203d 206e 702e          #b = np.
+00019bc0: 7a65 726f 7328 4461 7461 7365 745b 3a2c  zeros(Dataset[:,
+00019bd0: 3a2c 305d 2e73 6861 7065 202b 2028 3330  :,0].shape + (30
+00019be0: 2c29 290a 2020 2020 2020 2020 625b 3a2c  ,)).        b[:,
+00019bf0: 2032 3a2c 2030 5d20 3d20 696e 7465 6765   2:, 0] = intege
+00019c00: 7273 0a20 2020 2020 2020 2062 5b3a 2c3a  rs.        b[:,:
+00019c10: 325d 203d 2064 6174 6173 6574 5b3a 2c3a  2] = dataset[:,:
+00019c20: 325d 0a20 2020 2072 6574 7572 6e20 620a  2].    return b.
+00019c30: 6465 6620 7072 6f63 6573 735f 626c 6f73  def process_blos
+00019c40: 756d 2862 6c6f 7375 6d2c 6161 5f66 7265  um(blosum,aa_fre
+00019c50: 7173 2c61 6c69 676e 5f73 6571 5f6c 656e  qs,align_seq_len
+00019c60: 2c61 615f 7072 6f62 7329 3a0a 2020 2020  ,aa_probs):.    
+00019c70: 2222 220a 2020 2020 436f 6d70 7574 6573  """.    Computes
+00019c80: 2074 6865 206d 6174 7269 6365 7320 7265   the matrices re
+00019c90: 7175 6972 6564 2074 6f20 6275 696c 6420  quired to build 
+00019ca0: 6120 626c 6f73 756d 2065 6d62 6564 6469  a blosum embeddi
+00019cb0: 6e67 0a20 2020 203a 7061 7261 6d20 7465  ng.    :param te
+00019cc0: 6e73 6f72 2062 6c6f 7375 6d3a 2042 4c4f  nsor blosum: BLO
+00019cd0: 5355 4d20 6c69 6b65 6c69 686f 6f64 2020  SUM likelihood  
+00019ce0: 7363 6f72 6573 0a20 2020 203a 7061 7261  scores.    :para
+00019cf0: 6d20 7465 6e73 6f72 2061 615f 6672 6571  m tensor aa_freq
+00019d00: 7320 3a20 616d 696e 6f20 6163 6964 2066  s : amino acid f
+00019d10: 7265 7175 656e 6369 6573 2070 6572 2070  requencies per p
+00019d20: 6f73 6974 696f 6e0a 2020 2020 3a70 6172  osition.    :par
+00019d30: 616d 2061 6c69 676e 5f73 6571 5f6c 656e  am align_seq_len
+00019d40: 3a20 616c 6967 6e6d 656e 7420 6c65 6e67  : alignment leng
+00019d50: 7468 0a20 2020 203a 7061 7261 6d20 6161  th.    :param aa
+00019d60: 5f70 726f 6273 3a20 616d 696e 6f20 6163  _probs: amino ac
+00019d70: 6964 2070 726f 6261 6269 6c69 7469 6573  id probabilities
+00019d80: 2c20 7479 7065 7320 6f66 2061 6d69 6e6f  , types of amino
+00019d90: 2061 6369 6473 2069 6e20 7468 6520 616c   acids in the al
+00019da0: 6967 6e6d 656e 740a 2020 2020 3a6f 7574  ignment.    :out
+00019db0: 2074 656e 736f 7220 626c 6f73 756d 5f6d   tensor blosum_m
+00019dc0: 6178 205b 616c 6967 6e5f 6c65 6e2c 6161  ax [align_len,aa
+00019dd0: 5f70 726f 625d 3a20 626c 6f73 756d 206c  _prob]: blosum l
+00019de0: 696b 656c 6968 6f6f 6420 7363 6f72 6573  ikelihood scores
+00019df0: 2066 6f72 2074 6865 206d 6f73 7420 6672   for the most fr
+00019e00: 6571 7565 6e74 2061 6120 696e 2074 6865  equent aa in the
+00019e10: 2061 6c69 676e 6d65 6e74 2070 6f73 6974   alignment posit
+00019e20: 696f 6e0a 2020 2020 3a6f 7574 2074 656e  ion.    :out ten
+00019e30: 736f 7220 626c 6f73 756d 5f77 6569 6768  sor blosum_weigh
+00019e40: 7465 6420 5b61 6c69 676e 5f6c 656e 2c61  ted [align_len,a
+00019e50: 615f 7072 6f62 3a20 7765 6967 6874 6564  a_prob: weighted
+00019e60: 2061 7665 7261 6765 206f 6620 626c 6f73   average of blos
+00019e70: 756d 206c 696b 656c 6968 6f6f 6473 2061  um likelihoods a
+00019e80: 6363 6f72 6469 6e67 2074 6f20 7468 6520  ccording to the 
+00019e90: 6161 2066 7265 7175 656e 6379 0a20 2020  aa frequency.   
+00019ea0: 203a 6f75 7420 7661 7269 6162 6c65 5f63   :out variable_c
+00019eb0: 6f72 653a 205b 616c 6967 6e5f 6c65 6e5d  ore: [align_len]
+00019ec0: 203a 2063 6f75 6e74 7320 7468 6520 6e75   : counts the nu
+00019ed0: 6d62 6572 206f 6620 6469 6666 6572 656e  mber of differen
+00019ee0: 7420 656c 656d 656e 7473 2028 616d 696e  t elements (amin
+00019ef0: 6f20 6163 6964 2064 6976 6572 7369 7479  o acid diversity
+00019f00: 2920 7065 7220 616c 6967 6e6d 656e 7420  ) per alignment 
+00019f10: 706f 7369 7469 6f6e 2222 220a 0a20 2020  position"""..   
+00019f20: 2061 615f 6672 6571 735f 6d61 7820 3d20   aa_freqs_max = 
+00019f30: 746f 7263 682e 6172 676d 6178 2861 615f  torch.argmax(aa_
+00019f40: 6672 6571 732c 2064 696d 3d31 292e 7265  freqs, dim=1).re
+00019f50: 7065 6174 2861 615f 7072 6f62 732c 2031  peat(aa_probs, 1
+00019f60: 292e 7065 726d 7574 6528 312c 2030 2920  ).permute(1, 0) 
+00019f70: 235b 6d61 785f 6c65 6e2c 2061 615f 7072  #[max_len, aa_pr
+00019f80: 6f62 735d 0a20 2020 2062 6c6f 7375 6d5f  obs].    blosum_
+00019f90: 6578 7061 6e64 6564 203d 2062 6c6f 7375  expanded = blosu
+00019fa0: 6d5b 313a 2c20 313a 5d2e 7265 7065 6174  m[1:, 1:].repeat
+00019fb0: 2861 6c69 676e 5f73 6571 5f6c 656e 2c20  (align_seq_len, 
+00019fc0: 312c 2031 2920 2023 205b 6d61 785f 6c65  1, 1)  # [max_le
+00019fd0: 6e2c 6161 5f70 726f 6273 2c61 615f 7072  n,aa_probs,aa_pr
+00019fe0: 6f62 735d 0a20 2020 2062 6c6f 7375 6d5f  obs].    blosum_
+00019ff0: 6d61 7820 3d20 626c 6f73 756d 5f65 7870  max = blosum_exp
+0001a000: 616e 6465 642e 6761 7468 6572 2831 2c20  anded.gather(1, 
+0001a010: 6161 5f66 7265 7173 5f6d 6178 2e75 6e73  aa_freqs_max.uns
+0001a020: 7175 6565 7a65 2831 2929 2e73 7175 6565  queeze(1)).squee
+0001a030: 7a65 2831 2920 2023 205b 616c 6967 6e5f  ze(1)  # [align_
+0001a040: 7365 715f 6c65 6e2c 3231 5d20 5365 656d  seq_len,21] Seem
+0001a050: 7320 636f 7272 6563 740a 0a20 2020 2062  s correct..    b
+0001a060: 6c6f 7375 6d5f 7765 6967 6874 6564 203d  losum_weighted =
+0001a070: 2061 615f 6672 6571 735b 3a2c 3a2c 4e6f   aa_freqs[:,:,No
+0001a080: 6e65 5d2a 626c 6f73 756d 5f65 7870 616e  ne]*blosum_expan
+0001a090: 6465 6420 232d 2d3e 2072 6570 6c61 6365  ded #--> replace
+0001a0a0: 2030 2077 6974 6820 6e61 6e73 3f20 6f74   0 with nans? ot
+0001a0b0: 6865 7277 6973 6520 7468 6520 3020 6172  herwise the 0 ar
+0001a0c0: 6520 696e 2074 6865 206d 6561 6e20 6173  e in the mean as
+0001a0d0: 2077 656c 6c2e 2e2e 2e0a 2020 2020 626c   well.....    bl
+0001a0e0: 6f73 756d 5f77 6569 6768 7465 6420 3d20  osum_weighted = 
+0001a0f0: 626c 6f73 756d 5f77 6569 6768 7465 642e  blosum_weighted.
+0001a100: 6d65 616e 2864 696d 3d31 290a 0a20 2020  mean(dim=1)..   
+0001a110: 2076 6172 6961 626c 655f 7363 6f72 6520   variable_score 
+0001a120: 3d20 746f 7263 682e 636f 756e 745f 6e6f  = torch.count_no
+0001a130: 6e7a 6572 6f28 6161 5f66 7265 7173 2c20  nzero(aa_freqs, 
+0001a140: 6469 6d3d 3129 2f61 615f 7072 6f62 7320  dim=1)/aa_probs 
+0001a150: 2368 6967 6865 7220 7363 6f72 652c 206d  #higher score, m
+0001a160: 6f72 6520 7661 7269 6162 6c65 0a0a 2020  ore variable..  
+0001a170: 2020 7265 7475 726e 2062 6c6f 7375 6d5f    return blosum_
+0001a180: 6d61 782c 626c 6f73 756d 5f77 6569 6768  max,blosum_weigh
+0001a190: 7465 642c 2076 6172 6961 626c 655f 7363  ted, variable_sc
+0001a1a0: 6f72 650a 6465 6620 626c 6f73 756d 5f65  ore.def blosum_e
+0001a1b0: 6d62 6564 6469 6e67 5f65 6e63 6f64 6572  mbedding_encoder
+0001a1c0: 2862 6c6f 7375 6d2c 6161 5f66 7265 7173  (blosum,aa_freqs
+0001a1d0: 2c61 6c69 676e 5f73 6571 5f6c 656e 2c61  ,align_seq_len,a
+0001a1e0: 615f 7072 6f62 732c 6461 7461 7365 745f  a_probs,dataset_
+0001a1f0: 7472 6169 6e2c 206f 6e65 5f68 6f74 5f65  train, one_hot_e
+0001a200: 6e63 6f64 696e 6729 3a0a 2020 2020 2222  ncoding):.    ""
+0001a210: 2254 7261 6e73 666f 726d 7320 7468 6520  "Transforms the 
+0001a220: 7472 6169 6e20 6461 7461 7365 742c 2077  train dataset, w
+0001a230: 6869 6368 2069 7320 666f 726d 6564 2062  hich is formed b
+0001a240: 7920 6120 7465 6e73 6f72 206f 6620 696e  y a tensor of in
+0001a250: 7465 6765 7273 2074 6861 7420 7265 7072  tegers that repr
+0001a260: 6573 656e 7420 7468 6520 616d 696e 6f20  esent the amino 
+0001a270: 6163 6964 732c 2069 6e74 6f20 6120 6461  acids, into a da
+0001a280: 7461 7365 7420 7768 6572 6520 6561 6368  taset where each
+0001a290: 2061 6d69 6e6f 2061 6369 6420 2869 6e74   amino acid (int
+0001a2a0: 6567 6572 2920 6973 2065 7863 6861 6e67  eger) is exchang
+0001a2b0: 6564 2077 6974 6820 6974 7320 626c 6f73  ed with its blos
+0001a2c0: 756d 2073 636f 7265 0a20 2020 203a 6f75  um score.    :ou
+0001a2d0: 7420 7465 6e73 6f72 2061 615f 7472 6169  t tensor aa_trai
+0001a2e0: 6e5f 626c 6f73 756d 203a 2054 7261 696e  n_blosum : Train
+0001a2f0: 696e 6720 6461 7461 7365 7420 7769 7468  ing dataset with
+0001a300: 2074 6865 2062 6c6f 7375 6d20 7665 6374   the blosum vect
+0001a310: 6f72 7320 696e 7374 6561 6420 6f66 2074  ors instead of t
+0001a320: 6865 2061 6d69 6e6f 2061 6369 6473 2028  he amino acids (
+0001a330: 6e75 6d62 6572 7320 6f72 206f 6e65 2068  numbers or one h
+0001a340: 6f74 2072 6570 7265 7365 6e74 6174 696f  ot representatio
+0001a350: 6e29 2222 220a 2020 2020 6966 206f 6e65  n)""".    if one
+0001a360: 5f68 6f74 5f65 6e63 6f64 696e 673a 0a20  _hot_encoding:. 
+0001a370: 2020 2020 2020 2064 6174 6173 6574 5f74         dataset_t
+0001a380: 7261 696e 203d 2063 6f6e 7665 7274 5f74  rain = convert_t
+0001a390: 6f5f 696e 7465 6765 7273 2864 6174 6173  o_integers(datas
+0001a3a0: 6574 5f74 7261 696e 2c61 615f 7072 6f62  et_train,aa_prob
+0001a3b0: 732c 6178 6973 3d32 290a 2020 2020 616d  s,axis=2).    am
+0001a3c0: 696e 6f61 6369 6473 5f73 6571 7320 3d20  inoacids_seqs = 
+0001a3d0: 6461 7461 7365 745f 7472 6169 6e5b 3a2c  dataset_train[:,
+0001a3e0: 323a 2c30 5d2e 7265 7065 6174 2861 615f  2:,0].repeat(aa_
+0001a3f0: 7072 6f62 732c 312c 3129 2e70 6572 6d75  probs,1,1).permu
+0001a400: 7465 2831 2c32 2c30 2920 235b 4e2c 6d61  te(1,2,0) #[N,ma
+0001a410: 785f 6c65 6e2c 6161 2072 6570 6561 7465  x_len,aa repeate
+0001a420: 6420 6161 5f70 726f 6273 2074 696d 6573  d aa_probs times
+0001a430: 5d2d 2d73 6565 6d73 2063 6f72 7265 6374  ]--seems correct
+0001a440: 0a20 2020 2062 6c6f 7375 6d5f 6578 7061  .    blosum_expa
+0001a450: 6e64 6564 203d 2062 6c6f 7375 6d5b 313a  nded = blosum[1:
+0001a460: 2c20 313a 5d2e 7265 7065 6174 2864 6174  , 1:].repeat(dat
+0001a470: 6173 6574 5f74 7261 696e 2e73 6861 7065  aset_train.shape
+0001a480: 5b30 5d2c 616c 6967 6e5f 7365 715f 6c65  [0],align_seq_le
+0001a490: 6e2c 2031 2c20 3129 2020 2320 5b4e 2c6d  n, 1, 1)  # [N,m
+0001a4a0: 6178 5f6c 656e 2c61 615f 7072 6f62 732c  ax_len,aa_probs,
+0001a4b0: 6161 5f70 726f 6273 5d0a 2020 2020 6161  aa_probs].    aa
+0001a4c0: 5f74 7261 696e 5f62 6c6f 7375 6d20 3d20  _train_blosum = 
+0001a4d0: 626c 6f73 756d 5f65 7870 616e 6465 642e  blosum_expanded.
+0001a4e0: 6761 7468 6572 2833 2c20 616d 696e 6f61  gather(3, aminoa
+0001a4f0: 6369 6473 5f73 6571 732e 746f 2874 6f72  cids_seqs.to(tor
+0001a500: 6368 2e69 6e74 3634 292e 756e 7371 7565  ch.int64).unsque
+0001a510: 657a 6528 3329 292e 7371 7565 657a 6528  eze(3)).squeeze(
+0001a520: 2d31 2920 2023 5b4e 2c6d 6178 5f6c 656e  -1)  #[N,max_len
+0001a530: 2c61 615f 7072 6f62 735d 0a0a 0a20 2020  ,aa_probs]...   
+0001a540: 2072 6574 7572 6e20 6161 5f74 7261 696e   return aa_train
+0001a550: 5f62 6c6f 7375 6d0a 0a64 6566 2073 7472  _blosum..def str
+0001a560: 3262 6f6f 6c28 7629 3a0a 2020 2020 2222  2bool(v):.    ""
+0001a570: 2243 6f6e 7665 7274 7320 6120 7374 7269  "Converts a stri
+0001a580: 6e67 2069 6e74 6f20 6120 626f 6f6c 6561  ng into a boolea
+0001a590: 6e2c 2075 7365 6675 6c20 666f 7220 626f  n, useful for bo
+0001a5a0: 6f6c 6561 6e20 6172 6775 6d65 6e74 730a  olean arguments.
+0001a5b0: 2020 2020 3a70 6172 616d 2073 7472 2076      :param str v
+0001a5c0: 2222 220a 2020 2020 6966 2069 7369 6e73  """.    if isins
+0001a5d0: 7461 6e63 6528 762c 2062 6f6f 6c29 3a0a  tance(v, bool):.
+0001a5e0: 2020 2020 2020 2020 7265 7475 726e 2076          return v
+0001a5f0: 0a20 2020 2069 6620 762e 6c6f 7765 7228  .    if v.lower(
+0001a600: 2920 696e 2028 2779 6573 272c 2027 7472  ) in ('yes', 'tr
+0001a610: 7565 272c 2027 7427 2c20 2779 272c 2027  ue', 't', 'y', '
+0001a620: 3127 2c27 5472 7565 2729 3a0a 2020 2020  1','True'):.    
+0001a630: 2020 2020 7265 7475 726e 2054 7275 650a      return True.
+0001a640: 2020 2020 656c 6966 2076 2e6c 6f77 6572      elif v.lower
+0001a650: 2829 2069 6e20 2827 6e6f 272c 2027 6661  () in ('no', 'fa
+0001a660: 6c73 6527 2c20 2766 272c 2027 6e27 2c20  lse', 'f', 'n', 
+0001a670: 2730 272c 2746 616c 7365 2729 3a0a 2020  '0','False'):.  
+0001a680: 2020 2020 2020 7265 7475 726e 2046 616c        return Fal
+0001a690: 7365 0a20 2020 2065 6c73 653a 0a20 2020  se.    else:.   
+0001a6a0: 2020 2020 2072 6169 7365 2061 7267 7061       raise argpa
+0001a6b0: 7273 652e 4172 6775 6d65 6e74 5479 7065  rse.ArgumentType
+0001a6c0: 4572 726f 7228 2742 6f6f 6c65 616e 2076  Error('Boolean v
+0001a6d0: 616c 7565 2065 7870 6563 7465 642e 2729  alue expected.')
+0001a6e0: 0a64 6566 2073 7472 324e 6f6e 6528 7629  .def str2None(v)
+0001a6f0: 3a0a 2020 2020 2222 2243 6f6e 7665 7274  :.    """Convert
+0001a700: 7320 6120 7374 7269 6e67 2069 6e74 6f20  s a string into 
+0001a710: 4e6f 6e65 0a20 2020 203a 7061 7261 6d20  None.    :param 
+0001a720: 7374 7220 7622 2222 0a0a 2020 2020 6966  str v"""..    if
+0001a730: 2076 2e6c 6f77 6572 2829 2069 6e20 2827   v.lower() in ('
+0001a740: 4e6f 6e65 272c 276e 6f6e 6527 293a 0a20  None','none'):. 
+0001a750: 2020 2020 2020 2072 6574 7572 6e20 4e6f         return No
+0001a760: 6e65 0a20 2020 2065 6c69 6620 6973 696e  ne.    elif isin
+0001a770: 7374 616e 6365 2876 2c73 7472 293a 0a20  stance(v,str):. 
+0001a780: 2020 2020 2020 2072 6574 7572 6e20 760a         return v.
+0001a790: 2020 2020 656c 7365 3a0a 2020 2020 2020      else:.      
+0001a7a0: 2020 7620 3d20 6173 742e 6c69 7465 7261    v = ast.litera
+0001a7b0: 6c5f 6576 616c 2876 290a 2020 2020 2020  l_eval(v).      
+0001a7c0: 2020 7265 7475 726e 2076 0a64 6566 2073    return v.def s
+0001a7d0: 7175 6565 7a65 5f74 656e 736f 7228 7265  queeze_tensor(re
+0001a7e0: 7175 6972 6564 5f6e 6469 6d73 2c74 656e  quired_ndims,ten
+0001a7f0: 736f 7229 3a0a 2020 2020 2222 2253 7175  sor):.    """Squ
+0001a800: 6565 7a65 7320 6120 7465 6e73 6f72 2074  eezes a tensor t
+0001a810: 6f20 6d61 7463 6820 7265 7175 6972 6564  o match required
+0001a820: 5f6e 6469 6d20 7769 7468 6f75 7420 6c65  _ndim without le
+0001a830: 6674 6f76 6572 2065 6d70 7479 2064 696d  ftover empty dim
+0001a840: 656e 7369 6f6e 7320 2831 290a 2020 2020  ensions (1).    
+0001a850: 4558 414d 504c 453a 0a20 2020 2020 2020  EXAMPLE:.       
+0001a860: 2061 2e73 6861 7065 2069 7320 5b33 302c   a.shape is [30,
+0001a870: 312c 312c 3430 5d20 2363 6f6e 7461 696e  1,1,40] #contain
+0001a880: 7320 656d 7074 7920 6469 6d65 6e73 696f  s empty dimensio
+0001a890: 6e73 0a20 2020 2020 2020 2072 6571 7569  ns.        requi
+0001a8a0: 7265 645f 6e64 696d 733d 2032 0a20 2020  red_ndims= 2.   
+0001a8b0: 2020 2020 2061 5f73 7175 6565 7a65 6420       a_squeezed 
+0001a8c0: 3d20 7371 7565 657a 655f 7465 6e73 6f72  = squeeze_tensor
+0001a8d0: 2872 6571 7569 7265 645f 6e64 696d 732c  (required_ndims,
+0001a8e0: 6129 0a20 2020 2020 2020 2061 5f73 7175  a).        a_squ
+0001a8f0: 6565 7a65 642e 7368 6170 6520 3d20 5b33  eezed.shape = [3
+0001a900: 302c 3430 5d0a 2020 2020 2222 220a 2020  0,40].    """.  
+0001a910: 2020 7369 7a65 203d 2074 6f72 6368 2e74    size = torch.t
+0001a920: 656e 736f 7228 7465 6e73 6f72 2e73 6861  ensor(tensor.sha
+0001a930: 7065 290a 2020 2020 6e64 696d 7320 3d20  pe).    ndims = 
+0001a940: 6c65 6e28 7369 7a65 290a 2020 2020 6964  len(size).    id
+0001a950: 785f 6f6e 6573 203d 2028 7369 7a65 203d  x_ones = (size =
+0001a960: 3d20 3129 0a20 2020 2069 6620 5472 7565  = 1).    if True
+0001a970: 2069 6e20 6964 785f 6f6e 6573 3a0a 2020   in idx_ones:.  
+0001a980: 2020 2020 2020 6f6e 6573 5f70 6f73 203d        ones_pos =
+0001a990: 2073 697a 652e 746f 6c69 7374 2829 2e69   size.tolist().i
+0001a9a0: 6e64 6578 2831 290a 2020 2020 6966 206e  ndex(1).    if n
+0001a9b0: 6469 6d73 203e 2072 6571 7569 7265 645f  dims > required_
+0001a9c0: 6e64 696d 733a 0a20 2020 2020 2020 2077  ndims:.        w
+0001a9d0: 6869 6c65 206e 6469 6d73 203e 2072 6571  hile ndims > req
+0001a9e0: 7569 7265 645f 6e64 696d 733a 0a20 2020  uired_ndims:.   
+0001a9f0: 2020 2020 2020 2020 2069 6620 7465 6e73           if tens
+0001aa00: 6f72 2e73 6861 7065 5b30 5d20 3d3d 2031  or.shape[0] == 1
+0001aa10: 3a0a 2020 2020 2020 2020 2020 2020 2020  :.              
+0001aa20: 2020 7465 6e73 6f72 203d 2074 656e 736f    tensor = tenso
+0001aa30: 722e 7371 7565 657a 6528 3029 0a20 2020  r.squeeze(0).   
+0001aa40: 2020 2020 2020 2020 2020 2020 2073 697a               siz
+0001aa50: 6520 3d20 746f 7263 682e 7465 6e73 6f72  e = torch.tensor
+0001aa60: 2874 656e 736f 722e 7368 6170 6529 0a20  (tensor.shape). 
+0001aa70: 2020 2020 2020 2020 2020 2020 2020 206e                 n
+0001aa80: 6469 6d73 203d 206c 656e 2873 697a 6529  dims = len(size)
+0001aa90: 0a20 2020 2020 2020 2020 2020 2065 6c69  .            eli
+0001aaa0: 6620 5472 7565 2069 6e20 6964 785f 6f6e  f True in idx_on
+0001aab0: 6573 3a0a 2020 2020 2020 2020 2020 2020  es:.            
+0001aac0: 2020 2020 7465 6e73 6f72 203d 2074 656e      tensor = ten
+0001aad0: 736f 722e 7371 7565 657a 6528 6f6e 6573  sor.squeeze(ones
+0001aae0: 5f70 6f73 290a 2020 2020 2020 2020 2020  _pos).          
+0001aaf0: 2020 2020 2020 7369 7a65 203d 2074 6f72        size = tor
+0001ab00: 6368 2e74 656e 736f 7228 7465 6e73 6f72  ch.tensor(tensor
+0001ab10: 2e73 6861 7065 290a 2020 2020 2020 2020  .shape).        
+0001ab20: 2020 2020 2020 2020 6e64 696d 7320 3d20          ndims = 
+0001ab30: 6c65 6e28 7369 7a65 290a 2020 2020 2020  len(size).      
+0001ab40: 2020 2020 2020 656c 7365 3a0a 2020 2020        else:.    
+0001ab50: 2020 2020 2020 2020 2020 2020 6e64 696d              ndim
+0001ab60: 7320 3d20 7265 7175 6972 6564 5f6e 6469  s = required_ndi
+0001ab70: 6d73 0a0a 2020 2020 2020 2020 7265 7475  ms..        retu
+0001ab80: 726e 2074 656e 736f 720a 2020 2020 656c  rn tensor.    el
+0001ab90: 7365 3a0a 2020 2020 2020 2020 7265 7475  se:.        retu
+0001aba0: 726e 2074 656e 736f 720a 0a0a 0a0a 0a    rn tensor......
```

### Comparing `draupnir-0.0.28/src/draupnir.egg-info/PKG-INFO` & `draupnir-0.0.29/src/draupnir.egg-info/PKG-INFO`

 * *Files 13% similar despite different names*

```diff
@@ -1,23 +1,39 @@
 Metadata-Version: 2.1
 Name: draupnir
-Version: 0.0.28
+Version: 0.0.29
 Summary: Ancestral sequence reconstruction using a tree structured Ornstein Uhlenbeck variational autoencoder
 Home-page: https://github.com/LysSanzMoreta/DRAUPNIR_ASR
 Author: Lys Sanz Moreta
 Author-email: lys.sanz.moreta@outlook.com
 Project-URL: Changelog, https://github.com/LysSanzMoreta/DRAUPNIR_ASR/blob/master/CHANGELOG.rst
 Project-URL: Issue Tracker, https://github.com/LysSanzMoreta/DRAUPNIR_ASR/issues
 Classifier: Development Status :: 4 - Beta
 Classifier: License :: OSI Approved :: MIT License
 Classifier: Programming Language :: Python :: 3.8
 Classifier: Programming Language :: Python :: 3 :: Only
 Requires-Python: >=3.7, <4
 Description-Content-Type: text/markdown
 License-File: LICENSE
+Requires-Dist: pyro-ppl>1.6.0
+Requires-Dist: biopython>1.78
+Requires-Dist: pandas>1.0.1
+Requires-Dist: matplotlib>3.3.4
+Requires-Dist: ete3>3.1.1
+Requires-Dist: dgl>0.6.1
+Requires-Dist: dill>0.3.3
+Requires-Dist: seaborn>=0.11.2
+Requires-Dist: pytorch-ignite>0.4.4
+Requires-Dist: scipy>1.5.4
+Requires-Dist: scikit-learn>0.24.1
+Requires-Dist: umap-learn>0.5.2
+Requires-Dist: gdown>=4.3.1
+Requires-Dist: ProDy>2.0.0
+Requires-Dist: PyQt5>=5.15.7
+Requires-Dist: lxml>=4.9.1
 
 
 DRAUPNIR: "Beta library version for performing ASR using a tree-structured Variational Autoencoder"
 
 <p align="center">
 <img src="https://github.com/LysSanzMoreta/DRAUPNIR_ASR/blob/main/draupnir/src/draupnir/images/draupnir_logo.png" height="auto" width="790" style="border-radius:50%">
 </p>
@@ -150,17 +166,17 @@
 
 **Where are my ancestral sequences?**
 
 - In each of the folders there should be a fasta file <dataset-name>_sampled_nodes_seq.fasta
 
 - Each of the sequences in the file should be identified as <node-name-input-tree>//<tree-level-order>\_sample\_<sample-number>
 
-    -Node-name-input-tree: Original name of the node in the given input tree
+    -<node-name-input-tree>: Original name of the node in the given/user's input tree.
 
-    -Tree-level-order: Position of the node in tree-level order in the tree
+    -<tree-level-order>: Position of the node in the tree expressed in tree-level order.
     
     >Node_A1//1.0_sample_0
 
     
 **If this library is useful for your research please cite:**
 
 ```
```

### Comparing `draupnir-0.0.28/src/draupnir.egg-info/SOURCES.txt` & `draupnir-0.0.29/src/draupnir.egg-info/SOURCES.txt`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/tests/test_accuracy.py` & `draupnir-0.0.29/tests/test_accuracy.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/tests/test_examples.py` & `draupnir-0.0.29/tests/test_examples.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/tests/test_load.py` & `draupnir-0.0.29/tests/test_load.py`

 * *Files identical despite different names*

### Comparing `draupnir-0.0.28/tests/test_norworking.py` & `draupnir-0.0.29/tests/test_norworking.py`

 * *Files identical despite different names*

